diff --git a/tests/queries/0_stateless/01460_allow_dollar_and_number_in_identifier.reference b/tests/queries/0_stateless/01460_allow_dollar_and_number_in_identifier.reference
new file mode 100644
index 000000000000..f5b4a27081dd
--- /dev/null
+++ b/tests/queries/0_stateless/01460_allow_dollar_and_number_in_identifier.reference
@@ -0,0 +1,4 @@
+$alias$name$
+1
+1alias1name1
+1
diff --git a/tests/queries/0_stateless/01460_allow_dollar_and_number_in_identifier.sql b/tests/queries/0_stateless/01460_allow_dollar_and_number_in_identifier.sql
new file mode 100644
index 000000000000..5a3aa9968990
--- /dev/null
+++ b/tests/queries/0_stateless/01460_allow_dollar_and_number_in_identifier.sql
@@ -0,0 +1,2 @@
+SELECT 1 AS $alias$name$ FORMAT TSVWithNames;
+SELECT 1 AS 1alias1name1 FORMAT TSVWithNames;
