diff --git a/.gitmodules b/.gitmodules
index 32ce4983bb01..fefaa0eb5f88 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -10,7 +10,7 @@
 	url = https://github.com/lz4/lz4.git
 [submodule "contrib/librdkafka"]
 	path = contrib/librdkafka
-	url = https://github.com/edenhill/librdkafka.git
+	url = https://github.com/ClickHouse-Extras/librdkafka.git
 [submodule "contrib/cctz"]
 	path = contrib/cctz
 	url = https://github.com/ClickHouse-Extras/cctz.git
diff --git a/contrib/librdkafka b/contrib/librdkafka
index b0d91bd74abb..2090cbf56b71 160000
--- a/contrib/librdkafka
+++ b/contrib/librdkafka
@@ -1,1 +1,1 @@
-Subproject commit b0d91bd74abb5f0e1ee972d326a317ad610f6300
+Subproject commit 2090cbf56b715247ec2be7f768707a7ab1bf7ede
diff --git a/src/Storages/Kafka/ReadBufferFromKafkaConsumer.cpp b/src/Storages/Kafka/ReadBufferFromKafkaConsumer.cpp
index bbd93bc5aad0..2498b7196483 100644
--- a/src/Storages/Kafka/ReadBufferFromKafkaConsumer.cpp
+++ b/src/Storages/Kafka/ReadBufferFromKafkaConsumer.cpp
@@ -384,7 +384,8 @@ bool ReadBufferFromKafkaConsumer::poll()
         {
             messages = std::move(new_messages);
             current = messages.begin();
-            LOG_TRACE(log, "Polled batch of {} messages. Offset position: {}", messages.size(), consumer->get_offsets_position(consumer->get_assignment()));
+            LOG_TRACE(log, "Polled batch of {} messages. Offsets position: {}",
+                messages.size(), consumer->get_offsets_position(consumer->get_assignment()));
             break;
         }
     }
@@ -416,7 +417,7 @@ size_t ReadBufferFromKafkaConsumer::filterMessageErrors()
         return false;
     });
 
-    size_t skipped = std::distance(messages.end(), new_end);
+    size_t skipped = std::distance(new_end, messages.end());
     if (skipped)
     {
         LOG_ERROR(log, "There were {} messages with an error", skipped);
