diff --git a/tests/queries/0_stateless/02234_cast_to_ip_address.reference b/tests/queries/0_stateless/02234_cast_to_ip_address.reference
index 9023b36a9bfc..fa9c6bd0f943 100644
--- a/tests/queries/0_stateless/02234_cast_to_ip_address.reference
+++ b/tests/queries/0_stateless/02234_cast_to_ip_address.reference
@@ -11,6 +11,10 @@ IPv4 functions
 127.0.0.1
 127.0.0.1
 --
+1.2.3.4
+1.2.3.4
+0.0.0.0
+--
 127.0.0.1
 --
 0
diff --git a/tests/queries/0_stateless/02234_cast_to_ip_address.sql b/tests/queries/0_stateless/02234_cast_to_ip_address.sql
index 6c65fe86cc91..28f1afff57f5 100644
--- a/tests/queries/0_stateless/02234_cast_to_ip_address.sql
+++ b/tests/queries/0_stateless/02234_cast_to_ip_address.sql
@@ -20,6 +20,13 @@ SELECT toIPv4OrNull('127.0.0.1');
 
 SELECT '--';
 
+SELECT toIPv4(toIPv6('::ffff:1.2.3.4'));
+SELECT toIPv4(toIPv6('::afff:1.2.3.4')); --{serverError CANNOT_CONVERT_TYPE}
+SELECT toIPv4OrDefault(toIPv6('::ffff:1.2.3.4'));
+SELECT toIPv4OrDefault(toIPv6('::afff:1.2.3.4'));
+
+SELECT '--';
+
 SELECT cast('test' , 'IPv4'); --{serverError CANNOT_PARSE_IPV4}
 SELECT cast('127.0.0.1' , 'IPv4');
 
