diff --git a/tests/integration/test_postgresql_database_engine/test.py b/tests/integration/test_postgresql_database_engine/test.py
index 63e85afb1d41..d9f06f0295b5 100644
--- a/tests/integration/test_postgresql_database_engine/test.py
+++ b/tests/integration/test_postgresql_database_engine/test.py
@@ -330,7 +330,7 @@ def test_predefined_connection_configuration(started_cluster):
     node1.query(
         """
         DROP DATABASE postgres_database;
-        CREATE DATABASE postgres_database ENGINE = PostgreSQL(postgres1, use_tables_cache=1);
+        CREATE DATABASE postgres_database ENGINE = PostgreSQL(postgres1, use_table_cache=1);
         """
     )
     assert (
