diff --git a/tests/queries/0_stateless/02463_julian_day_ubsan.reference b/tests/queries/0_stateless/02463_julian_day_ubsan.reference
new file mode 100644
index 000000000000..e69de29bb2d1
diff --git a/tests/queries/0_stateless/02463_julian_day_ubsan.sql b/tests/queries/0_stateless/02463_julian_day_ubsan.sql
new file mode 100644
index 000000000000..a8583d7b0a82
--- /dev/null
+++ b/tests/queries/0_stateless/02463_julian_day_ubsan.sql
@@ -0,0 +1,1 @@
+SELECT fromModifiedJulianDay(9223372036854775807 :: Int64); -- { serverError 490 }
