diff --git a/extension/parquet/parquet-extension.cpp b/extension/parquet/parquet-extension.cpp
index f038346e165e..b2ae084fc347 100644
--- a/extension/parquet/parquet-extension.cpp
+++ b/extension/parquet/parquet-extension.cpp
@@ -471,7 +471,7 @@ unique_ptr<LocalFunctionData> ParquetWriteInitializeLocal(ClientContext &context
 }
 
 unique_ptr<TableFunctionRef> ParquetScanReplacement(const string &table_name, void *data) {
-	if (!StringUtil::EndsWith(table_name, ".parquet")) {
+	if (!StringUtil::EndsWith(StringUtil::Lower(table_name), ".parquet")) {
 		return nullptr;
 	}
 	auto table_function = make_unique<TableFunctionRef>();
diff --git a/src/function/table/read_csv.cpp b/src/function/table/read_csv.cpp
index 303f89aa1b35..54b1e7b10ca0 100644
--- a/src/function/table/read_csv.cpp
+++ b/src/function/table/read_csv.cpp
@@ -255,8 +255,9 @@ void ReadCSVTableFunction::RegisterFunction(BuiltinFunctions &set) {
 }
 
 unique_ptr<TableFunctionRef> ReadCSVReplacement(const string &table_name, void *data) {
-	if (!StringUtil::EndsWith(table_name, ".csv") && !StringUtil::EndsWith(table_name, ".tsv") &&
-	    !StringUtil::EndsWith(table_name, ".csv.gz")) {
+	auto lower_name = StringUtil::Lower(table_name);
+	if (!StringUtil::EndsWith(lower_name, ".csv") && !StringUtil::EndsWith(lower_name, ".tsv") &&
+	    !StringUtil::EndsWith(lower_name, ".csv.gz")) {
 		return nullptr;
 	}
 	auto table_function = make_unique<TableFunctionRef>();
