diff --git a/data/parquet-testing/CASE_INSENSITIVE.PARQUET b/data/parquet-testing/CASE_INSENSITIVE.PARQUET
new file mode 100644
index 000000000000..b533c0fdcfbd
Binary files /dev/null and b/data/parquet-testing/CASE_INSENSITIVE.PARQUET differ
diff --git a/test/sql/copy/parquet/case_insensitive_replacement.test b/test/sql/copy/parquet/case_insensitive_replacement.test
new file mode 100644
index 000000000000..2ddf80318032
--- /dev/null
+++ b/test/sql/copy/parquet/case_insensitive_replacement.test
@@ -0,0 +1,10 @@
+# name: test/sql/copy/parquet/case_insensitive_replacement.test
+# description: Issue #2543: Case insensitive replacement scan
+# group: [parquet]
+
+require parquet
+
+query I
+SELECT data FROM 'data/parquet-testing/CASE_INSENSITIVE.PARQUET'
+----
+\x00\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0A\x0B\x0C\x0D\x0E\x0F
