diff --git a/src/planner/binder/query_node/plan_recursive_cte_node.cpp b/src/planner/binder/query_node/plan_recursive_cte_node.cpp
index 81257f6a64c7..1d67f10b60ab 100644
--- a/src/planner/binder/query_node/plan_recursive_cte_node.cpp
+++ b/src/planner/binder/query_node/plan_recursive_cte_node.cpp
@@ -24,7 +24,8 @@ unique_ptr<LogicalOperator> Binder::CreatePlan(BoundRecursiveCTENode &node) {
 	left_node = CastLogicalOperatorToTypes(node.left->types, node.types, move(left_node));
 	right_node = CastLogicalOperatorToTypes(node.right->types, node.types, move(right_node));
 
-	if (node.right_binder->bind_context.cte_references[node.ctename] == nullptr) {
+	if (!node.right_binder->bind_context.cte_references[node.ctename] ||
+	    *node.right_binder->bind_context.cte_references[node.ctename] == 0) {
 		auto root = make_unique<LogicalSetOperation>(node.setop_index, node.types.size(), move(left_node),
 		                                             move(right_node), LogicalOperatorType::LOGICAL_UNION);
 		return VisitQueryNode(node, move(root));
