diff --git a/src/TableFunctions/TableFunctionExecutable.cpp b/src/TableFunctions/TableFunctionExecutable.cpp
index 9fa5ddf15c23..2c3802e86676 100644
--- a/src/TableFunctions/TableFunctionExecutable.cpp
+++ b/src/TableFunctions/TableFunctionExecutable.cpp
@@ -71,6 +71,9 @@ std::vector<size_t> TableFunctionExecutable::skipAnalysisForArguments(const Quer
     const auto & table_function_node_arguments = table_function_node.getArguments().getNodes();
     size_t table_function_node_arguments_size = table_function_node_arguments.size();
 
+    if (table_function_node_arguments_size <= 2)
+        return {};
+
     std::vector<size_t> result_indexes;
     result_indexes.reserve(table_function_node_arguments_size - 2);
     for (size_t i = 2; i < table_function_node_arguments_size; ++i)
