diff --git a/src/core_functions/scalar/list/list_cosine_similarity.cpp b/src/core_functions/scalar/list/list_cosine_similarity.cpp
index 72607752661c..97fcc5c033d6 100644
--- a/src/core_functions/scalar/list/list_cosine_similarity.cpp
+++ b/src/core_functions/scalar/list/list_cosine_similarity.cpp
@@ -17,6 +17,9 @@ static void ListCosineSimilarity(DataChunk &args, ExpressionState &, Vector &res
 	auto &left_child = ListVector::GetEntry(left);
 	auto &right_child = ListVector::GetEntry(right);
 
+	left_child.Flatten(left_count);
+	right_child.Flatten(right_count);
+
 	D_ASSERT(left_child.GetVectorType() == VectorType::FLAT_VECTOR);
 	D_ASSERT(right_child.GetVectorType() == VectorType::FLAT_VECTOR);
 
@@ -41,6 +44,10 @@ static void ListCosineSimilarity(DataChunk &args, ExpressionState &, Vector &res
 
 		    auto dimensions = left.length;
 
+		    if (dimensions == 0) {
+			    throw InvalidInputException("The cosine similarity for empty vectors is not defined");
+		    }
+
 		    NUMERIC_TYPE distance = 0;
 		    NUMERIC_TYPE norm_l = 0;
 		    NUMERIC_TYPE norm_r = 0;
diff --git a/src/core_functions/scalar/list/list_distance.cpp b/src/core_functions/scalar/list/list_distance.cpp
index aa70e4a1309d..476235ce096c 100644
--- a/src/core_functions/scalar/list/list_distance.cpp
+++ b/src/core_functions/scalar/list/list_distance.cpp
@@ -16,6 +16,9 @@ static void ListDistance(DataChunk &args, ExpressionState &, Vector &result) {
 	auto &left_child = ListVector::GetEntry(left);
 	auto &right_child = ListVector::GetEntry(right);
 
+	left_child.Flatten(left_count);
+	right_child.Flatten(right_count);
+
 	D_ASSERT(left_child.GetVectorType() == VectorType::FLAT_VECTOR);
 	D_ASSERT(right_child.GetVectorType() == VectorType::FLAT_VECTOR);
 
