diff --git a/tests/queries/0_stateless/01643_system_suspend.reference b/tests/queries/0_stateless/01643_system_suspend.reference
new file mode 100644
index 000000000000..d00491fd7e5b
--- /dev/null
+++ b/tests/queries/0_stateless/01643_system_suspend.reference
@@ -0,0 +1,1 @@
+1
diff --git a/tests/queries/0_stateless/01643_system_suspend.sql b/tests/queries/0_stateless/01643_system_suspend.sql
new file mode 100644
index 000000000000..c2cd37e6156b
--- /dev/null
+++ b/tests/queries/0_stateless/01643_system_suspend.sql
@@ -0,0 +1,5 @@
+CREATE TEMPORARY TABLE t (x DateTime);
+INSERT INTO t VALUES (now());
+SYSTEM SUSPEND FOR 1 SECOND;
+INSERT INTO t VALUES (now());
+SELECT max(x) - min(x) >= 1 FROM t;
