diff --git a/tests/queries/0_stateless/02886_binary_like.reference b/tests/queries/0_stateless/02886_binary_like.reference
new file mode 100644
index 000000000000..c23ceaf97849
--- /dev/null
+++ b/tests/queries/0_stateless/02886_binary_like.reference
@@ -0,0 +1,24 @@
+1
+1
+1
+1
+1
+0
+0
+1
+1
+1
+1
+1
+1
+0
+0
+1
+1
+1
+1
+1
+1
+0
+0
+1
diff --git a/tests/queries/0_stateless/02886_binary_like.sql b/tests/queries/0_stateless/02886_binary_like.sql
new file mode 100644
index 000000000000..ba11f1fc0604
--- /dev/null
+++ b/tests/queries/0_stateless/02886_binary_like.sql
@@ -0,0 +1,26 @@
+SELECT 'aяb' LIKE 'a_b';
+SELECT 'a\0b' LIKE 'a_b';
+SELECT 'a\0b' LIKE 'a\0b';
+SELECT 'a\0b' LIKE 'a%\0b';
+SELECT 'a\xFFb' LIKE 'a%\xFFb';
+SELECT 'a\xFFb' LIKE 'a%\xFF\xFEb';
+SELECT 'a\xFFb' LIKE '%a\xFF\xFEb';
+SELECT 'a\xFF\xFEb' LIKE '%a\xFF\xFEb';
+
+SELECT materialize('aяb') LIKE 'a_b';
+SELECT materialize('a\0b') LIKE 'a_b';
+SELECT materialize('a\0b') LIKE 'a\0b';
+SELECT materialize('a\0b') LIKE 'a%\0b';
+SELECT materialize('a\xFFb') LIKE 'a%\xFFb';
+SELECT materialize('a\xFFb') LIKE 'a%\xFF\xFEb';
+SELECT materialize('a\xFFb') LIKE '%a\xFF\xFEb';
+SELECT materialize('a\xFF\xFEb') LIKE '%a\xFF\xFEb';
+
+SELECT materialize('aяb') LIKE materialize('a_b');
+SELECT materialize('a\0b') LIKE materialize('a_b');
+SELECT materialize('a\0b') LIKE materialize('a\0b');
+SELECT materialize('a\0b') LIKE materialize('a%\0b');
+SELECT materialize('a\xFFb') LIKE materialize('a%\xFFb');
+SELECT materialize('a\xFFb') LIKE materialize('a%\xFF\xFEb');
+SELECT materialize('a\xFFb') LIKE materialize('%a\xFF\xFEb');
+SELECT materialize('a\xFF\xFEb') LIKE materialize('%a\xFF\xFEb');
