diff --git a/dbms/tests/queries/0_stateless/00976_shard_low_cardinality_achimbab.reference b/dbms/tests/queries/0_stateless/00976_shard_low_cardinality_achimbab.reference
new file mode 100644
index 000000000000..9972842f9827
--- /dev/null
+++ b/dbms/tests/queries/0_stateless/00976_shard_low_cardinality_achimbab.reference
@@ -0,0 +1,1 @@
+1	1
diff --git a/dbms/tests/queries/0_stateless/00976_shard_low_cardinality_achimbab.sql b/dbms/tests/queries/0_stateless/00976_shard_low_cardinality_achimbab.sql
new file mode 100644
index 000000000000..3736be891cdf
--- /dev/null
+++ b/dbms/tests/queries/0_stateless/00976_shard_low_cardinality_achimbab.sql
@@ -0,0 +1,6 @@
+DROP TABLE IF EXISTS han_1;
+CREATE TABLE han_1 (k Int32, date_dt LowCardinality(Nullable(String))) 
+ENGINE = MergeTree() PARTITION BY k ORDER BY k;
+INSERT INTO han_1 values (1, '2019-07-31');
+SELECT k, uniq(date_dt) FROM remote('127.0.0.{1,2}', currentDatabase(), han_1) GROUP BY k;
+DROP TABLE IF EXISTS han_1;
