diff --git a/data/parquet-testing/hive-partitioning/ci-column-names/COL1=Hannes/data_0.parquet b/data/parquet-testing/hive-partitioning/ci-column-names/COL1=Hannes/data_0.parquet
new file mode 100644
index 000000000000..c70efafac6d7
Binary files /dev/null and b/data/parquet-testing/hive-partitioning/ci-column-names/COL1=Hannes/data_0.parquet differ
diff --git a/data/parquet-testing/hive-partitioning/ci-column-names/COL1=Mark/data_0.parquet b/data/parquet-testing/hive-partitioning/ci-column-names/COL1=Mark/data_0.parquet
new file mode 100644
index 000000000000..f7f4c00d7b22
Binary files /dev/null and b/data/parquet-testing/hive-partitioning/ci-column-names/COL1=Mark/data_0.parquet differ
diff --git a/test/sql/copy/partitioned/hive_partition_case_insensitive_column.test b/test/sql/copy/partitioned/hive_partition_case_insensitive_column.test
new file mode 100644
index 000000000000..678e4ec78f6a
--- /dev/null
+++ b/test/sql/copy/partitioned/hive_partition_case_insensitive_column.test
@@ -0,0 +1,11 @@
+# name: test/sql/copy/partitioned/hive_partition_case_insensitive_column.test
+# description: Test when columns in the hive partitioned files differ only in case from the partitions themselves
+# group: [partitioned]
+
+require parquet
+
+query II
+SELECT * FROM  'data/parquet-testing/hive-partitioning/ci-column-names/**/*.parquet' ORDER BY ALL
+----
+Hannes	2
+Mark	1
