diff --git a/tests/queries/0_stateless/01560_mann_whitney.reference b/tests/queries/0_stateless/01560_mann_whitney.reference
index 5a718fdca977..dcd1882b0964 100644
--- a/tests/queries/0_stateless/01560_mann_whitney.reference
+++ b/tests/queries/0_stateless/01560_mann_whitney.reference
@@ -1,5 +1,5 @@
-(223,0.5426959774289524)
+(223,0.5426959774289482)
 223.0	0.5426959774289482
-223	0.5426959774289524
-223	0.5426959774289524
-223	0.5426959774289524
+223	0.5426959774289482
+223	0.5426959774289482
+223	0.5426959774289482
diff --git a/tests/queries/0_stateless/02293_ttest_large_samples.reference b/tests/queries/0_stateless/02293_ttest_large_samples.reference
new file mode 100644
index 000000000000..2e90a4a98b23
--- /dev/null
+++ b/tests/queries/0_stateless/02293_ttest_large_samples.reference
@@ -0,0 +1,2 @@
+-0.12709	0.89887
+-1.27088	0.20377
diff --git a/tests/queries/0_stateless/02293_ttest_large_samples.sql b/tests/queries/0_stateless/02293_ttest_large_samples.sql
new file mode 100644
index 000000000000..d4d919d6e7ce
--- /dev/null
+++ b/tests/queries/0_stateless/02293_ttest_large_samples.sql
@@ -0,0 +1,53 @@
+SELECT roundBankers(result.1, 5), roundBankers(result.2, 5) FROM (
+SELECT
+     studentTTest(sample, variant) as result
+FROM (
+SELECT
+    toFloat64(number) % 30 AS sample,
+    0 AS variant
+FROM system.numbers limit 500000
+
+UNION ALL
+
+SELECT
+    toFloat64(number) % 30 + 0.0022 AS sample,
+    1 AS variant
+FROM system.numbers limit 500000));
+
+
+SELECT roundBankers(result.1, 5), roundBankers(result.2, 5 ) FROM (
+SELECT
+     studentTTest(sample, variant) as result
+FROM (
+SELECT
+    toFloat64(number) % 30 AS sample,
+    0 AS variant
+FROM system.numbers limit 50000000
+
+UNION ALL
+
+SELECT
+    toFloat64(number) % 30 + 0.0022 AS sample,
+    1 AS variant
+FROM system.numbers limit 50000000));
+
+
+SELECT roundBankers(result.2, 1025)
+FROM
+(
+    SELECT studentTTest(sample, variant) AS result
+    FROM
+    (
+        SELECT
+            toFloat64(number) % 30 AS sample,
+            1048576 AS variant
+        FROM system.numbers
+        LIMIT 1
+        UNION ALL
+        SELECT
+            (toFloat64(number) % 7) + inf AS sample,
+            255 AS variant
+        FROM system.numbers
+        LIMIT 1023
+    )
+); -- { serverError 36 }
