diff --git a/tests/config/test_function.xml b/tests/config/test_function.xml
index 928cbd75c785..a50ab69422ac 100644
--- a/tests/config/test_function.xml
+++ b/tests/config/test_function.xml
@@ -13,4 +13,18 @@
         <command>cd /; clickhouse-local --input-format TabSeparated --output-format TabSeparated --structure 'x UInt64, y UInt64' --query "SELECT x + y FROM table"</command>
         <execute_direct>0</execute_direct>
     </function>
+    <function>
+        <type>executable</type>
+        <name>test_function_with_parameter</name>
+        <return_type>UInt64</return_type>
+        <argument>
+            <type>UInt64</type>
+        </argument>
+        <argument>
+            <type>UInt64</type>
+        </argument>
+        <format>TabSeparated</format>
+        <command>cd /; clickhouse-local --input-format TabSeparated --output-format TabSeparated --structure 'x UInt64, y UInt64' --query "SELECT x + y + {test_parameter : UInt64} FROM table"</command>
+        <execute_direct>0</execute_direct>
+    </function>
 </functions>
diff --git a/tests/integration/test_executable_user_defined_function/functions/test_function_config.xml b/tests/integration/test_executable_user_defined_function/functions/test_function_config.xml
index 5da2e854da80..d35cb173dd06 100644
--- a/tests/integration/test_executable_user_defined_function/functions/test_function_config.xml
+++ b/tests/integration/test_executable_user_defined_function/functions/test_function_config.xml
@@ -311,4 +311,15 @@
         <command>input_nullable.py</command>
     </function>
 
+    <function>
+        <type>executable</type>
+        <name>test_function_parameter_python</name>
+        <return_type>String</return_type>
+        <argument>
+            <type>UInt64</type>
+        </argument>
+        <format>TabSeparated</format>
+        <command>input_parameter.py {test_parameter:UInt64}</command>
+    </function>
+
 </functions>
diff --git a/tests/integration/test_executable_user_defined_function/test.py b/tests/integration/test_executable_user_defined_function/test.py
index f48547a14377..1f4e14470c37 100644
--- a/tests/integration/test_executable_user_defined_function/test.py
+++ b/tests/integration/test_executable_user_defined_function/test.py
@@ -263,3 +263,25 @@ def test_executable_function_input_nullable_python(started_cluster):
         )
         == "Key 0
Key Nullable
Key 2
"
     )
+
+
+def test_executable_function_parameter_python(started_cluster):
+    skip_test_msan(node)
+
+    assert node.query_and_get_error(
+        "SELECT test_function_parameter_python(2,2)(toUInt64(1))"
+    )
+    assert node.query_and_get_error("SELECT test_function_parameter_python(2,2)(1)")
+    assert node.query_and_get_error("SELECT test_function_parameter_python(1)")
+    assert node.query_and_get_error(
+        "SELECT test_function_parameter_python('test')(toUInt64(1))"
+    )
+
+    assert (
+        node.query("SELECT test_function_parameter_python('2')(toUInt64(1))")
+        == "Parameter 2 key 1
"
+    )
+    assert (
+        node.query("SELECT test_function_parameter_python(2)(toUInt64(1))")
+        == "Parameter 2 key 1
"
+    )
diff --git a/tests/integration/test_executable_user_defined_function/user_scripts/input_parameter.py b/tests/integration/test_executable_user_defined_function/user_scripts/input_parameter.py
new file mode 100755
index 000000000000..c9d6ff9f9d58
--- /dev/null
+++ b/tests/integration/test_executable_user_defined_function/user_scripts/input_parameter.py
@@ -0,0 +1,8 @@
+#!/usr/bin/python3
+
+import sys
+
+if __name__ == "__main__":
+    for line in sys.stdin:
+        print("Parameter " + str(sys.argv[1]) + " key " + str(line), end="")
+        sys.stdout.flush()
diff --git a/tests/queries/0_stateless/02315_executable_user_defined_function_parameter.reference b/tests/queries/0_stateless/02315_executable_user_defined_function_parameter.reference
new file mode 100644
index 000000000000..fd3c81a4d763
--- /dev/null
+++ b/tests/queries/0_stateless/02315_executable_user_defined_function_parameter.reference
@@ -0,0 +1,2 @@
+5
+5
diff --git a/tests/queries/0_stateless/02315_executable_user_defined_function_parameter.sql b/tests/queries/0_stateless/02315_executable_user_defined_function_parameter.sql
new file mode 100644
index 000000000000..f6e5678e6120
--- /dev/null
+++ b/tests/queries/0_stateless/02315_executable_user_defined_function_parameter.sql
@@ -0,0 +1,6 @@
+SELECT test_function_with_parameter('test')(1, 2); --{serverError 53}
+SELECT test_function_with_parameter(2, 2)(1, 2); --{serverError 36}
+SELECT test_function_with_parameter(1, 2); --{serverError 36}
+
+SELECT test_function_with_parameter(2)(1, 2);
+SELECT test_function_with_parameter('2')(1, 2);
