diff --git a/tests/queries/0_stateless/02883_read_in_reverse_order_virtual_column.reference b/tests/queries/0_stateless/02883_read_in_reverse_order_virtual_column.reference
new file mode 100644
index 000000000000..f77195f1f318
--- /dev/null
+++ b/tests/queries/0_stateless/02883_read_in_reverse_order_virtual_column.reference
@@ -0,0 +1,1 @@
+198401_1_1_0
diff --git a/tests/queries/0_stateless/02883_read_in_reverse_order_virtual_column.sql b/tests/queries/0_stateless/02883_read_in_reverse_order_virtual_column.sql
new file mode 100644
index 000000000000..76821c8797d4
--- /dev/null
+++ b/tests/queries/0_stateless/02883_read_in_reverse_order_virtual_column.sql
@@ -0,0 +1,10 @@
+DROP TABLE IF EXISTS t_reverse_order_virt_col;
+
+CREATE TABLE t_reverse_order_virt_col (`order_0` Decimal(76, 53), `p_time` Date)
+ENGINE = MergeTree PARTITION BY toYYYYMM(p_time)
+ORDER BY order_0;
+
+INSERT INTO t_reverse_order_virt_col SELECT number, '1984-01-01' FROM numbers(1000000);
+SELECT DISTINCT _part FROM (SELECT _part FROM t_reverse_order_virt_col ORDER BY order_0 DESC);
+
+DROP TABLE IF EXISTS t_reverse_order_virt_col;
