diff --git a/tests/queries/0_stateless/02723_param_exception_message_context.reference b/tests/queries/0_stateless/02723_param_exception_message_context.reference
new file mode 100644
index 000000000000..286f53421b59
--- /dev/null
+++ b/tests/queries/0_stateless/02723_param_exception_message_context.reference
@@ -0,0 +1,1 @@
+for query parameter 'x'
diff --git a/tests/queries/0_stateless/02723_param_exception_message_context.sh b/tests/queries/0_stateless/02723_param_exception_message_context.sh
new file mode 100755
index 000000000000..d37badd444fb
--- /dev/null
+++ b/tests/queries/0_stateless/02723_param_exception_message_context.sh
@@ -0,0 +1,9 @@
+#!/usr/bin/env bash
+# Tags: no-fasttest
+
+CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+# shellcheck source=../shell_config.sh
+. "$CURDIR"/../shell_config.sh
+
+# The exception message about unparsed parameter also tells about the name of the parameter.
+$CLICKHOUSE_CLIENT --param_x Hello --query "SELECT {x:Array(String)}" 2>&1 | rg -oF "for query parameter 'x'" | uniq
