diff --git a/tests/performance/decimal_parse.xml b/tests/performance/decimal_parse.xml
new file mode 100644
index 000000000000..19e940b13df6
--- /dev/null
+++ b/tests/performance/decimal_parse.xml
@@ -0,0 +1,3 @@
+<test>
+    <query>SELECT count() FROM zeros(10000000) WHERE NOT ignore(toDecimal32OrZero(toString(rand() % 10000), 5))</query>
+</test>
diff --git a/tests/queries/0_stateless/01260_ubsan_decimal_parse.reference b/tests/queries/0_stateless/01260_ubsan_decimal_parse.reference
new file mode 100644
index 000000000000..945da8ffd365
--- /dev/null
+++ b/tests/queries/0_stateless/01260_ubsan_decimal_parse.reference
@@ -0,0 +1,1 @@
+0.000000
diff --git a/tests/queries/0_stateless/01260_ubsan_decimal_parse.sql b/tests/queries/0_stateless/01260_ubsan_decimal_parse.sql
new file mode 100644
index 000000000000..2c7cda512e80
--- /dev/null
+++ b/tests/queries/0_stateless/01260_ubsan_decimal_parse.sql
@@ -0,0 +1,1 @@
+SELECT toDecimal32OrZero(CAST(-7174046, 'String'), 6);
