diff --git a/tests/queries/0_stateless/02096_bad_options_in_client_and_local.reference b/tests/queries/0_stateless/02096_bad_options_in_client_and_local.reference
index c4c0901b9df0..432299e9556c 100644
--- a/tests/queries/0_stateless/02096_bad_options_in_client_and_local.reference
+++ b/tests/queries/0_stateless/02096_bad_options_in_client_and_local.reference
@@ -8,5 +8,3 @@ OK
 OK
 OK
 OK
-OK
-OK
diff --git a/tests/queries/0_stateless/02096_bad_options_in_client_and_local.sh b/tests/queries/0_stateless/02096_bad_options_in_client_and_local.sh
index d37155e85062..753d56fb424b 100755
--- a/tests/queries/0_stateless/02096_bad_options_in_client_and_local.sh
+++ b/tests/queries/0_stateless/02096_bad_options_in_client_and_local.sh
@@ -9,8 +9,6 @@ ${CLICKHOUSE_LOCAL} --unknown-option 2>&1 | grep -F -q "UNRECOGNIZED_ARGUMENTS"
 
 ${CLICKHOUSE_LOCAL} --unknown-option-1 --unknown-option-2 2>&1 | grep -F -q "UNRECOGNIZED_ARGUMENTS" && echo "OK" || echo "FAIL"
 
-${CLICKHOUSE_LOCAL} -- --unknown-option 2>&1 | grep -F -q "BAD_ARGUMENTS" && echo "OK" || echo "FAIL"
-
 ${CLICKHOUSE_LOCAL} -- 'positional-argument' 2>&1 | grep -F -q "BAD_ARGUMENTS" && echo "OK" || echo "FAIL"
 
 ${CLICKHOUSE_LOCAL} -f 2>&1 | grep -F -q "Bad arguments" && echo "OK" || echo "FAIL"
@@ -22,8 +20,6 @@ ${CLICKHOUSE_CLIENT} --unknown-option 2>&1 | grep -F -q "UNRECOGNIZED_ARGUMENTS"
 
 ${CLICKHOUSE_CLIENT} --unknown-option-1 --unknown-option-2 2>&1 | grep -F -q "UNRECOGNIZED_ARGUMENTS" && echo "OK" || echo "FAIL"
 
-${CLICKHOUSE_CLIENT} -- --unknown-option 2>&1 | grep -F -q "BAD_ARGUMENTS" && echo "OK" || echo "FAIL"
-
 ${CLICKHOUSE_CLIENT} -- 'positional-argument' 2>&1 | grep -F -q "BAD_ARGUMENTS" && echo "OK" || echo "FAIL"
 
 ${CLICKHOUSE_CLIENT} --j 2>&1 | grep -F -q "Bad arguments" && echo "OK" || echo "FAIL"
diff --git a/tests/queries/0_stateless/02833_local_udf_options.reference b/tests/queries/0_stateless/02833_local_udf_options.reference
new file mode 100644
index 000000000000..19f0805d8de3
--- /dev/null
+++ b/tests/queries/0_stateless/02833_local_udf_options.reference
@@ -0,0 +1,1 @@
+qwerty
diff --git a/tests/queries/0_stateless/02833_local_udf_options.sh b/tests/queries/0_stateless/02833_local_udf_options.sh
new file mode 100755
index 000000000000..149b62d7e2cc
--- /dev/null
+++ b/tests/queries/0_stateless/02833_local_udf_options.sh
@@ -0,0 +1,11 @@
+#!/usr/bin/env bash
+
+set -e
+
+CUR_DIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+# shellcheck source=../shell_config.sh
+. "$CUR_DIR"/../shell_config.sh
+
+SCRIPTS_DIR=$CUR_DIR/scripts_udf
+
+$CLICKHOUSE_LOCAL -q 'select test_function()' -- --user_scripts_path=$SCRIPTS_DIR --user_defined_executable_functions_config=$SCRIPTS_DIR/function.xml
diff --git a/tests/queries/0_stateless/scripts_udf/function.xml b/tests/queries/0_stateless/scripts_udf/function.xml
new file mode 100644
index 000000000000..69a0abb5cec0
--- /dev/null
+++ b/tests/queries/0_stateless/scripts_udf/function.xml
@@ -0,0 +1,9 @@
+<functions>
+    <function>
+        <type>executable</type>
+        <name>test_function</name>
+        <return_type>String</return_type>
+        <format>TabSeparated</format>
+        <command>udf.sh</command>
+    </function>
+</functions>
diff --git a/tests/queries/0_stateless/scripts_udf/udf.sh b/tests/queries/0_stateless/scripts_udf/udf.sh
new file mode 100755
index 000000000000..add85833c3e7
--- /dev/null
+++ b/tests/queries/0_stateless/scripts_udf/udf.sh
@@ -0,0 +1,3 @@
+#!/bin/sh
+
+echo qwerty
