diff --git a/tests/queries/0_stateless/01420_logical_functions_materialized_null.reference b/tests/queries/0_stateless/01420_logical_functions_materialized_null.reference
new file mode 100644
index 000000000000..d82aab4099f4
--- /dev/null
+++ b/tests/queries/0_stateless/01420_logical_functions_materialized_null.reference
@@ -0,0 +1,7 @@
+\N
+1
+\N
+1
+\N
+1
+1
diff --git a/tests/queries/0_stateless/01420_logical_functions_materialized_null.sql b/tests/queries/0_stateless/01420_logical_functions_materialized_null.sql
new file mode 100644
index 000000000000..7c61295ca1f3
--- /dev/null
+++ b/tests/queries/0_stateless/01420_logical_functions_materialized_null.sql
@@ -0,0 +1,8 @@
+SELECT NULL AND 1;
+SELECT NULL OR 1;
+SELECT materialize(NULL) AND 1;
+SELECT materialize(NULL) OR 1;
+SELECT arrayJoin([NULL]) AND 1;
+SELECT arrayJoin([NULL]) OR 1;
+
+SELECT isConstant(arrayJoin([NULL]) AND 1);
