diff --git a/src/AggregateFunctions/AggregateFunctionNothing.h b/src/AggregateFunctions/AggregateFunctionNothing.h
index 8491e8edfaad..8280d72c0854 100644
--- a/src/AggregateFunctions/AggregateFunctionNothing.h
+++ b/src/AggregateFunctions/AggregateFunctionNothing.h
@@ -13,6 +13,11 @@ namespace DB
 {
 struct Settings;
 
+namespace ErrorCodes
+{
+    extern const int INCORRECT_DATA;
+}
+
 
 /** Aggregate function that takes arbitrary number of arbitrary arguments and does nothing.
   */
@@ -69,7 +74,8 @@ class AggregateFunctionNothing final : public IAggregateFunctionHelper<Aggregate
     {
         [[maybe_unused]] char symbol;
         readChar(symbol, buf);
-        assert(symbol == '\0');
+        if (symbol != '\0')
+            throw Exception(ErrorCodes::INCORRECT_DATA, "Incorrect state of aggregate function 'nothing', it should contain exactly one zero byte.");
     }
 
     void insertResultInto(AggregateDataPtr __restrict, IColumn & to, Arena *) const override
