diff --git a/tests/queries/0_stateless/03145_unicode_quotes.reference b/tests/queries/0_stateless/03145_unicode_quotes.reference
new file mode 100644
index 000000000000..8d33edf14e3d
--- /dev/null
+++ b/tests/queries/0_stateless/03145_unicode_quotes.reference
@@ -0,0 +1,5 @@
+This is an example of using English-style Unicode single quotes.
+Row 1:
+──────
+\c\\u\\\r\\\\l\\\\\y\\\\\\: This is \an \\example ‘of using English-style Unicode single quotes.\
+{"1":1}
diff --git a/tests/queries/0_stateless/03145_unicode_quotes.sql b/tests/queries/0_stateless/03145_unicode_quotes.sql
new file mode 100644
index 000000000000..34a465cd874b
--- /dev/null
+++ b/tests/queries/0_stateless/03145_unicode_quotes.sql
@@ -0,0 +1,9 @@
+-- They work:
+SELECT ‘This is an example of using English-style Unicode single quotes.’ AS “curly”;
+
+-- It is unspecified which escaping rules apply inside the literal in Unicode quotes, and currently none apply (similarly to heredocs)
+-- This could be changed.
+
+SELECT ‘This is \an \\example ‘of using English-style Unicode single quotes.\’ AS “\c\\u\\\r\\\\l\\\\\y\\\\\\” FORMAT Vertical;
+
+SELECT ‘’ = '' AS “1” FORMAT JSONLines;
