diff --git a/dbms/tests/queries/0_stateless/00597_push_down_predicate.reference b/dbms/tests/queries/0_stateless/00597_push_down_predicate.reference
index 9fde80689f13..480b1c4525cb 100644
--- a/dbms/tests/queries/0_stateless/00597_push_down_predicate.reference
+++ b/dbms/tests/queries/0_stateless/00597_push_down_predicate.reference
@@ -46,17 +46,19 @@ SELECT 
    id, 
    date, 
    value
FROM 
(
    SELECT 
        id, 

 SELECT 
    date, 
    id, 
    name, 
    value
FROM 
(
    SELECT 
        date, 
        id, 
        name, 
        value
    FROM test_00597
    WHERE id = 1
    UNION ALL
    SELECT 
        date, 
        id, 
        name, 
        value
    FROM test_00597
    WHERE id = 1
)
WHERE id = 1
 2000-01-01	1	test string 1	1
 2000-01-01	1	test string 1	1
-SELECT 
    date, 
    id, 
    name, 
    value, 
    date, 
    name, 
    value
FROM 
(
    SELECT 
        date, 
        id, 
        name, 
        value
    FROM test_00597
    WHERE id = 1
)
ANY LEFT JOIN 
(
    SELECT *
    FROM test_00597
) USING (id)
WHERE id = 1
+SELECT 
    date, 
    id, 
    name, 
    value, 
    date, 
    name, 
    value
FROM 
(
    SELECT 
        date, 
        id, 
        name, 
        value
    FROM test_00597
    WHERE id = 1
)
ANY LEFT JOIN 
(
    SELECT id
    FROM test_00597
) USING (id)
WHERE id = 1
 2000-01-01	1	test string 1	1	2000-01-01	test string 1	1
-SELECT 
    id, 
    date, 
    name, 
    value
FROM 
(
    SELECT toInt8(1) AS id
)
ANY LEFT JOIN 
(
    SELECT *
    FROM test_00597
) AS test_00597 USING (id)
WHERE value = 1
+SELECT 
    id, 
    date, 
    name, 
    value
FROM 
(
    SELECT toInt8(1) AS id
)
ANY LEFT JOIN 
(
    SELECT 
        date, 
        id, 
        name, 
        value
    FROM test_00597
) AS test_00597 USING (id)
WHERE value = 1
 1	2000-01-01	test string 1	1
 SELECT value
FROM 
(
    SELECT toInt8(1) AS id
)
ANY LEFT JOIN test_00597 AS b USING (id)
WHERE value = 1
 1
-SELECT 
    date, 
    id, 
    name, 
    value
FROM 
(
    SELECT 
        date, 
        id, 
        name, 
        value, 
        date, 
        name, 
        value
    FROM 
    (
        SELECT 
            date, 
            id, 
            name, 
            value
        FROM test_00597
        WHERE id = 1
    )
    ANY LEFT JOIN 
    (
        SELECT *
        FROM test_00597
    ) USING (id)
    WHERE id = 1
)
WHERE id = 1
+SELECT 
    date, 
    id, 
    name, 
    value
FROM 
(
    SELECT 
        date, 
        id, 
        name, 
        value, 
        date, 
        name, 
        value
    FROM 
    (
        SELECT 
            date, 
            id, 
            name, 
            value
        FROM test_00597
        WHERE id = 1
    )
    ANY LEFT JOIN 
    (
        SELECT id
        FROM test_00597
    ) USING (id)
    WHERE id = 1
)
WHERE id = 1
 2000-01-01	1	test string 1	1
-SELECT 
    date, 
    id, 
    name, 
    value, 
    b.date, 
    b.name, 
    b.value
FROM 
(
    SELECT 
        date, 
        id, 
        name, 
        value
    FROM test_00597
)
ANY LEFT JOIN 
(
    SELECT *
    FROM test_00597
) AS b USING (id)
WHERE b.id = 1
+SELECT 
    date, 
    id, 
    name, 
    value, 
    b.date, 
    b.name, 
    b.value
FROM 
(
    SELECT 
        date, 
        id, 
        name, 
        value
    FROM test_00597
)
ANY LEFT JOIN 
(
    SELECT 
        date, 
        id, 
        name, 
        value
    FROM test_00597
) AS b USING (id)
WHERE b.id = 1
 2000-01-01	1	test string 1	1	2000-01-01	test string 1	1
-SELECT 
    id, 
    date, 
    name, 
    value
FROM 
(
    SELECT 
        toInt8(1) AS id, 
        toDate(\'2000-01-01\') AS date
    FROM system.numbers
    LIMIT 1
)
ANY LEFT JOIN 
(
    SELECT *
    FROM test_00597
) AS b USING (date, id)
WHERE b.date = toDate(\'2000-01-01\')
+SELECT 
    id, 
    date, 
    name, 
    value
FROM 
(
    SELECT 
        toInt8(1) AS id, 
        toDate(\'2000-01-01\') AS date
    FROM system.numbers
    LIMIT 1
)
ANY LEFT JOIN 
(
    SELECT 
        date, 
        id, 
        name, 
        value
    FROM test_00597
) AS b USING (date, id)
WHERE b.date = toDate(\'2000-01-01\')
 1	2000-01-01	test string 1	1
-SELECT 
    date, 
    id, 
    name, 
    value, 
    `b.date`, 
    `b.id`, 
    `b.name`, 
    `b.value`
FROM 
(
    SELECT 
        date, 
        id, 
        name, 
        value, 
        b.date, 
        b.id, 
        b.name, 
        b.value
    FROM 
    (
        SELECT 
            date, 
            id, 
            name, 
            value
        FROM test_00597
        WHERE id = 1
    ) AS a
    ANY LEFT JOIN 
    (
        SELECT *
        FROM test_00597
    ) AS b ON id = b.id
    WHERE id = 1
)
WHERE id = 1
+SELECT 
    date, 
    id, 
    name, 
    value, 
    `b.date`, 
    `b.id`, 
    `b.name`, 
    `b.value`
FROM 
(
    SELECT 
        date, 
        id, 
        name, 
        value, 
        b.date, 
        b.id, 
        b.name, 
        b.value
    FROM 
    (
        SELECT 
            date, 
            id, 
            name, 
            value
        FROM test_00597
        WHERE id = 1
    ) AS a
    ANY LEFT JOIN 
    (
        SELECT 
            date, 
            id, 
            name, 
            value
        FROM test_00597
    ) AS b ON id = b.id
    WHERE id = 1
)
WHERE id = 1
 2000-01-01	1	test string 1	1	2000-01-01	1	test string 1	1
+SELECT 
    date, 
    id, 
    name, 
    value, 
    r.date, 
    r.name, 
    r.value
FROM 
(
    SELECT 
        date, 
        id, 
        name, 
        value
    FROM test_00597
)
SEMI LEFT JOIN 
(
    SELECT 
        date, 
        id, 
        name, 
        value
    FROM 
    (
        SELECT 
            date, 
            id, 
            name, 
            value
        FROM test_00597
        WHERE id = 1
    )
    WHERE id = 1
) AS r USING (id)
WHERE r.id = 1
+2000-01-01	1	test string 1	1	2000-01-01	test string 1	1
diff --git a/dbms/tests/queries/0_stateless/00597_push_down_predicate.sql b/dbms/tests/queries/0_stateless/00597_push_down_predicate.sql
index 997750831c94..774c855c369e 100644
--- a/dbms/tests/queries/0_stateless/00597_push_down_predicate.sql
+++ b/dbms/tests/queries/0_stateless/00597_push_down_predicate.sql
@@ -130,5 +130,9 @@ SELECT * FROM (SELECT toInt8(1) AS id, toDate('2000-01-01') AS date FROM system.
 ANALYZE SELECT * FROM (SELECT * FROM (SELECT * FROM test_00597) AS a ANY LEFT JOIN (SELECT * FROM test_00597) AS b  ON  a.id = b.id) WHERE id = 1;
 SELECT * FROM (SELECT * FROM (SELECT * FROM test_00597) AS a ANY LEFT JOIN (SELECT * FROM test_00597) AS b  ON  a.id = b.id) WHERE id = 1;
 
+-- Explain with join subquery
+ANALYZE SELECT * FROM (SELECT * FROM test_00597) ANY INNER JOIN (SELECT * FROM (SELECT * FROM test_00597)) as r USING id WHERE r.id = 1;
+SELECT * FROM (SELECT * FROM test_00597) ANY INNER JOIN (SELECT * FROM (SELECT * FROM test_00597)) as r USING id WHERE r.id = 1;
+
 DROP TABLE IF EXISTS test_00597;
 DROP TABLE IF EXISTS test_view_00597;
