diff --git a/tests/queries/0_stateless/01711_decimal_multiplication.reference b/tests/queries/0_stateless/01711_decimal_multiplication.reference
new file mode 100644
index 000000000000..37869329ca44
--- /dev/null
+++ b/tests/queries/0_stateless/01711_decimal_multiplication.reference
@@ -0,0 +1,4 @@
+2.0000
+2.0000
+2.0000
+2.0000
diff --git a/tests/queries/0_stateless/01711_decimal_multiplication.sql b/tests/queries/0_stateless/01711_decimal_multiplication.sql
new file mode 100644
index 000000000000..10d23599b4d9
--- /dev/null
+++ b/tests/queries/0_stateless/01711_decimal_multiplication.sql
@@ -0,0 +1,4 @@
+SELECT materialize(toDecimal64(4,4)) - materialize(toDecimal32(2,2));
+SELECT toDecimal64(4,4) - materialize(toDecimal32(2,2));
+SELECT materialize(toDecimal64(4,4)) - toDecimal32(2,2);
+SELECT toDecimal64(4,4) - toDecimal32(2,2);
