diff --git a/tests/queries/0_stateless/03312_issue_63093.reference b/tests/queries/0_stateless/03312_issue_63093.reference
new file mode 100644
index 000000000000..33844fb6e2f4
--- /dev/null
+++ b/tests/queries/0_stateless/03312_issue_63093.reference
@@ -0,0 +1,2 @@
+param-test OK
+param_test OK
diff --git a/tests/queries/0_stateless/03312_issue_63093.sh b/tests/queries/0_stateless/03312_issue_63093.sh
new file mode 100755
index 000000000000..6149e6b874c6
--- /dev/null
+++ b/tests/queries/0_stateless/03312_issue_63093.sh
@@ -0,0 +1,8 @@
+#!/usr/bin/env bash
+
+CUR_DIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+# shellcheck source=../shell_config.sh
+. "$CUR_DIR"/../shell_config.sh
+
+${CLICKHOUSE_CLIENT} --param-test "param-test OK" -q "SELECT {test:String}"
+${CLICKHOUSE_CLIENT} --param_test "param_test OK" -q "SELECT {test:String}"
