diff --git a/ydb/tests/functional/scheme_tests/canondata/tablet_scheme_tests.TestTabletSchemes.test_tablet_schemes_flat_hive_/flat_hive.schema b/ydb/tests/functional/scheme_tests/canondata/tablet_scheme_tests.TestTabletSchemes.test_tablet_schemes_flat_hive_/flat_hive.schema
index 6eeaca17c779..86fb8c6e5ab7 100644
--- a/ydb/tests/functional/scheme_tests/canondata/tablet_scheme_tests.TestTabletSchemes.test_tablet_schemes_flat_hive_/flat_hive.schema
+++ b/ydb/tests/functional/scheme_tests/canondata/tablet_scheme_tests.TestTabletSchemes.test_tablet_schemes_flat_hive_/flat_hive.schema
@@ -652,6 +652,11 @@
                 "ColumnId": 10,
                 "ColumnName": "Name",
                 "ColumnType": "String"
+            },
+            {
+                "ColumnId": 11,
+                "ColumnName": "BecomeUpOnRestart",
+                "ColumnType": "Bool"
             }
         ],
         "ColumnsDropped": [],
@@ -667,7 +672,8 @@
                     6,
                     8,
                     9,
-                    10
+                    10,
+                    11
                 ],
                 "RoomID": 0,
                 "Codec": 0,
