diff --git a/tests/queries/0_stateless/01925_broken_partition_id_zookeeper.reference b/tests/queries/0_stateless/01925_broken_partition_id_zookeeper.reference
new file mode 100644
index 000000000000..e69de29bb2d1
diff --git a/tests/queries/0_stateless/01925_broken_partition_id_zookeeper.sql b/tests/queries/0_stateless/01925_broken_partition_id_zookeeper.sql
new file mode 100644
index 000000000000..baf6c1fbf8f2
--- /dev/null
+++ b/tests/queries/0_stateless/01925_broken_partition_id_zookeeper.sql
@@ -0,0 +1,16 @@
+DROP TABLE IF EXISTS broken_partition;
+
+CREATE TABLE broken_partition
+(
+    date Date,
+    key UInt64
+)
+ENGINE = ReplicatedMergeTree('/clickhouse/test_01925_{database}/rmt', 'r1')
+ORDER BY tuple()
+PARTITION BY date;
+
+ALTER TABLE broken_partition DROP PARTITION ID '20210325_0_13241_6_12747'; --{serverError 248}
+
+ALTER TABLE broken_partition DROP PARTITION ID '20210325_0_13241_6_12747'; --{serverError 248}
+
+DROP TABLE IF EXISTS broken_partition;
