diff --git a/tests/queries/0_stateless/03461_clickhouse_local_extra_options.reference b/tests/queries/0_stateless/03461_clickhouse_local_extra_options.reference
new file mode 100644
index 000000000000..459fa4db3093
--- /dev/null
+++ b/tests/queries/0_stateless/03461_clickhouse_local_extra_options.reference
@@ -0,0 +1,2 @@
+666
+666
diff --git a/tests/queries/0_stateless/03461_clickhouse_local_extra_options.sh b/tests/queries/0_stateless/03461_clickhouse_local_extra_options.sh
new file mode 100755
index 000000000000..2af5b0707d77
--- /dev/null
+++ b/tests/queries/0_stateless/03461_clickhouse_local_extra_options.sh
@@ -0,0 +1,8 @@
+#!/usr/bin/env bash
+
+CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+# shellcheck source=../shell_config.sh
+. "$CURDIR"/../shell_config.sh
+
+${CLICKHOUSE_LOCAL} --query "SELECT getServerSetting('max_connections')" -- --max_connections=666
+${CLICKHOUSE_LOCAL} --query "SELECT getServerSetting('max_connections')" -- --max_connections 666
