diff --git a/data/parquet-testing/out_of_range_stats.parquet b/data/parquet-testing/out_of_range_stats.parquet
new file mode 100644
index 000000000000..3c3be2cb7f11
Binary files /dev/null and b/data/parquet-testing/out_of_range_stats.parquet differ
diff --git a/test/sql/copy/parquet/parquet_corrupt_stats.test b/test/sql/copy/parquet/parquet_corrupt_stats.test
new file mode 100644
index 000000000000..fb5a5362dfcd
--- /dev/null
+++ b/test/sql/copy/parquet/parquet_corrupt_stats.test
@@ -0,0 +1,10 @@
+# name: test/sql/copy/parquet/parquet_corrupt_stats.test
+# description: Test reading a Parquet file with stats that are out-of-range of the type
+# group: [parquet]
+
+require parquet
+
+query I
+FROM 'data/parquet-testing/out_of_range_stats.parquet'
+----
+255
