diff --git a/dbms/src/Storages/Kafka/KafkaBlockInputStream.cpp b/dbms/src/Storages/Kafka/KafkaBlockInputStream.cpp
index 80efc9f49965..c1252086b1de 100644
--- a/dbms/src/Storages/Kafka/KafkaBlockInputStream.cpp
+++ b/dbms/src/Storages/Kafka/KafkaBlockInputStream.cpp
@@ -199,8 +199,6 @@ Block KafkaBlockInputStream::readImpl()
 
 void KafkaBlockInputStream::readSuffixImpl()
 {
-    broken = false;
-
     if (commit_in_suffix)
         commit();
 }
@@ -211,6 +209,8 @@ void KafkaBlockInputStream::commit()
         return;
 
     buffer->commit();
+
+    broken = false;
 }
 
 }
diff --git a/dbms/src/Storages/Kafka/ReadBufferFromKafkaConsumer.cpp b/dbms/src/Storages/Kafka/ReadBufferFromKafkaConsumer.cpp
index 10f5fd0b47af..7edb4c0ac4f6 100644
--- a/dbms/src/Storages/Kafka/ReadBufferFromKafkaConsumer.cpp
+++ b/dbms/src/Storages/Kafka/ReadBufferFromKafkaConsumer.cpp
@@ -78,9 +78,18 @@ ReadBufferFromKafkaConsumer::ReadBufferFromKafkaConsumer(
 ReadBufferFromKafkaConsumer::~ReadBufferFromKafkaConsumer()
 {
     /// NOTE: see https://github.com/edenhill/librdkafka/issues/2077
-    consumer->unsubscribe();
-    consumer->unassign();
-    while (consumer->get_consumer_queue().next_event(100ms));
+    try
+    {
+        if (!consumer->get_subscription().empty())
+            consumer->unsubscribe();
+        if (!assignment.empty())
+            consumer->unassign();
+        while (consumer->get_consumer_queue().next_event(100ms));
+    }
+    catch (const cppkafka::HandleException & e)
+    {
+        LOG_ERROR(log, "Exception from ReadBufferFromKafkaConsumer destructor: " << e.what());
+    }
 }
 
 void ReadBufferFromKafkaConsumer::commit()
@@ -184,7 +193,17 @@ void ReadBufferFromKafkaConsumer::unsubscribe()
     current = messages.begin();
     BufferBase::set(nullptr, 0, 0);
 
-    consumer->unsubscribe();
+    // it should not raise exception as used in destructor
+    try
+    {
+        if (!consumer->get_subscription().empty())
+            consumer->unsubscribe();
+    }
+    catch (const cppkafka::HandleException & e)
+    {
+        LOG_ERROR(log, "Exception from ReadBufferFromKafkaConsumer::unsubscribe: " << e.what());
+    }
+
 }
 
 
