diff --git a/dbms/src/Interpreters/CrossToInnerJoinVisitor.cpp b/dbms/src/Interpreters/CrossToInnerJoinVisitor.cpp
index 596819dcde9d..54d5205c4c29 100644
--- a/dbms/src/Interpreters/CrossToInnerJoinVisitor.cpp
+++ b/dbms/src/Interpreters/CrossToInnerJoinVisitor.cpp
@@ -124,6 +124,10 @@ class CheckExpressionVisitorData
         {
             /// leave other comparisons as is
         }
+        else if (functionIsLikeOperator(node.name)) /// LIKE, NOT LIKE
+        {
+            /// leave as is
+        }
         else if (functionIsInOperator(node.name)) /// IN, NOT IN
         {
             if (auto ident = node.arguments->children.at(0)->as<ASTIdentifier>())
diff --git a/dbms/src/Interpreters/misc.h b/dbms/src/Interpreters/misc.h
index d5e2894bb4cf..e2f34375dc02 100644
--- a/dbms/src/Interpreters/misc.h
+++ b/dbms/src/Interpreters/misc.h
@@ -13,4 +13,9 @@ inline bool functionIsInOrGlobalInOperator(const std::string & name)
     return functionIsInOperator(name) || name == "globalIn" || name == "globalNotIn";
 }
 
+inline bool functionIsLikeOperator(const std::string & name)
+{
+    return name == "like" || name == "notLike";
+}
+
 }
