diff --git a/tests/queries/0_stateless/02210_toColumnTypeName_toLowCardinality_const.reference b/tests/queries/0_stateless/02210_toColumnTypeName_toLowCardinality_const.reference
index 2ac2f690f1be..1e3d3a50562b 100644
--- a/tests/queries/0_stateless/02210_toColumnTypeName_toLowCardinality_const.reference
+++ b/tests/queries/0_stateless/02210_toColumnTypeName_toLowCardinality_const.reference
@@ -1,1 +1,1 @@
-Const(UInt8)
+Const(ColumnLowCardinality)
diff --git a/tests/queries/0_stateless/02354_array_lowcardinality.reference b/tests/queries/0_stateless/02354_array_lowcardinality.reference
new file mode 100644
index 000000000000..d364bbdb5f49
--- /dev/null
+++ b/tests/queries/0_stateless/02354_array_lowcardinality.reference
@@ -0,0 +1,12 @@
+Array(LowCardinality(String))
+Array(LowCardinality(String))
+Array(LowCardinality(String))
+Array(LowCardinality(String))
+Array(String)
+Array(String)
+Array(String)
+Array(String)
+Map(LowCardinality(String), LowCardinality(String))
+Map(LowCardinality(String), LowCardinality(String))
+Map(LowCardinality(String), LowCardinality(String))
+Map(LowCardinality(String), LowCardinality(String))
diff --git a/tests/queries/0_stateless/02354_array_lowcardinality.sql b/tests/queries/0_stateless/02354_array_lowcardinality.sql
new file mode 100644
index 000000000000..1aa26b290506
--- /dev/null
+++ b/tests/queries/0_stateless/02354_array_lowcardinality.sql
@@ -0,0 +1,14 @@
+SELECT toTypeName([toLowCardinality('1'), toLowCardinality('2')]);
+SELECT toTypeName([materialize(toLowCardinality('1')), toLowCardinality('2')]);
+SELECT toTypeName([toLowCardinality('1'), materialize(toLowCardinality('2'))]);
+SELECT toTypeName([materialize(toLowCardinality('1')), materialize(toLowCardinality('2'))]);
+
+SELECT toTypeName([toLowCardinality('1'), '2']);
+SELECT toTypeName([materialize(toLowCardinality('1')), '2']);
+SELECT toTypeName([toLowCardinality('1'), materialize('2')]);
+SELECT toTypeName([materialize(toLowCardinality('1')), materialize('2')]);
+
+SELECT toTypeName(map(toLowCardinality('1'), toLowCardinality('2')));
+SELECT toTypeName(map(materialize(toLowCardinality('1')), toLowCardinality('2')));
+SELECT toTypeName(map(toLowCardinality('1'), materialize(toLowCardinality('2'))));
+SELECT toTypeName(map(materialize(toLowCardinality('1')), materialize(toLowCardinality('2'))));
diff --git a/tests/queries/0_stateless/02354_tuple_lowcardinality.reference b/tests/queries/0_stateless/02354_tuple_lowcardinality.reference
new file mode 100644
index 000000000000..435417d7f2ba
--- /dev/null
+++ b/tests/queries/0_stateless/02354_tuple_lowcardinality.reference
@@ -0,0 +1,4 @@
+Tuple(LowCardinality(String), LowCardinality(UInt8))
+Tuple(LowCardinality(String), LowCardinality(UInt8))
+Tuple(LowCardinality(String), LowCardinality(UInt8))
+Tuple(LowCardinality(String), LowCardinality(UInt8))
diff --git a/tests/queries/0_stateless/02354_tuple_lowcardinality.sql b/tests/queries/0_stateless/02354_tuple_lowcardinality.sql
new file mode 100644
index 000000000000..44b64aab3272
--- /dev/null
+++ b/tests/queries/0_stateless/02354_tuple_lowcardinality.sql
@@ -0,0 +1,6 @@
+SET allow_suspicious_low_cardinality_types = 1;
+
+SELECT toTypeName(tuple(toLowCardinality('1'), toLowCardinality(1)));
+SELECT toTypeName(tuple(materialize(toLowCardinality('1')), toLowCardinality(1)));
+SELECT toTypeName(tuple(toLowCardinality('1'), materialize(toLowCardinality(1))));
+SELECT toTypeName(tuple(materialize(toLowCardinality('1')), materialize(toLowCardinality(1))));
diff --git a/tests/queries/0_stateless/02355_column_type_name_lc.reference b/tests/queries/0_stateless/02355_column_type_name_lc.reference
new file mode 100644
index 000000000000..234a072299f9
--- /dev/null
+++ b/tests/queries/0_stateless/02355_column_type_name_lc.reference
@@ -0,0 +1,1 @@
+ColumnLowCardinality
diff --git a/tests/queries/0_stateless/02355_column_type_name_lc.sql b/tests/queries/0_stateless/02355_column_type_name_lc.sql
new file mode 100644
index 000000000000..13a6393aaf28
--- /dev/null
+++ b/tests/queries/0_stateless/02355_column_type_name_lc.sql
@@ -0,0 +1,1 @@
+SELECT toColumnTypeName(toLowCardinality(materialize('foo')));
