diff --git a/tests/queries/0_stateless/02100_now64_types_bug.reference b/tests/queries/0_stateless/02100_now64_types_bug.reference
new file mode 100644
index 000000000000..ae136939b645
--- /dev/null
+++ b/tests/queries/0_stateless/02100_now64_types_bug.reference
@@ -0,0 +1,3 @@
+2
+1
+0
diff --git a/tests/queries/0_stateless/02100_now64_types_bug.sql b/tests/queries/0_stateless/02100_now64_types_bug.sql
new file mode 100644
index 000000000000..ef91b19bb08c
--- /dev/null
+++ b/tests/queries/0_stateless/02100_now64_types_bug.sql
@@ -0,0 +1,8 @@
+SELECT x
+FROM
+(
+    SELECT if((number % NULL) = -2147483648, NULL, if(toInt64(toInt64(now64(if((number % NULL) = -2147483648, NULL, if(toInt64(now64(toInt64(9223372036854775807, now64(plus(NULL, NULL))), plus(NULL, NULL))) = (number % NULL), nan, toFloat64(number))), toInt64(9223372036854775807, toInt64(9223372036854775807, now64(plus(NULL, NULL))), now64(plus(NULL, NULL))), plus(NULL, NULL))), now64(toInt64(9223372036854775807, toInt64(0, now64(plus(NULL, NULL))), now64(plus(NULL, NULL))), plus(NULL, NULL))) = (number % NULL), nan, toFloat64(number))) AS x
+    FROM system.numbers
+    LIMIT 3
+)
+ORDER BY x DESC NULLS LAST
