diff --git a/tests/queries/0_stateless/02814_order_by_tuple_window_function.reference b/tests/queries/0_stateless/02814_order_by_tuple_window_function.reference
new file mode 100644
index 000000000000..d00491fd7e5b
--- /dev/null
+++ b/tests/queries/0_stateless/02814_order_by_tuple_window_function.reference
@@ -0,0 +1,1 @@
+1
diff --git a/tests/queries/0_stateless/02814_order_by_tuple_window_function.sql b/tests/queries/0_stateless/02814_order_by_tuple_window_function.sql
new file mode 100644
index 000000000000..8ba54fc11d0a
--- /dev/null
+++ b/tests/queries/0_stateless/02814_order_by_tuple_window_function.sql
@@ -0,0 +1,1 @@
+SELECT 1 ORDER BY tuple(count() OVER ());
