diff --git a/tools/rpkg/DESCRIPTION b/tools/rpkg/DESCRIPTION
index 20e1ce6d4005..cc16c972fd58 100644
--- a/tools/rpkg/DESCRIPTION
+++ b/tools/rpkg/DESCRIPTION
@@ -24,12 +24,12 @@ Depends:
     DBI,
     R (>= 3.5.0)
 Imports:
-    dbplyr,
     methods,
     utils
 Suggests: 
     callr,
     DBItest,
+    dbplyr,
     nycflights13,
     testthat
 Encoding: UTF-8
diff --git a/tools/rpkg/NAMESPACE b/tools/rpkg/NAMESPACE
index e4db36722b86..1c48e08e5fdf 100644
--- a/tools/rpkg/NAMESPACE
+++ b/tools/rpkg/NAMESPACE
@@ -35,6 +35,5 @@ exportMethods(show)
 import(DBI)
 import(methods)
 importFrom(DBI,dbConnect)
-importFrom(dbplyr,src_dbi)
 importFrom(utils,head)
 useDynLib(duckdb, .registration = TRUE)
diff --git a/tools/rpkg/R/Driver.R b/tools/rpkg/R/Driver.R
index b6e80cfb267f..71a4a342ea08 100644
--- a/tools/rpkg/R/Driver.R
+++ b/tools/rpkg/R/Driver.R
@@ -167,15 +167,12 @@ is_installed <- function(pkg) {
 
 
 #' @importFrom DBI dbConnect
-#' @importFrom dbplyr src_dbi
 #' @param path The file in which the DuckDB database should be stored
 #' @param create Create a new database if none is present in `path`
 #' @rdname duckdb_driver
 #' @export
 src_duckdb <- function(path = ":memory:", create = FALSE, read_only = FALSE) {
-  if (!is_installed("dbplyr")) {
-    stop("Need package `dbplyr` installed.")
-  }
+  requireNamespace("dbplyr", quietly = TRUE)
   if (path != ":memory:" && !create && !file.exists(path)) {
     stop("`path` '", path, "' must already exist, unless `create` = TRUE")
   }
