diff --git a/tests/queries/0_stateless/02015_async_inserts_1.reference b/tests/queries/0_stateless/02015_async_inserts_1.reference
index 85305b346574..e58540d7333d 100644
--- a/tests/queries/0_stateless/02015_async_inserts_1.reference
+++ b/tests/queries/0_stateless/02015_async_inserts_1.reference
@@ -10,3 +10,4 @@
 10	j
 11	k
 12	l
+13	m
diff --git a/tests/queries/0_stateless/02015_async_inserts_1.sh b/tests/queries/0_stateless/02015_async_inserts_1.sh
index b4310f5101cd..60b8a540703f 100755
--- a/tests/queries/0_stateless/02015_async_inserts_1.sh
+++ b/tests/queries/0_stateless/02015_async_inserts_1.sh
@@ -32,6 +32,8 @@ ${CLICKHOUSE_CURL} -sS "$url" -d 'INSERT INTO async_inserts FORMAT CSVWithNames
 12,"l"
 ' &
 
+${CLICKHOUSE_CURL} -sS "$url" -H 'X-Clickhouse-Format: XML' -d 'INSERT INTO async_inserts FORMAT JSONEachRow {"id": 13, "s": "m"}' &
+
 wait
 
 ${CLICKHOUSE_CLIENT} -q "SELECT * FROM async_inserts ORDER BY id"
