diff --git a/tests/queries/0_stateless/03164_analyzer_rewrite_aggregate_function_with_if.reference b/tests/queries/0_stateless/03164_analyzer_rewrite_aggregate_function_with_if.reference
new file mode 100644
index 000000000000..d00491fd7e5b
--- /dev/null
+++ b/tests/queries/0_stateless/03164_analyzer_rewrite_aggregate_function_with_if.reference
@@ -0,0 +1,1 @@
+1
diff --git a/tests/queries/0_stateless/03164_analyzer_rewrite_aggregate_function_with_if.sql b/tests/queries/0_stateless/03164_analyzer_rewrite_aggregate_function_with_if.sql
new file mode 100644
index 000000000000..52f767d8aae8
--- /dev/null
+++ b/tests/queries/0_stateless/03164_analyzer_rewrite_aggregate_function_with_if.sql
@@ -0,0 +1,1 @@
+SELECT countIf(multiIf(number < 2, NULL, if(number = 4, 1, 0))) FROM numbers(5);
