diff --git a/src/Processors/OffsetTransform.cpp b/src/Processors/OffsetTransform.cpp
index 0b2fb0a5cc40..40a0833ffb4a 100644
--- a/src/Processors/OffsetTransform.cpp
+++ b/src/Processors/OffsetTransform.cpp
@@ -96,13 +96,14 @@ OffsetTransform::Status OffsetTransform::preparePair(PortsData & data)
     auto & input = *data.input_port;
 
     /// Check can output.
-    bool output_finished = false;
+
     if (output.isFinished())
     {
-        output_finished = true;
+        input.close();
+        return Status::Finished;
     }
 
-    if (!output_finished && !output.canPush())
+    if (!output.canPush())
     {
         input.setNotNeeded();
         return Status::PortFull;
