diff --git a/src/server/hll_family_test.cc b/src/server/hll_family_test.cc
index 7d162fc4fa63..4b52924daa54 100644
--- a/src/server/hll_family_test.cc
+++ b/src/server/hll_family_test.cc
@@ -194,6 +194,8 @@ TEST_F(HllFamilyTest, MergeOverlapping) {
 }
 
 TEST_F(HllFamilyTest, MergeInvalid) {
+  GTEST_SKIP() << "TBD: MergeInvalid test fails with multi-shard runs, see #5004";
+
   EXPECT_EQ(CheckedInt({"pfadd", "key1", "1", "2", "3"}), 1);
   EXPECT_EQ(Run({"set", "key2", "..."}), "OK");
   EXPECT_THAT(Run({"pfmerge", "key1", "key2"}), ErrArg(HllFamily::kInvalidHllErr));
