diff --git a/tests/queries/0_stateless/01293_external_sorting_limit_bug.reference b/tests/queries/0_stateless/01293_external_sorting_limit_bug.reference
new file mode 100644
index 000000000000..e69de29bb2d1
diff --git a/tests/queries/0_stateless/01293_external_sorting_limit_bug.sql b/tests/queries/0_stateless/01293_external_sorting_limit_bug.sql
new file mode 100644
index 000000000000..570ec4b8c3ce
--- /dev/null
+++ b/tests/queries/0_stateless/01293_external_sorting_limit_bug.sql
@@ -0,0 +1,1 @@
+SELECT number FROM (SELECT number FROM system.numbers LIMIT 999990) ORDER BY number ASC LIMIT 100, 65535 SETTINGS max_bytes_before_external_sort = 1000000 format Null
