diff --git a/data/parquet-testing/arrow/lineitem-arrow.parquet b/data/parquet-testing/arrow/lineitem-arrow.parquet
new file mode 100644
index 000000000000..bf19dd5c8722
Binary files /dev/null and b/data/parquet-testing/arrow/lineitem-arrow.parquet differ
diff --git a/data/parquet-testing/decimals.parquet b/data/parquet-testing/decimals.parquet
new file mode 100644
index 000000000000..44c379c0c631
Binary files /dev/null and b/data/parquet-testing/decimals.parquet differ
diff --git a/test/sql/copy/parquet/decimal_filter.test b/test/sql/copy/parquet/decimal_filter.test
new file mode 100644
index 000000000000..4087e2d8a723
--- /dev/null
+++ b/test/sql/copy/parquet/decimal_filter.test
@@ -0,0 +1,28 @@
+# name: test/sql/copy/parquet/decimal_filter.test
+# description: Decimal filter pushdown into Parquet
+# group: [parquet]
+
+require parquet
+
+statement ok
+pragma enable_verification
+
+query IIII
+select * from 'data/parquet-testing/decimals.parquet'
+----
+0.1	0.1	0.1	0.1
+-0.1	-0.1	-0.1	-0.1
+
+loop i 1 5
+
+query IIII
+select * from 'data/parquet-testing/decimals.parquet' WHERE l${i}=0.1
+----
+0.1	0.1	0.1	0.1
+
+query IIII
+select * from 'data/parquet-testing/decimals.parquet' WHERE l${i}=-0.1
+----
+-0.1	-0.1	-0.1	-0.1
+
+endloop
\ No newline at end of file
diff --git a/test/sql/copy/parquet/lineitem_arrow.test b/test/sql/copy/parquet/lineitem_arrow.test
new file mode 100644
index 000000000000..4d3d318cb429
--- /dev/null
+++ b/test/sql/copy/parquet/lineitem_arrow.test
@@ -0,0 +1,33 @@
+# name: test/sql/copy/parquet/lineitem_arrow.test
+# description: Issue #2261: TPC-H Q6 fails on Parquet input
+# group: [parquet]
+
+require tpch
+
+require parquet
+
+statement ok
+CREATE TABLE lineitem AS SELECT * FROM 'data/parquet-testing/arrow/lineitem-arrow.parquet'
+
+query I nosort q01
+PRAGMA tpch(1)
+----
+
+query I nosort q06
+PRAGMA tpch(6)
+----
+
+statement ok
+DROP TABLE lineitem
+
+statement ok
+CREATE VIEW lineitem AS SELECT * FROM 'data/parquet-testing/arrow/lineitem-arrow.parquet'
+
+query I nosort q01
+PRAGMA tpch(1)
+----
+
+query I nosort q06
+PRAGMA tpch(6)
+----
+
