diff --git a/src/Formats/PrettyFormatHelpers.cpp b/src/Formats/PrettyFormatHelpers.cpp
index a56e0d5ca6da..acd9ce000870 100644
--- a/src/Formats/PrettyFormatHelpers.cpp
+++ b/src/Formats/PrettyFormatHelpers.cpp
@@ -5,7 +5,6 @@
 #include <Processors/Chunk.h>
 #include <Common/formatReadable.h>
 #include <Common/UTF8Helpers.h>
-#include <base/find_symbols.h>
 
 
 static constexpr const char * GRAY_COLOR = "\033[90m";
@@ -29,9 +28,12 @@ void writeReadableNumberTip(WriteBuffer & out, const IColumn & column, size_t ro
         return;
 
     auto value = column.getFloat64(row);
+    auto abs_value = abs(value);
     auto threshold = settings.pretty.single_large_number_tip_threshold;
 
-    if (threshold && isFinite(value) && abs(value) > threshold)
+    if (threshold && isFinite(value) && abs_value > threshold
+        /// Most (~99.5%) of 64-bit hash values are in this range, and it is not necessarily to highlight them:
+        && !(abs_value > 1e17 && abs_value < 1.844675e19))
     {
         if (color)
             writeCString(GRAY_COLOR, out);
