diff --git a/src/Functions/registerFunctions.cpp b/src/Functions/registerFunctions.cpp
index 35193e9be8d0..b2f038240aa0 100644
--- a/src/Functions/registerFunctions.cpp
+++ b/src/Functions/registerFunctions.cpp
@@ -54,9 +54,6 @@ void registerFunctionTupleHammingDistance(FunctionFactory & factory);
 void registerFunctionsStringHash(FunctionFactory & factory);
 void registerFunctionValidateNestedArraySizes(FunctionFactory & factory);
 void registerFunctionsSnowflake(FunctionFactory & factory);
-#if !defined(ARCADIA_BUILD)
-void registerFunctionBayesAB(FunctionFactory &);
-#endif
 void registerFunctionTid(FunctionFactory & factory);
 void registerFunctionLogTrace(FunctionFactory & factory);
 
@@ -122,10 +119,6 @@ void registerFunctions()
     registerFunctionValidateNestedArraySizes(factory);
     registerFunctionsSnowflake(factory);
 
-#if !defined(ARCADIA_BUILD)
-    registerFunctionBayesAB(factory);
-#endif
-
 #if USE_SSL
     registerFunctionEncrypt(factory);
     registerFunctionDecrypt(factory);
