diff --git a/test/sql/alter/rename_table/test_rename_table_view.test b/test/sql/alter/rename_table/test_rename_table_view.test
index 6efc0d0d5bd7..88bd2e21eefc 100644
--- a/test/sql/alter/rename_table/test_rename_table_view.test
+++ b/test/sql/alter/rename_table/test_rename_table_view.test
@@ -15,7 +15,7 @@ statement error
 ALTER TABLE v1 RENAME TO v2
 
 query I
-SELECT * FROM  v1
+SELECT * FROM v1
 ----
 999
 100
diff --git a/test/sql/alter/rename_view/test_rename_view.test b/test/sql/alter/rename_view/test_rename_view.test
new file mode 100644
index 000000000000..4d9e95569aba
--- /dev/null
+++ b/test/sql/alter/rename_view/test_rename_view.test
@@ -0,0 +1,60 @@
+# name: test/sql/alter/rename_view/test_rename_view.test
+# description: Test RENAME VIEW single transaction
+# group: [rename_view]
+
+statement ok
+CREATE TABLE tbl(i INTEGER);
+INSERT INTO tbl VALUES (999), (100);
+CREATE VIEW vw AS SELECT * FROM tbl;
+
+statement ok
+BEGIN TRANSACTION
+
+statement ok
+ALTER VIEW vw RENAME TO vw2
+
+query I
+SELECT * FROM vw2
+----
+999
+100
+
+statement error
+SELECT * FROM vw
+
+statement ok
+ROLLBACK
+
+query I
+SELECT * FROM vw;
+----
+999
+100
+
+statement error
+SELECT * FROM vw2
+
+statement ok
+BEGIN TRANSACTION;
+
+statement ok
+ALTER VIEW vw RENAME TO vw2
+
+statement ok
+COMMIT
+
+query I
+SELECT * FROM vw2
+----
+999
+100
+
+statement error
+SELECT * FROM vw
+
+query I
+SELECT * FROM vw2
+----
+999
+100
+
diff --git a/test/sql/alter/rename_view/test_rename_view_incorrect.test b/test/sql/alter/rename_view/test_rename_view_incorrect.test
new file mode 100644
index 000000000000..44565ac40459
--- /dev/null
+++ b/test/sql/alter/rename_view/test_rename_view_incorrect.test
@@ -0,0 +1,21 @@
+# name: test/sql/alter/rename_view/test_rename_view_incorrect.test
+# description: Test RENAME VIEW: view does not exist and rename to an already existing view
+# group: [rename_view]
+
+statement ok
+CREATE TABLE tbl(i INTEGER)
+
+statement ok
+CREATE VIEW vw AS SELECT * FROM tbl
+
+statement ok
+CREATE VIEW vw2 AS SELECT 1729 AS i
+
+# Renaming a non existing view
+statement error
+ALTER VIEW non_view RENAME TO vw
+
+# rename to an already existing view
+statement error
+ALTER VIEW vw2 RENAME TO vw
+
diff --git a/test/sql/alter/rename_view/test_rename_view_many_transactions.test b/test/sql/alter/rename_view/test_rename_view_many_transactions.test
new file mode 100644
index 000000000000..eae076444b7c
--- /dev/null
+++ b/test/sql/alter/rename_view/test_rename_view_many_transactions.test
@@ -0,0 +1,101 @@
+# name: test/sql/alter/rename_view/test_rename_view_many_transactions.test
+# description: Test RENAME VIEW four view renames and four parallel transactions
+# group: [rename_view]
+
+statement ok con1
+CREATE TABLE tbl1(i INTEGER)
+
+statement ok con1
+INSERT INTO tbl1 VALUES (999), (100)
+
+statement ok con1
+CREATE VIEW vw1 AS SELECT * FROM tbl1
+
+# rename chain
+# con2 starts a transaction now
+statement ok con2
+BEGIN TRANSACTION
+
+# rename in con1, con2 should still see "vw1"
+statement ok con1
+ALTER VIEW vw1 RENAME TO vw2
+
+# con3 starts a transaction now
+statement ok con3
+BEGIN TRANSACTION
+
+# rename in con1, con3 should still see "vw2"
+statement ok con1
+ALTER VIEW vw2 RENAME TO vw3
+
+# con4 starts a transaction now
+statement ok con4
+BEGIN TRANSACTION
+
+# rename in con1, con4 should still see "vw3"
+statement ok con1
+ALTER VIEW vw3 RENAME TO vw4
+
+# con2 sees ONLY vw1
+query I con2
+SELECT * FROM vw1
+----
+999
+100
+
+statement error con2
+SELECT * FROM vw2
+
+statement error con2
+SELECT * FROM vw3
+
+statement error con2
+SELECT * FROM vw4
+
+# con3 sees ONLY vw2
+statement error con3
+SELECT * FROM vw1
+
+query I con3
+SELECT * FROM vw2
+----
+999
+100
+
+statement error con3
+SELECT * FROM vw3
+
+statement error con3
+SELECT * FROM vw4
+
+# con4 sees ONLY vw3
+statement error con4
+SELECT * FROM vw1
+
+statement error con4
+SELECT * FROM vw2
+
+query I con4
+SELECT * FROM vw3
+----
+999
+100
+
+statement error con4
+SELECT * FROM vw4
+
+# con1 sees ONLY vw4
+statement error con1
+SELECT * FROM vw1
+
+statement error con1
+SELECT * FROM vw2
+
+statement error con1
+SELECT * FROM vw3
+
+query I con1
+SELECT * FROM vw4
+----
+999
+100
diff --git a/test/sql/alter/rename_view/test_rename_view_table.test b/test/sql/alter/rename_view/test_rename_view_table.test
new file mode 100644
index 000000000000..e2f448c62123
--- /dev/null
+++ b/test/sql/alter/rename_view/test_rename_view_table.test
@@ -0,0 +1,22 @@
+# name: test/sql/alter/rename_view/test_rename_view_table.test
+# description: Test RENAME VIEW with a table as entry
+# group: [rename_view]
+
+statement ok
+CREATE TABLE tbl(i INTEGER)
+
+statement ok
+INSERT INTO tbl VALUES (999), (100)
+
+statement ok
+CREATE VIEW v1 AS SELECT * FROM tbl
+
+statement error
+ALTER VIEW tbl RENAME TO tbl2
+
+query I
+SELECT * FROM v1
+----
+999
+100
+
diff --git a/test/sql/alter/rename_view/test_rename_view_transactions.test b/test/sql/alter/rename_view/test_rename_view_transactions.test
new file mode 100644
index 000000000000..a92e0cbe7820
--- /dev/null
+++ b/test/sql/alter/rename_view/test_rename_view_transactions.test
@@ -0,0 +1,64 @@
+# name: test/sql/alter/rename_view/test_rename_view_transactions.test
+# description: Test RENAME VIEW two parallel transactions
+# group: [rename_view]
+
+statement ok con1
+CREATE TABLE tbl(i INTEGER)
+
+statement ok con1
+INSERT INTO tbl VALUES (999), (100)
+
+statement ok con1
+CREATE VIEW vw AS SELECT * FROM tbl
+
+statement ok con1
+BEGIN TRANSACTION
+
+statement ok con2
+BEGIN TRANSACTION
+
+statement ok con1
+ALTER VIEW vw RENAME TO vw2
+
+query I con1
+SELECT * FROM vw2
+----
+999
+100
+
+statement error con1
+SELECT * FROM vw
+
+query I con2
+SELECT * FROM vw
+----
+999
+100
+
+statement error con2
+SELECT * FROM vw2
+
+statement ok con1
+COMMIT
+
+statement ok con2
+COMMIT
+
+statement error con1
+SELECT * FROM vw
+
+statement error con2
+SELECT * FROM vw
+
+query I con1
+SELECT * FROM vw2
+----
+999
+100
+
+query I con2
+SELECT * FROM vw2
+----
+999
+100
+
diff --git a/test/sql/storage/catalog/test_store_rename_table.test b/test/sql/storage/catalog/test_store_rename_table.test
index 7b624c0d29b7..aa358ca16847 100644
--- a/test/sql/storage/catalog/test_store_rename_table.test
+++ b/test/sql/storage/catalog/test_store_rename_table.test
@@ -1,5 +1,5 @@
-# name: test/sql/storage/catalog/test_rename_column.test
-# description: Test storage of alter table rename column
+# name: test/sql/storage/catalog/test_rename_table.test
+# description: Test storage of alter table rename table
 # group: [catalog]
 
 #  FIXME: right now rename is implemented as drop + create
diff --git a/test/sql/storage/catalog/test_store_rename_view.test b/test/sql/storage/catalog/test_store_rename_view.test
new file mode 100644
index 000000000000..e7f1d85eade6
--- /dev/null
+++ b/test/sql/storage/catalog/test_store_rename_view.test
@@ -0,0 +1,76 @@
+# name: test/sql/storage/catalog/test_rename_view.test
+# description: Test storage of alter view
+# group: [catalog]
+
+# load the DB from disk
+load __TEST_DIR__/test_rename_view.db
+
+statement ok
+CREATE TABLE test (a INTEGER, b INTEGER);
+
+statement ok
+INSERT INTO test VALUES (11, 22), (13, 22), (12, 21)
+
+statement ok
+CREATE VIEW vtest AS SELECT * FROM test
+
+statement ok
+BEGIN TRANSACTION
+
+query I
+SELECT a FROM vtest ORDER BY a
+----
+11
+12
+13
+
+statement ok
+ALTER VIEW vtest RENAME TO new_name
+
+query I
+SELECT a FROM new_name ORDER BY 1
+----
+11
+12
+13
+
+statement ok
+ROLLBACK
+
+# restart the database
+restart
+
+statement ok
+BEGIN TRANSACTION
+
+# verify that the table is still there in the original form
+query I
+SELECT a FROM vtest ORDER BY a
+----
+11
+12
+13
+
+# now repeat the process, but this time commit
+statement ok
+ALTER VIEW vtest RENAME TO new_name
+
+query I
+SELECT a FROM new_name ORDER BY 1
+----
+11
+12
+13
+
+statement ok
+COMMIT
+
+restart
+
+# after a restart, the renamed table is still here
+query I
+SELECT a FROM new_name ORDER BY 1
+----
+11
+12
+13
