diff --git a/tests/queries/0_stateless/01458_named_tuple_millin.reference b/tests/queries/0_stateless/01458_named_tuple_millin.reference
new file mode 100644
index 000000000000..b826566c74b2
--- /dev/null
+++ b/tests/queries/0_stateless/01458_named_tuple_millin.reference
@@ -0,0 +1,12 @@
+CREATE TABLE default.tuple
+(
+    `j` Tuple(    a Int8,     b String)
+)
+ENGINE = Memory
+j	Tuple(a Int8, b String)					
+CREATE TABLE default.tuple
+(
+    `j` Tuple(    a Int8,     b String)
+)
+ENGINE = Memory
+j	Tuple(a Int8, b String)					
diff --git a/tests/queries/0_stateless/01458_named_tuple_millin.sql b/tests/queries/0_stateless/01458_named_tuple_millin.sql
new file mode 100644
index 000000000000..7687dd4c158a
--- /dev/null
+++ b/tests/queries/0_stateless/01458_named_tuple_millin.sql
@@ -0,0 +1,19 @@
+DROP TABLE IF EXISTS tuple;
+
+CREATE TABLE tuple
+(
+    `j` Tuple(a Int8, b String)
+)
+ENGINE = Memory;
+
+SHOW CREATE TABLE tuple FORMAT TSVRaw;
+DESC tuple;
+DROP TABLE tuple;
+
+CREATE TABLE tuple
+ENGINE = Memory AS
+SELECT CAST((1, 'Test'), 'Tuple(a Int8,  b String)') AS j;
+
+SHOW CREATE TABLE tuple FORMAT TSVRaw;
+DESC tuple;
+DROP TABLE tuple;
