diff --git a/src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp b/src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp
index dd0080eea820..938f25b89fa8 100644
--- a/src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp
+++ b/src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp
@@ -429,7 +429,7 @@ SortingInputOrder buildInputOrderFromSortDescription(
         // ASC NULLS LAST ("in order") or DESC NULLS FIRST ("reverse")
         /// supported only this direction, other cases are represented as nulls_direction==-1
         /// Also actual for floating point values NaN.
-        const auto column_is_nullable = sorting_key.data_types[next_sort_key]->isNullable() || isFloat(*sorting_key.data_types[next_sort_key]);
+        const auto column_is_nullable = isNullableOrLowCardinalityNullable(sorting_key.data_types[next_sort_key])|| isFloat(*sorting_key.data_types[next_sort_key]);
         if (column_is_nullable && sort_column_description.nulls_direction == -1)
             break;
 
