diff --git a/src/server/json_family.cc b/src/server/json_family.cc
index a3774a882efd..5f14e9239945 100644
--- a/src/server/json_family.cc
+++ b/src/server/json_family.cc
@@ -505,6 +505,8 @@ OpStatus SetFullJson(const OpArgs& op_args, string_view key, string_view json_st
     return OpStatus::INVALID_JSON;
   }
 
+  op_args.GetDbSlice().RemoveExpire(op_args.db_cntx.db_index, it_res->it);
+
   if (JsonEnconding() == kEncodingJsonFlat) {
     flexbuffers::Builder fbb;
     json::FromJsonType(*parsed_json, &fbb);
