diff --git a/tests/queries/0_stateless/02535_analyzer_group_by_use_nulls.reference b/tests/queries/0_stateless/02535_analyzer_group_by_use_nulls.reference
index 63610604ddd9..858fbe988389 100644
--- a/tests/queries/0_stateless/02535_analyzer_group_by_use_nulls.reference
+++ b/tests/queries/0_stateless/02535_analyzer_group_by_use_nulls.reference
@@ -264,3 +264,44 @@ SETTINGS group_by_use_nulls = 1, max_bytes_before_external_sort=10;
 9	\N	9
 \N	0	20
 \N	1	25
+CREATE TABLE test
+ENGINE = ReplacingMergeTree
+PRIMARY KEY id
+AS SELECT number AS id FROM numbers(100);
+SELECT id
+FROM test
+GROUP BY id
+    WITH CUBE
+HAVING id IN (
+    SELECT id
+    FROM test
+)
+FORMAT `NUll`
+SETTINGS allow_experimental_analyzer = 1, group_by_use_nulls = true;
+SELECT id
+FROM test
+FINAL
+GROUP BY id
+    WITH CUBE
+HAVING id IN (
+    SELECT DISTINCT id
+    FROM test
+    FINAL
+)
+FORMAT `NUll`
+SETTINGS allow_experimental_analyzer = 1, group_by_use_nulls = true;
+SELECT id
+FROM test
+FINAL
+GROUP BY
+    GROUPING SETS ((id))
+ORDER BY
+    id IN (
+        SELECT DISTINCT id
+        FROM test
+        FINAL
+        LIMIT 4
+    ) ASC
+LIMIT 256 BY id
+FORMAT `NUll`
+SETTINGS allow_experimental_analyzer = 1, group_by_use_nulls=true;
diff --git a/tests/queries/0_stateless/02535_analyzer_group_by_use_nulls.sql b/tests/queries/0_stateless/02535_analyzer_group_by_use_nulls.sql
index a4d4f2f8bc94..4ae5df9629ad 100644
--- a/tests/queries/0_stateless/02535_analyzer_group_by_use_nulls.sql
+++ b/tests/queries/0_stateless/02535_analyzer_group_by_use_nulls.sql
@@ -83,3 +83,48 @@ GROUP BY
     )
 ORDER BY 1, tuple(val)
 SETTINGS group_by_use_nulls = 1, max_bytes_before_external_sort=10;
+
+CREATE TABLE test
+ENGINE = ReplacingMergeTree
+PRIMARY KEY id
+AS SELECT number AS id FROM numbers(100);
+
+SELECT id
+FROM test
+GROUP BY id
+    WITH CUBE
+HAVING id IN (
+    SELECT id
+    FROM test
+)
+FORMAT `NUll`
+SETTINGS allow_experimental_analyzer = 1, group_by_use_nulls = true;
+
+SELECT id
+FROM test
+FINAL
+GROUP BY id
+    WITH CUBE
+HAVING id IN (
+    SELECT DISTINCT id
+    FROM test
+    FINAL
+)
+FORMAT `NUll`
+SETTINGS allow_experimental_analyzer = 1, group_by_use_nulls = true;
+
+SELECT id
+FROM test
+FINAL
+GROUP BY
+    GROUPING SETS ((id))
+ORDER BY
+    id IN (
+        SELECT DISTINCT id
+        FROM test
+        FINAL
+        LIMIT 4
+    ) ASC
+LIMIT 256 BY id
+FORMAT `NUll`
+SETTINGS allow_experimental_analyzer = 1, group_by_use_nulls=true;
