diff --git a/src/Analyzer/Resolve/ScopeAliases.h b/src/Analyzer/Resolve/ScopeAliases.h
index baab843988b9..830ae72144b8 100644
--- a/src/Analyzer/Resolve/ScopeAliases.h
+++ b/src/Analyzer/Resolve/ScopeAliases.h
@@ -75,7 +75,12 @@ struct ScopeAliases
             if (jt == transitive_aliases.end())
                 return {};
 
-            key = &(getKey(jt->second, find_option));
+            const auto & new_key = getKey(jt->second, find_option);
+            /// Ignore potential cyclic aliases.
+            if (new_key == *key)
+                return {};
+
+            key = &new_key;
             it = alias_map.find(*key);
         }
 
