diff --git a/tests/queries/0_stateless/02896_cyclic_aliases_crash.reference b/tests/queries/0_stateless/02896_cyclic_aliases_crash.reference
index caf11f5c15a6..e537236478da 100644
--- a/tests/queries/0_stateless/02896_cyclic_aliases_crash.reference
+++ b/tests/queries/0_stateless/02896_cyclic_aliases_crash.reference
@@ -1,2 +1,3 @@
 1	2	3
 1	5
+300
diff --git a/tests/queries/0_stateless/02896_cyclic_aliases_crash.sql b/tests/queries/0_stateless/02896_cyclic_aliases_crash.sql
index 5fb628eeb67c..5440872e0524 100644
--- a/tests/queries/0_stateless/02896_cyclic_aliases_crash.sql
+++ b/tests/queries/0_stateless/02896_cyclic_aliases_crash.sql
@@ -30,3 +30,7 @@ WHERE (time_stamp_utc >= toDateTime('2024-04-25 00:00:00')) AND (time_stamp_utc
 GROUP BY time_stamp_utc
 ORDER BY Impressions DESC
 LIMIT 1000;
+
+drop table test_table;
+create table test_table engine MergeTree order by sum as select 100 as sum union all select 200 as sum;
+select sum as sum from (select sum(sum) as sum from test_table);
