diff --git a/ydb/tests/functional/canonical/canondata/test_sql.TestCanonicalFolder1.test_case_explain.script-script_/explain.script.plan b/ydb/tests/functional/canonical/canondata/test_sql.TestCanonicalFolder1.test_case_explain.script-script_/explain.script.plan
index 0faa53fd98e5..581652f1a138 100644
--- a/ydb/tests/functional/canonical/canondata/test_sql.TestCanonicalFolder1.test_case_explain.script-script_/explain.script.plan
+++ b/ydb/tests/functional/canonical/canondata/test_sql.TestCanonicalFolder1.test_case_explain.script-script_/explain.script.plan
@@ -520,29 +520,31 @@
                 "Plans": [
                     {
                         "Node Type": "Sink",
+                        "Operators": [
+                            {
+                                "Inputs": [],
+                                "Name": "Replace",
+                                "Path": "/local/base_explain_script_script/ScriptingTest",
+                                "SinkType": "KqpTableSink",
+                                "Table": "base_explain_script_script/ScriptingTest"
+                            }
+                        ],
                         "PlanNodeId": 2,
                         "Plans": [
                             {
-                                "Node Type": "ConstantExpr-Sink",
+                                "Node Type": "ConstantExpr",
                                 "Operators": [
                                     {
                                         "Inputs": [],
                                         "Iterator": "[{Key: 3,Value: \"Three\"},{Key: 4,Value: \"Four\"}]",
                                         "Name": "Iterator"
-                                    },
-                                    {
-                                        "Inputs": [],
-                                        "Name": "Replace",
-                                        "Path": "/local/base_explain_script_script/ScriptingTest",
-                                        "SinkType": "KqpTableSink",
-                                        "Table": "base_explain_script_script/ScriptingTest"
                                     }
                                 ],
-                                "PlanNodeId": 1,
-                                "Tables": [
-                                    "base_explain_script_script/ScriptingTest"
-                                ]
+                                "PlanNodeId": 1
                             }
+                        ],
+                        "Tables": [
+                            "base_explain_script_script/ScriptingTest"
                         ]
                     }
                 ],
@@ -572,29 +574,31 @@
                 "Plans": [
                     {
                         "Node Type": "Sink",
+                        "Operators": [
+                            {
+                                "Inputs": [],
+                                "Name": "Replace",
+                                "Path": "/local/base_explain_script_script/ScriptingTest",
+                                "SinkType": "KqpTableSink",
+                                "Table": "base_explain_script_script/ScriptingTest"
+                            }
+                        ],
                         "PlanNodeId": 2,
                         "Plans": [
                             {
-                                "Node Type": "ConstantExpr-Sink",
+                                "Node Type": "ConstantExpr",
                                 "Operators": [
                                     {
                                         "Inputs": [],
                                         "Iterator": "[{Key: 1,Value: \"One\"},{Key: 2,Value: \"Two\"}]",
                                         "Name": "Iterator"
-                                    },
-                                    {
-                                        "Inputs": [],
-                                        "Name": "Replace",
-                                        "Path": "/local/base_explain_script_script/ScriptingTest",
-                                        "SinkType": "KqpTableSink",
-                                        "Table": "base_explain_script_script/ScriptingTest"
                                     }
                                 ],
-                                "PlanNodeId": 1,
-                                "Tables": [
-                                    "base_explain_script_script/ScriptingTest"
-                                ]
+                                "PlanNodeId": 1
                             }
+                        ],
+                        "Tables": [
+                            "base_explain_script_script/ScriptingTest"
                         ]
                     }
                 ],
diff --git a/ydb/tests/functional/canonical/canondata/test_sql.TestCanonicalFolder1.test_case_join_group_by_lookup.script-script_/join_group_by_lookup.script.plan b/ydb/tests/functional/canonical/canondata/test_sql.TestCanonicalFolder1.test_case_join_group_by_lookup.script-script_/join_group_by_lookup.script.plan
index 67e5fb9fa820..d24bb2c08af4 100644
--- a/ydb/tests/functional/canonical/canondata/test_sql.TestCanonicalFolder1.test_case_join_group_by_lookup.script-script_/join_group_by_lookup.script.plan
+++ b/ydb/tests/functional/canonical/canondata/test_sql.TestCanonicalFolder1.test_case_join_group_by_lookup.script-script_/join_group_by_lookup.script.plan
@@ -282,29 +282,31 @@
                 "Plans": [
                     {
                         "Node Type": "Sink",
+                        "Operators": [
+                            {
+                                "Inputs": [],
+                                "Name": "Replace",
+                                "Path": "/local/base_join_group_by_lookup_script_script/Temp",
+                                "SinkType": "KqpTableSink",
+                                "Table": "base_join_group_by_lookup_script_script/Temp"
+                            }
+                        ],
                         "PlanNodeId": 2,
                         "Plans": [
                             {
-                                "Node Type": "ConstantExpr-Sink",
+                                "Node Type": "ConstantExpr",
                                 "Operators": [
                                     {
                                         "Inputs": [],
                                         "Iterator": "[{Group: 1,Value: \"One\"},{Group: 3,Value: \"Three\"}]",
                                         "Name": "Iterator"
-                                    },
-                                    {
-                                        "Inputs": [],
-                                        "Name": "Replace",
-                                        "Path": "/local/base_join_group_by_lookup_script_script/Temp",
-                                        "SinkType": "KqpTableSink",
-                                        "Table": "base_join_group_by_lookup_script_script/Temp"
                                     }
                                 ],
-                                "PlanNodeId": 1,
-                                "Tables": [
-                                    "base_join_group_by_lookup_script_script/Temp"
-                                ]
+                                "PlanNodeId": 1
                             }
+                        ],
+                        "Tables": [
+                            "base_join_group_by_lookup_script_script/Temp"
                         ]
                     }
                 ],
diff --git a/ydb/tests/functional/canonical/canondata/test_sql.TestCanonicalFolder1.test_case_simple_ct.script-script_/simple_ct.script.plan b/ydb/tests/functional/canonical/canondata/test_sql.TestCanonicalFolder1.test_case_simple_ct.script-script_/simple_ct.script.plan
index 10512bcaaa79..a767440d0255 100644
--- a/ydb/tests/functional/canonical/canondata/test_sql.TestCanonicalFolder1.test_case_simple_ct.script-script_/simple_ct.script.plan
+++ b/ydb/tests/functional/canonical/canondata/test_sql.TestCanonicalFolder1.test_case_simple_ct.script-script_/simple_ct.script.plan
@@ -11,29 +11,31 @@
                 "Plans": [
                     {
                         "Node Type": "Sink",
+                        "Operators": [
+                            {
+                                "Inputs": [],
+                                "Name": "Upsert",
+                                "Path": "/local/base_simple_ct_script_script/Questions",
+                                "SinkType": "KqpTableSink",
+                                "Table": "base_simple_ct_script_script/Questions"
+                            }
+                        ],
                         "PlanNodeId": 2,
                         "Plans": [
                             {
-                                "Node Type": "ConstantExpr-Sink",
+                                "Node Type": "ConstantExpr",
                                 "Operators": [
                                     {
                                         "Inputs": [],
                                         "Iterator": "[{idx: 1,text: \"to do or not to do\"}]",
                                         "Name": "Iterator"
-                                    },
-                                    {
-                                        "Inputs": [],
-                                        "Name": "Upsert",
-                                        "Path": "/local/base_simple_ct_script_script/Questions",
-                                        "SinkType": "KqpTableSink",
-                                        "Table": "base_simple_ct_script_script/Questions"
                                     }
                                 ],
-                                "PlanNodeId": 1,
-                                "Tables": [
-                                    "base_simple_ct_script_script/Questions"
-                                ]
+                                "PlanNodeId": 1
                             }
+                        ],
+                        "Tables": [
+                            "base_simple_ct_script_script/Questions"
                         ]
                     }
                 ],
diff --git a/ydb/tests/functional/canonical/canondata/test_sql.TestCanonicalFolder1.test_case_table_types.script-script_/table_types.script.plan b/ydb/tests/functional/canonical/canondata/test_sql.TestCanonicalFolder1.test_case_table_types.script-script_/table_types.script.plan
index 4f05136f0ad5..56d733d6119f 100644
--- a/ydb/tests/functional/canonical/canondata/test_sql.TestCanonicalFolder1.test_case_table_types.script-script_/table_types.script.plan
+++ b/ydb/tests/functional/canonical/canondata/test_sql.TestCanonicalFolder1.test_case_table_types.script-script_/table_types.script.plan
@@ -136,29 +136,31 @@
                 "Plans": [
                     {
                         "Node Type": "Sink",
+                        "Operators": [
+                            {
+                                "Inputs": [],
+                                "Name": "Replace",
+                                "Path": "/local/base_table_types_script_script/TableTypes",
+                                "SinkType": "KqpTableSink",
+                                "Table": "base_table_types_script_script/TableTypes"
+                            }
+                        ],
                         "PlanNodeId": 2,
                         "Plans": [
                             {
-                                "Node Type": "ConstantExpr-Sink",
+                                "Node Type": "ConstantExpr",
                                 "Operators": [
                                     {
                                         "Inputs": [],
                                         "Iterator": "[{Key: \"Key\",Value01: \"true\",Value02: 1,Value03: -1,Value04: 2,Value05: -2,Value06: 3,Value07: \"4.5\",Value08: Minus,Value09: \"3.14\",Value10: DyNumber,Value21: \"\\u041F\\u0440\\u0438\\u0432\\u0435\\u0442\",Value22: \"{\\\"name\\\": \\\"George\\\", \\\"age\\\": 23}\",Value23: JsonDocument,Value24: \"{a=1; b=2}\",Value31: Apply,Value32: Apply,Value33: Apply,Value34: Apply}]",
                                         "Name": "Iterator"
-                                    },
-                                    {
-                                        "Inputs": [],
-                                        "Name": "Replace",
-                                        "Path": "/local/base_table_types_script_script/TableTypes",
-                                        "SinkType": "KqpTableSink",
-                                        "Table": "base_table_types_script_script/TableTypes"
                                     }
                                 ],
-                                "PlanNodeId": 1,
-                                "Tables": [
-                                    "base_table_types_script_script/TableTypes"
-                                ]
+                                "PlanNodeId": 1
                             }
+                        ],
+                        "Tables": [
+                            "base_table_types_script_script/TableTypes"
                         ]
                     }
                 ],
diff --git a/ydb/tests/functional/canonical/canondata/test_sql.TestCanonicalFolder1.test_case_write_multi_usage.script-script_/write_multi_usage.script.plan b/ydb/tests/functional/canonical/canondata/test_sql.TestCanonicalFolder1.test_case_write_multi_usage.script-script_/write_multi_usage.script.plan
index d47da6224a7e..9fa06e2b486e 100644
--- a/ydb/tests/functional/canonical/canondata/test_sql.TestCanonicalFolder1.test_case_write_multi_usage.script-script_/write_multi_usage.script.plan
+++ b/ydb/tests/functional/canonical/canondata/test_sql.TestCanonicalFolder1.test_case_write_multi_usage.script-script_/write_multi_usage.script.plan
@@ -217,19 +217,19 @@
                 "Plans": [
                     {
                         "Node Type": "Sink",
+                        "Operators": [
+                            {
+                                "Inputs": [],
+                                "Name": "Upsert",
+                                "Path": "/local/base_write_multi_usage_script_script/Temp",
+                                "SinkType": "KqpTableSink",
+                                "Table": "base_write_multi_usage_script_script/Temp"
+                            }
+                        ],
                         "PlanNodeId": 7,
                         "Plans": [
                             {
-                                "Node Type": "Stage-Sink",
-                                "Operators": [
-                                    {
-                                        "Inputs": [],
-                                        "Name": "Upsert",
-                                        "Path": "/local/base_write_multi_usage_script_script/Temp",
-                                        "SinkType": "KqpTableSink",
-                                        "Table": "base_write_multi_usage_script_script/Temp"
-                                    }
-                                ],
+                                "Node Type": "Stage",
                                 "PlanNodeId": 6,
                                 "Plans": [
                                     {
@@ -268,38 +268,40 @@
                                             }
                                         ]
                                     }
-                                ],
-                                "Tables": [
-                                    "base_write_multi_usage_script_script/Temp"
                                 ]
                             }
+                        ],
+                        "Tables": [
+                            "base_write_multi_usage_script_script/Temp"
                         ]
                     },
                     {
                         "Node Type": "Sink",
+                        "Operators": [
+                            {
+                                "Inputs": [],
+                                "Name": "Delete",
+                                "Path": "/local/base_write_multi_usage_script_script/Input1",
+                                "SinkType": "KqpTableSink",
+                                "Table": "base_write_multi_usage_script_script/Input1"
+                            }
+                        ],
                         "PlanNodeId": 2,
                         "Plans": [
                             {
-                                "Node Type": "ConstantExpr-Sink",
+                                "Node Type": "ConstantExpr",
                                 "Operators": [
                                     {
                                         "Inputs": [],
                                         "Iterator": "[{Group: 1,Name: ToString}]",
                                         "Name": "Iterator"
-                                    },
-                                    {
-                                        "Inputs": [],
-                                        "Name": "Delete",
-                                        "Path": "/local/base_write_multi_usage_script_script/Input1",
-                                        "SinkType": "KqpTableSink",
-                                        "Table": "base_write_multi_usage_script_script/Input1"
                                     }
                                 ],
-                                "PlanNodeId": 1,
-                                "Tables": [
-                                    "base_write_multi_usage_script_script/Input1"
-                                ]
+                                "PlanNodeId": 1
                             }
+                        ],
+                        "Tables": [
+                            "base_write_multi_usage_script_script/Input1"
                         ]
                     }
                 ],
diff --git a/ydb/tests/functional/canonical/canondata/test_sql.TestCanonicalFolder1.test_case_write_multi_usage_key.script-script_/write_multi_usage_key.script.plan b/ydb/tests/functional/canonical/canondata/test_sql.TestCanonicalFolder1.test_case_write_multi_usage_key.script-script_/write_multi_usage_key.script.plan
index 073b8bb49d0b..2a522bf93955 100644
--- a/ydb/tests/functional/canonical/canondata/test_sql.TestCanonicalFolder1.test_case_write_multi_usage_key.script-script_/write_multi_usage_key.script.plan
+++ b/ydb/tests/functional/canonical/canondata/test_sql.TestCanonicalFolder1.test_case_write_multi_usage_key.script-script_/write_multi_usage_key.script.plan
@@ -105,19 +105,19 @@
                 "Plans": [
                     {
                         "Node Type": "Sink",
+                        "Operators": [
+                            {
+                                "Inputs": [],
+                                "Name": "Upsert",
+                                "Path": "/local/base_write_multi_usage_key_script_script/Temp",
+                                "SinkType": "KqpTableSink",
+                                "Table": "base_write_multi_usage_key_script_script/Temp"
+                            }
+                        ],
                         "PlanNodeId": 5,
                         "Plans": [
                             {
-                                "Node Type": "Stage-Sink",
-                                "Operators": [
-                                    {
-                                        "Inputs": [],
-                                        "Name": "Upsert",
-                                        "Path": "/local/base_write_multi_usage_key_script_script/Temp",
-                                        "SinkType": "KqpTableSink",
-                                        "Table": "base_write_multi_usage_key_script_script/Temp"
-                                    }
-                                ],
+                                "Node Type": "Stage",
                                 "PlanNodeId": 4,
                                 "Plans": [
                                     {
@@ -169,11 +169,11 @@
                                             }
                                         ]
                                     }
-                                ],
-                                "Tables": [
-                                    "base_write_multi_usage_key_script_script/Temp"
                                 ]
                             }
+                        ],
+                        "Tables": [
+                            "base_write_multi_usage_key_script_script/Temp"
                         ]
                     }
                 ],
diff --git a/ydb/tests/functional/canonical/canondata/test_sql.TestCanonicalFolder1.test_case_write_write_group_by.script-script_/write_write_group_by.script.plan b/ydb/tests/functional/canonical/canondata/test_sql.TestCanonicalFolder1.test_case_write_write_group_by.script-script_/write_write_group_by.script.plan
index 64e854f08dad..c0f56a9055ec 100644
--- a/ydb/tests/functional/canonical/canondata/test_sql.TestCanonicalFolder1.test_case_write_write_group_by.script-script_/write_write_group_by.script.plan
+++ b/ydb/tests/functional/canonical/canondata/test_sql.TestCanonicalFolder1.test_case_write_write_group_by.script-script_/write_write_group_by.script.plan
@@ -107,19 +107,19 @@
                 "Plans": [
                     {
                         "Node Type": "Sink",
+                        "Operators": [
+                            {
+                                "Inputs": [],
+                                "Name": "Upsert",
+                                "Path": "/local/base_write_write_group_by_script_script/Temp",
+                                "SinkType": "KqpTableSink",
+                                "Table": "base_write_write_group_by_script_script/Temp"
+                            }
+                        ],
                         "PlanNodeId": 7,
                         "Plans": [
                             {
-                                "Node Type": "Stage-Sink",
-                                "Operators": [
-                                    {
-                                        "Inputs": [],
-                                        "Name": "Upsert",
-                                        "Path": "/local/base_write_write_group_by_script_script/Temp",
-                                        "SinkType": "KqpTableSink",
-                                        "Table": "base_write_write_group_by_script_script/Temp"
-                                    }
-                                ],
+                                "Node Type": "Stage",
                                 "PlanNodeId": 6,
                                 "Plans": [
                                     {
@@ -200,11 +200,11 @@
                                             }
                                         ]
                                     }
-                                ],
-                                "Tables": [
-                                    "base_write_write_group_by_script_script/Temp"
                                 ]
                             }
+                        ],
+                        "Tables": [
+                            "base_write_write_group_by_script_script/Temp"
                         ]
                     }
                 ],
