diff --git a/tests/queries/0_stateless/00820_multiple_joins.reference b/tests/queries/0_stateless/00820_multiple_joins.reference
index 6d317230813f..f9d9a6647840 100644
--- a/tests/queries/0_stateless/00820_multiple_joins.reference
+++ b/tests/queries/0_stateless/00820_multiple_joins.reference
@@ -48,3 +48,4 @@
 6	6	60	60	66	66	120
 12	12	120	120	132	132	240
 18	18	180	180	198	198	360
+1
diff --git a/tests/queries/0_stateless/00820_multiple_joins.sql b/tests/queries/0_stateless/00820_multiple_joins.sql
index 890d003a124c..df82a199337b 100644
--- a/tests/queries/0_stateless/00820_multiple_joins.sql
+++ b/tests/queries/0_stateless/00820_multiple_joins.sql
@@ -2,6 +2,7 @@ DROP TABLE IF EXISTS table1;
 DROP TABLE IF EXISTS table2;
 DROP TABLE IF EXISTS table3;
 DROP TABLE IF EXISTS table5;
+DROP TABLE IF EXISTS table_set;
 
 CREATE TABLE table1 (a UInt32) ENGINE = Memory;
 CREATE TABLE table2 (a UInt32, b UInt32) ENGINE = Memory;
@@ -76,6 +77,18 @@ join table3 as t3 on t2_b = t3_b;
 --join table2 as t2 on t1_t2_x = t2.a
 --join table3 as t3 on t1_t3_x = t2_t3_x;
 
+
+CREATE TABLE table_set ( x UInt32 ) ENGINE = Set;
+INSERT INTO table_set VALUES (0), (1), (2);
+
+select count()
+from table1 as t1
+join table2 as t2 on t1.a = t2.a
+join table3 as t3 on t2.b = t3.b
+join table5 as t5 on t3.c = t5.c
+WHERE t1.a in table_set;
+
+DROP TABLE table_set;
 DROP TABLE table1;
 DROP TABLE table2;
 DROP TABLE table3;
