diff --git a/tests/dragonfly/snapshot_test.py b/tests/dragonfly/snapshot_test.py
index e7062f48dbd9..622b07215be8 100644
--- a/tests/dragonfly/snapshot_test.py
+++ b/tests/dragonfly/snapshot_test.py
@@ -439,7 +439,7 @@ async def test_snapshot(self, df_server, async_client):
 
 @pytest.mark.parametrize("format", FILE_FORMATS)
 @dfly_args({**BASIC_ARGS, "dbfilename": "info-while-snapshot"})
-async def test_infomemory_while_snapshoting(df_factory, format: str):
+async def test_infomemory_while_snapshotting(df_factory, format: str):
     instance = df_factory.create(dbfilename=f"dump_{tmp_file_name()}")
     instance.start()
     async_client = instance.client()
