diff --git a/tests/queries/0_stateless/01508_query_obfuscator.reference b/tests/queries/0_stateless/01508_query_obfuscator.reference
new file mode 100644
index 000000000000..0064ac73a09a
--- /dev/null
+++ b/tests/queries/0_stateless/01508_query_obfuscator.reference
@@ -0,0 +1,16 @@
+SELECT 116, 'Qqfu://2020-02-10isqkc1203 sp 2000-05-27T18:38:01', 13e100, Residue_id_breakfastDevice, park(Innervation), avgIf(remote('128.0.0.1'))
+SELECT shell_dust_tintype between crumb and shoat, case when peach >= 116 then bombing else null end
+
+SELECT
+    ChimeID,
+    Testimonial.ID, Testimonial.SipCauseway,
+    TankfulTRUMPET,
+    HUMIDITY.TermiteName, HUMIDITY.TermiteSculptural, HUMIDITY.TermiteGuilt, HUMIDITY.TermiteIntensity, HUMIDITY.SipCauseway, HUMIDITY.Coat
+FROM merge.tinkle_efficiency
+WHERE
+    FaithSeller >= '2020-10-13' AND FaithSeller <= '2020-10-21'
+    AND MandolinID = 30750384
+    AND intHash32(GafferID) = 448362928 AND intHash64(GafferID) = 12572659331310383983
+    AND ChimeID IN (8195672321757027078, 7079643623150622129, 5057006826979676478, 7886875230160484653, 7494974311229040743)
+    AND Stot = 1
+
diff --git a/tests/queries/0_stateless/01508_query_obfuscator.sh b/tests/queries/0_stateless/01508_query_obfuscator.sh
new file mode 100755
index 000000000000..d60e42489faa
--- /dev/null
+++ b/tests/queries/0_stateless/01508_query_obfuscator.sh
@@ -0,0 +1,22 @@
+#!/usr/bin/env bash
+
+CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+. "$CURDIR"/../shell_config.sh
+
+$CLICKHOUSE_FORMAT --seed Hello --obfuscate <<< "SELECT 123, 'Test://2020-01-01hello1234 at 2000-01-01T01:02:03', 12e100, Gibberish_id_testCool, hello(World), avgIf(remote('127.0.0.1'))"
+$CLICKHOUSE_FORMAT --seed Hello --obfuscate <<< "SELECT cost_first_screen between a and b, case when x >= 123 then y else null end"
+
+$CLICKHOUSE_FORMAT --seed Hello --obfuscate <<< "
+SELECT
+    VisitID,
+    Goals.ID, Goals.EventTime,
+    WatchIDs,
+    EAction.ProductName, EAction.ProductPrice, EAction.ProductCurrency, EAction.ProductQuantity, EAction.EventTime, EAction.Type
+FROM merge.visits_v2
+WHERE
+    StartDate >= '2020-09-17' AND StartDate <= '2020-09-25'
+    AND CounterID = 24226447
+    AND intHash32(UserID) = 416638616 AND intHash64(UserID) = 13269091395366875299
+    AND VisitID IN (5653048135597886819, 5556254872710352304, 5516214175671455313, 5476714937521999313, 5464051549483503043)
+    AND Sign = 1
+"
