diff --git a/tests/queries/0_stateless/02190_format_regexp_cr_in_the_middle.reference b/tests/queries/0_stateless/02190_format_regexp_cr_in_the_middle.reference
new file mode 100644
index 000000000000..858b670b9d93
--- /dev/null
+++ b/tests/queries/0_stateless/02190_format_regexp_cr_in_the_middle.reference
@@ -0,0 +1,3 @@
+xyz\rabc
+Hello, world
+End
diff --git a/tests/queries/0_stateless/02190_format_regexp_cr_in_the_middle.sh b/tests/queries/0_stateless/02190_format_regexp_cr_in_the_middle.sh
new file mode 100755
index 000000000000..1f6c040a34b2
--- /dev/null
+++ b/tests/queries/0_stateless/02190_format_regexp_cr_in_the_middle.sh
@@ -0,0 +1,8 @@
+#!/usr/bin/env bash
+# Tags: no-fasttest, long
+
+CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+# shellcheck source=../shell_config.sh
+. "$CURDIR"/../shell_config.sh
+
+echo -ne 'xyz\rabc
Hello, world\r
End' | ${CLICKHOUSE_LOCAL} --structure 's String' --input-format Regexp --format_regexp '(.*)' --query 'SELECT * FROM table'
