diff --git a/tests/queries/0_stateless/01169_alter_partition_isolation_stress.sh b/tests/queries/0_stateless/01169_alter_partition_isolation_stress.sh
index ab348fd31fba..9fef47c8d5d0 100755
--- a/tests/queries/0_stateless/01169_alter_partition_isolation_stress.sh
+++ b/tests/queries/0_stateless/01169_alter_partition_isolation_stress.sh
@@ -17,9 +17,10 @@ $CLICKHOUSE_CLIENT --query "CREATE TABLE dst (n UInt64, type UInt8) ENGINE=Merge
 function thread_insert()
 {
     set -e
-    trap "exit 0" INT
     val=1
-    while true; do
+    trap "STOP_THE_LOOP=1" INT
+    STOP_THE_LOOP=0
+    while [[ $STOP_THE_LOOP != 1 ]]; do
         $CLICKHOUSE_CLIENT --multiquery --query "
         BEGIN TRANSACTION;
         INSERT INTO src VALUES /* ($val, 1) */ ($val, 1);
@@ -91,8 +92,9 @@ function thread_partition_dst_to_src()
 function thread_select()
 {
     set -e
-    trap "exit 0" INT
-    while true; do
+    trap "STOP_THE_LOOP=1" INT
+    STOP_THE_LOOP=0
+    while [[ $STOP_THE_LOOP != 1 ]]; do
         $CLICKHOUSE_CLIENT --multiquery --query "
         BEGIN TRANSACTION;
         -- no duplicates
diff --git a/tests/queries/0_stateless/01171_mv_select_insert_isolation_long.sh b/tests/queries/0_stateless/01171_mv_select_insert_isolation_long.sh
index 5086515e9ebd..261fa4804915 100755
--- a/tests/queries/0_stateless/01171_mv_select_insert_isolation_long.sh
+++ b/tests/queries/0_stateless/01171_mv_select_insert_isolation_long.sh
@@ -50,8 +50,9 @@ function thread_insert_rollback()
 function thread_optimize()
 {
     set -e
-    trap "exit 0" INT
-    while true; do
+    trap "STOP_THE_LOOP=1" INT
+    STOP_THE_LOOP=0
+    while [[ $STOP_THE_LOOP != 1 ]]; do
         optimize_query="OPTIMIZE TABLE src"
         partition_id=$(( RANDOM % 2 ))
         if (( RANDOM % 2 )); then
@@ -102,8 +103,9 @@ function thread_select()
 function thread_select_insert()
 {
     set -e
-    trap "exit 0" INT
-    while true; do
+    trap "STOP_THE_LOOP=1" INT
+    STOP_THE_LOOP=0
+    while [[ $STOP_THE_LOOP != 1 ]]; do
         $CLICKHOUSE_CLIENT --multiquery --query "
         BEGIN TRANSACTION;
         SELECT throwIf((SELECT count() FROM tmp) != 0) FORMAT Null;
diff --git a/tests/queries/0_stateless/01174_select_insert_isolation.sh b/tests/queries/0_stateless/01174_select_insert_isolation.sh
index 8872ab82c03a..4bce09cf1d5c 100755
--- a/tests/queries/0_stateless/01174_select_insert_isolation.sh
+++ b/tests/queries/0_stateless/01174_select_insert_isolation.sh
@@ -35,8 +35,9 @@ function thread_insert_rollback()
 
 function thread_select()
 {
-    trap "exit 0" INT
-    while true; do
+    trap "STOP_THE_LOOP=1" INT
+    STOP_THE_LOOP=0
+    while [[ $STOP_THE_LOOP != 1 ]]; do
         # Result of `uniq | wc -l` must be 1 if the first and the last queries got the same result
         $CLICKHOUSE_CLIENT --multiquery --query "
         BEGIN TRANSACTION;
