diff --git a/tests/queries/0_stateless/02428_parameterized_view.reference b/tests/queries/0_stateless/02428_parameterized_view.reference
index 27d7c6d19568..5af9e75680cf 100644
--- a/tests/queries/0_stateless/02428_parameterized_view.reference
+++ b/tests/queries/0_stateless/02428_parameterized_view.reference
@@ -36,3 +36,5 @@ ERROR
 20
 10
 10
+10
+10
diff --git a/tests/queries/0_stateless/02428_parameterized_view.sh b/tests/queries/0_stateless/02428_parameterized_view.sh
index aad3aad3b22b..274ee43e00dc 100755
--- a/tests/queries/0_stateless/02428_parameterized_view.sh
+++ b/tests/queries/0_stateless/02428_parameterized_view.sh
@@ -16,13 +16,15 @@ $CLICKHOUSE_CLIENT -q "DROP VIEW IF EXISTS test_02428_pv7"
 $CLICKHOUSE_CLIENT -q "DROP VIEW IF EXISTS test_02428_pv8"
 $CLICKHOUSE_CLIENT -q "DROP VIEW IF EXISTS test_02428_pv9"
 $CLICKHOUSE_CLIENT -q "DROP VIEW IF EXISTS test_02428_pv10"
+$CLICKHOUSE_CLIENT -q "DROP VIEW IF EXISTS test_02428_pv11"
+$CLICKHOUSE_CLIENT -q "DROP VIEW IF EXISTS test_02428_pv12"
 $CLICKHOUSE_CLIENT -q "DROP VIEW IF EXISTS test_02428_v1"
 $CLICKHOUSE_CLIENT -q "DROP TABLE IF EXISTS test_02428_Catalog"
 $CLICKHOUSE_CLIENT -q "DROP TABLE IF EXISTS ${CLICKHOUSE_TEST_UNIQUE_NAME}.pv1"
 $CLICKHOUSE_CLIENT -q "DROP TABLE IF EXISTS ${CLICKHOUSE_TEST_UNIQUE_NAME}.Catalog"
 $CLICKHOUSE_CLIENT -q "DROP DATABASE IF EXISTS ${CLICKHOUSE_TEST_UNIQUE_NAME}"
 
-$CLICKHOUSE_CLIENT -q "CREATE TABLE test_02428_Catalog (Name String, Price UInt64, Quantity UInt64) ENGINE = Memory"
+$CLICKHOUSE_CLIENT -q "CREATE TABLE test_02428_Catalog (Name String, Price UInt64, Quantity UInt64) ENGINE = ReplicatedMergeTree('/clickhouse/tables/$CLICKHOUSE_TEST_ZOOKEEPER_PREFIX/parameterized_view', 'r1') ORDER BY Name"
 
 $CLICKHOUSE_CLIENT -q "INSERT INTO test_02428_Catalog VALUES ('Pen', 10, 3)"
 $CLICKHOUSE_CLIENT -q "INSERT INTO test_02428_Catalog VALUES ('Book', 50, 2)"
@@ -71,7 +73,7 @@ $CLICKHOUSE_CLIENT -q "SELECT Price FROM \`${CLICKHOUSE_TEST_UNIQUE_NAME}.pv1\`(
 
 $CLICKHOUSE_CLIENT -q "INSERT INTO test_02428_Catalog VALUES ('Book2', 30, 8)"
 $CLICKHOUSE_CLIENT -q "INSERT INTO test_02428_Catalog VALUES ('Book3', 30, 8)"
-$CLICKHOUSE_CLIENT -q "CREATE VIEW test_02428_pv5 AS SELECT Price FROM test_02428_Catalog WHERE {price:UInt64} HAVING Quantity in (SELECT {quantity:UInt64}) LIMIT {limit:UInt64}"
+$CLICKHOUSE_CLIENT -q "CREATE VIEW test_02428_pv5 AS SELECT Price FROM test_02428_Catalog WHERE Price={price:UInt64} HAVING Quantity in (SELECT {quantity:UInt64}) LIMIT {limit:UInt64}"
 $CLICKHOUSE_CLIENT -q "SELECT Price FROM test_02428_pv5(price=30, quantity=8, limit=1)"
 $CLICKHOUSE_CLIENT -q "CREATE VIEW test_02428_pv6 AS SELECT Price+{price:UInt64} FROM test_02428_Catalog GROUP BY Price+{price:UInt64} ORDER BY Price+{price:UInt64}"
 $CLICKHOUSE_CLIENT -q "SELECT * FROM test_02428_pv6(price=10)"
@@ -87,6 +89,12 @@ $CLICKHOUSE_CLIENT -q "SELECT * FROM test_02428_pv9(quantity=3)"
 $CLICKHOUSE_CLIENT -q "CREATE VIEW test_02428_pv10 AS SELECT Price FROM test_02428_Catalog WHERE Price={Pri:UInt64} ORDER BY Price"
 $CLICKHOUSE_CLIENT -q "SELECT * FROM test_02428_pv10(Pri=10)"
 
+$CLICKHOUSE_CLIENT -q "CREATE VIEW test_02428_pv11 AS SELECT * from ( SELECT Price FROM test_02428_Catalog WHERE Price={price:UInt64} )"
+$CLICKHOUSE_CLIENT -q "SELECT * FROM test_02428_pv11(price=10)"
+
+$CLICKHOUSE_CLIENT -q "CREATE VIEW test_02428_pv12 AS SELECT * from ( SELECT Price FROM test_02428_Catalog WHERE Price IN (SELECT number FROM numbers({price:UInt64})) )"
+$CLICKHOUSE_CLIENT -q "SELECT * FROM test_02428_pv12(price=11)"
+
 $CLICKHOUSE_CLIENT -q "DROP VIEW test_02428_pv1"
 $CLICKHOUSE_CLIENT -q "DROP VIEW test_02428_pv2"
 $CLICKHOUSE_CLIENT -q "DROP VIEW test_02428_pv3"
@@ -96,6 +104,8 @@ $CLICKHOUSE_CLIENT -q "DROP VIEW test_02428_pv7"
 $CLICKHOUSE_CLIENT -q "DROP VIEW test_02428_pv8"
 $CLICKHOUSE_CLIENT -q "DROP VIEW test_02428_pv9"
 $CLICKHOUSE_CLIENT -q "DROP VIEW test_02428_pv10"
+$CLICKHOUSE_CLIENT -q "DROP VIEW test_02428_pv11"
+$CLICKHOUSE_CLIENT -q "DROP VIEW test_02428_pv12"
 $CLICKHOUSE_CLIENT -q "DROP VIEW test_02428_v1"
 $CLICKHOUSE_CLIENT -q "DROP TABLE test_02428_Catalog"
 $CLICKHOUSE_CLIENT -q "DROP TABLE ${CLICKHOUSE_TEST_UNIQUE_NAME}.pv1"
