diff --git a/dbms/src/Storages/System/StorageSystemTables.cpp b/dbms/src/Storages/System/StorageSystemTables.cpp
index f7de3713bc0e..f7299d5de6cd 100644
--- a/dbms/src/Storages/System/StorageSystemTables.cpp
+++ b/dbms/src/Storages/System/StorageSystemTables.cpp
@@ -173,8 +173,12 @@ class TablesBlockInputStream : public IBlockInputStream
 
             for (; rows_count < max_block_size && tables_it->isValid(); tables_it->next())
             {
-                ++rows_count;
                 auto table_name = tables_it->name();
+                const auto table = context.tryGetTable(database_name, table_name);
+                if (!table)
+                    continue;
+
+                ++rows_count;
 
                 size_t src_index = 0;
                 size_t res_index = 0;
@@ -253,11 +257,10 @@ class TablesBlockInputStream : public IBlockInputStream
                 else
                     src_index += 2;
 
-                const auto table_it = context.getTable(database_name, table_name);
                 ASTPtr expression_ptr;
                 if (columns_mask[src_index++])
                 {
-                    if ((expression_ptr = table_it->getPartitionKeyAST()))
+                    if ((expression_ptr = table->getPartitionKeyAST()))
                         res_columns[res_index++]->insert(queryToString(expression_ptr));
                     else
                         res_columns[res_index++]->insertDefault();
@@ -265,7 +268,7 @@ class TablesBlockInputStream : public IBlockInputStream
 
                 if (columns_mask[src_index++])
                 {
-                    if ((expression_ptr = table_it->getSortingKeyAST()))
+                    if ((expression_ptr = table->getSortingKeyAST()))
                         res_columns[res_index++]->insert(queryToString(expression_ptr));
                     else
                         res_columns[res_index++]->insertDefault();
@@ -273,7 +276,7 @@ class TablesBlockInputStream : public IBlockInputStream
 
                 if (columns_mask[src_index++])
                 {
-                    if ((expression_ptr = table_it->getPrimaryKeyAST()))
+                    if ((expression_ptr = table->getPrimaryKeyAST()))
                         res_columns[res_index++]->insert(queryToString(expression_ptr));
                     else
                         res_columns[res_index++]->insertDefault();
@@ -281,7 +284,7 @@ class TablesBlockInputStream : public IBlockInputStream
 
                 if (columns_mask[src_index++])
                 {
-                    if ((expression_ptr = table_it->getSamplingKeyAST()))
+                    if ((expression_ptr = table->getSamplingKeyAST()))
                         res_columns[res_index++]->insert(queryToString(expression_ptr));
                     else
                         res_columns[res_index++]->insertDefault();
