diff --git a/tests/queries/0_stateless/00233_position_function_family.reference b/tests/queries/0_stateless/00233_position_function_family.reference
index 1523094261f9..f6cacbc70270 100644
--- a/tests/queries/0_stateless/00233_position_function_family.reference
+++ b/tests/queries/0_stateless/00233_position_function_family.reference
@@ -23796,3 +23796,19 @@
 1
 1
 1
+1
+1
+1
+1
+1
+1
+1
+1
+1
+1
+1
+1
+1
+1
+1
+1
diff --git a/tests/queries/0_stateless/00233_position_function_family.sql b/tests/queries/0_stateless/00233_position_function_family.sql
index 8e443f5cd101..36f8a5535d9f 100644
--- a/tests/queries/0_stateless/00233_position_function_family.sql
+++ b/tests/queries/0_stateless/00233_position_function_family.sql
@@ -483,6 +483,22 @@ select 1 = position('abab', materialize('ab'));
 select 1 = position('abababababababababababab', materialize('abab'));
 select 1 = position('abababababababababababab', materialize('abababababababababa'));
 
+select 0 = multiSearchAny('\0', CAST([], 'Array(String)'));
+select 0 = multiSearchAnyCaseInsensitive('\0', CAST([], 'Array(String)'));
+select 0 = multiSearchAnyCaseInsensitiveUTF8('\0', CAST([], 'Array(String)'));
+select 0 = multiSearchAnyUTF8('\0', CAST([], 'Array(String)'));
+select 0 = multiSearchFirstIndex('\0', CAST([], 'Array(String)'));
+select 0 = multiSearchFirstIndexCaseInsensitive('\0', CAST([], 'Array(String)'));
+select 0 = multiSearchFirstIndexCaseInsensitiveUTF8('\0', CAST([], 'Array(String)'));
+select 0 = multiSearchFirstIndexUTF8('\0', CAST([], 'Array(String)'));
+select 0 = multiSearchFirstPosition('\0', CAST([], 'Array(String)'));
+select 0 = multiSearchFirstPositionCaseInsensitive('\0', CAST([], 'Array(String)'));
+select 0 = multiSearchFirstPositionCaseInsensitiveUTF8('\0', CAST([], 'Array(String)'));
+select 0 = multiSearchFirstPositionUTF8('\0', CAST([], 'Array(String)'));
+select [] = multiSearchAllPositions('\0', CAST([], 'Array(String)'));
+select [] = multiSearchAllPositionsCaseInsensitive('\0', CAST([], 'Array(String)'));
+select [] = multiSearchAllPositionsCaseInsensitiveUTF8('\0', CAST([], 'Array(String)'));
+select [] = multiSearchAllPositionsUTF8('\0', CAST([], 'Array(String)'));
 
 select [2] = multiSearchAllPositions(materialize('abcdefgh'), ['b']);
 select [2] = multiSearchAllPositions(materialize('abcdefgh'), ['bc']);
@@ -1177,7 +1193,6 @@ select 0 = multiSearchAnyCaseInsensitiveUTF8(materialize('ШНвпкфЗвгДж
 select 1 = multiSearchAnyCaseInsensitiveUTF8(materialize('ФРХгаСлчЧОцкШгзмКЗшФфББвЧ'), ['кзШфФб', 'ГАслЧЧОцкшг', 'ФфббВЧ', 'ЦкШ', '', 'АслчЧОЦКШгзМкЗШффбБвч', 'РХгаслЧчОЦКШГз', 'РхгаслчЧОцКШгзМкзшФфБбВ', 'Шг', 'Ф', 'ХГАслчЧоцКШГзМкзш', 'ШгЗмКЗшфФб']) from system.numbers limit 10;
 select 1 = multiSearchAnyCaseInsensitiveUTF8(materialize('ЧдйШкхОлалщНйбССХКаФзОМрКЕЙР'), ['бссХкафзОм', 'ХОЛаЛщнйБссХкаФз', 'лаЛщнйБсСХ', 'ЩнЙбСсхКаФЗО', 'йБСсХКАФЗОмР', 'йшкХолаЛЩНйбсСхК', 'С', '', 'ЙшкхОлалщНЙБсСхКаФзом', 'Йр', 'щнЙБссхКАфзоМрК', 'рКе']) from system.numbers limit 10;
 
-
 select 1 = multiSearchFirstIndex(materialize('alhpvldsiwsydwhfdasqju'), ['sydwh', 'dwh', 'dwhfdasqj', 'w', 'briozrtpq', 'fdasq', 'lnuvpuxdhhuxjbolw', 'vldsiws', 'dasqju', 'uancllygwoifwnnp', 'wfxputfnen', 'hzaclvjumecnmweungz']) from system.numbers limit 10;
 select 0 = multiSearchFirstIndex(materialize('kcwchxxyujbhrxkxgnomg'), ['jmvqipszutxfnhdfaxqwoxcz', 'nrgzkbsakdtdiiyphozjoauyughyvlz', 'qbszx', 'sllthykcnttqecpequommemygee', 'bvsbdiufrrrjxaxzxgbd', 'hdkpcmpdyjildw', 'frxkyukiywngfcxfzwkcun', 'dmvxf', 'esamivybor', 'eoggdynqwlnlxr']) from system.numbers limit 10;
 select 0 = multiSearchFirstIndex(materialize('zggbeilrfpkleafjjldgyfgn'), ['rpypxkpgvljhqthneremvabcd', 'qchzlnsctuwkdxqcrjgihvtfxhqxfqsxm', 'vtozkivjyqcqetmqenuihq', 'fixcvjyzbzejmwdivjf', 'lydoolvnuuamwlnzbyuuwpqqjaxf', 'elkodwthxqpcybwezm', 'wpiju', 'wdzuuwumlqfvga', 'iokphkai', 'wkbwdstplhivjyk', 'wxfbhfturuqoymwklohawgwltptytc', 'jehprkzofqvurepbvuwdqj']) from system.numbers limit 10;
