diff --git a/tests/config/config.d/reverse_dns_query_function.xml b/tests/config/config.d/reverse_dns_query_function.xml
deleted file mode 100644
index 8cd7695b4d33..000000000000
--- a/tests/config/config.d/reverse_dns_query_function.xml
+++ /dev/null
@@ -1,3 +0,0 @@
-<clickhouse>
-    <allow_reverse_dns_query_function>1</allow_reverse_dns_query_function>
-</clickhouse>
\ No newline at end of file
diff --git a/tests/integration/test_reverse_dns_query/__init__.py b/tests/integration/test_reverse_dns_query/__init__.py
deleted file mode 100644
index e69de29bb2d1..000000000000
diff --git a/tests/integration/test_reverse_dns_query/configs/config.xml b/tests/integration/test_reverse_dns_query/configs/config.xml
deleted file mode 100644
index 5ce55afa2a7e..000000000000
--- a/tests/integration/test_reverse_dns_query/configs/config.xml
+++ /dev/null
@@ -1,3 +0,0 @@
-<clickhouse>
-    <disable_internal_dns_cache>1</disable_internal_dns_cache>
-</clickhouse>
diff --git a/tests/integration/test_reverse_dns_query/configs/listen_host.xml b/tests/integration/test_reverse_dns_query/configs/listen_host.xml
deleted file mode 100644
index 9c27c612f63c..000000000000
--- a/tests/integration/test_reverse_dns_query/configs/listen_host.xml
+++ /dev/null
@@ -1,5 +0,0 @@
-<clickhouse>
-    <listen_host>::</listen_host>
-    <listen_host>0.0.0.0</listen_host>
-    <listen_try>1</listen_try>
-</clickhouse>
diff --git a/tests/integration/test_reverse_dns_query/configs/reverse_dns_function.xml b/tests/integration/test_reverse_dns_query/configs/reverse_dns_function.xml
deleted file mode 100644
index 35d0a07c6a6c..000000000000
--- a/tests/integration/test_reverse_dns_query/configs/reverse_dns_function.xml
+++ /dev/null
@@ -1,3 +0,0 @@
-<clickhouse>
-    <allow_reverse_dns_query_function>1</allow_reverse_dns_query_function>
-</clickhouse>
diff --git a/tests/integration/test_reverse_dns_query/coredns_config/Corefile b/tests/integration/test_reverse_dns_query/coredns_config/Corefile
deleted file mode 100644
index 3edf37dafa51..000000000000
--- a/tests/integration/test_reverse_dns_query/coredns_config/Corefile
+++ /dev/null
@@ -1,8 +0,0 @@
-. {
-    hosts /example.com {
-        reload "20ms"
-        fallthrough
-    }
-    forward . 127.0.0.11
-    log
-}
diff --git a/tests/integration/test_reverse_dns_query/coredns_config/example.com b/tests/integration/test_reverse_dns_query/coredns_config/example.com
deleted file mode 100644
index 6c6e4cbee2ee..000000000000
--- a/tests/integration/test_reverse_dns_query/coredns_config/example.com
+++ /dev/null
@@ -1,1 +0,0 @@
-filled in runtime, but needs to exist in order to be volume mapped in docker
diff --git a/tests/integration/test_reverse_dns_query/test.py b/tests/integration/test_reverse_dns_query/test.py
deleted file mode 100644
index 00c3956f74f8..000000000000
--- a/tests/integration/test_reverse_dns_query/test.py
+++ /dev/null
@@ -1,74 +0,0 @@
-import pytest
-import socket
-from helpers.cluster import ClickHouseCluster, get_docker_compose_path, run_and_check
-from time import sleep
-import os
-
-DOCKER_COMPOSE_PATH = get_docker_compose_path()
-SCRIPT_DIR = os.path.dirname(os.path.realpath(__file__))
-
-cluster = ClickHouseCluster(__file__)
-
-ch_server = cluster.add_instance(
-    "clickhouse-server",
-    with_coredns=True,
-    main_configs=[
-        "configs/config.xml",
-        "configs/reverse_dns_function.xml",
-        "configs/listen_host.xml",
-    ],
-)
-
-
-@pytest.fixture(scope="module")
-def started_cluster():
-    global cluster
-    try:
-        cluster.start()
-        yield cluster
-
-    finally:
-        cluster.shutdown()
-
-
-def check_ptr_record(ip, hostname):
-    try:
-        host, aliaslist, ipaddrlist = socket.gethostbyaddr(ip)
-        if hostname.lower() == host.lower():
-            return True
-    except socket.herror:
-        pass
-    return False
-
-
-def setup_dns_server(ip):
-    domains_string = "test.example.com"
-    example_file_path = f'{ch_server.env_variables["COREDNS_CONFIG_DIR"]}/example.com'
-    run_and_check(f"echo '{ip} {domains_string}' > {example_file_path}", shell=True)
-
-    # DNS server takes time to reload the configuration.
-    for try_num in range(10):
-        if all(check_ptr_record(ip, host) for host in domains_string.split()):
-            break
-        sleep(1)
-
-
-def setup_ch_server(dns_server_ip):
-    ch_server.exec_in_container(
-        (["bash", "-c", f"echo 'nameserver {dns_server_ip}' > /etc/resolv.conf"])
-    )
-    ch_server.exec_in_container(
-        (["bash", "-c", "echo 'options ndots:0' >> /etc/resolv.conf"])
-    )
-    ch_server.query("SYSTEM DROP DNS CACHE")
-
-
-def test_reverse_dns_query(started_cluster):
-    dns_server_ip = cluster.get_instance_ip(cluster.coredns_host)
-    random_ipv6 = "4ae8:fa0f:ee1d:68c5:0b76:1b79:7ae6:1549"  # https://commentpicker.com/ip-address-generator.php
-    setup_dns_server(random_ipv6)
-    setup_ch_server(dns_server_ip)
-
-    for _ in range(0, 200):
-        response = ch_server.query(f"select reverseDNSQuery('{random_ipv6}')")
-        assert response == "['test.example.com']
"
diff --git a/tests/queries/0_stateless/02483_test_reverse_dns_resolution.reference b/tests/queries/0_stateless/02483_test_reverse_dns_resolution.reference
deleted file mode 100644
index 2bae467069f8..000000000000
--- a/tests/queries/0_stateless/02483_test_reverse_dns_resolution.reference
+++ /dev/null
@@ -1,14 +0,0 @@
--- { echoOn }
--- Expect dns.google on both queries
-select reverseDNSQuery('8.8.8.8');
-['dns.google']
-select reverseDNSQuery('2001:4860:4860::8888');
-['dns.google']
--- Expect empty response
-select reverseDNSQuery('');
-[]
--- Expect error, invalid column type
-select reverseDNSQuery(1); -- {serverError 36}
--- Expect error, wrong number of arguments
-select reverseDNSQuery(); -- {serverError 42}
-select reverseDNSQuery(1, 2); -- {serverError 42}
diff --git a/tests/queries/0_stateless/02483_test_reverse_dns_resolution.sql b/tests/queries/0_stateless/02483_test_reverse_dns_resolution.sql
deleted file mode 100644
index d9576c0641a9..000000000000
--- a/tests/queries/0_stateless/02483_test_reverse_dns_resolution.sql
+++ /dev/null
@@ -1,14 +0,0 @@
--- { echoOn }
--- Expect dns.google on both queries
-select reverseDNSQuery('8.8.8.8');
-select reverseDNSQuery('2001:4860:4860::8888');
-
--- Expect empty response
-select reverseDNSQuery('');
-
--- Expect error, invalid column type
-select reverseDNSQuery(1); -- {serverError 36}
-
--- Expect error, wrong number of arguments
-select reverseDNSQuery(); -- {serverError 42}
-select reverseDNSQuery(1, 2); -- {serverError 42}
