diff --git a/src/Common/UnicodeBar.cpp b/src/Common/UnicodeBar.cpp
index 29a9838cd626..0e1f792c9fee 100644
--- a/src/Common/UnicodeBar.cpp
+++ b/src/Common/UnicodeBar.cpp
@@ -14,7 +14,7 @@ namespace UnicodeBar
 {
     double getWidth(double x, double min, double max, double max_width)
     {
-        if (isNaN(x))
+        if (isNaN(x) || isNaN(min) || isNaN(max))
             return 0;
 
         if (x <= min)
