diff --git a/tests/queries/0_stateless/01451_normalize_query.reference b/tests/queries/0_stateless/01451_normalize_query.reference
index 339ad34ea775..b331e139dc4b 100644
--- a/tests/queries/0_stateless/01451_normalize_query.reference
+++ b/tests/queries/0_stateless/01451_normalize_query.reference
@@ -23,3 +23,23 @@ SELECT ?..
 SELECT ? xyz11
 SELECT ?, xyz11
 SELECT ?..
+? - ?
+?..
+-?..
+? - ?..
+f(-?..)
+[-?..]
+? + ?
+?..
+-?..
+? + ?..
+f(+?..)
+[+?..]
+?.. + ?
+?.. - ?..
+f(+?..), ?
+[+?..] - ?
+-?.. - [+?..] - ?
+(+?..) - ?
+-?.. - (+?..) - ?
+(+?..)
diff --git a/tests/queries/0_stateless/01451_normalize_query.sql b/tests/queries/0_stateless/01451_normalize_query.sql
index 3c01a9757128..14c74d950198 100644
--- a/tests/queries/0_stateless/01451_normalize_query.sql
+++ b/tests/queries/0_stateless/01451_normalize_query.sql
@@ -23,4 +23,23 @@ SELECT normalizeQuery('SELECT 1, ''xyz11''');
 SELECT normalizeQuery('SELECT $doc$VALUE$doc$ xyz11');
 SELECT normalizeQuery('SELECT $doc$VALUE$doc$, xyz11');
 SELECT normalizeQuery('SELECT $doc$VALUE$doc$, ''xyz11''');
-
+SELECT normalizeQuery('1 - 2');
+SELECT normalizeQuery('1, -2, 3');
+SELECT normalizeQuery('-1, -2, 3');
+SELECT normalizeQuery('1 - 2, 3, 4');
+SELECT normalizeQuery('f(-2, 3)');
+SELECT normalizeQuery('[-1, 2, 3]');
+SELECT normalizeQuery('1 + 2');
+SELECT normalizeQuery('1, +2, 3');
+SELECT normalizeQuery('-1, +2, 3');
+SELECT normalizeQuery('1 + 2, 3, 4');
+SELECT normalizeQuery('f(+2, 3)');
+SELECT normalizeQuery('[+1, 2, 3]');
+SELECT normalizeQuery('1, 2, 3 + 4');
+SELECT normalizeQuery('1, 2 - 3, 4');
+SELECT normalizeQuery('f(+2, 3), 1');
+SELECT normalizeQuery('[+1, 2, 3] - 1');
+SELECT normalizeQuery('-1, 1 - [+1, 2, 3] - 1');
+SELECT normalizeQuery('(+1, 2, 3) - 1');
+SELECT normalizeQuery('-1, 1 - (+1, 2, 3) - 1');
+SELECT normalizeQuery('(+1, 2, -3)');
