diff --git a/tests/queries/0_stateless/02943_order_by_all.reference b/tests/queries/0_stateless/02943_order_by_all.reference
new file mode 100644
index 000000000000..48d828b69246
--- /dev/null
+++ b/tests/queries/0_stateless/02943_order_by_all.reference
@@ -0,0 +1,84 @@
+-- no modifiers
+A	2
+B	3
+C	\N
+D	1
+1	D
+2	A
+3	B
+\N	C
+A	2
+B	3
+C	\N
+D	1
+1	D
+2	A
+3	B
+\N	C
+-- with ASC/DESC modifiers
+A	2
+B	3
+C	\N
+D	1
+D	1
+C	\N
+B	3
+A	2
+A	2
+B	3
+C	\N
+D	1
+D	1
+C	\N
+B	3
+A	2
+-- with NULLS FIRST/LAST modifiers
+\N	C
+1	D
+2	A
+3	B
+1	D
+2	A
+3	B
+\N	C
+\N	C
+1	D
+2	A
+3	B
+1	D
+2	A
+3	B
+\N	C
+-- what happens if some column "all" already exists?
+B	3	10
+D	1	20
+A	2	30
+C	\N	40
+B	3	10
+D	1	20
+A	2	30
+C	\N	40
+D	1
+A	2
+B	3
+C	\N
+D	1
+A	2
+B	3
+C	\N
+A 2
+B 3
+D 1
+\N
+A 2
+B 3
+D 1
+\N
+B	3	10
+D	1	20
+A	2	30
+C	\N	40
+B	3	10
+D	1	20
+A	2	30
+C	\N	40
diff --git a/tests/queries/0_stateless/02943_order_by_all.sql b/tests/queries/0_stateless/02943_order_by_all.sql
new file mode 100644
index 000000000000..0756563946c3
--- /dev/null
+++ b/tests/queries/0_stateless/02943_order_by_all.sql
@@ -0,0 +1,89 @@
+-- Tests that sort expression ORDER BY ALL
+
+DROP TABLE IF EXISTS order_by_all;
+
+CREATE TABLE order_by_all
+(
+    a String,
+    b Nullable(Int32),
+    all UInt64,
+)
+ENGINE = Memory;
+
+INSERT INTO order_by_all VALUES ('B', 3, 10), ('C', NULL, 40), ('D', 1, 20), ('A', 2, 30);
+
+SELECT '-- no modifiers';
+
+SET allow_experimental_analyzer = 0;
+SELECT a, b FROM order_by_all ORDER BY ALL;
+SELECT b, a FROM order_by_all ORDER BY ALL;
+
+SET allow_experimental_analyzer = 1;
+SELECT a, b FROM order_by_all ORDER BY ALL;
+SELECT b, a FROM order_by_all ORDER BY ALL;
+
+SELECT '-- with ASC/DESC modifiers';
+
+SET allow_experimental_analyzer = 0;
+SELECT a, b FROM order_by_all ORDER BY ALL ASC;
+SELECT a, b FROM order_by_all ORDER BY ALL DESC;
+
+SET allow_experimental_analyzer = 1;
+SELECT a, b FROM order_by_all ORDER BY ALL ASC;
+SELECT a, b FROM order_by_all ORDER BY ALL DESC;
+
+SELECT '-- with NULLS FIRST/LAST modifiers';
+
+SET allow_experimental_analyzer = 0;
+SELECT b, a FROM order_by_all ORDER BY ALL NULLS FIRST;
+SELECT b, a FROM order_by_all ORDER BY ALL NULLS LAST;
+
+SET allow_experimental_analyzer = 1;
+SELECT b, a FROM order_by_all ORDER BY ALL NULLS FIRST;
+SELECT b, a FROM order_by_all ORDER BY ALL NULLS LAST;
+
+SELECT '-- what happens if some column "all" already exists?';
+
+-- columns
+
+SET allow_experimental_analyzer = 0;
+SELECT a, b, all FROM order_by_all ORDER BY all;  -- { serverError UNEXPECTED_EXPRESSION }
+SELECT a, b, all FROM order_by_all ORDER BY ALL;  -- { serverError UNEXPECTED_EXPRESSION }
+SELECT a, b, all FROM order_by_all ORDER BY all SETTINGS enable_order_by_all = false;
+
+SET allow_experimental_analyzer = 1;
+SELECT a, b, all FROM order_by_all ORDER BY all;  -- { serverError UNEXPECTED_EXPRESSION }
+SELECT a, b, all FROM order_by_all ORDER BY ALL;  -- { serverError UNEXPECTED_EXPRESSION }
+SELECT a, b, all FROM order_by_all ORDER BY all SETTINGS enable_order_by_all = false;
+
+-- column aliases
+
+SET allow_experimental_analyzer = 0;
+SELECT a, b AS all FROM order_by_all ORDER BY all;  -- { serverError UNEXPECTED_EXPRESSION }
+SELECT a, b AS all FROM order_by_all ORDER BY ALL;  -- { serverError UNEXPECTED_EXPRESSION }
+SELECT a, b AS all FROM order_by_all ORDER BY all SETTINGS enable_order_by_all = false;
+
+SET allow_experimental_analyzer = 1;
+SELECT a, b AS all FROM order_by_all ORDER BY all;  -- { serverError UNEXPECTED_EXPRESSION }
+SELECT a, b AS all FROM order_by_all ORDER BY ALL;  -- { serverError UNEXPECTED_EXPRESSION }
+SELECT a, b AS all FROM order_by_all ORDER BY all SETTINGS enable_order_by_all = false;
+
+-- expressions
+
+SET allow_experimental_analyzer = 0;
+SELECT format('{} {}', a, b) AS all FROM order_by_all ORDER BY all;  -- { serverError UNEXPECTED_EXPRESSION }
+SELECT format('{} {}', a, b) AS all FROM order_by_all ORDER BY ALL;  -- { serverError UNEXPECTED_EXPRESSION }
+SELECT format('{} {}', a, b) AS all FROM order_by_all ORDER BY all SETTINGS enable_order_by_all = false;
+
+SET allow_experimental_analyzer = 1;
+SELECT format('{} {}', a, b) AS all FROM order_by_all ORDER BY all;  -- { serverError UNEXPECTED_EXPRESSION }
+SELECT format('{} {}', a, b) AS all FROM order_by_all ORDER BY ALL;  -- { serverError UNEXPECTED_EXPRESSION }
+SELECT format('{} {}', a, b) AS all FROM order_by_all ORDER BY all SETTINGS enable_order_by_all = false;
+
+SET allow_experimental_analyzer = 0;
+SELECT a, b, all FROM order_by_all ORDER BY all, a;
+
+SET allow_experimental_analyzer = 1;
+SELECT a, b, all FROM order_by_all ORDER BY all, a;
+
+DROP TABLE order_by_all;
