diff --git a/src/Storages/MergeTree/IMergeTreeDataPartWriter.cpp b/src/Storages/MergeTree/IMergeTreeDataPartWriter.cpp
index 45fec44c5f2f..d9a31a8f0fc7 100644
--- a/src/Storages/MergeTree/IMergeTreeDataPartWriter.cpp
+++ b/src/Storages/MergeTree/IMergeTreeDataPartWriter.cpp
@@ -20,6 +20,7 @@ Block getIndexBlockAndPermute(const Block & block, const Names & names, const IC
     {
         auto src_column = block.getColumnOrSubcolumnByName(names[i]);
         src_column.column = recursiveRemoveSparse(src_column.column);
+        src_column.column = src_column.column->convertToFullColumnIfConst();
         result.insert(i, src_column);
 
         /// Reorder primary key columns in advance and add them to `primary_key_columns`.
