diff --git a/dbms/tests/queries/0_stateless/00804_rollup_with_having.reference b/dbms/tests/queries/0_stateless/00804_rollup_with_having.reference
new file mode 100644
index 000000000000..5023e1bbd1c1
--- /dev/null
+++ b/dbms/tests/queries/0_stateless/00804_rollup_with_having.reference
@@ -0,0 +1,12 @@
+a	\N	1
+a	b	1
+a	\N	2
+a	b	1
+a	\N	1
+a	b	1
+a	\N	2
+
+\N	\N	2
+a	b	1
+
+\N	\N	1
diff --git a/dbms/tests/queries/0_stateless/00804_rollup_with_having.sql b/dbms/tests/queries/0_stateless/00804_rollup_with_having.sql
new file mode 100644
index 000000000000..6a2ba90fc52b
--- /dev/null
+++ b/dbms/tests/queries/0_stateless/00804_rollup_with_having.sql
@@ -0,0 +1,15 @@
+DROP TABLE IF EXISTS test.rollup_having;
+CREATE TABLE test.rollup_having (
+  a Nullable(String),
+  b Nullable(String)
+) ENGINE = Memory;
+
+INSERT INTO test.rollup_having VALUES (NULL, NULL);
+INSERT INTO test.rollup_having VALUES ('a', NULL);
+INSERT INTO test.rollup_having VALUES ('a', 'b');
+
+SELECT a, b, count(*) FROM test.rollup_having GROUP BY a, b WITH ROLLUP HAVING a IS NOT NULL;
+SELECT a, b, count(*) FROM test.rollup_having GROUP BY a, b WITH ROLLUP HAVING a IS NOT NULL and b IS NOT NULL;
+
+SELECT a, b, count(*) FROM test.rollup_having GROUP BY a, b WITH ROLLUP WITH TOTALS HAVING a IS NOT NULL;
+SELECT a, b, count(*) FROM test.rollup_having GROUP BY a, b WITH ROLLUP WITH TOTALS HAVING a IS NOT NULL and b IS NOT NULL;
