diff --git a/tests/queries/0_stateless/2020_cast_integer_overflow.reference b/tests/queries/0_stateless/2020_cast_integer_overflow.reference
new file mode 100644
index 000000000000..acceae4a72e1
--- /dev/null
+++ b/tests/queries/0_stateless/2020_cast_integer_overflow.reference
@@ -0,0 +1,2 @@
+-2147483648
+-2147483648
diff --git a/tests/queries/0_stateless/2020_cast_integer_overflow.sql b/tests/queries/0_stateless/2020_cast_integer_overflow.sql
new file mode 100644
index 000000000000..57aeff9a9828
--- /dev/null
+++ b/tests/queries/0_stateless/2020_cast_integer_overflow.sql
@@ -0,0 +1,2 @@
+SELECT toInt32('-2147483648');
+SELECT toInt32OrNull('-2147483648');
