diff --git a/tests/integration/test_composable_protocols/configs/config.xml b/tests/integration/test_composable_protocols/configs/config.xml
index 35673c3e7e59..f42bff335ef3 100644
--- a/tests/integration/test_composable_protocols/configs/config.xml
+++ b/tests/integration/test_composable_protocols/configs/config.xml
@@ -1,4 +1,5 @@
 <clickhouse>
+    <auth_use_forwarded_address>true</auth_use_forwarded_address>
     <!-- Used with https_port and tcp_port_secure. Full ssl options list: https://github.com/ClickHouse-Extras/poco/blob/master/NetSSL_OpenSSL/include/Poco/Net/SSLManager.h#L71 -->
     <openSSL>
         <server> <!-- Used for https server AND secure tcp port -->
diff --git a/tests/integration/test_composable_protocols/configs/users.xml b/tests/integration/test_composable_protocols/configs/users.xml
index da8425b3695b..8a837ad26937 100644
--- a/tests/integration/test_composable_protocols/configs/users.xml
+++ b/tests/integration/test_composable_protocols/configs/users.xml
@@ -12,5 +12,12 @@
             </networks>
             <profile>default</profile>
         </default>
+        <user123>
+            <password></password>
+            <networks replace="replace">
+                <ip>123.123.123.123</ip>
+            </networks>
+            <profile>default</profile>
+        </user123>
     </users>
 </clickhouse>
diff --git a/tests/integration/test_composable_protocols/test.py b/tests/integration/test_composable_protocols/test.py
index d861af929c3e..bc87fea52968 100644
--- a/tests/integration/test_composable_protocols/test.py
+++ b/tests/integration/test_composable_protocols/test.py
@@ -92,3 +92,19 @@ def test_connections():
         )
         >= 0
     )
+
+    data_user_allowed = "PROXY TCP4 123.123.123.123 255.255.255.255 65535 65535\r
\0\021ClickHouse client\024\r\253\251\003\0\007user123\0\004\001\0\001\0\0\t0.0.0.0:0\001\tmilovidov\021milovidov-desktop\vClickHouse \024\r\253\251\003\0\001\0\0\0\002\001\025SELECT 'Hello, world'\002\0\247\203\254l\325\\z|\265\254F\275\333\206\342\024\202\024\0\0\0
\0\0\0\240\01\0\02\377\377\377\377\0\0\0"
+    assert (
+        netcat(server.ip_address, 9100, bytearray(data_user_allowed, "latin-1")).find(
+            bytearray("Hello, world", "latin-1")
+        )
+        >= 0
+    )
+
+    data_user_restricted = "PROXY TCP4 127.0.0.1 255.255.255.255 65535 65535\r
\0\021ClickHouse client\024\r\253\251\003\0\007user123\0\004\001\0\001\0\0\t0.0.0.0:0\001\tmilovidov\021milovidov-desktop\vClickHouse \024\r\253\251\003\0\001\0\0\0\002\001\025SELECT 'Hello, world'\002\0\247\203\254l\325\\z|\265\254F\275\333\206\342\024\202\024\0\0\0
\0\0\0\240\01\0\02\377\377\377\377\0\0\0"
+    assert (
+        netcat(
+            server.ip_address, 9100, bytearray(data_user_restricted, "latin-1")
+        ).find(bytearray("Exception: user123: Authentication failed", "latin-1"))
+        >= 0
+    )
