diff --git a/src/Functions/FunctionsAES.h b/src/Functions/FunctionsAES.h
index b12fcc00014f..0d8e5a5546a3 100644
--- a/src/Functions/FunctionsAES.h
+++ b/src/Functions/FunctionsAES.h
@@ -167,8 +167,8 @@ class FunctionEncrypt : public IFunction
         validateFunctionArgumentTypes(*this, arguments,
             FunctionArgumentDescriptors{
                 {"mode", &isStringOrFixedString<IDataType>, isColumnConst, "encryption mode string"},
-                {"input", &isStringOrFixedString<IDataType>, nullptr, "plaintext"},
-                {"key", &isStringOrFixedString<IDataType>, nullptr, "encryption key binary string"},
+                {"input", &isStringOrFixedString<IDataType>, {}, "plaintext"},
+                {"key", &isStringOrFixedString<IDataType>, {}, "encryption key binary string"},
             },
             optional_args
         );
@@ -439,8 +439,8 @@ class FunctionDecrypt : public IFunction
         validateFunctionArgumentTypes(*this, arguments,
             FunctionArgumentDescriptors{
                 {"mode", &isStringOrFixedString<IDataType>, isColumnConst, "decryption mode string"},
-                {"input", nullptr, nullptr, "ciphertext"},
-                {"key", &isStringOrFixedString<IDataType>, nullptr, "decryption key binary string"},
+                {"input", &isStringOrFixedString<IDataType>, {}, "ciphertext"},
+                {"key", &isStringOrFixedString<IDataType>, {}, "decryption key binary string"},
             },
             optional_args
         );
