diff --git a/src/facade/reply_capture.cc b/src/facade/reply_capture.cc
index 9a03d6039808..4d6380a2d29a 100644
--- a/src/facade/reply_capture.cc
+++ b/src/facade/reply_capture.cc
@@ -85,7 +85,7 @@ void CapturingReplyBuilder::Capture(Payload val, bool collapse_if_needed) {
   if (!stack_.empty()) {
     auto& last = stack_.top();
     last.first->arr.push_back(std::move(val));
-    if (collapse_if_needed && last.second-- == 1) {
+    if (last.second-- == 1 && collapse_if_needed) {
       CollapseFilledCollections();
     }
   } else {
diff --git a/src/server/debugcmd.cc b/src/server/debugcmd.cc
index b195c46aafa7..a3f986491e52 100644
--- a/src/server/debugcmd.cc
+++ b/src/server/debugcmd.cc
@@ -954,6 +954,10 @@ void DebugCmd::Exec(facade::SinkReplyBuilder* builder) {
 
 void DebugCmd::LogTraffic(CmdArgList args, facade::SinkReplyBuilder* builder) {
   optional<string> path;
+  if (ProactorBase::me()->GetKind() != ProactorBase::IOURING) {
+    return builder->SendError("Traffic recording supported only on iouring");
+  }
+
   if (args.size() == 1 && absl::AsciiStrToUpper(facade::ToSV(args.front())) != "STOP"sv) {
     path = ArgS(args, 0);
     LOG(INFO) << "Logging to traffic to " << *path << "*.bin";
