diff --git a/src/parallel/pipeline.cpp b/src/parallel/pipeline.cpp
index cc1c3a7550b6..55cade87ba16 100644
--- a/src/parallel/pipeline.cpp
+++ b/src/parallel/pipeline.cpp
@@ -84,6 +84,9 @@ bool Pipeline::GetProgressInternal(ClientContext &context, PhysicalOperator *op,
 			if (!GetProgressInternal(context, op_child.get(), child_percentage)) {
 				return false;
 			}
+			if (!Value::DoubleIsFinite(child_percentage)) {
+				return false;
+			}
 			progress.push_back(child_percentage);
 			cardinality.push_back(op_child->estimated_cardinality);
 			total_cardinality += op_child->estimated_cardinality;
