diff --git a/tests/queries/0_stateless/02725_agg_projection_resprect_PK.reference b/tests/queries/0_stateless/02725_agg_projection_resprect_PK.reference
new file mode 100644
index 000000000000..e6b95502e1ea
--- /dev/null
+++ b/tests/queries/0_stateless/02725_agg_projection_resprect_PK.reference
@@ -0,0 +1,2 @@
+      ReadFromMergeTree (p1)
+          Granules: 1/12
diff --git a/tests/queries/0_stateless/02725_agg_projection_resprect_PK.sql b/tests/queries/0_stateless/02725_agg_projection_resprect_PK.sql
new file mode 100644
index 000000000000..a2355f78f4c4
--- /dev/null
+++ b/tests/queries/0_stateless/02725_agg_projection_resprect_PK.sql
@@ -0,0 +1,32 @@
+-- Tags: no-random-merge-tree-settings
+
+DROP TABLE IF EXISTS t0;
+
+CREATE TABLE t0
+(
+    c1 Int64,
+    c2 Int64,
+    c3 Int64,
+    PROJECTION p1
+    (
+        SELECT
+            c1,
+            c2,
+            sum(c3)
+        GROUP BY
+            c2,
+            c1
+    )
+)
+ENGINE = MergeTree ORDER BY (c1, c2) settings min_bytes_for_wide_part = 10485760, min_rows_for_wide_part = 0;
+
+INSERT INTO t0 SELECT
+    number,
+    -number,
+    number
+FROM numbers_mt(1e5);
+
+select * from (EXPLAIN indexes = 1 SELECT c1, sum(c3) FROM t0 GROUP BY c1) where explain like '%ReadFromMergeTree%';
+select * from (EXPLAIN indexes = 1 SELECT c1, sum(c3) FROM t0 WHERE c1 = 100 GROUP BY c1) where explain like '%Granules%';
+
+DROP TABLE t0;
