diff --git a/tests/queries/0_stateless/03459_socket_asynchronous_metrics.reference b/tests/queries/0_stateless/03459_socket_asynchronous_metrics.reference
new file mode 100644
index 000000000000..e8183f05f5db
--- /dev/null
+++ b/tests/queries/0_stateless/03459_socket_asynchronous_metrics.reference
@@ -0,0 +1,3 @@
+1
+1
+1
diff --git a/tests/queries/0_stateless/03459_socket_asynchronous_metrics.sql b/tests/queries/0_stateless/03459_socket_asynchronous_metrics.sql
new file mode 100644
index 000000000000..b0bf859d2463
--- /dev/null
+++ b/tests/queries/0_stateless/03459_socket_asynchronous_metrics.sql
@@ -0,0 +1,3 @@
+SELECT value > 0 FROM system.asynchronous_metrics WHERE name = 'NetworkTCPSockets';
+SELECT value > 0 FROM system.asynchronous_metrics WHERE name = 'NetworkTCPSockets_LISTEN';
+SELECT value > 0 FROM system.asynchronous_metrics WHERE name = 'NetworkTCPSocketRemoteAddresses';
