diff --git a/tests/queries/0_stateless/01322_any_input_optimize.reference b/tests/queries/0_stateless/01322_any_input_optimize.reference
index 209a4afb3a2b..4b6fbfd32c03 100644
--- a/tests/queries/0_stateless/01322_any_input_optimize.reference
+++ b/tests/queries/0_stateless/01322_any_input_optimize.reference
@@ -1,3 +1,30 @@
-9
 SELECT any(number) + (any(number) * 2)
-FROM numbers(3, 10)
+FROM numbers(1, 2)
+3
+SELECT anyLast(number) + (anyLast(number) * 2)
+FROM numbers(1, 2)
+6
+WITH any(number) * 3 AS x
+SELECT x
+FROM numbers(1, 2)
+3
+SELECT 
+    anyLast(number) * 3 AS x,
+    x
+FROM numbers(1, 2)
+6	6
+SELECT any(number + (number * 2))
+FROM numbers(1, 2)
+3
+SELECT anyLast(number + (number * 2))
+FROM numbers(1, 2)
+6
+WITH any(number * 3) AS x
+SELECT x
+FROM numbers(1, 2)
+3
+SELECT 
+    anyLast(number * 3) AS x,
+    x
+FROM numbers(1, 2)
+6	6
diff --git a/tests/queries/0_stateless/01322_any_input_optimize.sql b/tests/queries/0_stateless/01322_any_input_optimize.sql
index 65f09d65738d..4b8a55d4c7bc 100644
--- a/tests/queries/0_stateless/01322_any_input_optimize.sql
+++ b/tests/queries/0_stateless/01322_any_input_optimize.sql
@@ -1,4 +1,32 @@
-SET optimize_move_functions_out_of_any=1;
-SET enable_debug_queries=1;
-SELECT any(number + number * 2) FROM  numbers(3, 10);
-ANALYZE SELECT any(number + number * 2) FROM  numbers(3, 10);
+SET enable_debug_queries = 1;
+SET optimize_move_functions_out_of_any = 1;
+
+ANALYZE SELECT any(number + number * 2) FROM numbers(1, 2);
+SELECT any(number + number * 2) FROM numbers(1, 2);
+
+ANALYZE SELECT anyLast(number + number * 2) FROM numbers(1, 2);
+SELECT anyLast(number + number * 2) FROM numbers(1, 2);
+
+ANALYZE WITH any(number * 3) AS x SELECT x FROM numbers(1, 2);
+WITH any(number * 3) AS x SELECT x FROM numbers(1, 2);
+
+ANALYZE SELECT anyLast(number * 3) AS x, x FROM numbers(1, 2);
+SELECT anyLast(number * 3) AS x, x FROM numbers(1, 2);
+
+SELECT any(anyLast(number)) FROM numbers(1); -- { serverError 184 }
+
+SET optimize_move_functions_out_of_any = 0;
+
+ANALYZE SELECT any(number + number * 2) FROM numbers(1, 2);
+SELECT any(number + number * 2) FROM numbers(1, 2);
+
+ANALYZE SELECT anyLast(number + number * 2) FROM numbers(1, 2);
+SELECT anyLast(number + number * 2) FROM numbers(1, 2);
+
+ANALYZE WITH any(number * 3) AS x SELECT x FROM numbers(1, 2);
+WITH any(number * 3) AS x SELECT x FROM numbers(1, 2);
+
+ANALYZE SELECT anyLast(number * 3) AS x, x FROM numbers(1, 2);
+SELECT anyLast(number * 3) AS x, x FROM numbers(1, 2);
+
+SELECT any(anyLast(number)) FROM numbers(1); -- { serverError 184 }
diff --git a/tests/queries/0_stateless/01414_optimize_any_bug.reference b/tests/queries/0_stateless/01414_optimize_any_bug.reference
new file mode 100644
index 000000000000..573541ac9702
--- /dev/null
+++ b/tests/queries/0_stateless/01414_optimize_any_bug.reference
@@ -0,0 +1,1 @@
+0
diff --git a/tests/queries/0_stateless/01414_optimize_any_bug.sql b/tests/queries/0_stateless/01414_optimize_any_bug.sql
new file mode 100644
index 000000000000..6f6f291c504e
--- /dev/null
+++ b/tests/queries/0_stateless/01414_optimize_any_bug.sql
@@ -0,0 +1,18 @@
+DROP TABLE IF EXISTS test;
+
+CREATE TABLE test
+(
+    `Source.C1` Array(UInt64),
+    `Source.C2` Array(UInt64)
+)
+ENGINE = MergeTree()
+ORDER BY tuple();
+
+SET optimize_move_functions_out_of_any = 1;
+
+SELECT any(arrayFilter((c, d) -> (4 = d), `Source.C1`, `Source.C2`)[1]) AS x
+FROM test
+WHERE 0
+GROUP BY 42;
+
+DROP TABLE test;
