diff --git a/tests/queries/0_stateless/00250_tuple_comparison.reference b/tests/queries/0_stateless/00250_tuple_comparison.reference
index 05547aaf6a4d..078592f1f2a4 100644
--- a/tests/queries/0_stateless/00250_tuple_comparison.reference
+++ b/tests/queries/0_stateless/00250_tuple_comparison.reference
@@ -13,3 +13,4 @@
 1	0	0	0	1	1
 0	1	1	0	1	0
 0	1	0	1	0	1
+\N	\N	\N	\N	\N
diff --git a/tests/queries/0_stateless/00250_tuple_comparison.sql b/tests/queries/0_stateless/00250_tuple_comparison.sql
index 8f5763e2ebbe..2ee29ef58639 100644
--- a/tests/queries/0_stateless/00250_tuple_comparison.sql
+++ b/tests/queries/0_stateless/00250_tuple_comparison.sql
@@ -103,3 +103,9 @@ SELECT
     tuple(2) >  tuple(1),
     tuple(2) <= tuple(1),
     tuple(2) >= tuple(1);
+SELECT
+    tuple(NULL) < tuple(1),
+    tuple(NULL) = tuple(1),
+    tuple(NULL) <= tuple(1),
+    tuple(1, NULL) = tuple(2, 1),
+    tuple(1, NULL) < tuple(2, 1);
