diff --git a/tests/queries/0_stateless/01375_output_format_tsv_csv_with_names.reference b/tests/queries/0_stateless/01375_output_format_tsv_csv_with_names.reference
new file mode 100644
index 000000000000..6f1974ccd73d
--- /dev/null
+++ b/tests/queries/0_stateless/01375_output_format_tsv_csv_with_names.reference
@@ -0,0 +1,13 @@
+TSVWithNames
+number
+0
+1
+TSVWithNamesAndTypes
+number
+UInt64
+0
+1
+CSVWithNames
+"number"
+0
+1
diff --git a/tests/queries/0_stateless/01375_output_format_tsv_csv_with_names.sh b/tests/queries/0_stateless/01375_output_format_tsv_csv_with_names.sh
new file mode 100755
index 000000000000..de4486a88a5c
--- /dev/null
+++ b/tests/queries/0_stateless/01375_output_format_tsv_csv_with_names.sh
@@ -0,0 +1,18 @@
+#!/usr/bin/env bash
+
+CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+. "$CURDIR"/../shell_config.sh
+
+opts=(
+    --input-format CSV
+    -q 'SELECT number FROM numbers(2)'
+)
+
+echo 'TSVWithNames'
+${CLICKHOUSE_LOCAL} "${opts[@]}" --format TSVWithNames
+
+echo 'TSVWithNamesAndTypes'
+${CLICKHOUSE_LOCAL} "${opts[@]}" --format TSVWithNamesAndTypes
+
+echo 'CSVWithNames'
+${CLICKHOUSE_LOCAL} "${opts[@]}" --format CSVWithNames
diff --git a/tests/queries/0_stateless/01375_storage_file_tsv_csv_with_names_write_prefix.reference b/tests/queries/0_stateless/01375_storage_file_tsv_csv_with_names_write_prefix.reference
new file mode 100644
index 000000000000..4f2a79b99059
--- /dev/null
+++ b/tests/queries/0_stateless/01375_storage_file_tsv_csv_with_names_write_prefix.reference
@@ -0,0 +1,17 @@
+zero rows
+TSVWithNames
+TSVWithNamesAndTypes
+CSVWithNames
+multi clickhouse-local one file
+TSVWithNames
+0
+0
+0
+TSVWithNamesAndTypes
+0
+0
+0
+CSVWithNames
+0
+0
+0
diff --git a/tests/queries/0_stateless/01375_storage_file_tsv_csv_with_names_write_prefix.sh b/tests/queries/0_stateless/01375_storage_file_tsv_csv_with_names_write_prefix.sh
new file mode 100755
index 000000000000..d396981f873c
--- /dev/null
+++ b/tests/queries/0_stateless/01375_storage_file_tsv_csv_with_names_write_prefix.sh
@@ -0,0 +1,30 @@
+#!/usr/bin/env bash
+
+CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+. "$CURDIR"/../shell_config.sh
+
+# zero rows
+echo 'zero rows'
+for format in TSVWithNames TSVWithNamesAndTypes CSVWithNames; do
+    echo $format
+    ${CLICKHOUSE_LOCAL} --query="
+        CREATE TABLE ${format}_01375 ENGINE File($format, '01375_$format.tsv') AS SELECT * FROM numbers(1) WHERE number < 0;
+        SELECT * FROM ${format}_01375;
+        DROP TABLE ${format}_01375;
+    "
+    rm 01375_$format.tsv
+done
+
+# run multiple times to the same file
+echo 'multi clickhouse-local one file'
+for format in TSVWithNames TSVWithNamesAndTypes CSVWithNames; do
+    echo $format
+    for _ in {1..2}; do
+        ${CLICKHOUSE_LOCAL} --query="
+            CREATE TABLE ${format}_01375 ENGINE File($format, '01375_$format.tsv') AS SELECT * FROM numbers(1);
+            SELECT * FROM ${format}_01375;
+            DROP TABLE ${format}_01375;
+        "
+    done
+    rm 01375_$format.tsv
+done
diff --git a/tests/queries/0_stateless/01375_storage_file_write_prefix.reference b/tests/queries/0_stateless/01375_storage_file_write_prefix_csv_with_names.reference
similarity index 100%
rename from tests/queries/0_stateless/01375_storage_file_write_prefix.reference
rename to tests/queries/0_stateless/01375_storage_file_write_prefix_csv_with_names.reference
diff --git a/tests/queries/0_stateless/01375_storage_file_write_prefix.sql b/tests/queries/0_stateless/01375_storage_file_write_prefix_csv_with_names.sql
similarity index 100%
rename from tests/queries/0_stateless/01375_storage_file_write_prefix.sql
rename to tests/queries/0_stateless/01375_storage_file_write_prefix_csv_with_names.sql
diff --git a/tests/queries/0_stateless/01375_storage_file_write_prefix_tsv_with_names.reference b/tests/queries/0_stateless/01375_storage_file_write_prefix_tsv_with_names.reference
new file mode 100644
index 000000000000..ed9a18b93463
--- /dev/null
+++ b/tests/queries/0_stateless/01375_storage_file_write_prefix_tsv_with_names.reference
@@ -0,0 +1,30 @@
+0	0
+1	1
+2	2
+3	3
+4	4
+5	5
+6	6
+7	7
+8	8
+9	9
+0	0
+1	1
+2	2
+3	3
+4	4
+5	5
+6	6
+7	7
+8	8
+9	9
+0	0
+1	1
+2	2
+3	3
+4	4
+5	5
+6	6
+7	7
+8	8
+9	9
diff --git a/tests/queries/0_stateless/01375_storage_file_write_prefix_tsv_with_names.sql b/tests/queries/0_stateless/01375_storage_file_write_prefix_tsv_with_names.sql
new file mode 100644
index 000000000000..55a97eb6e957
--- /dev/null
+++ b/tests/queries/0_stateless/01375_storage_file_write_prefix_tsv_with_names.sql
@@ -0,0 +1,14 @@
+DROP TABLE IF EXISTS tmp_01375;
+DROP TABLE IF EXISTS table_tsv_01375;
+
+CREATE TABLE tmp_01375 (n UInt32, s String) ENGINE = Memory;
+CREATE TABLE table_tsv_01375 AS tmp_01375 ENGINE = File(TSVWithNames);
+
+INSERT INTO table_tsv_01375 SELECT number as n, toString(n) as s FROM numbers(10);
+INSERT INTO table_tsv_01375 SELECT number as n, toString(n) as s FROM numbers(10);
+INSERT INTO table_tsv_01375 SELECT number as n, toString(n) as s FROM numbers(10);
+
+SELECT * FROM table_tsv_01375;
+
+DROP TABLE IF EXISTS tmp_01375;
+DROP TABLE IF EXISTS table_tsv_01375;
