diff --git a/tests/queries/0_stateless/01637_nullable_fuzz3.reference b/tests/queries/0_stateless/01637_nullable_fuzz3.reference
index d9cf16b9d2ab..795a0159ac8d 100644
--- a/tests/queries/0_stateless/01637_nullable_fuzz3.reference
+++ b/tests/queries/0_stateless/01637_nullable_fuzz3.reference
@@ -1,2 +1,6 @@
 
 0
+
+0
+
+0	0	0
diff --git a/tests/queries/0_stateless/01637_nullable_fuzz3.sql b/tests/queries/0_stateless/01637_nullable_fuzz3.sql
index 21bf8999eaef..6cfd0fc7d1c7 100644
--- a/tests/queries/0_stateless/01637_nullable_fuzz3.sql
+++ b/tests/queries/0_stateless/01637_nullable_fuzz3.sql
@@ -1,4 +1,6 @@
 DROP TABLE IF EXISTS t;
 CREATE TABLE t (`item_id` UInt64, `price_sold` Float32, `date` Date) ENGINE = MergeTree ORDER BY item_id;
 SELECT item_id FROM (SELECT item_id FROM t GROUP BY item_id WITH TOTALS) AS l FULL OUTER JOIN (SELECT item_id FROM t GROUP BY item_id WITH TOTALS) AS r USING (item_id);
+SELECT item_id FROM (SELECT item_id FROM t GROUP BY item_id WITH TOTALS) AS l FULL OUTER JOIN (SELECT item_id FROM t GROUP BY item_id WITH TOTALS) AS r USING (item_id) SETTINGS join_use_nulls = '1';
+SELECT * FROM (SELECT item_id, sum(price_sold) as price_sold FROM t GROUP BY item_id WITH TOTALS) AS l FULL OUTER JOIN (SELECT item_id, sum(price_sold) as price_sold FROM t GROUP BY item_id WITH TOTALS) AS r USING (item_id) SETTINGS join_use_nulls = '1';
 DROP TABLE t;
