diff --git a/tests/queries/0_stateless/02868_select_support_from_keywords.reference b/tests/queries/0_stateless/02868_select_support_from_keywords.reference
new file mode 100644
index 000000000000..d2dcb047cf0a
--- /dev/null
+++ b/tests/queries/0_stateless/02868_select_support_from_keywords.reference
@@ -0,0 +1,1 @@
+CREATE VIEW default.test_view
(
    `date` Date,
    `__sign` Int8,
    `from` Float64,
    `to` Float64
) AS
WITH cte AS
    (
        SELECT
            date,
            __sign,
            from,
            to
        FROM default.test_table
        FINAL
    )
SELECT
    date,
    __sign,
    from,
    to
FROM cte
diff --git a/tests/queries/0_stateless/02868_select_support_from_keywords.sql b/tests/queries/0_stateless/02868_select_support_from_keywords.sql
new file mode 100644
index 000000000000..dc06651a8eb1
--- /dev/null
+++ b/tests/queries/0_stateless/02868_select_support_from_keywords.sql
@@ -0,0 +1,5 @@
+create table test_table ( `date` Date, `__sign` Int8, `from` Float64, `to` Float64 ) ENGINE = CollapsingMergeTree(__sign) PARTITION BY toYYYYMM(date) ORDER BY (date) SETTINGS index_granularity = 8192;
+create VIEW test_view AS WITH cte AS (SELECT date, __sign, "from", "to" FROM test_table FINAL) SELECT date, __sign, "from", "to" FROM cte;
+show create table test_view;
+drop table test_view;
+drop table test_table;
