diff --git a/tests/queries/0_stateless/01780_range_msan.reference b/tests/queries/0_stateless/01780_range_msan.reference
new file mode 100644
index 000000000000..e69de29bb2d1
diff --git a/tests/queries/0_stateless/01780_range_msan.sql b/tests/queries/0_stateless/01780_range_msan.sql
new file mode 100644
index 000000000000..dd0a35c3eea3
--- /dev/null
+++ b/tests/queries/0_stateless/01780_range_msan.sql
@@ -0,0 +1,1 @@
+SELECT range(toUInt256(1), 1); -- { serverError 44 }
