diff --git a/test/sql/table_function/duckdb_keywords.test b/test/sql/table_function/duckdb_keywords.test
new file mode 100644
index 000000000000..650510f56f8a
--- /dev/null
+++ b/test/sql/table_function/duckdb_keywords.test
@@ -0,0 +1,9 @@
+# name: test/sql/table_function/duckdb_keywords.test
+# description: Test duckdb_keywords function
+# group: [table_function]
+
+statement ok
+PRAGMA enable_verification
+
+statement ok
+SELECT * FROM duckdb_keywords();
