diff --git a/tests/queries/0_stateless/01925_join_materialized_columns.reference b/tests/queries/0_stateless/01925_join_materialized_columns.reference
index 1dfda3c769b1..5125b322aed4 100644
--- a/tests/queries/0_stateless/01925_join_materialized_columns.reference
+++ b/tests/queries/0_stateless/01925_join_materialized_columns.reference
@@ -22,3 +22,13 @@ fact1t1_val1	fact1t2_val2
 fact2t1_val2	fact2t1_val2
 -
 2020-02-02 13:00:00	2020-02-05 13:00:00
+-
+1
+1
+1
+1
+-
+2020-01-01 12:00:00
+2020-01-01 12:00:00
+2020-01-01 12:00:00
+2020-01-01 12:00:00
diff --git a/tests/queries/0_stateless/01925_join_materialized_columns.sql b/tests/queries/0_stateless/01925_join_materialized_columns.sql
index 6a34fef96ab5..7c56d5fea39c 100644
--- a/tests/queries/0_stateless/01925_join_materialized_columns.sql
+++ b/tests/queries/0_stateless/01925_join_materialized_columns.sql
@@ -2,17 +2,23 @@ DROP TABLE IF EXISTS t1;
 DROP TABLE IF EXISTS t2;
 
 CREATE TABLE t1 (
-    time DateTime, foo String, dimension_1 String,
+    time DateTime,
+    foo String,
+    dimension_1 String,
     dt Date MATERIALIZED toDate(time),
     dt1 Date MATERIALIZED toDayOfYear(time),
-    aliascol1 ALIAS foo || dimension_1
+    aliascol1 ALIAS foo || dimension_1,
+    time_alias DateTime ALIAS time
 ) ENGINE = MergeTree() PARTITION BY toYYYYMM(dt) ORDER BY (dt, foo);
 
 CREATE TABLE t2 (
-    time DateTime, bar String, dimension_2 String, 
+    time DateTime,
+    bar String,
+    dimension_2 String,
     dt Date MATERIALIZED toDate(time),
     dt2 Date MATERIALIZED toDayOfYear(time),
-    aliascol2 ALIAS bar || dimension_2
+    aliascol2 ALIAS bar || dimension_2,
+    time_alias DateTime ALIAS time
 ) ENGINE = MergeTree() PARTITION BY toYYYYMM(dt) ORDER BY (dt, bar);
 
 INSERT INTO t1 VALUES ('2020-01-01 12:00:00', 'fact1', 't1_val1'), ('2020-02-02 13:00:00', 'fact2', 't1_val2'), ('2020-01-01 13:00:00', 'fact3', 't1_val3');
@@ -35,3 +41,15 @@ SELECT '-';
 SELECT t1.aliascol1, t2.aliascol2 FROM t1 JOIN t2 ON t1.foo = t2.bar ORDER BY t1.time, t2.time;
 SELECT '-';
 SELECT t1.time, t2.time FROM t1 JOIN t2 ON t1.aliascol1 = t2.aliascol2 ORDER BY t1.time, t2.time;
+SELECT '-';
+SELECT count() FROM t1 JOIN t2 ON t1.time_alias = t2.time;
+SELECT count() FROM t1 JOIN t2 ON t1.time = t2.time_alias;
+SELECT count() FROM t1 JOIN t2 ON t1.time_alias = t2.time_alias;
+SELECT count() FROM t1 JOIN t2 USING (time_alias);
+SELECT '-';
+SELECT t1.time as talias FROM t1 JOIN t2 ON talias = t2.time;
+SELECT t1.time as talias FROM t1 JOIN t2 ON talias = t2.time_alias;
+SELECT t2.time as talias FROM t1 JOIN t2 ON t1.time = talias;
+SELECT t2.time as talias FROM t1 JOIN t2 ON t1.time_alias = talias;
+SELECT time as talias FROM t1 JOIN t2 ON t1.time = talias; -- { serverError AMBIGUOUS_COLUMN_NAME }
+SELECT time as talias FROM t1 JOIN t2 ON talias = t2.time; -- { serverError AMBIGUOUS_COLUMN_NAME }
