diff --git a/test/api/capi/test_capi_prepared.cpp b/test/api/capi/test_capi_prepared.cpp
index db2484d6e9c6..0ba93cb41db0 100644
--- a/test/api/capi/test_capi_prepared.cpp
+++ b/test/api/capi/test_capi_prepared.cpp
@@ -253,4 +253,13 @@ TEST_CASE("Test prepared statements in C API", "[capi]") {
 	memcpy(malloced_data, "hello\0", 6);
 	REQUIRE(string((char *)malloced_data) == "hello");
 	duckdb_free(malloced_data);
+
+	status = duckdb_prepare(tester.connection, "SELECT sum(i) FROM a WHERE i > ?", &stmt);
+	REQUIRE(status == DuckDBSuccess);
+	REQUIRE(stmt != nullptr);
+	REQUIRE(duckdb_nparams(stmt) == 1);
+	REQUIRE(duckdb_param_type(nullptr, 0) == DUCKDB_TYPE_INVALID);
+	REQUIRE(duckdb_param_type(stmt, 1) == DUCKDB_TYPE_INTEGER);
+
+	duckdb_destroy_prepare(&stmt);
 }
diff --git a/test/sql/copy/csv/empty_first_line.test b/test/sql/copy/csv/empty_first_line.test
new file mode 100644
index 000000000000..002706f84114
--- /dev/null
+++ b/test/sql/copy/csv/empty_first_line.test
@@ -0,0 +1,20 @@
+# name: test/sql/copy/csv/empty_first_line.test
+# description: Issue #4933: Reading a CSV file with an empty header is problematic
+# group: [csv]
+
+statement ok
+PRAGMA enable_verification
+
+query II
+SELECT * FROM read_csv_auto('data/csv/empty_first_line.csv', delim=' ');
+----
+a	1
+b	2
+c	3
+
+query I
+SELECT * FROM read_csv_auto('data/csv/empty_first_line.csv', delim='|');
+----
+a 1
+b 2
+c 3
