diff --git a/tests/queries/0_stateless/02710_date_diff_aliases.reference b/tests/queries/0_stateless/02710_date_diff_aliases.reference
new file mode 100644
index 000000000000..1eeb5a3a2fa9
--- /dev/null
+++ b/tests/queries/0_stateless/02710_date_diff_aliases.reference
@@ -0,0 +1,5 @@
+DATE_DIFF
+TIMESTAMP_DIFF
+date_diff
+timestampDiff
+timestamp_diff
diff --git a/tests/queries/0_stateless/02710_date_diff_aliases.sql b/tests/queries/0_stateless/02710_date_diff_aliases.sql
new file mode 100644
index 000000000000..c6b31c44f952
--- /dev/null
+++ b/tests/queries/0_stateless/02710_date_diff_aliases.sql
@@ -0,0 +1,7 @@
+SELECT name FROM system.functions
+WHERE name = 'date_diff'
+   OR name = 'DATE_DIFF'
+   OR name = 'timestampDiff'
+   OR name = 'timestamp_diff'
+   OR name = 'TIMESTAMP_DIFF'
+ORDER BY name;
diff --git a/tests/queries/0_stateless/02711_trim_aliases.reference b/tests/queries/0_stateless/02711_trim_aliases.reference
new file mode 100644
index 000000000000..fa0920cd079e
--- /dev/null
+++ b/tests/queries/0_stateless/02711_trim_aliases.reference
@@ -0,0 +1,3 @@
+ltrim
+rtrim
+trim
diff --git a/tests/queries/0_stateless/02711_trim_aliases.sql b/tests/queries/0_stateless/02711_trim_aliases.sql
new file mode 100644
index 000000000000..d0d739805fdc
--- /dev/null
+++ b/tests/queries/0_stateless/02711_trim_aliases.sql
@@ -0,0 +1,5 @@
+SELECT name FROM system.functions
+WHERE name = 'ltrim'
+   OR name = 'rtrim'
+   OR name = 'trim'
+ORDER BY name;
