diff --git a/src/server/transaction.cc b/src/server/transaction.cc
index 6fbaec62b0bb..834ee4c421d7 100644
--- a/src/server/transaction.cc
+++ b/src/server/transaction.cc
@@ -1445,12 +1445,14 @@ OpResult<KeyIndex> DetermineKeys(const CommandId* cid, CmdArgList args) {
     if (!absl::SimpleAtoi(num, &num_custom_keys) || num_custom_keys < 0)
       return OpStatus::INVALID_INT;
 
-    // TODO Fix this for Z family functions.
-    // Examples that crash: ZUNION 0 myset
     if (name == "ZDIFF" && num_custom_keys == 0) {
       return OpStatus::INVALID_INT;
     }
 
+    if (name == "ZUNION" && num_custom_keys == 0) {
+      return OpStatus::SYNTAX_ERR;
+    }
+
     if (args.size() < size_t(num_custom_keys) + num_keys_index + 1)
       return OpStatus::SYNTAX_ERR;
   }
diff --git a/src/server/zset_family.cc b/src/server/zset_family.cc
index 51197cf5d1ae..17346b044911 100644
--- a/src/server/zset_family.cc
+++ b/src/server/zset_family.cc
@@ -956,7 +956,7 @@ OpResult<void> FillAggType(string_view agg, SetOpArgs* op_args) {
 
 // Parse functions return the number of arguments read from CmdArgList
 OpResult<unsigned> ParseAggregate(CmdArgList args, bool store, SetOpArgs* op_args) {
-  if (args.size() < 1) {
+  if (args.size() <= 1) {
     return OpStatus::SYNTAX_ERR;
   }
 
