diff --git a/data/parquet-testing/struct_skip_test.parquet b/data/parquet-testing/struct_skip_test.parquet
new file mode 100644
index 000000000000..88988e74f439
Binary files /dev/null and b/data/parquet-testing/struct_skip_test.parquet differ
diff --git a/test/sql/copy/parquet/struct_column_reader_skip.test b/test/sql/copy/parquet/struct_column_reader_skip.test
new file mode 100644
index 000000000000..f94e4c53d00c
--- /dev/null
+++ b/test/sql/copy/parquet/struct_column_reader_skip.test
@@ -0,0 +1,13 @@
+# name: test/sql/copy/parquet/struct_column_reader_skip.test
+# description: Issue #3205: skip in struct column reader
+# group: [parquet]
+
+require parquet
+
+statement ok
+PRAGMA enable_verification
+
+query III
+SELECT my_map['A'], * FROM parquet_scan('data/parquet-testing/struct_skip_test.parquet') where filter == '0'
+----
+[NULL]	{A=NULL}	0
