diff --git a/tests/queries/0_stateless/03155_in_nested_subselects.reference b/tests/queries/0_stateless/03155_in_nested_subselects.reference
new file mode 100644
index 000000000000..5565ed6787ff
--- /dev/null
+++ b/tests/queries/0_stateless/03155_in_nested_subselects.reference
@@ -0,0 +1,4 @@
+0
+1
+0
+1
diff --git a/tests/queries/0_stateless/03155_in_nested_subselects.sql b/tests/queries/0_stateless/03155_in_nested_subselects.sql
new file mode 100644
index 000000000000..4f5ccd30aa3e
--- /dev/null
+++ b/tests/queries/0_stateless/03155_in_nested_subselects.sql
@@ -0,0 +1,19 @@
+-- https://github.com/ClickHouse/ClickHouse/issues/63833
+SET allow_experimental_analyzer = 1;
+
+create table Example (id Int32) engine = MergeTree ORDER BY id;
+INSERT INTO Example SELECT number AS id FROM numbers(2);
+
+create table Null engine=Null as Example ;
+--create table Null engine=MergeTree order by id as Example ;
+
+create materialized view Transform to Example as
+select * from Null 
+join ( select * FROM Example 
+       WHERE id IN (SELECT * FROM (SELECT * FROM (SELECT * FROM (SELECT * FROM (SELECT * FROM Null)))))
+     ) as old  
+using id;
+
+INSERT INTO Null SELECT number AS id FROM numbers(2);
+
+select * from Example;  -- should return 4 rows
