diff --git a/tests/queries/0_stateless/03100_lwu_39_after_replace_partition.reference b/tests/queries/0_stateless/03100_lwu_39_after_replace_partition.reference
new file mode 100644
index 000000000000..51993f072d58
--- /dev/null
+++ b/tests/queries/0_stateless/03100_lwu_39_after_replace_partition.reference
@@ -0,0 +1,2 @@
+2
+2
diff --git a/tests/queries/0_stateless/03100_lwu_39_after_replace_partition.sql b/tests/queries/0_stateless/03100_lwu_39_after_replace_partition.sql
new file mode 100644
index 000000000000..0b1df5b3d249
--- /dev/null
+++ b/tests/queries/0_stateless/03100_lwu_39_after_replace_partition.sql
@@ -0,0 +1,25 @@
+DROP TABLE IF EXISTS t_lwu_replace;
+
+SET allow_experimental_lightweight_update = 1;
+
+CREATE TABLE t_lwu_replace (c0 Int)
+ENGINE = MergeTree ORDER BY tuple()
+SETTINGS enable_block_number_column = 1, enable_block_offset_column = 1;
+
+INSERT INTO TABLE t_lwu_replace (c0) VALUES (1);
+ALTER TABLE t_lwu_replace REPLACE PARTITION ID '0' FROM t_lwu_replace;
+UPDATE t_lwu_replace SET c0 = 2 WHERE TRUE;
+
+SELECT * FROM t_lwu_replace ORDER BY c0;
+DROP TABLE IF EXISTS t_lwu_replace;
+
+CREATE TABLE t_lwu_replace (c0 Int)
+ENGINE = ReplicatedMergeTree('/clickhouse/tables/{database}/t_lwu_replace', '1') ORDER BY tuple()
+SETTINGS enable_block_number_column = 1, enable_block_offset_column = 1;
+
+INSERT INTO TABLE t_lwu_replace (c0) VALUES (1);
+ALTER TABLE t_lwu_replace REPLACE PARTITION ID '0' FROM t_lwu_replace;
+UPDATE t_lwu_replace SET c0 = 2 WHERE TRUE;
+
+SELECT * FROM t_lwu_replace ORDER BY c0;
+DROP TABLE IF EXISTS t_lwu_replace;
