diff --git a/src/catalog/dependency_manager.cpp b/src/catalog/dependency_manager.cpp
index 66042b4d9350..303d511b3744 100644
--- a/src/catalog/dependency_manager.cpp
+++ b/src/catalog/dependency_manager.cpp
@@ -513,6 +513,9 @@ void DependencyManager::AlterObject(CatalogTransaction transaction, CatalogEntry
 				disallow_alter = false;
 				break;
 			}
+			case AlterTableType::ADD_COLUMN: {
+				disallow_alter = false;
+			}
 			default:
 				break;
 			}
