diff --git a/tests/queries/0_stateless/01639_distributed_sync_insert_zero_rows.reference b/tests/queries/0_stateless/01639_distributed_sync_insert_zero_rows.reference
new file mode 100644
index 000000000000..ad86be865c1d
--- /dev/null
+++ b/tests/queries/0_stateless/01639_distributed_sync_insert_zero_rows.reference
@@ -0,0 +1,6 @@
+128
+256
+128
+256
+128
+256
diff --git a/tests/queries/0_stateless/01639_distributed_sync_insert_zero_rows.sql b/tests/queries/0_stateless/01639_distributed_sync_insert_zero_rows.sql
new file mode 100644
index 000000000000..7fb81efaac88
--- /dev/null
+++ b/tests/queries/0_stateless/01639_distributed_sync_insert_zero_rows.sql
@@ -0,0 +1,24 @@
+DROP TABLE IF EXISTS local;
+DROP TABLE IF EXISTS distributed;
+
+CREATE TABLE local (x UInt8) ENGINE = Memory;
+CREATE TABLE distributed AS local ENGINE = Distributed(test_cluster_two_shards, currentDatabase(), local, x);
+
+SET insert_distributed_sync = 1;
+
+INSERT INTO distributed SELECT number FROM numbers(256) WHERE number % 2 = 0;
+SELECT count() FROM local;
+SELECT count() FROM distributed;
+
+TRUNCATE TABLE local;
+INSERT INTO distributed SELECT number FROM numbers(256) WHERE number % 2 = 1;
+SELECT count() FROM local;
+SELECT count() FROM distributed;
+
+TRUNCATE TABLE local;
+INSERT INTO distributed SELECT number FROM numbers(256) WHERE number < 128;
+SELECT count() FROM local;
+SELECT count() FROM distributed;
+
+DROP TABLE local;
+DROP TABLE distributed;
diff --git a/tests/queries/0_stateless/arcadia_skip_list.txt b/tests/queries/0_stateless/arcadia_skip_list.txt
index 3d9b70a621e1..6c6636b923db 100644
--- a/tests/queries/0_stateless/arcadia_skip_list.txt
+++ b/tests/queries/0_stateless/arcadia_skip_list.txt
@@ -181,3 +181,4 @@
 01561_mann_whitney_scipy
 01601_custom_tld
 01636_nullable_fuzz2
+01639_distributed_sync_insert_zero_rows
