diff --git a/src/Storages/MutationCommands.cpp b/src/Storages/MutationCommands.cpp
index 2aa90a039bcc..8902707ab645 100644
--- a/src/Storages/MutationCommands.cpp
+++ b/src/Storages/MutationCommands.cpp
@@ -1,5 +1,6 @@
 #include <Storages/MutationCommands.h>
-#include <IO/Operators.h>
+#include <IO/WriteHelpers.h>
+#include <IO/ReadHelpers.h>
 #include <Parsers/formatAST.h>
 #include <Parsers/ExpressionListParsers.h>
 #include <Parsers/ParserAlterQuery.h>
@@ -133,13 +134,13 @@ void MutationCommands::writeText(WriteBuffer & out) const
 {
     WriteBufferFromOwnString commands_buf;
     formatAST(*ast(), commands_buf, /* hilite = */ false, /* one_line = */ true);
-    out << escape << commands_buf.str();
+    writeEscapedString(commands_buf.str(), out);
 }
 
 void MutationCommands::readText(ReadBuffer & in)
 {
     String commands_str;
-    in >> escape >> commands_str;
+    readEscapedString(commands_str, in);
 
     ParserAlterCommandList p_alter_commands;
     auto commands_ast = parseQuery(
