diff --git a/tests/queries/0_stateless/02241_array_first_last_or_null.reference b/tests/queries/0_stateless/02241_array_first_last_or_null.reference
index 2906b04ecd02..fc4a5ff8af5e 100644
--- a/tests/queries/0_stateless/02241_array_first_last_or_null.reference
+++ b/tests/queries/0_stateless/02241_array_first_last_or_null.reference
@@ -7,6 +7,9 @@ ArrayFirst non constant predicate
 \N
 2
 2
+ArrayFirst with Null
+2
+\N
 ArrayLast constant predicate
 \N
 \N
@@ -16,3 +19,6 @@ ArrayLast non constant predicate
 \N
 3
 3
+ArrayLast with Null
+2
+\N
diff --git a/tests/queries/0_stateless/02241_array_first_last_or_null.sql b/tests/queries/0_stateless/02241_array_first_last_or_null.sql
index 3230e4d483a9..aa8f0cdbf928 100644
--- a/tests/queries/0_stateless/02241_array_first_last_or_null.sql
+++ b/tests/queries/0_stateless/02241_array_first_last_or_null.sql
@@ -9,6 +9,10 @@ SELECT arrayFirstOrNull(x -> x >= 2, emptyArrayUInt8());
 SELECT arrayFirstOrNull(x -> x >= 2, [1, 2, 3]);
 SELECT arrayFirstOrNull(x -> x >= 2, materialize([1, 2, 3]));
 
+SELECT 'ArrayFirst with Null';
+SELECT arrayFirstOrNull((x,f) -> f, [1,2,3,NULL], [0,1,0,0]);
+SELECT arrayFirstOrNull((x,f) -> f, [1,2,3,NULL], [0,0,0,1]);
+
 SELECT 'ArrayLast constant predicate';
 SELECT arrayLastOrNull(x -> 1, emptyArrayUInt8());
 SELECT arrayLastOrNull(x -> 0, emptyArrayUInt8());
@@ -19,3 +23,7 @@ SELECT 'ArrayLast non constant predicate';
 SELECT arrayLastOrNull(x -> x >= 2, emptyArrayUInt8());
 SELECT arrayLastOrNull(x -> x >= 2, [1, 2, 3]);
 SELECT arrayLastOrNull(x -> x >= 2, materialize([1, 2, 3]));
+
+SELECT 'ArrayLast with Null';
+SELECT arrayLastOrNull((x,f) -> f, [1,2,3,NULL], [0,1,0,0]);
+SELECT arrayLastOrNull((x,f) -> f, [1,2,3,NULL], [0,1,0,1]);
\ No newline at end of file
