diff --git a/dbms/tests/queries/0_stateless/01079_new_range_reader_segfault.reference b/dbms/tests/queries/0_stateless/01079_new_range_reader_segfault.reference
new file mode 100644
index 000000000000..257cc5642cb1
--- /dev/null
+++ b/dbms/tests/queries/0_stateless/01079_new_range_reader_segfault.reference
@@ -0,0 +1,1 @@
+foo
diff --git a/dbms/tests/queries/0_stateless/01079_new_range_reader_segfault.sql b/dbms/tests/queries/0_stateless/01079_new_range_reader_segfault.sql
new file mode 100644
index 000000000000..9c3f073c72f3
--- /dev/null
+++ b/dbms/tests/queries/0_stateless/01079_new_range_reader_segfault.sql
@@ -0,0 +1,11 @@
+drop table if exists t;
+
+create table t (a Int) engine = MergeTree order by a;
+
+-- some magic to satisfy conditions to run optimizations in MergeTreeRangeReader
+insert into t select number < 20 ? 0 : 1 from numbers(50);
+alter table t add column s String default 'foo';
+
+select s from t prewhere a != 1 where rand() % 2 = 0 limit 1;
+
+drop table t;
