diff --git a/tests/queries/0_stateless/02896_optimize_array_exists_to_has_with_date.reference b/tests/queries/0_stateless/02896_optimize_array_exists_to_has_with_date.reference
new file mode 100644
index 000000000000..d00491fd7e5b
--- /dev/null
+++ b/tests/queries/0_stateless/02896_optimize_array_exists_to_has_with_date.reference
@@ -0,0 +1,1 @@
+1
diff --git a/tests/queries/0_stateless/02896_optimize_array_exists_to_has_with_date.sql b/tests/queries/0_stateless/02896_optimize_array_exists_to_has_with_date.sql
new file mode 100644
index 000000000000..b87a154b46ab
--- /dev/null
+++ b/tests/queries/0_stateless/02896_optimize_array_exists_to_has_with_date.sql
@@ -0,0 +1,1 @@
+SELECT arrayExists(date -> (date = '2022-07-31'), [toDate('2022-07-31')]) AS date_exists;
