diff --git a/dbms/tests/queries/0_stateless/01066_bit_count.reference b/dbms/tests/queries/0_stateless/01066_bit_count.reference
new file mode 100644
index 000000000000..4a3b084b4a26
--- /dev/null
+++ b/dbms/tests/queries/0_stateless/01066_bit_count.reference
@@ -0,0 +1,21 @@
+0
+1
+1
+2
+1
+2
+2
+3
+1
+2
+4
+0
+1
+8
+64
+32
+16
+8
+1	10	000000000000F03F
+-1	11	000000000000F0BF
+inf	11	000000000000F07F
diff --git a/dbms/tests/queries/0_stateless/01066_bit_count.sql b/dbms/tests/queries/0_stateless/01066_bit_count.sql
new file mode 100644
index 000000000000..d50b26575422
--- /dev/null
+++ b/dbms/tests/queries/0_stateless/01066_bit_count.sql
@@ -0,0 +1,13 @@
+SELECT bitCount(number) FROM numbers(10);
+SELECT avg(bitCount(number)) FROM numbers(256);
+
+SELECT bitCount(0);
+SELECT bitCount(1);
+SELECT bitCount(-1);
+
+SELECT bitCount(toInt64(-1));
+SELECT bitCount(toInt32(-1));
+SELECT bitCount(toInt16(-1));
+SELECT bitCount(toInt8(-1));
+
+SELECT x, bitCount(x), hex(reinterpretAsString(x)) FROM VALUES ('x Float64', (1), (-1), (inf));
