diff --git a/data/parquet-testing/timetz_4byte_stats.parquet b/data/parquet-testing/timetz_4byte_stats.parquet
new file mode 100644
index 000000000000..a3357557e9b7
Binary files /dev/null and b/data/parquet-testing/timetz_4byte_stats.parquet differ
diff --git a/test/sql/copy/parquet/timetz_parquet.test b/test/sql/copy/parquet/timetz_parquet.test
new file mode 100644
index 000000000000..4371535a1065
--- /dev/null
+++ b/test/sql/copy/parquet/timetz_parquet.test
@@ -0,0 +1,13 @@
+# name: test/sql/copy/parquet/timetz_parquet.test
+# description: Test correct reading of stats for timetz columns
+# group: [parquet]
+
+require parquet
+
+statement ok
+PRAGMA enable_verification
+
+query II
+select * from 'data/parquet-testing/timetz_4byte_stats.parquet' order by 1;
+----
+00:00:00+00	00:00:00+00
