diff --git a/src/execution/physical_plan/plan_comparison_join.cpp b/src/execution/physical_plan/plan_comparison_join.cpp
index 3f843120db58..9984b909eb7f 100644
--- a/src/execution/physical_plan/plan_comparison_join.cpp
+++ b/src/execution/physical_plan/plan_comparison_join.cpp
@@ -91,6 +91,10 @@ void CheckForPerfectJoinOpt(LogicalComparisonJoin &op, PerfectHashJoinStats &joi
 	    !ExtractNumericValue(NumericStats::Max(stats_build), max_value)) {
 		return;
 	}
+	if (max_value < min_value) {
+		// empty table
+		return;
+	}
 	int64_t build_range;
 	if (!TrySubtractOperator::Operation(max_value, min_value, build_range)) {
 		return;
