diff --git a/tests/queries/0_stateless/01926_bin_unbin.reference b/tests/queries/0_stateless/01926_bin_unbin.reference
index f84a858e4498..731d0223bb96 100644
--- a/tests/queries/0_stateless/01926_bin_unbin.reference
+++ b/tests/queries/0_stateless/01926_bin_unbin.reference
@@ -33,3 +33,7 @@
 1
 1
 1
+1
+1
+2D000000000000000A
+001011010000000000000000000000000000000000000000000000000000000000001010
diff --git a/tests/queries/0_stateless/01926_bin_unbin.sql b/tests/queries/0_stateless/01926_bin_unbin.sql
index 555770d09c65..e112f8bd8a4d 100644
--- a/tests/queries/0_stateless/01926_bin_unbin.sql
+++ b/tests/queries/0_stateless/01926_bin_unbin.sql
@@ -37,3 +37,9 @@ select bin(unbin('0')) == '00000000';
 select hex('') == bin('');
 select unhex('') == unbin('');
 select unhex('0') == unbin('0');
+
+-- hex and bin support AggregateFunction
+select hex(sumState(number)) == hex(toString(sumState(number))) from numbers(10);
+select hex(avgState(number)) == hex(toString(avgState(number))) from numbers(99);
+select hex(avgState(number)) from numbers(10);
+select bin(avgState(number)) from numbers(10);
