diff --git a/tools/pythonpkg/tests/fast/api/test_with_propagating_exceptions.py b/tools/pythonpkg/tests/fast/api/test_with_propagating_exceptions.py
new file mode 100644
index 000000000000..c9fa98d4d1d6
--- /dev/null
+++ b/tools/pythonpkg/tests/fast/api/test_with_propagating_exceptions.py
@@ -0,0 +1,18 @@
+import pytest
+import duckdb
+
+class TestWithPropagatingExceptions(object):
+
+    def test_with(self):
+        # Should propagate exception raised in the 'with duckdb.connect() ..'
+        with pytest.raises(duckdb.ParserException, match="syntax error at or near *"):
+            with duckdb.connect() as con:
+                print('before')
+                con.execute('invalid')
+                print('after')
+
+        # Does not raise an exception
+        with duckdb.connect() as con:
+            print('before')
+            con.execute('select 1')
+            print('after')
