diff --git a/src/Access/AccessControl.cpp b/src/Access/AccessControl.cpp
index 432fe19634e6..fca4146adf8c 100644
--- a/src/Access/AccessControl.cpp
+++ b/src/Access/AccessControl.cpp
@@ -621,7 +621,9 @@ AuthResult AccessControl::authenticate(const Credentials & credentials, const Po
         /// It is typical when users install ClickHouse, type some password and instantly forget it.
         if (credentials.getUserName().empty() || credentials.getUserName() == "default")
         {
-            error_code = ErrorCodes::REQUIRED_PASSWORD;
+            if (credentials.allowInteractiveBasicAuthenticationInTheBrowser())
+                error_code = ErrorCodes::REQUIRED_PASSWORD;
+
             message << R"(
 
 If you use ClickHouse Cloud, the password can be reset at https://clickhouse.cloud/
diff --git a/src/Access/Credentials.h b/src/Access/Credentials.h
index 52f33385a0e8..e7cbf0743e15 100644
--- a/src/Access/Credentials.h
+++ b/src/Access/Credentials.h
@@ -24,6 +24,7 @@ class Credentials
 
     const String & getUserName() const;
     bool isReady() const;
+    virtual bool allowInteractiveBasicAuthenticationInTheBrowser() const { return false; }
 
 protected:
     [[noreturn]] static void throwNotReady();
@@ -66,9 +67,12 @@ class BasicCredentials
     void setUserName(const String & user_name_);
     void setPassword(const String & password_);
     const String & getPassword() const;
+    bool allowInteractiveBasicAuthenticationInTheBrowser() const override { return allow_interactive_basic_authentication_in_the_browser; }
+    void enableInteractiveBasicAuthenticationInTheBrowser() { allow_interactive_basic_authentication_in_the_browser = true; }
 
 private:
     String password;
+    bool allow_interactive_basic_authentication_in_the_browser = false;
 };
 
 class CredentialsWithScramble : public Credentials
diff --git a/src/Server/HTTP/authenticateUserByHTTP.cpp b/src/Server/HTTP/authenticateUserByHTTP.cpp
index f96ffc2cb6a9..65a6851ecb2b 100644
--- a/src/Server/HTTP/authenticateUserByHTTP.cpp
+++ b/src/Server/HTTP/authenticateUserByHTTP.cpp
@@ -217,6 +217,9 @@ bool authenticateUserByHTTP(
         if (!basic_credentials)
             throw Exception(ErrorCodes::AUTHENTICATION_FAILED, "Invalid authentication: expected 'Basic' HTTP Authorization scheme");
 
+        if (request.get("Authorization", "") != "never")
+            basic_credentials->enableInteractiveBasicAuthenticationInTheBrowser();
+
         chassert(!user.empty());
         basic_credentials->setUserName(user);
         basic_credentials->setPassword(password);
