diff --git a/tests/queries/0_stateless/02187_async_inserts_all_formats.reference b/tests/queries/0_stateless/02187_async_inserts_all_formats.reference
index b4a5b6c3a426..92ce8dfd2c7c 100644
--- a/tests/queries/0_stateless/02187_async_inserts_all_formats.reference
+++ b/tests/queries/0_stateless/02187_async_inserts_all_formats.reference
@@ -14,8 +14,10 @@ JSONCompactStringsEachRow
 JSONCompactStringsEachRowWithNames
 JSONCompactStringsEachRowWithNamesAndTypes
 JSONEachRow
+JSONLines
 JSONStringsEachRow
 MsgPack
+NDJSON
 Native
 ORC
 Parquet
diff --git a/tests/queries/0_stateless/02267_jsonlines_ndjson_format.reference b/tests/queries/0_stateless/02267_jsonlines_ndjson_format.reference
new file mode 100644
index 000000000000..b8346a240da3
--- /dev/null
+++ b/tests/queries/0_stateless/02267_jsonlines_ndjson_format.reference
@@ -0,0 +1,32 @@
+{"number":"0"}
+{"number":"1"}
+{"number":"2"}
+{"number":"3"}
+{"number":"4"}
+{"number":"5"}
+{"number":"6"}
+{"number":"7"}
+{"number":"8"}
+{"number":"9"}
+{"number":"0"}
+{"number":"1"}
+{"number":"2"}
+{"number":"3"}
+{"number":"4"}
+{"number":"5"}
+{"number":"6"}
+{"number":"7"}
+{"number":"8"}
+{"number":"9"}
+{"n1":1,"n2":2}
+{"n1":1,"n2":2}
+{"n1":3,"n2":4}
+{"n1":3,"n2":4}
+{"n1":5,"n2":6}
+{"n1":5,"n2":6}
+{"n1":1,"n2":2}
+{"n1":1,"n2":2}
+{"n1":3,"n2":4}
+{"n1":3,"n2":4}
+{"n1":5,"n2":6}
+{"n1":5,"n2":6}
diff --git a/tests/queries/0_stateless/02267_jsonlines_ndjson_format.sql b/tests/queries/0_stateless/02267_jsonlines_ndjson_format.sql
new file mode 100644
index 000000000000..5e32758b3284
--- /dev/null
+++ b/tests/queries/0_stateless/02267_jsonlines_ndjson_format.sql
@@ -0,0 +1,14 @@
+SELECT * FROM numbers(10) FORMAT JSONLines;
+SELECT * FROM numbers(10) FORMAT NDJSON;
+
+DROP TABLE IF EXISTS 02267_t;
+
+CREATE TABLE 02267_t (n1 UInt32, n2 UInt32) ENGINE = Memory;
+
+INSERT INTO 02267_t FORMAT JSONLines {"n1": 1, "n2": 2} {"n1": 3, "n2": 4} {"n1": 5, "n2": 6};
+INSERT INTO 02267_t FORMAT NDJSON {"n1": 1, "n2": 2} {"n1": 3, "n2": 4} {"n1": 5, "n2": 6};
+
+SELECT * FROM 02267_t ORDER BY n1, n2 FORMAT JSONLines;
+SELECT * FROM 02267_t ORDER BY n1, n2 FORMAT NDJSON;
+
+DROP TABLE 02267_t;
