diff --git a/tests/queries/0_stateless/03522_nullable_partition_key.reference b/tests/queries/0_stateless/03522_nullable_partition_key.reference
new file mode 100644
index 000000000000..19441250f22d
--- /dev/null
+++ b/tests/queries/0_stateless/03522_nullable_partition_key.reference
@@ -0,0 +1,5 @@
+1	1_2_2_1
+\N	2b16b490a87466b6a6d34fd9027f3c3d_1_1_1
+1	1_1_1_1
+2	2_2_2_1
+\N	2b16b490a87466b6a6d34fd9027f3c3d_3_3_1
diff --git a/tests/queries/0_stateless/03522_nullable_partition_key.sql b/tests/queries/0_stateless/03522_nullable_partition_key.sql
new file mode 100644
index 000000000000..589a0642c38d
--- /dev/null
+++ b/tests/queries/0_stateless/03522_nullable_partition_key.sql
@@ -0,0 +1,10 @@
+CREATE TABLE t (c0 Nullable(Int)) ENGINE = MergeTree() ORDER BY (c0) PARTITION BY (c0) SETTINGS allow_nullable_key = 1;
+
+INSERT INTO TABLE t (c0) VALUES (NULL),(1);
+OPTIMIZE TABLE t FINAL;
+SELECT c0, _part FROM t ORDER BY ALL;
+
+CREATE TABLE taggr (c0 Nullable(Int)) ENGINE = AggregatingMergeTree() ORDER BY (c0) PARTITION BY (c0) SETTINGS allow_nullable_key = 1;
+INSERT INTO TABLE taggr (c0) VALUES (1), (2);
+INSERT INTO TABLE taggr (c0) VALUES (NULL), (1), (2);
+SELECT c0, _part FROM taggr FINAL ORDER BY ALL;
