diff --git a/tests/integration/test_storage_kafka/test.py b/tests/integration/test_storage_kafka/test.py
index 13577864870f..d89684e21315 100644
--- a/tests/integration/test_storage_kafka/test.py
+++ b/tests/integration/test_storage_kafka/test.py
@@ -840,28 +840,28 @@ def test_kafka_virtual_columns2(kafka_cluster):
                      kafka_format = 'JSONEachRow';
 
         CREATE MATERIALIZED VIEW test.view Engine=Log AS
-        SELECT value, _key, _topic, _partition, _offset, toUnixTimestamp(_timestamp) FROM test.kafka;
+        SELECT value, _key, _topic, _partition, _offset, toUnixTimestamp(_timestamp), toUnixTimestamp64Milli(_timestamp_ms) FROM test.kafka;
         ''')
 
     producer = KafkaProducer(bootstrap_servers="localhost:9092")
 
-    producer.send(topic='virt2_0', value=json.dumps({'value': 1}), partition=0, key='k1', timestamp_ms=1577836801000)
-    producer.send(topic='virt2_0', value=json.dumps({'value': 2}), partition=0, key='k2', timestamp_ms=1577836802000)
+    producer.send(topic='virt2_0', value=json.dumps({'value': 1}), partition=0, key='k1', timestamp_ms=1577836801001)
+    producer.send(topic='virt2_0', value=json.dumps({'value': 2}), partition=0, key='k2', timestamp_ms=1577836802002)
     producer.flush()
     time.sleep(1)
 
-    producer.send(topic='virt2_0', value=json.dumps({'value': 3}), partition=1, key='k3', timestamp_ms=1577836803000)
-    producer.send(topic='virt2_0', value=json.dumps({'value': 4}), partition=1, key='k4', timestamp_ms=1577836804000)
+    producer.send(topic='virt2_0', value=json.dumps({'value': 3}), partition=1, key='k3', timestamp_ms=1577836803003)
+    producer.send(topic='virt2_0', value=json.dumps({'value': 4}), partition=1, key='k4', timestamp_ms=1577836804004)
     producer.flush()
     time.sleep(1)
 
-    producer.send(topic='virt2_1', value=json.dumps({'value': 5}), partition=0, key='k5', timestamp_ms=1577836805000)
-    producer.send(topic='virt2_1', value=json.dumps({'value': 6}), partition=0, key='k6', timestamp_ms=1577836806000)
+    producer.send(topic='virt2_1', value=json.dumps({'value': 5}), partition=0, key='k5', timestamp_ms=1577836805005)
+    producer.send(topic='virt2_1', value=json.dumps({'value': 6}), partition=0, key='k6', timestamp_ms=1577836806006)
     producer.flush()
     time.sleep(1)
 
-    producer.send(topic='virt2_1', value=json.dumps({'value': 7}), partition=1, key='k7', timestamp_ms=1577836807000)
-    producer.send(topic='virt2_1', value=json.dumps({'value': 8}), partition=1, key='k8', timestamp_ms=1577836808000)
+    producer.send(topic='virt2_1', value=json.dumps({'value': 7}), partition=1, key='k7', timestamp_ms=1577836807007)
+    producer.send(topic='virt2_1', value=json.dumps({'value': 8}), partition=1, key='k8', timestamp_ms=1577836808008)
     producer.flush()
 
     time.sleep(10)
@@ -869,14 +869,14 @@ def test_kafka_virtual_columns2(kafka_cluster):
     result = instance.query("SELECT * FROM test.view ORDER BY value", ignore_error=True)
 
     expected = '''\
-1	k1	virt2_0	0	0	1577836801
-2	k2	virt2_0	0	1	1577836802
-3	k3	virt2_0	1	0	1577836803
-4	k4	virt2_0	1	1	1577836804
-5	k5	virt2_1	0	0	1577836805
-6	k6	virt2_1	0	1	1577836806
-7	k7	virt2_1	1	0	1577836807
-8	k8	virt2_1	1	1	1577836808
+1	k1	virt2_0	0	0	1577836801	1577836801001
+2	k2	virt2_0	0	1	1577836802	1577836802002
+3	k3	virt2_0	1	0	1577836803	1577836803003
+4	k4	virt2_0	1	1	1577836804	1577836804004
+5	k5	virt2_1	0	0	1577836805	1577836805005
+6	k6	virt2_1	0	1	1577836806	1577836806006
+7	k7	virt2_1	1	0	1577836807	1577836807007
+8	k8	virt2_1	1	1	1577836808	1577836808008
 '''
 
     assert TSV(result) == TSV(expected)
