diff --git a/src/server/transaction.cc b/src/server/transaction.cc
index 040a01ceb6fe..3db3f0abd4ea 100644
--- a/src/server/transaction.cc
+++ b/src/server/transaction.cc
@@ -82,6 +82,7 @@ Transaction::~Transaction() {
 OpStatus Transaction::InitByArgs(DbIndex index, CmdArgList args) {
   db_index_ = index;
   cmd_with_full_args_ = args;
+  local_result_ = OpStatus::OK;
 
   if (IsGlobal()) {
     unique_shard_cnt_ = shard_set->size();
