diff --git a/src/IO/parseDateTimeBestEffort.cpp b/src/IO/parseDateTimeBestEffort.cpp
index 063955cdd1ea..a3ea418dd9f5 100644
--- a/src/IO/parseDateTimeBestEffort.cpp
+++ b/src/IO/parseDateTimeBestEffort.cpp
@@ -119,6 +119,7 @@ ReturnType parseDateTimeBestEffortImpl(
     UInt8 time_zone_offset_hour = 0;
     UInt8 time_zone_offset_minute = 0;
 
+    bool is_am = false;
     bool is_pm = false;
 
     auto read_alpha_month = [&month] (const auto & alpha)
@@ -486,6 +487,7 @@ ReturnType parseDateTimeBestEffortImpl(
                     {
                         if (alpha[0] == 'A' || alpha[0] == 'a')
                         {
+                            is_am = true;
                         }
                         else if (alpha[0] == 'P' || alpha[0] == 'p')
                         {
@@ -560,6 +562,9 @@ ReturnType parseDateTimeBestEffortImpl(
     if (!check_date(is_leap_year, month, day_of_month))
         return on_error("Cannot read DateTime: unexpected date: " + std::to_string(year) + "-" + std::to_string(month) + "-" + std::to_string(day_of_month), ErrorCodes::CANNOT_PARSE_DATETIME);
 
+    if (is_am && hour == 12)
+        hour = 0;
+
     if (is_pm && hour < 12)
         hour += 12;
 
diff --git a/src/IO/parseDateTimeBestEffort.h b/src/IO/parseDateTimeBestEffort.h
index fcc0195c4d9f..61de31a02139 100644
--- a/src/IO/parseDateTimeBestEffort.h
+++ b/src/IO/parseDateTimeBestEffort.h
@@ -50,7 +50,7 @@ class ReadBuffer;
   *
   * single whitespace can be used as a separator
   *
-  * AM/PM - AM is ignored and PM means: add 12 hours if value is less than 12.
+  * AM/PM - AM means: subtract 12 hours if a value is 12 and PM means: add 12 hours if a value is less than 12.
   *
   * Jan/Feb/Mar/Apr/May/Jun/Jul/Aug/Sep/Oct/Nov/Dec - allowed to specify month
   * Mon/Tue/Wed/Thu/Fri/Sat/Sun - simply ignored.
