diff --git a/tests/queries/0_stateless/03566_analyzer_single_with_scope.reference b/tests/queries/0_stateless/03566_analyzer_single_with_scope.reference
new file mode 100644
index 000000000000..bea4d8d5eb90
--- /dev/null
+++ b/tests/queries/0_stateless/03566_analyzer_single_with_scope.reference
@@ -0,0 +1,8 @@
+1
+1
+1
+1
+ENABLED COMPATIBILITY MODE
+1
+1
+1
diff --git a/tests/queries/0_stateless/03566_analyzer_single_with_scope.sql b/tests/queries/0_stateless/03566_analyzer_single_with_scope.sql
new file mode 100644
index 000000000000..d8935f23249c
--- /dev/null
+++ b/tests/queries/0_stateless/03566_analyzer_single_with_scope.sql
@@ -0,0 +1,49 @@
+create table tab (x String, y UInt8) engine = MergeTree order by tuple();
+insert into tab select 'rue', 1;
+
+with ('t' || x) as y 
+  select 1 from tab where y = 'true' settings enable_analyzer=0;
+with ('t' || x) as y 
+  select 1 from tab where y = 'true' settings enable_analyzer=1;
+
+with ('t' || x) as y select * from 
+  (select 1 from tab where y = 'true') settings enable_analyzer=0;
+with ('t' || x) as y select * from 
+  (select 1 from tab where y = 'true') settings enable_analyzer=1; -- { serverError TYPE_MISMATCH }
+
+with
+  ('t' || x) as y,
+  'rue' as x
+select 1 from tab where y = 'true' settings enable_analyzer=1;
+
+with ('t' || x) as y,
+  'rue' as x 
+select * from 
+  (select 1 from tab where y = 'true') settings enable_analyzer=1; -- { serverError TYPE_MISMATCH }
+
+SET enable_scopes_for_with_statement = 0;
+
+SELECT 'ENABLED COMPATIBILITY MODE';
+
+with ('t' || x) as y select * from 
+  (select 1 from tab where y = 'true') settings enable_analyzer=1;
+
+with
+  ('t' || x) as y,
+  'rue' as x
+select 1 from tab where y = 'true' settings enable_analyzer=1;
+
+with ('t' || x) as y,
+  'rue' as x 
+select * from 
+  (select 1 from tab where y = 'true') settings enable_analyzer=1;
+
+with
+  ('t' || x) as y,
+  'rue' as x
+select 1 from tab where y = 'true' settings enable_analyzer=0; -- { serverError TYPE_MISMATCH }
+
+with ('t' || x) as y,
+  'rue' as x 
+select * from 
+  (select 1 from tab where y = 'true') settings enable_analyzer=0; -- { serverError TYPE_MISMATCH }
diff --git a/tests/queries/0_stateless/03567_analyzer_single_with_scope_cycle.reference b/tests/queries/0_stateless/03567_analyzer_single_with_scope_cycle.reference
new file mode 100644
index 000000000000..e69de29bb2d1
diff --git a/tests/queries/0_stateless/03567_analyzer_single_with_scope_cycle.sql b/tests/queries/0_stateless/03567_analyzer_single_with_scope_cycle.sql
new file mode 100644
index 000000000000..7ffa1cc8d34d
--- /dev/null
+++ b/tests/queries/0_stateless/03567_analyzer_single_with_scope_cycle.sql
@@ -0,0 +1,30 @@
+create table tab (x String, y UInt8) engine = MergeTree order by tuple();
+insert into tab select 'rue', 1;
+
+WITH
+    a as b
+SELECT 1 FROM (
+    WITH b as c 
+    SELECT 1 FROM (
+        WITH c as d
+        SELECT 1 FROM (
+            SELECT 1 FROM tab WHERE e = 'true'
+        )
+    )
+)
+SETTINGS allow_experimental_analyzer = 0; -- { serverError UNKNOWN_IDENTIFIER }
+
+SET enable_scopes_for_with_statement = 0;
+
+WITH
+    a as b
+SELECT 1 FROM (
+    WITH b as c 
+    SELECT 1 FROM (
+        WITH c as d
+        SELECT 1 FROM (
+            SELECT 1 FROM tab WHERE e = 'true'
+        )
+    )
+)
+SETTINGS allow_experimental_analyzer = 1; -- { serverError UNKNOWN_IDENTIFIER }
