diff --git a/tests/performance/quantile.xml b/tests/performance/quantile.xml
new file mode 100644
index 000000000000..4718c129db68
--- /dev/null
+++ b/tests/performance/quantile.xml
@@ -0,0 +1,29 @@
+<test>
+    <preconditions>
+        <table_exists>hits_100m_single</table_exists>
+    </preconditions>
+
+    <substitutions>
+        <substitution>
+           <name>key</name>
+           <values>
+               <value>SearchEngineID</value>
+               <value>RegionID</value>
+               <value>SearchPhrase</value>
+               <value>ClientIP</value>
+           </values>
+        </substitution>
+        <substitution>
+           <name>func</name>
+           <values>
+               <value>quantile</value>
+               <value>quantileExact</value>
+               <value>quantileTDigest</value>
+               <value>quantileTiming</value>
+               <value>quantileBFloat16</value>
+           </values>
+       </substitution>
+    </substitutions>
+
+    <query>SELECT {key} AS k, {func}(ResolutionWidth) FROM hits_100m_single GROUP BY k FORMAT Null</query>
+</test>
diff --git a/tests/performance/uniq.xml b/tests/performance/uniq.xml
index 0ddd2321ca62..d4521961b9a0 100644
--- a/tests/performance/uniq.xml
+++ b/tests/performance/uniq.xml
@@ -1,12 +1,9 @@
 <test>
-
     <preconditions>
         <table_exists>hits_100m_single</table_exists>
         <ram_size>30000000000</ram_size>
     </preconditions>
 
-
-
     <settings>
         <max_memory_usage>30000000000</max_memory_usage>
         <!--
@@ -36,7 +33,7 @@
                <value>SearchPhrase</value>
                <value>ClientIP</value>
            </values>
-       </substitution>
+        </substitution>
         <substitution>
            <name>func</name>
            <values>
diff --git a/tests/queries/0_stateless/00753_quantile_format.reference b/tests/queries/0_stateless/00753_quantile_format.reference
index 04bb0000b3ed..d3503745e4b2 100644
--- a/tests/queries/0_stateless/00753_quantile_format.reference
+++ b/tests/queries/0_stateless/00753_quantile_format.reference
@@ -10,7 +10,9 @@
 [30000]
 30000
 [30000]
-2016-06-15 23:01:04
-['2016-06-15 23:01:04']
-2016-06-15 23:01:04
-['2016-06-15 23:01:04']
+2016-06-15 23:00:16
+['2016-06-15 23:00:16']
+2016-06-15 23:00:16
+['2016-06-15 23:00:16']
+2016-04-02 17:23:12
+['2016-04-02 17:23:12']
diff --git a/tests/queries/0_stateless/00753_quantile_format.sql b/tests/queries/0_stateless/00753_quantile_format.sql
index 24cf0e4bf902..f72ecc8a49fd 100644
--- a/tests/queries/0_stateless/00753_quantile_format.sql
+++ b/tests/queries/0_stateless/00753_quantile_format.sql
@@ -1,7 +1,7 @@
 DROP TABLE IF EXISTS datetime;
 
-CREATE TABLE datetime (d DateTime) ENGINE = Memory;
-INSERT INTO datetime(d) VALUES(toDateTime('2016-06-15 23:00:00'));
+CREATE TABLE datetime (d DateTime('UTC')) ENGINE = Memory;
+INSERT INTO datetime(d) VALUES(toDateTime('2016-06-15 23:00:00', 'UTC'));
 
 SELECT quantile(0.2)(d) FROM datetime;
 SELECT quantiles(0.2)(d) FROM datetime;
@@ -27,4 +27,7 @@ SELECT quantilesTDigest(0.2)(d) FROM datetime;
 SELECT quantileTDigestWeighted(0.2)(d, 1) FROM datetime;
 SELECT quantilesTDigestWeighted(0.2)(d, 1) FROM datetime;
 
+SELECT quantileBFloat16(0.2)(d) FROM datetime;
+SELECT quantilesBFloat16(0.2)(d) FROM datetime;
+
 DROP TABLE datetime;
diff --git a/tests/queries/0_stateless/01813_quantileBfloat16_nans.reference b/tests/queries/0_stateless/01813_quantileBfloat16_nans.reference
new file mode 100644
index 000000000000..d00491fd7e5b
--- /dev/null
+++ b/tests/queries/0_stateless/01813_quantileBfloat16_nans.reference
@@ -0,0 +1,1 @@
+1
diff --git a/tests/queries/0_stateless/01813_quantileBfloat16_nans.sql b/tests/queries/0_stateless/01813_quantileBfloat16_nans.sql
new file mode 100644
index 000000000000..1f823f0b9902
--- /dev/null
+++ b/tests/queries/0_stateless/01813_quantileBfloat16_nans.sql
@@ -0,0 +1,16 @@
+SELECT DISTINCT
+    eq
+FROM
+    (
+        WITH
+            range(2 + number % 10) AS arr, -- minimum two elements, to avoid nan result --
+            arrayMap(x -> x = intDiv(number, 10) ? nan : x, arr) AS arr_with_nan,
+            arrayFilter(x -> x != intDiv(number, 10), arr) AS arr_filtered
+        SELECT
+            number,
+            arrayReduce('quantileBFloat16', arr_with_nan) AS q1,
+            arrayReduce('quantileBFloat16', arr_filtered) AS q2,
+            q1 = q2 AS eq
+        FROM
+            numbers(100)
+    );
diff --git a/tests/queries/1_stateful/00164_quantileBfloat16.reference b/tests/queries/1_stateful/00164_quantileBfloat16.reference
new file mode 100644
index 000000000000..f0aaff5cef68
--- /dev/null
+++ b/tests/queries/1_stateful/00164_quantileBfloat16.reference
@@ -0,0 +1,40 @@
+1704509	1384
+732797	1336
+598875	1384
+792887	1336
+3807842	1336
+25703952	1336
+716829	1384
+59183	1336
+33010362	1336
+800784	1336
+1704509	[1296,1384,1840,1960,3696]
+732797	[1232,1336,1840,1944,3664]
+598875	[1232,1384,1840,1944,3536]
+792887	[1296,1336,1840,1888,3696]
+3807842	[1232,1336,1840,1936,2032]
+25703952	[1012,1336,1840,1944,3696]
+716829	[1232,1384,1840,1944,3696]
+59183	[316,1336,1840,2008,2032]
+33010362	[1232,1336,1840,1936,2032]
+800784	[1232,1336,1840,1928,2032]
+1704509	1384
+732797	1336
+598875	1384
+792887	1336
+3807842	1336
+25703952	1336
+716829	1384
+59183	1336
+33010362	1336
+800784	1336
+1704509	[1296,1384,1840,1960,3696]
+732797	[1232,1336,1840,1944,3664]
+598875	[1232,1384,1840,1944,3536]
+792887	[1296,1336,1840,1888,3696]
+3807842	[1232,1336,1840,1936,2032]
+25703952	[1012,1336,1840,1944,3696]
+716829	[1232,1384,1840,1944,3696]
+59183	[316,1336,1840,2008,2032]
+33010362	[1232,1336,1840,1936,2032]
+800784	[1232,1336,1840,1928,2032]
diff --git a/tests/queries/1_stateful/00164_quantileBfloat16.sql b/tests/queries/1_stateful/00164_quantileBfloat16.sql
new file mode 100644
index 000000000000..b39bae3803c8
--- /dev/null
+++ b/tests/queries/1_stateful/00164_quantileBfloat16.sql
@@ -0,0 +1,6 @@
+SELECT CounterID AS k, quantileBFloat16(0.5)(ResolutionWidth) FROM test.hits GROUP BY k ORDER BY count() DESC, CounterID LIMIT 10;
+SELECT CounterID AS k, quantilesBFloat16(0.1, 0.5, 0.9, 0.99, 0.999)(ResolutionWidth) FROM test.hits GROUP BY k ORDER BY count() DESC, CounterID LIMIT 10;
+
+
+SELECT CounterID AS k, quantileBFloat16(0.5)(ResolutionWidth) FROM remote('127.0.0.{1,2}', test.hits) GROUP BY k ORDER BY count() DESC, CounterID LIMIT 10;
+SELECT CounterID AS k, quantilesBFloat16(0.1, 0.5, 0.9, 0.99, 0.999)(ResolutionWidth) FROM remote('127.0.0.{1,2}', test.hits) GROUP BY k ORDER BY count() DESC, CounterID LIMIT 10;
