diff --git a/.github/workflows/LinuxRelease.yml b/.github/workflows/LinuxRelease.yml
index b12b2f9ca1c5..52f67178ed90 100644
--- a/.github/workflows/LinuxRelease.yml
+++ b/.github/workflows/LinuxRelease.yml
@@ -552,6 +552,8 @@ jobs:
     env:
       BUILD_VISUALIZER: 1
       BUILD_HTTPFS: 1
+      BUILD_TPCH: 1
+      BUILD_TPCDS: 1
       S3_TEST_SERVER_AVAILABLE: 1
       GEN: ninja
 
diff --git a/extension/httpfs/s3fs.cpp b/extension/httpfs/s3fs.cpp
index 046287d923e3..20376c9989ef 100644
--- a/extension/httpfs/s3fs.cpp
+++ b/extension/httpfs/s3fs.cpp
@@ -912,7 +912,7 @@ string AWSListObjectV2::Request(string &path, HTTPParams &http_params, S3AuthPar
 
 	string req_params = "";
 	if (!continuation_token.empty()) {
-		req_params += "continuation-token=" + S3FileSystem::UrlEncode(continuation_token);
+		req_params += "continuation-token=" + S3FileSystem::UrlEncode(continuation_token, true);
 		req_params += "&";
 	}
 	req_params += "encoding-type=url&list-type=2";
