diff --git a/tests/queries/0_stateless/03555_inconsistent_formatting_ttl.reference b/tests/queries/0_stateless/03555_inconsistent_formatting_ttl.reference
new file mode 100644
index 000000000000..e69de29bb2d1
diff --git a/tests/queries/0_stateless/03555_inconsistent_formatting_ttl.sql b/tests/queries/0_stateless/03555_inconsistent_formatting_ttl.sql
new file mode 100644
index 000000000000..45a54932102f
--- /dev/null
+++ b/tests/queries/0_stateless/03555_inconsistent_formatting_ttl.sql
@@ -0,0 +1,3 @@
+DROP TABLE IF EXISTS t0;
+CREATE TABLE t0 (c0 Date) ENGINE = MergeTree() ORDER BY () TTL (materialize(c0));
+DROP TABLE t0;
