diff --git a/tests/queries/0_stateless/02025_subcolumns_compact_parts.reference b/tests/queries/0_stateless/02025_subcolumns_compact_parts.reference
new file mode 100644
index 000000000000..431d62da742e
--- /dev/null
+++ b/tests/queries/0_stateless/02025_subcolumns_compact_parts.reference
@@ -0,0 +1,8 @@
+0
+0
+0
+0
+0
+0
+40000
+219970
diff --git a/tests/queries/0_stateless/02025_subcolumns_compact_parts.sql b/tests/queries/0_stateless/02025_subcolumns_compact_parts.sql
new file mode 100644
index 000000000000..7d1957a1efc6
--- /dev/null
+++ b/tests/queries/0_stateless/02025_subcolumns_compact_parts.sql
@@ -0,0 +1,14 @@
+DROP TABLE IF EXISTS t_comp_subcolumns;
+
+CREATE TABLE t_comp_subcolumns (id UInt32, n Nullable(String), arr Array(Array(UInt32)))
+ENGINE = MergeTree ORDER BY id;
+
+INSERT INTO t_comp_subcolumns SELECT number, 'a', [range(number % 11), range(number % 13)] FROM numbers(20000);
+
+SELECT sum(n.null) FROM t_comp_subcolumns;
+SELECT n.null FROM t_comp_subcolumns LIMIT 10000, 5;
+
+SELECT sum(arr.size0) FROM t_comp_subcolumns;
+SELECT sumArray(arr.size1) FROM t_comp_subcolumns;
+
+DROP TABLE t_comp_subcolumns;
