diff --git a/tests/dragonfly/eval_test.py b/tests/dragonfly/eval_test.py
index f687a16e40f9..acc89142006c 100644
--- a/tests/dragonfly/eval_test.py
+++ b/tests/dragonfly/eval_test.py
@@ -264,3 +264,12 @@ async def test_lua_auto_async(async_client: aioredis.Redis):
 
     flushes = (await async_client.info("transaction"))["eval_squashed_flushes"]
     assert 1 <= flushes <= 3  # all 100 commands are executed in at most 3 batches
+
+
+@dfly_args({"proactor_threads": 4, "lua_auto_async": None})
+async def test_lua_redis_log_noop(async_client: aioredis.Redis):
+    res = await async_client.eval("redis.log('nonsense', 'nonsense')", 0)
+    assert res == None
+
+    res = await async_client.eval("redis.log(redis.LOG_WARNING, 'warn')", 0)
+    assert res == None
