diff --git a/dbms/tests/queries/0_stateless/00960_eval_ml_method_const.reference b/dbms/tests/queries/0_stateless/00960_eval_ml_method_const.reference
new file mode 100644
index 000000000000..573541ac9702
--- /dev/null
+++ b/dbms/tests/queries/0_stateless/00960_eval_ml_method_const.reference
@@ -0,0 +1,1 @@
+0
diff --git a/dbms/tests/queries/0_stateless/00960_eval_ml_method_const.sql b/dbms/tests/queries/0_stateless/00960_eval_ml_method_const.sql
new file mode 100644
index 000000000000..401c83af917e
--- /dev/null
+++ b/dbms/tests/queries/0_stateless/00960_eval_ml_method_const.sql
@@ -0,0 +1,1 @@
+WITH (SELECT stochasticLinearRegressionState(1, 2, 3)) AS model SELECT evalMLMethod(model, toFloat64(1), toFloat64(1));
