diff --git a/tests/queries/0_stateless/02353_isnullable.reference b/tests/queries/0_stateless/02353_isnullable.reference
new file mode 100644
index 000000000000..74240c4b196d
--- /dev/null
+++ b/tests/queries/0_stateless/02353_isnullable.reference
@@ -0,0 +1,4 @@
+0
+1
+1
+1
diff --git a/tests/queries/0_stateless/02353_isnullable.sql b/tests/queries/0_stateless/02353_isnullable.sql
new file mode 100644
index 000000000000..567e294d4987
--- /dev/null
+++ b/tests/queries/0_stateless/02353_isnullable.sql
@@ -0,0 +1,5 @@
+SELECT isNullable(3);
+SELECT isNullable(toNullable(3));
+
+SELECT isNullable(NULL);
+SELECT isNullable(materialize(NULL));
