diff --git a/tests/queries/0_stateless/02893_trash_optimization.reference b/tests/queries/0_stateless/02893_trash_optimization.reference
new file mode 100644
index 000000000000..573541ac9702
--- /dev/null
+++ b/tests/queries/0_stateless/02893_trash_optimization.reference
@@ -0,0 +1,1 @@
+0
diff --git a/tests/queries/0_stateless/02893_trash_optimization.sql b/tests/queries/0_stateless/02893_trash_optimization.sql
new file mode 100644
index 000000000000..a61bc86eeb95
--- /dev/null
+++ b/tests/queries/0_stateless/02893_trash_optimization.sql
@@ -0,0 +1,3 @@
+SELECT *
+FROM merge('system', '^one$') AS one
+WHERE (one.dummy = 0) OR (one.dummy = 1);
