diff --git a/tests/queries/0_stateless/01017_bithamming_distance.reference b/tests/queries/0_stateless/01017_bithamming_distance.reference
index cc2d4f39154d..3e82e0d28648 100644
--- a/tests/queries/0_stateless/01017_bithamming_distance.reference
+++ b/tests/queries/0_stateless/01017_bithamming_distance.reference
@@ -13,3 +13,18 @@
 5
 9
 9
+0
+6
+6
+0
+6
+47
+0
+6
+47
+0
+6
+26
+0
+6
+26
diff --git a/tests/queries/0_stateless/01017_bithamming_distance.sql b/tests/queries/0_stateless/01017_bithamming_distance.sql
index 4b36894b97ca..11f4f27d9a8a 100644
--- a/tests/queries/0_stateless/01017_bithamming_distance.sql
+++ b/tests/queries/0_stateless/01017_bithamming_distance.sql
@@ -18,3 +18,27 @@ SELECT bitHammingDistance(n2, 100) FROM defaults;
 SELECT bitHammingDistance(n3, n4) FROM defaults;
 
 DROP TABLE defaults;
+
+DROP TABLE IF EXISTS test_string;
+
+CREATE TABLE test_string
+(
+	s1 String,
+	s2 String,
+	s3 FixedString(10),
+	s4 FixedString(10),
+) ENGINE = Memory;
+
+INSERT INTO test_string VALUES ('hello', 'hello', 'hello', 'hello') ('hello', 'hellow', 'hello', 'hellow') ('clickhouse', '012345', 'clickhouse', '012345');
+
+SELECT bitHammingDistance('hello', 'hello');
+SELECT bitHammingDistance('hello', 'hellow');
+SELECT bitHammingDistance(toFixedString('hello', 6), toFixedString('hellow', 6));
+
+SELECT bitHammingDistance(s1, s2) FROM test_string;
+SELECT bitHammingDistance(s3, s4) FROM test_string;
+
+SELECT bitHammingDistance('hello', s2) FROM test_string;
+SELECT bitHammingDistance(s4, toFixedString('hello', 10)) FROM test_string;
+
+DROP TABLE test_string;
