diff --git a/tests/queries/0_stateless/02003_bug_from_23515.reference b/tests/queries/0_stateless/02003_bug_from_23515.reference
new file mode 100644
index 000000000000..d00491fd7e5b
--- /dev/null
+++ b/tests/queries/0_stateless/02003_bug_from_23515.reference
@@ -0,0 +1,1 @@
+1
diff --git a/tests/queries/0_stateless/02003_bug_from_23515.sql b/tests/queries/0_stateless/02003_bug_from_23515.sql
new file mode 100644
index 000000000000..febd59b0755e
--- /dev/null
+++ b/tests/queries/0_stateless/02003_bug_from_23515.sql
@@ -0,0 +1,10 @@
+drop table if exists tab;
+create table tab (a LowCardinality(String), b LowCardinality(String)) engine = MergeTree partition by a order by tuple() settings min_bytes_for_wide_part = 0, min_rows_for_wide_part = 0;
+
+insert into tab values ('1', 'a'), ('2', 'b');
+SELECT a = '1' FROM tab WHERE a = '1' and b='a';
+
+-- Fuzzed
+SELECT * FROM tab WHERE (a = '1') AND 0 AND (b = 'a');
+
+drop table if exists tab;
