diff --git a/tools/pythonpkg/tests/fast/test_get_table_names.py b/tools/pythonpkg/tests/fast/test_get_table_names.py
index 968a13c49a9f..00752d184358 100644
--- a/tools/pythonpkg/tests/fast/test_get_table_names.py
+++ b/tools/pythonpkg/tests/fast/test_get_table_names.py
@@ -11,5 +11,5 @@ def test_table_success(self, duckdb_cursor):
     def test_table_fail(self, duckdb_cursor):
         conn = duckdb.connect()
         conn.close()
-        with pytest.raises(duckdb.ConnectionException, match="Connection has already been closed"):
+        with pytest.raises(duckdb.ConnectionException, match="Connection already closed"):
             table_names = conn.get_table_names("SELECT * FROM my_table1, my_table2, my_table3")
diff --git a/tools/pythonpkg/tests/fast/test_runtime_error.py b/tools/pythonpkg/tests/fast/test_runtime_error.py
index 9f1f5378c5e4..98c5a332bda5 100644
--- a/tools/pythonpkg/tests/fast/test_runtime_error.py
+++ b/tools/pythonpkg/tests/fast/test_runtime_error.py
@@ -2,7 +2,7 @@
 import pytest
 from conftest import NumpyPandas, ArrowPandas
 
-closed = lambda: pytest.raises(duckdb.ConnectionException, match='Connection has already been closed')
+closed = lambda: pytest.raises(duckdb.ConnectionException, match='Connection already closed')
 no_result_set = lambda: pytest.raises(duckdb.InvalidInputException, match='No open result set')
 
 
