diff --git a/tests/queries/0_stateless/02285_executable_user_defined_function_group_by.reference b/tests/queries/0_stateless/02285_executable_user_defined_function_group_by.reference
new file mode 100644
index 000000000000..1c59de1a03b9
--- /dev/null
+++ b/tests/queries/0_stateless/02285_executable_user_defined_function_group_by.reference
@@ -0,0 +1,10 @@
+0
+2
+4
+6
+8
+10
+12
+14
+16
+18
diff --git a/tests/queries/0_stateless/02285_executable_user_defined_function_group_by.sql b/tests/queries/0_stateless/02285_executable_user_defined_function_group_by.sql
new file mode 100644
index 000000000000..96db8c088dc4
--- /dev/null
+++ b/tests/queries/0_stateless/02285_executable_user_defined_function_group_by.sql
@@ -0,0 +1,1 @@
+SELECT test_function(number, number) as a FROM numbers(10) GROUP BY a ORDER BY a;
