diff --git a/tests/queries/0_stateless/01305_duplicate_order_by_and_distinct.reference b/tests/queries/0_stateless/01305_duplicate_order_by_and_distinct.reference
index 5e6fdfcdefb6..c076e56a91b9 100644
--- a/tests/queries/0_stateless/01305_duplicate_order_by_and_distinct.reference
+++ b/tests/queries/0_stateless/01305_duplicate_order_by_and_distinct.reference
@@ -31,7 +31,7 @@ ORDER BY number ASC
 SELECT number
 FROM 
 (
-    SELECT DISTINCT number
+    SELECT number
     FROM 
     (
         SELECT DISTINCT number % 2 AS number
diff --git a/tests/queries/0_stateless/01306_disable_duplicate_order_by_and_distinct_optimize_for_distributed_table.reference b/tests/queries/0_stateless/01306_disable_duplicate_order_by_and_distinct_optimize_for_distributed_table.reference
index aa47d0d46d47..44e0be8e3569 100644
--- a/tests/queries/0_stateless/01306_disable_duplicate_order_by_and_distinct_optimize_for_distributed_table.reference
+++ b/tests/queries/0_stateless/01306_disable_duplicate_order_by_and_distinct_optimize_for_distributed_table.reference
@@ -1,2 +1,4 @@
 0
 0
+0
+0
diff --git a/tests/queries/0_stateless/01306_disable_duplicate_order_by_and_distinct_optimize_for_distributed_table.sql b/tests/queries/0_stateless/01306_disable_duplicate_order_by_and_distinct_optimize_for_distributed_table.sql
index e1467bacf2f3..2daf199f82e6 100644
--- a/tests/queries/0_stateless/01306_disable_duplicate_order_by_and_distinct_optimize_for_distributed_table.sql
+++ b/tests/queries/0_stateless/01306_disable_duplicate_order_by_and_distinct_optimize_for_distributed_table.sql
@@ -1,4 +1,6 @@
 set optimize_duplicate_order_by_and_distinct = 1;
+SET distributed_group_by_no_merge = 0;
+
 SELECT DISTINCT number
 FROM
 (
@@ -8,7 +10,19 @@ FROM
     SETTINGS distributed_group_by_no_merge = 1
 );
 
+SET distributed_group_by_no_merge = 1;
+
+SELECT DISTINCT number
+FROM
+(
+    SELECT DISTINCT number
+    FROM remote('127.0.0.{1,2}', system.numbers)
+    LIMIT 1
+);
+
 set optimize_duplicate_order_by_and_distinct = 0;
+SET distributed_group_by_no_merge = 0;
+
 SELECT DISTINCT number
 FROM
 (
@@ -18,3 +32,13 @@ FROM
     SETTINGS distributed_group_by_no_merge = 1
 );
 
+SET distributed_group_by_no_merge = 1;
+
+set optimize_duplicate_order_by_and_distinct = 0;
+SELECT DISTINCT number
+FROM
+(
+    SELECT DISTINCT number
+    FROM remote('127.0.0.{1,2}', system.numbers)
+    LIMIT 1
+);
diff --git a/tests/queries/0_stateless/01455_duplicate_distinct_optimization.reference b/tests/queries/0_stateless/01455_duplicate_distinct_optimization.reference
new file mode 100644
index 000000000000..c5a06ca0cd3c
--- /dev/null
+++ b/tests/queries/0_stateless/01455_duplicate_distinct_optimization.reference
@@ -0,0 +1,136 @@
+SELECT DISTINCT number
+FROM numbers(1)
+SELECT number
+FROM 
+(
+    SELECT DISTINCT number
+    FROM numbers(1)
+)
+SELECT DISTINCT number * 2
+FROM 
+(
+    SELECT DISTINCT 
+        number * 2,
+        number
+    FROM numbers(1)
+)
+SELECT number
+FROM 
+(
+    SELECT DISTINCT number * 2 AS number
+    FROM numbers(1)
+)
+SELECT 
+    b,
+    a
+FROM 
+(
+    SELECT DISTINCT 
+        number % 2 AS a,
+        number % 3 AS b
+    FROM numbers(100)
+)
+SELECT DISTINCT a
+FROM 
+(
+    SELECT DISTINCT 
+        number % 2 AS a,
+        number % 3 AS b
+    FROM numbers(100)
+)
+SELECT a
+FROM 
+(
+    SELECT DISTINCT a
+    FROM 
+    (
+        SELECT DISTINCT 
+            number % 2 AS a,
+            number % 3 AS b
+        FROM numbers(100)
+    )
+)
+SELECT DISTINCT a
+FROM 
+(
+    SELECT 
+        a,
+        b
+    FROM 
+    (
+        SELECT DISTINCT 
+            number % 2 AS a,
+            number % 3 AS b
+        FROM numbers(100)
+    )
+)
+SELECT 
+    a,
+    b
+FROM 
+(
+    SELECT 
+        b,
+        a
+    FROM 
+    (
+        SELECT DISTINCT 
+            number AS a,
+            number AS b
+        FROM numbers(1)
+    )
+)
+SELECT 
+    a,
+    b
+FROM 
+(
+    SELECT 
+        b,
+        a,
+        a + b
+    FROM 
+    (
+        SELECT DISTINCT 
+            number % 2 AS a,
+            number % 3 AS b
+        FROM numbers(100)
+    )
+)
+SELECT DISTINCT a
+FROM 
+(
+    SELECT a
+    FROM 
+    (
+        SELECT DISTINCT 
+            number % 2 AS a,
+            number % 3 AS b
+        FROM numbers(100)
+    )
+)
+SELECT DISTINCT number
+FROM 
+(
+    SELECT DISTINCT number
+    FROM numbers(1)
+) AS t1
+CROSS JOIN numbers(2) AS t2
+SELECT number
+FROM 
+(
+    SELECT DISTINCT number
+    FROM numbers(1) AS t1
+    CROSS JOIN numbers(2) AS t2
+)
+SELECT DISTINCT number
+FROM 
+(
+    SELECT DISTINCT number
+    FROM numbers(1)
+    UNION ALL
+    SELECT DISTINCT number
+    FROM numbers(2)
+)
+0
+1
diff --git a/tests/queries/0_stateless/01455_duplicate_distinct_optimization.sql b/tests/queries/0_stateless/01455_duplicate_distinct_optimization.sql
new file mode 100644
index 000000000000..10937819de28
--- /dev/null
+++ b/tests/queries/0_stateless/01455_duplicate_distinct_optimization.sql
@@ -0,0 +1,33 @@
+SET enable_debug_queries = 1;
+SET optimize_duplicate_order_by_and_distinct = 1;
+
+ANALYZE SELECT DISTINCT number FROM numbers(1);
+ANALYZE SELECT DISTINCT number FROM (SELECT DISTINCT number FROM numbers(1));
+ANALYZE SELECT DISTINCT number * 2 FROM (SELECT DISTINCT number * 2, number FROM numbers(1));
+ANALYZE SELECT DISTINCT number FROM (SELECT DISTINCT number * 2 AS number FROM numbers(1));
+ANALYZE SELECT DISTINCT b, a FROM (SELECT DISTINCT number % 2 AS a, number % 3 AS b FROM numbers(100));
+ANALYZE SELECT DISTINCT a FROM (SELECT DISTINCT number % 2 AS a, number % 3 AS b FROM numbers(100));
+ANALYZE SELECT DISTINCT a FROM (SELECT DISTINCT a FROM (SELECT DISTINCT number % 2 AS a, number % 3 AS b FROM numbers(100)));
+ANALYZE SELECT DISTINCT a FROM (SELECT DISTINCT a, b FROM (SELECT DISTINCT number % 2 AS a, number % 3 AS b FROM numbers(100)));
+ANALYZE SELECT DISTINCT a, b FROM (SELECT DISTINCT b, a FROM (SELECT DISTINCT number a, number b FROM numbers(1)));
+ANALYZE SELECT DISTINCT a, b FROM (SELECT b, a, a + b FROM (SELECT DISTINCT number % 2 AS a, number % 3 AS b FROM numbers(100)));
+ANALYZE SELECT DISTINCT a FROM (SELECT a FROM (SELECT DISTINCT number % 2 AS a, number % 3 AS b FROM numbers(100)));
+ANALYZE SELECT DISTINCT number FROM (SELECT DISTINCT number FROM numbers(1)) t1 CROSS JOIN numbers(2) t2;
+ANALYZE SELECT DISTINCT number FROM (SELECT DISTINCT number FROM numbers(1) t1 CROSS JOIN numbers(2) t2);
+
+ANALYZE SELECT DISTINCT number FROM
+(
+    (SELECT DISTINCT number FROM numbers(1))
+    UNION ALL
+    (SELECT DISTINCT number FROM numbers(2))
+);
+
+--
+
+SELECT DISTINCT number FROM
+(
+    (SELECT DISTINCT number FROM numbers(1))
+    UNION ALL
+    (SELECT DISTINCT number FROM numbers(2))
+)
+ORDER BY number;
