diff --git a/tests/queries/0_stateless/02242_optimize_to_subcolumns_no_storage.reference b/tests/queries/0_stateless/02242_optimize_to_subcolumns_no_storage.reference
new file mode 100644
index 000000000000..0cfbf08886fc
--- /dev/null
+++ b/tests/queries/0_stateless/02242_optimize_to_subcolumns_no_storage.reference
@@ -0,0 +1,1 @@
+2
diff --git a/tests/queries/0_stateless/02242_optimize_to_subcolumns_no_storage.sql b/tests/queries/0_stateless/02242_optimize_to_subcolumns_no_storage.sql
new file mode 100644
index 000000000000..e6e4663c5aaf
--- /dev/null
+++ b/tests/queries/0_stateless/02242_optimize_to_subcolumns_no_storage.sql
@@ -0,0 +1,3 @@
+SET optimize_functions_to_subcolumns = 1;
+SELECT count(*) FROM numbers(2) AS n1, numbers(3) AS n2, numbers(4) AS n3
+WHERE (n1.number = n2.number) AND (n2.number = n3.number);
