diff --git a/tests/queries/0_stateless/01287_max_execution_speed.sql b/tests/queries/0_stateless/01287_max_execution_speed.sql
index 89c3050a2565..8fdd16e0a590 100644
--- a/tests/queries/0_stateless/01287_max_execution_speed.sql
+++ b/tests/queries/0_stateless/01287_max_execution_speed.sql
@@ -13,6 +13,7 @@ SELECT count() FROM system.numbers; -- { serverError TOO_SLOW }
 SET min_execution_speed_bytes = 0;
 SELECT 'Ok (2)';
 
+SET max_execution_time = 600;
 SET max_execution_speed = 1000000;
 SET max_block_size = 100;
 
diff --git a/tests/queries/0_stateless/03509_stripe_log_compatible_types.reference b/tests/queries/0_stateless/03509_stripe_log_compatible_types.reference
new file mode 100644
index 000000000000..6494e611154a
--- /dev/null
+++ b/tests/queries/0_stateless/03509_stripe_log_compatible_types.reference
@@ -0,0 +1,1 @@
+2022-09-21 03:03:24
diff --git a/tests/queries/0_stateless/03509_stripe_log_compatible_types.sql b/tests/queries/0_stateless/03509_stripe_log_compatible_types.sql
new file mode 100644
index 000000000000..d278011674c9
--- /dev/null
+++ b/tests/queries/0_stateless/03509_stripe_log_compatible_types.sql
@@ -0,0 +1,14 @@
+DROP TABLE IF EXISTS test_tz;
+
+CREATE TABLE test_tz
+(
+    `dt` DateTime('UTC')
+)
+ENGINE = StripeLog;
+
+INSERT INTO test_tz VALUES ('2022-09-21 03:03:24');
+
+SELECT *
+FROM test_tz;
+
+DROP TABLE test_tz;
