diff --git a/tests/queries/0_stateless/02116_global_in_time_limit.reference b/tests/queries/0_stateless/02116_global_in_time_limit.reference
new file mode 100644
index 000000000000..e69de29bb2d1
diff --git a/tests/queries/0_stateless/02116_global_in_time_limit.sh b/tests/queries/0_stateless/02116_global_in_time_limit.sh
new file mode 100755
index 000000000000..2db8353dde57
--- /dev/null
+++ b/tests/queries/0_stateless/02116_global_in_time_limit.sh
@@ -0,0 +1,16 @@
+#!/usr/bin/env bash
+
+CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+# shellcheck source=../shell_config.sh
+. "$CURDIR"/../shell_config.sh
+
+# This test is pretty depend on query execution time
+# In release, it works about 14 seconds with no timeout.
+
+function run
+{
+  ${CLICKHOUSE_CLIENT} -q "select number from remote('127.0.0.2', numbers(10)) where number global in (select number + 9 from numbers(400000000)) settings max_execution_time=3" 2>&1 | echo > /dev/null
+}
+
+export -f run
+timeout 6 bash -c run
