diff --git a/tests/queries/0_stateless/02895_forbid_create_inverted_index.reference b/tests/queries/0_stateless/02895_forbid_create_inverted_index.reference
new file mode 100644
index 000000000000..e69de29bb2d1
diff --git a/tests/queries/0_stateless/02895_forbid_create_inverted_index.sql b/tests/queries/0_stateless/02895_forbid_create_inverted_index.sql
new file mode 100644
index 000000000000..dc92d9198fb8
--- /dev/null
+++ b/tests/queries/0_stateless/02895_forbid_create_inverted_index.sql
@@ -0,0 +1,13 @@
+SET allow_experimental_inverted_index = 0;
+DROP TABLE IF EXISTS tab;
+CREATE TABLE tab
+(
+    `key` UInt64,
+    `str` String
+)
+ENGINE = MergeTree
+ORDER BY key;
+
+ALTER TABLE tab ADD INDEX inv_idx(str) TYPE inverted(0); -- { serverError SUPPORT_IS_DISABLED }
+
+DROP TABLE tab;
