diff --git a/tests/queries/0_stateless/01322_cast_keep_nullable.reference b/tests/queries/0_stateless/01322_cast_keep_nullable.reference
new file mode 100644
index 000000000000..8ad99a101709
--- /dev/null
+++ b/tests/queries/0_stateless/01322_cast_keep_nullable.reference
@@ -0,0 +1,10 @@
+0	Int32
+0	Int32
+1	Nullable(Int32)
+1	Nullable(Int32)
+2	Nullable(Float32)
+2	Nullable(UInt8)
+3	Nullable(Int32)
+\N	Nullable(Int32)
+42	Nullable(Int32)
+\N	Nullable(Int32)
diff --git a/tests/queries/0_stateless/01322_cast_keep_nullable.sql b/tests/queries/0_stateless/01322_cast_keep_nullable.sql
new file mode 100644
index 000000000000..109187174693
--- /dev/null
+++ b/tests/queries/0_stateless/01322_cast_keep_nullable.sql
@@ -0,0 +1,19 @@
+SET cast_keep_nullable = 0;
+
+SELECT CAST(toNullable(toInt32(0)) AS Int32) as x, toTypeName(x);
+SELECT CAST(toNullable(toInt8(0)) AS Int32) as x, toTypeName(x);
+
+SET cast_keep_nullable = 1;
+
+SELECT CAST(toNullable(toInt32(1)) AS Int32) as x, toTypeName(x);
+SELECT CAST(toNullable(toInt8(1)) AS Int32) as x, toTypeName(x);
+
+SELECT CAST(toNullable(toFloat32(2)), 'Float32') as x, toTypeName(x);
+SELECT CAST(toNullable(toFloat32(2)), 'UInt8') as x, toTypeName(x);
+SELECT CAST(toNullable(toFloat32(2)), 'UUID') as x, toTypeName(x); -- { serverError 70 }
+
+SELECT CAST(if(1 = 1, toNullable(toInt8(3)), NULL) AS Int32) as x, toTypeName(x);
+SELECT CAST(if(1 = 0, toNullable(toInt8(3)), NULL) AS Int32) as x, toTypeName(x);
+
+SELECT CAST(a, 'Int32') as x, toTypeName(x) FROM (SELECT materialize(CAST(42, 'Nullable(UInt8)')) AS a);
+SELECT CAST(a, 'Int32') as x, toTypeName(x) FROM (SELECT materialize(CAST(NULL, 'Nullable(UInt8)')) AS a);
