diff --git a/src/Common/Arena.h b/src/Common/Arena.h
index aab5ee4a35aa..7604091442e7 100644
--- a/src/Common/Arena.h
+++ b/src/Common/Arena.h
@@ -3,11 +3,11 @@
 #include <cstring>
 #include <memory>
 #include <vector>
-#include <boost/noncopyable.hpp>
 #include <Core/Defines.h>
-#include <Common/memcpySmall.h>
-#include <Common/ProfileEvents.h>
+#include <boost/noncopyable.hpp>
 #include <Common/Allocator.h>
+#include <Common/ProfileEvents.h>
+#include <Common/memcpySmall.h>
 
 #if __has_include(<sanitizer/asan_interface.h>) && defined(ADDRESS_SANITIZER)
 #   include <sanitizer/asan_interface.h>
@@ -180,7 +180,7 @@ class Arena : private boost::noncopyable
     char * alloc(size_t size)
     {
         used_bytes += size;
-        if (unlikely(head.empty() || static_cast<std::ptrdiff_t>(size) > head.end - head.pos))
+        if (unlikely(head.empty() || size > head.remaining()))
             addMemoryChunk(size);
 
         char * res = head.pos;
@@ -193,6 +193,9 @@ class Arena : private boost::noncopyable
     char * alignedAlloc(size_t size, size_t alignment)
     {
         used_bytes += size;
+        if (unlikely(head.empty() || size > head.remaining()))
+            addMemoryChunk(size + alignment);
+
         do
         {
             void * head_pos = head.pos;
