diff --git a/tests/queries/0_stateless/02733_fix_distinct_in_order_bug_49622.reference b/tests/queries/0_stateless/02733_fix_distinct_in_order_bug_49622.reference
new file mode 100644
index 000000000000..19f58f6ca911
--- /dev/null
+++ b/tests/queries/0_stateless/02733_fix_distinct_in_order_bug_49622.reference
@@ -0,0 +1,2 @@
+	1
+	2
diff --git a/tests/queries/0_stateless/02733_fix_distinct_in_order_bug_49622.sql b/tests/queries/0_stateless/02733_fix_distinct_in_order_bug_49622.sql
new file mode 100644
index 000000000000..9501a2c07619
--- /dev/null
+++ b/tests/queries/0_stateless/02733_fix_distinct_in_order_bug_49622.sql
@@ -0,0 +1,15 @@
+set optimize_distinct_in_order=1;
+
+DROP TABLE IF EXISTS test_string;
+
+CREATE TABLE test_string
+(
+    `c1` String,
+    `c2` String
+)
+ENGINE = MergeTree
+ORDER BY c1;
+
+INSERT INTO test_string(c1, c2) VALUES ('1',  ''), ('2', '');
+
+SELECT DISTINCT c2, c1 FROM test_string;
