diff --git a/tests/queries/0_stateless/02456_summing_mt_lc.reference b/tests/queries/0_stateless/02456_summing_mt_lc.reference
new file mode 100644
index 000000000000..439cf103b389
--- /dev/null
+++ b/tests/queries/0_stateless/02456_summing_mt_lc.reference
@@ -0,0 +1,2 @@
+1	6	2020-01-01 00:00:00
+2	6	2020-01-02 00:00:00
diff --git a/tests/queries/0_stateless/02456_summing_mt_lc.sql b/tests/queries/0_stateless/02456_summing_mt_lc.sql
new file mode 100644
index 000000000000..297eb64c6425
--- /dev/null
+++ b/tests/queries/0_stateless/02456_summing_mt_lc.sql
@@ -0,0 +1,20 @@
+SET allow_suspicious_low_cardinality_types = 1;
+
+DROP TABLE IF EXISTS t_summing_lc;
+
+CREATE TABLE t_summing_lc
+(
+    `key` UInt32,
+    `val` LowCardinality(UInt32),
+    `date` DateTime
+)
+ENGINE = SummingMergeTree(val)
+PARTITION BY date
+ORDER BY key;
+
+INSERT INTO t_summing_lc VALUES (1, 1, '2020-01-01'), (2, 1, '2020-01-02'), (1, 5, '2020-01-01'), (2, 5, '2020-01-02');
+
+OPTIMIZE TABLE t_summing_lc FINAL;
+SELECT * FROM t_summing_lc ORDER BY key;
+
+DROP TABLE t_summing_lc;
