diff --git a/tests/queries/0_stateless/03100_lwu_37_update_all_columns.reference b/tests/queries/0_stateless/03100_lwu_37_update_all_columns.reference
new file mode 100644
index 000000000000..d00491fd7e5b
--- /dev/null
+++ b/tests/queries/0_stateless/03100_lwu_37_update_all_columns.reference
@@ -0,0 +1,1 @@
+1
diff --git a/tests/queries/0_stateless/03100_lwu_37_update_all_columns.sql b/tests/queries/0_stateless/03100_lwu_37_update_all_columns.sql
new file mode 100644
index 000000000000..1e85178389b5
--- /dev/null
+++ b/tests/queries/0_stateless/03100_lwu_37_update_all_columns.sql
@@ -0,0 +1,10 @@
+DROP TABLE IF EXISTS t0;
+CREATE TABLE t0 (c0 Int) ENGINE = MergeTree() ORDER BY tuple() SETTINGS enable_block_number_column = 1, enable_block_offset_column = 1;
+
+INSERT INTO TABLE t0 (c0) VALUES (0);
+
+SET allow_experimental_lightweight_update = 1;
+UPDATE t0 SET c0 = 1 WHERE 1;
+SELECT c0 FROM t0 ORDER BY c0;
+
+DROP TABLE IF EXISTS t0;
