diff --git a/test/sql/aggregate/distinct/test_distinct_on.test b/test/sql/aggregate/distinct/test_distinct_on.test
index f5b78ca70443..074e8212f4e2 100644
--- a/test/sql/aggregate/distinct/test_distinct_on.test
+++ b/test/sql/aggregate/distinct/test_distinct_on.test
@@ -131,3 +131,23 @@ statement error
 SELECT DISTINCT ON (2) i FROM integers
 ----
 Binder Error: ORDER term out of range - should be between 1 and 1
+
+# DISTINCT ON constant returns an error
+statement error
+SELECT DISTINCT ON(i, 'literal') i FROM integers
+----
+DISTINCT ON non-integer literal has no effect
+
+statement ok
+SET order_by_non_integer_literal=true
+
+query I
+SELECT DISTINCT ON(i, 'literal') i FROM integers ORDER BY ALL
+----
+2
+4
+
+statement error
+PREPARE v1 AS select distinct on (?) 42;
+----
+Parameter not supported in DISTINCT ON clause
