diff --git a/tests/queries/0_stateless/01073_bad_alter_partition.reference b/tests/queries/0_stateless/01073_bad_alter_partition.reference
index 9008cf749445..580df2ddf089 100644
--- a/tests/queries/0_stateless/01073_bad_alter_partition.reference
+++ b/tests/queries/0_stateless/01073_bad_alter_partition.reference
@@ -1,3 +1,25 @@
+1	2020-01-01
 1	2020-01-02
+1	2020-01-03
+1	2020-01-04
+1	2020-01-05
+1	2020-01-06
+3	2020-01-01
 3	2020-01-02
+3	2020-01-03
+3	2020-01-04
+3	2020-01-05
+3	2020-01-06
+4	2020-01-01
 4	2020-01-02
+4	2020-01-03
+4	2020-01-05
+4	2020-01-06
+5	2020-01-01
+5	2020-01-02
+5	2020-01-03
+5	2020-01-06
+6	2020-01-01
+6	2020-01-02
+6	2020-01-03
+6	2020-01-06
diff --git a/tests/queries/0_stateless/01073_bad_alter_partition.sql b/tests/queries/0_stateless/01073_bad_alter_partition.sql
index 58fc39fd68b5..2e3cd47d6a01 100644
--- a/tests/queries/0_stateless/01073_bad_alter_partition.sql
+++ b/tests/queries/0_stateless/01073_bad_alter_partition.sql
@@ -1,19 +1,22 @@
 DROP TABLE IF EXISTS merge_tree;
 CREATE TABLE merge_tree (d Date) ENGINE = MergeTree ORDER BY d PARTITION BY d;
 
-INSERT INTO merge_tree VALUES ('2020-01-02');
-SELECT 1, * FROM merge_tree;
+INSERT INTO merge_tree VALUES ('2020-01-01'), ('2020-01-02'), ('2020-01-03'), ('2020-01-04'), ('2020-01-05'), ('2020-01-06');
+SELECT 1, * FROM merge_tree ORDER BY d;
 
 -- ALTER TABLE merge_tree DROP PARTITION 2020-01-02; -- This does not even parse
 -- SELECT 2, * FROM merge_tree;
 
-ALTER TABLE merge_tree DROP PARTITION 20200102;
-SELECT 3, * FROM merge_tree;
+ALTER TABLE merge_tree DROP PARTITION 20200103; -- unfortunately, this works, but not as user expected.
+SELECT 3, * FROM merge_tree ORDER BY d;
 
-ALTER TABLE merge_tree DROP PARTITION '20200102'; -- { serverError 38 }
-SELECT 4, * FROM merge_tree;
+ALTER TABLE merge_tree DROP PARTITION '20200104';
+SELECT 4, * FROM merge_tree ORDER BY d;
 
-ALTER TABLE merge_tree DROP PARTITION '2020-01-02';
-SELECT 5, * FROM merge_tree;
+ALTER TABLE merge_tree DROP PARTITION '2020-01-05';
+SELECT 5, * FROM merge_tree ORDER BY d;
+
+ALTER TABLE merge_tree DROP PARTITION '202001-06'; -- { serverError 38 }
+SELECT 6, * FROM merge_tree ORDER BY d;
 
 DROP TABLE merge_tree;
diff --git a/tests/queries/0_stateless/02112_parse_date_yyyymmdd.reference b/tests/queries/0_stateless/02112_parse_date_yyyymmdd.reference
new file mode 100644
index 000000000000..dc49af313b77
--- /dev/null
+++ b/tests/queries/0_stateless/02112_parse_date_yyyymmdd.reference
@@ -0,0 +1,1 @@
+2021-02-03	2021-03-04	2021-04-05
diff --git a/tests/queries/0_stateless/02112_parse_date_yyyymmdd.sh b/tests/queries/0_stateless/02112_parse_date_yyyymmdd.sh
new file mode 100755
index 000000000000..0c18697eb06a
--- /dev/null
+++ b/tests/queries/0_stateless/02112_parse_date_yyyymmdd.sh
@@ -0,0 +1,7 @@
+#!/usr/bin/env bash
+
+CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+# shellcheck source=../shell_config.sh
+. "$CURDIR"/../shell_config.sh
+
+echo '20210203,2021-03-04,20210405' | $CLICKHOUSE_LOCAL --input-format CSV --structure 'a Date, b Date, c Date' --query 'SELECT * FROM table'
