diff --git a/tests/queries/0_stateless/02008_complex_key_range_hashed_dictionary.reference b/tests/queries/0_stateless/02008_complex_key_range_hashed_dictionary.reference
new file mode 100644
index 000000000000..41a12f889060
--- /dev/null
+++ b/tests/queries/0_stateless/02008_complex_key_range_hashed_dictionary.reference
@@ -0,0 +1,58 @@
+Dictionary not nullable
+dictGet
+0.33
+0.42
+0.46
+0.2
+0.4
+dictHas
+1
+1
+1
+0
+select columns from dictionary
+allColumns
+2019-05-05	2019-05-20	1	1	0.33
+2019-05-21	2019-05-30	1	1	0.42
+2019-05-21	2019-05-30	2	2	0.46
+noColumns
+1
+1
+1
+onlySpecificColumns
+1	2019-05-05	0.33
+1	2019-05-21	0.42
+2	2019-05-21	0.46
+onlySpecificColumn
+0.33
+0.42
+0.46
+Dictionary nullable
+dictGet
+0.33
+0.42
+\N
+0.2
+0.4
+dictHas
+1
+1
+1
+0
+select columns from dictionary
+allColumns
+2019-05-05	2019-05-20	1	1	0.33
+2019-05-21	2019-05-30	1	1	0.42
+2019-05-21	2019-05-30	2	2	\N
+noColumns
+1
+1
+1
+onlySpecificColumns
+1	2019-05-05	0.33
+1	2019-05-21	0.42
+2	2019-05-21	\N
+onlySpecificColumn
+0.33
+0.42
+\N
diff --git a/tests/queries/0_stateless/02008_complex_key_range_hashed_dictionary.sql b/tests/queries/0_stateless/02008_complex_key_range_hashed_dictionary.sql
new file mode 100644
index 000000000000..ba6ed30b6092
--- /dev/null
+++ b/tests/queries/0_stateless/02008_complex_key_range_hashed_dictionary.sql
@@ -0,0 +1,109 @@
+DROP TABLE IF EXISTS date_table;
+CREATE TABLE date_table
+(
+  CountryID UInt64,
+  CountryKey String,
+  StartDate Date,
+  EndDate Date,
+  Tax Float64
+)
+ENGINE = MergeTree()
+ORDER BY CountryID;
+
+INSERT INTO date_table VALUES(1, '1', toDate('2019-05-05'), toDate('2019-05-20'), 0.33);
+INSERT INTO date_table VALUES(1, '1', toDate('2019-05-21'), toDate('2019-05-30'), 0.42);
+INSERT INTO date_table VALUES(2, '2', toDate('2019-05-21'), toDate('2019-05-30'), 0.46);
+
+DROP DICTIONARY IF EXISTS range_dictionary;
+CREATE DICTIONARY range_dictionary
+(
+  CountryID UInt64,
+  CountryKey String,
+  StartDate Date,
+  EndDate Date,
+  Tax Float64 DEFAULT 0.2
+)
+PRIMARY KEY CountryID, CountryKey
+SOURCE(CLICKHOUSE(TABLE 'date_table'))
+LIFETIME(MIN 1 MAX 1000)
+LAYOUT(COMPLEX_KEY_RANGE_HASHED())
+RANGE(MIN StartDate MAX EndDate);
+
+SELECT 'Dictionary not nullable';
+SELECT 'dictGet';
+SELECT dictGet('range_dictionary', 'Tax', (toUInt64(1), '1'), toDate('2019-05-15'));
+SELECT dictGet('range_dictionary', 'Tax', (toUInt64(1), '1'), toDate('2019-05-29'));
+SELECT dictGet('range_dictionary', 'Tax', (toUInt64(2), '2'), toDate('2019-05-29'));
+SELECT dictGet('range_dictionary', 'Tax', (toUInt64(2), '2'), toDate('2019-05-31'));
+SELECT dictGetOrDefault('range_dictionary', 'Tax', (toUInt64(2), '2'), toDate('2019-05-31'), 0.4);
+SELECT 'dictHas';
+SELECT dictHas('range_dictionary', (toUInt64(1), '1'), toDate('2019-05-15'));
+SELECT dictHas('range_dictionary', (toUInt64(1), '1'), toDate('2019-05-29'));
+SELECT dictHas('range_dictionary', (toUInt64(2), '2'), toDate('2019-05-29'));
+SELECT dictHas('range_dictionary', (toUInt64(2), '2'), toDate('2019-05-31'));
+SELECT 'select columns from dictionary';
+SELECT 'allColumns';
+SELECT * FROM range_dictionary;
+SELECT 'noColumns';
+SELECT 1 FROM range_dictionary;
+SELECT 'onlySpecificColumns';
+SELECT CountryID, StartDate, Tax FROM range_dictionary;
+SELECT 'onlySpecificColumn';
+SELECT Tax FROM range_dictionary;
+
+DROP TABLE date_table;
+DROP DICTIONARY range_dictionary;
+
+CREATE TABLE date_table
+(
+  CountryID UInt64,
+  CountryKey String,
+  StartDate Date,
+  EndDate Date,
+  Tax Nullable(Float64)
+)
+ENGINE = MergeTree()
+ORDER BY CountryID;
+
+INSERT INTO date_table VALUES(1, '1', toDate('2019-05-05'), toDate('2019-05-20'), 0.33);
+INSERT INTO date_table VALUES(1, '1', toDate('2019-05-21'), toDate('2019-05-30'), 0.42);
+INSERT INTO date_table VALUES(2, '2', toDate('2019-05-21'), toDate('2019-05-30'), NULL);
+
+CREATE DICTIONARY range_dictionary_nullable
+(
+  CountryID UInt64,
+  CountryKey String,
+  StartDate Date,
+  EndDate Date,
+  Tax Nullable(Float64) DEFAULT 0.2
+)
+PRIMARY KEY CountryID, CountryKey
+SOURCE(CLICKHOUSE(TABLE 'date_table'))
+LIFETIME(MIN 1 MAX 1000)
+LAYOUT(COMPLEX_KEY_RANGE_HASHED())
+RANGE(MIN StartDate MAX EndDate);
+
+SELECT 'Dictionary nullable';
+SELECT 'dictGet';
+SELECT dictGet('range_dictionary_nullable', 'Tax', (toUInt64(1), '1'), toDate('2019-05-15'));
+SELECT dictGet('range_dictionary_nullable', 'Tax', (toUInt64(1), '1'), toDate('2019-05-29'));
+SELECT dictGet('range_dictionary_nullable', 'Tax', (toUInt64(2), '2'), toDate('2019-05-29'));
+SELECT dictGet('range_dictionary_nullable', 'Tax', (toUInt64(2), '2'), toDate('2019-05-31'));
+SELECT dictGetOrDefault('range_dictionary_nullable', 'Tax', (toUInt64(2), '2'), toDate('2019-05-31'), 0.4);
+SELECT 'dictHas';
+SELECT dictHas('range_dictionary_nullable', (toUInt64(1), '1'), toDate('2019-05-15'));
+SELECT dictHas('range_dictionary_nullable', (toUInt64(1), '1'), toDate('2019-05-29'));
+SELECT dictHas('range_dictionary_nullable', (toUInt64(2), '2'), toDate('2019-05-29'));
+SELECT dictHas('range_dictionary_nullable', (toUInt64(2), '2'), toDate('2019-05-31'));
+SELECT 'select columns from dictionary';
+SELECT 'allColumns';
+SELECT * FROM range_dictionary_nullable;
+SELECT 'noColumns';
+SELECT 1 FROM range_dictionary_nullable;
+SELECT 'onlySpecificColumns';
+SELECT CountryID, StartDate, Tax FROM range_dictionary_nullable;
+SELECT 'onlySpecificColumn';
+SELECT Tax FROM range_dictionary_nullable;
+
+DROP TABLE date_table;
+DROP DICTIONARY range_dictionary_nullable;
