diff --git a/tests/queries/0_stateless/02155_nested_lc_defalut_bug.reference b/tests/queries/0_stateless/02155_nested_lc_defalut_bug.reference
new file mode 100644
index 000000000000..fe99b0a65850
--- /dev/null
+++ b/tests/queries/0_stateless/02155_nested_lc_defalut_bug.reference
@@ -0,0 +1,1 @@
+1	['a','b']	[3,4]	['','']
diff --git a/tests/queries/0_stateless/02155_nested_lc_defalut_bug.sql b/tests/queries/0_stateless/02155_nested_lc_defalut_bug.sql
new file mode 100644
index 000000000000..45cb9f96b95b
--- /dev/null
+++ b/tests/queries/0_stateless/02155_nested_lc_defalut_bug.sql
@@ -0,0 +1,8 @@
+DROP TABLE IF EXISTS nested_test;
+CREATE TABLE nested_test (x UInt32, `nest.col1` Array(String), `nest.col2` Array(Int8)) ENGINE = MergeTree ORDER BY x;
+
+ALTER TABLE nested_test ADD COLUMN `nest.col3` Array(LowCardinality(String));
+INSERT INTO nested_test (x, `nest.col1`, `nest.col2`) values (1, ['a', 'b'], [3, 4]);
+SELECT * FROM nested_test;
+
+DROP TABLE IF EXISTS nested_test;
