diff --git a/dbms/tests/queries/0_stateless/00420_null_in_scalar_subqueries.reference b/dbms/tests/queries/0_stateless/00420_null_in_scalar_subqueries.reference
index e4076fb6ab0a..167d32f88425 100644
--- a/dbms/tests/queries/0_stateless/00420_null_in_scalar_subqueries.reference
+++ b/dbms/tests/queries/0_stateless/00420_null_in_scalar_subqueries.reference
@@ -5,3 +5,5 @@
 \N
 \N
 \N
+a	b
+\N	1
diff --git a/dbms/tests/queries/0_stateless/00420_null_in_scalar_subqueries.sql b/dbms/tests/queries/0_stateless/00420_null_in_scalar_subqueries.sql
index 77f3fb51b303..4bfa8dc4ecb1 100644
--- a/dbms/tests/queries/0_stateless/00420_null_in_scalar_subqueries.sql
+++ b/dbms/tests/queries/0_stateless/00420_null_in_scalar_subqueries.sql
@@ -5,3 +5,4 @@ SELECT (SELECT Null WHERE nuLL IS NOT NULL);
 SELECT (SELECT Null WHERE 1);
 SELECT CAST(NULL as Null);
 SELECT (SELECT CAST(NULL as Null) WHERE 0);
+SELECT (SELECT 1 WHERE 0) AS a, (SELECT 1 WHERE 1) AS b FORMAT TSVWithNames;
