diff --git a/tests/queries/0_stateless/02426_to_string_nullable_fixedstring.reference b/tests/queries/0_stateless/02426_to_string_nullable_fixedstring.reference
new file mode 100644
index 000000000000..bc0ca9332834
--- /dev/null
+++ b/tests/queries/0_stateless/02426_to_string_nullable_fixedstring.reference
@@ -0,0 +1,11 @@
+a	Nullable(FixedString(1))	a
+0	Nullable(FixedString(1))	\N
+1	Nullable(FixedString(1))	1
+2	Nullable(FixedString(1))	2
+3	Nullable(FixedString(1))	\N
+4	Nullable(FixedString(1))	4
+5	Nullable(FixedString(1))	5
+6	Nullable(FixedString(1))	\N
+7	Nullable(FixedString(1))	7
+8	Nullable(FixedString(1))	8
+9	Nullable(FixedString(1))	\N
diff --git a/tests/queries/0_stateless/02426_to_string_nullable_fixedstring.sql b/tests/queries/0_stateless/02426_to_string_nullable_fixedstring.sql
new file mode 100644
index 000000000000..f6dcc8f92bec
--- /dev/null
+++ b/tests/queries/0_stateless/02426_to_string_nullable_fixedstring.sql
@@ -0,0 +1,2 @@
+SELECT CAST('a', 'Nullable(FixedString(1))') as s,  toTypeName(s), toString(s);
+SELECT number, toTypeName(s), toString(s) FROM (SELECT number, if(number % 3 = 0, NULL, toFixedString(toString(number), 1)) AS s from numbers(10)) ORDER BY number;
