diff --git a/src/Parsers/ExpressionListParsers.cpp b/src/Parsers/ExpressionListParsers.cpp
index f50cf71f54de..afe85f069c76 100644
--- a/src/Parsers/ExpressionListParsers.cpp
+++ b/src/Parsers/ExpressionListParsers.cpp
@@ -18,6 +18,8 @@ const char * ParserMultiplicativeExpression::operators[] =
     "*",     "multiply",
     "/",     "divide",
     "%",     "modulo",
+    "MOD",   "modulo",
+    "DIV",   "intDiv",
     nullptr
 };
 
