diff --git a/tests/queries/0_stateless/00977_int_div.reference b/tests/queries/0_stateless/00977_int_div.reference
index 39e135d8eeda..e5882f4aaba0 100644
--- a/tests/queries/0_stateless/00977_int_div.reference
+++ b/tests/queries/0_stateless/00977_int_div.reference
@@ -81,3 +81,43 @@
 -1
 -1
 -1
+-1
+0
+0
+0
+0
+0
+0
+0
+0
+0
+-1
+-2
+-3
+-4
+-5
+-6
+-7
+-8
+-9
+-10
+-1
+-2
+-3
+-4
+-5
+-6
+-7
+-8
+-9
+-10
+-1
+-1
+-1
+-1
+-1
+-1
+-1
+-1
+-1
+-1
diff --git a/tests/queries/0_stateless/00977_int_div.sql b/tests/queries/0_stateless/00977_int_div.sql
index 08085020ca5f..4184475e3a03 100644
--- a/tests/queries/0_stateless/00977_int_div.sql
+++ b/tests/queries/0_stateless/00977_int_div.sql
@@ -23,3 +23,9 @@ SELECT intDiv(toInt64(number), -1) FROM numbers(1, 10);
 SELECT intDivOrZero(toInt64(number), -1) FROM numbers(1, 10);
 SELECT intDiv(number, -number) FROM numbers(1, 10);
 SELECT intDivOrZero(number, -number) FROM numbers(1, 10);
+
+SELECT -1 DIV number FROM numbers(1, 10);
+SELECT toInt32(number) DIV -1 FROM numbers(1, 10);
+SELECT toInt64(number) DIV -1 FROM numbers(1, 10);
+SELECT number DIV -number FROM numbers(1, 10);
+SELECT -1 DIV 0;  -- { serverError 153 }
diff --git a/tests/queries/0_stateless/01412_mod_float.reference b/tests/queries/0_stateless/01412_mod_float.reference
index 77e3431f7a97..94f7f6d3af71 100644
--- a/tests/queries/0_stateless/01412_mod_float.reference
+++ b/tests/queries/0_stateless/01412_mod_float.reference
@@ -1,3 +1,7 @@
 1	-1	1	-1
 0.125	-0.125	0.125	-0.125
 1	-1	1	-1
+1	-1	1	-1
+0.125	-0.125	0.125	-0.125
+nan
+nan
diff --git a/tests/queries/0_stateless/01412_mod_float.sql b/tests/queries/0_stateless/01412_mod_float.sql
index 147cb88ff240..f2a5f2fcee69 100644
--- a/tests/queries/0_stateless/01412_mod_float.sql
+++ b/tests/queries/0_stateless/01412_mod_float.sql
@@ -1,3 +1,7 @@
 WITH 8.5 AS a, 2.5 AS b SELECT a % b, -a % b, a % -b, -a % -b;
 WITH 10.125 AS a, 2.5 AS b SELECT a % b, -a % b, a % -b, -a % -b;
 WITH 8.5 AS a, 2.5 AS b SELECT mod(a, b), MOD(-a, b), modulo(a, -b), moduloOrZero(-a, -b);
+WITH 8.5 AS a, 2.5 AS b SELECT a MOD b, -a MOD b, a MOD -b, -a MOD -b;
+WITH 10.125 AS a, 2.5 AS b SELECT a MOD b, -a MOD b, a MOD -b, -a MOD -b;
+SELECT 3.5 % 0;
+SELECT 3.5 MOD 0;
