diff --git a/tests/queries/0_stateless/01427_pk_and_expression_with_different_type.reference b/tests/queries/0_stateless/01427_pk_and_expression_with_different_type.reference
new file mode 100644
index 000000000000..e69de29bb2d1
diff --git a/tests/queries/0_stateless/01427_pk_and_expression_with_different_type.sql b/tests/queries/0_stateless/01427_pk_and_expression_with_different_type.sql
new file mode 100644
index 000000000000..d905eac4ce2f
--- /dev/null
+++ b/tests/queries/0_stateless/01427_pk_and_expression_with_different_type.sql
@@ -0,0 +1,4 @@
+DROP TABLE IF EXISTS pk;
+CREATE TABLE pk (x DateTime) ENGINE = MergeTree ORDER BY toStartOfMinute(x) SETTINGS index_granularity = 1;
+SELECT * FROM pk WHERE x >= toDateTime(120) AND x <= toDateTime(NULL);
+DROP TABLE pk;
