diff --git a/tests/queries/0_stateless/02576_predicate_push_down_sorting_fix.reference b/tests/queries/0_stateless/02576_predicate_push_down_sorting_fix.reference
index d391c365ea7b..263a76a33f3a 100644
--- a/tests/queries/0_stateless/02576_predicate_push_down_sorting_fix.reference
+++ b/tests/queries/0_stateless/02576_predicate_push_down_sorting_fix.reference
@@ -10,14 +10,20 @@ Positions: 2
   Header: ignore(2_UInt8) UInt8
           __table2.number UInt64
   Sort description: ignore(2_UInt8) ASC
-    Filter (( + (Before ORDER BY + (Projection + Change column names to column identifiers))))
+    Expression (( + (Before ORDER BY + (Projection + Change column names to column identifiers))))
     Header: ignore(2_UInt8) UInt8
             __table2.number UInt64
-    Filter column: ignore(2_UInt8)
     Actions: INPUT : 0 -> number UInt64 : 0
              COLUMN Const(UInt8) -> 2_UInt8 UInt8 : 1
              ALIAS number :: 0 -> __table2.number UInt64 : 2
              FUNCTION ignore(2_UInt8 :: 1) -> ignore(2_UInt8) UInt8 : 0
     Positions: 0 2
-      ReadFromSystemNumbers
+      Filter ((( + (Before ORDER BY + (Projection + Change column names to column identifiers))))[split])
       Header: number UInt64
+      Filter column: _filter_ignore(2_UInt8)_0 (removed)
+      Actions: COLUMN Const(UInt8) -> 2_UInt8 UInt8 : 0
+               FUNCTION ignore(2_UInt8 :: 0) -> ignore(2_UInt8) UInt8 : 1
+               ALIAS ignore(2_UInt8) :: 1 -> _filter_ignore(2_UInt8)_0 UInt8 : 0
+      Positions: 0
+        ReadFromSystemNumbers
+        Header: number UInt64
diff --git a/tests/queries/0_stateless/03519_fulter_push_down_duplicate_column_name_bug.reference b/tests/queries/0_stateless/03519_fulter_push_down_duplicate_column_name_bug.reference
new file mode 100644
index 000000000000..0ee2d1e918a2
--- /dev/null
+++ b/tests/queries/0_stateless/03519_fulter_push_down_duplicate_column_name_bug.reference
@@ -0,0 +1,10 @@
+0	0
+1	1
+2	1
+3	1
+4	1
+5	1
+6	1
+7	1
+8	1
+9	1
diff --git a/tests/queries/0_stateless/03519_fulter_push_down_duplicate_column_name_bug.sql b/tests/queries/0_stateless/03519_fulter_push_down_duplicate_column_name_bug.sql
new file mode 100644
index 000000000000..8965bf1050c1
--- /dev/null
+++ b/tests/queries/0_stateless/03519_fulter_push_down_duplicate_column_name_bug.sql
@@ -0,0 +1,8 @@
+SELECT
+    number,
+    countIf(1, number > 0)
+FROM numbers(10)
+GROUP BY number
+HAVING (count() <= 10) AND 1
+ORDER BY number ASC
+SETTINGS enable_analyzer = 1;
