diff --git a/tests/queries/0_stateless/00835_if_generic_case.reference b/tests/queries/0_stateless/00835_if_generic_case.reference
index 8f9c3f6ef13b..de2d38897a31 100644
--- a/tests/queries/0_stateless/00835_if_generic_case.reference
+++ b/tests/queries/0_stateless/00835_if_generic_case.reference
@@ -4,8 +4,8 @@
 2000-01-01 00:00:00	2000-01-02	2000-01-02 00:00:00
 2000-01-01 00:00:00	2000-01-02	2000-01-02 00:00:00
 2000-01-01 00:00:00	2000-01-02	2000-01-02 00:00:00
-2000-01-01 00:00:00	2000-01-02	2000-01-02
-2000-01-01 00:00:00	2000-01-02	2000-01-02
+2000-01-01 00:00:00	2000-01-02	2000-01-02 00:00:00
+2000-01-01 00:00:00	2000-01-02	2000-01-02 00:00:00
 2000-01-01 00:00:00	2000-01-02	2000-01-01 00:00:00
 2000-01-01 00:00:00	2000-01-02	2000-01-01 00:00:00
 2000-01-01 00:00:00	2000-01-02	2000-01-01 00:00:00
diff --git a/tests/queries/0_stateless/02901_remove_nullable_crash_analyzer.reference b/tests/queries/0_stateless/02901_remove_nullable_crash_analyzer.reference
new file mode 100644
index 000000000000..44e0be8e3569
--- /dev/null
+++ b/tests/queries/0_stateless/02901_remove_nullable_crash_analyzer.reference
@@ -0,0 +1,4 @@
+0
+0
+0
+0
diff --git a/tests/queries/0_stateless/02901_remove_nullable_crash_analyzer.sql b/tests/queries/0_stateless/02901_remove_nullable_crash_analyzer.sql
new file mode 100644
index 000000000000..96f88a23a098
--- /dev/null
+++ b/tests/queries/0_stateless/02901_remove_nullable_crash_analyzer.sql
@@ -0,0 +1,6 @@
+SELECT 1 % ( CASE WHEN 1 THEN (1 IS NOT NULL + *) ELSE NULL END );
+SELECT CASE 1 WHEN FALSE THEN 1 ELSE CASE WHEN 1 THEN 1 - (CASE 1 WHEN 1 THEN 1 ELSE 1 END) END % 1 END;
+
+SELECT 1 % if(1, dummy, NULL); -- { serverError ILLEGAL_DIVISION }
+SELECT sum(multiIf(1, dummy, NULL));
+SELECT sum(multiIf(1, dummy, NULL)) OVER ();
