diff --git a/data/parquet-testing/hive-partitioning/duplicate_names/x=1/x=2/test.parquet b/data/parquet-testing/hive-partitioning/duplicate_names/x=1/x=2/test.parquet
new file mode 100644
index 000000000000..ea9710486eac
Binary files /dev/null and b/data/parquet-testing/hive-partitioning/duplicate_names/x=1/x=2/test.parquet differ
diff --git a/data/parquet-testing/hive-partitioning/duplicate_names/x=2/x=1/test.parquet b/data/parquet-testing/hive-partitioning/duplicate_names/x=2/x=1/test.parquet
new file mode 100644
index 000000000000..afa7bb5c0ed6
Binary files /dev/null and b/data/parquet-testing/hive-partitioning/duplicate_names/x=2/x=1/test.parquet differ
diff --git a/data/parquet-testing/hive-partitioning/empty_string/key=NULL/part.parquet b/data/parquet-testing/hive-partitioning/empty_string/key=NULL/part.parquet
new file mode 100644
index 000000000000..97e54f67ee29
Binary files /dev/null and b/data/parquet-testing/hive-partitioning/empty_string/key=NULL/part.parquet differ
diff --git a/test/sql/copy/parquet/parquet_hive_empty.test b/test/sql/copy/parquet/parquet_hive_empty.test
index ab11846371b7..4e7f0d2fc999 100644
--- a/test/sql/copy/parquet/parquet_hive_empty.test
+++ b/test/sql/copy/parquet/parquet_hive_empty.test
@@ -6,9 +6,32 @@ require parquet
 
 query II
 select * 
-from parquet_scan('data/parquet-testing/hive-partitioning/empty_string/*/*.parquet', hive_partitioning=1)
+from parquet_scan('data/parquet-testing/hive-partitioning/empty_string/*/*.parquet')
 ORDER BY ALL
 ----
 a	a
 b	(empty)
+c	NULL
 
+# filter on hive partitioning with NULL values
+query II
+select *
+from parquet_scan('data/parquet-testing/hive-partitioning/empty_string/*/*.parquet')
+WHERE key IS NULL
+----
+c	NULL
+
+
+query II
+select *
+from parquet_scan('data/parquet-testing/hive-partitioning/empty_string/*/*.parquet')
+WHERE key='a'
+----
+a	a
+
+query II
+select *
+from parquet_scan('data/parquet-testing/hive-partitioning/empty_string/*/*.parquet')
+WHERE key=''
+----
+b	(empty)
diff --git a/test/sql/copy/partitioned/hive_partition_duplicate_name.test b/test/sql/copy/partitioned/hive_partition_duplicate_name.test
new file mode 100644
index 000000000000..2f3f52e8f043
--- /dev/null
+++ b/test/sql/copy/partitioned/hive_partition_duplicate_name.test
@@ -0,0 +1,15 @@
+# name: test/sql/copy/partitioned/hive_partition_duplicate_name.test
+# description: Test partitioning names with duplicate keys
+# group: [partitioned]
+
+require parquet
+
+# we just use the first partitioning key by default
+query III
+select *
+from parquet_scan('data/parquet-testing/hive-partitioning/duplicate_names/**/*.parquet')
+ORDER BY ALL
+----
+1	value1	1
+2	value2	2
+
