diff --git a/tests/queries/0_stateless/01913_replace_dictionary.reference b/tests/queries/0_stateless/01913_replace_dictionary.reference
new file mode 100644
index 000000000000..2d33c16ccc26
--- /dev/null
+++ b/tests/queries/0_stateless/01913_replace_dictionary.reference
@@ -0,0 +1,2 @@
+0	Value0
+0	Value1
diff --git a/tests/queries/0_stateless/01913_replace_dictionary.sql b/tests/queries/0_stateless/01913_replace_dictionary.sql
new file mode 100644
index 000000000000..22b0bd002ae7
--- /dev/null
+++ b/tests/queries/0_stateless/01913_replace_dictionary.sql
@@ -0,0 +1,50 @@
+DROP DATABASE IF EXISTS 01913_db;
+CREATE DATABASE 01913_db ENGINE=Atomic;
+
+DROP TABLE IF EXISTS 01913_db.test_source_table_1;
+CREATE TABLE 01913_db.test_source_table_1
+(
+    id UInt64,
+    value String
+) ENGINE=TinyLog;
+
+INSERT INTO 01913_db.test_source_table_1 VALUES (0, 'Value0');
+
+DROP DICTIONARY IF EXISTS 01913_db.test_dictionary;
+CREATE DICTIONARY 01913_db.test_dictionary
+(
+    id UInt64,
+    value String
+)
+PRIMARY KEY id
+LAYOUT(DIRECT())
+SOURCE(CLICKHOUSE(DB '01913_db' TABLE 'test_source_table_1'));
+
+SELECT * FROM 01913_db.test_dictionary;
+
+DROP TABLE IF EXISTS 01913_db.test_source_table_2;
+CREATE TABLE 01913_db.test_source_table_2
+(
+    id UInt64,
+    value_1 String
+) ENGINE=TinyLog;
+
+INSERT INTO 01913_db.test_source_table_2 VALUES (0, 'Value1');
+
+REPLACE DICTIONARY 01913_db.test_dictionary
+(
+    id UInt64,
+    value_1 String
+)
+PRIMARY KEY id
+LAYOUT(HASHED())
+SOURCE(CLICKHOUSE(DB '01913_db' TABLE 'test_source_table_2'))
+LIFETIME(0);
+
+SELECT * FROM 01913_db.test_dictionary;
+
+DROP DICTIONARY 01913_db.test_dictionary;
+DROP TABLE 01913_db.test_source_table_1;
+DROP TABLE 01913_db.test_source_table_2;
+
+DROP DATABASE 01913_db;
diff --git a/tests/queries/0_stateless/01914_exchange_dictionaries.reference b/tests/queries/0_stateless/01914_exchange_dictionaries.reference
new file mode 100644
index 000000000000..9278d0abeed4
--- /dev/null
+++ b/tests/queries/0_stateless/01914_exchange_dictionaries.reference
@@ -0,0 +1,4 @@
+1	Table1
+2	Table2
+2	Table2
+1	Table1
diff --git a/tests/queries/0_stateless/01914_exchange_dictionaries.sql b/tests/queries/0_stateless/01914_exchange_dictionaries.sql
new file mode 100644
index 000000000000..ba0c70d13be4
--- /dev/null
+++ b/tests/queries/0_stateless/01914_exchange_dictionaries.sql
@@ -0,0 +1,34 @@
+DROP TABLE IF EXISTS table_1;
+CREATE TABLE table_1 (id UInt64, value String) ENGINE=TinyLog;
+
+DROP TABLE IF EXISTS table_2;
+CREATE TABLE table_2 (id UInt64, value String) ENGINE=TinyLog;
+
+INSERT INTO table_1 VALUES (1, 'Table1');
+INSERT INTO table_2 VALUES (2, 'Table2');
+
+DROP DICTIONARY IF EXISTS dictionary_1;
+CREATE DICTIONARY dictionary_1 (id UInt64, value String)
+PRIMARY KEY id
+LAYOUT(DIRECT())
+SOURCE(CLICKHOUSE(TABLE 'table_1'));
+
+DROP DICTIONARY IF EXISTS dictionary_2;
+CREATE DICTIONARY dictionary_2 (id UInt64, value String)
+PRIMARY KEY id
+LAYOUT(DIRECT())
+SOURCE(CLICKHOUSE(TABLE 'table_2'));
+
+SELECT * FROM dictionary_1;
+SELECT * FROM dictionary_2;
+
+EXCHANGE DICTIONARIES dictionary_1 AND dictionary_2;
+
+SELECT * FROM dictionary_1;
+SELECT * FROM dictionary_2;
+
+DROP DICTIONARY dictionary_1;
+DROP DICTIONARY dictionary_2;
+
+DROP TABLE table_1;
+DROP TABLE table_2;
diff --git a/tests/queries/0_stateless/01915_create_or_replace_dictionary.reference b/tests/queries/0_stateless/01915_create_or_replace_dictionary.reference
new file mode 100644
index 000000000000..2d33c16ccc26
--- /dev/null
+++ b/tests/queries/0_stateless/01915_create_or_replace_dictionary.reference
@@ -0,0 +1,2 @@
+0	Value0
+0	Value1
diff --git a/tests/queries/0_stateless/01915_create_or_replace_dictionary.sql b/tests/queries/0_stateless/01915_create_or_replace_dictionary.sql
new file mode 100644
index 000000000000..5d5515f4f8a9
--- /dev/null
+++ b/tests/queries/0_stateless/01915_create_or_replace_dictionary.sql
@@ -0,0 +1,50 @@
+DROP DATABASE IF EXISTS 01915_db;
+CREATE DATABASE 01915_db ENGINE=Atomic;
+
+DROP TABLE IF EXISTS 01915_db.test_source_table_1;
+CREATE TABLE 01915_db.test_source_table_1
+(
+    id UInt64,
+    value String
+) ENGINE=TinyLog;
+
+INSERT INTO 01915_db.test_source_table_1 VALUES (0, 'Value0');
+
+DROP DICTIONARY IF EXISTS 01915_db.test_dictionary;
+CREATE OR REPLACE DICTIONARY 01915_db.test_dictionary
+(
+    id UInt64,
+    value String
+)
+PRIMARY KEY id
+LAYOUT(DIRECT())
+SOURCE(CLICKHOUSE(DB '01915_db' TABLE 'test_source_table_1'));
+
+SELECT * FROM 01915_db.test_dictionary;
+
+DROP TABLE IF EXISTS 01915_db.test_source_table_2;
+CREATE TABLE 01915_db.test_source_table_2
+(
+    id UInt64,
+    value_1 String
+) ENGINE=TinyLog;
+
+INSERT INTO 01915_db.test_source_table_2 VALUES (0, 'Value1');
+
+CREATE OR REPLACE DICTIONARY 01915_db.test_dictionary
+(
+    id UInt64,
+    value_1 String
+)
+PRIMARY KEY id
+LAYOUT(HASHED())
+SOURCE(CLICKHOUSE(DB '01915_db' TABLE 'test_source_table_2'))
+LIFETIME(0);
+
+SELECT * FROM 01915_db.test_dictionary;
+
+DROP DICTIONARY 01915_db.test_dictionary;
+DROP TABLE 01915_db.test_source_table_1;
+DROP TABLE 01915_db.test_source_table_2;
+
+DROP DATABASE 01915_db;
diff --git a/tests/queries/skip_list.json b/tests/queries/skip_list.json
index 52c2d4684981..fcfd5192ce9e 100644
--- a/tests/queries/skip_list.json
+++ b/tests/queries/skip_list.json
@@ -516,7 +516,10 @@
         "01913_if_int_decimal",
         "01913_join_push_down_bug",
         "01921_with_fill_with_totals",
-        "01924_argmax_bitmap_state"
+        "01924_argmax_bitmap_state",
+        "01913_replace_dictionary",
+        "01914_exchange_dictionaries",
+        "01915_create_or_replace_dictionary"
     ],
     "parallel":
     [
