diff --git a/tools/shell/shell-test.py b/tools/shell/shell-test.py
index fd46a1845953..222abddfa027 100644
--- a/tools/shell/shell-test.py
+++ b/tools/shell/shell-test.py
@@ -414,6 +414,19 @@ def tf():
 
 shutil.rmtree(target_dir)
 
+# test using .import with a CSV file containing invalid UTF8
+
+duckdb_nonsensecsv = 'duckdbtest_nonsensecsv.csv'
+with open(duckdb_nonsensecsv, 'wb+') as f:
+     f.write(b'\xFF
')
+test('''
+.nullvalue NULL
+CREATE TABLE test(i INTEGER);
+.import duckdbtest_nonsensecsv.csv test
+SELECT * FROM test;
+''', out="NULL")
+os.remove(duckdb_nonsensecsv)
+
 # dump blobs: FIXME
 # test('''
 # CREATE TABLE a (b BLOB);
