diff --git a/tests/queries/0_stateless/03537_kusto_ubsan.reference b/tests/queries/0_stateless/03537_kusto_ubsan.reference
new file mode 100644
index 000000000000..e69de29bb2d1
diff --git a/tests/queries/0_stateless/03537_kusto_ubsan.sql b/tests/queries/0_stateless/03537_kusto_ubsan.sql
new file mode 100644
index 000000000000..f1573eb9f780
--- /dev/null
+++ b/tests/queries/0_stateless/03537_kusto_ubsan.sql
@@ -0,0 +1,2 @@
+SET interval_output_format = 'kusto';
+SELECT INTERVAL 3508 MONTH; -- { clientError BAD_ARGUMENTS }
