diff --git a/src/facade/reply_builder.h b/src/facade/reply_builder.h
index 534ac2f036d8..181e0338f742 100644
--- a/src/facade/reply_builder.h
+++ b/src/facade/reply_builder.h
@@ -442,7 +442,7 @@ class RedisReplyBuilder2Base : public SinkReplyBuilder2, public RedisReplyBuilde
   }
 
   void StartAggregate() override {
-    aggregators_.emplace_back(SinkReplyBuilder2::ReplyAggregator(this));
+    aggregators_.emplace_back(this);
   }
 
   void StopAggregate() override {
