diff --git a/src/storage/local_storage.cpp b/src/storage/local_storage.cpp
index 83618f33b69a..f44527e6cd89 100644
--- a/src/storage/local_storage.cpp
+++ b/src/storage/local_storage.cpp
@@ -440,6 +440,9 @@ void LocalStorage::Flush(DataTable &table, LocalTableStorage &storage) {
 		return;
 	}
 	if (storage.row_groups->GetTotalRows() <= storage.deleted_rows) {
+		// all rows that we added were deleted
+		// rollback any partial blocks that are still outstanding
+		storage.Rollback();
 		return;
 	}
 	idx_t append_count = storage.row_groups->GetTotalRows() - storage.deleted_rows;
