diff --git a/tests/queries/0_stateless/01398_any_with_alias.sql b/tests/queries/0_stateless/01398_any_with_alias.sql
index 32e67f3f4b6a..a65b8132c671 100644
--- a/tests/queries/0_stateless/01398_any_with_alias.sql
+++ b/tests/queries/0_stateless/01398_any_with_alias.sql
@@ -1,3 +1,5 @@
+SET optimize_move_functions_out_of_any = 1;
+
 SELECT any(number * number) AS n FROM numbers(100) FORMAT CSVWithNames;
 EXPLAIN SYNTAX SELECT any(number * number) AS n FROM numbers(100);
 
diff --git a/tests/queries/0_stateless/01470_columns_transformers.reference b/tests/queries/0_stateless/01470_columns_transformers.reference
index a103d62167b4..ae0adb3ba602 100644
--- a/tests/queries/0_stateless/01470_columns_transformers.reference
+++ b/tests/queries/0_stateless/01470_columns_transformers.reference
@@ -27,9 +27,9 @@ SELECT
     avg(k)
 FROM columns_transformers
 SELECT
-    toDate(any(i)),
-    toDate(any(j)),
-    toDate(any(k))
+    any(toDate(i)),
+    any(toDate(j)),
+    any(toDate(k))
 FROM columns_transformers AS a
 SELECT
     length(toString(j)),
@@ -44,9 +44,9 @@ SELECT
     avg(k)
 FROM columns_transformers
 SELECT
-    toDate(any(i)),
-    toDate(any(j)),
-    toDate(any(k))
+    any(toDate(i)),
+    any(toDate(j)),
+    any(toDate(k))
 FROM columns_transformers AS a
 SELECT
     sum(i + 1 AS i),
@@ -59,9 +59,9 @@ SELECT
     avg(k)
 FROM columns_transformers
 SELECT
-    toDate(any(i)),
-    toDate(any(j)),
-    toDate(any(k))
+    any(toDate(i)),
+    any(toDate(j)),
+    any(toDate(k))
 FROM columns_transformers AS a
 SELECT
     (i + 1) + 1 AS i,
diff --git a/tests/queries/0_stateless/01591_window_functions.reference b/tests/queries/0_stateless/01591_window_functions.reference
index ce56860ed8bf..aad784b1ac1e 100644
--- a/tests/queries/0_stateless/01591_window_functions.reference
+++ b/tests/queries/0_stateless/01591_window_functions.reference
@@ -122,7 +122,7 @@ select * from (select * from numbers(5) order by rand()) order by count() over (
 2
 3
 4
-select * from (select * from numbers(5) order by rand()) group by number order by sum(any(number + 1)) over (order by min(number) desc) desc;
+select * from (select * from numbers(5) order by rand()) group by number order by sum(any(number) + 1) over (order by min(number) desc) desc;
 
 -- different windows
 -- an explain test would also be helpful, but it's too immature now and I don't
diff --git a/tests/queries/0_stateless/01591_window_functions.sql b/tests/queries/0_stateless/01591_window_functions.sql
index 082a6652a650..30094690a92f 100644
--- a/tests/queries/0_stateless/01591_window_functions.sql
+++ b/tests/queries/0_stateless/01591_window_functions.sql
@@ -41,7 +41,7 @@ select * from (select * from numbers(5) order by rand()) order by count() over (
 -- Aggregate functions as window function arguments. This query is semantically
 -- the same as the above one, only we replace `number` with
 -- `any(number) group by number` and so on.
-select * from (select * from numbers(5) order by rand()) group by number order by sum(any(number + 1)) over (order by min(number) desc) desc;
+select * from (select * from numbers(5) order by rand()) group by number order by sum(any(number) + 1) over (order by min(number) desc) desc;
 
 -- different windows
 -- an explain test would also be helpful, but it's too immature now and I don't
diff --git a/tests/queries/0_stateless/01650_any_null_if.reference b/tests/queries/0_stateless/01650_any_null_if.reference
new file mode 100644
index 000000000000..e965047ad7c5
--- /dev/null
+++ b/tests/queries/0_stateless/01650_any_null_if.reference
@@ -0,0 +1,1 @@
+Hello
diff --git a/tests/queries/0_stateless/01650_any_null_if.sql b/tests/queries/0_stateless/01650_any_null_if.sql
new file mode 100644
index 000000000000..17f57e92032b
--- /dev/null
+++ b/tests/queries/0_stateless/01650_any_null_if.sql
@@ -0,0 +1,6 @@
+SELECT any(nullIf(s, '')) FROM (SELECT arrayJoin(['', 'Hello']) AS s);
+
+SET optimize_move_functions_out_of_any = 0;
+EXPLAIN SYNTAX select any(nullIf('', ''), 'some text'); -- { serverError 42 }
+SET optimize_move_functions_out_of_any = 1;
+EXPLAIN SYNTAX select any(nullIf('', ''), 'some text'); -- { serverError 42 }
