diff --git a/tests/queries/0_stateless/00700_decimal_arithm.reference b/tests/queries/0_stateless/00700_decimal_arithm.reference
index 811946c87e06..109c0632fb1f 100644
--- a/tests/queries/0_stateless/00700_decimal_arithm.reference
+++ b/tests/queries/0_stateless/00700_decimal_arithm.reference
@@ -10,18 +10,18 @@
 63	21	-42	882	-882	2	0	2	0
 63	21	-42	882	-882	2	0	2	0
 1.00305798474369219219752355409390731264	-0.16305798474369219219752355409390731264	1.490591730234615865843651857942052864	-1.38847100762815390390123822295304634368	1.38847100762815390390123822295304634368	0.02	0.005
-63.42	21.42	-41.58	890.82	-890.82	2.02	0.505	2.02	0.505
-63.42	21.42	-41.58	890.82	-890.82	2.02	0.505	2.02	0.505
-63.42	21.42	-41.58	890.82	-890.82	2.02	0.505	2.02	0.505
-63.42	21.42	-41.58	890.82	-890.82	2.02	0.5	2.02	0.5
+63.42	21.42	-41.58	890.82	-890.82	2.02	0.505	2	0
+63.42	21.42	-41.58	890.82	-890.82	2.02	0.505	2	0
+63.42	21.42	-41.58	890.82	-890.82	2.02	0.505	2	0
+63.42	21.42	-41.58	890.82	-890.82	2.02	0.5	2	0
 63	-21	42	882	-882	0	2	0	2
 63	-21	42	882	-882	0	2	0	2
 63	-21	42	882	-882	0	2	0	2
 1.00305798474369219219752355409390731264	0.16305798474369219219752355409390731264	-1.490591730234615865843651857942052864	-1.38847100762815390390123822295304634368	1.38847100762815390390123822295304634368	-0.00000000000000000000000000000000000001	0.00000000000000000000000000000000000001
-63.42	-21.42	41.58	890.82	-890.82	0.495	1.98	0.495	1.98
+63.42	-21.42	41.58	890.82	-890.82	0.495	1.98	0	1
 63.42	-21.42	41.58	890.82	-890.82
-63.42	-21.42	41.58	890.82	-890.82	0.495049504950495049	1.980198019801980198	0.495049504950495049	1.980198019801980198
-63.42	-21.42	41.58	890.82	-890.82	0.49	1.98	0.49	1.98
+63.42	-21.42	41.58	890.82	-890.82	0.495049504950495049	1.980198019801980198	0	1
+63.42	-21.42	41.58	890.82	-890.82	0.49	1.98	0	1
 -42	42	42	42	0.42	0.42	0.42	42.42	42.42	42.42
 0	0	0	0	0	0	0	0	0	0
 42	-42	-42	-42	-0.42	-0.42	-0.42	-42.42	-42.42	-42.42
diff --git a/tests/queries/0_stateless/01717_int_div_float_too_large_ubsan.sql b/tests/queries/0_stateless/01717_int_div_float_too_large_ubsan.sql
index c4f26a079f09..dc1e5b370506 100644
--- a/tests/queries/0_stateless/01717_int_div_float_too_large_ubsan.sql
+++ b/tests/queries/0_stateless/01717_int_div_float_too_large_ubsan.sql
@@ -1,2 +1,2 @@
-SELECT intDiv(9223372036854775807, 0.9998999834060669); -- { serverError 153 }
-SELECT intDiv(9223372036854775807, 1.);  -- { serverError 153 }
+SELECT intDiv(18446744073709551615, 0.9998999834060669); -- { serverError 153 }
+SELECT intDiv(18446744073709551615, 1.);  -- { serverError 153 }
diff --git a/tests/queries/0_stateless/02975_intdiv_with_decimal.reference b/tests/queries/0_stateless/02975_intdiv_with_decimal.reference
new file mode 100644
index 000000000000..5540734ae4ca
--- /dev/null
+++ b/tests/queries/0_stateless/02975_intdiv_with_decimal.reference
@@ -0,0 +1,68 @@
+2
+2
+1
+2
+2
+2
+2
+2
+2
+2
+2
+1
+2
+1
+2
+1
+2
+2
+2
+2
+2
+2
+2
+2
+2
+2
+1
+1
+1
+1
+2
+2
+2
+2
+2
+2
+1
+2
+2
+2
+2
+2
+2
+2
+2
+1
+2
+1
+2
+1
+2
+2
+2
+2
+2
+2
+2
+2
+2
+2
+1
+1
+1
+1
+2
+2
+2
+2
diff --git a/tests/queries/0_stateless/02975_intdiv_with_decimal.sql b/tests/queries/0_stateless/02975_intdiv_with_decimal.sql
new file mode 100644
index 000000000000..0911a4812515
--- /dev/null
+++ b/tests/queries/0_stateless/02975_intdiv_with_decimal.sql
@@ -0,0 +1,70 @@
+--intDiv--
+SELECT intDiv(4,2);
+SELECT intDiv(toDecimal32(4.4, 2), 2);
+SELECT intDiv(4, toDecimal32(2.2, 2));
+SELECT intDiv(toDecimal32(4.4, 2), 2);
+SELECT intDiv(toDecimal32(4.4, 2), toDecimal32(2.2, 2));
+SELECT intDiv(toDecimal64(4.4, 3), 2);
+SELECT intDiv(toDecimal64(4.4, 3), toDecimal32(2.2, 2));
+SELECT intDiv(toDecimal128(4.4, 4), 2);
+SELECT intDiv(toDecimal128(4.4, 4), toDecimal32(2.2, 2));
+SELECT intDiv(toDecimal256(4.4, 5), 2);
+SELECT intDiv(toDecimal256(4.4, 5), toDecimal32(2.2, 2));
+SELECT intDiv(4, toDecimal64(2.2, 2));
+SELECT intDiv(toDecimal32(4.4, 2), toDecimal64(2.2, 2));
+SELECT intDiv(4, toDecimal128(2.2, 3));
+SELECT intDiv(toDecimal32(4.4, 2), toDecimal128(2.2, 2));
+SELECT intDiv(4, toDecimal256(2.2, 4));
+SELECT intDiv(toDecimal32(4.4, 2), toDecimal256(2.2, 2));
+SELECT intDiv(toDecimal64(4.4, 2), toDecimal64(2.2, 2));
+SELECT intDiv(toDecimal128(4.4, 2), toDecimal64(2.2, 2));
+SELECT intDiv(toDecimal256(4.4, 2), toDecimal64(2.2, 2));
+SELECT intDiv(toDecimal64(4.4, 2), toDecimal128(2.2, 2));
+SELECT intDiv(toDecimal128(4.4, 2), toDecimal128(2.2, 2));
+SELECT intDiv(toDecimal256(4.4, 2), toDecimal128(2.2, 2));
+SELECT intDiv(toDecimal64(4.4, 2), toDecimal256(2.2, 2));
+SELECT intDiv(toDecimal128(4.4, 2), toDecimal256(2.2, 2));
+SELECT intDiv(toDecimal256(4.4, 2), toDecimal256(2.2, 2));
+SELECT intDiv(4.2, toDecimal32(2.2, 2));
+SELECT intDiv(4.2, toDecimal64(2.2, 2));
+SELECT intDiv(4.2, toDecimal128(2.2, 2));
+SELECT intDiv(4.2, toDecimal256(2.2, 2));
+SELECT intDiv(toDecimal32(4.4, 2), 2.2);
+SELECT intDiv(toDecimal64(4.4, 2), 2.2);
+SELECT intDiv(toDecimal128(4.4, 2), 2.2);
+SELECT intDiv(toDecimal256(4.4, 2), 2.2);
+--intDivOrZero--
+SELECT intDivOrZero(4,2);
+SELECT intDivOrZero(toDecimal32(4.4, 2), 2);
+SELECT intDivOrZero(4, toDecimal32(2.2, 2));
+SELECT intDivOrZero(toDecimal32(4.4, 2), 2);
+SELECT intDivOrZero(toDecimal32(4.4, 2), toDecimal32(2.2, 2));
+SELECT intDivOrZero(toDecimal64(4.4, 3), 2);
+SELECT intDivOrZero(toDecimal64(4.4, 3), toDecimal32(2.2, 2));
+SELECT intDivOrZero(toDecimal128(4.4, 4), 2);
+SELECT intDivOrZero(toDecimal128(4.4, 4), toDecimal32(2.2, 2));
+SELECT intDivOrZero(toDecimal256(4.4, 5), 2);
+SELECT intDivOrZero(toDecimal256(4.4, 5), toDecimal32(2.2, 2));
+SELECT intDivOrZero(4, toDecimal64(2.2, 2));
+SELECT intDivOrZero(toDecimal32(4.4, 2), toDecimal64(2.2, 2));
+SELECT intDivOrZero(4, toDecimal128(2.2, 3));
+SELECT intDivOrZero(toDecimal32(4.4, 2), toDecimal128(2.2, 2));
+SELECT intDivOrZero(4, toDecimal256(2.2, 4));
+SELECT intDivOrZero(toDecimal32(4.4, 2), toDecimal256(2.2, 2));
+SELECT intDivOrZero(toDecimal64(4.4, 2), toDecimal64(2.2, 2));
+SELECT intDivOrZero(toDecimal128(4.4, 2), toDecimal64(2.2, 2));
+SELECT intDivOrZero(toDecimal256(4.4, 2), toDecimal64(2.2, 2));
+SELECT intDivOrZero(toDecimal64(4.4, 2), toDecimal128(2.2, 2));
+SELECT intDivOrZero(toDecimal128(4.4, 2), toDecimal128(2.2, 2));
+SELECT intDivOrZero(toDecimal256(4.4, 2), toDecimal128(2.2, 2));
+SELECT intDivOrZero(toDecimal64(4.4, 2), toDecimal256(2.2, 2));
+SELECT intDivOrZero(toDecimal128(4.4, 2), toDecimal256(2.2, 2));
+SELECT intDivOrZero(toDecimal256(4.4, 2), toDecimal256(2.2, 2));
+SELECT intDivOrZero(4.2, toDecimal32(2.2, 2));
+SELECT intDivOrZero(4.2, toDecimal64(2.2, 2));
+SELECT intDivOrZero(4.2, toDecimal128(2.2, 2));
+SELECT intDivOrZero(4.2, toDecimal256(2.2, 2));
+SELECT intDivOrZero(toDecimal32(4.4, 2), 2.2);
+SELECT intDivOrZero(toDecimal64(4.4, 2), 2.2);
+SELECT intDivOrZero(toDecimal128(4.4, 2), 2.2);
+SELECT intDivOrZero(toDecimal256(4.4, 2), 2.2);
