diff --git a/tools/rpkg/tests/testthat/test_dbinfo.R b/tools/rpkg/tests/testthat/test_dbinfo.R
new file mode 100644
index 000000000000..fee51cc2d459
--- /dev/null
+++ b/tools/rpkg/tests/testthat/test_dbinfo.R
@@ -0,0 +1,17 @@
+library("testthat")
+library("DBI")
+test_that("dbGetInfo returns something meaningful", {
+  dbdir <- tempfile()
+  drv <- duckdb::duckdb(dbdir)
+
+  info_drv <- dbGetInfo(drv)
+  expect_equal(info_drv$dbname, dbdir)
+  expect_true(grepl("\\d+\\.\\d+\\.\\d+", info_drv$client.version))
+  expect_true(grepl("\\d+\\.\\d+\\.\\d+", info_drv$driver.version))
+
+  con <- dbConnect(drv)
+  on.exit(dbDisconnect(con, shutdown = TRUE))
+  info_con <- dbGetInfo(con)
+  expect_equal(info_con$dbname, dbdir)
+  expect_true(grepl("\\d+\\.\\d+\\.\\d+", info_con$db.version))
+})
\ No newline at end of file
diff --git a/tools/rpkg/tests/testthat/test_timezone.R b/tools/rpkg/tests/testthat/test_timezone.R
index 2ba3ed638e75..b2ca5760ccdd 100644
--- a/tools/rpkg/tests/testthat/test_timezone.R
+++ b/tools/rpkg/tests/testthat/test_timezone.R
@@ -26,6 +26,8 @@ test_that("timezone_out works with a specified timezone", {
 })
 
 test_that("timezone_out works with '' and converts to local timezime", {
+  unlockBinding(".sys.timezone", baseenv())
+  on.exit(lockBinding(".sys.timezone", baseenv()), add = TRUE)
   withr::local_timezone("Pacific/Tahiti")
   con <- dbConnect(duckdb(), timezone_out = "")
   on.exit(dbDisconnect(con, shutdown = TRUE))
@@ -36,6 +38,8 @@ test_that("timezone_out works with '' and converts to local timezime", {
 })
 
 test_that("timezone_out works with Sys.timezone", {
+  unlockBinding(".sys.timezone", baseenv())
+  on.exit(lockBinding(".sys.timezone", baseenv()), add = TRUE)
   withr::local_timezone("Pacific/Tahiti")
   con <- dbConnect(duckdb(), timezone_out = Sys.timezone())
   on.exit(dbDisconnect(con, shutdown = TRUE))
@@ -64,6 +68,8 @@ test_that("timezone_out works with a specified timezone and tz_out_convert = 'fo
 })
 
 test_that("timezone_out works with '' and tz_out_convert = 'force': forces local timezime", {
+  unlockBinding(".sys.timezone", baseenv())
+  on.exit(lockBinding(".sys.timezone", baseenv()), add = TRUE)
   withr::local_timezone("Pacific/Tahiti")
   con <- dbConnect(duckdb(), timezone_out = "", tz_out_convert = 'force')
   on.exit(dbDisconnect(con, shutdown = TRUE))
@@ -74,6 +80,8 @@ test_that("timezone_out works with '' and tz_out_convert = 'force': forces local
 })
 
 test_that("timezone_out works with a specified local timezone and tz_out_convert = 'force': forces local timezime", {
+  unlockBinding(".sys.timezone", baseenv())
+  on.exit(lockBinding(".sys.timezone", baseenv()), add = TRUE)
   withr::local_timezone("Pacific/Tahiti")
   con <- dbConnect(duckdb(), timezone_out = Sys.timezone(), tz_out_convert = 'force')
   on.exit(dbDisconnect(con, shutdown = TRUE))
@@ -98,4 +106,3 @@ test_that("timezone_out gives a warning with NULL timezone, and converts to UTC"
 test_that("dbConnect fails when tz_out_convert is misspecified", {
   expect_error(dbConnect(duckdb(), tz_out_convert = "nope"))
 })
-
