diff --git a/tests/queries/0_stateless/02766_bitshift_with_const_arguments.sql b/tests/queries/0_stateless/02766_bitshift_with_const_arguments.sql
index 91e8624057c4..6b2961f05555 100644
--- a/tests/queries/0_stateless/02766_bitshift_with_const_arguments.sql
+++ b/tests/queries/0_stateless/02766_bitshift_with_const_arguments.sql
@@ -10,7 +10,7 @@ DROP TABLE IF EXISTS t1;
 CREATE TABLE t0 (vkey UInt32, pkey UInt32, c0 UInt32) engine = TinyLog;
 CREATE TABLE t1 (vkey UInt32) ENGINE = AggregatingMergeTree  ORDER BY vkey;
 INSERT INTO t0 VALUES (15, 25000, 58);
-SELECT ref_5.pkey AS c_2_c2392_6 FROM t0 AS ref_5 WHERE 'J[' < multiIf(ref_5.pkey IN ( SELECT 1 ), bitShiftLeft(multiIf(ref_5.c0 > NULL, '1', ')'), 40), NULL); -- { serverError ARGUMENT_OUT_OF_BOUND }
+SELECT ref_5.pkey AS c_2_c2392_6 FROM t0 AS ref_5 WHERE 'J[' < multiIf(ref_5.pkey IN ( SELECT 1 ), bitShiftLeft(multiIf(ref_5.c0 > NULL, '1', ')'), 40), NULL);
 DROP TABLE t0;
 DROP TABLE t1;
 
diff --git a/tests/queries/0_stateless/03198_bit_shift_throws_error_for_out_of_bounds.reference b/tests/queries/0_stateless/03198_bit_shift_throws_error_for_out_of_bounds.reference
index 33b8cd6ee260..1fda82a97478 100644
--- a/tests/queries/0_stateless/03198_bit_shift_throws_error_for_out_of_bounds.reference
+++ b/tests/queries/0_stateless/03198_bit_shift_throws_error_for_out_of_bounds.reference
@@ -1,3 +1,9 @@
 -- bitShiftRight
+0
+
+\0\0\0\0\0\0\0\0
 -- bitShiftLeft
+0
+
+\0\0\0\0\0\0\0\0
 OK
diff --git a/tests/queries/0_stateless/03198_bit_shift_throws_error_for_out_of_bounds.sql b/tests/queries/0_stateless/03198_bit_shift_throws_error_for_out_of_bounds.sql
index aec017536739..340cc1292e46 100644
--- a/tests/queries/0_stateless/03198_bit_shift_throws_error_for_out_of_bounds.sql
+++ b/tests/queries/0_stateless/03198_bit_shift_throws_error_for_out_of_bounds.sql
@@ -1,17 +1,17 @@
 SELECT '-- bitShiftRight';
 SELECT bitShiftRight(1, -1); -- { serverError ARGUMENT_OUT_OF_BOUND }
-SELECT bitShiftRight(toUInt8(1), 8 + 1); -- { serverError ARGUMENT_OUT_OF_BOUND }
+SELECT bitShiftRight(toUInt8(1), 8 + 1);
 SELECT bitShiftRight('hola', -1); -- { serverError ARGUMENT_OUT_OF_BOUND }
-SELECT bitShiftRight('hola', 4 * 8 + 1); -- { serverError ARGUMENT_OUT_OF_BOUND }
+SELECT bitShiftRight('hola', 4 * 8 + 1);
 SELECT bitShiftRight(toFixedString('hola', 8), -1); -- { serverError ARGUMENT_OUT_OF_BOUND }
-SELECT bitShiftRight(toFixedString('hola', 8),  8 * 8 + 1); -- { serverError ARGUMENT_OUT_OF_BOUND }
+SELECT bitShiftRight(toFixedString('hola', 8),  8 * 8 + 1);
 
 SELECT '-- bitShiftLeft';
 SELECT bitShiftLeft(1, -1); -- { serverError ARGUMENT_OUT_OF_BOUND }
-SELECT bitShiftLeft(toUInt8(1), 8 + 1); -- { serverError ARGUMENT_OUT_OF_BOUND }
+SELECT bitShiftLeft(toUInt8(1), 8 + 1);
 SELECT bitShiftLeft('hola', -1); -- { serverError ARGUMENT_OUT_OF_BOUND }
-SELECT bitShiftLeft('hola', 4 * 8 + 1); -- { serverError ARGUMENT_OUT_OF_BOUND }
+SELECT bitShiftLeft('hola', 4 * 8 + 1);
 SELECT bitShiftLeft(toFixedString('hola', 8), -1); -- { serverError ARGUMENT_OUT_OF_BOUND }
-SELECT bitShiftLeft(toFixedString('hola', 8),  8 * 8 + 1); -- { serverError ARGUMENT_OUT_OF_BOUND }
+SELECT bitShiftLeft(toFixedString('hola', 8),  8 * 8 + 1);
 
 SELECT 'OK';
\ No newline at end of file
