diff --git a/test/sql/copy/csv/unicode_filename.test b/test/sql/copy/csv/unicode_filename.test
new file mode 100644
index 000000000000..a2df68173d9e
--- /dev/null
+++ b/test/sql/copy/csv/unicode_filename.test
@@ -0,0 +1,38 @@
+# name: test/sql/copy/csv/unicode_filename.test
+# description: Issue #2628: CSV with Chinese name cannot copy into DuckDB in Windows
+# group: [csv]
+
+statement ok
+PRAGMA enable_verification
+
+query II
+SELECT * FROM 'data/csv/issue2628_中文.csv'
+----
+1	1
+2	2
+3	3
+
+# unicode glob
+query II
+SELECT * FROM 'data/csv/*中文.csv'
+----
+1	1
+2	2
+3	3
+
+# unicode directory
+query II
+SELECT * FROM 'data/csv/中文/*.csv' ORDER BY 1
+----
+1	1
+2	2
+3	3
+4	4
+
+query II
+SELECT * FROM 'data/csv/中*/*.csv' ORDER BY 1
+----
+1	1
+2	2
+3	3
+4	4
\ No newline at end of file
diff --git a/test/sql/storage/unicode_filename.test b/test/sql/storage/unicode_filename.test
new file mode 100644
index 000000000000..67fe51cdb9fa
--- /dev/null
+++ b/test/sql/storage/unicode_filename.test
@@ -0,0 +1,104 @@
+# name: test/sql/storage/unicode_filename.test
+# description: Test storage with a unicode filename
+# group: [storage]
+
+# load the DB from disk
+load __TEST_DIR__/中文.db
+
+# create a database and insert values
+statement ok
+CREATE TABLE test (a INTEGER, b INTEGER);
+
+statement ok
+INSERT INTO test VALUES (11, 22), (13, 22), (12, 21), (NULL, NULL)
+
+statement ok
+CREATE TABLE test2 (a INTEGER);
+
+statement ok
+INSERT INTO test2 VALUES (13), (12), (11)
+
+query II
+SELECT * FROM test ORDER BY a
+----
+NULL	NULL
+11	22
+12	21
+13	22
+
+query I
+SELECT * FROM test2 ORDER BY a
+----
+11
+12
+13
+
+loop i 0 10
+
+statement ok
+CHECKPOINT;
+
+query II
+SELECT * FROM test ORDER BY a
+----
+NULL	NULL
+11	22
+12	21
+13	22
+
+query I
+SELECT * FROM test2 ORDER BY a
+----
+11
+12
+13
+
+endloop
+
+# verify that the tables are still there after a restart
+restart
+
+query II
+SELECT * FROM test ORDER BY a
+----
+NULL	NULL
+11	22
+12	21
+13	22
+
+query I
+SELECT * FROM test2 ORDER BY a
+----
+11
+12
+13
+
+# append a row
+statement ok
+INSERT INTO test VALUES (14, 23)
+
+query II
+SELECT * FROM test ORDER BY a
+----
+NULL	NULL
+11	22
+12	21
+13	22
+14	23
+
+restart
+
+query II
+SELECT * FROM test ORDER BY a
+----
+NULL	NULL
+11	22
+12	21
+13	22
+14	23
+
+statement ok
+DROP TABLE test
+
+statement ok
+DROP TABLE test2
