diff --git a/tests/integration/test_s3_plain_rewritable/test.py b/tests/integration/test_s3_plain_rewritable/test.py
index c23241b7f798..f02f20d954ff 100644
--- a/tests/integration/test_s3_plain_rewritable/test.py
+++ b/tests/integration/test_s3_plain_rewritable/test.py
@@ -179,3 +179,38 @@ def restart(node):
     )
 
     assert len(list(it)) == 0
+
+
+@pytest.mark.parametrize(
+    "storage_policy",
+    [
+        pytest.param("s3_plain_rewritable"),
+        pytest.param("cache_s3_plain_rewritable"),
+        pytest.param("s3_plain_rewritable_with_metadata_cache"),
+    ],
+)
+def test_projections(storage_policy):
+    node = cluster.instances["node1"]
+    table_name = "prj_test"
+    node.query(f"DROP TABLE IF EXISTS {table_name} SYNC")
+    node.query(
+        f"""
+            CREATE TABLE {table_name} (
+                x UInt32,
+                y UInt32,
+                PROJECTION prj_y (SELECT x, y ORDER BY y)
+            )
+            ENGINE = MergeTree
+            PRIMARY KEY x
+            SETTINGS storage_policy = '{storage_policy}';
+        """
+    )
+    node.query(
+        f"""
+            INSERT INTO {table_name} SELECT number, number FROM system.numbers limit 100000;
+        """
+    )
+    assert int(node.query(f"SELECT count(*) FROM {table_name} WHERE x=100")) == 1
+    assert int(node.query(f"SELECT count(*) FROM {table_name} WHERE y=100")) == 1
+
+    node.query(f"DROP TABLE {table_name} SYNC")
