diff --git a/tests/queries/0_stateless/02891_array_shingles.reference b/tests/queries/0_stateless/02891_array_shingles.reference
new file mode 100644
index 000000000000..04f7bfa331f2
--- /dev/null
+++ b/tests/queries/0_stateless/02891_array_shingles.reference
@@ -0,0 +1,41 @@
+-- negative tests
+-- const and non-const inputs
+Row 1:
+──────
+arr:                                            [1,2,3,4,5]
+len:                                            1
+arrayShingles([1, 2, 3, 4, 5], 1):              [[1],[2],[3],[4],[5]]
+arrayShingles(materialize([1, 2, 3, 4, 5]), 1): [[1],[2],[3],[4],[5]]
+Row 1:
+──────
+arr:                                            [1,2,3,4,5]
+len:                                            3
+arrayShingles([1, 2, 3, 4, 5], 3):              [[1,2,3],[2,3,4],[3,4,5]]
+arrayShingles(materialize([1, 2, 3, 4, 5]), 3): [[1,2,3],[2,3,4],[3,4,5]]
+Row 1:
+──────
+arr:                                            [1,2,3,4,5]
+len:                                            5
+arrayShingles([1, 2, 3, 4, 5], 5):              [[1,2,3,4,5]]
+arrayShingles(materialize([1, 2, 3, 4, 5]), 5): [[1,2,3,4,5]]
+Row 1:
+──────
+arr:                                                        ['ab','c','de','','hi']
+len:                                                        1
+arrayShingles(['ab', 'c', 'de', '', 'hi'], 1):              [['ab'],['c'],['de'],[''],['hi']]
+arrayShingles(materialize(['ab', 'c', 'de', '', 'hi']), 1): [['ab'],['c'],['de'],[''],['hi']]
+Row 1:
+──────
+arr:                                                        ['ab','c','de','','hi']
+len:                                                        3
+arrayShingles(['ab', 'c', 'de', '', 'hi'], 3):              [['ab','c','de'],['c','de',''],['de','','hi']]
+arrayShingles(materialize(['ab', 'c', 'de', '', 'hi']), 3): [['ab','c','de'],['c','de',''],['de','','hi']]
+Row 1:
+──────
+arr:                                                        ['ab','c','de','','hi']
+len:                                                        5
+arrayShingles(['ab', 'c', 'de', '', 'hi'], 5):              [['ab','c','de','','hi']]
+arrayShingles(materialize(['ab', 'c', 'de', '', 'hi']), 5): [['ab','c','de','','hi']]
+-- special cases
+[[2],[1]]
+[[2],[1]]
diff --git a/tests/queries/0_stateless/02891_array_shingles.sql b/tests/queries/0_stateless/02891_array_shingles.sql
new file mode 100644
index 000000000000..f199a445a304
--- /dev/null
+++ b/tests/queries/0_stateless/02891_array_shingles.sql
@@ -0,0 +1,23 @@
+SELECT '-- negative tests';
+SELECT arrayShingles(); -- { serverError NUMBER_OF_ARGUMENTS_DOESNT_MATCH }
+SELECT arrayShingles([1, 2, 3, 4, 5]); -- { serverError NUMBER_OF_ARGUMENTS_DOESNT_MATCH }
+SELECT arrayShingles([1, 2, 3, 4, 5], 2, 3); -- { serverError NUMBER_OF_ARGUMENTS_DOESNT_MATCH }
+SELECT arrayShingles([1, 2, 3, 4, 5], 'str'); -- { serverError ILLEGAL_TYPE_OF_ARGUMENT }
+SELECT arrayShingles((1, 2, 3, 4, 5), 0); -- { serverError ILLEGAL_TYPE_OF_ARGUMENT }
+SELECT arrayShingles([1, 2, 3, 4, 5], 0); -- { serverError BAD_ARGUMENTS }
+SELECT arrayShingles([1, 2, 3, 4, 5], -2); -- { serverError BAD_ARGUMENTS }
+SELECT arrayShingles([1, 2, 3, 4, 5], 6); -- { serverError BAD_ARGUMENTS }
+SELECT arrayShingles([], 1); -- { serverError BAD_ARGUMENTS }
+
+SELECT '-- const and non-const inputs';
+SELECT [1, 2, 3, 4, 5] AS arr, 1 AS len, arrayShingles(arr, len), arrayShingles(materialize(arr), len) FORMAT Vertical;
+SELECT [1, 2, 3, 4, 5] AS arr, 3 AS len, arrayShingles(arr, len), arrayShingles(materialize(arr), len) FORMAT Vertical;
+SELECT [1, 2 ,3, 4, 5] AS arr, 5 AS len, arrayShingles(arr, len), arrayShingles(materialize(arr), len) FORMAT Vertical;
+
+SELECT ['ab', 'c', 'de', '', 'hi'] AS arr, 1 AS len, arrayShingles(arr, len), arrayShingles(materialize(arr), len) FORMAT Vertical;
+SELECT ['ab', 'c', 'de', '', 'hi'] AS arr, 3 AS len, arrayShingles(arr, len), arrayShingles(materialize(arr), len) FORMAT Vertical;
+SELECT ['ab', 'c', 'de', '', 'hi'] AS arr, 5 AS len, arrayShingles(arr, len), arrayShingles(materialize(arr), len) FORMAT Vertical;
+
+SELECT '-- special cases';
+SELECT arrayShingles([toNullable(2), toNullable(1)], 1);
+SELECT arrayShingles([toLowCardinality(2), toLowCardinality(1)], 1);
