diff --git a/tests/queries/0_stateless/01152_cross_replication.reference b/tests/queries/0_stateless/01152_cross_replication.reference
new file mode 100644
index 000000000000..389d14ff28b6
--- /dev/null
+++ b/tests/queries/0_stateless/01152_cross_replication.reference
@@ -0,0 +1,10 @@
+localhost	9000	0		0	0
+localhost	9000	0		0	0
+demo_loan_01568
+demo_loan_01568
+CREATE TABLE shard_0.demo_loan_01568
(
    `id` Int64 COMMENT \'id\',
    `date_stat` Date COMMENT \'date of stat\',
    `customer_no` String COMMENT \'customer no\',
    `loan_principal` Float64 COMMENT \'loan principal\'
)
ENGINE = ReplacingMergeTree
PARTITION BY toYYYYMM(date_stat)
ORDER BY id
SETTINGS index_granularity = 8192
+CREATE TABLE shard_1.demo_loan_01568
(
    `id` Int64 COMMENT \'id\',
    `date_stat` Date COMMENT \'date of stat\',
    `customer_no` String COMMENT \'customer no\',
    `loan_principal` Float64 COMMENT \'loan principal\'
)
ENGINE = ReplacingMergeTree
PARTITION BY toYYYYMM(date_stat)
ORDER BY id
SETTINGS index_granularity = 8192
+1	2021-04-13	qwerty	3.14159
+2	2021-04-14	asdfgh	2.71828
+2	2021-04-14	asdfgh	2.71828
+1	2021-04-13	qwerty	3.14159
diff --git a/tests/queries/0_stateless/01152_cross_replication.sql b/tests/queries/0_stateless/01152_cross_replication.sql
new file mode 100644
index 000000000000..23507c41fd02
--- /dev/null
+++ b/tests/queries/0_stateless/01152_cross_replication.sql
@@ -0,0 +1,30 @@
+DROP DATABASE IF EXISTS shard_0;
+DROP DATABASE IF EXISTS shard_1;
+SET distributed_ddl_output_mode='none';
+DROP TABLE IF EXISTS demo_loan_01568_dist;
+
+CREATE DATABASE shard_0;
+CREATE DATABASE shard_1;
+
+CREATE TABLE demo_loan_01568 ON CLUSTER test_cluster_two_shards_different_databases ( `id` Int64 COMMENT 'id', `date_stat` Date COMMENT 'date of stat', `customer_no` String COMMENT 'customer no', `loan_principal` Float64 COMMENT 'loan principal' ) ENGINE=ReplacingMergeTree() ORDER BY id PARTITION BY toYYYYMM(date_stat); -- { serverError 371 }
+SET distributed_ddl_output_mode='throw';
+CREATE TABLE shard_0.demo_loan_01568 ON CLUSTER test_cluster_two_shards_different_databases ( `id` Int64 COMMENT 'id', `date_stat` Date COMMENT 'date of stat', `customer_no` String COMMENT 'customer no', `loan_principal` Float64 COMMENT 'loan principal' ) ENGINE=ReplacingMergeTree() ORDER BY id PARTITION BY toYYYYMM(date_stat);
+CREATE TABLE shard_1.demo_loan_01568 ON CLUSTER test_cluster_two_shards_different_databases ( `id` Int64 COMMENT 'id', `date_stat` Date COMMENT 'date of stat', `customer_no` String COMMENT 'customer no', `loan_principal` Float64 COMMENT 'loan principal' ) ENGINE=ReplacingMergeTree() ORDER BY id PARTITION BY toYYYYMM(date_stat);
+SET distributed_ddl_output_mode='none';
+
+SHOW TABLES FROM shard_0;
+SHOW TABLES FROM shard_1;
+SHOW CREATE TABLE shard_0.demo_loan_01568;
+SHOW CREATE TABLE shard_1.demo_loan_01568;
+
+CREATE TABLE demo_loan_01568_dist AS shard_0.demo_loan_01568 ENGINE=Distributed('test_cluster_two_shards_different_databases', '', 'demo_loan_01568', id % 2);
+INSERT INTO demo_loan_01568_dist VALUES (1, '2021-04-13', 'qwerty', 3.14159), (2, '2021-04-14', 'asdfgh', 2.71828);
+SYSTEM FLUSH DISTRIBUTED demo_loan_01568_dist;
+SELECT * FROM demo_loan_01568_dist ORDER BY id;
+
+SELECT * FROM shard_0.demo_loan_01568;
+SELECT * FROM shard_1.demo_loan_01568;
+
+DROP DATABASE shard_0;
+DROP DATABASE shard_1;
+DROP TABLE demo_loan_01568_dist;
diff --git a/tests/queries/skip_list.json b/tests/queries/skip_list.json
index d41a41bd5244..59ee6ae89f45 100644
--- a/tests/queries/skip_list.json
+++ b/tests/queries/skip_list.json
@@ -557,6 +557,7 @@
         "01135_default_and_alter_zookeeper",
         "01148_zookeeper_path_macros_unfolding",
         "01150_ddl_guard_rwr",
+        "01152_cross_replication",
         "01185_create_or_replace_table",
         "01190_full_attach_syntax",
         "01191_rename_dictionary",
