diff --git a/tests/queries/0_stateless/02181_sql_user_defined_functions_invalid_lambda.sql b/tests/queries/0_stateless/02181_sql_user_defined_functions_invalid_lambda.sql
index 115f3b2f9c06..79cc8fd906a0 100644
--- a/tests/queries/0_stateless/02181_sql_user_defined_functions_invalid_lambda.sql
+++ b/tests/queries/0_stateless/02181_sql_user_defined_functions_invalid_lambda.sql
@@ -1,4 +1,4 @@
-CREATE FUNCTION 02181_invalid_lambda AS lambda(((x * 2) AS x_doubled) + x_doubled); --{serverError UNSUPPORTED_METHOD}
-CREATE FUNCTION 02181_invalid_lambda AS lambda(x); --{serverError UNSUPPORTED_METHOD}
-CREATE FUNCTION 02181_invalid_lambda AS lambda(); --{serverError UNSUPPORTED_METHOD}
-CREATE FUNCTION 02181_invalid_lambda AS lambda(tuple(x)) --{serverError UNSUPPORTED_METHOD}
+CREATE FUNCTION 02181_invalid_lambda AS lambda(((x * 2) AS x_doubled) + x_doubled); --{serverError BAD_ARGUMENTS}
+CREATE FUNCTION 02181_invalid_lambda AS lambda(x); --{serverError BAD_ARGUMENTS}
+CREATE FUNCTION 02181_invalid_lambda AS lambda(); --{serverError BAD_ARGUMENTS}
+CREATE FUNCTION 02181_invalid_lambda AS lambda(tuple(x)) --{serverError BAD_ARGUMENTS}
diff --git a/tests/queries/0_stateless/02292_create_function_validate.sql b/tests/queries/0_stateless/02292_create_function_validate.sql
index a1f0afe055e9..56b546789868 100644
--- a/tests/queries/0_stateless/02292_create_function_validate.sql
+++ b/tests/queries/0_stateless/02292_create_function_validate.sql
@@ -1,1 +1,1 @@
-create function foo as x -- { serverError UNSUPPORTED_METHOD }
+create function foo as x -- { serverError BAD_ARGUMENTS }
diff --git a/tests/queries/0_stateless/03518_bad_sql_udf.reference b/tests/queries/0_stateless/03518_bad_sql_udf.reference
new file mode 100644
index 000000000000..e69de29bb2d1
diff --git a/tests/queries/0_stateless/03518_bad_sql_udf.sql b/tests/queries/0_stateless/03518_bad_sql_udf.sql
new file mode 100644
index 000000000000..2c60941d0768
--- /dev/null
+++ b/tests/queries/0_stateless/03518_bad_sql_udf.sql
@@ -0,0 +1,1 @@
+CREATE OR REPLACE FUNCTION 03518_bad_sql_udf AS lambda(identity(x), x); -- { serverError BAD_ARGUMENTS }
