diff --git a/src/AggregateFunctions/AggregateFunctionLargestTriangleThreeBuckets.cpp b/src/AggregateFunctions/AggregateFunctionLargestTriangleThreeBuckets.cpp
index 813b13b6f7bd..99425a27ef08 100644
--- a/src/AggregateFunctions/AggregateFunctionLargestTriangleThreeBuckets.cpp
+++ b/src/AggregateFunctions/AggregateFunctionLargestTriangleThreeBuckets.cpp
@@ -329,7 +329,7 @@ class AggregateFunctionLargestTriangleThreeBuckets final : public IAggregateFunc
                 return [](IColumn & column, Float64 value)
                 {
                     auto & col = assert_cast<ColumnDateTime64 &>(column);
-                    col.getData().push_back(static_cast<UInt64>(value));
+                    col.getData().push_back(static_cast<Int64>(value));
                 };
             default:
                 return [](IColumn & column, Float64 value)
