diff --git a/tests/queries/0_stateless/02842_largestTriangleThreeBuckets_aggregate_function.reference b/tests/queries/0_stateless/02842_largestTriangleThreeBuckets_aggregate_function.reference
index cf7f77642531..1e304d84612c 100644
--- a/tests/queries/0_stateless/02842_largestTriangleThreeBuckets_aggregate_function.reference
+++ b/tests/queries/0_stateless/02842_largestTriangleThreeBuckets_aggregate_function.reference
@@ -15,3 +15,4 @@
 (9978,978)	9978	10
 (9988,988)	9988	10
 (9999,999)	9999	11
+[(0,'1900-01-01 00:00:00.000')]
diff --git a/tests/queries/0_stateless/02842_largestTriangleThreeBuckets_aggregate_function.sql b/tests/queries/0_stateless/02842_largestTriangleThreeBuckets_aggregate_function.sql
index d5ef564469e0..438302dc1def 100644
--- a/tests/queries/0_stateless/02842_largestTriangleThreeBuckets_aggregate_function.sql
+++ b/tests/queries/0_stateless/02842_largestTriangleThreeBuckets_aggregate_function.sql
@@ -61,4 +61,6 @@ SELECT
   point_x - neighbor(point_x, -1) AS point_x_diff_with_previous_row
 FROM largestTriangleTreeBucketsBucketSizeTest LIMIT 990, 10;
 
+SELECT largestTriangleThreeBuckets(1)(0, '1900-01-01 00:00:00'::DateTime64);
+
 DROP TABLE largestTriangleTreeBucketsBucketSizeTest;
