diff --git a/dbms/tests/queries/0_stateless/00374_json_each_row_input_with_noisy_fields.reference b/dbms/tests/queries/0_stateless/00374_json_each_row_input_with_noisy_fields.reference
index 4150b9e80b14..e9d98aa456e2 100644
--- a/dbms/tests/queries/0_stateless/00374_json_each_row_input_with_noisy_fields.reference
+++ b/dbms/tests/queries/0_stateless/00374_json_each_row_input_with_noisy_fields.reference
@@ -3,3 +3,4 @@
 0	
 1	ok
 1	ok
+1	ok
diff --git a/dbms/tests/queries/0_stateless/00374_json_each_row_input_with_noisy_fields.sh b/dbms/tests/queries/0_stateless/00374_json_each_row_input_with_noisy_fields.sh
index 0b4b33143ae4..7b34db0b9404 100755
--- a/dbms/tests/queries/0_stateless/00374_json_each_row_input_with_noisy_fields.sh
+++ b/dbms/tests/queries/0_stateless/00374_json_each_row_input_with_noisy_fields.sh
@@ -1,11 +1,16 @@
 #!/usr/bin/env bash
-clickhouse-client -n --query "DROP TABLE IF EXISTS test.json_noisy; CREATE TABLE test.json_noisy (d1 UInt8, d2 String) ENGINE = Memory"
+set -e
+
+clickhouse-client -q "DROP TABLE IF EXISTS test.json_noisy"
+clickhouse-client -q "CREATE TABLE test.json_noisy (d1 UInt8, d2 String) ENGINE = Memory"
 
 echo '{"d1" : 1, "d2" : "ok"}
 { }
 {"t1" : 0, "t2":true,"t3":false, "t4":null,"t5":[],"t6":"trash" }
 {"d2":"ok","t1":[[[]],true, null, false, "1","2",9.03,101], "t2":[["1","2"]], "d1":"1"}
-{"d2":"ok","t1":[[[]],true, null, false, "1","2", 0.03, 1], "d1":"1", "t2":["1","2"]}' \
-| clickhouse-client -n --query "SET input_format_skip_unknown_fields = 1; INSERT INTO test.json_noisy FORMAT JSONEachRow"
+{"d2":"ok","t1":[[[]],true, null, false, "1","2", 0.03, 1], "d1":"1", "t2":["1","2"]}
+{"t0" : -0.1, "t1" : +1, "t2" : 0, "t3" : [0.0, -0.1], "d2" : "ok", "d1" : 1}' \
+| clickhouse-client --input_format_skip_unknown_fields=1 -q "INSERT INTO test.json_noisy FORMAT JSONEachRow"
 
-clickhouse-client -n --query "SELECT * FROM test.json_noisy; DROP TABLE IF EXISTS test.json_noisy;"
+clickhouse-client --max_threads=1 -q "SELECT * FROM test.json_noisy"
+clickhouse-client -q "DROP TABLE IF EXISTS test.json_noisy"
