diff --git a/tests/queries/0_stateless/02129_add_column_add_ttl.reference b/tests/queries/0_stateless/02129_add_column_add_ttl.reference
new file mode 100644
index 000000000000..8b3280ef095d
--- /dev/null
+++ b/tests/queries/0_stateless/02129_add_column_add_ttl.reference
@@ -0,0 +1,41 @@
+0		2021-01-01	0
+0		2021-01-01	0
+1		2021-01-01	0
+1		2021-01-01	0
+2		2021-01-01	0
+2		2021-01-01	0
+3		2021-01-01	0
+3		2021-01-01	0
+4		2021-01-01	0
+4		2021-01-01	0
+5		2021-01-01	0
+5		2021-01-01	0
+6		2021-01-01	0
+6		2021-01-01	0
+7		2021-01-01	0
+7		2021-01-01	0
+8		2021-01-01	0
+8		2021-01-01	0
+9		2021-01-01	0
+9		2021-01-01	0
+==========
+0		2021-01-01	0
+0		2021-01-01	0
+1		2021-01-01	0
+1		2021-01-01	0
+2		2021-01-01	0
+2		2021-01-01	0
+3		2021-01-01	0
+3		2021-01-01	0
+4		2021-01-01	0
+4		2021-01-01	0
+5		2021-01-01	0
+5		2021-01-01	0
+6		2021-01-01	0
+6		2021-01-01	0
+7		2021-01-01	0
+7		2021-01-01	0
+8		2021-01-01	0
+8		2021-01-01	0
+9		2021-01-01	0
+9		2021-01-01	0
diff --git a/tests/queries/0_stateless/02129_add_column_add_ttl.sql b/tests/queries/0_stateless/02129_add_column_add_ttl.sql
new file mode 100644
index 000000000000..7a6dd928a3f5
--- /dev/null
+++ b/tests/queries/0_stateless/02129_add_column_add_ttl.sql
@@ -0,0 +1,31 @@
+drop table if exists ttl_test_02129;
+
+create table ttl_test_02129(a Int64, b String, d Date)
+Engine=MergeTree partition by d order by a
+settings min_bytes_for_wide_part = 0, min_rows_for_wide_part = 0, materialize_ttl_recalculate_only = 0;
+
+insert into ttl_test_02129 select number, '', '2021-01-01' from numbers(10);
+alter table ttl_test_02129 add column c Int64 settings mutations_sync=2;
+
+insert into ttl_test_02129 select number, '', '2021-01-01', 0 from numbers(10);
+alter table  ttl_test_02129 modify TTL (d + INTERVAL 1 MONTH) DELETE WHERE c=1 settings mutations_sync=2;
+
+select * from ttl_test_02129 order by a, b, d, c;
+drop table ttl_test_02129;
+
+drop table if exists ttl_test_02129;
+
+select '==========';
+
+create table ttl_test_02129(a Int64, b String, d Date)
+Engine=MergeTree partition by d order by a
+settings min_bytes_for_wide_part = 0, min_rows_for_wide_part = 0, materialize_ttl_recalculate_only = 1;
+
+insert into ttl_test_02129 select number, '', '2021-01-01' from numbers(10);
+alter table ttl_test_02129 add column c Int64 settings mutations_sync=2;
+
+insert into ttl_test_02129 select number, '', '2021-01-01', 0 from numbers(10);
+alter table  ttl_test_02129 modify TTL (d + INTERVAL 1 MONTH) DELETE WHERE c=1 settings mutations_sync=2;
+
+select * from ttl_test_02129 order by a, b, d, c;
+drop table ttl_test_02129;
