diff --git a/src/Functions/sleep.h b/src/Functions/sleep.h
index 6dca6b5f84c9..65566e36d1f5 100644
--- a/src/Functions/sleep.h
+++ b/src/Functions/sleep.h
@@ -78,8 +78,8 @@ class FunctionSleep : public IFunction
 
         Float64 seconds = applyVisitor(FieldVisitorConvertToNumber<Float64>(), assert_cast<const ColumnConst &>(*col).getField());
 
-        if (seconds < 0)
-            throw Exception("Cannot sleep negative amount of time (not implemented)", ErrorCodes::BAD_ARGUMENTS);
+        if (seconds < 0 || !std::isfinite(seconds))
+            throw Exception("Cannot sleep infinite or negative amount of time (not implemented)", ErrorCodes::BAD_ARGUMENTS);
 
         size_t size = col->size();
 
