diff --git a/tests/queries/0_stateless/02812_subquery_operators.reference b/tests/queries/0_stateless/02812_subquery_operators.reference
new file mode 100644
index 000000000000..aed0a046f996
--- /dev/null
+++ b/tests/queries/0_stateless/02812_subquery_operators.reference
@@ -0,0 +1,6 @@
+
+Hello
+Hello
+123
+1
+	['\0']	[]	\0	['']
diff --git a/tests/queries/0_stateless/02812_subquery_operators.sql b/tests/queries/0_stateless/02812_subquery_operators.sql
new file mode 100644
index 000000000000..b0638b43e899
--- /dev/null
+++ b/tests/queries/0_stateless/02812_subquery_operators.sql
@@ -0,0 +1,6 @@
+SELECT singleValueOrNull(toNullable(''));
+SELECT singleValueOrNull(toNullable('Hello'));
+SELECT singleValueOrNull((SELECT 'Hello'));
+SELECT singleValueOrNull(toNullable(123));
+SELECT '' = ALL (SELECT toNullable(''));
+SELECT '', ['\0'], [], singleValueOrNull(( SELECT '\0' ) ), [''];
