diff --git a/tests/queries/0_stateless/03366_with_fill_dag.reference b/tests/queries/0_stateless/03366_with_fill_dag.reference
new file mode 100644
index 000000000000..a1c69bb348dc
--- /dev/null
+++ b/tests/queries/0_stateless/03366_with_fill_dag.reference
@@ -0,0 +1,14 @@
+0
+1
+2
+3
+4
+5
+6
+7
+8
+9
+0	1
+1	1
+2	1
+3	1
diff --git a/tests/queries/0_stateless/03366_with_fill_dag.sql b/tests/queries/0_stateless/03366_with_fill_dag.sql
new file mode 100644
index 000000000000..32f7e92f5ce3
--- /dev/null
+++ b/tests/queries/0_stateless/03366_with_fill_dag.sql
@@ -0,0 +1,16 @@
+SELECT number
+FROM numbers(10)
+ORDER BY
+    number ASC WITH FILL STEP 1,
+    'aaa' ASC
+LIMIT 1 BY number;
+
+WITH 1 AS one
+SELECT
+    number AS num,
+    one
+FROM numbers(4)
+ORDER BY
+    num ASC WITH FILL STEP 1,
+    one ASC
+INTERPOLATE ( one AS 42 );
