diff --git a/src/Common/tests/gtest_optimize_re.cpp b/src/Common/tests/gtest_optimize_re.cpp
index d6735c3ccfe9..b0d66d7486df 100644
--- a/src/Common/tests/gtest_optimize_re.cpp
+++ b/src/Common/tests/gtest_optimize_re.cpp
@@ -19,6 +19,11 @@ TEST(OptimizeRE, analyze)
     };
     test_f("abc", "abc", {}, true, true);
     test_f("c([^k]*)de", "");
+    test_f("a|XYZ", "", {"a", "XYZ"});
+    test_f("XYZ|a", "", {"XYZ", "a"});
+    test_f("[Ff]|XYZ", "", {"", "XYZ"});
+    test_f("XYZ|[Ff]", "", {"XYZ", ""});
+    test_f("XYZ|ABC|[Ff]", "", {"XYZ", "ABC", ""});
     test_f("(?-s)bob", "bob", {}, false, true);
     test_f("(?s)bob", "bob", {}, false, true);
     test_f("(?ssss", "");
diff --git a/tests/queries/0_stateless/03443_index_match_alternatives.reference b/tests/queries/0_stateless/03443_index_match_alternatives.reference
new file mode 100644
index 000000000000..001c06f0001b
--- /dev/null
+++ b/tests/queries/0_stateless/03443_index_match_alternatives.reference
@@ -0,0 +1,22 @@
+-- Without index
+John
+John
+John
+John
+-- With index
+John
+John
+John
+John
+-- Assert selected granules
+Granules: 3/3
+Granules: 1/3
+
+Granules: 3/3
+Granules: 1/3
+
+Granules: 3/3
+Granules: 3/3
+
+Granules: 3/3
+Granules: 3/3
diff --git a/tests/queries/0_stateless/03443_index_match_alternatives.sql b/tests/queries/0_stateless/03443_index_match_alternatives.sql
new file mode 100644
index 000000000000..fe147975fbb2
--- /dev/null
+++ b/tests/queries/0_stateless/03443_index_match_alternatives.sql
@@ -0,0 +1,39 @@
+DROP TABLE IF EXISTS 03443_data;
+
+CREATE TABLE 03443_data
+(
+    id Int32,
+    name String,
+    INDEX idx_name name TYPE ngrambf_v1(1, 1024, 3, 0) GRANULARITY 1
+)
+ENGINE = MergeTree ORDER BY id SETTINGS index_granularity = 1
+AS
+SELECT 1, 'John' UNION ALL
+SELECT 2, 'Ksenia' UNION ALL
+SELECT 3, 'Alice';
+
+SELECT '-- Without index';
+SELECT name FROM 03443_data WHERE match(name, 'J|XYZ') SETTINGS use_skip_indexes = 0;
+SELECT name FROM 03443_data WHERE match(name, 'XYZ|J') SETTINGS use_skip_indexes = 0;
+SELECT name FROM 03443_data WHERE match(name, '[J]|XYZ') SETTINGS use_skip_indexes = 0;
+SELECT name FROM 03443_data WHERE match(name, 'XYZ|[J]') SETTINGS use_skip_indexes = 0;
+
+SELECT '-- With index';
+SELECT name FROM 03443_data WHERE match(name, 'J|XYZ') SETTINGS use_skip_indexes = 1;
+SELECT name FROM 03443_data WHERE match(name, 'XYZ|J') SETTINGS use_skip_indexes = 1;
+SELECT name FROM 03443_data WHERE match(name, '[J]|XYZ') SETTINGS use_skip_indexes = 1;
+SELECT name FROM 03443_data WHERE match(name, 'XYZ|[J]') SETTINGS use_skip_indexes = 1;
+
+SELECT '-- Assert selected granules';
+
+SET parallel_replicas_local_plan = 1;
+
+SELECT trim(leading ' ' from explain) FROM (EXPLAIN indexes=1 SELECT name FROM 03443_data WHERE match(name, 'J|XYZ')) WHERE explain LIKE '%Granules: %' SETTINGS use_skip_indexes = 1;
+SELECT '';
+SELECT trim(leading ' ' from explain) FROM (EXPLAIN indexes=1 SELECT name FROM 03443_data WHERE match(name, 'XYZ|J')) WHERE explain LIKE '%Granules: %' SETTINGS use_skip_indexes = 1;
+SELECT '';
+SELECT trim(leading ' ' from explain) FROM (EXPLAIN indexes=1 SELECT name FROM 03443_data WHERE match(name, '[J]|XYZ')) WHERE explain LIKE '%Granules: %' SETTINGS use_skip_indexes = 1;
+SELECT '';
+SELECT trim(leading ' ' from explain) FROM (EXPLAIN indexes=1 SELECT name FROM 03443_data WHERE match(name, 'XYZ|[J]')) WHERE explain LIKE '%Granules: %' SETTINGS use_skip_indexes = 1;
+
+DROP TABLE 03443_data;
