diff --git a/tests/queries/0_stateless/01265_datetime_string_comparison_felix_mueller.reference b/tests/queries/0_stateless/01265_datetime_string_comparison_felix_mueller.reference
new file mode 100644
index 000000000000..a76c5db14aff
--- /dev/null
+++ b/tests/queries/0_stateless/01265_datetime_string_comparison_felix_mueller.reference
@@ -0,0 +1,4 @@
+1557136800	1557169200
+1
+1
+1
diff --git a/tests/queries/0_stateless/01265_datetime_string_comparison_felix_mueller.sql b/tests/queries/0_stateless/01265_datetime_string_comparison_felix_mueller.sql
new file mode 100644
index 000000000000..41cc5b532390
--- /dev/null
+++ b/tests/queries/0_stateless/01265_datetime_string_comparison_felix_mueller.sql
@@ -0,0 +1,29 @@
+DROP TABLE IF EXISTS tztest;
+
+CREATE TABLE tztest
+(
+    timeBerlin DateTime('Europe/Berlin'), 
+    timeLA DateTime('America/Los_Angeles')
+)
+ENGINE = Memory;
+
+INSERT INTO tztest (timeBerlin, timeLA) VALUES ('2019-05-06 12:00:00', '2019-05-06 12:00:00');
+
+SELECT
+    toUnixTimestamp(timeBerlin),
+    toUnixTimestamp(timeLA)
+FROM tztest;
+
+SELECT 1
+FROM tztest
+WHERE timeBerlin = '2019-05-06 12:00:00';
+
+SELECT 1
+FROM tztest
+WHERE timeLA = '2019-05-06 12:00:00';
+
+SELECT 1
+FROM tztest
+WHERE '2019-05-06 12:00:00' = timeBerlin;
+
+DROP TABLE tztest;
