diff --git a/src/Interpreters/PartLog.cpp b/src/Interpreters/PartLog.cpp
index cf231cbd7c9a..881fcae4de66 100644
--- a/src/Interpreters/PartLog.cpp
+++ b/src/Interpreters/PartLog.cpp
@@ -135,6 +135,7 @@ NamesAndAliases PartLogElement::getNamesAndAliases()
     {
         {"ProfileEvents.Names", {std::make_shared<DataTypeArray>(std::make_shared<DataTypeString>())}, "mapKeys(ProfileEvents)"},
         {"ProfileEvents.Values", {std::make_shared<DataTypeArray>(std::make_shared<DataTypeUInt64>())}, "mapValues(ProfileEvents)"},
+        {"name", {std::make_shared<DataTypeString>()}, "part_name"},
     };
 }
 
diff --git a/src/Storages/System/StorageSystemPartsBase.cpp b/src/Storages/System/StorageSystemPartsBase.cpp
index fb7f4f9066ab..0979b9d9371f 100644
--- a/src/Storages/System/StorageSystemPartsBase.cpp
+++ b/src/Storages/System/StorageSystemPartsBase.cpp
@@ -304,6 +304,7 @@ StorageSystemPartsBase::StorageSystemPartsBase(const StorageID & table_id_, Name
     /// Add aliases for old column names for backwards compatibility.
     add_alias("bytes", "bytes_on_disk");
     add_alias("marks_size", "marks_bytes");
+    add_alias("part_name", "name");
 
     StorageInMemoryMetadata storage_metadata;
     storage_metadata.setColumns(tmp_columns);
