diff --git a/tools/jdbc/src/test/java/org/duckdb/test/TestDuckDBJDBC.java b/tools/jdbc/src/test/java/org/duckdb/test/TestDuckDBJDBC.java
index aaebc706f05b..86e9ac53f648 100644
--- a/tools/jdbc/src/test/java/org/duckdb/test/TestDuckDBJDBC.java
+++ b/tools/jdbc/src/test/java/org/duckdb/test/TestDuckDBJDBC.java
@@ -1288,6 +1288,10 @@ public static void test_utf_string_bug1271() throws Exception {
 		Statement stmt = conn.createStatement();
 
 		ResultSet rs = stmt.executeQuery("SELECT 'Mühleisen', '🦆', '🦄ྀི123456789'");
+		assertEquals(rs.getMetaData().getColumnName(1), "Mühleisen");
+		assertEquals(rs.getMetaData().getColumnName(2), "🦆");
+		assertEquals(rs.getMetaData().getColumnName(3), "🦄ྀི123456789");
+
 		assertTrue(rs.next());
 
 		assertEquals(rs.getString(1), "Mühleisen");
