diff --git a/docker/test/performance-comparison/config/users.d/perf-comparison-tweaks-users.xml b/docker/test/performance-comparison/config/users.d/perf-comparison-tweaks-users.xml
index cb591f1a184b..e780a99ecde7 100644
--- a/docker/test/performance-comparison/config/users.d/perf-comparison-tweaks-users.xml
+++ b/docker/test/performance-comparison/config/users.d/perf-comparison-tweaks-users.xml
@@ -34,9 +34,4 @@
             <memory_profiler_step>0</memory_profiler_step>
         </default>
     </profiles>
-    <users>
-        <default>
-            <access_management>1</access_management>
-        </default>
-    </users>
 </clickhouse>
diff --git a/tests/integration/test_named_collections/configs/users.d/users_no_default_access.xml b/tests/integration/test_named_collections/configs/users.d/users_no_default_access.xml
index b8f38f04ca94..dcac83188dcb 100644
--- a/tests/integration/test_named_collections/configs/users.d/users_no_default_access.xml
+++ b/tests/integration/test_named_collections/configs/users.d/users_no_default_access.xml
@@ -4,6 +4,7 @@
             <password></password>
             <profile>default</profile>
             <quota>default</quota>
+            <named_collection_control>0</named_collection_control>
         </default>
     </users>
 </clickhouse>
diff --git a/tests/integration/test_storage_s3/test.py b/tests/integration/test_storage_s3/test.py
index 01ade1acc4d5..3dd3c9e39d02 100644
--- a/tests/integration/test_storage_s3/test.py
+++ b/tests/integration/test_storage_s3/test.py
@@ -944,13 +944,6 @@ def test_predefined_connection_configuration(started_cluster):
     instance.query("GRANT SELECT ON *.* TO user")
 
     instance.query(f"drop table if exists {name}", user="user")
-    error = instance.query_and_get_error(
-        f"CREATE TABLE {name} (id UInt32) ENGINE = S3(s3_conf1, format='CSV')"
-    )
-    assert (
-        "To execute this query, it's necessary to have the grant NAMED COLLECTION ON s3_conf1"
-        in error
-    )
     error = instance.query_and_get_error(
         f"CREATE TABLE {name} (id UInt32) ENGINE = S3(s3_conf1, format='CSV')",
         user="user",
@@ -975,11 +968,6 @@ def test_predefined_connection_configuration(started_cluster):
     )
     assert result == instance.query("SELECT number FROM numbers(10)")
 
-    error = instance.query_and_get_error("SELECT * FROM s3(no_collection)")
-    assert (
-        "To execute this query, it's necessary to have the grant NAMED COLLECTION ON no_collection"
-        in error
-    )
     error = instance.query_and_get_error("SELECT * FROM s3(no_collection)", user="user")
     assert (
         "To execute this query, it's necessary to have the grant NAMED COLLECTION ON no_collection"
