diff --git a/tests/queries/0_stateless/02532_profileevents_server_startup_time.reference b/tests/queries/0_stateless/02532_profileevents_server_startup_time.reference
new file mode 100644
index 000000000000..e3c0063769c5
--- /dev/null
+++ b/tests/queries/0_stateless/02532_profileevents_server_startup_time.reference
@@ -0,0 +1,1 @@
+ServerStartupMilliseconds
diff --git a/tests/queries/0_stateless/02532_profileevents_server_startup_time.sql b/tests/queries/0_stateless/02532_profileevents_server_startup_time.sql
new file mode 100644
index 000000000000..d7c97d08c151
--- /dev/null
+++ b/tests/queries/0_stateless/02532_profileevents_server_startup_time.sql
@@ -0,0 +1,1 @@
+SELECT event FROM system.events WHERE event = 'ServerStartupMilliseconds'
\ No newline at end of file
