diff --git a/tests/queries/0_stateless/02508_bad_graphite.reference b/tests/queries/0_stateless/02508_bad_graphite.reference
new file mode 100644
index 000000000000..e69de29bb2d1
diff --git a/tests/queries/0_stateless/02508_bad_graphite.sql b/tests/queries/0_stateless/02508_bad_graphite.sql
new file mode 100644
index 000000000000..a0ca9dcf6908
--- /dev/null
+++ b/tests/queries/0_stateless/02508_bad_graphite.sql
@@ -0,0 +1,6 @@
+DROP TABLE IF EXISTS test_graphite;
+create table test_graphite (key UInt32, Path String, Time DateTime('UTC'), Value UInt8, Version UInt32, col UInt64)
+    engine = GraphiteMergeTree('graphite_rollup') order by key;
+
+INSERT INTO test_graphite (key) VALUES (0); -- { serverError BAD_ARGUMENTS }
+DROP TABLE test_graphite;
