diff --git a/src/server/stream_family.cc b/src/server/stream_family.cc
index cafb245589cd..d73c51ad2b04 100644
--- a/src/server/stream_family.cc
+++ b/src/server/stream_family.cc
@@ -762,6 +762,7 @@ void StreamFamily::XRange(CmdArgList args, ConnectionContext* cntx) {
 }
 
 void StreamFamily::XRevRange(CmdArgList args, ConnectionContext* cntx) {
+  swap(args[1], args[2]);
   XRangeGeneric(std::move(args), true, cntx);
 }
 
