diff --git a/test/optimizer/distributivity_rule.test b/test/optimizer/distributivity_rule.test
index 8fec60298d97..ca12e4dfbdee 100644
--- a/test/optimizer/distributivity_rule.test
+++ b/test/optimizer/distributivity_rule.test
@@ -31,10 +31,6 @@ query I nosort distributivity3
 EXPLAIN SELECT X OR X FROM test
 ----
 
-query I nosort distributivity3
-EXPLAIN SELECT X FROM test
-----
-
 query I nosort distributivity3
 EXPLAIN SELECT X OR X OR X OR X FROM test
 ----
@@ -43,13 +39,13 @@ query I nosort distributivity3
 EXPLAIN SELECT X OR (X OR (X OR X)) FROM test
 ----
 
-# X OR (X AND A) => X AND A
+# X OR (X AND A) => X
 query I nosort distributivity4
 EXPLAIN SELECT X OR (X AND A) FROM test
 ----
 
 query I nosort distributivity4
-EXPLAIN SELECT X AND A FROM test
+EXPLAIN SELECT X OR X FROM test
 ----
 
 statement ok
diff --git a/test/sql/function/string/test_issue_1812.test b/test/sql/function/string/test_issue_1812.test
new file mode 100644
index 000000000000..787a24f1249f
--- /dev/null
+++ b/test/sql/function/string/test_issue_1812.test
@@ -0,0 +1,58 @@
+# name: test/sql/function/string/test_issue_1812.test
+# description: Test LIKE statement
+# group: [string]
+
+statement ok
+PRAGMA enable_verification
+
+statement ok
+CREATE TABLE t (str VARCHAR);
+
+statement ok
+INSERT INTO t VALUES ('hello1'), ('hello2'), ('hello3'), ('world1'), ('world2'), ('world3');
+
+query I
+SELECT COUNT(*) FROM t WHERE str LIKE '%o%'; -- 6
+----
+6
+
+query I
+SELECT COUNT(*) FROM t WHERE str LIKE '%rld%'; -- 3
+----
+3
+
+query I
+SELECT COUNT(*) FROM t WHERE str LIKE '%o%' OR (str LIKE '%o%' AND str LIKE '%rld%');
+----
+6
+
+query I
+SELECT COUNT(*) FROM t
+WHERE (str LIKE '%o%' AND str LIKE '%rld%')
+    OR str LIKE '%o%';
+----
+6
+
+query I
+SELECT COUNT(*) FROM t
+WHERE (str LIKE '%o%' AND str LIKE '%rld%')
+   OR (str LIKE '%o%')
+   OR (str LIKE '%o%');
+----
+6
+
+query I
+SELECT COUNT(*) FROM t
+WHERE (str LIKE '%o%' AND str LIKE '%rld%')
+   OR (str LIKE '%o%')
+   OR (str LIKE '%o%' AND str LIKE 'blabla%');
+----
+6
+
+query I
+SELECT COUNT(*) FROM t
+WHERE (str LIKE '%o%' AND str LIKE '%1%')
+   OR (str LIKE '%o%' AND str LIKE '%1%' AND str LIKE 'blabla%')
+   OR (str LIKE '%o%' AND str LIKE '%1%' AND str LIKE 'blabla2%')
+----
+2
