diff --git a/src/Storages/StorageBuffer.cpp b/src/Storages/StorageBuffer.cpp
index 4b038692b777..659df2026c80 100644
--- a/src/Storages/StorageBuffer.cpp
+++ b/src/Storages/StorageBuffer.cpp
@@ -221,17 +221,20 @@ Pipe StorageBuffer::read(
                     columns_intersection, destination_metadata_snapshot, query_info,
                     context, processed_stage, max_block_size, num_streams);
 
-                pipe_from_dst.addSimpleTransform([&](const Block & stream_header)
+                if (!pipe_from_dst.empty())
                 {
-                    return std::make_shared<AddingMissedTransform>(stream_header, header_after_adding_defaults,
-                        metadata_snapshot->getColumns(), context);
-                });
-
-                pipe_from_dst.addSimpleTransform([&](const Block & stream_header)
-                {
-                    return std::make_shared<ConvertingTransform>(
-                        stream_header, header, ConvertingTransform::MatchColumnsMode::Name);
-                });
+                    pipe_from_dst.addSimpleTransform([&](const Block & stream_header)
+                    {
+                        return std::make_shared<AddingMissedTransform>(stream_header, header_after_adding_defaults,
+                            metadata_snapshot->getColumns(), context);
+                    });
+
+                    pipe_from_dst.addSimpleTransform([&](const Block & stream_header)
+                    {
+                        return std::make_shared<ConvertingTransform>(
+                            stream_header, header, ConvertingTransform::MatchColumnsMode::Name);
+                    });
+                }
             }
         }
 
