diff --git a/tests/queries/0_stateless/01514_empty_buffer_different_types.reference b/tests/queries/0_stateless/01514_empty_buffer_different_types.reference
new file mode 100644
index 000000000000..78981922613b
--- /dev/null
+++ b/tests/queries/0_stateless/01514_empty_buffer_different_types.reference
@@ -0,0 +1,1 @@
+a
diff --git a/tests/queries/0_stateless/01514_empty_buffer_different_types.sql b/tests/queries/0_stateless/01514_empty_buffer_different_types.sql
new file mode 100644
index 000000000000..3afadbcd33f1
--- /dev/null
+++ b/tests/queries/0_stateless/01514_empty_buffer_different_types.sql
@@ -0,0 +1,12 @@
+set send_logs_level = 'error';
+
+DROP TABLE IF EXISTS merge_tree_table1;
+CREATE TABLE merge_tree_table1 (`s` LowCardinality(String), x UInt32) ENGINE = MergeTree ORDER BY x settings index_granularity = 1;
+CREATE TABLE buffer_table1 ( `s` String , x UInt32) ENGINE = Buffer(currentDatabase(), 'merge_tree_table1', 16, 10, 60, 10, 1000, 1048576, 2097152);
+SELECT s FROM buffer_table1;
+
+insert into merge_tree_table1 values ('a', 1);
+select s from buffer_table1 where x = 1;
+select s from buffer_table1 where x = 2;
+
+DROP TABLE IF EXISTS merge_tree_table1;
