diff --git a/extension/parquet/parquet_reader.cpp b/extension/parquet/parquet_reader.cpp
index 9e6872afa166..e65e19dd894c 100644
--- a/extension/parquet/parquet_reader.cpp
+++ b/extension/parquet/parquet_reader.cpp
@@ -147,7 +147,7 @@ static unique_ptr<ColumnReader> CreateReaderRecursive(const FileMetaData *file_m
 
 	if (s_ele.__isset.repetition_type) {
 		if (s_ele.repetition_type != FieldRepetitionType::REQUIRED) {
-			max_define = depth;
+			max_define++;
 		}
 		if (s_ele.repetition_type == FieldRepetitionType::REPEATED) {
 			max_repeat++;
