diff --git a/test/sql/copy/parquet/data/bug1618_struct_strings.parquet b/test/sql/copy/parquet/data/bug1618_struct_strings.parquet
new file mode 100644
index 000000000000..7a0da50fc851
Binary files /dev/null and b/test/sql/copy/parquet/data/bug1618_struct_strings.parquet differ
diff --git a/test/sql/copy/parquet/parquet_1618_struct_strings.test b/test/sql/copy/parquet/parquet_1618_struct_strings.test
new file mode 100644
index 000000000000..8381be047fb7
--- /dev/null
+++ b/test/sql/copy/parquet/parquet_1618_struct_strings.test
@@ -0,0 +1,24 @@
+# name: test/sql/copy/parquet/parquet_1618_struct_strings.test
+# description: Unclear what went wrong here in the past, but its fixed, and let's make sure it never happens again
+# group: [parquet]
+
+require parquet
+
+query I
+SELECT "inner"['str_field'] FROM parquet_scan('test/sql/copy/parquet/data/bug1618_struct_strings.parquet')
+----
+hello
+NULL
+
+query I
+SELECT "inner"['f64_field'] FROM parquet_scan('test/sql/copy/parquet/data/bug1618_struct_strings.parquet')
+----
+NULL
+1.23
+
+query I
+SELECT "inner" FROM parquet_scan('test/sql/copy/parquet/data/bug1618_struct_strings.parquet')
+----
+<str_field: hello, f64_field: NULL>
+<str_field: NULL, f64_field: 1.230000>
+
diff --git a/test/sql/copy/parquet/parquet_1619.test b/test/sql/copy/parquet/parquet_1619.test
index 9c6b8749040e..3b8d37590be3 100644
--- a/test/sql/copy/parquet/parquet_1619.test
+++ b/test/sql/copy/parquet/parquet_1619.test
@@ -7,17 +7,17 @@ require parquet
 query I
 select struct_extract("inner", 'f64_field') from parquet_scan('test/sql/copy/parquet/data/struct.parquet');
 ----
-1.23
 NULL
+1.23
 
 query I
 select ("inner")."f64_field" from parquet_scan('test/sql/copy/parquet/data/struct.parquet');
 ----
-1.23
 NULL
+1.23
 
 query I
 select "inner"['f64_field'] from parquet_scan('test/sql/copy/parquet/data/struct.parquet');
 ----
-1.23
 NULL
+1.23
