diff --git a/tests/queries/0_stateless/01773_case_sensitive_version.reference b/tests/queries/0_stateless/01773_case_sensitive_version.reference
new file mode 100644
index 000000000000..72749c905a31
--- /dev/null
+++ b/tests/queries/0_stateless/01773_case_sensitive_version.reference
@@ -0,0 +1,1 @@
+1	1	1
diff --git a/tests/queries/0_stateless/01773_case_sensitive_version.sql b/tests/queries/0_stateless/01773_case_sensitive_version.sql
new file mode 100644
index 000000000000..27fa1c27b2ac
--- /dev/null
+++ b/tests/queries/0_stateless/01773_case_sensitive_version.sql
@@ -0,0 +1,1 @@
+SELECT version()=Version(), VERSION()=Version(), vErSiOn()=VeRsIoN();
diff --git a/tests/queries/0_stateless/01774_case_sensitive_connection_id.reference b/tests/queries/0_stateless/01774_case_sensitive_connection_id.reference
new file mode 100644
index 000000000000..95bea1a178ce
--- /dev/null
+++ b/tests/queries/0_stateless/01774_case_sensitive_connection_id.reference
@@ -0,0 +1,1 @@
+0	0	0	0	0	0
diff --git a/tests/queries/0_stateless/01774_case_sensitive_connection_id.sql b/tests/queries/0_stateless/01774_case_sensitive_connection_id.sql
new file mode 100644
index 000000000000..5a4f2b5853a0
--- /dev/null
+++ b/tests/queries/0_stateless/01774_case_sensitive_connection_id.sql
@@ -0,0 +1,1 @@
+SELECT connection_id(), CONNECTION_ID(), CoNnEcTiOn_Id(), connectionid(), CONNECTIONID(), CoNnEcTiOnId();
