diff --git a/tests/queries/0_stateless/03271_decimal_monotonic_day_of_week.reference b/tests/queries/0_stateless/03271_decimal_monotonic_day_of_week.reference
new file mode 100644
index 000000000000..b8626c4cff28
--- /dev/null
+++ b/tests/queries/0_stateless/03271_decimal_monotonic_day_of_week.reference
@@ -0,0 +1,1 @@
+4
diff --git a/tests/queries/0_stateless/03271_decimal_monotonic_day_of_week.sql b/tests/queries/0_stateless/03271_decimal_monotonic_day_of_week.sql
new file mode 100644
index 000000000000..66e085c502d6
--- /dev/null
+++ b/tests/queries/0_stateless/03271_decimal_monotonic_day_of_week.sql
@@ -0,0 +1,7 @@
+DROP TABLE IF EXISTS decimal_dt;
+
+CREATE TABLE decimal_dt (timestamp DateTime64(9)) ENGINE=MergeTree() ORDER BY timestamp;
+INSERT INTO decimal_dt VALUES (toDate('2024-11-11')),(toDate('2024-11-12')),(toDate('2024-11-13')),(toDate('2024-11-14')),(toDate('2024-11-15')),(toDate('2024-11-16')),(toDate('2024-11-17'));
+SELECT count() FROM decimal_dt WHERE toDayOfWeek(timestamp) > 3;
+
+DROP TABLE IF EXISTS decimal_dt;
