diff --git a/tests/queries/0_stateless/01451_normalize_query.reference b/tests/queries/0_stateless/01451_normalize_query.reference
new file mode 100644
index 000000000000..67aa3a97998c
--- /dev/null
+++ b/tests/queries/0_stateless/01451_normalize_query.reference
@@ -0,0 +1,22 @@
+SELECT ?
+SELECT ?
+SELECT ?..
+SELECT ?.., 
+SELECT ?.., 
+SELECT ?..
+SELECT ?.. WHERE ?
+SELECT ?.. WHERE ? = ?
+SELECT ?.. WHERE ? = ? AND (x, y)
+SELECT ?.. WHERE ? = ? AND (?, y)
+[?..]
+[?.., x]
+SELECT ?.. WHERE ? = ? AND (?, y) LIMIT ?..
+SELECT ? AS `xyz`
+SELECT ? AS `xyz1`
+SELECT ? AS `xyz11`
+SELECT ? AS `?`
+SELECT ? AS xyz1
+SELECT ? AS xyz11
+SELECT ? xyz11
+SELECT ?, xyz11
+SELECT ?..
diff --git a/tests/queries/0_stateless/01451_normalize_query.sql b/tests/queries/0_stateless/01451_normalize_query.sql
new file mode 100644
index 000000000000..d1e45dea967b
--- /dev/null
+++ b/tests/queries/0_stateless/01451_normalize_query.sql
@@ -0,0 +1,22 @@
+SELECT normalizeQuery('SELECT 1');
+SELECT normalizeQuery('SELECT  1');
+SELECT normalizeQuery('SELECT  1, 1, 1');
+SELECT normalizeQuery('SELECT 1, 1, 1, /* Hwllo */');
+SELECT normalizeQuery('SELECT 1, 1, 1, /* Hello */');
+SELECT normalizeQuery('SELECT 1, 1, 1, /* Hello */ \'abc\'');
+SELECT normalizeQuery('SELECT 1, 1, 1, /* Hello */ \'abc\' WHERE 1');
+SELECT normalizeQuery('SELECT 1, 1, 1, /* Hello */ \'abc\' WHERE 1 = 1');
+SELECT normalizeQuery('SELECT 1, 1, 1, /* Hello */ \'abc\' WHERE 1 = 1 AND (x, y)');
+SELECT normalizeQuery('SELECT 1, 1, 1, /* Hello */ \'abc\' WHERE 1 = 1 AND (1, y)');
+SELECT normalizeQuery('[1, 2, 3]');
+SELECT normalizeQuery('[1, 2, 3, x]');
+SELECT normalizeQuery('SELECT 1, 1, 1, /* Hello */ \'abc\' WHERE 1 = 1 AND (1, y) LIMIT 1, 1');
+SELECT normalizeQuery('SELECT 1 AS `xyz`');
+SELECT normalizeQuery('SELECT 1 AS `xyz1`');
+SELECT normalizeQuery('SELECT 1 AS `xyz11`');
+SELECT normalizeQuery('SELECT 1 AS xyz111');
+SELECT normalizeQuery('SELECT 1 AS xyz1');
+SELECT normalizeQuery('SELECT 1 AS xyz11');
+SELECT normalizeQuery('SELECT 1 xyz11');
+SELECT normalizeQuery('SELECT 1, xyz11');
+SELECT normalizeQuery('SELECT 1, ''xyz11''');
diff --git a/tests/queries/0_stateless/01452_normalized_query_hash.reference b/tests/queries/0_stateless/01452_normalized_query_hash.reference
new file mode 100644
index 000000000000..fcb49fa99454
--- /dev/null
+++ b/tests/queries/0_stateless/01452_normalized_query_hash.reference
@@ -0,0 +1,7 @@
+1
+1
+1
+1
+1
+1
+1
diff --git a/tests/queries/0_stateless/01452_normalized_query_hash.sql b/tests/queries/0_stateless/01452_normalized_query_hash.sql
new file mode 100644
index 000000000000..a888d2b87b5b
--- /dev/null
+++ b/tests/queries/0_stateless/01452_normalized_query_hash.sql
@@ -0,0 +1,7 @@
+SELECT normalizedQueryHash('SELECT 1') = normalizedQueryHash('SELECT 2');
+SELECT normalizedQueryHash('SELECT  1') != normalizedQueryHash('SELECT  1, 1, 1');
+SELECT normalizedQueryHash('SELECT 1, 1, 1, /* Hello */ \'abc\'') = normalizedQueryHash('SELECT 2, 3');
+SELECT normalizedQueryHash('[1, 2, 3]') = normalizedQueryHash('[1, ''x'']');
+SELECT normalizedQueryHash('[1, 2, 3, x]') != normalizedQueryHash('[1, x]');
+SELECT normalizedQueryHash('SELECT 1 AS `xyz`') != normalizedQueryHash('SELECT 1 AS `abc`');
+SELECT normalizedQueryHash('SELECT 1 AS xyz111') = normalizedQueryHash('SELECT 2 AS xyz234');
