diff --git a/dbms/src/DataTypes/DataTypeSet.h b/dbms/src/DataTypes/DataTypeSet.h
index 21ab50a7c88f..7ef0d9312796 100644
--- a/dbms/src/DataTypes/DataTypeSet.h
+++ b/dbms/src/DataTypes/DataTypeSet.h
@@ -1,6 +1,7 @@
 #pragma once
 
 #include <DataTypes/IDataTypeDummy.h>
+#include <Columns/ColumnSet.h>
 
 
 namespace DB
@@ -18,6 +19,9 @@ class DataTypeSet final : public IDataTypeDummy
     bool equals(const IDataType & rhs) const override { return typeid(rhs) == typeid(*this); }
     bool isParametric() const override { return true; }
 
+    // Used for expressions analysis.
+    MutableColumnPtr createColumn() const override { return ColumnSet::create(0, nullptr); }
+
     // Used only for debugging, making it DUMPABLE
     Field getDefault() const override { return Tuple(); }
 };
