diff --git a/src/Functions/format.cpp b/src/Functions/format.cpp
index 41b6d65023b0..036ff9f0c57a 100644
--- a/src/Functions/format.cpp
+++ b/src/Functions/format.cpp
@@ -39,6 +39,7 @@ class FormatFunction : public IFunction
 
     size_t getNumberOfArguments() const override { return 0; }
 
+    bool useDefaultImplementationForConstants() const override { return true; }
     ColumnNumbers getArgumentsThatAreAlwaysConstant() const override { return {0}; }
 
     DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
