diff --git a/dbms/tests/queries/0_stateless/00568_empty_function_with_fixed_string.reference b/dbms/tests/queries/0_stateless/00568_empty_function_with_fixed_string.reference
index ae8f58a27fb2..9c4bdb5d5ec3 100644
--- a/dbms/tests/queries/0_stateless/00568_empty_function_with_fixed_string.reference
+++ b/dbms/tests/queries/0_stateless/00568_empty_function_with_fixed_string.reference
@@ -1,2 +1,3 @@
 \0\0\0\0	1
 \0abc	0
+32768	1
diff --git a/dbms/tests/queries/0_stateless/00568_empty_function_with_fixed_string.sql b/dbms/tests/queries/0_stateless/00568_empty_function_with_fixed_string.sql
index 9ceb0fa4c750..7abc9067b7cc 100644
--- a/dbms/tests/queries/0_stateless/00568_empty_function_with_fixed_string.sql
+++ b/dbms/tests/queries/0_stateless/00568_empty_function_with_fixed_string.sql
@@ -1,2 +1,8 @@
 SELECT toFixedString('', 4) AS str, empty(str) AS is_empty;
 SELECT toFixedString('\0abc', 4) AS str, empty(str) AS is_empty;
+
+DROP TABLE IF EXISTS test.defaulted;
+CREATE TABLE test.defaulted (v6 FixedString(16)) ENGINE=Memory;
+INSERT INTO test.defaulted SELECT toFixedString('::0', 16) FROM numbers(32768);
+SELECT count(), notEmpty(v6) e FROM test.defaulted GROUP BY e;
+DROP TABLE test.defaulted;
