diff --git a/tools/rpkg/R/Connection.R b/tools/rpkg/R/Connection.R
index 4d29e7f4bb6e..35ae97d29dc3 100644
--- a/tools/rpkg/R/Connection.R
+++ b/tools/rpkg/R/Connection.R
@@ -181,7 +181,7 @@ setMethod(
         col_idx <- 1
         for (name in col_names) {
             if (name %in% names(field.types)) {
-                cols <- c(cols, sprintf("#%d::%s %s", col_idx, field.types[name], name))
+                cols <- c(cols, sprintf("#%d::%s %s", col_idx, field.types[name], dbQuoteIdentifier(conn, name)))
             }
             else {
                 cols <- c(cols, sprintf("#%d", col_idx))
