diff --git a/tests/queries/0_stateless/00612_pk_in_tuple.reference b/tests/queries/0_stateless/00612_pk_in_tuple.reference
index 74e8e642f20f..294ab7491c10 100644
--- a/tests/queries/0_stateless/00612_pk_in_tuple.reference
+++ b/tests/queries/0_stateless/00612_pk_in_tuple.reference
@@ -35,3 +35,4 @@ max(key) from tab_00612 any left join (select key, arrayJoin(n.x) as val from ta
 max(key) from tab_00612 any left join (select key, arrayJoin(n.x) as val from tab_00612) using key where (key, val) in ((1, 1), (2, 2))
 2
 1
+1
diff --git a/tests/queries/0_stateless/00612_pk_in_tuple.sql b/tests/queries/0_stateless/00612_pk_in_tuple.sql
index 499474d1b0a4..081e8a75092e 100644
--- a/tests/queries/0_stateless/00612_pk_in_tuple.sql
+++ b/tests/queries/0_stateless/00612_pk_in_tuple.sql
@@ -43,4 +43,6 @@ CREATE TABLE tab_00612 (key1 Int32, id1  Int64, c1 Int64) ENGINE = MergeTree  PA
 insert into tab_00612 values ( -1, 1, 0 );
 SELECT count(*) FROM  tab_00612 PREWHERE id1 IN (1);
 
+SELECT count() FROM tab_00612 WHERE (key1, id1) IN (-1, 1) AND (key1, 1) IN (-1, 1) SETTINGS force_primary_key = 1;
+
 drop table tab_00612;
diff --git a/tests/queries/0_stateless/00612_pk_in_tuple_perf.reference b/tests/queries/0_stateless/00612_pk_in_tuple_perf.reference
new file mode 100644
index 000000000000..5bea74275a13
--- /dev/null
+++ b/tests/queries/0_stateless/00612_pk_in_tuple_perf.reference
@@ -0,0 +1,4 @@
+1
+		"rows_read": 2,
+1
+		"rows_read": 2,
diff --git a/tests/queries/0_stateless/00612_pk_in_tuple_perf.sh b/tests/queries/0_stateless/00612_pk_in_tuple_perf.sh
new file mode 100755
index 000000000000..8f3e89098e62
--- /dev/null
+++ b/tests/queries/0_stateless/00612_pk_in_tuple_perf.sh
@@ -0,0 +1,45 @@
+#!/usr/bin/env bash
+
+
+CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+. $CURDIR/../shell_config.sh
+
+
+
+$CLICKHOUSE_CLIENT --multiquery <<EOF
+DROP TABLE IF EXISTS pk_in_tuple_perf;
+CREATE TABLE pk_in_tuple_perf
+(
+    v UInt64,
+    u UInt32
+) ENGINE = MergeTree()
+ORDER BY v
+SETTINGS index_granularity = 1;
+
+INSERT INTO pk_in_tuple_perf SELECT number, number * 10 FROM numbers(100);
+EOF
+
+query="SELECT count() FROM pk_in_tuple_perf WHERE (v, u) IN ((2, 10), (2, 20))"
+
+$CLICKHOUSE_CLIENT --query "$query"
+$CLICKHOUSE_CLIENT --query "$query FORMAT JSON" | grep "rows_read"
+
+## Test with non-const args in tuple
+
+$CLICKHOUSE_CLIENT --multiquery <<EOF
+DROP TABLE IF EXISTS pk_in_tuple_perf_non_const;
+CREATE TABLE pk_in_tuple_perf_non_const
+(
+    d Date,
+    u UInt32
+) ENGINE = MergeTree()
+ORDER BY (u, d)
+SETTINGS index_granularity = 1;
+
+INSERT INTO pk_in_tuple_perf_non_const SELECT today() - number, number FROM numbers(100);
+EOF
+
+query="SELECT count() FROM pk_in_tuple_perf_non_const WHERE (u, d) IN ((0, today()), (1, today()))"
+
+$CLICKHOUSE_CLIENT --query "$query"
+$CLICKHOUSE_CLIENT --query "$query FORMAT JSON" | grep "rows_read"
