diff --git a/tests/queries/0_stateless/02884_string_distance_function.reference b/tests/queries/0_stateless/02884_string_distance_function.reference
index e8353d894815..cedc23cc84d1 100644
--- a/tests/queries/0_stateless/02884_string_distance_function.reference
+++ b/tests/queries/0_stateless/02884_string_distance_function.reference
@@ -1,9 +1,7 @@
 const arguments byteHammingDistance
 0
-const arguments byteEditDistance
+const arguments editDistance
 6
-const arguments byteJaccardIndex
-0.4
 byteHammingDistance
 1
 7
@@ -32,17 +30,12 @@ mismatches(alias)
 6
 3
 10
-byteJaccardIndex
-0.8571428571428571
-0.8571428571428571
-0
-0.4
-byteEditDistance
+editDistance
 1
 1
 7
 6
-byteLevenshteinDistance
+levenshteinDistance
 1
 1
 7
diff --git a/tests/queries/0_stateless/02884_string_distance_function.sql b/tests/queries/0_stateless/02884_string_distance_function.sql
index d19f8a9f43c5..1ddb9bfbafda 100644
--- a/tests/queries/0_stateless/02884_string_distance_function.sql
+++ b/tests/queries/0_stateless/02884_string_distance_function.sql
@@ -1,9 +1,9 @@
 select 'const arguments byteHammingDistance';
 select byteHammingDistance('abcd', 'abcd');
-select 'const arguments byteEditDistance';
-select byteEditDistance('clickhouse', 'mouse');
-select 'const arguments byteJaccardIndex';
-select byteJaccardIndex('clickhouse', 'mouse');
+select 'const arguments editDistance';
+select editDistance('clickhouse', 'mouse');
+/*select 'const arguments jaccardIndex';
+select jaccardIndex('clickhouse', 'mouse');*/
 
 drop table if exists t;
 create table t
@@ -25,13 +25,13 @@ select mismatches(s1, s2) from t;
 select mismatches('abc', s2) from t;
 select mismatches(s2, 'def') from t;
 
-select 'byteJaccardIndex';
-select byteJaccardIndex(s1, s2) from t;
-select 'byteEditDistance';
-select byteEditDistance(s1, s2) from t;
-select 'byteLevenshteinDistance';
-select byteLevenshteinDistance(s1, s2) from t;
+/*select 'byteJaccardIndex';
+select byteJaccardIndex(s1, s2) from t;*/
+select 'editDistance';
+select editDistance(s1, s2) from t;
+select 'levenshteinDistance';
+select levenshteinDistance(s1, s2) from t;
 
-SELECT byteEditDistance(randomString(power(2, 17)), 'abc'); -- { serverError TOO_LARGE_STRING_SIZE}
+SELECT editDistance(randomString(power(2, 17)), 'abc'); -- { serverError TOO_LARGE_STRING_SIZE}
 
 drop table t;
