diff --git a/tests/queries/0_stateless/03254_part_log_partition_column_is_set.reference b/tests/queries/0_stateless/03254_part_log_partition_column_is_set.reference
new file mode 100644
index 000000000000..c355e3b06a67
--- /dev/null
+++ b/tests/queries/0_stateless/03254_part_log_partition_column_is_set.reference
@@ -0,0 +1,3 @@
+1	1	11
+1	2	12
+0
diff --git a/tests/queries/0_stateless/03254_part_log_partition_column_is_set.sql b/tests/queries/0_stateless/03254_part_log_partition_column_is_set.sql
new file mode 100644
index 000000000000..32cab8f8dd86
--- /dev/null
+++ b/tests/queries/0_stateless/03254_part_log_partition_column_is_set.sql
@@ -0,0 +1,16 @@
+DROP TABLE IF EXISTS test;
+CREATE TABLE test (x UInt8, y UInt8, z String DEFAULT toString(x)) PARTITION BY x ORDER BY x;
+INSERT INTO test (x, y) VALUES (1, 1);
+INSERT INTO test (x, y) VALUES (1, 2);
+OPTIMIZE TABLE test FINAL;
+INSERT INTO test (x, y) VALUES (2, 1);
+ALTER TABLE test DROP PARTITION 2;
+SET mutations_sync = 1;
+ALTER TABLE test UPDATE z = x || y WHERE 1;
+SELECT * FROM test ORDER BY ALL;
+TRUNCATE TABLE test;
+DROP TABLE test SYNC;
+SYSTEM FLUSH LOGS;
+
+-- SELECT * FROM system.part_log WHERE database = currentDatabase() FORMAT Vertical;
+SELECT DISTINCT throwIf(empty(partition)) FROM system.part_log WHERE database = currentDatabase();
