diff --git a/tests/queries/0_stateless/02475_analyzer_join_tree_subquery.reference b/tests/queries/0_stateless/02475_analyzer_join_tree_subquery.reference
new file mode 100644
index 000000000000..623ba8fde995
--- /dev/null
+++ b/tests/queries/0_stateless/02475_analyzer_join_tree_subquery.reference
@@ -0,0 +1,3 @@
+45
+--
+45
diff --git a/tests/queries/0_stateless/02475_analyzer_join_tree_subquery.sql b/tests/queries/0_stateless/02475_analyzer_join_tree_subquery.sql
new file mode 100644
index 000000000000..eda905291669
--- /dev/null
+++ b/tests/queries/0_stateless/02475_analyzer_join_tree_subquery.sql
@@ -0,0 +1,7 @@
+SET allow_experimental_analyzer = 1;
+
+WITH subquery AS (SELECT sum(number) FROM numbers(10)) SELECT * FROM subquery;
+
+SELECT '--';
+
+WITH subquery AS (SELECT sum(number) FROM numbers(10)) SELECT (SELECT * FROM subquery);
