diff --git a/src/main/client_context.cpp b/src/main/client_context.cpp
index afe4a2b133bb..9e28a452fa6c 100644
--- a/src/main/client_context.cpp
+++ b/src/main/client_context.cpp
@@ -1013,6 +1013,7 @@ void ClientContext::TryBindRelation(Relation &relation, vector<ColumnDefinition>
 	D_ASSERT(!relation.GetAlias().empty());
 	D_ASSERT(!relation.ToString().empty());
 #endif
+	client_data->http_state = make_uniq<HTTPState>();
 	RunFunctionInTransaction([&]() {
 		// bind the expressions
 		auto binder = Binder::CreateBinder(*this);
