diff --git a/tests/queries/0_stateless/03033_analyzer_query_parameters.sh b/tests/queries/0_stateless/03033_analyzer_query_parameters.sh
index c821791e4378..cf46067df993 100755
--- a/tests/queries/0_stateless/03033_analyzer_query_parameters.sh
+++ b/tests/queries/0_stateless/03033_analyzer_query_parameters.sh
@@ -4,5 +4,5 @@ CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
 # shellcheck source=../shell_config.sh
 . "$CURDIR"/../shell_config.sh
 
-clickhouse-local --param_rounding 1 --query "SELECT 1 AS x ORDER BY x WITH FILL STEP {rounding:UInt32} SETTINGS allow_experimental_analyzer = 1"
-clickhouse-local --param_rounding 1 --query "SELECT 1 AS x ORDER BY x WITH FILL STEP {rounding:UInt32} SETTINGS allow_experimental_analyzer = 0"
+${CLICKHOUSE_LOCAL} --param_rounding 1 --query "SELECT 1 AS x ORDER BY x WITH FILL STEP {rounding:UInt32} SETTINGS allow_experimental_analyzer = 1"
+${CLICKHOUSE_LOCAL} --param_rounding 1 --query "SELECT 1 AS x ORDER BY x WITH FILL STEP {rounding:UInt32} SETTINGS allow_experimental_analyzer = 0"
diff --git a/tests/queries/0_stateless/03036_udf_user_defined_directory_in_client.reference b/tests/queries/0_stateless/03036_udf_user_defined_directory_in_client.reference
new file mode 100644
index 000000000000..251d054748a5
--- /dev/null
+++ b/tests/queries/0_stateless/03036_udf_user_defined_directory_in_client.reference
@@ -0,0 +1,1 @@
+Unknown function
diff --git a/tests/queries/0_stateless/03036_udf_user_defined_directory_in_client.sh b/tests/queries/0_stateless/03036_udf_user_defined_directory_in_client.sh
new file mode 100755
index 000000000000..e0a145d84566
--- /dev/null
+++ b/tests/queries/0_stateless/03036_udf_user_defined_directory_in_client.sh
@@ -0,0 +1,15 @@
+#!/usr/bin/env bash
+
+CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+# shellcheck source=../shell_config.sh
+. "$CURDIR"/../shell_config.sh
+
+${CLICKHOUSE_CLIENT} --query "DROP TABLE IF EXISTS test"
+${CLICKHOUSE_CLIENT} --query "CREATE TABLE test (s String) ENGINE = Memory"
+
+# Calling an unknown function should not lead to creation of a 'user_defined' directory in the current directory
+${CLICKHOUSE_CLIENT} --query "INSERT INTO test VALUES (xyz('abc'))" 2>&1 | grep -o -F 'Unknown function'
+
+ls -ld user_defined 2> /dev/null
+
+${CLICKHOUSE_CLIENT} --query "DROP TABLE test"
