diff --git a/src/execution/operator/helper/physical_execute.cpp b/src/execution/operator/helper/physical_execute.cpp
index 64211d045b7a..ea574393e703 100644
--- a/src/execution/operator/helper/physical_execute.cpp
+++ b/src/execution/operator/helper/physical_execute.cpp
@@ -9,5 +9,5 @@ void PhysicalExecute::GetChunkInternal(ClientContext &context, DataChunk &chunk,
 }
 
 unique_ptr<PhysicalOperatorState> PhysicalExecute::GetOperatorState() {
-	return make_unique<PhysicalOperatorState>(plan->children.size() > 0 ? plan->children[0].get() : nullptr);
+	return plan->GetOperatorState();
 }
