diff --git a/test/sql/timezone/test_icu_calendar.test b/test/sql/timezone/test_icu_calendar.test
index 8f9a3cda037b..e635c7ebf4e2 100644
--- a/test/sql/timezone/test_icu_calendar.test
+++ b/test/sql/timezone/test_icu_calendar.test
@@ -84,3 +84,31 @@ Showa	1989-01-06 09:00:00+09	{'era': 234, 'year': 64, 'month': 1, 'day': 6}
 Heisei	1989-01-08 09:00:00+09	{'era': 235, 'year': 1, 'month': 1, 'day': 8}
 Reiwa	2019-05-01 09:00:00+09	{'era': 236, 'year': 1, 'month': 5, 'day': 1}
 Reiwa	2022-01-01 09:00:00+09	{'era': 236, 'year': 4, 'month': 1, 'day': 1}
+
+# Issue #5234
+statement ok
+SET CALENDAR='islamic-umalqura';
+
+query I
+SELECT strftime(TIMESTAMPTZ '-260722-3-4 0:3:52',TIMESTAMP '-285441-5-3 8:3:4');
+----
+285442-05-03 (BC) 08:03:04
+
+# Issue #5235
+statement ok
+SET CALENDAR='indian';
+
+query I
+SELECT TIMESTAMPTZ '-276069-9-30 0:0:00 America/Whitehorse';
+----
+276070-10-01 (BC) 18:19:11+09:18
+
+# Issue #5237
+statement ok
+PRAGMA CALENDAR='japanese';
+
+query I
+SELECT strftime(TIMESTAMPTZ '-23831-1-15 2:5:17 America/La_Paz',TIMETZ '0:8:29 America/Cayman');
+----
+00:08:29+00
+
