diff --git a/tests/queries/0_stateless/01373_summing_merge_tree_exclude_partition_key.reference b/tests/queries/0_stateless/01373_summing_merge_tree_exclude_partition_key.reference
new file mode 100644
index 000000000000..3732e9403cec
--- /dev/null
+++ b/tests/queries/0_stateless/01373_summing_merge_tree_exclude_partition_key.reference
@@ -0,0 +1,20 @@
+---
+1000000
+---
+17
+---
+0	17
+---
+0	17
+1	17
+2	17
+3	17
+4	17
+5	17
+6	17
+7	17
+8	17
+9	17
+10	17
+11	17
+12	17
diff --git a/tests/queries/0_stateless/01373_summing_merge_tree_exclude_partition_key.sql b/tests/queries/0_stateless/01373_summing_merge_tree_exclude_partition_key.sql
new file mode 100644
index 000000000000..60c988a2e2f9
--- /dev/null
+++ b/tests/queries/0_stateless/01373_summing_merge_tree_exclude_partition_key.sql
@@ -0,0 +1,26 @@
+DROP TABLE IF EXISTS tt_01373;
+
+CREATE TABLE tt_01373
+(a Int64, d Int64, val Int64) 
+ENGINE = SummingMergeTree PARTITION BY (a) ORDER BY (d);
+
+SYSTEM STOP MERGES;
+
+INSERT INTO tt_01373 SELECT number%13, number%17, 1 from numbers(1000000);
+
+SELECT '---';
+SELECT count(*) FROM tt_01373;
+
+SELECT '---';
+SELECT count(*) FROM tt_01373 FINAL;
+
+SELECT '---';
+SELECT a, count() FROM tt_01373 FINAL GROUP BY a ORDER BY a;
+
+SYSTEM START MERGES;
+
+OPTIMIZE TABLE tt_01373 FINAL;
+SELECT '---';
+SELECT a, count() FROM tt_01373 GROUP BY a ORDER BY a;
+
+DROP TABLE IF EXISTS tt_01373;
diff --git a/tests/queries/0_stateless/01373_summing_merge_tree_explicit_columns_definition.reference b/tests/queries/0_stateless/01373_summing_merge_tree_explicit_columns_definition.reference
new file mode 100644
index 000000000000..e69de29bb2d1
diff --git a/tests/queries/0_stateless/01373_summing_merge_tree_explicit_columns_definition.sql b/tests/queries/0_stateless/01373_summing_merge_tree_explicit_columns_definition.sql
new file mode 100644
index 000000000000..cc456b3a2573
--- /dev/null
+++ b/tests/queries/0_stateless/01373_summing_merge_tree_explicit_columns_definition.sql
@@ -0,0 +1,7 @@
+DROP TABLE IF EXISTS tt_error_1373;
+
+CREATE TABLE tt_error_1373
+( a   Int64, d   Int64, val Int64 ) 
+ENGINE = SummingMergeTree((a, val)) PARTITION BY (a) ORDER BY (d); -- { serverError 36 }
+
+DROP TABLE IF EXISTS tt_error_1373;
\ No newline at end of file
