diff --git a/test/sql/function/list/list_zip.test b/test/sql/function/list/list_zip.test
index 243b14b1df26..52f3810e0a38 100644
--- a/test/sql/function/list/list_zip.test
+++ b/test/sql/function/list/list_zip.test
@@ -315,3 +315,12 @@ SELECT list_zip('')
 statement error
 SELECT list_zip(3, 4)
 ----
+
+statement error
+SELECT list_zip(FALSE)
+----
+
+statement error
+SELECT list_zip(TRUE)
+----
+
