diff --git a/tests/queries/0_stateless/00539_functions_for_working_with_json.reference b/tests/queries/0_stateless/00539_functions_for_working_with_json.reference
index ee7fb68b7c27..c0399f8ab2e0 100644
--- a/tests/queries/0_stateless/00539_functions_for_working_with_json.reference
+++ b/tests/queries/0_stateless/00539_functions_for_working_with_json.reference
@@ -9,5 +9,7 @@ test"string
  "test_string"
  "test\\"string"
  "test\\"string"
+ "{"
+ "["
  ["]", "2", "3"]
  {"nested" : [1,2,3]}
diff --git a/tests/queries/0_stateless/00539_functions_for_working_with_json.sql b/tests/queries/0_stateless/00539_functions_for_working_with_json.sql
index 8a4d1794293a..514b5f2e5ead 100644
--- a/tests/queries/0_stateless/00539_functions_for_working_with_json.sql
+++ b/tests/queries/0_stateless/00539_functions_for_working_with_json.sql
@@ -11,5 +11,7 @@ SELECT visitParamExtractRaw('{"myparam":"test_string"}', 'myparam');
 SELECT visitParamExtractRaw('{"myparam": "test_string"}', 'myparam');
 SELECT visitParamExtractRaw('{"myparam": "test\\"string"}', 'myparam');
 SELECT visitParamExtractRaw('{"myparam": "test\\"string", "other":123}', 'myparam');
+SELECT visitParamExtractRaw('{"myparam": "{"}', 'myparam');
+SELECT visitParamExtractRaw('{"myparam": "["}', 'myparam');
 SELECT visitParamExtractRaw('{"myparam": ["]", "2", "3"], "other":123}', 'myparam');
 SELECT visitParamExtractRaw('{"myparam": {"nested" : [1,2,3]}, "other":123}', 'myparam');
