diff --git a/tests/queries/0_stateless/00273_quantiles.reference b/tests/queries/0_stateless/00273_quantiles.reference
index 616e06841e4f..aefb00846486 100644
--- a/tests/queries/0_stateless/00273_quantiles.reference
+++ b/tests/queries/0_stateless/00273_quantiles.reference
@@ -4,7 +4,7 @@
 [500]
 [0,1,10,50,100,200,300,400,500,600,700,800,900,950,990,999,1000]
 [0,1,10,50,100,200,300,400,500,600,700,800,900,950,990,999,1000]
-[0,0.50100005,9.51,49.55,99.6,199.7,299.8,399.9,500,600.1,700.2,800.3,900.4,950.45,990.49,999.499,1000]
+[0,1,10,50,99.6,199.7,299.8,399.9,500,600.1,700.2,800.3,900.4,950,990,999,1000]
 [0,1,10,50,100,200,300,400,500,600,700,800,900,950,990,999,1000]
 1	333334	[699144.2,835663,967429.2]	[699999,833333,966666]
 2	266667	[426549.5,536255.5,638957.6]	[426665,533332,639999]
diff --git a/tests/queries/0_stateless/00725_quantiles_shard.reference b/tests/queries/0_stateless/00725_quantiles_shard.reference
index 6974bee97355..ec404bb89a19 100644
--- a/tests/queries/0_stateless/00725_quantiles_shard.reference
+++ b/tests/queries/0_stateless/00725_quantiles_shard.reference
@@ -1,4 +1,4 @@
 [4.5,8.100000000000001]
 [5,9]
-[4.5,8.5]
+[4,8]
 [4.5,8.100000000000001]
