diff --git a/src/server/engine_shard.cc b/src/server/engine_shard.cc
index cd07cc6b5b11..8702cc123149 100644
--- a/src/server/engine_shard.cc
+++ b/src/server/engine_shard.cc
@@ -764,7 +764,9 @@ void EngineShard::Heartbeat() {
   if (db_slice.WillBlockOnJournalWrite() || !can_acquire_global_lock) {
     const auto elapsed = std::chrono::system_clock::now() - start;
     if (elapsed > std::chrono::seconds(1)) {
-      LOG_EVERY_T(WARNING, 5) << "Stalled heartbeat() fiber for " << elapsed.count() << " seconds";
+      const auto elapsed_seconds = std::chrono::duration_cast<std::chrono::seconds>(elapsed);
+      LOG_EVERY_T(WARNING, 5) << "Stalled heartbeat() fiber for " << elapsed_seconds.count()
+                              << " seconds";
     }
     return;
   }
diff --git a/src/server/journal/journal.h b/src/server/journal/journal.h
index 21fd18d77d6c..e3986b492de5 100644
--- a/src/server/journal/journal.h
+++ b/src/server/journal/journal.h
@@ -46,7 +46,7 @@ class Journal {
 class JournalFlushGuard {
  public:
   explicit JournalFlushGuard(Journal* journal) : journal_(journal) {
-    if (journal_) {
+    if (journal_ && counter_ == 0) {
       journal_->SetFlushMode(false);
     }
     util::fb2::detail::EnterFiberAtomicSection();
