diff --git a/tests/queries/0_stateless/02841_not_ready_set_constraints.reference b/tests/queries/0_stateless/02841_not_ready_set_constraints.reference
index d81cc0710eb6..daaac9e30302 100644
--- a/tests/queries/0_stateless/02841_not_ready_set_constraints.reference
+++ b/tests/queries/0_stateless/02841_not_ready_set_constraints.reference
@@ -1,1 +1,2 @@
 42
+42
diff --git a/tests/queries/0_stateless/02841_not_ready_set_constraints.sql b/tests/queries/0_stateless/02841_not_ready_set_constraints.sql
index ecdf4d506353..274940f50a3b 100644
--- a/tests/queries/0_stateless/02841_not_ready_set_constraints.sql
+++ b/tests/queries/0_stateless/02841_not_ready_set_constraints.sql
@@ -17,3 +17,27 @@ ENGINE = MergeTree ORDER BY conversation;
 INSERT INTO t2(conversation) VALUES (42);
 
 select * from t2;
+
+drop table t1;
+
+INSERT INTO t2(conversation) VALUES (42); -- { serverError UNKNOWN_TABLE }
+
+drop table t2;
+
+CREATE TABLE t2 (
+    `conversation` UInt64,
+    CONSTRAINT constraint_conversation CHECK conversation IN (SELECT id FROM t1)
+)
+ENGINE = MergeTree ORDER BY conversation;
+
+INSERT INTO t2(conversation) VALUES (42); -- { serverError UNKNOWN_TABLE }
+
+CREATE TABLE t1 (
+    `id` UInt64
+)
+ENGINE = MergeTree ORDER BY id;
+
+INSERT INTO t1(id) VALUES (42);
+
+INSERT INTO t2(conversation) VALUES (42);
+select * from t2;
