diff --git a/yt/yt/core/actions/unittests/future_ut.cpp b/yt/yt/core/actions/unittests/future_ut.cpp
index 6f5debceb722..de1f661e1dcd 100644
--- a/yt/yt/core/actions/unittests/future_ut.cpp
+++ b/yt/yt/core/actions/unittests/future_ut.cpp
@@ -33,6 +33,15 @@ struct TNonAssignable
     const int Value = 0;
 };
 
+// void Foo()
+// {
+//     auto a = NYT::MakeFuture<std::string>({})
+//     .ApplyUnique(BIND([](std::string&&) {
+//         return NYT::MakeFuture(std::make_unique<int>(42));
+//     }))
+//     .GetUnique();
+// }
+
 TEST_F(TFutureTest, NoncopyableGet)
 {
     auto f = MakeFuture<std::unique_ptr<int>>(std::make_unique<int>(1));
