diff --git a/tests/queries/0_stateless/01640_distributed_async_insert_compression.reference b/tests/queries/0_stateless/01640_distributed_async_insert_compression.reference
new file mode 100644
index 000000000000..fef47999e2e2
--- /dev/null
+++ b/tests/queries/0_stateless/01640_distributed_async_insert_compression.reference
@@ -0,0 +1,2 @@
+256
+512
diff --git a/tests/queries/0_stateless/01640_distributed_async_insert_compression.sql b/tests/queries/0_stateless/01640_distributed_async_insert_compression.sql
new file mode 100644
index 000000000000..b0a674b83894
--- /dev/null
+++ b/tests/queries/0_stateless/01640_distributed_async_insert_compression.sql
@@ -0,0 +1,16 @@
+DROP TABLE IF EXISTS local;
+DROP TABLE IF EXISTS distributed;
+
+CREATE TABLE local (x UInt8) ENGINE = Memory;
+CREATE TABLE distributed AS local ENGINE = Distributed(test_cluster_two_shards, currentDatabase(), local, x);
+
+SET insert_distributed_sync = 0, network_compression_method = 'zstd';
+
+INSERT INTO distributed SELECT number FROM numbers(256);
+SYSTEM FLUSH DISTRIBUTED distributed;
+
+SELECT count() FROM local;
+SELECT count() FROM distributed;
+
+DROP TABLE local;
+DROP TABLE distributed;
