diff --git a/tests/queries/0_stateless/01532_min_max_with_modifiers.reference b/tests/queries/0_stateless/01532_min_max_with_modifiers.reference
new file mode 100644
index 000000000000..9e1683ca6f9c
--- /dev/null
+++ b/tests/queries/0_stateless/01532_min_max_with_modifiers.reference
@@ -0,0 +1,20 @@
+totals
+1	1	1
+2	2	2
+3	3	3
+
+0	1	3
+rollup
+1	1	1
+2	2	2
+3	3	3
+0	1	3
+cube
+1	1	1
+2	2	2
+3	3	3
+0	1	3
+=======
+1	1	2	1
+2	2	3	1
+0	1	3	2
diff --git a/tests/queries/0_stateless/01532_min_max_with_modifiers.sql b/tests/queries/0_stateless/01532_min_max_with_modifiers.sql
new file mode 100644
index 000000000000..0c8651c0f010
--- /dev/null
+++ b/tests/queries/0_stateless/01532_min_max_with_modifiers.sql
@@ -0,0 +1,18 @@
+SELECT 'totals';
+SELECT number % 3 + 1 AS n, min(n), max(n) FROM numbers(100) GROUP BY n WITH TOTALS;
+SELECT 'rollup';
+SELECT number % 3 + 1 AS n, min(n), max(n) FROM numbers(100) GROUP BY n WITH ROLLUP;
+SELECT 'cube';
+SELECT number % 3 + 1 AS n, min(n), max(n) FROM numbers(100) GROUP BY n WITH CUBE;
+SELECT '=======';
+
+SELECT
+    x,
+    min(x) AS lower,
+    max(x) + 1 AS upper,
+    upper - lower AS range
+FROM 
+(
+    SELECT arrayJoin([1, 2]) AS x
+) 
+GROUP BY x WITH ROLLUP;
