diff --git a/test/sql/attach/attach_reserved.test b/test/sql/attach/attach_reserved.test
index 986f7801ce2b..f1a4709ec6b6 100644
--- a/test/sql/attach/attach_reserved.test
+++ b/test/sql/attach/attach_reserved.test
@@ -18,6 +18,16 @@ CREATE TABLE temp_db.integers(i INTEGER);
 statement ok
 DETACH temp_db;
 
+# attach a new database called temp
+statement ok
+ATTACH DATABASE '__TEST_DIR__/system.db';
+
+statement ok
+CREATE TABLE system_db.integers(i INTEGER);
+
+statement ok
+DETACH system_db;
+
 # explicitly selecting these aliases leads to an error
 statement error
 ATTACH DATABASE ':memory:' AS temp;
@@ -28,3 +38,8 @@ statement error
 ATTACH DATABASE ':memory:' AS main;
 ----
 reserved name
+
+statement error
+ATTACH DATABASE ':memory:' AS system;
+----
+reserved name
