diff --git a/tests/integration/test_http_handlers_config/test.py b/tests/integration/test_http_handlers_config/test.py
index e73324ada8f5..9c64bd41b23d 100644
--- a/tests/integration/test_http_handlers_config/test.py
+++ b/tests/integration/test_http_handlers_config/test.py
@@ -65,14 +65,28 @@ def test_dynamic_query_handler():
             ).status_code
         )
 
+        res_default = cluster.instance.http_request(
+            "test_dynamic_handler_get?max_threads=1&get_dynamic_handler_query="
+            + test_query,
+            method="GET",
+            headers={"XXX": "xxx"},
+        )
+        assert 200 == res_default.status_code
         assert (
-            200
-            == cluster.instance.http_request(
-                "test_dynamic_handler_get?max_threads=1&get_dynamic_handler_query="
-                + test_query,
-                method="GET",
-                headers={"XXX": "xxx"},
-            ).status_code
+            "text/tab-separated-values; charset=UTF-8"
+            == res_default.headers["content-type"]
+        )
+
+        res_custom_ct = cluster.instance.http_request(
+            "test_dynamic_handler_get_custom_content_type?max_threads=1&get_dynamic_handler_query="
+            + test_query,
+            method="GET",
+            headers={"XXX": "xxx"},
+        )
+        assert 200 == res_custom_ct.status_code
+        assert (
+            "application/whatever; charset=cp1337"
+            == res_custom_ct.headers["content-type"]
         )
 
 
@@ -116,22 +130,22 @@ def test_predefined_query_handler():
             ).status_code
         )
 
+        res1 = cluster.instance.http_request(
+            "test_predefined_handler_get?max_threads=1&setting_name=max_threads",
+            method="GET",
+            headers={"XXX": "xxx"},
+        )
+        assert b"max_threads\t1
" == res1.content
         assert (
-            b"max_threads\t1
"
-            == cluster.instance.http_request(
-                "test_predefined_handler_get?max_threads=1&setting_name=max_threads",
-                method="GET",
-                headers={"XXX": "xxx"},
-            ).content
+            "text/tab-separated-values; charset=UTF-8" == res1.headers["content-type"]
         )
 
-        assert (
-            b"max_final_threads\t1
max_threads\t1
"
-            == cluster.instance.http_request(
-                "query_param_with_url/max_threads?max_threads=1&max_final_threads=1",
-                headers={"XXX": "max_final_threads"},
-            ).content
+        res2 = cluster.instance.http_request(
+            "query_param_with_url/max_threads?max_threads=1&max_final_threads=1",
+            headers={"XXX": "max_final_threads"},
         )
+        assert b"max_final_threads\t1
max_threads\t1
" == res2.content
+        assert "application/generic+one" == res2.headers["content-type"]
 
 
 def test_fixed_static_handler():
diff --git a/tests/integration/test_http_handlers_config/test_dynamic_handler/config.xml b/tests/integration/test_http_handlers_config/test_dynamic_handler/config.xml
index 60ae98e177b9..68e5ba302dba 100644
--- a/tests/integration/test_http_handlers_config/test_dynamic_handler/config.xml
+++ b/tests/integration/test_http_handlers_config/test_dynamic_handler/config.xml
@@ -11,5 +11,15 @@
                 <query_param_name>get_dynamic_handler_query</query_param_name>
             </handler>
         </rule>
+        <rule>
+            <headers><XXX>xxx</XXX></headers>
+            <methods>GET</methods>
+            <url>/test_dynamic_handler_get_custom_content_type</url>
+            <handler>
+                <type>dynamic_query_handler</type>
+                <query_param_name>get_dynamic_handler_query</query_param_name>
+                <content_type>application/whatever; charset=cp1337</content_type>
+            </handler>
+        </rule>
     </http_handlers>
 </clickhouse>
diff --git a/tests/integration/test_http_handlers_config/test_predefined_handler/config.xml b/tests/integration/test_http_handlers_config/test_predefined_handler/config.xml
index 4aa41a777d4e..96fc018ae3e1 100644
--- a/tests/integration/test_http_handlers_config/test_predefined_handler/config.xml
+++ b/tests/integration/test_http_handlers_config/test_predefined_handler/config.xml
@@ -19,6 +19,7 @@
             <handler>
                 <type>predefined_query_handler</type>
                 <query>SELECT name, value FROM system.settings WHERE name = {setting_name_1:String} OR name = {setting_name_2:String}</query>
+                <content_type>application/generic+one</content_type>
             </handler>
         </rule>
     </http_handlers>
