diff --git a/src/Functions/formatDateTime.cpp b/src/Functions/formatDateTime.cpp
index f63ecbf6146e..1c5dc05c5aa6 100644
--- a/src/Functions/formatDateTime.cpp
+++ b/src/Functions/formatDateTime.cpp
@@ -205,13 +205,13 @@ class FunctionFormatDateTimeImpl : public IFunction
             return 4;
         }
 
-        /// Cast content from integer to string, and append result string to buffer.
-        /// Make sure digits number in result string is no less than total_digits by padding leading '0'
+        /// Casts val from integer to string, then appends result string to buffer.
+        /// Makes sure digits number in result string is no less than min_digits by padding leading '0'.
         /// Notice: '-' is not counted as digit.
         /// For example:
-        /// val = -123, total_digits = 2 => dest = "-123"
-        /// val = -123, total_digits = 3 => dest = "-123"
-        /// val = -123, total_digits = 4 => dest = "-0123"
+        /// val = -123, min_digits = 2 => dest = "-123"
+        /// val = -123, min_digits = 3 => dest = "-123"
+        /// val = -123, min_digits = 4 => dest = "-0123"
         static size_t writeNumberWithPadding(char * dest, std::integral auto val, size_t min_digits)
         {
             using T = decltype(val);
@@ -226,9 +226,10 @@ class FunctionFormatDateTimeImpl : public IFunction
                 ++digits;
             }
 
-            /// Possible sign
             size_t pos = 0;
             n = val;
+
+            /// Possible sign
             if constexpr (is_signed_v<T>)
                 if (val < 0)
                 {
@@ -245,16 +246,17 @@ class FunctionFormatDateTimeImpl : public IFunction
             }
 
             /// Digits
+            size_t digits_written = 0;
             while (w >= 100)
             {
                 w /= 100;
 
                 writeNumber2(dest + pos, n / w);
                 pos += 2;
-
+                digits_written += 2;
                 n = n % w;
             }
-            if (n)
+            if (digits_written < digits)
             {
                 dest[pos] = '0' + n;
                 ++pos;
