diff --git a/dbms/src/Parsers/ASTShowTablesQuery.cpp b/dbms/src/Parsers/ASTShowTablesQuery.cpp
index 7742cfcf06b3..4409ec63e28e 100644
--- a/dbms/src/Parsers/ASTShowTablesQuery.cpp
+++ b/dbms/src/Parsers/ASTShowTablesQuery.cpp
@@ -30,7 +30,7 @@ void ASTShowTablesQuery::formatQueryImpl(const FormatSettings & settings, Format
                 << backQuoteIfNeed(from);
 
         if (!like.empty())
-            settings.ostr << (settings.hilite ? hilite_keyword : "") << " LIKE " << (settings.hilite ? hilite_none : "")
+            settings.ostr << (settings.hilite ? hilite_keyword : "") << (not_like ? " NOT" : "") << " LIKE " << (settings.hilite ? hilite_none : "")
                 << std::quoted(like, '\'');
 
         if (limit_length)
