diff --git a/tests/queries/0_stateless/01386_negative_float_constant_key_condition.reference b/tests/queries/0_stateless/01386_negative_float_constant_key_condition.reference
index 44e0be8e3569..bb0b1cf658d1 100644
--- a/tests/queries/0_stateless/01386_negative_float_constant_key_condition.reference
+++ b/tests/queries/0_stateless/01386_negative_float_constant_key_condition.reference
@@ -1,4 +1,3 @@
 0
 0
 0
-0
diff --git a/tests/queries/0_stateless/01386_negative_float_constant_key_condition.sql b/tests/queries/0_stateless/01386_negative_float_constant_key_condition.sql
index 216f43c42855..c2191d6ab968 100644
--- a/tests/queries/0_stateless/01386_negative_float_constant_key_condition.sql
+++ b/tests/queries/0_stateless/01386_negative_float_constant_key_condition.sql
@@ -12,7 +12,7 @@ SETTINGS index_granularity = 8192;
 INSERT INTO t0 VALUES (0, 0);
 
 SELECT t0.c1 FROM t0 WHERE NOT (t0.c1 OR (t0.c0 AND -1524532316));
-SELECT t0.c1 FROM t0 WHERE NOT (t0.c1 OR (t0.c0 AND -1.0));
+SELECT t0.c1 FROM t0 WHERE NOT (t0.c1 OR (t0.c0 AND -1.0)); -- { serverError 70 }
 SELECT t0.c1 FROM t0 WHERE NOT (t0.c1 OR (t0.c0 AND inf));
 SELECT t0.c1 FROM t0 WHERE NOT (t0.c1 OR (t0.c0 AND nan));
 
diff --git a/tests/queries/0_stateless/01804_uniq_up_to_ubsan.reference b/tests/queries/0_stateless/01804_uniq_up_to_ubsan.reference
new file mode 100644
index 000000000000..e69de29bb2d1
diff --git a/tests/queries/0_stateless/01804_uniq_up_to_ubsan.sql b/tests/queries/0_stateless/01804_uniq_up_to_ubsan.sql
new file mode 100644
index 000000000000..fcbe585b70ab
--- /dev/null
+++ b/tests/queries/0_stateless/01804_uniq_up_to_ubsan.sql
@@ -0,0 +1,2 @@
+SELECT uniqUpTo(1e100)(number) FROM numbers(5); -- { serverError 70 }
+SELECT uniqUpTo(-1e100)(number) FROM numbers(5); -- { serverError 70 }
