diff --git a/tests/queries/0_stateless/03095_group_by_server_constants_bug.reference b/tests/queries/0_stateless/03095_group_by_server_constants_bug.reference
new file mode 100644
index 000000000000..80ab3c879bb0
--- /dev/null
+++ b/tests/queries/0_stateless/03095_group_by_server_constants_bug.reference
@@ -0,0 +1,1 @@
+r1	2
diff --git a/tests/queries/0_stateless/03095_group_by_server_constants_bug.sql b/tests/queries/0_stateless/03095_group_by_server_constants_bug.sql
new file mode 100644
index 000000000000..9f9fda1ef620
--- /dev/null
+++ b/tests/queries/0_stateless/03095_group_by_server_constants_bug.sql
@@ -0,0 +1,5 @@
+SELECT serverUUID() AS s, count() FROM remote('127.0.0.{1,2}', system.one) GROUP BY s format Null;
+
+select getMacro('replica') as s, count() from remote('127.0.0.{1,2}', system.one) group by s;
+
+select uptime() as s, count() FROM remote('127.0.0.{1,2}', system.one) group by s format Null;
