diff --git a/tests/queries/0_stateless/03538_analyzer_lag_lead_functions.reference b/tests/queries/0_stateless/03538_analyzer_lag_lead_functions.reference
new file mode 100644
index 000000000000..3584257c2410
--- /dev/null
+++ b/tests/queries/0_stateless/03538_analyzer_lag_lead_functions.reference
@@ -0,0 +1,13 @@
+   ┏━━━━━━━━┳━━━━━━┳━━━━━━━━┳━━━━━━━━━┳━━━━━━┳━━━━━━━━━━┳━━━━━━━━━┓
+   ┃ number ┃  lag ┃ lagASC ┃ lagDESC ┃ lead ┃ leadDESC ┃ leadASC ┃
+   ┡━━━━━━━━╇━━━━━━╇━━━━━━━━╇━━━━━━━━━╇━━━━━━╇━━━━━━━━━━╇━━━━━━━━━┩
+1. │      0 │ 8472 │   8472 │       1 │    1 │     8472 │       1 │
+   ├────────┼──────┼────────┼─────────┼──────┼──────────┼─────────┤
+2. │      1 │    0 │      0 │       2 │    2 │        0 │       2 │
+   ├────────┼──────┼────────┼─────────┼──────┼──────────┼─────────┤
+3. │      2 │    1 │      1 │       3 │    3 │        1 │       3 │
+   ├────────┼──────┼────────┼─────────┼──────┼──────────┼─────────┤
+4. │      3 │    2 │      2 │       4 │    4 │        2 │       4 │
+   ├────────┼──────┼────────┼─────────┼──────┼──────────┼─────────┤
+5. │      4 │    3 │      3 │    8472 │ 8472 │        3 │    8472 │
+   └────────┴──────┴────────┴─────────┴──────┴──────────┴─────────┘
diff --git a/tests/queries/0_stateless/03538_analyzer_lag_lead_functions.sql b/tests/queries/0_stateless/03538_analyzer_lag_lead_functions.sql
new file mode 100644
index 000000000000..ad9117fbe28f
--- /dev/null
+++ b/tests/queries/0_stateless/03538_analyzer_lag_lead_functions.sql
@@ -0,0 +1,38 @@
+set enable_analyzer = 1;
+
+SELECT number
+    ,lag(number, 1, 8472) OVER () lag
+    ,lag(number, 1, 8472) OVER (ORDER BY number ASC) lagASC
+    ,lag(number, 1, 8472) OVER (ORDER BY number DESC) lagDESC
+    ,lead(number, 1, 8472) OVER () lead
+    ,lead(number, 1, 8472) OVER (ORDER BY number DESC) leadDESC
+    ,lead(number, 1, 8472) OVER (ORDER BY number ASC) leadASC
+FROM numbers(5)
+ORDER BY number
+FORMAT Pretty;
+
+SELECT number
+    ,lead(number, 1, 8472) OVER (ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING) lead
+FROM numbers(5)
+ORDER BY number
+FORMAT Pretty; -- { serverError BAD_ARGUMENTS }
+
+SELECT number
+    ,lag(number, 1, 8472) OVER (ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING) lag
+FROM numbers(5)
+ORDER BY number
+FORMAT Pretty; -- { serverError BAD_ARGUMENTS }
+
+set enable_analyzer = 0;
+
+SELECT number
+    ,lead(number, 1, 8472) OVER () lead
+FROM numbers(5)
+ORDER BY number
+FORMAT Pretty; -- { serverError NOT_IMPLEMENTED }
+
+SELECT number
+    ,lag(number, 1, 8472) OVER () lag
+FROM numbers(5)
+ORDER BY number
+FORMAT Pretty; -- { serverError NOT_IMPLEMENTED }
