diff --git a/tests/queries/0_stateless/01450_set_null_const.reference b/tests/queries/0_stateless/01450_set_null_const.reference
new file mode 100644
index 000000000000..573541ac9702
--- /dev/null
+++ b/tests/queries/0_stateless/01450_set_null_const.reference
@@ -0,0 +1,1 @@
+0
diff --git a/tests/queries/0_stateless/01450_set_null_const.sql b/tests/queries/0_stateless/01450_set_null_const.sql
new file mode 100644
index 000000000000..c47176a880fa
--- /dev/null
+++ b/tests/queries/0_stateless/01450_set_null_const.sql
@@ -0,0 +1,7 @@
+DROP TABLE IF EXISTS test_mtree;
+
+CREATE TABLE test_mtree (`x` String, INDEX idx x TYPE set(10) GRANULARITY 1) ENGINE = MergeTree ORDER BY tuple();
+INSERT INTO test_mtree VALUES ('Hello, world');
+SELECT count() FROM test_mtree WHERE x = NULL;
+
+DROP TABLE test_mtree;
