diff --git a/tests/queries/0_stateless/01303_aggregate_function_nothing_serde.reference b/tests/queries/0_stateless/01303_aggregate_function_nothing_serde.reference
new file mode 100644
index 000000000000..dec7d2fabd2e
--- /dev/null
+++ b/tests/queries/0_stateless/01303_aggregate_function_nothing_serde.reference
@@ -0,0 +1,1 @@
+\N
diff --git a/tests/queries/0_stateless/01303_aggregate_function_nothing_serde.sql b/tests/queries/0_stateless/01303_aggregate_function_nothing_serde.sql
new file mode 100644
index 000000000000..e3b9d7037964
--- /dev/null
+++ b/tests/queries/0_stateless/01303_aggregate_function_nothing_serde.sql
@@ -0,0 +1,1 @@
+select sum(NULL) from remote('127.0.0.{1,2}', system.one);
