diff --git a/tests/queries/0_stateless/01455_nullable_type_with_if_agg_combinator.reference b/tests/queries/0_stateless/01455_nullable_type_with_if_agg_combinator.reference
new file mode 100644
index 000000000000..77f38b722ce3
--- /dev/null
+++ b/tests/queries/0_stateless/01455_nullable_type_with_if_agg_combinator.reference
@@ -0,0 +1,3 @@
+\N	Nullable(UInt8)
+\N	Nullable(UInt8)
+0	UInt8
diff --git a/tests/queries/0_stateless/01455_nullable_type_with_if_agg_combinator.sql b/tests/queries/0_stateless/01455_nullable_type_with_if_agg_combinator.sql
new file mode 100644
index 000000000000..852660117f54
--- /dev/null
+++ b/tests/queries/0_stateless/01455_nullable_type_with_if_agg_combinator.sql
@@ -0,0 +1,6 @@
+-- Value nullable
+SELECT anyIf(CAST(number, 'Nullable(UInt8)'), number = 3) AS a, toTypeName(a) FROM numbers(2);
+-- Value and condition nullable
+SELECT anyIf(number, number = 3) AS a, toTypeName(a) FROM (SELECT CAST(number, 'Nullable(UInt8)') AS number FROM numbers(2));
+-- Condition nullable
+SELECT anyIf(CAST(number, 'UInt8'), number = 3) AS a, toTypeName(a) FROM (SELECT CAST(number, 'Nullable(UInt8)') AS number FROM numbers(2));
diff --git a/tests/queries/0_stateless/01556_if_null.reference b/tests/queries/0_stateless/01556_if_null.reference
index a0c5e7faf406..adc63ecf47b8 100644
--- a/tests/queries/0_stateless/01556_if_null.reference
+++ b/tests/queries/0_stateless/01556_if_null.reference
@@ -1,1 +1,1 @@
-([1],[5])	4	4
+([1],[4])	4	4
diff --git a/tests/queries/0_stateless/01576_if_null_external_aggregation.reference b/tests/queries/0_stateless/01576_if_null_external_aggregation.reference
new file mode 100644
index 000000000000..e69de29bb2d1
diff --git a/tests/queries/0_stateless/01576_if_null_external_aggregation.sql b/tests/queries/0_stateless/01576_if_null_external_aggregation.sql
new file mode 100644
index 000000000000..b9c36a9ceccb
--- /dev/null
+++ b/tests/queries/0_stateless/01576_if_null_external_aggregation.sql
@@ -0,0 +1,7 @@
+SET max_bytes_before_external_group_by = 200000000;
+
+SET max_memory_usage = 1500000000;
+SET max_threads = 12;
+
+SELECT bitAnd(number, pow(2, 20) - 1) as k, argMaxIf(k, number % 2 = 0 ? number : Null, number > 42),  uniq(number) AS u FROM numbers(1000000) GROUP BY k format Null;
+
