diff --git a/tests/queries/0_stateless/02962_indexHint_rpn_construction.reference b/tests/queries/0_stateless/02962_indexHint_rpn_construction.reference
new file mode 100644
index 000000000000..e69de29bb2d1
diff --git a/tests/queries/0_stateless/02962_indexHint_rpn_construction.sql b/tests/queries/0_stateless/02962_indexHint_rpn_construction.sql
new file mode 100644
index 000000000000..3532bea57fa0
--- /dev/null
+++ b/tests/queries/0_stateless/02962_indexHint_rpn_construction.sql
@@ -0,0 +1,20 @@
+CREATE TABLE tab
+(
+    `foo` Array(LowCardinality(String)),
+    INDEX idx foo TYPE bloom_filter GRANULARITY 1
+)
+ENGINE = MergeTree
+PRIMARY KEY tuple();
+
+INSERT INTO tab SELECT if(number % 2, ['value'], [])
+FROM system.numbers
+LIMIT 10000;
+
+SELECT *
+FROM tab
+PREWHERE indexHint(indexHint(-1, 0.))
+WHERE has(foo, 'b');
+
+SELECT *
+FROM tab
+PREWHERE indexHint(0);
