diff --git a/src/server/cluster/cluster_family_test.cc b/src/server/cluster/cluster_family_test.cc
index e26dba47a683..965dc34ca0ef 100644
--- a/src/server/cluster/cluster_family_test.cc
+++ b/src/server/cluster/cluster_family_test.cc
@@ -188,7 +188,7 @@ TEST_F(ClusterFamilyTest, ClusterConfigNoReplicas) {
                                         "abcd1234")))));
 
   EXPECT_EQ(Run({"cluster", "nodes"}),
-            "abcd1234 10.0.0.1:7000@7000 master - 0 0 0 connected 0-16383\r
");
+            "abcd1234 10.0.0.1:7000@7000 master - 0 0 0 connected 0-16383
");
 }
 
 TEST_F(ClusterFamilyTest, ClusterConfigFull) {
@@ -259,8 +259,8 @@ TEST_F(ClusterFamilyTest, ClusterConfigFull) {
                                         "wxyz")))));
 
   EXPECT_EQ(Run({"cluster", "nodes"}),
-            "abcd1234 10.0.0.1:7000@7000 master - 0 0 0 connected 0-16383\r
"
-            "wxyz 10.0.0.10:8000@8000 slave abcd1234 0 0 0 connected\r
");
+            "abcd1234 10.0.0.1:7000@7000 master - 0 0 0 connected 0-16383
"
+            "wxyz 10.0.0.10:8000@8000 slave abcd1234 0 0 0 connected
");
 }
 
 TEST_F(ClusterFamilyTest, ClusterConfigFullMultipleInstances) {
@@ -383,10 +383,10 @@ TEST_F(ClusterFamilyTest, ClusterConfigFullMultipleInstances) {
                                             "qwerty")))))));
 
   EXPECT_THAT(Run({"cluster", "nodes"}),
-              "abcd1234 10.0.0.1:7000@7000 master - 0 0 0 connected 0-10000\r
"
-              "wxyz 10.0.0.10:8000@8000 slave abcd1234 0 0 0 connected\r
"
-              "efgh7890 10.0.0.2:7001@7001 master - 0 0 0 connected 10001-16383\r
"
-              "qwerty 10.0.0.11:8001@8001 slave efgh7890 0 0 0 connected\r
");
+              "abcd1234 10.0.0.1:7000@7000 master - 0 0 0 connected 0-10000
"
+              "wxyz 10.0.0.10:8000@8000 slave abcd1234 0 0 0 connected
"
+              "efgh7890 10.0.0.2:7001@7001 master - 0 0 0 connected 10001-16383
"
+              "qwerty 10.0.0.11:8001@8001 slave efgh7890 0 0 0 connected
");
 
   absl::InsecureBitGen eng;
   while (true) {
@@ -818,7 +818,7 @@ TEST_F(ClusterFamilyEmulatedTest, ClusterSlots) {
 
 TEST_F(ClusterFamilyEmulatedTest, ClusterNodes) {
   EXPECT_THAT(Run({"cluster", "nodes"}),
-              GetMyId() + " fake-host:6379@6379 myself,master - 0 0 0 connected 0-16383\r
");
+              GetMyId() + " fake-host:6379@6379 myself,master - 0 0 0 connected 0-16383
");
 }
 
 }  // namespace
