diff --git a/src/server/generic_family_test.cc b/src/server/generic_family_test.cc
index 966958015b1c..4eb323a1590b 100644
--- a/src/server/generic_family_test.cc
+++ b/src/server/generic_family_test.cc
@@ -734,4 +734,12 @@ TEST_F(GenericFamilyTest, FieldTtl) {
   EXPECT_EQ(-3, CheckedInt({"fieldttl", "k2", "f4"}));
 }
 
+TEST_F(GenericFamilyTest, RandomKey) {
+  auto resp = Run({"randomkey"});
+  EXPECT_EQ(resp.type, RespExpr::NIL);
+
+  resp = Run({"set", "k1", "1"});
+  EXPECT_EQ(Run({"randomkey"}), "k1");
+}
+
 }  // namespace dfly
