diff --git a/test/sql/copy/csv/csv_external_access.test b/test/sql/copy/csv/csv_external_access.test
new file mode 100644
index 000000000000..743f41f7fbec
--- /dev/null
+++ b/test/sql/copy/csv/csv_external_access.test
@@ -0,0 +1,28 @@
+# name: test/sql/copy/csv/csv_external_access.test
+# description: Test that enable_external_access blocks CSV readers
+# group: [csv]
+
+statement ok
+CREATE TABLE date_test(d date);
+
+statement ok
+COPY date_test FROM 'test/sql/copy/csv/data/test/date.csv';
+
+statement ok
+SET enable_external_access=false;
+
+statement error
+SELECT * FROM read_csv('test/sql/copy/csv/data/test/date.csv', columns = {'d': 'DATE'});
+
+statement error
+SELECT * FROM read_csv_auto('test/sql/copy/csv/data/test/date.csv');
+
+statement error
+COPY date_test FROM 'test/sql/copy/csv/data/test/date.csv';
+
+statement error
+COPY date_test TO '__TEST_DIR__/date.csv'
+
+# we also can't just enable external access again
+statement error
+SET enable_external_access=true;
diff --git a/test/sql/copy/parquet/parquet_external_access.test b/test/sql/copy/parquet/parquet_external_access.test
new file mode 100644
index 000000000000..82731a917cd1
--- /dev/null
+++ b/test/sql/copy/parquet/parquet_external_access.test
@@ -0,0 +1,42 @@
+# name: test/sql/copy/parquet/parquet_external_access.test
+# description: Test that enable_external_access blocks Parquet reads
+# group: [parquet]
+
+require parquet
+
+# we cannot read parquet files
+statement ok
+CREATE TABLE lineitem AS SELECT * FROM 'data/parquet-testing/arrow/lineitem-arrow.parquet'
+
+statement ok
+SET enable_external_access=false;
+
+# we cannot read parquet files
+statement error
+SELECT * FROM 'data/parquet-testing/arrow/lineitem-arrow.parquet'
+
+# or their metadata
+statement error
+SELECT * FROM parquet_metadata('data/parquet-testing/arrow/lineitem-arrow.parquet')
+
+statement error
+SELECT * FROM parquet_schema('data/parquet-testing/arrow/lineitem-arrow.parquet')
+
+# also not in a list
+statement error
+SELECT * FROM parquet_scan(['data/parquet-testing/arrow/lineitem-arrow.parquet', 'data/parquet-testing/arrow/lineitem-arrow.parquet'])
+
+# neither can we glob
+statement error
+SELECT * FROM glob('data/parquet-testing/arrow/lineitem-arrow.parquet')
+
+# or copy to/from...
+statement error
+COPY lineitem FROM 'data/parquet-testing/arrow/lineitem-arrow.parquet'
+
+statement error
+COPY lineitem TO '__TEST_DIR__/lineitem.parquet'
+
+# we also can't just enable external access again
+statement error
+SET enable_external_access=true;
diff --git a/test/sql/export/export_external_access.test b/test/sql/export/export_external_access.test
new file mode 100644
index 000000000000..585c11fa3797
--- /dev/null
+++ b/test/sql/export/export_external_access.test
@@ -0,0 +1,24 @@
+# name: test/sql/export/export_external_access.test
+# description: Test export database
+# group: [export]
+
+statement ok
+BEGIN TRANSACTION
+
+statement ok
+CREATE TABLE integers(i INTEGER, j INTEGER, CHECK(i+j<10))
+
+statement ok
+EXPORT DATABASE '__TEST_DIR__/export_permissions_test' (FORMAT CSV)
+
+statement ok
+ROLLBACK
+
+statement ok
+SET enable_external_access=false
+
+statement error
+IMPORT DATABASE '__TEST_DIR__/export_permissions_test'
+
+statement error
+EXPORT DATABASE '__TEST_DIR__/export_permissions_test2' (FORMAT CSV)
