diff --git a/src/server/server_family_test.cc b/src/server/server_family_test.cc
index f84659270673..88708f4faca2 100644
--- a/src/server/server_family_test.cc
+++ b/src/server/server_family_test.cc
@@ -310,4 +310,10 @@ TEST_F(ServerFamilyTest, ClientTrackingSelectDB) {
   EXPECT_EQ(GetInvalidationMessage("IO0", 0).key, "C");
 }
 
+TEST_F(ServerFamilyTest, ClientTrackingNonTransactionalBug) {
+  Run({"HELLO", "3"});
+  Run({"CLIENT", "TRACKING", "ON"});
+
+  Run({"CLUSTER", "SLOTS"});
+}
 }  // namespace dfly
