diff --git a/tests/queries/0_stateless/02813_array_agg.reference b/tests/queries/0_stateless/02813_array_agg.reference
new file mode 100644
index 000000000000..202f6aa4bac1
--- /dev/null
+++ b/tests/queries/0_stateless/02813_array_agg.reference
@@ -0,0 +1,6 @@
+['hello, world!','hello, world!','hello, world!','hello, world!','hello, world!']
+['hello, world!']
+['hello, world!']
+['hello, world!']
+['hello, world!']
+['hello, world!']
diff --git a/tests/queries/0_stateless/02813_array_agg.sql b/tests/queries/0_stateless/02813_array_agg.sql
new file mode 100644
index 000000000000..91d8d0774d3b
--- /dev/null
+++ b/tests/queries/0_stateless/02813_array_agg.sql
@@ -0,0 +1,10 @@
+drop table if exists t;
+create table t (n Int32, s String) engine=MergeTree order by n;
+
+insert into t select number, 'hello, world!' from numbers (5);
+
+select array_agg(s) from t;
+
+select aRray_Agg(s) from t group by n;
+
+drop table t;
