diff --git a/src/Dictionaries/ExternalQueryBuilder.cpp b/src/Dictionaries/ExternalQueryBuilder.cpp
index 792c4e3e9074..a31301cd0f3b 100644
--- a/src/Dictionaries/ExternalQueryBuilder.cpp
+++ b/src/Dictionaries/ExternalQueryBuilder.cpp
@@ -401,7 +401,7 @@ std::string ExternalQueryBuilder::composeLoadKeysQuery(
         {
             writeString("SELECT * FROM (", out);
             writeString(query, out);
-            writeString(") WHERE ", out);
+            writeString(") AS subquery WHERE ", out);
             composeKeysCondition(key_columns, requested_rows, method, partition_key_prefix, out);
             writeString(";", out);
 
