diff --git a/src/execution/operator/set/physical_recursive_cte.cpp b/src/execution/operator/set/physical_recursive_cte.cpp
index 6b694b7a4dac..a49b6cfb452d 100644
--- a/src/execution/operator/set/physical_recursive_cte.cpp
+++ b/src/execution/operator/set/physical_recursive_cte.cpp
@@ -124,6 +124,9 @@ void PhysicalRecursiveCTE::ExecuteRecursivePipelines(ExecutionContext &context)
 
 	while (true) {
 		executor.WorkOnTasks();
+		if (executor.HasError()) {
+			executor.ThrowException();
+		}
 		bool finished = true;
 		for (auto &event : events) {
 			if (!event->IsFinished()) {
