diff --git a/tests/queries/0_stateless/01497_mutation_support_for_storage_memory.reference b/tests/queries/0_stateless/01497_mutation_support_for_storage_memory.reference
index 5aaf21f91370..29777b0c05c0 100644
--- a/tests/queries/0_stateless/01497_mutation_support_for_storage_memory.reference
+++ b/tests/queries/0_stateless/01497_mutation_support_for_storage_memory.reference
@@ -8,7 +8,9 @@
 3	3
 4	4
 5	5
+5
 2	2
 3	3
 4	4
 5	5
+4
diff --git a/tests/queries/0_stateless/01497_mutation_support_for_storage_memory.sql b/tests/queries/0_stateless/01497_mutation_support_for_storage_memory.sql
index 408487ed205f..0e2861b8e940 100644
--- a/tests/queries/0_stateless/01497_mutation_support_for_storage_memory.sql
+++ b/tests/queries/0_stateless/01497_mutation_support_for_storage_memory.sql
@@ -13,8 +13,12 @@ ALTER TABLE defaults UPDATE n = 100 WHERE s = '1';
 
 SELECT * FROM defaults;
 
+SELECT count(*) FROM defaults;
+
 ALTER TABLE defaults DELETE WHERE n = 100;
 
 SELECT * FROM defaults;
 
+SELECT count(*) FROM defaults;
+
 DROP TABLE defaults;
