diff --git a/programs/local/LocalServer.cpp b/programs/local/LocalServer.cpp
index dbf153eeb81b..1fee37985a79 100644
--- a/programs/local/LocalServer.cpp
+++ b/programs/local/LocalServer.cpp
@@ -273,11 +273,12 @@ try
     global_context->setCurrentDatabase(default_database);
     applyCmdOptions(*global_context);
 
-    String path = global_context->getPath();
-    if (!path.empty())
+    if (config().has("path"))
     {
+        String path = global_context->getPath();
+
         /// Lock path directory before read
-        status.emplace(global_context->getPath() + "status", StatusFile::write_full_info);
+        status.emplace(path + "status", StatusFile::write_full_info);
 
         LOG_DEBUG(log, "Loading metadata from {}", path);
         Poco::File(path + "data/").createDirectories();
