diff --git a/tests/queries/0_stateless/01647_clickhouse_local_hung.reference b/tests/queries/0_stateless/01647_clickhouse_local_hung.reference
new file mode 100644
index 000000000000..8925955b88e8
--- /dev/null
+++ b/tests/queries/0_stateless/01647_clickhouse_local_hung.reference
@@ -0,0 +1,1 @@
+    100 14
diff --git a/tests/queries/0_stateless/01647_clickhouse_local_hung.sh b/tests/queries/0_stateless/01647_clickhouse_local_hung.sh
new file mode 100755
index 000000000000..04f32055ab68
--- /dev/null
+++ b/tests/queries/0_stateless/01647_clickhouse_local_hung.sh
@@ -0,0 +1,10 @@
+#!/usr/bin/env bash
+
+set -e
+
+CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+# shellcheck source=../shell_config.sh
+. "$CURDIR"/../shell_config.sh
+
+
+for _ in {1..100}; do echo 'Hello, world!' | ${CLICKHOUSE_LOCAL} --query "SELECT * FROM table" --structure 's String' | wc -c; done | uniq -c
