diff --git a/tests/integration/test_limit_materialized_view_count/__init__.py b/tests/integration/test_limit_materialized_view_count/__init__.py
new file mode 100644
index 000000000000..e69de29bb2d1
diff --git a/tests/integration/test_limit_materialized_view_count/configs/max_num_limit.xml b/tests/integration/test_limit_materialized_view_count/configs/max_num_limit.xml
new file mode 100644
index 000000000000..7fcf9005a088
--- /dev/null
+++ b/tests/integration/test_limit_materialized_view_count/configs/max_num_limit.xml
@@ -0,0 +1,3 @@
+<clickhouse>
+    <max_materialized_views_count_for_table>1</max_materialized_views_count_for_table>
+</clickhouse>
diff --git a/tests/integration/test_limit_materialized_view_count/test.py b/tests/integration/test_limit_materialized_view_count/test.py
new file mode 100644
index 000000000000..c14c5b2055e8
--- /dev/null
+++ b/tests/integration/test_limit_materialized_view_count/test.py
@@ -0,0 +1,49 @@
+import pytest
+from helpers.cluster import ClickHouseCluster
+
+cluster = ClickHouseCluster(__file__)
+
+node = cluster.add_instance(
+    "node",
+    main_configs=["configs/max_num_limit.xml"],
+    stay_alive=True,
+)
+
+config = """<clickhouse>
+    <max_materialized_views_count_for_table>2</max_materialized_views_count_for_table>
+</clickhouse>
+"""
+
+
+@pytest.fixture(scope="module")
+def started_cluster():
+    try:
+        cluster.start()
+        for _, node in cluster.instances.items():
+            node.query(
+                f"""
+                CREATE TABLE test_tb (a String) ENGINE = MergeTree ORDER BY a;
+                """
+            )
+        yield cluster
+    finally:
+        cluster.shutdown()
+
+
+def test_limit_materialized_view_count(started_cluster):
+    node.query(
+        "CREATE MATERIALIZED VIEW test_view1 ENGINE = MergeTree ORDER BY a AS SELECT * FROM test_tb;"
+    )
+    assert "Too many materialized views" in node.query_and_get_error(
+        "CREATE MATERIALIZED VIEW test_view2 ENGINE = MergeTree ORDER BY a AS SELECT * FROM test_tb;"
+    )
+
+    node.replace_config("/etc/clickhouse-server/config.d/max_num_limit.xml", config)
+    node.restart_clickhouse()
+
+    node.query(
+        "CREATE MATERIALIZED VIEW test_view2 ENGINE = MergeTree ORDER BY a AS SELECT * FROM test_tb;"
+    )
+    assert "Too many materialized views" in node.query_and_get_error(
+        "CREATE MATERIALIZED VIEW test_view3 ENGINE = MergeTree ORDER BY a AS SELECT * FROM test_tb;"
+    )
