diff --git a/tests/clickhouse-test b/tests/clickhouse-test
index a3d7e0e922de..1203ad3730a7 100755
--- a/tests/clickhouse-test
+++ b/tests/clickhouse-test
@@ -59,6 +59,7 @@ MESSAGES_TO_RETRY = [
     "is already started to be removing by another replica right now",
     # This is from LSan, and it indicates its own internal problem:
     "Unable to get registers from thread",
+    "You can retry",
 ]
 
 MAX_RETRIES = 3
