diff --git a/src/Functions/FunctionsComparison.h b/src/Functions/FunctionsComparison.h
index be0875581a5c..9b2328065fc7 100644
--- a/src/Functions/FunctionsComparison.h
+++ b/src/Functions/FunctionsComparison.h
@@ -1033,6 +1033,9 @@ class FunctionComparison : public IFunction
             size_t tuple_size,
             size_t input_rows_count) const
     {
+        if (0 == tuple_size)
+            throw Exception(ErrorCodes::NOT_IMPLEMENTED, "Comparison of zero-sized tuples is not implemented");
+
         ColumnsWithTypeAndName less_columns(tuple_size);
         ColumnsWithTypeAndName equal_columns(tuple_size - 1);
         ColumnsWithTypeAndName tmp_columns(2);
diff --git a/src/Functions/if.cpp b/src/Functions/if.cpp
index e03b27b3c394..5e1e7067e86e 100644
--- a/src/Functions/if.cpp
+++ b/src/Functions/if.cpp
@@ -668,6 +668,9 @@ class FunctionIf : public FunctionIfBase
         temporary_columns[0] = arguments[0];
 
         size_t tuple_size = type1.getElements().size();
+        if (tuple_size == 0)
+            return ColumnTuple::create(input_rows_count);
+
         Columns tuple_columns(tuple_size);
 
         for (size_t i = 0; i < tuple_size; ++i)
