diff --git a/src/Analyzer/Passes/QueryAnalysisPass.cpp b/src/Analyzer/Passes/QueryAnalysisPass.cpp
index 5cdd18b4b2f9..c6fbd728b8f7 100644
--- a/src/Analyzer/Passes/QueryAnalysisPass.cpp
+++ b/src/Analyzer/Passes/QueryAnalysisPass.cpp
@@ -970,6 +970,10 @@ class QueryExpressionsAliasVisitor : public InDepthQueryTreeVisitor<QueryExpress
         if (!node->hasAlias())
             return;
 
+        // We should not resolve expressions to WindowNode
+        if (node->getNodeType() == QueryTreeNodeType::WINDOW)
+            return;
+
         const auto & alias = node->getAlias();
 
         if (is_lambda_node)
