diff --git a/tests/integration/test_catboost_model_reload/__init__.py b/tests/integration/test_catboost_model_reload/__init__.py
new file mode 100644
index 000000000000..e69de29bb2d1
diff --git a/tests/integration/test_catboost_model_reload/config/catboost_lib.xml b/tests/integration/test_catboost_model_reload/config/catboost_lib.xml
new file mode 100644
index 000000000000..745be7cebe62
--- /dev/null
+++ b/tests/integration/test_catboost_model_reload/config/catboost_lib.xml
@@ -0,0 +1,3 @@
+<yandex>
+    <catboost_dynamic_library_path>/etc/clickhouse-server/model/libcatboostmodel.so</catboost_dynamic_library_path>
+</yandex>
diff --git a/tests/integration/test_catboost_model_reload/config/models_config.xml b/tests/integration/test_catboost_model_reload/config/models_config.xml
new file mode 100644
index 000000000000..e84ca8b52855
--- /dev/null
+++ b/tests/integration/test_catboost_model_reload/config/models_config.xml
@@ -0,0 +1,3 @@
+<yandex>
+    <models_config>/etc/clickhouse-server/model/model_config.xml</models_config>
+</yandex>
diff --git a/tests/integration/test_catboost_model_reload/model/conjunction.cbm b/tests/integration/test_catboost_model_reload/model/conjunction.cbm
new file mode 100644
index 000000000000..7b75fb5f8866
Binary files /dev/null and b/tests/integration/test_catboost_model_reload/model/conjunction.cbm differ
diff --git a/tests/integration/test_catboost_model_reload/model/disjunction.cbm b/tests/integration/test_catboost_model_reload/model/disjunction.cbm
new file mode 100644
index 000000000000..8145c24637fc
Binary files /dev/null and b/tests/integration/test_catboost_model_reload/model/disjunction.cbm differ
diff --git a/tests/integration/test_catboost_model_reload/model/libcatboostmodel.so b/tests/integration/test_catboost_model_reload/model/libcatboostmodel.so
new file mode 100755
index 000000000000..388d9f887b46
Binary files /dev/null and b/tests/integration/test_catboost_model_reload/model/libcatboostmodel.so differ
diff --git a/tests/integration/test_catboost_model_reload/model/model_config.xml b/tests/integration/test_catboost_model_reload/model/model_config.xml
new file mode 100644
index 000000000000..7cbda165ce90
--- /dev/null
+++ b/tests/integration/test_catboost_model_reload/model/model_config.xml
@@ -0,0 +1,8 @@
+<models>
+    <model>
+        <type>catboost</type>
+        <name>model</name>
+        <path>/etc/clickhouse-server/model/model.cbm</path>
+        <lifetime>0</lifetime>
+    </model>
+</models>
diff --git a/tests/integration/test_catboost_model_reload/test.py b/tests/integration/test_catboost_model_reload/test.py
new file mode 100644
index 000000000000..8283e6af9750
--- /dev/null
+++ b/tests/integration/test_catboost_model_reload/test.py
@@ -0,0 +1,74 @@
+import os
+import sys
+import time
+
+import pytest
+
+sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
+SCRIPT_DIR = os.path.dirname(os.path.realpath(__file__))
+
+from helpers.cluster import ClickHouseCluster
+
+cluster = ClickHouseCluster(__file__)
+node = cluster.add_instance('node', stay_alive=True, main_configs=['config/models_config.xml', 'config/catboost_lib.xml'])
+
+def copy_file_to_container(local_path, dist_path, container_id):
+    os.system("docker cp {local} {cont_id}:{dist}".format(local=local_path, cont_id=container_id, dist=dist_path))
+
+@pytest.fixture(scope="module")
+def started_cluster():
+    try:
+        cluster.start()
+
+        copy_file_to_container(os.path.join(SCRIPT_DIR, 'model/.'), '/etc/clickhouse-server/model', node.docker_id)
+        node.query("CREATE TABLE binary (x UInt64, y UInt64) ENGINE = TinyLog()")
+        node.query("INSERT INTO binary VALUES (1, 1), (1, 0), (0, 1), (0, 0)")
+
+        node.restart_clickhouse()
+
+        yield cluster
+
+    finally:
+        cluster.shutdown()
+
+def test_model_reload(started_cluster):
+    node.exec_in_container(["bash", "-c", "rm -f /etc/clickhouse-server/model/model.cbm"])
+    node.exec_in_container(["bash", "-c", "ln /etc/clickhouse-server/model/conjunction.cbm /etc/clickhouse-server/model/model.cbm"])
+    node.query("SYSTEM RELOAD MODEL model")
+
+    result = node.query("""
+        WITH modelEvaluate('model', toFloat64(x), toFloat64(y)) as prediction, exp(prediction) / (1 + exp(prediction)) as probability
+        SELECT if(probability > 0.5, 1, 0) FROM binary;
+        """)
+    assert result == '1
0
0
0
'
+
+    node.exec_in_container(["bash", "-c", "rm /etc/clickhouse-server/model/model.cbm"])
+    node.exec_in_container(["bash", "-c", "ln /etc/clickhouse-server/model/disjunction.cbm /etc/clickhouse-server/model/model.cbm"])
+    node.query("SYSTEM RELOAD MODEL model")
+
+    result = node.query("""
+        WITH modelEvaluate('model', toFloat64(x), toFloat64(y)) as prediction, exp(prediction) / (1 + exp(prediction)) as probability
+        SELECT if(probability > 0.5, 1, 0) FROM binary;
+        """)
+    assert result == '1
1
1
0
'
+
+def test_models_reload(started_cluster):
+    node.exec_in_container(["bash", "-c", "rm -f /etc/clickhouse-server/model/model.cbm"])
+    node.exec_in_container(["bash", "-c", "ln /etc/clickhouse-server/model/conjunction.cbm /etc/clickhouse-server/model/model.cbm"])
+    node.query("SYSTEM RELOAD MODELS")
+
+    result = node.query("""
+        WITH modelEvaluate('model', toFloat64(x), toFloat64(y)) as prediction, exp(prediction) / (1 + exp(prediction)) as probability
+        SELECT if(probability > 0.5, 1, 0) FROM binary;
+        """)
+    assert result == '1
0
0
0
'
+
+    node.exec_in_container(["bash", "-c", "rm /etc/clickhouse-server/model/model.cbm"])
+    node.exec_in_container(["bash", "-c", "ln /etc/clickhouse-server/model/disjunction.cbm /etc/clickhouse-server/model/model.cbm"])
+    node.query("SYSTEM RELOAD MODELS")
+
+    result = node.query("""
+        WITH modelEvaluate('model', toFloat64(x), toFloat64(y)) as prediction, exp(prediction) / (1 + exp(prediction)) as probability
+        SELECT if(probability > 0.5, 1, 0) FROM binary;
+        """)
+    assert result == '1
1
1
0
'
\ No newline at end of file
diff --git a/tests/queries/0_stateless/01271_show_privileges.reference b/tests/queries/0_stateless/01271_show_privileges.reference
index c8b8662dc3ee..892bd95d2d96 100644
--- a/tests/queries/0_stateless/01271_show_privileges.reference
+++ b/tests/queries/0_stateless/01271_show_privileges.reference
@@ -82,6 +82,7 @@ SYSTEM DROP CACHE	['DROP CACHE']	\N	SYSTEM
 SYSTEM RELOAD CONFIG	['RELOAD CONFIG']	GLOBAL	SYSTEM RELOAD
 SYSTEM RELOAD SYMBOLS	['RELOAD SYMBOLS']	GLOBAL	SYSTEM RELOAD
 SYSTEM RELOAD DICTIONARY	['SYSTEM RELOAD DICTIONARIES','RELOAD DICTIONARY','RELOAD DICTIONARIES']	GLOBAL	SYSTEM RELOAD
+SYSTEM RELOAD MODEL	['SYSTEM RELOAD MODELS','RELOAD MODEL','RELOAD MODELS']	GLOBAL	SYSTEM RELOAD
 SYSTEM RELOAD EMBEDDED DICTIONARIES	['RELOAD EMBEDDED DICTIONARIES']	GLOBAL	SYSTEM RELOAD
 SYSTEM RELOAD	[]	\N	SYSTEM
 SYSTEM MERGES	['SYSTEM STOP MERGES','SYSTEM START MERGES','STOP_MERGES','START MERGES']	TABLE	SYSTEM
