diff --git a/src/Processors/Transforms/buildPushingToViewsChain.cpp b/src/Processors/Transforms/buildPushingToViewsChain.cpp
index b68ec5472801..a8890f5bccb4 100644
--- a/src/Processors/Transforms/buildPushingToViewsChain.cpp
+++ b/src/Processors/Transforms/buildPushingToViewsChain.cpp
@@ -430,6 +430,7 @@ static QueryPipeline process(Block block, ViewRuntimeData & view, const ViewsDat
     InterpreterSelectQuery select(view.query, local_context, SelectQueryOptions());
     auto pipeline = select.buildQueryPipeline();
     pipeline.resize(1);
+    pipeline.dropTotalsAndExtremes();
 
     /// Squashing is needed here because the materialized view query can generate a lot of blocks
     /// even when only one block is inserted into the parent table (e.g. if the query is a GROUP BY
