diff --git a/tests/queries/0_stateless/02366_window_function_order_by.reference b/tests/queries/0_stateless/02366_window_function_order_by.reference
new file mode 100644
index 000000000000..89f6c6c911a2
--- /dev/null
+++ b/tests/queries/0_stateless/02366_window_function_order_by.reference
@@ -0,0 +1,37 @@
+-- { echoOn }
+SELECT groupArray(tuple(value)) OVER ()
+FROM (select number value from numbers(10))
+ORDER BY value ASC;
+[(0),(1),(2),(3),(4),(5),(6),(7),(8),(9)]
+[(0),(1),(2),(3),(4),(5),(6),(7),(8),(9)]
+[(0),(1),(2),(3),(4),(5),(6),(7),(8),(9)]
+[(0),(1),(2),(3),(4),(5),(6),(7),(8),(9)]
+[(0),(1),(2),(3),(4),(5),(6),(7),(8),(9)]
+[(0),(1),(2),(3),(4),(5),(6),(7),(8),(9)]
+[(0),(1),(2),(3),(4),(5),(6),(7),(8),(9)]
+[(0),(1),(2),(3),(4),(5),(6),(7),(8),(9)]
+[(0),(1),(2),(3),(4),(5),(6),(7),(8),(9)]
+[(0),(1),(2),(3),(4),(5),(6),(7),(8),(9)]
+SELECT count() OVER (ORDER BY number + 1) FROM numbers(10) ORDER BY number;
+1
+2
+3
+4
+5
+6
+7
+8
+9
+10
+SELECT count() OVER (ORDER BY number + 1) + 1 as foo FROM numbers(10)
+ORDER BY foo;
+2
+3
+4
+5
+6
+7
+8
+9
+10
+11
diff --git a/tests/queries/0_stateless/02366_window_function_order_by.sql b/tests/queries/0_stateless/02366_window_function_order_by.sql
new file mode 100644
index 000000000000..a3a02355cf71
--- /dev/null
+++ b/tests/queries/0_stateless/02366_window_function_order_by.sql
@@ -0,0 +1,9 @@
+-- { echoOn }
+SELECT groupArray(tuple(value)) OVER ()
+FROM (select number value from numbers(10))
+ORDER BY value ASC;
+
+SELECT count() OVER (ORDER BY number + 1) FROM numbers(10) ORDER BY number;
+
+SELECT count() OVER (ORDER BY number + 1) + 1 as foo FROM numbers(10)
+ORDER BY foo;
