diff --git a/src/parser/transform/statement/transform_create_view.cpp b/src/parser/transform/statement/transform_create_view.cpp
index 8504ecfc9a91..24a26c022d0d 100644
--- a/src/parser/transform/statement/transform_create_view.cpp
+++ b/src/parser/transform/statement/transform_create_view.cpp
@@ -21,7 +21,7 @@ unique_ptr<CreateStatement> Transformer::TransformCreateView(duckdb_libpgquery::
 	}
 	info->on_conflict = TransformOnConflict(stmt.onconflict);
 
-	info->query = TransformSelectStmt(*PGPointerCast<duckdb_libpgquery::PGSelectStmt>(stmt.query), false);
+	info->query = TransformSelectStmt(*stmt.query, false);
 
 	PivotEntryCheck("view");
 
