diff --git a/tests/queries/0_stateless/03205_hashing_empty_tuples.reference b/tests/queries/0_stateless/03205_hashing_empty_tuples.reference
new file mode 100644
index 000000000000..e24b5809aeed
--- /dev/null
+++ b/tests/queries/0_stateless/03205_hashing_empty_tuples.reference
@@ -0,0 +1,22 @@
+16324913028386710556
+16324913028386710556
+5049034479224883533
+7385293435322750976
+12248912094175844631
+5049034479224883533
+5887129541803688833
+5887129541803688833
+13747979201178469747
+5887129541803688833
+15520217392480966957
+16324913028386710556
+16324913028386710556
+5049034479224883533
+7385293435322750976
+12248912094175844631
+5049034479224883533
+5887129541803688833
+5887129541803688833
+13747979201178469747
+5887129541803688833
+15520217392480966957
diff --git a/tests/queries/0_stateless/03205_hashing_empty_tuples.sql b/tests/queries/0_stateless/03205_hashing_empty_tuples.sql
new file mode 100644
index 000000000000..4a97f30ced31
--- /dev/null
+++ b/tests/queries/0_stateless/03205_hashing_empty_tuples.sql
@@ -0,0 +1,23 @@
+SELECT sipHash64(());
+SELECT sipHash64((), ());
+SELECT sipHash64((), 1);
+SELECT sipHash64(1, ());
+SELECT sipHash64(1, (), 1);
+SELECT sipHash64((), 1, ());
+SELECT sipHash64((), (1, 2));
+SELECT sipHash64((), (1, 2));
+SELECT sipHash64((1, 2), ());
+SELECT sipHash64((), (1, 2), ());
+SELECT sipHash64((1, 2), (), (3, 4));
+
+SELECT sipHash64(materialize(()));
+SELECT sipHash64(materialize(()), materialize(()));
+SELECT sipHash64(materialize(()), 1);
+SELECT sipHash64(1, materialize(()));
+SELECT sipHash64(1, materialize(()), 1);
+SELECT sipHash64((), 1, materialize(()));
+SELECT sipHash64(materialize(()), (1, 2));
+SELECT sipHash64(materialize(()), (1, 2));
+SELECT sipHash64((1, 2), materialize(()));
+SELECT sipHash64(materialize(()), (1, 2), ());
+SELECT sipHash64((1, 2), materialize(()), (3, 4));
