diff --git a/src/Interpreters/ReplaceQueryParameterVisitor.cpp b/src/Interpreters/ReplaceQueryParameterVisitor.cpp
index 5c29c722f88e..2a629cefb01f 100644
--- a/src/Interpreters/ReplaceQueryParameterVisitor.cpp
+++ b/src/Interpreters/ReplaceQueryParameterVisitor.cpp
@@ -25,14 +25,18 @@ namespace ErrorCodes
 
 
 void ReplaceQueryParameterVisitor::visit(ASTPtr & ast)
+{
+    if (ast->as<ASTQueryParameter>())
+        visitQueryParameter(ast);
+    else
+        visitChildren(ast);
+}
+
+
+void ReplaceQueryParameterVisitor::visitChildren(ASTPtr & ast)
 {
     for (auto & child : ast->children)
-    {
-        if (child->as<ASTQueryParameter>())
-            visitQueryParameter(child);
-        else
-            visit(child);
-    }
+        visit(child);
 }
 
 const String & ReplaceQueryParameterVisitor::getParamValue(const String & name)
diff --git a/src/Interpreters/ReplaceQueryParameterVisitor.h b/src/Interpreters/ReplaceQueryParameterVisitor.h
index 1931d4c0ba8c..3a84cd22acd9 100644
--- a/src/Interpreters/ReplaceQueryParameterVisitor.h
+++ b/src/Interpreters/ReplaceQueryParameterVisitor.h
@@ -22,6 +22,7 @@ class ReplaceQueryParameterVisitor
     const NameToNameMap & query_parameters;
     const String & getParamValue(const String & name);
     void visitQueryParameter(ASTPtr & ast);
+    void visitChildren(ASTPtr & ast);
 };
 
 }
