diff --git a/programs/CMakeLists.txt b/programs/CMakeLists.txt
index 657426973331..baf8270d1bfb 100644
--- a/programs/CMakeLists.txt
+++ b/programs/CMakeLists.txt
@@ -207,7 +207,7 @@ if (TARGET clickhouse-server AND TARGET copy-headers)
 endif ()
 
 if (ENABLE_TESTS AND USE_GTEST)
-    set (CLICKHOUSE_ALL_TESTS_TARGETS local_date_time_comparison unit_tests_libcommon unit_tests_dbms hashing_write_buffer hashing_read_buffer in_join_subqueries_preprocessor expression_analyzer)
+    set (CLICKHOUSE_ALL_TESTS_TARGETS local_date_time_comparison unit_tests_libcommon unit_tests_dbms hashing_write_buffer hashing_read_buffer in_join_subqueries_preprocessor)
     add_custom_target (clickhouse-tests ALL DEPENDS ${CLICKHOUSE_ALL_TESTS_TARGETS})
     add_dependencies(clickhouse-bundle clickhouse-tests)
 endif()
