diff --git a/tests/queries/0_stateless/02519_monotonicity_fuzz.reference b/tests/queries/0_stateless/02519_monotonicity_fuzz.reference
new file mode 100644
index 000000000000..9459d4ba2a0d
--- /dev/null
+++ b/tests/queries/0_stateless/02519_monotonicity_fuzz.reference
@@ -0,0 +1,1 @@
+1.1
diff --git a/tests/queries/0_stateless/02519_monotonicity_fuzz.sql b/tests/queries/0_stateless/02519_monotonicity_fuzz.sql
new file mode 100644
index 000000000000..4a0860702bb7
--- /dev/null
+++ b/tests/queries/0_stateless/02519_monotonicity_fuzz.sql
@@ -0,0 +1,5 @@
+DROP TABLE IF EXISTS t;
+CREATE TABLE t (x Decimal(18, 3)) ENGINE = MergeTree ORDER BY x;
+INSERT INTO t VALUES (1.1);
+SELECT * FROM t WHERE toUInt64(x) = 1;
+DROP TABLE t;
