diff --git a/tests/queries/0_stateless/01836_date_time_keep_default_timezone_on_operations_den_crane.reference b/tests/queries/0_stateless/01836_date_time_keep_default_timezone_on_operations_den_crane.reference
new file mode 100644
index 000000000000..fc624e3510f7
--- /dev/null
+++ b/tests/queries/0_stateless/01836_date_time_keep_default_timezone_on_operations_den_crane.reference
@@ -0,0 +1,6 @@
+DateTime
+DateTime
+DateTime(\'UTC\')
+DateTime64(3)
+DateTime64(3)
+DateTime64(3, \'UTC\')
diff --git a/tests/queries/0_stateless/01836_date_time_keep_default_timezone_on_operations_den_crane.sql b/tests/queries/0_stateless/01836_date_time_keep_default_timezone_on_operations_den_crane.sql
new file mode 100644
index 000000000000..be47cfb0411e
--- /dev/null
+++ b/tests/queries/0_stateless/01836_date_time_keep_default_timezone_on_operations_den_crane.sql
@@ -0,0 +1,26 @@
+SELECT toTypeName(now());
+SELECT toTypeName(now() - 1);
+SELECT toTypeName(now('UTC') - 1);
+
+SELECT toTypeName(now64(3));
+SELECT toTypeName(now64(3) - 1);
+SELECT toTypeName(toTimeZone(now64(3), 'UTC') - 1);
+
+DROP TABLE IF EXISTS tt_null;
+DROP TABLE IF EXISTS tt;
+DROP TABLE IF EXISTS tt_mv;
+
+create table tt_null(p String) engine = Null;
+
+create table tt(p String,tmin AggregateFunction(min, DateTime)) 
+engine = AggregatingMergeTree  order by p;
+
+create materialized view tt_mv to tt as 
+select p, minState(now() - interval 30 minute) as tmin
+from tt_null group by p;
+
+insert into tt_null values('x');
+
+DROP TABLE tt_null;
+DROP TABLE tt;
+DROP TABLE tt_mv;
