diff --git a/tests/queries/0_stateless/02241_short_circuit_short_column.reference b/tests/queries/0_stateless/02241_short_circuit_short_column.reference
new file mode 100644
index 000000000000..c25c8fb59d0e
--- /dev/null
+++ b/tests/queries/0_stateless/02241_short_circuit_short_column.reference
@@ -0,0 +1,20 @@
+1	\N
+1	\N
+1	\N
+1	\N
+1	\N
+1	\N
+1	\N
+1	\N
+1	\N
+1	\N
+0
+0
+0
+0
+0
+0
+0
+0
+0
+0
diff --git a/tests/queries/0_stateless/02241_short_circuit_short_column.sql b/tests/queries/0_stateless/02241_short_circuit_short_column.sql
new file mode 100644
index 000000000000..311307fe5057
--- /dev/null
+++ b/tests/queries/0_stateless/02241_short_circuit_short_column.sql
@@ -0,0 +1,2 @@
+SELECT 65536 AND 2147483647, throwIf((((1048575 AND throwIf((0. AND NULL) AND (((65536 AND 257) AND (1.1754943508222875e-38 AND 1024) AND -2147483649) AND NULL) AND (10 AND NULL)) AND (((65536 AND 257) AND (1.1754943508222875e-38 AND 1024) AND -1) AND NULL) AND 65535) AND -1) AND NULL) AND (NULL AND NULL), NULL < number) FROM numbers(10);
+SELECT NULL AND throwIf((0 AND NULL) AND 2147483646 AND NULL AND NULL) AND -2147483649 AND (NULL AND NULL) AND NULL FROM system.numbers LIMIT 10;
