diff --git a/tests/queries/0_stateless/02471_wrong_date_monotonicity.reference b/tests/queries/0_stateless/02471_wrong_date_monotonicity.reference
new file mode 100644
index 000000000000..d00491fd7e5b
--- /dev/null
+++ b/tests/queries/0_stateless/02471_wrong_date_monotonicity.reference
@@ -0,0 +1,1 @@
+1
diff --git a/tests/queries/0_stateless/02471_wrong_date_monotonicity.sql b/tests/queries/0_stateless/02471_wrong_date_monotonicity.sql
new file mode 100644
index 000000000000..40d64e53309a
--- /dev/null
+++ b/tests/queries/0_stateless/02471_wrong_date_monotonicity.sql
@@ -0,0 +1,5 @@
+DROP TABLE IF EXISTS tdm__fuzz_23;
+CREATE TABLE tdm__fuzz_23 (`x` UInt256) ENGINE = MergeTree ORDER BY x SETTINGS write_final_mark = 0;
+INSERT INTO tdm__fuzz_23 FORMAT Values (1);
+SELECT count(x) FROM tdm__fuzz_23 WHERE toDate(x) < toDate(now(), 'Asia/Istanbul') SETTINGS max_rows_to_read = 1;
+DROP TABLE tdm__fuzz_23;
