diff --git a/tests/queries/0_stateless/02517_union_columns_order.reference b/tests/queries/0_stateless/02517_union_columns_order.reference
new file mode 100644
index 000000000000..e69de29bb2d1
diff --git a/tests/queries/0_stateless/02517_union_columns_order.sql b/tests/queries/0_stateless/02517_union_columns_order.sql
new file mode 100644
index 000000000000..c02dacfa0dfb
--- /dev/null
+++ b/tests/queries/0_stateless/02517_union_columns_order.sql
@@ -0,0 +1,32 @@
+CREATE TABLE t1 (c0 Int32, PRIMARY KEY (c0)) ENGINE = MergeTree;
+SELECT DISTINCT *
+FROM
+(
+    SELECT DISTINCT
+        cos(sign(exp(t1.c0))),
+        -min2(pow(t1.c0, t1.c0), intDiv(t1.c0, t1.c0)),
+        t1.c0,
+        t1.c0,
+        erf(abs(-t1.c0))
+    FROM t1
+    WHERE t1.c0 > 0
+    UNION ALL
+    SELECT DISTINCT
+        cos(sign(exp(t1.c0))),
+        -min2(pow(t1.c0, t1.c0), intDiv(t1.c0, t1.c0)),
+        t1.c0,
+        t1.c0,
+        erf(abs(-t1.c0))
+    FROM t1
+    WHERE NOT (t1.c0 > 0)
+    UNION ALL
+    SELECT DISTINCT
+        cos(sign(exp(t1.c0))),
+        -min2(pow(t1.c0, t1.c0), intDiv(t1.c0, t1.c0)),
+        t1.c0,
+        t1.c0,
+        erf(abs(-t1.c0))
+    FROM t1
+    WHERE t1.c0 > (0 IS NULL)
+);
+
