diff --git a/tests/queries/0_stateless/03360_bool_remote.reference b/tests/queries/0_stateless/03360_bool_remote.reference
new file mode 100644
index 000000000000..b979d62f4fa8
--- /dev/null
+++ b/tests/queries/0_stateless/03360_bool_remote.reference
@@ -0,0 +1,3 @@
+true
+true
+true
diff --git a/tests/queries/0_stateless/03360_bool_remote.sql b/tests/queries/0_stateless/03360_bool_remote.sql
new file mode 100644
index 000000000000..69c9b6da2ce6
--- /dev/null
+++ b/tests/queries/0_stateless/03360_bool_remote.sql
@@ -0,0 +1,3 @@
+SELECT true AS x FROM remote('127.0.0.{1,2}', system.one) LIMIT 1;
+SELECT materialize(true) AS x FROM remote('127.0.0.{1,2}', system.one) LIMIT 1;
+SELECT true AS x FROM remote('127.0.0.{1,2}', system.one) GROUP BY x;
