diff --git a/test/sql/attach/attach_view_search_path.test b/test/sql/attach/attach_view_search_path.test
new file mode 100644
index 000000000000..01a1f6eeddd8
--- /dev/null
+++ b/test/sql/attach/attach_view_search_path.test
@@ -0,0 +1,78 @@
+# name: test/sql/attach/attach_view_search_path.test
+# description: Test ATTACH with search path
+# group: [attach]
+
+# avoid loading a storage database because it changes the initial database name
+require noforcestorage
+
+statement ok
+PRAGMA enable_verification
+
+statement ok
+ATTACH DATABASE '__TEST_DIR__/view_search_path.db' AS view_search_path;
+
+statement ok
+USE view_search_path
+
+statement ok
+CREATE TABLE my_tbl(i INTEGER)
+
+statement ok
+INSERT INTO my_tbl VALUES (42)
+
+statement ok
+CREATE VIEW my_view AS FROM my_tbl
+
+query I
+FROM my_view
+----
+42
+
+statement ok
+CREATE SCHEMA my_schema
+
+statement ok
+USE my_schema
+
+statement ok
+CREATE TABLE my_tbl(i INTEGER)
+
+statement ok
+INSERT INTO my_tbl VALUES (84)
+
+statement ok
+CREATE VIEW my_view AS FROM my_tbl
+
+query I
+FROM my_view
+----
+84
+
+statement ok
+USE memory
+
+query I
+FROM view_search_path.my_view
+----
+42
+
+query I
+FROM view_search_path.my_schema.my_view
+----
+84
+
+statement ok
+DETACH view_search_path
+
+statement ok
+ATTACH DATABASE '__TEST_DIR__/view_search_path.db' AS view_search_path;
+
+query I
+FROM view_search_path.my_view
+----
+42
+
+query I
+FROM view_search_path.my_schema.my_view
+----
+84
diff --git a/test/sql/attach/attach_views.test b/test/sql/attach/attach_views.test
index 6311e9e702fb..c64c0fbf6505 100644
--- a/test/sql/attach/attach_views.test
+++ b/test/sql/attach/attach_views.test
@@ -2,7 +2,7 @@
 # description: Test views in an attached database
 # group: [attach]
 
-require skip_reload
+require noforcestorage
 
 statement ok
 PRAGMA enable_verification
@@ -46,7 +46,7 @@ SELECT * FROM ${prefix}.v1
 
 # reference tables from different databases in view
 statement ok
-CREATE OR REPLACE VIEW ${prefix}.v1 AS SELECT * FROM ${prefix}.t1 UNION ALL FROM t1 ORDER BY ALL
+CREATE OR REPLACE VIEW ${prefix}.v1 AS SELECT * FROM ${prefix}.t1 UNION ALL FROM memory.t1 ORDER BY ALL
 
 query I
 SELECT * FROM ${prefix}.v1
