diff --git a/test/sql/prepared/prepared_select.test b/test/sql/prepared/prepared_select.test
new file mode 100644
index 000000000000..4ef01ca014c5
--- /dev/null
+++ b/test/sql/prepared/prepared_select.test
@@ -0,0 +1,62 @@
+# name: test/sql/prepared/prepared_select.test
+# description: Test order of unmarked prepared statements in select
+# group: [prepared]
+
+# select - from - where with subquery
+statement ok
+PREPARE s1 AS SELECT ?::VARCHAR FROM (SELECT ?::INTEGER) tbl(i) WHERE i > ?::INTEGER
+
+query I
+EXECUTE s1('hello', 2, 1)
+----
+hello
+
+# select - from - where - groupby - having
+statement ok
+PREPARE s2 AS
+SELECT FIRST(?::VARCHAR)
+FROM (VALUES (?::INTEGER)) tbl(i)
+WHERE i > ?::INTEGER
+GROUP BY i % ?::INTEGER
+HAVING SUM(i)::VARCHAR <> ?::VARCHAR
+
+query I
+EXECUTE s2('hello', 2, 1, 2, 'blabla')
+----
+hello
+
+query I
+EXECUTE s2('hello', 2, 1, 2, '2')
+----
+
+# union
+statement ok
+PREPARE s3 AS
+SELECT LENGTH(?::VARCHAR)
+UNION ALL
+SELECT ?::INTEGER
+ORDER BY 1
+
+query I
+EXECUTE s3('hello', 3)
+----
+3
+5
+
+# IN clause
+statement ok
+PREPARE s4 AS SELECT ?::INTEGER IN (?::INTEGER, ?::INTEGER, ?::INTEGER)
+
+query I
+EXECUTE s4(1, 2, 3, 1)
+----
+1
+
+# IN clause with subquery
+statement ok
+PREPARE s5 AS SELECT ?::INTEGER IN (SELECT i FROM (VALUES (?::INTEGER), (?::INTEGER), (?::INTEGER)) tbl(i))
+
+query I
+EXECUTE s5(1, 2, 3, 1)
+----
+1
diff --git a/test/sql/prepared/prepared_update.test b/test/sql/prepared/prepared_update.test
new file mode 100644
index 000000000000..038e1bf153d2
--- /dev/null
+++ b/test/sql/prepared/prepared_update.test
@@ -0,0 +1,42 @@
+# name: test/sql/prepared/prepared_update.test
+# description: Test order of unmarked prepared statements
+# group: [prepared]
+
+statement ok
+CREATE TABLE integers(i INTEGER, j VARCHAR)
+
+statement ok
+INSERT INTO integers VALUES (1, 'hello')
+
+statement ok
+PREPARE s1 AS UPDATE integers SET i=?, j=?;
+
+statement ok
+EXECUTE s1(2, 'world')
+
+query II
+SELECT * FROM integers
+----
+2	world
+
+statement ok
+PREPARE s2 AS UPDATE integers SET j=? WHERE i=?;
+
+statement ok
+EXECUTE s2('test', 2)
+
+query II
+SELECT * FROM integers
+----
+2	test
+
+statement ok
+PREPARE s3 AS UPDATE integers SET j=? WHERE i=? AND j=?;
+
+statement ok
+EXECUTE s3('test2', 2, 'test')
+
+query II
+SELECT * FROM integers
+----
+2	test2
