diff --git a/test/sql/types/struct/test_struct.test b/test/sql/types/struct/test_struct.test
index 4ac122e03a14..3b65e671d127 100644
--- a/test/sql/types/struct/test_struct.test
+++ b/test/sql/types/struct/test_struct.test
@@ -175,3 +175,9 @@ SELECT STRUCT_EXTRACT(STRUCT_PACK(xx := e, yy := g), '42') FROM struct_data
 
 statement error
 CREATE TABLE test AS SELECT e, STRUCT_PACK(e) FROM struct_data
+
+statement ok
+SELECT STRUCT_PACK(i := range, j := 'asdf') FROM range(2);
+
+statement ok
+SELECT STRUCT_PACK(i := range, j := 42) FROM range(2);
