diff --git a/tests/queries/0_stateless/03314_has_column_in_table_alias_column.reference b/tests/queries/0_stateless/03314_has_column_in_table_alias_column.reference
new file mode 100644
index 000000000000..d00491fd7e5b
--- /dev/null
+++ b/tests/queries/0_stateless/03314_has_column_in_table_alias_column.reference
@@ -0,0 +1,1 @@
+1
diff --git a/tests/queries/0_stateless/03314_has_column_in_table_alias_column.sql b/tests/queries/0_stateless/03314_has_column_in_table_alias_column.sql
new file mode 100644
index 000000000000..f85f0038cf6d
--- /dev/null
+++ b/tests/queries/0_stateless/03314_has_column_in_table_alias_column.sql
@@ -0,0 +1,4 @@
+DROP TABLE IF EXISTS test;
+CREATE TABLE test(`a` String, `alias_col_a` String ALIAS a) ENGINE = MergeTree ORDER BY a;
+SELECT hasColumnInTable(currentDatabase(), 'test', 'alias_col_a');
+DROP TABLE test;
