diff --git a/tests/integration/test_backup_restore_on_cluster/configs/disallow_concurrency.xml b/tests/integration/test_backup_restore_on_cluster/configs/disallow_concurrency.xml
index 144be77c9f92..9e67f54f8e85 100644
--- a/tests/integration/test_backup_restore_on_cluster/configs/disallow_concurrency.xml
+++ b/tests/integration/test_backup_restore_on_cluster/configs/disallow_concurrency.xml
@@ -9,7 +9,7 @@
     </storage_configuration>
     <backups>
         <allowed_disk>backups</allowed_disk>
+        <allow_concurrent_backups>false</allow_concurrent_backups>
+        <allow_concurrent_restores>false</allow_concurrent_restores>
     </backups>
-    <allow_concurrent_backups>false</allow_concurrent_backups>
-    <allow_concurrent_restores>false</allow_concurrent_restores>
 </clickhouse>
diff --git a/tests/integration/test_backup_restore_on_cluster/test_disallow_concurrency.py b/tests/integration/test_backup_restore_on_cluster/test_disallow_concurrency.py
index 8f514b95d0b0..43e7682ec1d0 100644
--- a/tests/integration/test_backup_restore_on_cluster/test_disallow_concurrency.py
+++ b/tests/integration/test_backup_restore_on_cluster/test_disallow_concurrency.py
@@ -148,10 +148,14 @@ def test_concurrent_backups_on_different_nodes():
 
     backup_name = new_backup_name()
 
-    nodes[1].query(f"BACKUP TABLE tbl ON CLUSTER 'cluster' TO {backup_name} ASYNC")
+    id = (
+        nodes[1]
+        .query(f"BACKUP TABLE tbl ON CLUSTER 'cluster' TO {backup_name} ASYNC")
+        .split("\t")[0]
+    )
     assert_eq_with_retry(
         nodes[1],
-        f"SELECT status FROM system.backups WHERE status == 'CREATING_BACKUP'",
+        f"SELECT status FROM system.backups WHERE status == 'CREATING_BACKUP' AND id = '{id}'",
         "CREATING_BACKUP",
     )
     assert "Concurrent backups not supported" in nodes[2].query_and_get_error(
