diff --git a/tests/queries/0_stateless/02002_system_table_with_tuple.reference b/tests/queries/0_stateless/02002_system_table_with_tuple.reference
new file mode 100644
index 000000000000..d00491fd7e5b
--- /dev/null
+++ b/tests/queries/0_stateless/02002_system_table_with_tuple.reference
@@ -0,0 +1,1 @@
+1
diff --git a/tests/queries/0_stateless/02002_system_table_with_tuple.sh b/tests/queries/0_stateless/02002_system_table_with_tuple.sh
new file mode 100755
index 000000000000..2abf5b8b5253
--- /dev/null
+++ b/tests/queries/0_stateless/02002_system_table_with_tuple.sh
@@ -0,0 +1,9 @@
+#!/usr/bin/env bash
+
+CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+# shellcheck source=../shell_config.sh
+. "$CURDIR"/../shell_config.sh
+
+$CLICKHOUSE_CLIENT -q "SELECT count() > 0 FROM system.tables \
+    WHERE database IN ('system', '$CLICKHOUSE_DATABASE') \
+    SETTINGS legacy_column_name_of_tuple_literal = 1"
