diff --git a/tests/queries/0_stateless/01412_optimize_deduplicate_bug.reference b/tests/queries/0_stateless/01412_optimize_deduplicate_bug.reference
new file mode 100644
index 000000000000..9972842f9827
--- /dev/null
+++ b/tests/queries/0_stateless/01412_optimize_deduplicate_bug.reference
@@ -0,0 +1,1 @@
+1	1
diff --git a/tests/queries/0_stateless/01412_optimize_deduplicate_bug.sql b/tests/queries/0_stateless/01412_optimize_deduplicate_bug.sql
new file mode 100644
index 000000000000..b75b31243c4c
--- /dev/null
+++ b/tests/queries/0_stateless/01412_optimize_deduplicate_bug.sql
@@ -0,0 +1,10 @@
+drop table if exists tesd_dedupl;
+
+create table tesd_dedupl (x UInt32, y UInt32) engine = MergeTree order by x;
+insert into tesd_dedupl values (1, 1);
+insert into tesd_dedupl values (1, 1);
+
+OPTIMIZE TABLE tesd_dedupl DEDUPLICATE;
+select * from tesd_dedupl;
+
+drop table if exists tesd_dedupl;
