diff --git a/test/sql/copy/s3/http_proxy.test b/test/sql/copy/s3/http_proxy.test
index c8458762b1d5..aece0a9e16c6 100644
--- a/test/sql/copy/s3/http_proxy.test
+++ b/test/sql/copy/s3/http_proxy.test
@@ -38,6 +38,12 @@ value-1
 statement ok
 set http_proxy='blabla:1337'
 
+statement ok
+set http_proxy_username='xxx'
+
+statement ok
+set http_proxy_password='yyy'
+
 statement error
 FROM 's3://test-bucket/proxy-test/test.parquet'
 ----
@@ -47,6 +53,12 @@ Could not establish connection
 statement ok
 set http_proxy='${HTTP_PROXY_PUBLIC}'
 
+statement ok
+RESET http_proxy_username
+
+statement ok
+RESET http_proxy_password
+
 query I
 FROM 's3://test-bucket/proxy-test/test.parquet'
 ----
