diff --git a/tests/queries/0_stateless/02354_read_in_order_prewhere.reference b/tests/queries/0_stateless/02354_read_in_order_prewhere.reference
new file mode 100644
index 000000000000..7d5543bf9cca
--- /dev/null
+++ b/tests/queries/0_stateless/02354_read_in_order_prewhere.reference
@@ -0,0 +1,10 @@
+1
+1
+1
+2001
+2001
+1
+1
+1
+2001
+2001
diff --git a/tests/queries/0_stateless/02354_read_in_order_prewhere.sql b/tests/queries/0_stateless/02354_read_in_order_prewhere.sql
new file mode 100644
index 000000000000..c5abd5945f32
--- /dev/null
+++ b/tests/queries/0_stateless/02354_read_in_order_prewhere.sql
@@ -0,0 +1,30 @@
+drop table if exists order;
+
+CREATE TABLE order
+(
+    ID Int64,
+    Type Int64,
+    Num UInt64,
+    t DateTime
+)
+ENGINE = MergeTree()
+PARTITION BY toYYYYMMDD(t)
+ORDER BY (ID, Type, Num);
+
+system stop merges order;
+
+insert into order select number%2000, 1, number, (1656700561 - intDiv(intHash32(number), 1000)) from numbers(100000);
+insert into order select number%2000, 1, number, (1656700561 - intDiv(intHash32(number), 1000)) from numbers(100000);
+insert into order select number%2000, 1, number, (1656700561 - intDiv(intHash32(number), 1000)) from numbers(100000);
+
+SELECT Num
+FROM order
+WHERE Type = 1 AND ID = 1
+ORDER BY Num ASC limit 5;
+
+SELECT Num
+FROM order
+PREWHERE Type = 1
+WHERE ID = 1
+ORDER BY Num ASC limit 5;
+
