diff --git a/tests/queries/0_stateless/25340_grace_hash_limit_race.reference b/tests/queries/0_stateless/02677_grace_hash_limit_race.reference
similarity index 100%
rename from tests/queries/0_stateless/25340_grace_hash_limit_race.reference
rename to tests/queries/0_stateless/02677_grace_hash_limit_race.reference
diff --git a/tests/queries/0_stateless/25340_grace_hash_limit_race.sql b/tests/queries/0_stateless/02677_grace_hash_limit_race.sql
similarity index 100%
rename from tests/queries/0_stateless/25340_grace_hash_limit_race.sql
rename to tests/queries/0_stateless/02677_grace_hash_limit_race.sql
diff --git a/tests/queries/0_stateless/02678_explain_pipeline_graph.reference b/tests/queries/0_stateless/02678_explain_pipeline_graph_with_projection.reference
similarity index 100%
rename from tests/queries/0_stateless/02678_explain_pipeline_graph.reference
rename to tests/queries/0_stateless/02678_explain_pipeline_graph_with_projection.reference
diff --git a/tests/queries/0_stateless/02678_explain_pipeline_graph.sql b/tests/queries/0_stateless/02678_explain_pipeline_graph_with_projection.sql
similarity index 50%
rename from tests/queries/0_stateless/02678_explain_pipeline_graph.sql
rename to tests/queries/0_stateless/02678_explain_pipeline_graph_with_projection.sql
index 48cfbf2b3493..e8b7405d6024 100644
--- a/tests/queries/0_stateless/02678_explain_pipeline_graph.sql
+++ b/tests/queries/0_stateless/02678_explain_pipeline_graph_with_projection.sql
@@ -1,7 +1,12 @@
--- The server does not crash after these queries:
-
 DROP TABLE IF EXISTS t1;
 CREATE TABLE t1(ID UInt64, name String) engine=MergeTree order by ID;
+
 insert into t1(ID, name) values (1, 'abc'), (2, 'bbb');
+
+-- The returned node order is uncertain
 explain pipeline graph=1 select count(ID) from t1 FORMAT Null;
+explain pipeline graph=1 select sum(1) from t1 FORMAT Null;
+explain pipeline graph=1 select min(ID) from t1 FORMAT Null;
+explain pipeline graph=1 select max(ID) from t1 FORMAT Null;
+
 DROP TABLE t1;
