diff --git a/src/AggregateFunctions/AggregateFunctionGroupBitmapData.h b/src/AggregateFunctions/AggregateFunctionGroupBitmapData.h
index 39ab846e61ef..ec945d418f29 100644
--- a/src/AggregateFunctions/AggregateFunctionGroupBitmapData.h
+++ b/src/AggregateFunctions/AggregateFunctionGroupBitmapData.h
@@ -229,7 +229,7 @@ class RoaringBitmapWithSmallSet : private boost::noncopyable
         {
             for (const auto & x : small)
             {
-                if (!rb->contains(static_cast<Value>(x.getValue())))
+                if (!r1.rb->contains(static_cast<Value>(x.getValue())))
                     buffer.push_back(x.getValue());
             }
 
