diff --git a/tests/queries/0_stateless/03405_ssd_cache_incorrect_min_max_lifetimes_and_block_size.sql b/tests/queries/0_stateless/03405_ssd_cache_incorrect_min_max_lifetimes_and_block_size.sql
index 8a20d1146dc4..84010bf4bec1 100644
--- a/tests/queries/0_stateless/03405_ssd_cache_incorrect_min_max_lifetimes_and_block_size.sql
+++ b/tests/queries/0_stateless/03405_ssd_cache_incorrect_min_max_lifetimes_and_block_size.sql
@@ -1,18 +1,19 @@
 -- Tags: no-ordinary-database
 
--- Tests that various conditions are checked during creation of a ssd_cache layout of a dictionary
+-- Tests that various conditions are checked during creation of 'ssd_cache' and 'complex_key_ssd_cache' dictionaries
 
 -- Github issue #78314
 
-DROP DICTIONARY IF EXISTS dx;
+DROP DICTIONARY IF EXISTS dict;
 
 SELECT 'BLOCK_SIZE is a negative value.';
-CREATE DICTIONARY dx (col Int64 default null) PRIMARY KEY (col) SOURCE(NULL()) LAYOUT(SSD_CACHE(BLOCK_SIZE -1)) LIFETIME(1); -- { serverError BAD_ARGUMENTS }
+CREATE DICTIONARY dict (col Int64 default null) PRIMARY KEY (col) SOURCE(NULL()) LAYOUT(SSD_CACHE(BLOCK_SIZE -1)) LIFETIME(1); -- { serverError BAD_ARGUMENTS }
+CREATE DICTIONARY dict (col Int64 default null) PRIMARY KEY (col) SOURCE(NULL()) LAYOUT(COMPLEX_KEY_SSD_CACHE(BLOCK_SIZE -1)) LIFETIME(1); -- { serverError BAD_ARGUMENTS }
 
 SELECT 'BLOCK_SIZE is zero.';
-CREATE DICTIONARY dx (col Int64 default null) PRIMARY KEY (col) SOURCE(NULL()) LAYOUT(SSD_CACHE(BLOCK_SIZE 0)) LIFETIME(1); -- { serverError BAD_ARGUMENTS }
+CREATE DICTIONARY dict (col Int64 default null) PRIMARY KEY (col) SOURCE(NULL()) LAYOUT(SSD_CACHE(BLOCK_SIZE 0)) LIFETIME(1); -- { serverError BAD_ARGUMENTS }
+CREATE DICTIONARY dict (col Int64 default null) PRIMARY KEY (col) SOURCE(NULL()) LAYOUT(COMPLEX_KEY_SSD_CACHE(BLOCK_SIZE 0)) LIFETIME(1); -- { serverError BAD_ARGUMENTS }
 
 SELECT 'WRITE_BUFFER_SIZE is zero.';
-CREATE DICTIONARY dx (col Int64 default null) PRIMARY KEY (col) SOURCE(NULL()) LAYOUT(SSD_CACHE(WRITE_BUFFER_SIZE 0)) LIFETIME(1); -- { serverError BAD_ARGUMENTS }
-
-DROP DICTIONARY IF EXISTS dx;
+CREATE DICTIONARY dict (col Int64 default null) PRIMARY KEY (col) SOURCE(NULL()) LAYOUT(SSD_CACHE(WRITE_BUFFER_SIZE 0)) LIFETIME(1); -- { serverError BAD_ARGUMENTS }
+CREATE DICTIONARY dict (col Int64 default null) PRIMARY KEY (col) SOURCE(NULL()) LAYOUT(COMPLEX_KEY_SSD_CACHE(WRITE_BUFFER_SIZE 0)) LIFETIME(1); -- { serverError BAD_ARGUMENTS }
diff --git a/tests/queries/0_stateless/03406_dictionary_incorrect_min_max_lifetimes.sql b/tests/queries/0_stateless/03406_dictionary_incorrect_min_max_lifetimes.sql
index 4de8843907e3..a0aec5e78e6c 100644
--- a/tests/queries/0_stateless/03406_dictionary_incorrect_min_max_lifetimes.sql
+++ b/tests/queries/0_stateless/03406_dictionary_incorrect_min_max_lifetimes.sql
@@ -4,19 +4,18 @@
 
 -- Github issue #78314
 
-DROP DICTIONARY IF EXISTS dx;
+DROP DICTIONARY IF EXISTS dict;
 DROP TABLE IF EXISTS tbl;
 
 CREATE TABLE tbl (col Int) ENGINE = Memory;
 
 SELECT 'MIN is a negative value.';
-CREATE DICTIONARY dx (col Int DEFAULT 1) PRIMARY KEY (col) SOURCE(CLICKHOUSE(TABLE 'tbl')) LAYOUT(HASHED_ARRAY()) LIFETIME(MIN -1 MAX 0); -- { clientError SYNTAX_ERROR }
+CREATE DICTIONARY dict (col Int DEFAULT 1) PRIMARY KEY (col) SOURCE(CLICKHOUSE(TABLE 'tbl')) LAYOUT(HASHED_ARRAY()) LIFETIME(MIN -1 MAX 0); -- { clientError SYNTAX_ERROR }
 
 SELECT 'MAX is a negative value.';
-CREATE DICTIONARY dx (col Int DEFAULT 1) PRIMARY KEY (col) SOURCE(CLICKHOUSE(TABLE 'tbl')) LAYOUT(HASHED_ARRAY()) LIFETIME(MIN 0 MAX -1); -- { clientError SYNTAX_ERROR }
+CREATE DICTIONARY dict (col Int DEFAULT 1) PRIMARY KEY (col) SOURCE(CLICKHOUSE(TABLE 'tbl')) LAYOUT(HASHED_ARRAY()) LIFETIME(MIN 0 MAX -1); -- { clientError SYNTAX_ERROR }
 
 SELECT 'MIN is greater than MAX.';
-CREATE DICTIONARY dx (col Int DEFAULT 1) PRIMARY KEY (col) SOURCE(CLICKHOUSE(TABLE 'tbl')) LAYOUT(HASHED_ARRAY()) LIFETIME(MIN 1 MAX 0); -- { serverError BAD_ARGUMENTS }
+CREATE DICTIONARY dict (col Int DEFAULT 1) PRIMARY KEY (col) SOURCE(CLICKHOUSE(TABLE 'tbl')) LAYOUT(HASHED_ARRAY()) LIFETIME(MIN 1 MAX 0); -- { serverError BAD_ARGUMENTS }
 
-DROP DICTIONARY IF EXISTS dx;
-DROP TABLE IF EXISTS tbl;
+DROP TABLE tbl;
