diff --git a/tests/queries/0_stateless/00257_shard_no_aggregates_and_constant_keys.reference b/tests/queries/0_stateless/00257_shard_no_aggregates_and_constant_keys.reference
index 63b8a9d14fc6..fc77ed8a2412 100644
--- a/tests/queries/0_stateless/00257_shard_no_aggregates_and_constant_keys.reference
+++ b/tests/queries/0_stateless/00257_shard_no_aggregates_and_constant_keys.reference
@@ -8,13 +8,13 @@
 40
 41
 
-0
+41
 2	42
 
 2	42
 43
 
-0
+43
 11
 
 11
diff --git a/tests/queries/0_stateless/03215_analyzer_materialized_constants_bug.reference b/tests/queries/0_stateless/03215_analyzer_materialized_constants_bug.reference
new file mode 100644
index 000000000000..584e34c0cdea
--- /dev/null
+++ b/tests/queries/0_stateless/03215_analyzer_materialized_constants_bug.reference
@@ -0,0 +1,3 @@
+ba
+\N
+1	111111111111111111111111111111111111111
diff --git a/tests/queries/0_stateless/03215_analyzer_materialized_constants_bug.sql b/tests/queries/0_stateless/03215_analyzer_materialized_constants_bug.sql
new file mode 100644
index 000000000000..b2fd69d75d0b
--- /dev/null
+++ b/tests/queries/0_stateless/03215_analyzer_materialized_constants_bug.sql
@@ -0,0 +1,26 @@
+SET allow_experimental_analyzer = 1;
+
+SELECT concat(materialize(toLowCardinality('b')), 'a') FROM remote('127.0.0.{1,2}', system, one) GROUP BY 'a';
+
+SELECT concat(NULLIF(1, materialize(toLowCardinality(1))), concat(NULLIF(1, 1))) FROM remote('127.0.0.{1,2}', system, one) GROUP BY concat(NULLIF(1, 1));
+
+DROP TABLE IF EXISTS test__fuzz_21;
+CREATE TABLE test__fuzz_21
+(
+    `x` Decimal(18, 10)
+)
+ENGINE = MergeTree
+ORDER BY x;
+
+INSERT INTO test__fuzz_21 VALUES (1), (2), (3);
+
+WITH (
+        SELECT CAST(toFixedString(toFixedString(materialize(toFixedString('111111111111111111111111111111111111111', 39)), 39), 39), 'UInt128')
+    ) AS v
+SELECT
+    coalesce(materialize(toLowCardinality(toNullable(1))), 10, NULL),
+    max(v)
+FROM remote('127.0.0.{1,2}', currentDatabase(), test__fuzz_21)
+GROUP BY
+    coalesce(NULL),
+    coalesce(1, 10, 10, materialize(NULL));
