diff --git a/src/parser/parsed_expression_iterator.cpp b/src/parser/parsed_expression_iterator.cpp
index 6d807800c206..8b29ed7fe560 100644
--- a/src/parser/parsed_expression_iterator.cpp
+++ b/src/parser/parsed_expression_iterator.cpp
@@ -102,6 +102,9 @@ void ParsedExpressionIterator::EnumerateChildren(
 		if (star_expr.expr) {
 			callback(star_expr.expr);
 		}
+		for (auto &item : star_expr.replace_list) {
+			callback(item.second);
+		}
 		break;
 	}
 	case ExpressionClass::SUBQUERY: {
