diff --git a/src/function/cast/string_cast.cpp b/src/function/cast/string_cast.cpp
index 9f8b5ee29811..f3a19c4273bf 100644
--- a/src/function/cast/string_cast.cpp
+++ b/src/function/cast/string_cast.cpp
@@ -342,8 +342,8 @@ bool VectorStringToMap::StringToNestedTypeCastLoop(const string_t *source_data,
 		vector_cast_data.all_converted = false;
 	}
 
-	auto &key_validity = FlatVector::Validity(result_key_child);
 	if (!vector_cast_data.all_converted) {
+		auto &key_validity = FlatVector::Validity(result_key_child);
 		for (idx_t row_idx = 0; row_idx < count; row_idx++) {
 			if (!result_mask.RowIsValid(row_idx)) {
 				continue;
