diff --git a/tests/queries/0_stateless/01355_if_fixed_string.reference b/tests/queries/0_stateless/01355_if_fixed_string.reference
new file mode 100644
index 000000000000..43c8af518b4c
--- /dev/null
+++ b/tests/queries/0_stateless/01355_if_fixed_string.reference
@@ -0,0 +1,40 @@
+0\0\0\0\0	String
+1\0	String
+-2\0\0\0	String
+3\0	String
+-4\0\0\0	String
+5\0	String
+-6\0\0\0	String
+7\0	String
+-8\0\0\0	String
+9\0	String
+0\0	FixedString(2)
+1\0	FixedString(2)
+-2	FixedString(2)
+3\0	FixedString(2)
+-4	FixedString(2)
+5\0	FixedString(2)
+-6	FixedString(2)
+7\0	FixedString(2)
+-8	FixedString(2)
+9\0	FixedString(2)
+0	String
+1	String
+-2	String
+3	String
+-4	String
+5	String
+-6	String
+7	String
+-8	String
+9	String
+0\0	FixedString(2)
+1\0	FixedString(2)
+-2	FixedString(2)
+3\0	FixedString(2)
+-4	FixedString(2)
+5\0	FixedString(2)
+-6	FixedString(2)
+7\0	FixedString(2)
+-8	FixedString(2)
+9\0	FixedString(2)
diff --git a/tests/queries/0_stateless/01355_if_fixed_string.sql b/tests/queries/0_stateless/01355_if_fixed_string.sql
new file mode 100644
index 000000000000..a0afcc5f1972
--- /dev/null
+++ b/tests/queries/0_stateless/01355_if_fixed_string.sql
@@ -0,0 +1,5 @@
+SELECT if(number % 2, toFixedString(toString(number), 2), toFixedString(toString(-number), 5)) AS x, toTypeName(x) FROM system.numbers LIMIT 10;
+SELECT if(number % 2, toFixedString(toString(number), 2), toFixedString(toString(-number), 2)) AS x, toTypeName(x) FROM system.numbers LIMIT 10;
+
+SELECT multiIf(number % 2, toFixedString(toString(number), 2), toFixedString(toString(-number), 5)) AS x, toTypeName(x) FROM system.numbers LIMIT 10;
+SELECT multiIf(number % 2, toFixedString(toString(number), 2), toFixedString(toString(-number), 2)) AS x, toTypeName(x) FROM system.numbers LIMIT 10;
