diff --git a/tests/queries/0_stateless/02458_key_condition_not_like_prefix.reference b/tests/queries/0_stateless/02458_key_condition_not_like_prefix.reference
new file mode 100644
index 000000000000..02357561e024
--- /dev/null
+++ b/tests/queries/0_stateless/02458_key_condition_not_like_prefix.reference
@@ -0,0 +1,2 @@
+200000
+200000
diff --git a/tests/queries/0_stateless/02458_key_condition_not_like_prefix.sql b/tests/queries/0_stateless/02458_key_condition_not_like_prefix.sql
new file mode 100644
index 000000000000..e821b16ed5ca
--- /dev/null
+++ b/tests/queries/0_stateless/02458_key_condition_not_like_prefix.sql
@@ -0,0 +1,12 @@
+CREATE TABLE data (str String) ENGINE=MergeTree ORDER BY str;
+INSERT INTO data (str) SELECT 'aa' FROM numbers(100000);
+INSERT INTO data (str) SELECT 'ba' FROM numbers(100000);
+INSERT INTO data (str) SELECT 'ca' FROM numbers(100000);
+SELECT count() FROM data WHERE str NOT LIKE 'a%' SETTINGS force_primary_key=1;
+SELECT count() FROM data WHERE str NOT LIKE 'a%%' SETTINGS force_primary_key=1;
+SELECT count() FROM data WHERE str NOT LIKE 'a' SETTINGS force_primary_key=1; -- { serverError 277 }
+SELECT count() FROM data WHERE str NOT LIKE '%a' SETTINGS force_primary_key=1; -- { serverError 277 }
+SELECT count() FROM data WHERE str NOT LIKE 'a_' SETTINGS force_primary_key=1; -- { serverError 277 }
+SELECT count() FROM data WHERE str NOT LIKE 'a%_' SETTINGS force_primary_key=1; -- { serverError 277 }
+SELECT count() FROM data WHERE str NOT LIKE '_a' SETTINGS force_primary_key=1; -- { serverError 277 }
+SELECT count() FROM data WHERE str NOT LIKE 'a%\_' SETTINGS force_primary_key=1; -- { serverError 277 }
