diff --git a/tests/queries/0_stateless/01764_table_function_dictionary.reference b/tests/queries/0_stateless/01764_table_function_dictionary.reference
new file mode 100644
index 000000000000..b8e844ab3e9d
--- /dev/null
+++ b/tests/queries/0_stateless/01764_table_function_dictionary.reference
@@ -0,0 +1,2 @@
+0	0
+1	1
diff --git a/tests/queries/0_stateless/01764_table_function_dictionary.sql b/tests/queries/0_stateless/01764_table_function_dictionary.sql
new file mode 100644
index 000000000000..0168566077d2
--- /dev/null
+++ b/tests/queries/0_stateless/01764_table_function_dictionary.sql
@@ -0,0 +1,25 @@
+DROP TABLE IF EXISTS table_function_dictionary_source_table;
+CREATE TABLE table_function_dictionary_source_table
+(
+   id UInt64,
+   value UInt64
+)
+ENGINE = TinyLog;
+
+INSERT INTO table_function_dictionary_source_table VALUES (0, 0);
+INSERT INTO table_function_dictionary_source_table VALUES (1, 1);
+
+DROP DICTIONARY IF EXISTS table_function_dictionary_test_dictionary;
+CREATE DICTIONARY table_function_dictionary_test_dictionary
+(
+   id UInt64,
+   value UInt64 DEFAULT 0
+)
+PRIMARY KEY id
+SOURCE(CLICKHOUSE(HOST 'localhost' PORT tcpPort() USER 'default' TABLE 'table_function_dictionary_source_table'))
+LAYOUT(DIRECT());
+
+SELECT * FROM dictionary('table_function_dictionary_test_dictionary');
+
+DROP TABLE table_function_dictionary_source_table;
+DROP DICTIONARY table_function_dictionary_test_dictionary;
