diff --git a/ydb/core/tablet_flat/test/libs/table/wrap_part.h b/ydb/core/tablet_flat/test/libs/table/wrap_part.h
index 5bc92e9d68d8..6d90e7b820f4 100644
--- a/ydb/core/tablet_flat/test/libs/table/wrap_part.h
+++ b/ydb/core/tablet_flat/test/libs/table/wrap_part.h
@@ -104,7 +104,8 @@ namespace NTest {
         EReady SkipToRowVersion(TRowVersion rowVersion) noexcept
         {
             TIteratorStats stats;
-            Ready = Iter->SkipToRowVersion(rowVersion, stats, /* committed */ nullptr, /* observer */ nullptr);
+            Ready = Iter->SkipToRowVersion(rowVersion, stats, /* committed */ nullptr, /* observer */ nullptr,
+                /* decided */ ITransactionSet::None);
 
             if (Ready == EReady::Data)
                 Ready = RollUp();
