diff --git a/src/core/dense_set.cc b/src/core/dense_set.cc
index c3273ba88a34..d78b572a6e30 100644
--- a/src/core/dense_set.cc
+++ b/src/core/dense_set.cc
@@ -648,6 +648,8 @@ void DenseSet::Delete(DensePtr* prev, DensePtr* ptr) {
 
       DenseLinkKey* plink = prev->AsLink();
       DensePtr tmp = DensePtr::From(plink);
+      // Transfer TTL flag
+      tmp.SetTtl(prev->HasTtl());
       DCHECK(ObjectAllocSize(tmp.GetObject()));
 
       FreeLink(plink);
