diff --git a/test/sql/parser/test_value_functions.test b/test/sql/parser/test_value_functions.test
index 4ad0492d078d..b25be2238c16 100644
--- a/test/sql/parser/test_value_functions.test
+++ b/test/sql/parser/test_value_functions.test
@@ -46,3 +46,13 @@ statement error
 SELECT useRrr
 ----
 useRrr
+
+# value function in having
+query II
+select
+  1 as one,
+  max(cast('1000-05-01 00:00:00' as timestamp)) as mt
+group by one
+having max(cast('1000-05-01 00:00:00' as timestamp))  <= current_timestamp;
+----
+1	1000-05-01 00:00:00
