diff --git a/tests/queries/0_stateless/02096_sample_by_tuple.reference b/tests/queries/0_stateless/02096_sample_by_tuple.reference
new file mode 100644
index 000000000000..e69de29bb2d1
diff --git a/tests/queries/0_stateless/02096_sample_by_tuple.sql b/tests/queries/0_stateless/02096_sample_by_tuple.sql
new file mode 100644
index 000000000000..4996c9b83849
--- /dev/null
+++ b/tests/queries/0_stateless/02096_sample_by_tuple.sql
@@ -0,0 +1,7 @@
+DROP TABLE IF EXISTS t;
+
+CREATE TABLE t (n UInt8) ENGINE=MergeTree ORDER BY n SAMPLE BY tuple(); -- { serverError 80 }
+
+CREATE TABLE t (n UInt8) ENGINE=MergeTree ORDER BY tuple();
+
+ALTER TABLE t MODIFY SAMPLE BY tuple(); -- { serverError 80 }
