diff --git a/tests/queries/0_stateless/00900_long_parquet_load.reference b/tests/queries/0_stateless/00900_long_parquet_load.reference
index 50631bfab592..2316afeb2e70 100644
--- a/tests/queries/0_stateless/00900_long_parquet_load.reference
+++ b/tests/queries/0_stateless/00900_long_parquet_load.reference
@@ -142,6 +142,39 @@ abc	5	2	1	[1,2]
 1552
 1552
 1552
+=== Try load data from fixed_array_int.parquet
+idx1	[73,33,88,34,94,96,11,90,20,17]
+idx10	[10,15,52,43,22,50,42,87,19,91]
+idx2	[10,18,74,39,2,55,13,41,6,42]
+idx3	[48,27,74,82,70,46,18,78,63,73]
+idx4	[87,56,32,98,25,18,66,21,20,2]
+idx5	[67,84,56,84,50,21,93,90,37,42]
+idx6	[53,14,44,96,40,71,26,74,27,25]
+idx7	[46,1,21,11,83,74,11,63,28,49]
+idx8	[55,30,50,19,12,95,5,83,71,34]
+idx9	[90,53,83,54,35,87,73,74,98,50]
+=== Try load data from fixed_array_nested_list_int.parquet
+idx1	[[34,11,56,20,90,20,17,55,54,99],[22,18,15,77,68,51,30,76,9,40],[67,24,2,27,72,53,99,57,67,96],[95,77,60,47,68,56,91,28,90,38],[77,8,75,89,84,81,60,82,71,26],[41,89,60,15,43,81,31,68,71,65],[84,97,22,9,26,59,97,10,16,60],[13,82,16,47,32,74,34,78,90,15],[35,44,50,18,82,71,20,68,89,41],[60,59,7,65,18,23,24,37,48,45]]
+idx10	[[10,72,13,85,89,15,51,73,64,49],[78,28,54,57,10,1,3,35,23,15],[97,93,76,87,86,21,30,9,58,21],[27,23,35,71,4,68,90,14,93,87],[59,98,66,94,8,90,16,20,33,10],[73,84,79,37,75,50,64,74,79,31],[20,64,79,51,20,41,7,69,5,22],[37,60,1,99,45,43,56,26,15,83],[7,96,6,58,7,59,57,6,22,92],[97,11,71,55,32,79,88,20,58,31]]
+idx2	[[72,32,17,44,69,37,95,64,47,70],[15,4,28,31,80,63,40,93,17,87],[28,46,56,21,65,20,5,76,18,61],[66,44,5,66,71,73,26,70,47,22],[9,9,62,21,74,75,32,71,2,61],[66,81,43,46,10,78,66,31,34,17],[30,5,10,15,56,52,36,3,88,16],[45,84,39,12,39,13,43,66,67,16],[27,18,68,80,87,19,98,39,90,54],[48,96,24,67,66,91,3,10,7,66]]
+idx3	[[54,93,69,82,45,54,35,59,20,39],[80,3,15,33,45,22,45,11,28,94],[68,26,89,38,49,28,29,59,93,57],[11,62,53,12,16,83,2,55,65,6],[37,49,56,93,33,77,22,53,20,22],[4,75,2,5,34,5,90,67,2,79],[22,2,80,14,13,44,33,11,31,24],[11,54,75,33,60,20,10,51,56,33],[72,5,84,78,7,95,21,2,88,75],[69,88,52,85,70,96,51,69,48,20]]
+idx4	[[65,58,31,62,38,29,32,71,87,49],[66,89,59,83,29,32,59,49,68,34],[81,54,14,69,39,84,95,37,15,47],[15,51,15,38,90,61,76,51,11,24],[64,93,27,26,29,15,15,4,95,36],[60,61,57,66,61,25,21,97,34,36],[7,34,30,4,8,29,68,48,58,64],[77,19,92,77,81,21,52,1,86,92],[16,91,1,40,57,44,76,10,78,10],[88,13,70,61,2,76,78,69,90,24]]
+idx5	[[52,98,63,41,2,6,56,80,72,15],[11,64,89,57,65,92,63,39,76,51],[47,25,33,28,58,22,83,19,54,63],[61,8,20,24,83,44,33,32,65,90],[51,75,72,55,54,5,43,90,64,30],[86,87,48,65,84,67,82,94,2,9],[1,75,83,79,64,63,52,51,33,72],[12,84,92,27,66,32,40,48,34,55],[59,91,19,77,70,56,10,82,40,60],[79,83,14,82,42,22,72,54,12,9]]
+idx6	[[71,60,6,31,40,66,78,6,23,14],[62,88,38,68,37,37,63,18,57,3],[45,93,1,49,99,33,10,91,6,17],[13,40,57,85,48,14,74,89,43,49],[18,17,39,75,52,30,48,4,13,55],[3,34,73,71,75,58,6,73,73,31],[28,74,26,3,2,49,50,60,27,79],[35,28,12,10,1,21,61,70,65,37],[30,27,51,85,89,84,73,48,4,71],[1,86,23,68,82,9,6,95,14,25]]
+idx7	[[58,24,79,16,32,36,36,91,18,22],[85,29,36,16,75,79,71,70,6,39],[47,15,82,30,55,14,49,47,38,13],[28,54,95,82,25,16,44,82,33,11],[58,16,87,96,65,3,10,68,87,15],[94,84,65,50,21,78,8,78,89,72],[39,41,67,23,21,83,43,94,31,15],[67,58,73,87,58,71,52,10,30,90],[69,65,72,89,51,8,39,80,49,79],[32,36,76,11,88,87,75,55,33,74]]
+idx8	[[8,26,45,3,25,83,88,77,98,38],[52,76,79,94,6,74,73,31,93,53],[89,24,62,83,35,24,60,24,41,14],[52,13,13,32,87,77,19,20,52,47],[50,46,66,30,26,85,91,50,98,83],[87,44,5,11,25,52,9,55,15,37],[18,24,36,72,84,10,13,59,16,65],[93,4,19,13,75,64,73,29,81,25],[61,65,1,45,75,88,19,4,73,32],[23,20,27,55,13,34,97,80,8,19]]
+idx9	[[32,6,44,30,66,48,93,67,62,16],[49,67,73,32,86,78,28,88,25,60],[70,10,95,16,10,49,12,47,22,62],[95,86,63,43,4,96,3,34,60,53],[60,75,96,66,96,82,62,31,78,72],[17,44,13,7,45,54,27,84,10,68],[43,49,86,10,57,99,32,72,73,41],[57,79,15,62,79,87,67,69,75,59],[45,18,2,19,45,96,8,86,71,97],[20,20,11,15,82,29,16,12,87,87]]
+=== Try load data from fixed_array_str.parquet
+idx1	['str10','str11','str12','str13','str14','str15','str16','str17','str18','str19']
+idx10	['str100','str101','str102','str103','str104','str105','str106','str107','str108','str109']
+idx2	['str20','str21','str22','str23','str24','str25','str26','str27','str28','str29']
+idx3	['str30','str31','str32','str33','str34','str35','str36','str37','str38','str39']
+idx4	['str40','str41','str42','str43','str44','str45','str46','str47','str48','str49']
+idx5	['str50','str51','str52','str53','str54','str55','str56','str57','str58','str59']
+idx6	['str60','str61','str62','str63','str64','str65','str66','str67','str68','str69']
+idx7	['str70','str71','str72','str73','str74','str75','str76','str77','str78','str79']
+idx8	['str80','str81','str82','str83','str84','str85','str86','str87','str88','str89']
+idx9	['str90','str91','str92','str93','str94','str95','str96','str97','str98','str99']
 === Try load data from fixed_length_decimal.parquet
 1
 2
diff --git a/tests/queries/0_stateless/data_parquet/fixed_array_int.parquet b/tests/queries/0_stateless/data_parquet/fixed_array_int.parquet
new file mode 100644
index 000000000000..835b05bbf56e
Binary files /dev/null and b/tests/queries/0_stateless/data_parquet/fixed_array_int.parquet differ
diff --git a/tests/queries/0_stateless/data_parquet/fixed_array_int.parquet.columns b/tests/queries/0_stateless/data_parquet/fixed_array_int.parquet.columns
new file mode 100644
index 000000000000..2c383e8343f8
--- /dev/null
+++ b/tests/queries/0_stateless/data_parquet/fixed_array_int.parquet.columns
@@ -0,0 +1,1 @@
+`idx` String, `lst` Array(Int64)
\ No newline at end of file
diff --git a/tests/queries/0_stateless/data_parquet/fixed_array_nested_list_int.parquet b/tests/queries/0_stateless/data_parquet/fixed_array_nested_list_int.parquet
new file mode 100644
index 000000000000..8317f1d5d564
Binary files /dev/null and b/tests/queries/0_stateless/data_parquet/fixed_array_nested_list_int.parquet differ
diff --git a/tests/queries/0_stateless/data_parquet/fixed_array_nested_list_int.parquet.columns b/tests/queries/0_stateless/data_parquet/fixed_array_nested_list_int.parquet.columns
new file mode 100644
index 000000000000..3e1c9533a324
--- /dev/null
+++ b/tests/queries/0_stateless/data_parquet/fixed_array_nested_list_int.parquet.columns
@@ -0,0 +1,1 @@
+`idx` String, `lst` Array(Array(Int64))
\ No newline at end of file
diff --git a/tests/queries/0_stateless/data_parquet/fixed_array_str.parquet b/tests/queries/0_stateless/data_parquet/fixed_array_str.parquet
new file mode 100644
index 000000000000..2a6bec05799a
Binary files /dev/null and b/tests/queries/0_stateless/data_parquet/fixed_array_str.parquet differ
diff --git a/tests/queries/0_stateless/data_parquet/fixed_array_str.parquet.columns b/tests/queries/0_stateless/data_parquet/fixed_array_str.parquet.columns
new file mode 100644
index 000000000000..67a7b6bc04d5
--- /dev/null
+++ b/tests/queries/0_stateless/data_parquet/fixed_array_str.parquet.columns
@@ -0,0 +1,1 @@
+`idx` String, `lst` Array(String)
\ No newline at end of file
