diff --git a/tests/queries/0_stateless/01417_update_permutation_crash.reference b/tests/queries/0_stateless/01417_update_permutation_crash.reference
new file mode 100644
index 000000000000..2b4f3eda2e0c
--- /dev/null
+++ b/tests/queries/0_stateless/01417_update_permutation_crash.reference
@@ -0,0 +1,1 @@
+(1,1,0)
diff --git a/tests/queries/0_stateless/01417_update_permutation_crash.sql b/tests/queries/0_stateless/01417_update_permutation_crash.sql
new file mode 100644
index 000000000000..f59237817349
--- /dev/null
+++ b/tests/queries/0_stateless/01417_update_permutation_crash.sql
@@ -0,0 +1,1 @@
+select tuple(1, 1, number) as t from numbers_mt(1000001) order by t, number limit 1;
