diff --git a/tests/queries/0_stateless/00718_format_datetime.reference b/tests/queries/0_stateless/00718_format_datetime.reference
index f22c953e739d..43c630ac83b9 100644
--- a/tests/queries/0_stateless/00718_format_datetime.reference
+++ b/tests/queries/0_stateless/00718_format_datetime.reference
@@ -48,9 +48,9 @@ no formatting pattern	no formatting pattern
 000000
 000000
 000000
-123
+123000
+123456
 123456
-123456789
 0
 0
 0
@@ -58,8 +58,8 @@ no formatting pattern	no formatting pattern
 123
 123456
 123456789
-2022-12-08 18:11:29.123400000
-2022-12-08 18:11:29.1
+2022-12-08 18:11:29.123400
+2022-12-08 18:11:29.100000
 2022-12-08 18:11:29.000000
 2022-12-08 18:11:29.000000
 2022-12-08 00:00:00.000000
diff --git a/tests/queries/0_stateless/00718_format_datetime_1.reference b/tests/queries/0_stateless/00718_format_datetime_1.reference
index e495b69ddfcc..78367f4cabb7 100644
--- a/tests/queries/0_stateless/00718_format_datetime_1.reference
+++ b/tests/queries/0_stateless/00718_format_datetime_1.reference
@@ -1,5 +1,5 @@
-1900-01-01 00:00:00.000
-1962-12-08 18:11:29.123
-1969-12-31 23:59:59.999
-1970-01-01 00:00:00.000
-1970-01-01 00:00:00.001
+1900-01-01 00:00:00.000000
+1962-12-08 18:11:29.123000
+1969-12-31 23:59:59.999000
+1970-01-01 00:00:00.000000
+1970-01-01 00:00:00.001000
diff --git a/tests/queries/0_stateless/03287_format_datetime_mysqlfraction.reference b/tests/queries/0_stateless/03287_format_datetime_mysqlfraction.reference
new file mode 100644
index 000000000000..ccfbaa46a840
--- /dev/null
+++ b/tests/queries/0_stateless/03287_format_datetime_mysqlfraction.reference
@@ -0,0 +1,14 @@
+Default settings
+123456
+123456
+123400
+120000
+100000
+000000
+Compatibility settings
+12345678
+123456
+1234
+12
+1
+000000
diff --git a/tests/queries/0_stateless/03287_format_datetime_mysqlfraction.sql b/tests/queries/0_stateless/03287_format_datetime_mysqlfraction.sql
new file mode 100644
index 000000000000..9ed5a1151df9
--- /dev/null
+++ b/tests/queries/0_stateless/03287_format_datetime_mysqlfraction.sql
@@ -0,0 +1,17 @@
+-- Test for issue #72879
+
+SELECT 'Default settings';
+SELECT formatDateTime(toDateTime64('1970-01-01 00:00:00.12345678', 8, 'UTC'), '%f') SETTINGS formatdatetime_f_prints_scale_number_of_digits = 0;
+SELECT formatDateTime(toDateTime64('1970-01-01 00:00:00.123456', 6, 'UTC'), '%f') SETTINGS formatdatetime_f_prints_scale_number_of_digits = 0;
+SELECT formatDateTime(toDateTime64('1970-01-01 00:00:00.1234', 4, 'UTC'), '%f') SETTINGS formatdatetime_f_prints_scale_number_of_digits = 0;
+SELECT formatDateTime(toDateTime64('1970-01-01 00:00:00.12', 2, 'UTC'), '%f') SETTINGS formatdatetime_f_prints_scale_number_of_digits = 0;
+SELECT formatDateTime(toDateTime64('1970-01-01 00:00:00.1', 1, 'UTC'), '%f') SETTINGS formatdatetime_f_prints_scale_number_of_digits = 0;
+SELECT formatDateTime(toDateTime64('1970-01-01 00:00:00', 0, 'UTC'), '%f') SETTINGS formatdatetime_f_prints_scale_number_of_digits = 0;
+
+SELECT 'Compatibility settings';
+SELECT formatDateTime(toDateTime64('1970-01-01 00:00:00.12345678', 8, 'UTC'), '%f') SETTINGS formatdatetime_f_prints_scale_number_of_digits = 1;
+SELECT formatDateTime(toDateTime64('1970-01-01 00:00:00.123456', 6, 'UTC'), '%f') SETTINGS formatdatetime_f_prints_scale_number_of_digits = 1;
+SELECT formatDateTime(toDateTime64('1970-01-01 00:00:00.1234', 4, 'UTC'), '%f') SETTINGS formatdatetime_f_prints_scale_number_of_digits = 1;
+SELECT formatDateTime(toDateTime64('1970-01-01 00:00:00.12', 2, 'UTC'), '%f') SETTINGS formatdatetime_f_prints_scale_number_of_digits = 1;
+SELECT formatDateTime(toDateTime64('1970-01-01 00:00:00.1', 1, 'UTC'), '%f') SETTINGS formatdatetime_f_prints_scale_number_of_digits = 1;
+SELECT formatDateTime(toDateTime64('1970-01-01 00:00:00', 0, 'UTC'), '%f') SETTINGS formatdatetime_f_prints_scale_number_of_digits = 1;
