diff --git a/test/sql/json/scalar/test_json_extract.test b/test/sql/json/scalar/test_json_extract.test
index 6e13039fb2e6..7bb2717919b1 100644
--- a/test/sql/json/scalar/test_json_extract.test
+++ b/test/sql/json/scalar/test_json_extract.test
@@ -314,3 +314,41 @@ query T
 execute q1('a')
 ----
 1
+
+# test issue 11997
+query I
+select json_extract_string(json('{"j[so]n_\"key": 67}'), '$."j[so]n_\"key"');
+----
+67
+
+query I
+select '{"\"duck\"": 42}'->'$."\"duck\""';
+----
+42
+
+query I
+select '{"\"du\\ck\"": 42}'->'$."\"du\\ck\""';
+----
+42
+
+query I
+select '{"\"du\\ck\"": 42}'->'$."\"du\ck\""';
+----
+42
+
+query I
+select '{"du\\ck": 42}'->'$.du\ck';
+----
+42
+
+# characters other than \\ or \" get ignored (for now)
+query I
+select '{"\"du
ck\"": 42}'->'$."\"du
ck\""';
+----
+NULL
+
+# need to use chr(10) for 

+query I
+select '{"\"du
ck\"": 42}'->('$."\"du' || chr(10) || 'ck\""');
+----
+42
