diff --git a/src/Storages/RocksDB/StorageEmbeddedRocksDB.cpp b/src/Storages/RocksDB/StorageEmbeddedRocksDB.cpp
index 05faef2c1250..a0e14cd4a841 100644
--- a/src/Storages/RocksDB/StorageEmbeddedRocksDB.cpp
+++ b/src/Storages/RocksDB/StorageEmbeddedRocksDB.cpp
@@ -594,7 +594,7 @@ class ReadFromEmbeddedRocksDB : public SourceStepWithFilter
     size_t num_streams;
 
     FieldVectorPtr keys;
-    bool all_scan = false;
+    bool all_scan = true;
 };
 
 void StorageEmbeddedRocksDB::read(
diff --git a/src/Storages/StorageKeeperMap.cpp b/src/Storages/StorageKeeperMap.cpp
index 6720fc75bd8a..ffa2c6534c90 100644
--- a/src/Storages/StorageKeeperMap.cpp
+++ b/src/Storages/StorageKeeperMap.cpp
@@ -632,7 +632,7 @@ class ReadFromKeeperMap : public SourceStepWithFilter
     bool with_version_column;
 
     FieldVectorPtr keys;
-    bool all_scan = false;
+    bool all_scan = true;
 
     template<typename KeyContainerPtr>
     void initializePipelineImpl(QueryPipelineBuilder & pipeline, KeyContainerPtr key_container);
diff --git a/src/Storages/StorageRedis.cpp b/src/Storages/StorageRedis.cpp
index d9db5df8d490..74a0b568695e 100644
--- a/src/Storages/StorageRedis.cpp
+++ b/src/Storages/StorageRedis.cpp
@@ -251,7 +251,7 @@ class ReadFromRedis : public SourceStepWithFilter
     size_t num_streams;
 
     FieldVectorPtr keys;
-    bool all_scan = false;
+    bool all_scan = true;
 };
 
 void StorageRedis::read(
