diff --git a/tests/integration/test_storage_redis/test.py b/tests/integration/test_storage_redis/test.py
index 5a931aa2f4ee..5eef47f2b073 100644
--- a/tests/integration/test_storage_redis/test.py
+++ b/tests/integration/test_storage_redis/test.py
@@ -461,3 +461,6 @@ def check_query(query, read_type, keys_count, rows_read):
     check_query("SELECT * FROM test_get_keys", "FullScan", 0, 3)
     check_query("SELECT * FROM test_get_keys WHERE k = 1", "GetKeys", 1, 1)
     check_query("SELECT * FROM test_get_keys WHERE k in (3, 5)", "GetKeys", 2, 1)
+
+    plan = node.query("EXPLAIN actions=1, optimize=0 SELECT * FROM test_get_keys")
+    assert 'ReadType: FullScan' in plan
diff --git a/tests/queries/0_stateless/02375_rocksdb_with_filters.reference b/tests/queries/0_stateless/02375_rocksdb_with_filters.reference
index 77201bc4c43d..7d04ab283327 100644
--- a/tests/queries/0_stateless/02375_rocksdb_with_filters.reference
+++ b/tests/queries/0_stateless/02375_rocksdb_with_filters.reference
@@ -1,5 +1,7 @@
     ReadFromEmbeddedRocksDB
     ReadType: FullScan
+ReadFromEmbeddedRocksDB
+ReadType:FullScan
 1
 "rows_read":1,
     ReadFromEmbeddedRocksDB
diff --git a/tests/queries/0_stateless/02375_rocksdb_with_filters.sh b/tests/queries/0_stateless/02375_rocksdb_with_filters.sh
index 3a03fb89accb..36fcd0b5f6dd 100755
--- a/tests/queries/0_stateless/02375_rocksdb_with_filters.sh
+++ b/tests/queries/0_stateless/02375_rocksdb_with_filters.sh
@@ -11,6 +11,7 @@ $CLICKHOUSE_CLIENT --query="CREATE TABLE rocksdb_with_filter (key String, value
 $CLICKHOUSE_CLIENT --query="INSERT INTO rocksdb_with_filter (*) SELECT n.number, n.number*10 FROM numbers(10000) n;"
 
 $CLICKHOUSE_CLIENT --query "EXPLAIN actions=1 SELECT value FROM rocksdb_with_filter LIMIT 1" | grep -A 2 "ReadFromEmbeddedRocksDB"
+$CLICKHOUSE_CLIENT --query "EXPLAIN actions=1,optimize=0 SELECT value FROM rocksdb_with_filter" | grep -A 2 "ReadFromEmbeddedRocksDB" | tr -d "[:blank:]"
 
 $CLICKHOUSE_CLIENT --query "SELECT count() FROM rocksdb_with_filter WHERE key = '5000'"
 $CLICKHOUSE_CLIENT --query "SELECT value FROM rocksdb_with_filter WHERE key = '5000' FORMAT JSON" | grep "rows_read" | tr -d "[:blank:]"
diff --git a/tests/queries/0_stateless/03541_keeper_map_filter_keys.reference b/tests/queries/0_stateless/03541_keeper_map_filter_keys.reference
index 12a3bbcf41a9..c603377835be 100644
--- a/tests/queries/0_stateless/03541_keeper_map_filter_keys.reference
+++ b/tests/queries/0_stateless/03541_keeper_map_filter_keys.reference
@@ -1,5 +1,7 @@
     ReadFromKeeperMap
     ReadType: FullScan
+ReadFromKeeperMap
+ReadType:FullScan
 1
 "rows_read":1,
     ReadFromKeeperMap
diff --git a/tests/queries/0_stateless/03541_keeper_map_filter_keys.sh b/tests/queries/0_stateless/03541_keeper_map_filter_keys.sh
index b862bda95799..39949e133373 100755
--- a/tests/queries/0_stateless/03541_keeper_map_filter_keys.sh
+++ b/tests/queries/0_stateless/03541_keeper_map_filter_keys.sh
@@ -11,6 +11,7 @@ $CLICKHOUSE_CLIENT --query="CREATE TABLE keeper_map_with_filter (key String, val
 $CLICKHOUSE_CLIENT --query="INSERT INTO keeper_map_with_filter (*) SELECT n.number, n.number*10 FROM numbers(10) n;"
 
 $CLICKHOUSE_CLIENT --query "EXPLAIN actions=1 SELECT value FROM keeper_map_with_filter LIMIT 1" | grep -A 2 "ReadFromKeeperMap"
+$CLICKHOUSE_CLIENT --query "EXPLAIN actions=1,optimize=0 SELECT value FROM keeper_map_with_filter" | grep -A 2 "ReadFromKeeperMap" | tr -d "[:blank:]"
 
 $CLICKHOUSE_CLIENT --query "SELECT count() FROM keeper_map_with_filter WHERE key = '5'"
 $CLICKHOUSE_CLIENT --query "SELECT value FROM keeper_map_with_filter WHERE key = '5' FORMAT JSON" | grep "rows_read" | tr -d "[:blank:]"
