diff --git a/tests/queries/0_stateless/02459_read_in_order_bufer.reference b/tests/queries/0_stateless/02459_read_in_order_bufer.reference
new file mode 100644
index 000000000000..b040bdf61679
--- /dev/null
+++ b/tests/queries/0_stateless/02459_read_in_order_bufer.reference
@@ -0,0 +1,5 @@
+9
+8
+7
+6
+5
diff --git a/tests/queries/0_stateless/02459_read_in_order_bufer.sql b/tests/queries/0_stateless/02459_read_in_order_bufer.sql
new file mode 100644
index 000000000000..5a6e0a3dbc6c
--- /dev/null
+++ b/tests/queries/0_stateless/02459_read_in_order_bufer.sql
@@ -0,0 +1,13 @@
+CREATE TABLE mytable_stored (`a` UInt8) ENGINE = MergeTree ORDER BY a;
+CREATE TABLE mytable (`a` UInt8) ENGINE = Buffer(currentDatabase(), 'mytable_stored', 4, 600, 3600, 10, 100, 10000, 10000000);
+INSERT INTO mytable VALUES (0);
+INSERT INTO mytable VALUES (1);
+INSERT INTO mytable VALUES (2);
+INSERT INTO mytable VALUES (3);
+INSERT INTO mytable VALUES (4);
+INSERT INTO mytable VALUES (5);
+INSERT INTO mytable VALUES (6);
+INSERT INTO mytable VALUES (7);
+INSERT INTO mytable VALUES (8);
+INSERT INTO mytable VALUES (9);
+SELECT a FROM mytable ORDER BY a DESC LIMIT 5;
