diff --git a/tests/queries/0_stateless/02131_skip_index_not_materialized.reference b/tests/queries/0_stateless/02131_skip_index_not_materialized.reference
new file mode 100644
index 000000000000..d00491fd7e5b
--- /dev/null
+++ b/tests/queries/0_stateless/02131_skip_index_not_materialized.reference
@@ -0,0 +1,1 @@
+1
diff --git a/tests/queries/0_stateless/02131_skip_index_not_materialized.sql b/tests/queries/0_stateless/02131_skip_index_not_materialized.sql
new file mode 100644
index 000000000000..cae0b1d9fb39
--- /dev/null
+++ b/tests/queries/0_stateless/02131_skip_index_not_materialized.sql
@@ -0,0 +1,12 @@
+DROP TABLE IF EXISTS t_index_non_materialized;
+
+CREATE TABLE t_index_non_materialized (a UInt32) ENGINE = MergeTree ORDER BY tuple();
+
+INSERT INTO t_index_non_materialized VALUES (1);
+
+ALTER TABLE t_index_non_materialized ADD INDEX ind_set (a) TYPE set(1) GRANULARITY 1;
+ALTER TABLE t_index_non_materialized ADD INDEX ind_minmax (a) TYPE minmax() GRANULARITY 1;
+
+SELECT count() FROM t_index_non_materialized WHERE a = 1;
+
+DROP TABLE t_index_non_materialized;
