diff --git a/tests/queries/0_stateless/01881_union_header_mismatch_bug.sql b/tests/queries/0_stateless/01881_union_header_mismatch_bug.sql
index 9a220ffd49fc..bf8749fb046b 100644
--- a/tests/queries/0_stateless/01881_union_header_mismatch_bug.sql
+++ b/tests/queries/0_stateless/01881_union_header_mismatch_bug.sql
@@ -28,3 +28,8 @@ WHERE number IN
     SELECT number
     FROM numbers(5)
 ) order by label, number;
+
+SELECT NULL FROM
+(SELECT [1048575, NULL] AS ax, 2147483648 AS c) t1 ARRAY JOIN ax
+INNER JOIN (SELECT NULL AS c) t2 USING (c);
+
