diff --git a/tests/queries/0_stateless/02499_extract_key_value_pairs_multiple_input.reference b/tests/queries/0_stateless/02499_extract_key_value_pairs_multiple_input.reference
index 9daeafff7793..125afa194276 100644
--- a/tests/queries/0_stateless/02499_extract_key_value_pairs_multiple_input.reference
+++ b/tests/queries/0_stateless/02499_extract_key_value_pairs_multiple_input.reference
@@ -304,3 +304,27 @@ WITH
 SELECT
     x;
 {}
+-- check str_to_map alias (it is case-insensitive)
+WITH
+    sTr_tO_mAp('name:neymar, age:31 team:psg,nationality:brazil') AS s_map,
+    CAST(
+            arrayMap(
+                    (x) -> (x, s_map[x]), arraySort(mapKeys(s_map))
+                ),
+            'Map(String,String)'
+        ) AS x
+SELECT
+    x;
+{'age':'31','name':'neymar','nationality':'brazil','team':'psg'}
+-- check mapFromString alias
+WITH
+    mapFromString('name:neymar, age:31 team:psg,nationality:brazil') AS s_map,
+    CAST(
+            arrayMap(
+                    (x) -> (x, s_map[x]), arraySort(mapKeys(s_map))
+                ),
+            'Map(String,String)'
+        ) AS x
+SELECT
+    x;
+{'age':'31','name':'neymar','nationality':'brazil','team':'psg'}
diff --git a/tests/queries/0_stateless/02499_extract_key_value_pairs_multiple_input.sql b/tests/queries/0_stateless/02499_extract_key_value_pairs_multiple_input.sql
index 7d681e36a4bb..0a1500b17968 100644
--- a/tests/queries/0_stateless/02499_extract_key_value_pairs_multiple_input.sql
+++ b/tests/queries/0_stateless/02499_extract_key_value_pairs_multiple_input.sql
@@ -426,3 +426,27 @@ WITH
         ) AS x
 SELECT
     x;
+
+-- check str_to_map alias (it is case-insensitive)
+WITH
+    sTr_tO_mAp('name:neymar, age:31 team:psg,nationality:brazil') AS s_map,
+    CAST(
+            arrayMap(
+                    (x) -> (x, s_map[x]), arraySort(mapKeys(s_map))
+                ),
+            'Map(String,String)'
+        ) AS x
+SELECT
+    x;
+
+-- check mapFromString alias
+WITH
+    mapFromString('name:neymar, age:31 team:psg,nationality:brazil') AS s_map,
+    CAST(
+            arrayMap(
+                    (x) -> (x, s_map[x]), arraySort(mapKeys(s_map))
+                ),
+            'Map(String,String)'
+        ) AS x
+SELECT
+    x;
