diff --git a/tests/queries/0_stateless/02815_alias_to_length.reference b/tests/queries/0_stateless/02815_alias_to_length.reference
new file mode 100644
index 000000000000..de958a364ef9
--- /dev/null
+++ b/tests/queries/0_stateless/02815_alias_to_length.reference
@@ -0,0 +1,4 @@
+4
+4
+15
+4
diff --git a/tests/queries/0_stateless/02815_alias_to_length.sql b/tests/queries/0_stateless/02815_alias_to_length.sql
new file mode 100644
index 000000000000..780ac7dac6dd
--- /dev/null
+++ b/tests/queries/0_stateless/02815_alias_to_length.sql
@@ -0,0 +1,6 @@
+SELECT OCTET_LENGTH('1234');
+SELECT OcTet_lenGtH('1234');
+SELECT OCTET_LENGTH('你好，世界');
+
+-- This is a implementation-specific behavior of getting the length of an array.
+SELECT OCTET_LENGTH([1,2,3,4]);
