diff --git a/.gitmodules b/.gitmodules
index 618cfe6e76b4..ebeef312ae81 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -65,12 +65,6 @@
 [submodule "contrib/libgsasl"]
 	path = contrib/libgsasl
 	url = https://github.com/ClickHouse/libgsasl.git
-[submodule "contrib/libcxx"]
-	path = contrib/libcxx
-	url = https://github.com/ClickHouse/libcxx.git
-[submodule "contrib/libcxxabi"]
-	path = contrib/libcxxabi
-	url = https://github.com/ClickHouse/libcxxabi.git
 [submodule "contrib/snappy"]
 	path = contrib/snappy
 	url = https://github.com/ClickHouse/snappy.git
diff --git a/contrib/libcxx b/contrib/libcxx
deleted file mode 160000
index 4db7f838afd3..000000000000
--- a/contrib/libcxx
+++ /dev/null
@@ -1,1 +0,0 @@
-Subproject commit 4db7f838afd3139eb3761694b04d31275df45d2d
diff --git a/contrib/libcxx-cmake/CMakeLists.txt b/contrib/libcxx-cmake/CMakeLists.txt
index 8dc154e9d91a..21ed76f8b6fc 100644
--- a/contrib/libcxx-cmake/CMakeLists.txt
+++ b/contrib/libcxx-cmake/CMakeLists.txt
@@ -1,6 +1,6 @@
 include(CheckCXXCompilerFlag)
 
-set(LIBCXX_SOURCE_DIR "${ClickHouse_SOURCE_DIR}/contrib/libcxx")
+set(LIBCXX_SOURCE_DIR "${ClickHouse_SOURCE_DIR}/contrib/llvm-project/libcxx")
 
 set(SRCS
 "${LIBCXX_SOURCE_DIR}/src/algorithm.cpp"
diff --git a/contrib/libcxxabi b/contrib/libcxxabi
deleted file mode 160000
index a736a6b3c6a7..000000000000
--- a/contrib/libcxxabi
+++ /dev/null
@@ -1,1 +0,0 @@
-Subproject commit a736a6b3c6a7b8aae2ebad629ca21b2c55b4820e
diff --git a/contrib/libcxxabi-cmake/CMakeLists.txt b/contrib/libcxxabi-cmake/CMakeLists.txt
index a59452eee9a1..0473527912ea 100644
--- a/contrib/libcxxabi-cmake/CMakeLists.txt
+++ b/contrib/libcxxabi-cmake/CMakeLists.txt
@@ -1,4 +1,4 @@
-set(LIBCXXABI_SOURCE_DIR "${ClickHouse_SOURCE_DIR}/contrib/libcxxabi")
+set(LIBCXXABI_SOURCE_DIR "${ClickHouse_SOURCE_DIR}/contrib/llvm-project/libcxxabi")
 
 set(SRCS
 "${LIBCXXABI_SOURCE_DIR}/src/abort_message.cpp"
diff --git a/contrib/llvm-project b/contrib/llvm-project
index 3a39038345a4..e61a81aa6fc5 160000
--- a/contrib/llvm-project
+++ b/contrib/llvm-project
@@ -1,1 +1,1 @@
-Subproject commit 3a39038345a400e7e767811b142a94355d511215
+Subproject commit e61a81aa6fc529b469e2a54b7ce788606e138b5d
