diff --git a/tools/pythonpkg/tests/test_dbapi07.py b/tools/pythonpkg/tests/test_dbapi07.py
new file mode 100644
index 000000000000..951d6bf81db1
--- /dev/null
+++ b/tools/pythonpkg/tests/test_dbapi07.py
@@ -0,0 +1,9 @@
+# timestamp ms precision
+
+import numpy 
+
+
+class TestNumpyTimestampMilliseconds(object):
+    def test_numpy_timestamp(self, duckdb_cursor):
+        res = duckdb_cursor.execute("SELECT TIMESTAMP '2020-01-10 00:00:00.500' as test_time").fetchnumpy()
+        assert res['test_time'] == numpy.datetime64('2020-01-10 00:00:00.500')
