diff --git a/tests/queries/0_stateless/01851_hedged_connections_external_tables.reference b/tests/queries/0_stateless/01851_hedged_connections_external_tables.reference
new file mode 100644
index 000000000000..573541ac9702
--- /dev/null
+++ b/tests/queries/0_stateless/01851_hedged_connections_external_tables.reference
@@ -0,0 +1,1 @@
+0
diff --git a/tests/queries/0_stateless/01851_hedged_connections_external_tables.sql b/tests/queries/0_stateless/01851_hedged_connections_external_tables.sql
new file mode 100644
index 000000000000..c4625720e598
--- /dev/null
+++ b/tests/queries/0_stateless/01851_hedged_connections_external_tables.sql
@@ -0,0 +1,1 @@
+select number from remote('127.0.0.{3|2}', numbers(2)) where number global in (select number from numbers(1))　settings async_socket_for_remote=1, use_hedged_requests = 1, sleep_in_send_data_ms=10, receive_data_timeout_ms=1;
