diff --git a/tests/queries/0_stateless/03572_empty_tuple_in_nested_type.reference b/tests/queries/0_stateless/03572_empty_tuple_in_nested_type.reference
new file mode 100644
index 000000000000..6ea92a0bcbcd
--- /dev/null
+++ b/tests/queries/0_stateless/03572_empty_tuple_in_nested_type.reference
@@ -0,0 +1,9 @@
+-- { echo ON }
+
+DROP TABLE IF EXISTS t0;
+CREATE TABLE t0 (c0 Array(Tuple())) ENGINE = Memory;
+SET max_insert_block_size = 4;
+INSERT INTO TABLE t0 (c0) VALUES ([()]), ([()]), ([()]), ([()]), ([()]), ([()]), ([()]::Array(Tuple())), ([()]), ([(), ()]), ([()]);
+DROP TABLE t0;
+SELECT [(), ()];
+[(),()]
diff --git a/tests/queries/0_stateless/03572_empty_tuple_in_nested_type.sql b/tests/queries/0_stateless/03572_empty_tuple_in_nested_type.sql
new file mode 100644
index 000000000000..9fbb20a4eb61
--- /dev/null
+++ b/tests/queries/0_stateless/03572_empty_tuple_in_nested_type.sql
@@ -0,0 +1,13 @@
+-- { echo ON }
+
+DROP TABLE IF EXISTS t0;
+
+CREATE TABLE t0 (c0 Array(Tuple())) ENGINE = Memory;
+
+SET max_insert_block_size = 4;
+
+INSERT INTO TABLE t0 (c0) VALUES ([()]), ([()]), ([()]), ([()]), ([()]), ([()]), ([()]::Array(Tuple())), ([()]), ([(), ()]), ([()]);
+
+DROP TABLE t0;
+
+SELECT [(), ()];
