diff --git a/src/parser/parsed_data/alter_table_info.cpp b/src/parser/parsed_data/alter_table_info.cpp
index def8a8fffe65..b6cd24e8b41c 100644
--- a/src/parser/parsed_data/alter_table_info.cpp
+++ b/src/parser/parsed_data/alter_table_info.cpp
@@ -148,7 +148,8 @@ unique_ptr<AlterInfo> ChangeColumnTypeInfo::Deserialize(Deserializer &source, st
 // SetDefaultInfo
 //===--------------------------------------------------------------------===//
 unique_ptr<AlterInfo> SetDefaultInfo::Copy() const {
-	return make_unique_base<AlterInfo, SetDefaultInfo>(schema, name, column_name, expression->Copy());
+	return make_unique_base<AlterInfo, SetDefaultInfo>(schema, name, column_name,
+	                                                   expression ? expression->Copy() : nullptr);
 }
 
 void SetDefaultInfo::Serialize(Serializer &serializer) {
