diff --git a/test/sql/alter/default/drop_default.test b/test/sql/alter/default/drop_default.test
new file mode 100644
index 000000000000..7e566cf479b1
--- /dev/null
+++ b/test/sql/alter/default/drop_default.test
@@ -0,0 +1,30 @@
+# name: test/sql/alter/default/drop_default.test
+# description: Test ALTER TABLE DROP DEFAULT
+# group: [default]
+
+statement ok
+PRAGMA enable_verification;
+
+statement ok
+CREATE TABLE data(id INTEGER, x INTEGER);
+
+statement ok
+ALTER TABLE data ALTER COLUMN id DROP DEFAULT;
+
+statement ok
+INSERT INTO data VALUES (1, 0), (2, 1);
+
+statement ok
+ALTER TABLE data ALTER COLUMN id DROP DEFAULT;
+
+statement ok
+ALTER TABLE data ALTER COLUMN id DROP DEFAULT;
+
+statement ok
+ALTER TABLE data ALTER COLUMN x DROP DEFAULT;
+
+statement ok
+ALTER TABLE data ALTER COLUMN x DROP DEFAULT;
+
+statement error
+ALTER TABLE data ALTER COLUMN j DROP DEFAULT;
diff --git a/test/sql/alter/test_set_default.test b/test/sql/alter/default/test_set_default.test
similarity index 93%
rename from test/sql/alter/test_set_default.test
rename to test/sql/alter/default/test_set_default.test
index 1d7fedc43a2d..dc7ad9787c71 100644
--- a/test/sql/alter/test_set_default.test
+++ b/test/sql/alter/default/test_set_default.test
@@ -1,6 +1,6 @@
-# name: test/sql/alter/test_set_default.test
+# name: test/sql/alter/default/test_set_default.test
 # description: Test ALTER TABLE SET DEFAULT
-# group: [alter]
+# group: [default]
 
 statement ok
 CREATE TABLE test(i INTEGER, j INTEGER)
