diff --git a/tests/queries/0_stateless/00405_pretty_formats.reference b/tests/queries/0_stateless/00405_pretty_formats.reference
index bad93c19bb60..d1b080eddc7f 100644
--- a/tests/queries/0_stateless/00405_pretty_formats.reference
+++ b/tests/queries/0_stateless/00405_pretty_formats.reference
@@ -136,7 +136,7 @@
     ┡━━━━━━━╇━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
  [90m6. [0m│     5 │ 5     │ (5,'5') │               2 │
     └───────┴───────┴─────────┴─────────────────┘
-  Showed first 6.
+  Showed 6 out of 10 rows.
    ┌─[1mhello[0m─┬─[1mworld[0m─┬─[1mtuple[0m───┬─[1msometimes_nulls[0m─┐
 [90m1. [0m│     0 │ 0     │ (0,'0') │            ᴺᵁᴸᴸ │
 [90m2. [0m│     1 │ 1     │ (1,'1') │               1 │
@@ -147,7 +147,7 @@
     ┌─[1mhello[0m─┬─[1mworld[0m─┬─[1mtuple[0m───┬─[1msometimes_nulls[0m─┐
  [90m6. [0m│     5 │ 5     │ (5,'5') │               2 │
     └───────┴───────┴─────────┴─────────────────┘
-  Showed first 6.
+  Showed 6 out of 10 rows.
     [1mhello[0m   [1mworld[0m   [1mtuple[0m     [1msometimes_nulls[0m
 
 [90m1. [0m     0   0       (0,'0')              ᴺᵁᴸᴸ 
@@ -159,16 +159,17 @@
 
  [90m6. [0m     5   5       (5,'5')                 2 
 
-Showed first 6.
-   ┌─[1mhello[0m─┬─[1mworld[0m─┬─[1mtuple[0m───┬─[1msometimes_nulls[0m─┐
-[90m1. [0m│     0 │ 0     │ (0,'0') │            ᴺᵁᴸᴸ │
-[90m2. [0m│     1 │ 1     │ (1,'1') │               1 │
-[90m3. [0m│     2 │ 2     │ (2,'2') │               2 │
-[90m4. [0m│     3 │ 3     │ (3,'3') │            ᴺᵁᴸᴸ │
-[90m5. [0m│     4 │ 4     │ (4,'4') │               1 │
-[90m6. [0m│     5 │ 5     │ (5,'5') │               2 │
-   └───────┴───────┴─────────┴─────────────────┘
-  Showed first 6.
+Showed 6 out of 10 rows.
+    ┌─[1mhello[0m─┬─[1mworld[0m─┬─[1mtuple[0m───┬─[1msometimes_nulls[0m─┐
+ [90m1. [0m│     0 │ 0     │ (0,'0') │            ᴺᵁᴸᴸ │
+ [90m2. [0m│     1 │ 1     │ (1,'1') │               1 │
+ [90m3. [0m│     2 │ 2     │ (2,'2') │               2 │
+    ─       ─       ─         ─                 ─
+ [90m8. [0m│     7 │ 7     │ (7,'7') │               1 │
+ [90m9. [0m│     8 │ 8     │ (8,'8') │               2 │
+[90m10. [0m│     9 │ 9     │ (9,'9') │            ᴺᵁᴸᴸ │
+    └───────┴───────┴─────────┴─────────────────┘
+  Showed 6 out of 10 rows.
    ┏━━━━━━━┳━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
    ┃ hello ┃ world ┃ tuple   ┃ sometimes_nulls ┃
    ┡━━━━━━━╇━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
@@ -187,7 +188,7 @@ Showed first 6.
     ┡━━━━━━━╇━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
  6. │     5 │ 5     │ (5,'5') │               2 │
     └───────┴───────┴─────────┴─────────────────┘
-  Showed first 6.
+  Showed 6 out of 10 rows.
    ┌─hello─┬─world─┬─tuple───┬─sometimes_nulls─┐
 1. │     0 │ 0     │ (0,'0') │            ᴺᵁᴸᴸ │
 2. │     1 │ 1     │ (1,'1') │               1 │
@@ -198,7 +199,7 @@ Showed first 6.
     ┌─hello─┬─world─┬─tuple───┬─sometimes_nulls─┐
  6. │     5 │ 5     │ (5,'5') │               2 │
     └───────┴───────┴─────────┴─────────────────┘
-  Showed first 6.
+  Showed 6 out of 10 rows.
     hello   world   tuple     sometimes_nulls
 
 1.      0   0       (0,'0')              ᴺᵁᴸᴸ 
@@ -210,7 +211,7 @@ Showed first 6.
 
  6.      5   5       (5,'5')                 2 
 
-Showed first 6.
+Showed 6 out of 10 rows.
    +-------+-------+---------+-----------------+
    | [1mhello[0m | [1mworld[0m | [1mtuple  [0m | [1msometimes_nulls[0m |
    +-------+-------+---------+-----------------+
@@ -229,7 +230,7 @@ Showed first 6.
     +-------+-------+---------+-----------------+
  [90m6. [0m|     5 | 5     | (5,'5') |               2 |
     +-------+-------+---------+-----------------+
-  Showed first 6.
+  Showed 6 out of 10 rows.
    +-[1mhello[0m-+-[1mworld[0m-+-[1mtuple[0m---+-[1msometimes_nulls[0m-+
 [90m1. [0m|     0 | 0     | (0,'0') |            NULL |
 [90m2. [0m|     1 | 1     | (1,'1') |               1 |
@@ -240,7 +241,7 @@ Showed first 6.
     +-[1mhello[0m-+-[1mworld[0m-+-[1mtuple[0m---+-[1msometimes_nulls[0m-+
  [90m6. [0m|     5 | 5     | (5,'5') |               2 |
     +-------+-------+---------+-----------------+
-  Showed first 6.
+  Showed 6 out of 10 rows.
     [1mhello[0m   [1mworld[0m   [1mtuple[0m     [1msometimes_nulls[0m
 
 [90m1. [0m     0   0       (0,'0')              NULL 
@@ -252,16 +253,17 @@ Showed first 6.
 
  [90m6. [0m     5   5       (5,'5')                 2 
 
-Showed first 6.
-   +-[1mhello[0m-+-[1mworld[0m-+-[1mtuple[0m---+-[1msometimes_nulls[0m-+
-[90m1. [0m|     0 | 0     | (0,'0') |            NULL |
-[90m2. [0m|     1 | 1     | (1,'1') |               1 |
-[90m3. [0m|     2 | 2     | (2,'2') |               2 |
-[90m4. [0m|     3 | 3     | (3,'3') |            NULL |
-[90m5. [0m|     4 | 4     | (4,'4') |               1 |
-[90m6. [0m|     5 | 5     | (5,'5') |               2 |
-   +-------+-------+---------+-----------------+
-  Showed first 6.
+Showed 6 out of 10 rows.
+    +-[1mhello[0m-+-[1mworld[0m-+-[1mtuple[0m---+-[1msometimes_nulls[0m-+
+ [90m1. [0m|     0 | 0     | (0,'0') |            NULL |
+ [90m2. [0m|     1 | 1     | (1,'1') |               1 |
+ [90m3. [0m|     2 | 2     | (2,'2') |               2 |
+    -       -       -         -                 -
+ [90m8. [0m|     7 | 7     | (7,'7') |               1 |
+ [90m9. [0m|     8 | 8     | (8,'8') |               2 |
+[90m10. [0m|     9 | 9     | (9,'9') |            NULL |
+    +-------+-------+---------+-----------------+
+  Showed 6 out of 10 rows.
    +-------+-------+---------+-----------------+
    | hello | world | tuple   | sometimes_nulls |
    +-------+-------+---------+-----------------+
@@ -280,7 +282,7 @@ Showed first 6.
     +-------+-------+---------+-----------------+
  6. |     5 | 5     | (5,'5') |               2 |
     +-------+-------+---------+-----------------+
-  Showed first 6.
+  Showed 6 out of 10 rows.
    +-hello-+-world-+-tuple---+-sometimes_nulls-+
 1. |     0 | 0     | (0,'0') |            NULL |
 2. |     1 | 1     | (1,'1') |               1 |
@@ -291,7 +293,7 @@ Showed first 6.
     +-hello-+-world-+-tuple---+-sometimes_nulls-+
  6. |     5 | 5     | (5,'5') |               2 |
     +-------+-------+---------+-----------------+
-  Showed first 6.
+  Showed 6 out of 10 rows.
     hello   world   tuple     sometimes_nulls
 
 1.      0   0       (0,'0')              NULL 
@@ -303,4 +305,4 @@ Showed first 6.
 
  6.      5   5       (5,'5')                 2 
 
-Showed first 6.
+Showed 6 out of 10 rows.
diff --git a/tests/queries/0_stateless/03300_pretty_vertical_cut.reference b/tests/queries/0_stateless/03300_pretty_vertical_cut.reference
new file mode 100644
index 000000000000..1c2e9171217b
--- /dev/null
+++ b/tests/queries/0_stateless/03300_pretty_vertical_cut.reference
@@ -0,0 +1,114 @@
+    ┏━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━┓
+    ┃ number ┃ concat('Hello', number) ┃
+    ┡━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━┩
+ 1. │      0 │ Hello0                  │
+    ├────────┼─────────────────────────┤
+ 2. │      1 │ Hello1                  │
+    ├────────┼─────────────────────────┤
+ 3. │      2 │ Hello2                  │
+    ├────────┼─────────────────────────┤
+ 4. │      3 │ Hello3                  │
+    ├────────┼─────────────────────────┤
+ 5. │      4 │ Hello4                  │
+    ├────────┼─────────────────────────┤
+    ─        ─                         ─
+    ├────────┼─────────────────────────┤
+21. │     20 │ Hello20                 │
+    ├────────┼─────────────────────────┤
+22. │     21 │ Hello21                 │
+    ├────────┼─────────────────────────┤
+23. │     22 │ Hello22                 │
+    ├────────┼─────────────────────────┤
+24. │     23 │ Hello23                 │
+    ├────────┼─────────────────────────┤
+25. │     24 │ Hello24                 │
+    └────────┴─────────────────────────┘
+  Showed 10 out of 25 rows.
+    ┌─number─┬─concat('Hello', number)─┐
+ 1. │      0 │ Hello0                  │
+ 2. │      1 │ Hello1                  │
+ 3. │      2 │ Hello2                  │
+ 4. │      3 │ Hello3                  │
+ 5. │      4 │ Hello4                  │
+    ─        ─                         ─
+21. │     20 │ Hello20                 │
+22. │     21 │ Hello21                 │
+23. │     22 │ Hello22                 │
+24. │     23 │ Hello23                 │
+25. │     24 │ Hello24                 │
+    └────────┴─────────────────────────┘
+  Showed 10 out of 25 rows.
+     number   concat('Hello', number)
+
+ 1.       0   Hello0                  
+ 2.       1   Hello1                  
+ 3.       2   Hello2                  
+ 4.       3   Hello3                  
+ 5.       4   Hello4                  
+ - - - - - - - - - - - - - - - - - - 
+21.      20   Hello20                 
+22.      21   Hello21                 
+23.      22   Hello22                 
+24.      23   Hello23                 
+25.      24   Hello24                 
+
+Showed 10 out of 25 rows.
+    ┏━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━┓
+    ┃ number ┃ concat('Hello', number) ┃
+    ┡━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━┩
+ 1. │      0 │ Hello0                  │
+    ├────────┼─────────────────────────┤
+ 2. │      1 │ Hello1                  │
+    ├────────┼─────────────────────────┤
+ 3. │      2 │ Hello2                  │
+    ├────────┼─────────────────────────┤
+ 4. │      3 │ Hello3                  │
+    ├────────┼─────────────────────────┤
+ 5. │      4 │ Hello4                  │
+    ├────────┼─────────────────────────┤
+ 6. │      5 │ Hello5                  │
+    ├────────┼─────────────────────────┤
+    ─        ─                         ─
+    ├────────┼─────────────────────────┤
+21. │     20 │ Hello20                 │
+    ├────────┼─────────────────────────┤
+22. │     21 │ Hello21                 │
+    ├────────┼─────────────────────────┤
+23. │     22 │ Hello22                 │
+    ├────────┼─────────────────────────┤
+24. │     23 │ Hello23                 │
+    ├────────┼─────────────────────────┤
+25. │     24 │ Hello24                 │
+    └────────┴─────────────────────────┘
+  Showed 11 out of 25 rows.
+    ┌─number─┬─concat('Hello', number)─┐
+ 1. │      0 │ Hello0                  │
+ 2. │      1 │ Hello1                  │
+ 3. │      2 │ Hello2                  │
+ 4. │      3 │ Hello3                  │
+ 5. │      4 │ Hello4                  │
+ 6. │      5 │ Hello5                  │
+    ─        ─                         ─
+21. │     20 │ Hello20                 │
+22. │     21 │ Hello21                 │
+23. │     22 │ Hello22                 │
+24. │     23 │ Hello23                 │
+25. │     24 │ Hello24                 │
+    └────────┴─────────────────────────┘
+  Showed 11 out of 25 rows.
+     number   concat('Hello', number)
+
+ 1.       0   Hello0                  
+ 2.       1   Hello1                  
+ 3.       2   Hello2                  
+ 4.       3   Hello3                  
+ 5.       4   Hello4                  
+ 6.       5   Hello5                  
+ - - - - - - - - - - - - - - - - - - 
+21.      20   Hello20                 
+22.      21   Hello21                 
+23.      22   Hello22                 
+24.      23   Hello23                 
+25.      24   Hello24                 
+
+Showed 11 out of 25 rows.
diff --git a/tests/queries/0_stateless/03300_pretty_vertical_cut.sql b/tests/queries/0_stateless/03300_pretty_vertical_cut.sql
new file mode 100644
index 000000000000..8fc4f72a68b3
--- /dev/null
+++ b/tests/queries/0_stateless/03300_pretty_vertical_cut.sql
@@ -0,0 +1,11 @@
+SET output_format_pretty_max_rows = 10;
+
+SELECT number, 'Hello'||number FROM numbers(25) FORMAT Pretty;
+SELECT number, 'Hello'||number FROM numbers(25) FORMAT PrettyCompact;
+SELECT number, 'Hello'||number FROM numbers(25) FORMAT PrettySpace;
+
+SET output_format_pretty_max_rows = 11;
+
+SELECT number, 'Hello'||number FROM numbers(25) FORMAT Pretty;
+SELECT number, 'Hello'||number FROM numbers(25) FORMAT PrettyCompact;
+SELECT number, 'Hello'||number FROM numbers(25) FORMAT PrettySpace;
