diff --git a/src/Columns/ColumnUnique.h b/src/Columns/ColumnUnique.h
index bfa80b5e3b21..72904c5ab8fc 100644
--- a/src/Columns/ColumnUnique.h
+++ b/src/Columns/ColumnUnique.h
@@ -301,7 +301,7 @@ size_t ColumnUnique<ColumnType>::getNullValueIndex() const
 template <typename ColumnType>
 size_t ColumnUnique<ColumnType>::uniqueInsert(const Field & x)
 {
-    if (x.getType() == Field::Types::Null)
+    if (x.isNull())
         return getNullValueIndex();
 
     if (valuesHaveFixedSize())
