diff --git a/tests/queries/0_stateless/01344_alter_enum_partition_key.reference b/tests/queries/0_stateless/01344_alter_enum_partition_key.reference
new file mode 100644
index 000000000000..bf8a7949282f
--- /dev/null
+++ b/tests/queries/0_stateless/01344_alter_enum_partition_key.reference
@@ -0,0 +1,12 @@
+hello	test
+1_1_1_0	hello	1
+hello	test
+goodbye	test
+3_2_2_1	goodbye	3
+1_1_1_1	hello	1
+1	test
+3	test
+111	abc
+1_1_1_2	1	1
+111_3_3_1	111	111
+3_2_2_2	3	3
diff --git a/tests/queries/0_stateless/01344_alter_enum_partition_key.sql b/tests/queries/0_stateless/01344_alter_enum_partition_key.sql
new file mode 100644
index 000000000000..14bc3b5cdaaa
--- /dev/null
+++ b/tests/queries/0_stateless/01344_alter_enum_partition_key.sql
@@ -0,0 +1,36 @@
+DROP TABLE IF EXISTS test;
+CREATE TABLE test (x Enum('hello' = 1, 'world' = 2), y String) ENGINE = MergeTree PARTITION BY x ORDER BY y;
+INSERT INTO test VALUES ('hello', 'test');
+
+SELECT * FROM test;
+SELECT name, partition, partition_id FROM system.parts WHERE database = currentDatabase() AND table = 'test' AND active ORDER BY partition;
+
+ALTER TABLE test MODIFY COLUMN x Enum('hello' = 1, 'world' = 2, 'goodbye' = 3);
+INSERT INTO test VALUES ('goodbye', 'test');
+OPTIMIZE TABLE test FINAL;
+SELECT * FROM test ORDER BY x;
+SELECT name, partition, partition_id FROM system.parts WHERE database = currentDatabase() AND table = 'test' AND active ORDER BY partition;
+
+ALTER TABLE test MODIFY COLUMN x Enum('hello' = 1, 'world' = 2); -- { serverError 524 }
+ALTER TABLE test MODIFY COLUMN x Enum('hello' = 1, 'world' = 2, 'test' = 3); -- { serverError 524 }
+ALTER TABLE test MODIFY COLUMN x Enum('hello' = 1, 'world' = 2, 'goodbye' = 4); -- { serverError 524 }
+
+ALTER TABLE test MODIFY COLUMN x Int8;
+INSERT INTO test VALUES (111, 'abc');
+OPTIMIZE TABLE test FINAL;
+SELECT * FROM test ORDER BY x;
+SELECT name, partition, partition_id FROM system.parts WHERE database = currentDatabase() AND table = 'test' AND active ORDER BY partition;
+
+ALTER TABLE test MODIFY COLUMN x Enum8('' = 1); -- { serverError 524 }
+ALTER TABLE test MODIFY COLUMN x Enum16('' = 1); -- { serverError 524 }
+
+ALTER TABLE test MODIFY COLUMN x UInt64; -- { serverError 524 }
+ALTER TABLE test MODIFY COLUMN x String; -- { serverError 524 }
+ALTER TABLE test MODIFY COLUMN x Nullable(Int64); -- { serverError 524 }
+
+ALTER TABLE test RENAME COLUMN x TO z; -- { serverError 47 }
+ALTER TABLE test RENAME COLUMN y TO z; -- { serverError 47 }
+ALTER TABLE test DROP COLUMN x; -- { serverError 47 }
+ALTER TABLE test DROP COLUMN y; -- { serverError 47 }
+
+DROP TABLE test;
diff --git a/tests/queries/0_stateless/01346_alter_enum_partition_key_replicated.reference b/tests/queries/0_stateless/01346_alter_enum_partition_key_replicated.reference
new file mode 100644
index 000000000000..60c6076aef0c
--- /dev/null
+++ b/tests/queries/0_stateless/01346_alter_enum_partition_key_replicated.reference
@@ -0,0 +1,24 @@
+hello	test
+hello	test
+1_0_0_0	hello	1
+1_0_0_0	hello	1
+hello	test
+goodbye	test
+hello	test
+goodbye	test
+3_0_0_1	goodbye	3
+1_0_0_1	hello	1
+3_0_0_1	goodbye	3
+1_0_0_1	hello	1
+1	test
+3	test
+111	abc
+1	test
+3	test
+111	abc
+1_0_0_2	1	1
+111_0_0_1	111	111
+3_0_0_2	3	3
+1_0_0_2	1	1
+111_0_0_1	111	111
+3_0_0_2	3	3
diff --git a/tests/queries/0_stateless/01346_alter_enum_partition_key_replicated.sql b/tests/queries/0_stateless/01346_alter_enum_partition_key_replicated.sql
new file mode 100644
index 000000000000..8929fb79d174
--- /dev/null
+++ b/tests/queries/0_stateless/01346_alter_enum_partition_key_replicated.sql
@@ -0,0 +1,48 @@
+DROP TABLE IF EXISTS test;
+DROP TABLE IF EXISTS test2;
+
+CREATE TABLE test (x Enum('hello' = 1, 'world' = 2), y String) ENGINE = ReplicatedMergeTree('/clickhouse/test/table', 'r1') PARTITION BY x ORDER BY y;
+CREATE TABLE test2 (x Enum('hello' = 1, 'world' = 2), y String) ENGINE = ReplicatedMergeTree('/clickhouse/test/table', 'r2') PARTITION BY x ORDER BY y;
+INSERT INTO test VALUES ('hello', 'test');
+
+SELECT * FROM test;
+SYSTEM SYNC REPLICA test2;
+SELECT * FROM test2;
+SELECT name, partition, partition_id FROM system.parts WHERE database = currentDatabase() AND table = 'test' AND active ORDER BY partition;
+SELECT name, partition, partition_id FROM system.parts WHERE database = currentDatabase() AND table = 'test2' AND active ORDER BY partition;
+
+ALTER TABLE test MODIFY COLUMN x Enum('hello' = 1, 'world' = 2, 'goodbye' = 3);
+INSERT INTO test VALUES ('goodbye', 'test');
+OPTIMIZE TABLE test FINAL;
+SELECT * FROM test ORDER BY x;
+SYSTEM SYNC REPLICA test2;
+SELECT * FROM test2 ORDER BY x;
+SELECT name, partition, partition_id FROM system.parts WHERE database = currentDatabase() AND table = 'test' AND active ORDER BY partition;
+SELECT name, partition, partition_id FROM system.parts WHERE database = currentDatabase() AND table = 'test2' AND active ORDER BY partition;
+
+ALTER TABLE test MODIFY COLUMN x Enum('hello' = 1, 'world' = 2); -- { serverError 524 }
+ALTER TABLE test MODIFY COLUMN x Enum('hello' = 1, 'world' = 2, 'test' = 3); -- { serverError 524 }
+ALTER TABLE test MODIFY COLUMN x Enum('hello' = 1, 'world' = 2, 'goodbye' = 4); -- { serverError 524 }
+
+ALTER TABLE test MODIFY COLUMN x Int8;
+INSERT INTO test VALUES (111, 'abc');
+OPTIMIZE TABLE test FINAL;
+SELECT * FROM test ORDER BY x;
+SYSTEM SYNC REPLICA test2;
+SELECT * FROM test2 ORDER BY x;
+SELECT name, partition, partition_id FROM system.parts WHERE database = currentDatabase() AND table = 'test' AND active ORDER BY partition;
+SELECT name, partition, partition_id FROM system.parts WHERE database = currentDatabase() AND table = 'test2' AND active ORDER BY partition;
+
+ALTER TABLE test MODIFY COLUMN x Enum8('' = 1); -- { serverError 524 }
+ALTER TABLE test MODIFY COLUMN x Enum16('' = 1); -- { serverError 524 }
+
+ALTER TABLE test MODIFY COLUMN x UInt64; -- { serverError 524 }
+ALTER TABLE test MODIFY COLUMN x String; -- { serverError 524 }
+ALTER TABLE test MODIFY COLUMN x Nullable(Int64); -- { serverError 524 }
+
+ALTER TABLE test RENAME COLUMN x TO z; -- { serverError 47 }
+ALTER TABLE test RENAME COLUMN y TO z; -- { serverError 47 }
+ALTER TABLE test DROP COLUMN x; -- { serverError 47 }
+ALTER TABLE test DROP COLUMN y; -- { serverError 47 }
+
+DROP TABLE test;
