diff --git a/tests/queries/0_stateless/03296_bfloat16_ubsan.reference b/tests/queries/0_stateless/03296_bfloat16_ubsan.reference
new file mode 100644
index 000000000000..ff4401616211
--- /dev/null
+++ b/tests/queries/0_stateless/03296_bfloat16_ubsan.reference
@@ -0,0 +1,1 @@
+-256
diff --git a/tests/queries/0_stateless/03296_bfloat16_ubsan.sql b/tests/queries/0_stateless/03296_bfloat16_ubsan.sql
new file mode 100644
index 000000000000..d1cac3d8ea8e
--- /dev/null
+++ b/tests/queries/0_stateless/03296_bfloat16_ubsan.sql
@@ -0,0 +1,2 @@
+SET allow_experimental_bfloat16_type = 1;
+SELECT (65535::BFloat16)::Int16; -- The result is implementation defined on overflow.
