diff --git a/tools/rpkg/DESCRIPTION b/tools/rpkg/DESCRIPTION
index 074201739c7e..20e1ce6d4005 100644
--- a/tools/rpkg/DESCRIPTION
+++ b/tools/rpkg/DESCRIPTION
@@ -26,12 +26,12 @@ Depends:
 Imports:
     dbplyr,
     methods,
-    testthat,
     utils
 Suggests: 
     callr,
     DBItest,
-    nycflights13
+    nycflights13,
+    testthat
 Encoding: UTF-8
 LazyData: true
 RoxygenNote: 7.1.1
diff --git a/tools/rpkg/R/Driver.R b/tools/rpkg/R/Driver.R
index 981397ad6409..b6e80cfb267f 100644
--- a/tools/rpkg/R/Driver.R
+++ b/tools/rpkg/R/Driver.R
@@ -2,7 +2,7 @@ DBDIR_MEMORY <- ":memory:"
 
 #' @title DuckDB Driver
 #'
-#' @description A DuckDB database instance. 
+#' @description A DuckDB database instance.
 #'
 #' @param dbdir The file in which the DuckDB database should be stored
 #' @param read_only Whether the database file should be opened in read-only mode
@@ -12,7 +12,6 @@ DBDIR_MEMORY <- ":memory:"
 #' @export
 #' @examples
 #' \dontrun{
-#' #' library(DBI)
 #' duckdb::duckdb()
 #' }
 #'
diff --git a/tools/rpkg/man/duckdb_driver.Rd b/tools/rpkg/man/duckdb_driver.Rd
index 7ae1043091b3..c06eebe926c5 100644
--- a/tools/rpkg/man/duckdb_driver.Rd
+++ b/tools/rpkg/man/duckdb_driver.Rd
@@ -67,7 +67,6 @@ A DuckDB database instance.
 }
 \examples{
 \dontrun{
-#' library(DBI)
 duckdb::duckdb()
 }
 
