diff --git a/tests/queries/0_stateless/01821_dictionary_primary_key_wrong_order.reference b/tests/queries/0_stateless/01821_dictionary_primary_key_wrong_order.reference
new file mode 100644
index 000000000000..9833cbcc9b6b
--- /dev/null
+++ b/tests/queries/0_stateless/01821_dictionary_primary_key_wrong_order.reference
@@ -0,0 +1,1 @@
+1	20
diff --git a/tests/queries/0_stateless/01821_dictionary_primary_key_wrong_order.sql b/tests/queries/0_stateless/01821_dictionary_primary_key_wrong_order.sql
new file mode 100644
index 000000000000..636426fcc915
--- /dev/null
+++ b/tests/queries/0_stateless/01821_dictionary_primary_key_wrong_order.sql
@@ -0,0 +1,24 @@
+DROP TABLE IF EXISTS dictionary_primary_key_source_table;
+CREATE TABLE dictionary_primary_key_source_table
+(
+    identifier UInt64,
+    v UInt64
+) ENGINE = TinyLog;
+
+INSERT INTO dictionary_primary_key_source_table VALUES (20, 1);
+
+DROP DICTIONARY IF EXISTS flat_dictionary;
+CREATE DICTIONARY flat_dictionary
+(
+    identifier UInt64,
+    v UInt64
+)
+PRIMARY KEY v
+SOURCE(CLICKHOUSE(HOST 'localhost' PORT tcpPort() TABLE 'dictionary_primary_key_source_table'))
+LIFETIME(MIN 1 MAX 1000)
+LAYOUT(FLAT());
+
+SELECT * FROM flat_dictionary;
+
+DROP DICTIONARY flat_dictionary;
+DROP TABLE dictionary_primary_key_source_table;
