diff --git a/tests/queries/0_stateless/01436_storage_merge_with_join_push_down.reference b/tests/queries/0_stateless/01436_storage_merge_with_join_push_down.reference
new file mode 100644
index 000000000000..aa47d0d46d47
--- /dev/null
+++ b/tests/queries/0_stateless/01436_storage_merge_with_join_push_down.reference
@@ -0,0 +1,2 @@
+0
+0
diff --git a/tests/queries/0_stateless/01436_storage_merge_with_join_push_down.sql b/tests/queries/0_stateless/01436_storage_merge_with_join_push_down.sql
new file mode 100644
index 000000000000..a3c598c6d837
--- /dev/null
+++ b/tests/queries/0_stateless/01436_storage_merge_with_join_push_down.sql
@@ -0,0 +1,30 @@
+DROP TABLE IF EXISTS test1;
+DROP TABLE IF EXISTS test1_distributed;
+DROP TABLE IF EXISTS test_merge;
+
+SET enable_optimize_predicate_expression = 1;
+
+CREATE TABLE test1 (id Int64, name String) ENGINE MergeTree PARTITION BY (id) ORDER BY (id);
+CREATE TABLE test1_distributed AS test1 ENGINE = Distributed(test_cluster_two_shards_localhost, default, test1);
+CREATE TABLE test_merge AS test1 ENGINE = Merge('default', 'test1_distributed');
+
+SELECT count() FROM test_merge
+JOIN (SELECT 'anystring' AS name) AS n
+USING name
+WHERE id = 1;
+
+DROP TABLE test1;
+DROP TABLE test_merge;
+
+
+CREATE TABLE test1 (id Int64, name String) ENGINE MergeTree PARTITION BY (id) ORDER BY (id);
+CREATE TABLE test_merge AS test1 ENGINE = Merge('default', 'test1');
+
+SELECT count() FROM test_merge
+JOIN (SELECT 'anystring' AS name) AS n
+USING name
+WHERE id = 1;
+
+DROP TABLE test1;
+DROP TABLE test_merge;
+DROP TABLE test1_distributed;
