diff --git a/tests/queries/0_stateless/01060_avro.reference b/tests/queries/0_stateless/01060_avro.reference
index a21e7a3a1012..0550967a224b 100644
--- a/tests/queries/0_stateless/01060_avro.reference
+++ b/tests/queries/0_stateless/01060_avro.reference
@@ -19,6 +19,10 @@
 = references
 "a1","c1"
 "a2","c2"
+= nested
+1,"b1",2.2,2.3,"c3"
+2.3,"b1",1,"c3"
+not found
 = compression
 1000
 1000
diff --git a/tests/queries/0_stateless/01060_avro.sh b/tests/queries/0_stateless/01060_avro.sh
index 15e97abfa520..a64b28847315 100755
--- a/tests/queries/0_stateless/01060_avro.sh
+++ b/tests/queries/0_stateless/01060_avro.sh
@@ -27,6 +27,11 @@ cat $DATA_DIR/logical_types.avro | ${CLICKHOUSE_LOCAL} --input-format Avro --out
 echo = references
 cat $DATA_DIR/references.avro | ${CLICKHOUSE_LOCAL} --input-format Avro --output-format CSV -S "a String, c String" -q 'select * from table'
 
+echo = nested
+cat $DATA_DIR/nested.avro | ${CLICKHOUSE_LOCAL} --input-format Avro --output-format CSV -S 'a Int64, "b.a" String, "b.b" Double, "b.c" Double, c String' -q 'select * from table'
+cat $DATA_DIR/nested.avro | ${CLICKHOUSE_LOCAL} --input-format Avro --output-format CSV -S '"b.c" Double, "b.a" String, a Int64, c String' -q 'select * from table'
+cat $DATA_DIR/nested.avro | ${CLICKHOUSE_LOCAL} --input-format Avro --output-format CSV -S '"b" Double' -q 'select * from table' 2>&1 | grep -i 'not found' -o
+
 echo = compression
 cat $DATA_DIR/simple.null.avro | ${CLICKHOUSE_LOCAL} --input-format Avro --output-format CSV -S 'a Int64' -q 'select count() from table'
 cat $DATA_DIR/simple.deflate.avro | ${CLICKHOUSE_LOCAL} --input-format Avro --output-format CSV -S 'a Int64' -q 'select count() from table'
@@ -68,4 +73,4 @@ ${CLICKHOUSE_LOCAL} -q "select toInt64(number) as a from numbers(0)  format Avro
 ${CLICKHOUSE_LOCAL} -q "select toInt64(number) as a from numbers(1000)  format Avro" | ${CLICKHOUSE_LOCAL} --input-format Avro --output-format CSV -S "$S4" -q 'select count() from table'
 
 # type supported via conversion
-${CLICKHOUSE_LOCAL}  -q "select toInt16(123) as a format Avro" | wc -c
\ No newline at end of file
+${CLICKHOUSE_LOCAL}  -q "select toInt16(123) as a format Avro" | wc -c | tr -d ' '
\ No newline at end of file
diff --git a/tests/queries/0_stateless/data_avro/generate_avro.sh b/tests/queries/0_stateless/data_avro/generate_avro.sh
index 6ec26efc0492..b6ec75ad4dd0 100755
--- a/tests/queries/0_stateless/data_avro/generate_avro.sh
+++ b/tests/queries/0_stateless/data_avro/generate_avro.sh
@@ -8,6 +8,7 @@ avro-tools fromjson  --schema-file complex.avsc complex.json > complex.avro
 avro-tools fromjson  --schema-file logical_types.avsc logical_types.json > logical_types.avro
 avro-tools fromjson  --schema-file empty.avsc empty.json > empty.avro
 avro-tools fromjson  --schema-file references.avsc references.json >  references.avro
+avro-tools fromjson  --schema-file nested.avsc nested.json > nested.avro
 
 #compression
 avro-tools fromjson --codec null  --schema-file simple.avsc simple.json > simple.null.avro
diff --git a/tests/queries/0_stateless/data_avro/nested.avro b/tests/queries/0_stateless/data_avro/nested.avro
new file mode 100644
index 000000000000..1415c45d3282
Binary files /dev/null and b/tests/queries/0_stateless/data_avro/nested.avro differ
diff --git a/tests/queries/0_stateless/data_avro/nested.avsc b/tests/queries/0_stateless/data_avro/nested.avsc
new file mode 100644
index 000000000000..966dc6defb3e
--- /dev/null
+++ b/tests/queries/0_stateless/data_avro/nested.avsc
@@ -0,0 +1,17 @@
+{
+    "type": "record",
+    "name": "main",
+    "fields": [
+        {"name": "a", "type": "long"},
+        {"name": "b", "type": {
+            "type": "record",
+            "name": "sub1",
+            "fields": [
+                {"name": "a", "type": "string"},
+                {"name": "b", "type": "double"},
+                {"name": "c", "type": "double"}
+            ]
+        }},
+        {"name": "c", "type": "string"}
+    ]
+  }
\ No newline at end of file
diff --git a/tests/queries/0_stateless/data_avro/nested.json b/tests/queries/0_stateless/data_avro/nested.json
new file mode 100644
index 000000000000..63a7bc40e4bd
--- /dev/null
+++ b/tests/queries/0_stateless/data_avro/nested.json
@@ -0,0 +1,1 @@
+{"a":1, "b": {"a":"b1", "b": 2.2, "c": 2.3}, "c": "c3"}
