diff --git a/test/issues/general/test_14232.test b/test/issues/general/test_14232.test
index 2c6664ccbfb5..35fc46de7dbf 100644
--- a/test/issues/general/test_14232.test
+++ b/test/issues/general/test_14232.test
@@ -59,7 +59,7 @@ INSERT INTO t2 (t2a, t2b, t2c, t2d, t2e, t2f, t2g, t2h, t2i) VALUES
 ('t1f', 19, NULL, 19, 17.0, 25, 26.00, '2014-10-04 01:01:00', '2014-10-04'),
 ('t1b', NULL, 16, 19, 17.0, 25, 26.00, '2014-05-04 01:01:00', NULL)
 
-query II
+query II rowsort
 SELECT t1a,
 t1b
 FROM t1
diff --git a/test/issues/general/test_14286.test b/test/issues/general/test_14286.test
new file mode 100644
index 000000000000..1cf3095789b1
--- /dev/null
+++ b/test/issues/general/test_14286.test
@@ -0,0 +1,16 @@
+# name: test/issues/general/test_14286.test
+# description: Issue 14286 - An error was encountered using both distinct and struct
+# group: [general]
+
+statement ok
+select
+    category,
+    array_agg(distinct name) filter(where id != 5)            as a_list,
+    array_agg(name)          filter(where id != 5)            as b_list,
+    array_agg({'id': id, 'name': name, 'catetory': category}) as c_list
+from (
+    select 1 as id, '大熊猫' as name, '熊' as category union all
+    select 2 as id, '大熊猫' as name, '猫' as category union all
+    select 3 as id, '小熊猫' as name, '猫' as category
+) t
+group by category;
