diff --git a/tests/queries/0_stateless/01906_lc_in_bug.reference b/tests/queries/0_stateless/01906_lc_in_bug.reference
new file mode 100644
index 000000000000..9fe1650abf0f
--- /dev/null
+++ b/tests/queries/0_stateless/01906_lc_in_bug.reference
@@ -0,0 +1,2 @@
+1	0
+3	1
diff --git a/tests/queries/0_stateless/01906_lc_in_bug.sql b/tests/queries/0_stateless/01906_lc_in_bug.sql
new file mode 100644
index 000000000000..f8f41da31ae8
--- /dev/null
+++ b/tests/queries/0_stateless/01906_lc_in_bug.sql
@@ -0,0 +1,8 @@
+drop table if exists tab;
+create table tab (x LowCardinality(String)) engine = MergeTree order by tuple();
+
+insert into tab values ('a'), ('bb'), ('a'), ('cc');
+
+select count() as c, x in ('a', 'bb') as g from tab group by g order by c;
+
+drop table if exists tab;
