diff --git a/tests/queries/0_stateless/01861_explain_pipeline.reference b/tests/queries/0_stateless/01861_explain_pipeline.reference
index 427b3eaefc08..6874d5f2f3d7 100644
--- a/tests/queries/0_stateless/01861_explain_pipeline.reference
+++ b/tests/queries/0_stateless/01861_explain_pipeline.reference
@@ -18,11 +18,11 @@ ExpressionTransform × 2
   ExpressionTransform × 2
     ReplacingSorted
       FilterSortedStreamByRange
-      Description: filter values in [(5), +inf)
+      Description: filter values in ((5), +inf]
         ExpressionTransform
           MergeTreeInOrder 0 → 1
             ReplacingSorted 2 → 1
               FilterSortedStreamByRange × 2
-              Description: filter values in [-inf, (5))
+              Description: filter values in (-inf, (5)]
                 ExpressionTransform × 2
                   MergeTreeInOrder × 2 0 → 1
diff --git a/tests/queries/0_stateless/02780_final_streams_data_skipping_index.reference b/tests/queries/0_stateless/02780_final_streams_data_skipping_index.reference
index 5242c6253250..e7866cbc75d9 100644
--- a/tests/queries/0_stateless/02780_final_streams_data_skipping_index.reference
+++ b/tests/queries/0_stateless/02780_final_streams_data_skipping_index.reference
@@ -10,12 +10,12 @@ ExpressionTransform × 2
     ExpressionTransform × 2
       AggregatingSortedTransform 2 → 1
         FilterSortedStreamByRange × 2
-        Description: filter values in [(999424), +inf)
+        Description: filter values in ((999424), +inf]
           ExpressionTransform × 2
             MergeTreeInOrder × 2 0 → 1
               AggregatingSortedTransform
                 FilterSortedStreamByRange
-                Description: filter values in [-inf, (999424))
+                Description: filter values in (-inf, (999424)]
                   ExpressionTransform
                     MergeTreeInOrder 0 → 1
 EXPLAIN PIPELINE SELECT * FROM data FINAL WHERE v1 >= now() - INTERVAL 180 DAY
@@ -29,11 +29,11 @@ ExpressionTransform × 2
     ExpressionTransform × 2
       AggregatingSortedTransform 2 → 1
         FilterSortedStreamByRange × 2
-        Description: filter values in [(999424), +inf)
+        Description: filter values in ((999424), +inf]
           ExpressionTransform × 2
             MergeTreeInOrder × 2 0 → 1
               AggregatingSortedTransform
                 FilterSortedStreamByRange
-                Description: filter values in [-inf, (999424))
+                Description: filter values in (-inf, (999424)]
                   ExpressionTransform
                     MergeTreeInOrder 0 → 1
diff --git a/tests/queries/0_stateless/02867_nullable_primary_key_final.reference b/tests/queries/0_stateless/02867_nullable_primary_key_final.reference
new file mode 100644
index 000000000000..2e55b120f6eb
--- /dev/null
+++ b/tests/queries/0_stateless/02867_nullable_primary_key_final.reference
@@ -0,0 +1,2 @@
+2023-09-01	2500000000
+2023-09-01	166167
diff --git a/tests/queries/0_stateless/02867_nullable_primary_key_final.sql b/tests/queries/0_stateless/02867_nullable_primary_key_final.sql
new file mode 100644
index 000000000000..05677789459f
--- /dev/null
+++ b/tests/queries/0_stateless/02867_nullable_primary_key_final.sql
@@ -0,0 +1,57 @@
+DROP TABLE IF EXISTS t;
+
+CREATE TABLE t
+(
+    `d` Nullable(Date),
+    `f1` Nullable(String),
+    `f2` Nullable(String),
+    `c` Nullable(Int64)
+)
+ENGINE = ReplacingMergeTree
+ORDER BY (f1, f2, d)
+SETTINGS allow_nullable_key = 1;
+
+INSERT INTO t SELECT
+    toDate('2023-09-10', 'UTC') AS d,
+    [number % 99999, NULL][number % 2] AS f1,
+    ['x', NULL][number % 2] AS f2,
+    [number, NULL][number % 2] AS c
+FROM numbers(100000);
+
+SELECT
+    date_trunc('month', d),
+    SUM(c)
+FROM t
+FINAL
+WHERE f2 = 'x'
+GROUP BY 1;
+
+DROP TABLE t;
+
+CREATE TABLE t
+(
+    `d` Nullable(Date),
+    `f1` Nullable(String),
+    `f2` Nullable(String),
+    `c` Nullable(Int64)
+)
+ENGINE = SummingMergeTree
+ORDER BY (f1, f2, d)
+SETTINGS allow_nullable_key = 1, index_granularity = 1;
+
+INSERT INTO t SELECT
+    toDate('2023-09-10', 'UTC') AS d,
+    NULL AS f1,
+    ['x', 'y', 'z'][number % 3] AS f2,
+    number AS c
+FROM numbers(1000);
+
+SELECT
+    date_trunc('month', d),
+    SUM(c)
+FROM t
+FINAL
+WHERE f2 = 'x'
+GROUP BY 1;
+
+DROP TABLE t;
