diff --git a/src/common/arrow_wrapper.cpp b/src/common/arrow_wrapper.cpp
index 1cb6da5f0ccb..0fc0eedbca64 100644
--- a/src/common/arrow_wrapper.cpp
+++ b/src/common/arrow_wrapper.cpp
@@ -101,8 +101,9 @@ int ResultArrowArrayStreamWrapper::MyStreamGetNext(struct ArrowArrayStream *stre
 	if (result.type == QueryResultType::STREAM_RESULT) {
 		auto &stream_result = (StreamQueryResult &)result;
 		if (!stream_result.IsOpen()) {
-			my_stream->last_error = "Query Stream is closed";
-			return -1;
+			// Nothing to output
+			out->release = nullptr;
+			return 0;
 		}
 	}
 	unique_ptr<DataChunk> chunk_result = result.Fetch();
