--- a/src/function/scalar/date/date_diff.cpp
+++ b/src/function/scalar/date/date_diff.cpp
@@ -51,4 +51,4 @@ struct DayOperator {
 		template <class TA, class TB, class TR>
 		static inline TR Operation(TA startdate, TB enddate) {
-			return Date::EpochDays(enddate) - Date::EpochDays(startdate);
+			return static_cast<TR>(static_cast<int64_t>(Date::EpochDays(enddate)) - static_cast<int64_t>(Date::EpochDays(startdate)));
 		}
 	};
 
