diff --git a/tests/queries/0_stateless/01710_minmax_count_projection_modify_partition_key.reference b/tests/queries/0_stateless/01710_minmax_count_projection_modify_partition_key.reference
new file mode 100644
index 000000000000..60f9fb089181
--- /dev/null
+++ b/tests/queries/0_stateless/01710_minmax_count_projection_modify_partition_key.reference
@@ -0,0 +1,3 @@
+x	1
+x	1
+y	1
diff --git a/tests/queries/0_stateless/01710_minmax_count_projection_modify_partition_key.sql b/tests/queries/0_stateless/01710_minmax_count_projection_modify_partition_key.sql
new file mode 100644
index 000000000000..5297fb3e194f
--- /dev/null
+++ b/tests/queries/0_stateless/01710_minmax_count_projection_modify_partition_key.sql
@@ -0,0 +1,13 @@
+DROP TABLE IF EXISTS test;
+
+CREATE TABLE test (type Enum('x'), s String) ENGINE = MergeTree ORDER BY s PARTITION BY type;
+INSERT INTO test VALUES ('x', 'Hello');
+
+SELECT type, count() FROM test GROUP BY type ORDER BY type;
+
+ALTER TABLE test MODIFY COLUMN type Enum('x', 'y');
+INSERT INTO test VALUES ('y', 'World');
+
+SELECT type, count() FROM test GROUP BY type ORDER BY type;
+
+DROP TABLE test;
