diff --git a/tests/queries/0_stateless/01296_codecs_bad_arguments.reference b/tests/queries/0_stateless/01296_codecs_bad_arguments.reference
new file mode 100644
index 000000000000..d00491fd7e5b
--- /dev/null
+++ b/tests/queries/0_stateless/01296_codecs_bad_arguments.reference
@@ -0,0 +1,1 @@
+1
diff --git a/tests/queries/0_stateless/01296_codecs_bad_arguments.sql b/tests/queries/0_stateless/01296_codecs_bad_arguments.sql
new file mode 100644
index 000000000000..d7eb53300ecd
--- /dev/null
+++ b/tests/queries/0_stateless/01296_codecs_bad_arguments.sql
@@ -0,0 +1,15 @@
+DROP TABLE IF EXISTS delta_table;
+DROP TABLE IF EXISTS zstd_table;
+DROP TABLE IF EXISTS lz4_table;
+
+CREATE TABLE delta_table (`id` UInt64 CODEC(Delta(tuple()))) ENGINE = MergeTree() ORDER BY tuple(); --{serverError 433}
+CREATE TABLE zstd_table (`id` UInt64 CODEC(ZSTD(tuple()))) ENGINE = MergeTree() ORDER BY tuple(); --{serverError 433}
+CREATE TABLE lz4_table (`id` UInt64 CODEC(LZ4HC(tuple()))) ENGINE = MergeTree() ORDER BY tuple(); --{serverError 433}
+
+CREATE TABLE lz4_table (`id` UInt64 CODEC(LZ4(tuple()))) ENGINE = MergeTree() ORDER BY tuple(); --{serverError 378}
+
+SELECT 1;
+
+DROP TABLE IF EXISTS delta_table;
+DROP TABLE IF EXISTS zstd_table;
+DROP TABLE IF EXISTS lz4_table;
