diff --git a/src/Storages/StorageMaterializedView.cpp b/src/Storages/StorageMaterializedView.cpp
index 56844192ee9d..49111e02b110 100644
--- a/src/Storages/StorageMaterializedView.cpp
+++ b/src/Storages/StorageMaterializedView.cpp
@@ -412,6 +412,11 @@ StoragePtr StorageMaterializedView::tryGetTargetTable() const
     return DatabaseCatalog::instance().tryGetTable(target_table_id, getContext());
 }
 
+NamesAndTypesList StorageMaterializedView::getVirtuals() const
+{
+    return getTargetTable()->getVirtuals();
+}
+
 Strings StorageMaterializedView::getDataPaths() const
 {
     if (auto table = tryGetTargetTable())
diff --git a/src/Storages/StorageMaterializedView.h b/src/Storages/StorageMaterializedView.h
index c110d0b211ca..395560c1ca73 100644
--- a/src/Storages/StorageMaterializedView.h
+++ b/src/Storages/StorageMaterializedView.h
@@ -71,6 +71,9 @@ class StorageMaterializedView final : public shared_ptr_helper<StorageMaterializ
     StoragePtr getTargetTable() const;
     StoragePtr tryGetTargetTable() const;
 
+    /// Get the virtual column of the target table;
+    NamesAndTypesList getVirtuals() const override;
+
     ActionLock getActionLock(StorageActionBlockType type) override;
 
     Pipe read(
