diff --git a/src/Storages/System/StorageSystemDetachedTables.cpp b/src/Storages/System/StorageSystemDetachedTables.cpp
index a1ed491aff3f..e423aec53f94 100644
--- a/src/Storages/System/StorageSystemDetachedTables.cpp
+++ b/src/Storages/System/StorageSystemDetachedTables.cpp
@@ -93,7 +93,11 @@ class DetachedTablesBlockSource : public ISource
             }
 
              if (rows_count == max_block_size)
+             {
+                if (!detached_tables_it->isValid())
+                    ++database_idx;
                 break;
+             }
         }
 
         if (databases->size() == database_idx)
