diff --git a/contrib/googletest b/contrib/googletest
index 356f2d264a48..e7e591764bab 160000
--- a/contrib/googletest
+++ b/contrib/googletest
@@ -1,1 +1,1 @@
-Subproject commit 356f2d264a485db2fcc50ec1c672e0d37b6cb39b
+Subproject commit e7e591764baba0a0c3c9ad0014430e7a27331d16
diff --git a/src/Compression/tests/gtest_compressionCodec.cpp b/src/Compression/tests/gtest_compressionCodec.cpp
index e9470536ae88..0acd15e48c35 100644
--- a/src/Compression/tests/gtest_compressionCodec.cpp
+++ b/src/Compression/tests/gtest_compressionCodec.cpp
@@ -749,7 +749,7 @@ auto RandomishGenerator = [](auto i)
 {
     using T = decltype(i);
     double sin_value = sin(static_cast<double>(i * i)) * i;
-    if (sin_value < std::numeric_limits<T>::lowest() || sin_value > std::numeric_limits<T>::max())
+    if (sin_value < std::numeric_limits<T>::lowest() || sin_value > static_cast<double>(std::numeric_limits<T>::max()))
         return T{};
     return T(sin_value);
 };
