diff --git a/tests/queries/0_stateless/02735_system_zookeeper_auxiliary.reference b/tests/queries/0_stateless/02735_system_zookeeper_auxiliary.reference
new file mode 100644
index 000000000000..32c67aa6bb56
--- /dev/null
+++ b/tests/queries/0_stateless/02735_system_zookeeper_auxiliary.reference
@@ -0,0 +1,4 @@
+default
+zookeeper2
+1
+1
diff --git a/tests/queries/0_stateless/02735_system_zookeeper_auxiliary.sql b/tests/queries/0_stateless/02735_system_zookeeper_auxiliary.sql
new file mode 100644
index 000000000000..fee2d60162e8
--- /dev/null
+++ b/tests/queries/0_stateless/02735_system_zookeeper_auxiliary.sql
@@ -0,0 +1,20 @@
+-- Tags: no-fasttest, no-replicated-database, no-shared-merge-tree
+-- no-shared-merge-tree -- smt doesn't support aux zookeepers
+
+DROP TABLE IF EXISTS test_system_zookeeper_auxiliary;
+
+CREATE TABLE test_system_zookeeper_auxiliary (
+    key UInt64
+)
+ENGINE ReplicatedMergeTree('zookeeper2:/clickhouse/{database}/02731_test_system_zookeeper_auxiliary/{shard}', '{replica}')
+ORDER BY tuple();
+
+SELECT DISTINCT zookeeperName FROM system.zookeeper WHERE path = '/' AND zookeeperName = 'default';
+SELECT DISTINCT zookeeperName FROM system.zookeeper WHERE path = '/' AND zookeeperName = 'zookeeper2';
+
+SELECT count() FROM system.zookeeper WHERE path IN '/' AND zookeeperName = 'zookeeper3'; -- { serverError BAD_ARGUMENTS }
+
+SELECT count() = 0 FROM system.zookeeper WHERE path IN '/' AND zookeeperName = 'default' AND zookeeperName = 'zookeeper2';
+SELECT count() > 0 FROM system.zookeeper WHERE path IN '/' AND zookeeperName = 'zookeeper2' AND zookeeperName = 'zookeeper2';
+
+DROP TABLE IF EXISTS test_system_zookeeper_auxiliary;
diff --git a/tests/queries/0_stateless/02976_system_zookeeper_filters.reference b/tests/queries/0_stateless/02976_system_zookeeper_filters.reference
index a9e2f17562ae..fcb49fa99454 100644
--- a/tests/queries/0_stateless/02976_system_zookeeper_filters.reference
+++ b/tests/queries/0_stateless/02976_system_zookeeper_filters.reference
@@ -4,3 +4,4 @@
 1
 1
 1
+1
diff --git a/tests/queries/0_stateless/02976_system_zookeeper_filters.sql b/tests/queries/0_stateless/02976_system_zookeeper_filters.sql
index 3664c817da74..508f224b233c 100644
--- a/tests/queries/0_stateless/02976_system_zookeeper_filters.sql
+++ b/tests/queries/0_stateless/02976_system_zookeeper_filters.sql
@@ -15,3 +15,8 @@ SELECT count() > 0 FROM system.zookeeper WHERE name LIKE '%_%';
 SELECT count() > 0 FROM system.zookeeper WHERE value LIKE '%';
 SELECT count() > 0 FROM system.zookeeper WHERE path LIKE '/%';
 SELECT count() > 0 FROM system.zookeeper WHERE path = '/';
+
+SET allow_unrestricted_reads_from_keeper = DEFAULT;
+
+SELECT count() > 0 FROM system.zookeeper WHERE path = '/' AND zookeeperName  = 'unknown'; -- { serverError BAD_ARGUMENTS }
+SELECT count() > 0 FROM system.zookeeper WHERE path = '/' AND zookeeperName  = 'default';
