diff --git a/tests/queries/0_stateless/03311_min_max_count_projection_with_constant_partition_expr.reference b/tests/queries/0_stateless/03311_min_max_count_projection_with_constant_partition_expr.reference
new file mode 100644
index 000000000000..e69de29bb2d1
diff --git a/tests/queries/0_stateless/03311_min_max_count_projection_with_constant_partition_expr.sql b/tests/queries/0_stateless/03311_min_max_count_projection_with_constant_partition_expr.sql
new file mode 100644
index 000000000000..9257a7dbf6bf
--- /dev/null
+++ b/tests/queries/0_stateless/03311_min_max_count_projection_with_constant_partition_expr.sql
@@ -0,0 +1,5 @@
+DROP TABLE IF EXISTS t0;
+
+CREATE TABLE t0 (c0 Int) ENGINE = MergeTree() PARTITION BY (EXISTS (SELECT 1)) ORDER BY tuple();
+
+DROP TABLE t0;
