diff --git a/tests/queries/0_stateless/01948_heredoc.reference b/tests/queries/0_stateless/01948_heredoc.reference
index 2f40275e1d25..7e5b6119661c 100644
--- a/tests/queries/0_stateless/01948_heredoc.reference
+++ b/tests/queries/0_stateless/01948_heredoc.reference
@@ -7,8 +7,5 @@ $do$ $ doc$ $doc $ $doco$
 $do$ $ doc$ $doc $ $doco$	$do$ $ doc$ $doc $ $doco$
 ТЕСТ
 该类型的引擎
-VALUE
-VALUE
 
value1
value2
value3

 \'\\xc3\\x28\'
-\'\\xc3\\x28\'
diff --git a/tests/queries/0_stateless/01948_heredoc.sql b/tests/queries/0_stateless/01948_heredoc.sql
index 4a4ced004e34..9abc0025cf52 100644
--- a/tests/queries/0_stateless/01948_heredoc.sql
+++ b/tests/queries/0_stateless/01948_heredoc.sql
@@ -9,9 +9,6 @@ SELECT $doc$$do$ $ doc$ $doc $ $doco$$doc$, $doc$$do$ $ doc$ $doc $ $doco$$doc$;
 SELECT $doc$ТЕСТ$doc$;
 SELECT $doc$该类型的引擎$doc$;
 
-SELECT $РАЗДЕЛИТЕЛЬ$VALUE$РАЗДЕЛИТЕЛЬ$;
-SELECT $该类型的引擎$VALUE$该类型的引擎$;
-
 SELECT $$
 value1
 value2
@@ -19,4 +16,3 @@ value3
 $$;
 
 SELECT $doc$'\xc3\x28'$doc$;
-SELECT $\xc3\x28$'\xc3\x28'$\xc3\x28$;
diff --git a/tests/queries/0_stateless/03580_heredoc_ambiguity.reference b/tests/queries/0_stateless/03580_heredoc_ambiguity.reference
new file mode 100644
index 000000000000..23d944e59a7a
--- /dev/null
+++ b/tests/queries/0_stateless/03580_heredoc_ambiguity.reference
@@ -0,0 +1,3 @@
+Hello	World
+Hello	World
+Hello	World
diff --git a/tests/queries/0_stateless/03580_heredoc_ambiguity.sql b/tests/queries/0_stateless/03580_heredoc_ambiguity.sql
new file mode 100644
index 000000000000..b18a6770f5fd
--- /dev/null
+++ b/tests/queries/0_stateless/03580_heredoc_ambiguity.sql
@@ -0,0 +1,4 @@
+SET param_$1 = 'Hello', param_$2 = 'World';
+
+SELECT {$1:String} AS x, {$2:String} AS y; -- {$1:String} AS x, {$2:String} AS y
+SELECT {$1:String} AS x, {$2:String} AS y UNION ALL SELECT {$1:String} AS x, {$2:String} AS y;
