diff --git a/tests/config/config.d/custom_settings_prefixes.xml b/tests/config/config.d/custom_settings_prefixes.xml
index ec66c1a930ec..9baaa4e09f3c 100644
--- a/tests/config/config.d/custom_settings_prefixes.xml
+++ b/tests/config/config.d/custom_settings_prefixes.xml
@@ -1,4 +1,4 @@
 <clickhouse>
     <!-- Comma-separated list of prefixes for user-defined settings. -->
-    <custom_settings_prefixes>custom_</custom_settings_prefixes>
+    <custom_settings_prefixes>SQL_,custom_</custom_settings_prefixes>
 </clickhouse>
diff --git a/tests/queries/0_stateless/02737_sql_auto_is_null.reference b/tests/queries/0_stateless/02737_sql_auto_is_null.reference
new file mode 100644
index 000000000000..573541ac9702
--- /dev/null
+++ b/tests/queries/0_stateless/02737_sql_auto_is_null.reference
@@ -0,0 +1,1 @@
+0
diff --git a/tests/queries/0_stateless/02737_sql_auto_is_null.sql b/tests/queries/0_stateless/02737_sql_auto_is_null.sql
new file mode 100644
index 000000000000..22f1a9524ca8
--- /dev/null
+++ b/tests/queries/0_stateless/02737_sql_auto_is_null.sql
@@ -0,0 +1,2 @@
+SET SQL_AUTO_IS_NULL = 0;
+SELECT getSetting('SQL_AUTO_IS_NULL');
