diff --git a/src/Storages/checkAndGetLiteralArgument.cpp b/src/Storages/checkAndGetLiteralArgument.cpp
index 1aa942548a75..5baf47fe91a9 100644
--- a/src/Storages/checkAndGetLiteralArgument.cpp
+++ b/src/Storages/checkAndGetLiteralArgument.cpp
@@ -12,7 +12,15 @@ namespace ErrorCodes
 template <typename T>
 T checkAndGetLiteralArgument(const ASTPtr & arg, const String & arg_name)
 {
-    return checkAndGetLiteralArgument<T>(*arg->as<ASTLiteral>(), arg_name);
+    if (arg && arg->as<ASTLiteral>())
+        return checkAndGetLiteralArgument<T>(*arg->as<ASTLiteral>(), arg_name);
+
+    throw Exception(
+        ErrorCodes::BAD_ARGUMENTS,
+        "Argument '{}' must be a literal, get {} (value: {})",
+        arg_name,
+        arg ? arg->getID() : "NULL",
+        arg ? arg->formatForErrorMessage() : "NULL");
 }
 
 template <typename T>
