diff --git a/tests/queries/0_stateless/02811_invalid_embedded_rocksdb_create.reference b/tests/queries/0_stateless/02811_invalid_embedded_rocksdb_create.reference
new file mode 100644
index 000000000000..e69de29bb2d1
diff --git a/tests/queries/0_stateless/02811_invalid_embedded_rocksdb_create.sql b/tests/queries/0_stateless/02811_invalid_embedded_rocksdb_create.sql
new file mode 100644
index 000000000000..a87ac5e0de08
--- /dev/null
+++ b/tests/queries/0_stateless/02811_invalid_embedded_rocksdb_create.sql
@@ -0,0 +1,2 @@
+-- Tags: no-fasttest
+CREATE TABLE dict (`k` String, `v` String) ENGINE = EmbeddedRocksDB(k) PRIMARY KEY k; -- {serverError 36}
