diff --git a/tests/queries/0_stateless/03262_analyzer_materialized_view_in_with_cte.reference b/tests/queries/0_stateless/03262_analyzer_materialized_view_in_with_cte.reference
new file mode 100644
index 000000000000..5ddf8439af5c
--- /dev/null
+++ b/tests/queries/0_stateless/03262_analyzer_materialized_view_in_with_cte.reference
@@ -0,0 +1,1 @@
+1	2	\N	test
diff --git a/tests/queries/0_stateless/03262_analyzer_materialized_view_in_with_cte.sql b/tests/queries/0_stateless/03262_analyzer_materialized_view_in_with_cte.sql
new file mode 100644
index 000000000000..4543d336d14f
--- /dev/null
+++ b/tests/queries/0_stateless/03262_analyzer_materialized_view_in_with_cte.sql
@@ -0,0 +1,63 @@
+SET allow_experimental_analyzer = 1;
+
+DROP TABLE IF EXISTS mv_test;
+DROP TABLE IF EXISTS mv_test_target;
+DROP VIEW IF EXISTS mv_test_mv;
+
+CREATE TABLE mv_test
+(
+    `id` UInt64,
+    `ref_id` UInt64,
+    `final_id` Nullable(UInt64),
+    `display` String
+)
+ENGINE = Log;
+
+CREATE TABLE mv_test_target
+(
+    `id` UInt64,
+    `ref_id` UInt64,
+    `final_id` Nullable(UInt64),
+    `display` String
+)
+ENGINE = Log;
+
+CREATE MATERIALIZED VIEW mv_test_mv TO mv_test_target
+(
+    `id` UInt64,
+    `ref_id` UInt64,
+    `final_id` Nullable(UInt64),
+    `display` String
+)
+AS WITH
+    tester AS
+    (
+        SELECT
+            id,
+            ref_id,
+            final_id,
+            display
+        FROM mv_test
+    ),
+    id_set AS
+    (
+        SELECT
+            display,
+            max(id) AS max_id
+        FROM mv_test
+        GROUP BY display
+    )
+SELECT *
+FROM tester
+WHERE id IN (
+    SELECT max_id
+    FROM id_set
+);
+
+INSERT INTO mv_test ( id, ref_id, display) values ( 1, 2, 'test');
+
+SELECT * FROM mv_test_target;
+
+DROP VIEW mv_test_mv;
+DROP TABLE mv_test_target;
+DROP TABLE mv_test;
diff --git a/tests/queries/0_stateless/03263_analyzer_materialized_view_cte_nested.reference b/tests/queries/0_stateless/03263_analyzer_materialized_view_cte_nested.reference
new file mode 100644
index 000000000000..0cfbf08886fc
--- /dev/null
+++ b/tests/queries/0_stateless/03263_analyzer_materialized_view_cte_nested.reference
@@ -0,0 +1,1 @@
+2
diff --git a/tests/queries/0_stateless/03263_analyzer_materialized_view_cte_nested.sql b/tests/queries/0_stateless/03263_analyzer_materialized_view_cte_nested.sql
new file mode 100644
index 000000000000..4ea853a7c22a
--- /dev/null
+++ b/tests/queries/0_stateless/03263_analyzer_materialized_view_cte_nested.sql
@@ -0,0 +1,19 @@
+SET allow_experimental_analyzer = 1;
+
+DROP TABLE IF EXISTS test_table;
+DROP VIEW IF EXISTS test_mv;
+
+CREATE TABLE test_table ENGINE = MergeTree ORDER BY tuple() AS SELECT 1 as col1;
+
+CREATE MATERIALIZED VIEW test_mv ENGINE = MergeTree ORDER BY tuple() AS
+WITH
+    subquery_on_source AS (SELECT col1 AS aliased FROM test_table),
+    output AS (SELECT * FROM test_table WHERE col1 IN (SELECT aliased FROM subquery_on_source))
+SELECT * FROM output;
+
+INSERT INTO test_table VALUES (2);
+
+SELECT * FROM test_mv;
+
+DROP VIEW test_mv;
+DROP TABLE test_table;
