diff --git a/tests/queries/0_stateless/02389_analyzer_nested_lambda.sql b/tests/queries/0_stateless/02389_analyzer_nested_lambda.sql
index 48e84246d1c9..be4b64888ca2 100644
--- a/tests/queries/0_stateless/02389_analyzer_nested_lambda.sql
+++ b/tests/queries/0_stateless/02389_analyzer_nested_lambda.sql
@@ -127,3 +127,14 @@ SELECT arrayMap(x -> splitByChar(toString(id), arrayMap(x -> toString(1), [NULL]
 DROP TABLE test_table;
 
 -- { echoOff }
+
+SELECT
+    groupArray(number) AS counts,
+    arraySum(arrayMap(x -> (x + 1), counts)) as hello,
+    arrayMap(x -> (x / hello), counts) AS res
+FROM numbers(1000000) FORMAT Null;
+
+SELECT
+  arrayWithConstant(pow(10,6), 1) AS nums,
+  arrayMap(x -> x, nums) AS m,
+  arrayMap(x -> x + arraySum(m), m) AS res FORMAT Null;
