diff --git a/tests/dragonfly/server_family_test.py b/tests/dragonfly/server_family_test.py
index 020dcb60b399..f3a9dde65f20 100644
--- a/tests/dragonfly/server_family_test.py
+++ b/tests/dragonfly/server_family_test.py
@@ -52,3 +52,11 @@ def test_multi_eval(client):
     except Exception as e:
         msg = str(e)
         assert "Dragonfly does not allow execution of" in msg
+
+
+def test_connection_name(client):
+    name = client.execute_command("CLIENT GETNAME")
+    assert not name
+    client.execute_command("CLIENT SETNAME test_conn_name")
+    name = client.execute_command("CLIENT GETNAME")
+    assert name == "test_conn_name"
