diff --git a/tests/integration/test_storage_s3/s3_mock/mock_s3.py b/tests/integration/test_storage_s3/s3_mock/mock_s3.py
index 35b477d6b108..5b422f6a73a5 100644
--- a/tests/integration/test_storage_s3/s3_mock/mock_s3.py
+++ b/tests/integration/test_storage_s3/s3_mock/mock_s3.py
@@ -1,4 +1,11 @@
-from bottle import abort, route, run, request
+from bottle import abort, route, run, request, response
+
+
+@route('/redirected/<_path>')
+def infinite_redirect(_path):
+    response.set_header("Location", request.url)
+    response.status = 307
+    return 'Redirected'
 
 
 @route('/<_bucket>/<_path>')
diff --git a/tests/integration/test_storage_s3/test.py b/tests/integration/test_storage_s3/test.py
index b25e5907e62a..56e500fed136 100644
--- a/tests/integration/test_storage_s3/test.py
+++ b/tests/integration/test_storage_s3/test.py
@@ -336,3 +336,19 @@ def test_get_csv_default(cluster):
     instance = cluster.instances["dummy"]  # type: ClickHouseInstance
     result = run_query(instance, get_query)
     assert result == '1\t2\t3
'
+
+
+def test_infinite_redirect(cluster):
+    bucket = "redirected"
+    table_format = "column1 UInt32, column2 UInt32, column3 UInt32"
+    filename = "test.csv"
+    get_query = "select * from s3('http://resolver:8080/{bucket}/{file}', 'CSV', '{table_format}')".format(
+        bucket="redirected",
+        file=filename,
+        table_format=table_format)
+    instance = cluster.instances["dummy"]  # type: ClickHouseInstance
+    try:
+        result = run_query(instance, get_query)
+    except Exception as e:
+        assert str(e).find("Too many redirects while trying to access") != -1
+
