diff --git a/test/optimizer/compressed_materialization.test_coverage b/test/optimizer/compressed_materialization.test_slow
similarity index 98%
rename from test/optimizer/compressed_materialization.test_coverage
rename to test/optimizer/compressed_materialization.test_slow
index 4088c1efab2d..a39f17f8a4af 100644
--- a/test/optimizer/compressed_materialization.test_coverage
+++ b/test/optimizer/compressed_materialization.test_slow
@@ -1,4 +1,4 @@
-# name: test/optimizer/compressed_materialization.test_coverage
+# name: test/optimizer/compressed_materialization.test_slow
 # description: Compressed materialization test
 # group: [optimizer]
 
@@ -18,7 +18,7 @@ Binder Error: Compressed materialization functions are for internal use only!
 statement ok
 create table t0 as select range%400000 a, range%400000 b from range(500000);
 
-query III
+query III rowsort
 select * from (
       select *, row_number() OVER () as row_number from (
           SELECT * FROM t0 ORDER BY 1) ta
diff --git a/test/sql/topn/tpch_top_n.test_slow b/test/sql/topn/tpch_top_n.test_slow
new file mode 100644
index 000000000000..a3657bb05faf
--- /dev/null
+++ b/test/sql/topn/tpch_top_n.test_slow
@@ -0,0 +1,78 @@
+# name: test/sql/topn/tpch_top_n.test_slow
+# description: Test Top N NULLS FIRST/LAST with few rows
+# group: [topn]
+
+require tpch
+
+statement ok
+CALL dbgen(sf=1);
+
+query I
+select l_quantity
+from lineitem
+where l_linestatus = 'O'
+order by l_quantity limit 10 offset 100;
+----
+1.00
+1.00
+1.00
+1.00
+1.00
+1.00
+1.00
+1.00
+1.00
+1.00
+
+query I
+select l_quantity
+from lineitem
+where l_linestatus = 'O'
+order by l_quantity limit 10 offset 1000000;
+----
+17.00
+17.00
+17.00
+17.00
+17.00
+17.00
+17.00
+17.00
+17.00
+17.00
+
+query I
+select sum(l_quantity)
+from lineitem
+group by l_orderkey
+order by sum(l_quantity) desc
+limit 10 offset 100;
+----
+297.00
+296.00
+296.00
+296.00
+296.00
+296.00
+296.00
+296.00
+295.00
+295.00
+
+query I
+select sum(l_quantity)
+from lineitem
+group by l_orderkey
+order by sum(l_quantity) desc
+limit 10 offset 100000;
+----
+195.00
+195.00
+195.00
+195.00
+195.00
+195.00
+195.00
+195.00
+195.00
+195.00
