diff --git a/src/server/multi_test.cc b/src/server/multi_test.cc
index c48633633c0d..225ae5523122 100644
--- a/src/server/multi_test.cc
+++ b/src/server/multi_test.cc
@@ -1083,12 +1083,22 @@ TEST_F(MultiEvalTest, ScriptSquashingUknownCmd) {
 }
 
 TEST_F(MultiEvalTest, MultiAndEval) {
+  if (auto mode = absl::GetFlag(FLAGS_multi_exec_mode); mode == Transaction::NON_ATOMIC) {
+    GTEST_SKIP() << "Skipped MultiAndEval test because multi_exec_mode is non atomic";
+    return;
+  }
+
   // We had a bug in borrowing interpreters which caused a crash in this scenario
   Run({"multi"});
   Run({"eval", "return redis.call('set', 'x', 'y1')", "1", "x"});
   Run({"exec"});
 
   Run({"eval", "return redis.call('set', 'x', 'y1')", "1", "x"});
+
+  Run({"multi"});
+  Run({"eval", "return 'OK';", "0"});
+  auto resp = Run({"exec"});
+  EXPECT_EQ(resp, "OK");
 }
 
 }  // namespace dfly
