diff --git a/tests/queries/0_stateless/02674_and_consistency.reference b/tests/queries/0_stateless/02674_and_consistency.reference
new file mode 100644
index 000000000000..e74ab1928c5b
--- /dev/null
+++ b/tests/queries/0_stateless/02674_and_consistency.reference
@@ -0,0 +1,4 @@
+10
+#45218
+10
+=
diff --git a/tests/queries/0_stateless/02674_and_consistency.sql b/tests/queries/0_stateless/02674_and_consistency.sql
new file mode 100644
index 000000000000..5988832ba68b
--- /dev/null
+++ b/tests/queries/0_stateless/02674_and_consistency.sql
@@ -0,0 +1,21 @@
+SELECT SUM(number)
+FROM
+(
+    SELECT 10 AS number
+)
+GROUP BY number
+HAVING 1 AND sin(SUMOrNull(number))
+SETTINGS enable_optimize_predicate_expression = 0;
+
+select '#45218';
+
+SELECT SUM(number)
+FROM
+(
+    SELECT 10 AS number
+)
+GROUP BY cos(min2(number, number) % number) - number
+HAVING ((-sign(-233841197)) IS NOT NULL) AND sin(lcm(SUM(number), SUM(number)) >= ('372497213' IS NOT NULL))
+SETTINGS aggregate_functions_null_for_empty = 1, enable_optimize_predicate_expression = 0;
+
+select '=';
