diff --git a/src/Interpreters/InterpreterSelectQuery.cpp b/src/Interpreters/InterpreterSelectQuery.cpp
index a0d6fe67b396..5c096a7dbacc 100644
--- a/src/Interpreters/InterpreterSelectQuery.cpp
+++ b/src/Interpreters/InterpreterSelectQuery.cpp
@@ -1777,6 +1777,7 @@ void InterpreterSelectQuery::executeFetchColumns(QueryProcessingStage::Enum proc
     bool optimize_trivial_count =
         syntax_analyzer_result->optimize_trivial_count
         && (settings.max_parallel_replicas <= 1)
+        && !settings.allow_experimental_query_deduplication
         && storage
         && storage->getName() != "MaterializedMySQL"
         && !row_policy_filter
diff --git a/src/Storages/MergeTree/MergeTreeData.cpp b/src/Storages/MergeTree/MergeTreeData.cpp
index 25a0bf582aed..7bfff3b50884 100644
--- a/src/Storages/MergeTree/MergeTreeData.cpp
+++ b/src/Storages/MergeTree/MergeTreeData.cpp
@@ -4720,6 +4720,10 @@ std::optional<ProjectionCandidate> MergeTreeData::getQueryProcessingStageWithAgg
     if (select_query->sampleSize())
         return std::nullopt;
 
+    // Currently projection don't support deduplication when moving parts between shards.
+    if (settings.allow_experimental_query_deduplication)
+        return std::nullopt;
+
     // Currently projections don't support ARRAY JOIN yet.
     if (select_query->arrayJoinExpressionList().first)
         return std::nullopt;
