diff --git a/test/sql/fts/issue_12330.test b/test/sql/fts/issue_12330.test
new file mode 100644
index 000000000000..60433e1bc943
--- /dev/null
+++ b/test/sql/fts/issue_12330.test
@@ -0,0 +1,34 @@
+# name: test/sql/fts/issue_12330.test
+# description: Issue 12330: BM25 matching scores seems to be invalid
+# group: [fts]
+
+# issue #7384 and #8141
+
+require fts
+
+statement ok
+CREATE OR REPLACE TABLE documents (
+    id VARCHAR,
+    content VARCHAR
+);
+
+statement ok
+INSERT INTO documents VALUES
+    ('doc1', 'DuckDB database lorem'),
+    ('doc2', 'DuckDB database ipsum'),
+    ('doc3', 'DuckDB database ipsum dolor');
+
+statement ok
+PRAGMA create_fts_index('documents', 'id', 'content');
+
+query I
+SELECT
+    id
+FROM
+    documents
+ORDER BY
+    fts_main_documents.match_bm25(id, 'DuckDB database ipsum') DESC;
+----
+doc2
+doc3
+doc1
diff --git a/test/sql/fts/test_indexing.test_slow b/test/sql/fts/test_indexing.test_slow
index bc99f0e02983..7ff6928d1a93 100644
--- a/test/sql/fts/test_indexing.test_slow
+++ b/test/sql/fts/test_indexing.test_slow
@@ -135,8 +135,8 @@ doc1	0.36835264087244074
 query III
 SELECT score, id, body FROM (SELECT *, fts_main_documents.match_bm25(id, 'quacked barked') AS score FROM documents) sq WHERE score IS NOT NULL ORDER BY score DESC
 ----
-0.37543634550460314	doc2	 BÁRKING+BÁRKING+BÁRKING+BÁRKING
-0.3683526408724408	doc1	 QUÁCKING+QUÁCKING+QUÁCKING
+0.7208701623069375	doc2	 BÁRKING+BÁRKING+BÁRKING+BÁRKING
+0.7072688384898254	doc1	 QUÁCKING+QUÁCKING+QUÁCKING
 
 # drop and re-create, but index both the 'body' and 'author' column this time
 statement ok
diff --git a/test/sql/fts/test_issue_10254.test b/test/sql/fts/test_issue_10254.test
index 0d5cd0cef1b1..fd149745a10a 100644
--- a/test/sql/fts/test_issue_10254.test
+++ b/test/sql/fts/test_issue_10254.test
@@ -21,5 +21,5 @@ PRAGMA create_fts_index('data', 'id', 'context', 'question', stemmer='russian',
 query I
 SELECT id FROM (SELECT *, fts_main_data.match_bm25(id, 'Какие') AS score FROM data) sq WHERE score IS NOT NULL ORDER BY score DESC;
 ----
-1
 0
+1
