diff --git a/src/Interpreters/FillingRow.cpp b/src/Interpreters/FillingRow.cpp
index 7efbb62b0d20..21b5b04bca3f 100644
--- a/src/Interpreters/FillingRow.cpp
+++ b/src/Interpreters/FillingRow.cpp
@@ -16,6 +16,10 @@ bool less(const Field & lhs, const Field & rhs, int direction)
 
 bool equals(const Field & lhs, const Field & rhs)
 {
+    /// This will treat NaNs as equal
+    if (lhs.getType() == rhs.getType())
+        return lhs == rhs;
+
     return applyVisitor(FieldVisitorAccurateEquals(), lhs, rhs);
 }
 
