diff --git a/tests/queries/0_stateless/03239_nan_with_fill.reference b/tests/queries/0_stateless/03239_nan_with_fill.reference
new file mode 100644
index 000000000000..10c8c316670f
--- /dev/null
+++ b/tests/queries/0_stateless/03239_nan_with_fill.reference
@@ -0,0 +1,3 @@
+nan
+nan
+nan
diff --git a/tests/queries/0_stateless/03239_nan_with_fill.sql b/tests/queries/0_stateless/03239_nan_with_fill.sql
new file mode 100644
index 000000000000..69eaf9b0c42b
--- /dev/null
+++ b/tests/queries/0_stateless/03239_nan_with_fill.sql
@@ -0,0 +1,3 @@
+SELECT nan ORDER BY 1 WITH FILL;
+SELECT -nan ORDER BY 1 WITH FILL;
+SELECT 0./0. ORDER BY 1 WITH FILL;
