diff --git a/tests/queries/0_stateless/03261_minmax_indices_by_default_table_copy.reference b/tests/queries/0_stateless/03261_minmax_indices_by_default_table_copy.reference
new file mode 100644
index 000000000000..e69de29bb2d1
diff --git a/tests/queries/0_stateless/03261_minmax_indices_by_default_table_copy.sql b/tests/queries/0_stateless/03261_minmax_indices_by_default_table_copy.sql
new file mode 100644
index 000000000000..41bd81d8ffe8
--- /dev/null
+++ b/tests/queries/0_stateless/03261_minmax_indices_by_default_table_copy.sql
@@ -0,0 +1,18 @@
+-- Test for issue #75677
+
+DROP TABLE IF EXISTS tab1;
+
+CREATE TABLE tab1 (
+    a Int32,
+    b String,
+    c Float64)
+ENGINE MergeTree
+ORDER BY a
+SETTINGS
+    add_minmax_index_for_numeric_columns = 1,
+    add_minmax_index_for_string_columns = 1;
+
+CREATE TABLE tab2 AS tab1;
+
+DROP TABLE tab1;
+DROP TABLE tab2;
