diff --git a/tests/queries/0_stateless/02141_clickhouse_local_interactive_table.reference b/tests/queries/0_stateless/02141_clickhouse_local_interactive_table.reference
index 277f4df04ab0..8b5f06cd5a3d 100644
--- a/tests/queries/0_stateless/02141_clickhouse_local_interactive_table.reference
+++ b/tests/queries/0_stateless/02141_clickhouse_local_interactive_table.reference
@@ -1,2 +1,1 @@
 CREATE TEMPORARY TABLE `table`
(
    `key` String
)
ENGINE = File(TSVWithNamesAndTypes, \'/dev/null\', auto)
-CREATE TEMPORARY TABLE `table`
(
    `key` String
)
ENGINE = File(TSVWithNamesAndTypes, \'/dev/null\', auto)
diff --git a/tests/queries/0_stateless/02141_clickhouse_local_interactive_table.sh b/tests/queries/0_stateless/02141_clickhouse_local_interactive_table.sh
index 0ec16b332bb4..5b08a26d76e1 100755
--- a/tests/queries/0_stateless/02141_clickhouse_local_interactive_table.sh
+++ b/tests/queries/0_stateless/02141_clickhouse_local_interactive_table.sh
@@ -5,4 +5,3 @@ CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
 . "$CURDIR"/../shell_config.sh
 
 $CLICKHOUSE_LOCAL --file /dev/null --structure "key String" --input-format TSVWithNamesAndTypes --interactive <<<'show create temporary table table'
-$CLICKHOUSE_LOCAL --database foo --file /dev/null --structure "key String" --input-format TSVWithNamesAndTypes --interactive <<<'show create temporary table table'
diff --git a/tests/queries/0_stateless/03533_clickhouse_local_database_argument.reference b/tests/queries/0_stateless/03533_clickhouse_local_database_argument.reference
new file mode 100644
index 000000000000..4e3f26e9db41
--- /dev/null
+++ b/tests/queries/0_stateless/03533_clickhouse_local_database_argument.reference
@@ -0,0 +1,9 @@
+Hello, world
+Hello, world
+Hello, world
+Hello, world
+Hello, world
+Hello, world
+Hello from a file
+Hello from a file
+Hello from a file
diff --git a/tests/queries/0_stateless/03533_clickhouse_local_database_argument.sh b/tests/queries/0_stateless/03533_clickhouse_local_database_argument.sh
new file mode 100755
index 000000000000..231420574eca
--- /dev/null
+++ b/tests/queries/0_stateless/03533_clickhouse_local_database_argument.sh
@@ -0,0 +1,31 @@
+#!/usr/bin/env bash
+
+CUR_DIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+# shellcheck source=../shell_config.sh
+. "$CUR_DIR"/../shell_config.sh
+
+$CLICKHOUSE_LOCAL --path "${CLICKHOUSE_TMP}" --query "
+CREATE DATABASE test;
+USE test;
+CREATE TABLE t (s String) ORDER BY ();
+INSERT INTO t VALUES ('Hello, world');
+SELECT * FROM t;
+"
+
+# We can switch to the previously created database using a command-line argument:
+
+$CLICKHOUSE_LOCAL --path "${CLICKHOUSE_TMP}" --query "SELECT * FROM test.t;"
+$CLICKHOUSE_LOCAL --path "${CLICKHOUSE_TMP}" --query "USE test; SELECT * FROM t;"
+$CLICKHOUSE_LOCAL --path "${CLICKHOUSE_TMP}" --database default --query "USE test; SELECT * FROM t;"
+$CLICKHOUSE_LOCAL --path "${CLICKHOUSE_TMP}" --database test --query "SELECT * FROM t;"
+$CLICKHOUSE_LOCAL --path "${CLICKHOUSE_TMP}" --database system --query "USE test; SELECT * FROM t;"
+
+# Only default database is configured as a filesystem overlay:
+
+echo "Hello from a file" > "${CLICKHOUSE_TMP}/file.csv"
+
+$CLICKHOUSE_LOCAL --path "${CLICKHOUSE_TMP}" --query "SELECT * FROM '${CLICKHOUSE_TMP}/file.csv'"
+$CLICKHOUSE_LOCAL --path "${CLICKHOUSE_TMP}" --query "SELECT * FROM default.\`${CLICKHOUSE_TMP}/file.csv\`"
+$CLICKHOUSE_LOCAL --path "${CLICKHOUSE_TMP}" --database test --query "SELECT * FROM default.\`${CLICKHOUSE_TMP}/file.csv\`"
+
+$CLICKHOUSE_LOCAL --path "${CLICKHOUSE_TMP}" --query "DROP DATABASE test;"
