diff --git a/tests/queries/0_stateless/02990_format_select_from_explain.reference b/tests/queries/0_stateless/02990_format_select_from_explain.reference
new file mode 100644
index 000000000000..7c8dcef3824a
--- /dev/null
+++ b/tests/queries/0_stateless/02990_format_select_from_explain.reference
@@ -0,0 +1,9 @@
+SELECT explain
+FROM
+(
+    SELECT *
+    FROM viewExplain('EXPLAIN AST', '', (
+        SELECT *
+        FROM system.numbers
+    ))
+)
diff --git a/tests/queries/0_stateless/02990_format_select_from_explain.sh b/tests/queries/0_stateless/02990_format_select_from_explain.sh
new file mode 100755
index 000000000000..4955b7337883
--- /dev/null
+++ b/tests/queries/0_stateless/02990_format_select_from_explain.sh
@@ -0,0 +1,7 @@
+#!/usr/bin/env bash
+
+CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+# shellcheck source=../shell_config.sh
+. "$CURDIR"/../shell_config.sh
+
+${CLICKHOUSE_FORMAT} --query "SELECT explain FROM (EXPLAIN AST SELECT * FROM system.numbers)"
diff --git a/tests/queries/0_stateless/02993_lazy_index_loading.reference b/tests/queries/0_stateless/02993_lazy_index_loading.reference
new file mode 100644
index 000000000000..5bc329ae4eb7
--- /dev/null
+++ b/tests/queries/0_stateless/02993_lazy_index_loading.reference
@@ -0,0 +1,4 @@
+100000000	140000000
+0	0
+1
+100000000	100000000
diff --git a/tests/queries/0_stateless/02993_lazy_index_loading.sql b/tests/queries/0_stateless/02993_lazy_index_loading.sql
new file mode 100644
index 000000000000..7de4af9ef0e4
--- /dev/null
+++ b/tests/queries/0_stateless/02993_lazy_index_loading.sql
@@ -0,0 +1,18 @@
+DROP TABLE IF EXISTS test;
+CREATE TABLE test (s String) ENGINE = MergeTree ORDER BY s SETTINGS index_granularity = 1;
+
+INSERT INTO test SELECT randomString(1000) FROM numbers(100000);
+SELECT round(primary_key_bytes_in_memory, -7), round(primary_key_bytes_in_memory_allocated, -7) FROM system.parts WHERE database = currentDatabase() AND table = 'test';
+
+DETACH TABLE test;
+SET max_memory_usage = '50M';
+ATTACH TABLE test;
+
+SELECT primary_key_bytes_in_memory, primary_key_bytes_in_memory_allocated FROM system.parts WHERE database = currentDatabase() AND table = 'test';
+
+SET max_memory_usage = '200M';
+SELECT s != '' FROM test LIMIT 1;
+
+SELECT round(primary_key_bytes_in_memory, -7), round(primary_key_bytes_in_memory_allocated, -7) FROM system.parts WHERE database = currentDatabase() AND table = 'test';
+
+DROP TABLE test;
