diff --git a/tools/pythonpkg/tests/fast/test_type.py b/tools/pythonpkg/tests/fast/test_type.py
index 718b9defe310..d87a1cae38db 100644
--- a/tools/pythonpkg/tests/fast/test_type.py
+++ b/tools/pythonpkg/tests/fast/test_type.py
@@ -88,6 +88,12 @@ def test_struct_type(self):
         type = duckdb.struct_type([BIGINT, BOOLEAN])
         assert str(type) == 'STRUCT(v1 BIGINT, v2 BOOLEAN)'
 
+    def test_incomplete_struct_type(self):
+        with pytest.raises(
+            duckdb.InvalidInputException, match='Could not convert empty dictionary to a duckdb STRUCT type'
+        ):
+            type = duckdb.typing.DuckDBPyType(dict())
+
     def test_map_type(self):
         type = duckdb.map_type(duckdb.sqltype("BIGINT"), duckdb.sqltype("DECIMAL(10, 2)"))
         assert str(type) == 'MAP(BIGINT, DECIMAL(10,2))'
