diff --git a/tools/pythonpkg/tests/fast/api/test_duckdb_connection.py b/tools/pythonpkg/tests/fast/api/test_duckdb_connection.py
index c5f8e6375d69..76e9b750b4e8 100644
--- a/tools/pythonpkg/tests/fast/api/test_duckdb_connection.py
+++ b/tools/pythonpkg/tests/fast/api/test_duckdb_connection.py
@@ -69,6 +69,21 @@ def test_cursor(self):
             # 'tbl' no longer exists
             duckdb.table("tbl")
 
+    def test_cursor_lifetime(self):
+        con = duckdb.connect()
+
+        def use_cursors():
+            cursors = []
+            for _ in range(10):
+                cursors.append(con.cursor())
+
+            for cursor in cursors:
+                print("closing cursor")
+                cursor.close()
+
+        use_cursors()
+        con.close()
+
     def test_df(self):
         ref = [([1, 2, 3],)]
         duckdb.execute("select [1,2,3]")
