diff --git a/tests/queries/0_stateless/01375_storage_file_write_prefix.reference b/tests/queries/0_stateless/01375_storage_file_write_prefix.reference
new file mode 100644
index 000000000000..ed9a18b93463
--- /dev/null
+++ b/tests/queries/0_stateless/01375_storage_file_write_prefix.reference
@@ -0,0 +1,30 @@
+0	0
+1	1
+2	2
+3	3
+4	4
+5	5
+6	6
+7	7
+8	8
+9	9
+0	0
+1	1
+2	2
+3	3
+4	4
+5	5
+6	6
+7	7
+8	8
+9	9
+0	0
+1	1
+2	2
+3	3
+4	4
+5	5
+6	6
+7	7
+8	8
+9	9
diff --git a/tests/queries/0_stateless/01375_storage_file_write_prefix.sql b/tests/queries/0_stateless/01375_storage_file_write_prefix.sql
new file mode 100644
index 000000000000..b335db03f6c6
--- /dev/null
+++ b/tests/queries/0_stateless/01375_storage_file_write_prefix.sql
@@ -0,0 +1,14 @@
+DROP TABLE IF EXISTS tmp_01375;
+DROP TABLE IF EXISTS table_csv_01375;
+
+CREATE TABLE tmp_01375 (n UInt32, s String) ENGINE = Memory;
+CREATE TABLE table_csv_01375 AS tmp_01375 ENGINE = File(CSVWithNames);
+
+INSERT INTO table_csv_01375 SELECT number as n, toString(n) as s FROM numbers(10);
+INSERT INTO table_csv_01375 SELECT number as n, toString(n) as s FROM numbers(10);
+INSERT INTO table_csv_01375 SELECT number as n, toString(n) as s FROM numbers(10);
+
+SELECT * FROM table_csv_01375;
+
+DROP TABLE IF EXISTS tmp_01375;
+DROP TABLE IF EXISTS table_csv_01375;
