diff --git a/src/Interpreters/PreparedSets.cpp b/src/Interpreters/PreparedSets.cpp
index c2c01b912f95..e0551dff2adb 100644
--- a/src/Interpreters/PreparedSets.cpp
+++ b/src/Interpreters/PreparedSets.cpp
@@ -198,7 +198,11 @@ SetPtr FutureSetFromSubquery::buildOrderedSetInplace(const ContextPtr & context)
     CompletedPipelineExecutor executor(pipeline);
     executor.execute();
 
-    set_and_key->set->checkIsCreated();
+    /// SET may not be created successfully at this step because of the sub-query timeout, but if we have
+    /// timeout_overflow_mode set to `break`, no exception is thrown, and the executor just stops executing
+    /// the pipeline without setting `set_and_key->set->is_created` to true.
+    if (!set_and_key->set->isCreated())
+        return nullptr;
 
     return set_and_key->set;
 }
