diff --git a/tools/shell/tests/conftest.py b/tools/shell/tests/conftest.py
index 71a7cd50ac24..84dba9b4e37a 100644
--- a/tools/shell/tests/conftest.py
+++ b/tools/shell/tests/conftest.py
@@ -36,6 +36,12 @@ def check_stdout(self, expected: Union[str, List[str], bytes]):
         assert self.status_code == 0
         assert expected in self.stdout
 
+    def check_not_exist(self, not_exist: Union[str, List[str], bytes]):
+        if isinstance(not_exist, list):
+            not_exist = '
'.join(not_exist)
+        assert self.status_code == 0
+        assert not_exist not in self.stdout
+
     def check_stderr(self, expected: str):
         assert expected in self.stderr
 
diff --git a/tools/shell/tests/test_shell_basics.py b/tools/shell/tests/test_shell_basics.py
index df2a6a9d8c5b..e36edbb582b0 100644
--- a/tools/shell/tests/test_shell_basics.py
+++ b/tools/shell/tests/test_shell_basics.py
@@ -619,10 +619,16 @@ def test_mode_insert(shell):
     test = (
         ShellTest(shell)
         .statement(".mode insert")
-        .statement("SELECT NULL, 42, 'fourty-two', 42.0;")
+        .statement("SELECT NULL, 42, 'fourty-two', 42.0, 3.14, 2.71;")
     )
     result = test.run()
     result.check_stdout('fourty-two')
+    result.check_stdout('3.14')
+    result.check_stdout('2.71')
+    result.check_not_exist('3.140000')
+    result.check_not_exist('2.709999')
+    result.check_not_exist('3.139999')
+    result.check_not_exist('2.710000')
 
 def test_mode_line(shell):
     test = (
