diff --git a/tests/queries/0_stateless/01384_bloom_filter_bad_arguments.reference b/tests/queries/0_stateless/01384_bloom_filter_bad_arguments.reference
new file mode 100644
index 000000000000..e69de29bb2d1
diff --git a/tests/queries/0_stateless/01384_bloom_filter_bad_arguments.sql b/tests/queries/0_stateless/01384_bloom_filter_bad_arguments.sql
new file mode 100644
index 000000000000..5b18f28883a3
--- /dev/null
+++ b/tests/queries/0_stateless/01384_bloom_filter_bad_arguments.sql
@@ -0,0 +1,10 @@
+DROP TABLE IF EXISTS test;
+
+create table test (a String, index a a type tokenbf_v1(0, 2, 0) granularity 1) engine MergeTree order by a; -- { serverError 36 }
+create table test (a String, index a a type tokenbf_v1(2, 0, 0) granularity 1) engine MergeTree order by a; -- { serverError 36 }
+create table test (a String, index a a type tokenbf_v1(0, 1, 1) granularity 1) engine MergeTree order by a; -- { serverError 36 }
+create table test (a String, index a a type tokenbf_v1(1, 0, 1) granularity 1) engine MergeTree order by a; -- { serverError 36 }
+
+create table test (a String, index a a type tokenbf_v1(0.1, 2, 0) granularity 1) engine MergeTree order by a; -- { serverError 36 }
+create table test (a String, index a a type tokenbf_v1(-1, 2, 0) granularity 1) engine MergeTree order by a; -- { serverError 36 }
+create table test (a String, index a a type tokenbf_v1(0xFFFFFFFF, 2, 0) granularity 1) engine MergeTree order by a; -- { serverError 36 }
