diff --git a/tests/queries/0_stateless/01550_create_map_type.sql b/tests/queries/0_stateless/01550_create_map_type.sql
index 7a08886a5d8d..695a40c42b4a 100644
--- a/tests/queries/0_stateless/01550_create_map_type.sql
+++ b/tests/queries/0_stateless/01550_create_map_type.sql
@@ -63,3 +63,4 @@ SELECT sum(m['1']), sum(m['7']), sum(m['100']) FROM table_map;
 
 DROP TABLE IF EXISTS table_map;
 
+SELECT CAST(([2, 1, 1023], ['', '']), 'Map(UInt8, String)') AS map, map[10] -- { serverError 53}
