diff --git a/tests/queries/0_stateless/01937_nested_chinese.reference b/tests/queries/0_stateless/01937_nested_chinese.reference
new file mode 100644
index 000000000000..54b6175d7fc6
--- /dev/null
+++ b/tests/queries/0_stateless/01937_nested_chinese.reference
@@ -0,0 +1,12 @@
+id	String					
+products.产品	Array(Array(String))					
+products.销量	Array(Array(Int32))					
+id	String					
+products.产品	Array(Array(String))					
+products.销量	Array(Array(Int32))					
+id	String					
+products.产品	Array(String)					
+products.销量	Array(Int32)					
+p.产品	Array(String)					
+p.销量	Array(Int32)					
+0
diff --git a/tests/queries/0_stateless/01937_nested_chinese.sql b/tests/queries/0_stateless/01937_nested_chinese.sql
new file mode 100644
index 000000000000..94c6598480e3
--- /dev/null
+++ b/tests/queries/0_stateless/01937_nested_chinese.sql
@@ -0,0 +1,8 @@
+CREATE TEMPORARY TABLE test (`id` String, `products` Nested (`产品` Array(String), `销量` Array(Int32)));
+
+DESCRIBE test;
+DESCRIBE (SELECT * FROM test);
+DESCRIBE (SELECT * FROM test ARRAY JOIN products);
+DESCRIBE (SELECT p.`产品`, p.`销量` FROM test ARRAY JOIN products AS p);
+SELECT * FROM test ARRAY JOIN products;
+SELECT count() FROM (SELECT * FROM test ARRAY JOIN products);
