diff --git a/tests/queries/0_stateless/03254_uniq_exact_two_level_negative_zero.reference b/tests/queries/0_stateless/03254_uniq_exact_two_level_negative_zero.reference
new file mode 100644
index 000000000000..771c05369c19
--- /dev/null
+++ b/tests/queries/0_stateless/03254_uniq_exact_two_level_negative_zero.reference
@@ -0,0 +1,1 @@
+7992019
diff --git a/tests/queries/0_stateless/03254_uniq_exact_two_level_negative_zero.sql b/tests/queries/0_stateless/03254_uniq_exact_two_level_negative_zero.sql
new file mode 100644
index 000000000000..3237818d2901
--- /dev/null
+++ b/tests/queries/0_stateless/03254_uniq_exact_two_level_negative_zero.sql
@@ -0,0 +1,1 @@
+WITH number % 1000 = 0 ? (rand() % 2 ? 0.0 : -0.0) : number::Float64 AS x SELECT length(uniqExactState(x)::String) FROM numbers(1000000);
