diff --git a/ydb/core/tablet_flat/test/libs/table/test_part.h b/ydb/core/tablet_flat/test/libs/table/test_part.h
index 3b1eab2f7d2d..ab856b912cfd 100644
--- a/ydb/core/tablet_flat/test/libs/table/test_part.h
+++ b/ydb/core/tablet_flat/test/libs/table/test_part.h
@@ -52,6 +52,13 @@ namespace NTest {
             return Store->GetPageSize(groupId.Index, pageId);
         }
 
+        ui64 GetPageSize(ELargeObj lob, ui64 ref) const override
+        {
+            Y_UNUSED(lob);
+            Y_UNUSED(ref);
+            return 0;
+        }
+
         NPage::EPage GetPageType(NPage::TPageId pageId, NPage::TGroupId groupId) const override
         {
             return Store->GetPageType(groupId.Index, pageId);
