diff --git a/tests/queries/0_stateless/02137_mv_into_join.reference b/tests/queries/0_stateless/02137_mv_into_join.reference
new file mode 100644
index 000000000000..1228a2322e6c
--- /dev/null
+++ b/tests/queries/0_stateless/02137_mv_into_join.reference
@@ -0,0 +1,3 @@
+sku_0001	black	women	nice shirt
+sku_0001_black	sku_0001	black	women	nice shirt
+sku_0001_black	sku_0001	black	women	nice shirt
diff --git a/tests/queries/0_stateless/02137_mv_into_join.sql b/tests/queries/0_stateless/02137_mv_into_join.sql
new file mode 100644
index 000000000000..cca896ac6226
--- /dev/null
+++ b/tests/queries/0_stateless/02137_mv_into_join.sql
@@ -0,0 +1,17 @@
+CREATE TABLE main ( `id` String, `color` String, `section` String, `description` String) ENGINE = MergeTree ORDER BY tuple();
+CREATE TABLE destination_join ( `key` String, `id` String, `color` String, `section` String, `description` String) ENGINE = Join(ANY, LEFT, key);
+CREATE TABLE destination_set (`key` String) ENGINE = Set;
+
+CREATE MATERIALIZED VIEW mv_to_join TO `destination_join` AS SELECT concat(id, '_', color) AS key, * FROM main;
+CREATE MATERIALIZED VIEW mv_to_set TO `destination_set` AS SELECT key FROM destination_join;
+
+INSERT INTO main VALUES ('sku_0001','black','women','nice shirt');
+SELECT * FROM main;
+SELECT * FROM destination_join;
+SELECT * FROM destination_join WHERE key in destination_set;
+
+DROP TABLE mv_to_set;
+DROP TABLE destination_set;
+DROP TABLE mv_to_join;
+DROP TABLE destination_join;
+DROP TABLE main;
