diff --git a/src/facade/dragonfly_connection.cc b/src/facade/dragonfly_connection.cc
index 367423c00be8..21849e476cc7 100644
--- a/src/facade/dragonfly_connection.cc
+++ b/src/facade/dragonfly_connection.cc
@@ -698,6 +698,10 @@ auto Connection::FromArgs(RespVec args, mi_heap_t* heap) -> Request* {
   return req;
 }
 
+void Connection::ShutdownSelf() {
+  util::Connection::Shutdown();
+}
+
 void RespToArgList(const RespVec& src, CmdArgVec* dest) {
   dest->resize(src.size());
   for (size_t i = 0; i < src.size(); ++i) {
diff --git a/src/facade/dragonfly_connection.h b/src/facade/dragonfly_connection.h
index 5578d5ed85da..f7258af59d36 100644
--- a/src/facade/dragonfly_connection.h
+++ b/src/facade/dragonfly_connection.h
@@ -72,6 +72,8 @@ class Connection : public util::Connection {
   std::string GetClientInfo() const;
   uint32 GetClientId() const;
 
+  void ShutdownSelf();
+
  protected:
   void OnShutdown() override;
   void OnPreMigrateThread() override;
diff --git a/src/server/main_service.cc b/src/server/main_service.cc
index 49f1ad7cde41..a8efec664055 100644
--- a/src/server/main_service.cc
+++ b/src/server/main_service.cc
@@ -721,6 +721,7 @@ void Service::Quit(CmdArgList args, ConnectionContext* cntx) {
 
   SinkReplyBuilder* builder = cntx->reply_builder();
   builder->CloseConnection();
+  cntx->owner()->ShutdownSelf();
 }
 
 void Service::Multi(CmdArgList args, ConnectionContext* cntx) {
