diff --git a/tools/pythonpkg/tests/fast/arrow/test_arrow_offsets.py b/tools/pythonpkg/tests/fast/arrow/test_arrow_offsets.py
index fe34a7e1c06a..782a68042561 100644
--- a/tools/pythonpkg/tests/fast/arrow/test_arrow_offsets.py
+++ b/tools/pythonpkg/tests/fast/arrow/test_arrow_offsets.py
@@ -671,3 +671,17 @@ def test_struct_of_list(self, duckdb_cursor, col1_null):
             assert res[0] == None
         else:
             assert res[0]['a'][-1] == '131072'
+
+    def test_bools_with_offset(self, duckdb_cursor):
+        bools = [False, False, False, False, True, False, False, False, False, False]
+        bool_array = pa.array(bools, type=pa.bool_())
+
+        # Create schema using the same fields as the C++ version
+        schema = pa.schema([pa.field("bools", pa.bool_())])
+
+        # Create a RecordBatch with the arrays
+        record_batch = pa.RecordBatch.from_arrays([bool_array], schema=schema)
+
+        temp_record = record_batch.slice(4, 1)
+        res = duckdb_cursor.sql("select bools from temp_record").fetchall()
+        assert res == [(True,)]
