diff --git a/tests/queries/0_stateless/02346_to_hour_monotonicity_fix.reference b/tests/queries/0_stateless/02346_to_hour_monotonicity_fix.reference
new file mode 100644
index 000000000000..5d49984b71ba
--- /dev/null
+++ b/tests/queries/0_stateless/02346_to_hour_monotonicity_fix.reference
@@ -0,0 +1,1 @@
+5	8	42
diff --git a/tests/queries/0_stateless/02346_to_hour_monotonicity_fix.sql b/tests/queries/0_stateless/02346_to_hour_monotonicity_fix.sql
new file mode 100644
index 000000000000..5eb396970d78
--- /dev/null
+++ b/tests/queries/0_stateless/02346_to_hour_monotonicity_fix.sql
@@ -0,0 +1,8 @@
+drop table if exists test_tz_hour;
+
+create table test_tz_hour(t DateTime, x String) engine MergeTree partition by toYYYYMMDD(t) order by x;
+insert into test_tz_hour select toDateTime('2021-06-01 00:00:00', 'UTC') + number * 600, 'x' from numbers(1e3);
+
+select toHour(toTimeZone(t, 'UTC')) as toHour_UTC, toHour(toTimeZone(t, 'Asia/Jerusalem')) as toHour_Israel, count() from test_tz_hour where toHour_Israel = 8 group by toHour_UTC, toHour_Israel;
+
+drop table test_tz_hour;
