diff --git a/.gitmodules b/.gitmodules
index c7ff7cea7db9..ecccf0633e22 100644
--- a/.gitmodules
+++ b/.gitmodules
@@ -84,7 +84,7 @@
 	url = https://github.com/google/brotli.git
 [submodule "contrib/h3"]
 	path = contrib/h3
-	url = https://github.com/uber/h3
+	url = https://github.com/ClickHouse-Extras/h3
 [submodule "contrib/hyperscan"]
 	path = contrib/hyperscan
 	url = https://github.com/ClickHouse-Extras/hyperscan.git
diff --git a/contrib/h3 b/contrib/h3
index 6cfd649e8c0d..e209086ae1b5 160000
--- a/contrib/h3
+++ b/contrib/h3
@@ -1,1 +1,1 @@
-Subproject commit 6cfd649e8c0d3ed913e8aae928a669fc3b8a2365
+Subproject commit e209086ae1b5477307f545a0f6111780edc59940
diff --git a/contrib/h3-cmake/CMakeLists.txt b/contrib/h3-cmake/CMakeLists.txt
index 80e76311f704..2911d7283f02 100644
--- a/contrib/h3-cmake/CMakeLists.txt
+++ b/contrib/h3-cmake/CMakeLists.txt
@@ -16,6 +16,7 @@ ${H3_SOURCE_DIR}/lib/mathExtensions.c
 ${H3_SOURCE_DIR}/lib/polygon.c
 ${H3_SOURCE_DIR}/lib/vec2d.c
 ${H3_SOURCE_DIR}/lib/vec3d.c
+${H3_SOURCE_DIR}/lib/vertex.c
 ${H3_SOURCE_DIR}/lib/vertexGraph.c
 )
 
