diff --git a/tests/queries/0_stateless/03302_any_enum_aggregation.reference b/tests/queries/0_stateless/03302_any_enum_aggregation.reference
new file mode 100644
index 000000000000..37d2e903991e
--- /dev/null
+++ b/tests/queries/0_stateless/03302_any_enum_aggregation.reference
@@ -0,0 +1,16 @@
+Issue 68605
+LOW
+LOW		0
+LOW		0
+Empty Enum8 table:
+MEDIUM
+MEDIUM
+Enum8 table with HIGH value:
+HIGH
+HIGH
+Empty Enum16 table:
+MEDIUM
+MEDIUM
+Enum16 table with HIGH value:
+HIGH
+HIGH
diff --git a/tests/queries/0_stateless/03302_any_enum_aggregation.sql b/tests/queries/0_stateless/03302_any_enum_aggregation.sql
new file mode 100644
index 000000000000..d1ac1deed34f
--- /dev/null
+++ b/tests/queries/0_stateless/03302_any_enum_aggregation.sql
@@ -0,0 +1,32 @@
+SELECT 'Issue 68605';
+DROP TABLE IF EXISTS test_33602;
+CREATE TABLE test_33602 (name String, score UInt8, user_level  Enum8('LOW' = 1, 'MEDIUM' = 2, 'HIGH' = 3)) ENGINE=Memory;
+SELECT any(user_level) FROM test_33602;
+SELECT any(user_level), any(name), any(score) FROM test_33602;
+SELECT anyLast(user_level), anyLast(name), anyLast(score) FROM test_33602;
+DROP TABLE test_33602;
+
+SELECT 'Empty Enum8 table:';
+DROP TABLE IF EXISTS test_33602_t0a;
+CREATE TABLE test_33602_t0a (e Enum8('LOW' = 123, 'MEDIUM' = 12, 'HIGH' = 33)) ENGINE=Memory;
+SELECT any(e) FROM test_33602_t0a;
+SELECT anyLast(e) FROM test_33602_t0a;
+
+SELECT 'Enum8 table with HIGH value:';
+INSERT INTO test_33602_t0a VALUES('HIGH');
+SELECT any(e) FROM test_33602_t0a;
+SELECT anyLast(e) FROM test_33602_t0a;
+DROP TABLE test_33602_t0a;
+
+
+SELECT 'Empty Enum16 table:';
+DROP TABLE IF EXISTS test_33602_t0b;
+CREATE TABLE test_33602_t0b (e Enum16('LOW' = 123, 'MEDIUM' = 12, 'HIGH' = 33)) ENGINE=Memory;
+SELECT any(e) FROM test_33602_t0b;
+SELECT anyLast(e) FROM test_33602_t0b;
+
+SELECT 'Enum16 table with HIGH value:';
+INSERT INTO test_33602_t0b VALUES('HIGH');
+SELECT any(e) FROM test_33602_t0b;
+SELECT anyLast(e) FROM test_33602_t0b;
+DROP TABLE test_33602_t0b;
