diff --git a/tests/queries/0_stateless/01198_plus_inf.reference b/tests/queries/0_stateless/01198_plus_inf.reference
new file mode 100644
index 000000000000..f726b8429b69
--- /dev/null
+++ b/tests/queries/0_stateless/01198_plus_inf.reference
@@ -0,0 +1,3 @@
+inf
+-inf
+inf
diff --git a/tests/queries/0_stateless/01198_plus_inf.sql b/tests/queries/0_stateless/01198_plus_inf.sql
new file mode 100644
index 000000000000..e06faa2fd21c
--- /dev/null
+++ b/tests/queries/0_stateless/01198_plus_inf.sql
@@ -0,0 +1,3 @@
+SELECT DISTINCT toFloat64(arrayJoin(['+inf', '+Inf', '+INF', '+infinity', '+Infinity']));
+SELECT DISTINCT toFloat64(arrayJoin(['-inf', '-Inf', '-INF', '-infinity', '-Infinity']));
+SELECT DISTINCT toFloat64(arrayJoin(['inf', 'Inf', 'INF', 'infinity', 'Infinity']));
