diff --git a/tests/queries/0_stateless/00754_distributed_optimize_skip_select_on_unused_shards.reference b/tests/queries/0_stateless/00754_distributed_optimize_skip_select_on_unused_shards.reference
index aa5098932309..add8c239ade3 100644
--- a/tests/queries/0_stateless/00754_distributed_optimize_skip_select_on_unused_shards.reference
+++ b/tests/queries/0_stateless/00754_distributed_optimize_skip_select_on_unused_shards.reference
@@ -1,7 +1,7 @@
 OK
 OK
 1
-FAIL
+OK
 0
 4
 2
@@ -9,8 +9,8 @@ FAIL
 1
 1
 4
-FAIL
-FAIL
-FAIL
-FAIL
-FAIL
+OK
+OK
+OK
+OK
+OK
diff --git a/tests/queries/0_stateless/00754_distributed_optimize_skip_select_on_unused_shards_with_prewhere.reference b/tests/queries/0_stateless/00754_distributed_optimize_skip_select_on_unused_shards_with_prewhere.reference
index 611f0fd25850..4c66ccfd2a20 100644
--- a/tests/queries/0_stateless/00754_distributed_optimize_skip_select_on_unused_shards_with_prewhere.reference
+++ b/tests/queries/0_stateless/00754_distributed_optimize_skip_select_on_unused_shards_with_prewhere.reference
@@ -1,15 +1,15 @@
 OK
 OK
 1
-FAIL
+OK
 0
 1
 4
 4
 2
 4
-FAIL
-FAIL
-FAIL
-FAIL
-FAIL
+OK
+OK
+OK
+OK
+OK
diff --git a/tests/queries/0_stateless/01339_client_unrecognized_option.sh b/tests/queries/0_stateless/01339_client_unrecognized_option.sh
index b3488d010c2d..0cfb705185e5 100755
--- a/tests/queries/0_stateless/01339_client_unrecognized_option.sh
+++ b/tests/queries/0_stateless/01339_client_unrecognized_option.sh
@@ -11,7 +11,7 @@ $CLICKHOUSE_CLIENT -xyzgarbage 2>&1 | grep -q "UNRECOGNIZED_ARGUMENTS" && echo '
 
 $CLICKHOUSE_CLIENT --xyzgarbage 2>&1 | grep -q "UNRECOGNIZED_ARGUMENTS" && echo 'OK' || echo 'FAIL'
 
-cat /etc/passwd | sed 's/:/\t/g' | $CLICKHOUSE_CLIENT --query="SELECT shell, count() AS c FROM passwd GROUP BY shell ORDER BY c DESC" --external --file=- --name=passwd --structure='login String, unused String, uid UInt16, gid UInt16, comment String, home String, shell String' xyzgarbage 2>&1 | grep -q "BAD_ARGUMENTS" && echo 'OK' || echo 'FAIL'
+cat /etc/passwd | sed 's/:/\t/g' | $CLICKHOUSE_CLIENT --query="SELECT shell, count() AS c FROM passwd GROUP BY shell ORDER BY c DESC" --external --file=- --name=passwd --structure='login String, unused String, uid UInt16, gid UInt16, comment String, home String, shell String' xyzgarbage 2>&1 | grep -q "SYNTAX_ERROR" && echo 'OK' || echo 'FAIL'
 
 cat /etc/passwd | sed 's/:/\t/g' | $CLICKHOUSE_CLIENT --query="SELECT shell, count() AS c FROM passwd GROUP BY shell ORDER BY c DESC" --external -xyzgarbage --file=- --name=passwd --structure='login String, unused String, uid UInt16, gid UInt16, comment String, home String, shell String' 2>&1 | grep -q "Bad arguments" && echo 'OK' || echo 'FAIL'
 
diff --git a/tests/queries/0_stateless/02751_multiquery_with_argument.reference b/tests/queries/0_stateless/02751_multiquery_with_argument.reference
index df9771b6bd3a..843fffb476ce 100644
--- a/tests/queries/0_stateless/02751_multiquery_with_argument.reference
+++ b/tests/queries/0_stateless/02751_multiquery_with_argument.reference
@@ -18,9 +18,6 @@ Bad arguments
 Bad arguments
 Bad arguments
 BAD_ARGUMENTS
-Bad arguments
 BAD_ARGUMENTS
 Bad arguments
 Bad arguments
-Bad arguments
-Bad arguments
diff --git a/tests/queries/0_stateless/02751_multiquery_with_argument.sh b/tests/queries/0_stateless/02751_multiquery_with_argument.sh
index ce53ede33319..d742cc0ad90e 100755
--- a/tests/queries/0_stateless/02751_multiquery_with_argument.sh
+++ b/tests/queries/0_stateless/02751_multiquery_with_argument.sh
@@ -30,9 +30,6 @@ $CLICKHOUSE_LOCAL -n --multiquery "SELECT 307; SELECT 308;" 2>&1 | grep -o 'Bad
 $CLICKHOUSE_LOCAL --multiquery "SELECT 309; SELECT 310;" --multiquery  2>&1 | grep -o 'Bad arguments'
 $CLICKHOUSE_LOCAL --multiquery "SELECT 311;" --multiquery "SELECT 312;" 2>&1 | grep -o 'Bad arguments'
 $CLICKHOUSE_LOCAL --multiquery "SELECT 313;" -n "SELECT 314;" 2>&1 | grep -o 'BAD_ARGUMENTS'
-$CLICKHOUSE_LOCAL --multiquery "SELECT 315;" --query "SELECT 316;" 2>&1 | grep -o 'Bad arguments'
 $CLICKHOUSE_LOCAL -n "SELECT 320" --query "SELECT 317;" 2>&1 | grep -o 'BAD_ARGUMENTS'
-$CLICKHOUSE_LOCAL --query --multiquery --multiquery "SELECT 318;" 2>&1 | grep -o 'Bad arguments'
-$CLICKHOUSE_LOCAL --query --multiquery "SELECT 319;" 2>&1 | grep -o 'Bad arguments'
 $CLICKHOUSE_LOCAL --query -n "SELECT 400;" 2>&1 | grep -o 'Bad arguments'
-$CLICKHOUSE_LOCAL --query -n --multiquery "SELECT 401;" 2>&1 | grep -o 'Bad arguments'
\ No newline at end of file
+$CLICKHOUSE_LOCAL --query -n --multiquery "SELECT 401;" 2>&1 | grep -o 'Bad arguments'
diff --git a/tests/queries/0_stateless/02771_multiple_query_arguments.reference b/tests/queries/0_stateless/02771_multiple_query_arguments.reference
new file mode 100644
index 000000000000..5cad23947c86
--- /dev/null
+++ b/tests/queries/0_stateless/02771_multiple_query_arguments.reference
@@ -0,0 +1,17 @@
+101
+101
+202
+202
+Multi-statements are not allowed
+Empty query
+Bad arguments
+Syntax error
+101
+101
+202
+202
+303
+303
+303
+Bad arguments
+Syntax error
diff --git a/tests/queries/0_stateless/02771_multiple_query_arguments.sh b/tests/queries/0_stateless/02771_multiple_query_arguments.sh
new file mode 100755
index 000000000000..435e0a333155
--- /dev/null
+++ b/tests/queries/0_stateless/02771_multiple_query_arguments.sh
@@ -0,0 +1,21 @@
+#!/usr/bin/env bash
+
+CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+# shellcheck source=../shell_config.sh
+. "$CURDIR"/../shell_config.sh
+
+# clickhouse-client
+$CLICKHOUSE_CLIENT --query "SELECT 101" --query "SELECT 101"
+$CLICKHOUSE_CLIENT --query "SELECT 202;" --query "SELECT 202;"
+$CLICKHOUSE_CLIENT --query "SELECT 303" --query "SELECT 303; SELECT 303" 2>&1 | grep -o 'Multi-statements are not allowed'
+$CLICKHOUSE_CLIENT --query "" --query "" 2>&1 | grep -o 'Empty query'
+$CLICKHOUSE_CLIENT --query "SELECT 303" --query 2>&1 | grep -o 'Bad arguments'
+$CLICKHOUSE_CLIENT --query "SELECT 303" --query "SELE" 2>&1 | grep -o 'Syntax error'
+
+# clickhouse-local
+$CLICKHOUSE_LOCAL --query "SELECT 101" --query "SELECT 101"
+$CLICKHOUSE_LOCAL --query "SELECT 202;" --query "SELECT 202;"
+$CLICKHOUSE_LOCAL --query "SELECT 303" --query "SELECT 303; SELECT 303" 2>&1 # behaves differently than clickhouse-client, TODO make it consistent
+$CLICKHOUSE_LOCAL --query "" --query "" 2>&1 # behaves equally different than clickhouse-client TODO
+$CLICKHOUSE_LOCAL --query "SELECT 303" --query 2>&1 | grep -o 'Bad arguments'
+$CLICKHOUSE_LOCAL --query "SELECT 303" --query "SELE" 2>&1 | grep -o 'Syntax error'
