diff --git a/tests/queries/0_stateless/03322_nan_in_aggregator.reference b/tests/queries/0_stateless/03322_nan_in_aggregator.reference
new file mode 100644
index 000000000000..9ef72f6f37ce
--- /dev/null
+++ b/tests/queries/0_stateless/03322_nan_in_aggregator.reference
@@ -0,0 +1,2 @@
+500
+500
diff --git a/tests/queries/0_stateless/03322_nan_in_aggregator.sql b/tests/queries/0_stateless/03322_nan_in_aggregator.sql
new file mode 100644
index 000000000000..7afa6bac66f1
--- /dev/null
+++ b/tests/queries/0_stateless/03322_nan_in_aggregator.sql
@@ -0,0 +1,9 @@
+SET send_logs_level='fatal';
+
+-- max_bytes_ratio_before_external_group_by = nan - regarding fix in Aggregator.cpp, 
+-- max_bytes_ratio_before_external_sort = nan - regarding fix in SortingStep.cpp
+SELECT uniqExact(CAST(number, 'String'))
+FROM numbers(1000.)
+GROUP BY CAST(number % 2, 'String')
+ORDER BY uniqExact(CAST(number, 'String')) ASC NULLS FIRST
+SETTINGS max_bytes_ratio_before_external_group_by = nan, max_bytes_ratio_before_external_sort = nan;
