diff --git a/src/Interpreters/OptimizeIfWithConstantConditionVisitor.cpp b/src/Interpreters/OptimizeIfWithConstantConditionVisitor.cpp
index 993c9945039d..0440c52797c4 100644
--- a/src/Interpreters/OptimizeIfWithConstantConditionVisitor.cpp
+++ b/src/Interpreters/OptimizeIfWithConstantConditionVisitor.cpp
@@ -28,6 +28,8 @@ static bool tryExtractConstValueFromCondition(const ASTPtr & condition, bool & v
     }
 
     /// cast of numeric constant in condition to UInt8
+    /// Note: this solution is ad-hoc and only implemented for yandex.metrica use case.
+    /// We should allow any constant condition (or maybe remove this optimization completely) later.
     if (const auto * function = condition->as<ASTFunction>())
     {
         if (isFunctionCast(function))
@@ -49,6 +51,16 @@ static bool tryExtractConstValueFromCondition(const ASTPtr & condition, bool & v
                 }
             }
         }
+        else if (function->name == "toUInt8" || function->name == "toInt8")
+        {
+            if (const auto * expr_list = function->arguments->as<ASTExpressionList>())
+            {
+                if (expr_list->children.size() != 1)
+                    throw Exception(ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH, "Function {} must have exactly two arguments", function->name);
+
+                return tryExtractConstValueFromCondition(expr_list->children.at(0), value);
+            }
+        }
     }
 
     return false;
