diff --git a/tests/queries/0_stateless/25402_projection_and_ast_optimizations_bug.reference b/tests/queries/0_stateless/25402_projection_and_ast_optimizations_bug.reference
new file mode 100644
index 000000000000..9049324c3921
--- /dev/null
+++ b/tests/queries/0_stateless/25402_projection_and_ast_optimizations_bug.reference
@@ -0,0 +1,1 @@
+-2.5574077246549023	0.6663667453928805	1
diff --git a/tests/queries/0_stateless/25402_projection_and_ast_optimizations_bug.sql b/tests/queries/0_stateless/25402_projection_and_ast_optimizations_bug.sql
new file mode 100644
index 000000000000..5589fbeeb9e1
--- /dev/null
+++ b/tests/queries/0_stateless/25402_projection_and_ast_optimizations_bug.sql
@@ -0,0 +1,6 @@
+drop table if exists t1;
+CREATE TABLE t1 (c0 Int32) ENGINE = MergeTree() ORDER BY c0 PARTITION BY (- (c0));
+insert into t1 values(1);
+SELECT (- ((((tan (t1.c0)))+(t1.c0)))), (cos ((sin (pow(t1.c0,t1.c0))))), ((gcd((- (t1.c0)),((t1.c0)+(t1.c0))))*((- ((- (t1.c0)))))) FROM t1 GROUP BY (sqrt ((- (t1.c0)))), t1.c0, pow((erf ((- (t1.c0)))),t1.c0);
+drop table t1;
+
