diff --git a/tests/queries/0_stateless/01557_max_parallel_replicas_no_sample.reference b/tests/queries/0_stateless/01557_max_parallel_replicas_no_sample.reference
new file mode 100644
index 000000000000..2715babfff20
--- /dev/null
+++ b/tests/queries/0_stateless/01557_max_parallel_replicas_no_sample.reference
@@ -0,0 +1,4 @@
+Hello
+1000
+1000
+1000
diff --git a/tests/queries/0_stateless/01557_max_parallel_replicas_no_sample.sql b/tests/queries/0_stateless/01557_max_parallel_replicas_no_sample.sql
new file mode 100644
index 000000000000..d86f692a1eaa
--- /dev/null
+++ b/tests/queries/0_stateless/01557_max_parallel_replicas_no_sample.sql
@@ -0,0 +1,22 @@
+DROP TABLE IF EXISTS t;
+CREATE TABLE t (x String) ENGINE = MergeTree ORDER BY x;
+INSERT INTO t VALUES ('Hello');
+
+SET max_parallel_replicas = 3;
+SELECT * FROM remote('127.0.0.{2|3|4}', currentDatabase(), t);
+
+DROP TABLE t;
+
+CREATE TABLE t (x String) ENGINE = MergeTree ORDER BY cityHash64(x) SAMPLE BY cityHash64(x);
+INSERT INTO t SELECT toString(number) FROM numbers(1000);
+
+SET max_parallel_replicas = 1;
+SELECT count() FROM remote('127.0.0.{2|3|4}', currentDatabase(), t);
+
+SET max_parallel_replicas = 2;
+SELECT count() FROM remote('127.0.0.{2|3|4}', currentDatabase(), t);
+
+SET max_parallel_replicas = 3;
+SELECT count() FROM remote('127.0.0.{2|3|4}', currentDatabase(), t);
+
+DROP TABLE t;
diff --git a/tests/queries/0_stateless/arcadia_skip_list.txt b/tests/queries/0_stateless/arcadia_skip_list.txt
index 8ece63c419cc..900cc82b33f8 100644
--- a/tests/queries/0_stateless/arcadia_skip_list.txt
+++ b/tests/queries/0_stateless/arcadia_skip_list.txt
@@ -159,3 +159,4 @@
 01547_query_log_current_database
 01548_query_log_query_execution_ms
 01552_dict_fixedstring
+01557_max_parallel_replicas_no_sample.sql
