diff --git a/tests/queries/0_stateless/02503_in_lc_const_args_bug.reference b/tests/queries/0_stateless/02503_in_lc_const_args_bug.reference
new file mode 100644
index 000000000000..8baef1b4abc4
--- /dev/null
+++ b/tests/queries/0_stateless/02503_in_lc_const_args_bug.reference
@@ -0,0 +1,1 @@
+abc
diff --git a/tests/queries/0_stateless/02503_in_lc_const_args_bug.sql b/tests/queries/0_stateless/02503_in_lc_const_args_bug.sql
new file mode 100644
index 000000000000..6756e3815867
--- /dev/null
+++ b/tests/queries/0_stateless/02503_in_lc_const_args_bug.sql
@@ -0,0 +1,2 @@
+SELECT substr(toLowCardinality('abc'), 1 in 1) AS x GROUP BY x;
+
