diff --git a/tests/queries/0_stateless/01774_tuple_null_in.reference b/tests/queries/0_stateless/01774_tuple_null_in.reference
new file mode 100644
index 000000000000..aa47d0d46d47
--- /dev/null
+++ b/tests/queries/0_stateless/01774_tuple_null_in.reference
@@ -0,0 +1,2 @@
+0
+0
diff --git a/tests/queries/0_stateless/01774_tuple_null_in.sql b/tests/queries/0_stateless/01774_tuple_null_in.sql
new file mode 100644
index 000000000000..a9cc39e8840c
--- /dev/null
+++ b/tests/queries/0_stateless/01774_tuple_null_in.sql
@@ -0,0 +1,2 @@
+SELECT (NULL, NULL) = (8, 0) OR (NULL, NULL) = (3, 2) OR (NULL, NULL) = (0, 0) OR (NULL, NULL) = (3, 1);
+SELECT (NULL, NULL) IN ((NULL, 0), (3, 1), (3, 2), (8, 0), (NULL, NULL));
