diff --git a/src/server/dragonfly_test.cc b/src/server/dragonfly_test.cc
index 739ce25f4263..25308541c173 100644
--- a/src/server/dragonfly_test.cc
+++ b/src/server/dragonfly_test.cc
@@ -262,17 +262,19 @@ TEST_F(DflyEngineTest, Hello) {
   resp = Run({"hello", "2"});
   ASSERT_THAT(resp, ArrLen(14));
 
-  EXPECT_THAT(resp.GetVec(),
-              ElementsAre("server", "redis", "version", "6.2.11", "dragonfly_version",
-                          ArgType(RespExpr::STRING), "proto", IntArg(2), "id",
-                          ArgType(RespExpr::INT64), "mode", "standalone", "role", "master"));
+  EXPECT_THAT(
+      resp.GetVec(),
+      ElementsAre("server", "redis", "version", "6.2.11", "dragonfly_version",
+                  ArgType(RespExpr::STRING), "proto", IntArg(2), "id", ArgType(RespExpr::INT64),
+                  "mode", testing::AnyOf("standalone", "cluster"), "role", "master"));
 
   resp = Run({"hello", "3"});
   ASSERT_THAT(resp, ArrLen(14));
-  EXPECT_THAT(resp.GetVec(),
-              ElementsAre("server", "redis", "version", "6.2.11", "dragonfly_version",
-                          ArgType(RespExpr::STRING), "proto", IntArg(3), "id",
-                          ArgType(RespExpr::INT64), "mode", "standalone", "role", "master"));
+  EXPECT_THAT(
+      resp.GetVec(),
+      ElementsAre("server", "redis", "version", "6.2.11", "dragonfly_version",
+                  ArgType(RespExpr::STRING), "proto", IntArg(3), "id", ArgType(RespExpr::INT64),
+                  "mode", testing::AnyOf("standalone", "cluster"), "role", "master"));
 
   EXPECT_THAT(Run({"hello", "2", "AUTH", "uname", "pwd"}),
               ErrArg("WRONGPASS invalid username-password pair or user is disabled."));
