diff --git a/tests/queries/0_stateless/03513_filter_push_down_rand_bug.reference b/tests/queries/0_stateless/03513_filter_push_down_rand_bug.reference
new file mode 100644
index 000000000000..573541ac9702
--- /dev/null
+++ b/tests/queries/0_stateless/03513_filter_push_down_rand_bug.reference
@@ -0,0 +1,1 @@
+0
diff --git a/tests/queries/0_stateless/03513_filter_push_down_rand_bug.sql b/tests/queries/0_stateless/03513_filter_push_down_rand_bug.sql
new file mode 100644
index 000000000000..972160b5c81c
--- /dev/null
+++ b/tests/queries/0_stateless/03513_filter_push_down_rand_bug.sql
@@ -0,0 +1,21 @@
+drop table if exists users_items;
+CREATE TABLE users_items (user_id UInt64) ENGINE = Log;
+INSERT INTO users_items SELECT bitAnd(number, 15) from numbers(64);
+
+SELECT sum(in_sample)
+FROM
+(
+    WITH RandomUsers AS
+        (
+            SELECT
+                user_id,
+                rand() % 2 AS in_sample
+            FROM users_items
+            GROUP BY user_id
+        )
+    SELECT
+        user_id,
+        in_sample
+    FROM RandomUsers
+    WHERE in_sample = 0
+);
