diff --git a/tests/queries/0_stateless/01389_filter_by_virtual_columns.reference b/tests/queries/0_stateless/01389_filter_by_virtual_columns.reference
new file mode 100644
index 000000000000..573541ac9702
--- /dev/null
+++ b/tests/queries/0_stateless/01389_filter_by_virtual_columns.reference
@@ -0,0 +1,1 @@
+0
diff --git a/tests/queries/0_stateless/01389_filter_by_virtual_columns.sql b/tests/queries/0_stateless/01389_filter_by_virtual_columns.sql
new file mode 100644
index 000000000000..43ce3ad40e5f
--- /dev/null
+++ b/tests/queries/0_stateless/01389_filter_by_virtual_columns.sql
@@ -0,0 +1,2 @@
+SELECT count() FROM system.parts WHERE table = NULL AND database = currentDatabase();
+SELECT DISTINCT marks FROM system.parts WHERE (table = NULL) AND (database = currentDatabase()) AND (active = 1);
