diff --git a/tests/queries/0_stateless/03008_deduplication_cases_from_docs.sql b/tests/queries/0_stateless/03008_deduplication_cases_from_docs.sql
index 7927a6b1edf9..1bd7ef09c356 100644
--- a/tests/queries/0_stateless/03008_deduplication_cases_from_docs.sql
+++ b/tests/queries/0_stateless/03008_deduplication_cases_from_docs.sql
@@ -1,4 +1,3 @@
--- #########
 select 'Different materialized view insert into one underlayed table equal data.';
 
 DROP TABLE IF EXISTS dst;
@@ -81,8 +80,6 @@ DROP TABLE mv_first;
 DROP TABLE mv_dst;
 DROP TABLE dst;
 
-
--- #########
 select 'Different insert operations generate the same data after transformation in underlied table of materialized view.';
 
 DROP TABLE IF EXISTS dst;
@@ -151,8 +148,6 @@ ORDER by all;
 DROP TABLE mv_dst;
 DROP TABLE dst;
 
-
--- #########
 select 'Indentical blocks in insertion with `insert_deduplication_token`';
 
 DROP TABLE IF EXISTS dst;
@@ -217,8 +212,6 @@ ORDER by all;
 
 DROP TABLE dst;
 
-
--- #########
 select 'Indentical blocks in insertion';
 
 DROP TABLE IF EXISTS dst;
@@ -250,8 +243,6 @@ ORDER by all;
 
 DROP TABLE dst;
 
-
--- #########
 select 'Indentical blocks after materialised view`s transformation';
 
 DROP TABLE IF EXISTS dst;
