diff --git a/test/arrow/parquet_test.cpp b/test/arrow/parquet_test.cpp
index cb31760d6432..b70f927feeef 100644
--- a/test/arrow/parquet_test.cpp
+++ b/test/arrow/parquet_test.cpp
@@ -152,7 +152,6 @@ TEST_CASE("Test Parquet Long Files", "[arrow]") {
 }
 
 TEST_CASE("Test Parquet Files", "[arrow]") {
-
 	std::vector<std::string> skip {"aws2.parquet"};    //! Not supported by arrow
 	skip.emplace_back("datapage_v2.snappy.parquet");   //! Not supported by arrow
 	skip.emplace_back("broken-arrow.parquet");         //! Arrow can't read this
@@ -160,6 +159,8 @@ TEST_CASE("Test Parquet Files", "[arrow]") {
 	skip.emplace_back("fixed.parquet");                //! Can't roundtrip Fixed-size Binaries
 	skip.emplace_back("leftdate3_192_loop_1.parquet"); //! This is just crazy slow
 	skip.emplace_back("bug687_nulls.parquet");         //! This is just crazy slow
+	skip.emplace_back("nullbyte.parquet");             //! Null byte in file
+	skip.emplace_back("nullbyte_multiple.parquet");    //! Null byte in file
 
 	duckdb::DuckDB db;
 	duckdb::Connection conn {db};
diff --git a/test/sql/copy/parquet/data/nullbyte.parquet b/test/sql/copy/parquet/data/nullbyte.parquet
new file mode 100644
index 000000000000..4a478dd01d54
Binary files /dev/null and b/test/sql/copy/parquet/data/nullbyte.parquet differ
diff --git a/test/sql/copy/parquet/data/nullbyte_multiple.parquet b/test/sql/copy/parquet/data/nullbyte_multiple.parquet
new file mode 100644
index 000000000000..6fcf7b42167d
Binary files /dev/null and b/test/sql/copy/parquet/data/nullbyte_multiple.parquet differ
diff --git a/test/sql/copy/parquet/parquet_nullbyte.test b/test/sql/copy/parquet/parquet_nullbyte.test
new file mode 100644
index 000000000000..4f84d0a899f0
--- /dev/null
+++ b/test/sql/copy/parquet/parquet_nullbyte.test
@@ -0,0 +1,20 @@
+# name: test/sql/copy/parquet/parquet_nullbyte.test
+# description: Test reading parquet files with null bytes in strings
+# group: [parquet]
+
+require parquet
+
+statement ok
+PRAGMA enable_verification
+
+query II
+select * from parquet_scan('test/sql/copy/parquet/data/nullbyte.parquet')
+----
+42	hello
+
+query II
+select * from parquet_scan('test/sql/copy/parquet/data/nullbyte_multiple.parquet')
+----
+1	hello
+2	this is a long
+3	this string has no null bytes
