diff --git a/tests/queries/0_stateless/01123_parse_date_time_best_effort_even_more.reference b/tests/queries/0_stateless/01123_parse_date_time_best_effort_even_more.reference
new file mode 100644
index 000000000000..558ba34abcdb
--- /dev/null
+++ b/tests/queries/0_stateless/01123_parse_date_time_best_effort_even_more.reference
@@ -0,0 +1,2 @@
+2018-08-18 07:22:16
+2018-08-16 07:22:16
diff --git a/tests/queries/0_stateless/01123_parse_date_time_best_effort_even_more.sql b/tests/queries/0_stateless/01123_parse_date_time_best_effort_even_more.sql
new file mode 100644
index 000000000000..a4f6f1734026
--- /dev/null
+++ b/tests/queries/0_stateless/01123_parse_date_time_best_effort_even_more.sql
@@ -0,0 +1,2 @@
+SELECT toTimeZone(parseDateTimeBestEffort('Thu, 18 Aug 2018 07:22:16 GMT'), 'UTC');
+SELECT toTimeZone(parseDateTimeBestEffort('Tue, 16 Aug 2018 07:22:16 GMT'), 'UTC');
