diff --git a/dbms/tests/queries/0_stateless/01060_defaults_all_columns.reference b/dbms/tests/queries/0_stateless/01060_defaults_all_columns.reference
new file mode 100644
index 000000000000..68b4657ca606
--- /dev/null
+++ b/dbms/tests/queries/0_stateless/01060_defaults_all_columns.reference
@@ -0,0 +1,4 @@
+1	hello
+2	test42
+42	test42
+42	world
diff --git a/dbms/tests/queries/0_stateless/01060_defaults_all_columns.sql b/dbms/tests/queries/0_stateless/01060_defaults_all_columns.sql
new file mode 100644
index 000000000000..afbb01b8cb25
--- /dev/null
+++ b/dbms/tests/queries/0_stateless/01060_defaults_all_columns.sql
@@ -0,0 +1,10 @@
+DROP TABLE IF EXISTS defaults_all_columns;
+
+CREATE TABLE defaults_all_columns (n UInt8 DEFAULT 42, s String DEFAULT concat('test', CAST(n, 'String'))) ENGINE = Memory;
+
+INSERT INTO defaults_all_columns FORMAT JSONEachRow {"n": 1, "s": "hello"} {};
+INSERT INTO defaults_all_columns FORMAT JSONEachRow {"n": 2}, {"s": "world"};
+
+SELECT * FROM defaults_all_columns ORDER BY n, s;
+
+DROP TABLE defaults_all_columns;
diff --git a/dbms/tests/queries/bugs/01060_defaults_all_columns.reference b/dbms/tests/queries/bugs/01060_defaults_all_columns.reference
new file mode 100644
index 000000000000..7b1fdfb68171
--- /dev/null
+++ b/dbms/tests/queries/bugs/01060_defaults_all_columns.reference
@@ -0,0 +1,4 @@
+1	hello
+2	test2
+42	test42
+42	world
