diff --git a/tests/queries/0_stateless/02030_quantiles_underflow.reference b/tests/queries/0_stateless/02030_quantiles_underflow.reference
new file mode 100644
index 000000000000..907330658e6d
--- /dev/null
+++ b/tests/queries/0_stateless/02030_quantiles_underflow.reference
@@ -0,0 +1,1 @@
+[-1717986918.2,1.8]
diff --git a/tests/queries/0_stateless/02030_quantiles_underflow.sql b/tests/queries/0_stateless/02030_quantiles_underflow.sql
new file mode 100644
index 000000000000..cb5ad3dac871
--- /dev/null
+++ b/tests/queries/0_stateless/02030_quantiles_underflow.sql
@@ -0,0 +1,6 @@
+SELECT
+    arrayMap(y -> round(y, 1), quantilesExactInclusive(0.1, 0.9)(x)) AS q
+FROM
+(
+    SELECT arrayJoin([-2147483648, 1, 2]) AS x
+);
