diff --git a/test/sql/parallelism/interquery/concurrent_update_drop.test_slow b/test/sql/parallelism/interquery/concurrent_update_drop.test_slow
new file mode 100644
index 000000000000..845558d1a967
--- /dev/null
+++ b/test/sql/parallelism/interquery/concurrent_update_drop.test_slow
@@ -0,0 +1,21 @@
+# name: test/sql/parallelism/interquery/concurrent_update_drop.test_slow
+# description: Test concurrent updates and drops
+# group: [interquery]
+
+statement ok
+CREATE TABLE t1(i INTEGER)
+
+statement ok
+INSERT INTO t1 VALUES (1), (2), (3)
+
+concurrentloop threadid 0 2
+
+onlyif threadid=0
+statement ok
+UPDATE t1 SET i = 4 WHERE i = 2
+
+onlyif threadid=1
+statement ok
+DROP TABLE t1
+
+endloop
\ No newline at end of file
