diff --git a/tests/queries/0_stateless/00493_substring_of_enum.reference b/tests/queries/0_stateless/00493_substring_of_enum.reference
new file mode 100644
index 000000000000..17bc960b3182
--- /dev/null
+++ b/tests/queries/0_stateless/00493_substring_of_enum.reference
@@ -0,0 +1,155 @@
+-- Positive offsets (slice from left)
+Offset: 	1	Length: 	1	hello	shark	h	s
+Offset: 	1	Length: 	2	hello	shark	he	sh
+Offset: 	1	Length: 	3	hello	shark	hel	sha
+Offset: 	1	Length: 	4	hello	shark	hell	shar
+Offset: 	1	Length: 	5	hello	shark	hello	shark
+Offset: 	1	Length: 	6	hello	shark	hello	shark
+Offset: 	2	Length: 	1	ello	hark	e	h
+Offset: 	2	Length: 	2	ello	hark	el	ha
+Offset: 	2	Length: 	3	ello	hark	ell	har
+Offset: 	2	Length: 	4	ello	hark	ello	hark
+Offset: 	2	Length: 	5	ello	hark	ello	hark
+Offset: 	2	Length: 	6	ello	hark	ello	hark
+Offset: 	3	Length: 	1	llo	ark	l	a
+Offset: 	3	Length: 	2	llo	ark	ll	ar
+Offset: 	3	Length: 	3	llo	ark	llo	ark
+Offset: 	3	Length: 	4	llo	ark	llo	ark
+Offset: 	3	Length: 	5	llo	ark	llo	ark
+Offset: 	3	Length: 	6	llo	ark	llo	ark
+Offset: 	4	Length: 	1	lo	rk	l	r
+Offset: 	4	Length: 	2	lo	rk	lo	rk
+Offset: 	4	Length: 	3	lo	rk	lo	rk
+Offset: 	4	Length: 	4	lo	rk	lo	rk
+Offset: 	4	Length: 	5	lo	rk	lo	rk
+Offset: 	4	Length: 	6	lo	rk	lo	rk
+Offset: 	5	Length: 	1	o	k	o	k
+Offset: 	5	Length: 	2	o	k	o	k
+Offset: 	5	Length: 	3	o	k	o	k
+Offset: 	5	Length: 	4	o	k	o	k
+Offset: 	5	Length: 	5	o	k	o	k
+Offset: 	5	Length: 	6	o	k	o	k
+Offset: 	6	Length: 	1				
+Offset: 	6	Length: 	2				
+Offset: 	6	Length: 	3				
+Offset: 	6	Length: 	4				
+Offset: 	6	Length: 	5				
+Offset: 	6	Length: 	6				
+Offset: 	1	Length: 	1	world	eagle	w	e
+Offset: 	1	Length: 	2	world	eagle	wo	ea
+Offset: 	1	Length: 	3	world	eagle	wor	eag
+Offset: 	1	Length: 	4	world	eagle	worl	eagl
+Offset: 	1	Length: 	5	world	eagle	world	eagle
+Offset: 	1	Length: 	6	world	eagle	world	eagle
+Offset: 	2	Length: 	1	orld	agle	o	a
+Offset: 	2	Length: 	2	orld	agle	or	ag
+Offset: 	2	Length: 	3	orld	agle	orl	agl
+Offset: 	2	Length: 	4	orld	agle	orld	agle
+Offset: 	2	Length: 	5	orld	agle	orld	agle
+Offset: 	2	Length: 	6	orld	agle	orld	agle
+Offset: 	3	Length: 	1	rld	gle	r	g
+Offset: 	3	Length: 	2	rld	gle	rl	gl
+Offset: 	3	Length: 	3	rld	gle	rld	gle
+Offset: 	3	Length: 	4	rld	gle	rld	gle
+Offset: 	3	Length: 	5	rld	gle	rld	gle
+Offset: 	3	Length: 	6	rld	gle	rld	gle
+Offset: 	4	Length: 	1	ld	le	l	l
+Offset: 	4	Length: 	2	ld	le	ld	le
+Offset: 	4	Length: 	3	ld	le	ld	le
+Offset: 	4	Length: 	4	ld	le	ld	le
+Offset: 	4	Length: 	5	ld	le	ld	le
+Offset: 	4	Length: 	6	ld	le	ld	le
+Offset: 	5	Length: 	1	d	e	d	e
+Offset: 	5	Length: 	2	d	e	d	e
+Offset: 	5	Length: 	3	d	e	d	e
+Offset: 	5	Length: 	4	d	e	d	e
+Offset: 	5	Length: 	5	d	e	d	e
+Offset: 	5	Length: 	6	d	e	d	e
+Offset: 	6	Length: 	1				
+Offset: 	6	Length: 	2				
+Offset: 	6	Length: 	3				
+Offset: 	6	Length: 	4				
+Offset: 	6	Length: 	5				
+Offset: 	6	Length: 	6				
+-- Negative offsets (slice from right)
+Offset: 	-1	Length: 	1	o	k	o	k
+Offset: 	-1	Length: 	2	o	k	o	k
+Offset: 	-1	Length: 	3	o	k	o	k
+Offset: 	-1	Length: 	4	o	k	o	k
+Offset: 	-1	Length: 	5	o	k	o	k
+Offset: 	-1	Length: 	6	o	k	o	k
+Offset: 	-2	Length: 	1	lo	rk	l	r
+Offset: 	-2	Length: 	2	lo	rk	lo	rk
+Offset: 	-2	Length: 	3	lo	rk	lo	rk
+Offset: 	-2	Length: 	4	lo	rk	lo	rk
+Offset: 	-2	Length: 	5	lo	rk	lo	rk
+Offset: 	-2	Length: 	6	lo	rk	lo	rk
+Offset: 	-3	Length: 	1	llo	ark	l	a
+Offset: 	-3	Length: 	2	llo	ark	ll	ar
+Offset: 	-3	Length: 	3	llo	ark	llo	ark
+Offset: 	-3	Length: 	4	llo	ark	llo	ark
+Offset: 	-3	Length: 	5	llo	ark	llo	ark
+Offset: 	-3	Length: 	6	llo	ark	llo	ark
+Offset: 	-4	Length: 	1	ello	hark	e	h
+Offset: 	-4	Length: 	2	ello	hark	el	ha
+Offset: 	-4	Length: 	3	ello	hark	ell	har
+Offset: 	-4	Length: 	4	ello	hark	ello	hark
+Offset: 	-4	Length: 	5	ello	hark	ello	hark
+Offset: 	-4	Length: 	6	ello	hark	ello	hark
+Offset: 	-5	Length: 	1	hello	shark	h	s
+Offset: 	-5	Length: 	2	hello	shark	he	sh
+Offset: 	-5	Length: 	3	hello	shark	hel	sha
+Offset: 	-5	Length: 	4	hello	shark	hell	shar
+Offset: 	-5	Length: 	5	hello	shark	hello	shark
+Offset: 	-5	Length: 	6	hello	shark	hello	shark
+Offset: 	-6	Length: 	1	hello	shark		
+Offset: 	-6	Length: 	2	hello	shark	h	s
+Offset: 	-6	Length: 	3	hello	shark	he	sh
+Offset: 	-6	Length: 	4	hello	shark	hel	sha
+Offset: 	-6	Length: 	5	hello	shark	hell	shar
+Offset: 	-6	Length: 	6	hello	shark	hello	shark
+Offset: 	-1	Length: 	1	d	e	d	e
+Offset: 	-1	Length: 	2	d	e	d	e
+Offset: 	-1	Length: 	3	d	e	d	e
+Offset: 	-1	Length: 	4	d	e	d	e
+Offset: 	-1	Length: 	5	d	e	d	e
+Offset: 	-1	Length: 	6	d	e	d	e
+Offset: 	-2	Length: 	1	ld	le	l	l
+Offset: 	-2	Length: 	2	ld	le	ld	le
+Offset: 	-2	Length: 	3	ld	le	ld	le
+Offset: 	-2	Length: 	4	ld	le	ld	le
+Offset: 	-2	Length: 	5	ld	le	ld	le
+Offset: 	-2	Length: 	6	ld	le	ld	le
+Offset: 	-3	Length: 	1	rld	gle	r	g
+Offset: 	-3	Length: 	2	rld	gle	rl	gl
+Offset: 	-3	Length: 	3	rld	gle	rld	gle
+Offset: 	-3	Length: 	4	rld	gle	rld	gle
+Offset: 	-3	Length: 	5	rld	gle	rld	gle
+Offset: 	-3	Length: 	6	rld	gle	rld	gle
+Offset: 	-4	Length: 	1	orld	agle	o	a
+Offset: 	-4	Length: 	2	orld	agle	or	ag
+Offset: 	-4	Length: 	3	orld	agle	orl	agl
+Offset: 	-4	Length: 	4	orld	agle	orld	agle
+Offset: 	-4	Length: 	5	orld	agle	orld	agle
+Offset: 	-4	Length: 	6	orld	agle	orld	agle
+Offset: 	-5	Length: 	1	world	eagle	w	e
+Offset: 	-5	Length: 	2	world	eagle	wo	ea
+Offset: 	-5	Length: 	3	world	eagle	wor	eag
+Offset: 	-5	Length: 	4	world	eagle	worl	eagl
+Offset: 	-5	Length: 	5	world	eagle	world	eagle
+Offset: 	-5	Length: 	6	world	eagle	world	eagle
+Offset: 	-6	Length: 	1	world	eagle		
+Offset: 	-6	Length: 	2	world	eagle	w	e
+Offset: 	-6	Length: 	3	world	eagle	wo	ea
+Offset: 	-6	Length: 	4	world	eagle	wor	eag
+Offset: 	-6	Length: 	5	world	eagle	worl	eagl
+Offset: 	-6	Length: 	6	world	eagle	world	eagle
+-- Zero offset/length
+Offset: 	0	Length: 	0				
+Offset: 	0	Length: 	1				
+Offset: 	1	Length: 	0	hello	shark		
+Offset: 	0	Length: 	0				
+Offset: 	0	Length: 	1				
+Offset: 	1	Length: 	0	world	eagle		
+-- Constant enums
+f	fo
diff --git a/tests/queries/0_stateless/00493_substring_of_enum.sql b/tests/queries/0_stateless/00493_substring_of_enum.sql
new file mode 100644
index 000000000000..cdda76e88a7f
--- /dev/null
+++ b/tests/queries/0_stateless/00493_substring_of_enum.sql
@@ -0,0 +1,32 @@
+DROP TABLE IF EXISTS tab;
+CREATE TABLE tab(e8 Enum8('hello' = -5, 'world' = 15), e16 Enum16('shark' = -999, 'eagle' = 9999)) ENGINE MergeTree ORDER BY tuple();
+INSERT INTO TABLE tab VALUES ('hello', 'shark'), ('world', 'eagle');
+
+SELECT '-- Positive offsets (slice from left)';
+WITH cte AS (SELECT number + 1 AS n FROM system.numbers LIMIT 6),
+     permutations AS (SELECT c1.n AS offset, c2.n AS length FROM cte AS c1 CROSS JOIN cte AS c2)
+SELECT 'Offset: ', p.offset, 'Length: ', p.length,
+       substring(e8, p.offset) AS s1, substring(e16, p.offset) AS s2,
+       substring(e8, p.offset, p.length) AS s3, substring(e16, p.offset, p.length) AS s4
+FROM tab LEFT JOIN permutations AS p ON true;
+
+SELECT '-- Negative offsets (slice from right)';
+WITH cte AS (SELECT number + 1 AS n FROM system.numbers LIMIT 6),
+     permutations AS (SELECT -c1.n AS offset, c2.n AS length FROM cte AS c1 CROSS JOIN cte AS c2)
+SELECT 'Offset: ', p.offset, 'Length: ', p.length,
+       substring(e8, p.offset) AS s1, substring(e16, p.offset) AS s2,
+       substring(e8, p.offset, p.length) AS s3, substring(e16, p.offset, p.length) AS s4
+FROM tab LEFT JOIN permutations AS p ON true;
+
+SELECT '-- Zero offset/length';
+WITH cte AS (SELECT number AS n FROM system.numbers LIMIT 2),
+     permutations AS (SELECT c1.n AS offset, c2.n AS length FROM cte AS c1 CROSS JOIN cte AS c2 LIMIT 3)
+SELECT 'Offset: ', p.offset, 'Length: ', p.length,
+       substring(e8, p.offset) AS s1, substring(e16, p.offset) AS s2,
+       substring(e8, p.offset, p.length) AS s3, substring(e16, p.offset, p.length) AS s4
+FROM tab LEFT JOIN permutations AS p ON true;
+
+SELECT '-- Constant enums';
+SELECT substring(CAST('foo', 'Enum8(\'foo\' = 1)'), 1, 1), substring(CAST('foo', 'Enum16(\'foo\' = 1111)'), 1, 2);
+
+DROP TABLE tab;
