diff --git a/tests/queries/0_stateless/03021_get_client_http_header.reference b/tests/queries/0_stateless/03021_get_client_http_header.reference
new file mode 100644
index 000000000000..099bf1fa4e37
--- /dev/null
+++ b/tests/queries/0_stateless/03021_get_client_http_header.reference
@@ -0,0 +1,26 @@
+-- It works.
+application/x-www-form-urlencoded
+-- It supports non constant arguments.
+application/x-www-form-urlencoded
+*/*
+-- Empty string for non-existent headers.
+application/x-www-form-urlencoded
+
+-- I can use my own headers.
+wtf
+-- Some headers cannot be obtained.
+
+
+-- The setting matters.
+FUNCTION_NOT_ALLOWED
+-- The setting is not enabled by default.
+FUNCTION_NOT_ALLOWED
+-- Are headers case-sentitive?
+application/x-www-form-urlencoded
+
+-- Using it from non-HTTP does not make sense.
+
+
+-- Does it work for distributed queries? (not yet, but maybe it will be needed later)
+
+
diff --git a/tests/queries/0_stateless/03021_get_client_http_header.sh b/tests/queries/0_stateless/03021_get_client_http_header.sh
new file mode 100755
index 000000000000..71486c725b4e
--- /dev/null
+++ b/tests/queries/0_stateless/03021_get_client_http_header.sh
@@ -0,0 +1,37 @@
+#!/usr/bin/env bash
+
+CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+# shellcheck source=../shell_config.sh
+. "$CURDIR"/../shell_config.sh
+
+echo "-- It works."
+${CLICKHOUSE_CURL} "${CLICKHOUSE_URL}&allow_get_client_http_header=1" -d "SELECT getClientHTTPHeader('Content-Type')"
+
+echo "-- It supports non constant arguments."
+${CLICKHOUSE_CURL} "${CLICKHOUSE_URL}&allow_get_client_http_header=1" -d "SELECT getClientHTTPHeader(arrayJoin(['Content-Type', 'Accept']))"
+
+echo "-- Empty string for non-existent headers."
+${CLICKHOUSE_CURL} "${CLICKHOUSE_URL}&allow_get_client_http_header=1" -d "SELECT getClientHTTPHeader(arrayJoin(['Content-Type', 'Upyachka']))"
+
+echo "-- I can use my own headers."
+${CLICKHOUSE_CURL} -H 'Upyachka: wtf' "${CLICKHOUSE_URL}&allow_get_client_http_header=1" -d "SELECT getClientHTTPHeader('Upyachka')"
+
+echo "-- Some headers cannot be obtained."
+${CLICKHOUSE_CURL} -H 'X-ClickHouse-WTF: Secret' "${CLICKHOUSE_URL}&allow_get_client_http_header=1" -d "SELECT getClientHTTPHeader('X-ClickHouse-WTF')"
+${CLICKHOUSE_CURL} -H 'Authentication: Secret' "${CLICKHOUSE_URL}&allow_get_client_http_header=1" -d "SELECT getClientHTTPHeader('Authentication')"
+
+echo "-- The setting matters."
+${CLICKHOUSE_CURL} "${CLICKHOUSE_URL}&allow_get_client_http_header=0" -d "SELECT getClientHTTPHeader(arrayJoin(['Content-Type', 'Accept']))" | grep -o -F 'FUNCTION_NOT_ALLOWED'
+
+echo "-- The setting is not enabled by default."
+${CLICKHOUSE_CURL} "${CLICKHOUSE_URL}" -d "SELECT getClientHTTPHeader(arrayJoin(['Content-Type', 'Accept']))" | grep -o -F 'FUNCTION_NOT_ALLOWED'
+
+echo "-- Are headers case-sentitive?"
+${CLICKHOUSE_CURL} "${CLICKHOUSE_URL}&allow_get_client_http_header=1" -d "SELECT getClientHTTPHeader(arrayJoin(['Content-Type', 'content-type']))"
+
+echo "-- Using it from non-HTTP does not make sense."
+${CLICKHOUSE_CLIENT} --allow_get_client_http_header true --query "SELECT getClientHTTPHeader('Host')"
+${CLICKHOUSE_LOCAL} --allow_get_client_http_header true --query "SELECT getClientHTTPHeader('Host')"
+
+echo "-- Does it work for distributed queries? (not yet, but maybe it will be needed later)"
+${CLICKHOUSE_CURL} "${CLICKHOUSE_URL}&allow_get_client_http_header=1&prefer_localhost_replica=0" -d "SELECT getClientHTTPHeader(name) FROM clusterAllReplicas('test_cluster_one_shard_two_replicas', view(SELECT 'Content-Type' AS name))"
