diff --git a/tests/queries/0_stateless/00498_array_functions_concat_slice_push_pop.reference b/tests/queries/0_stateless/00498_array_functions_concat_slice_push_pop.reference
index f757a86aeee0..f0d66bd49eb8 100644
--- a/tests/queries/0_stateless/00498_array_functions_concat_slice_push_pop.reference
+++ b/tests/queries/0_stateless/00498_array_functions_concat_slice_push_pop.reference
@@ -217,12 +217,12 @@ arraySlice(arr1, -4, 2)
 arraySlice(arr1, -4, -1)
 []	2	4	[]
 []	2	4	[]
-[1,2,3,4,5]	2	4	[]
-[1,2,3,4,5,6,7]	2	4	[4,5]
-[1,2,3,4,5,6,7]	2	4	[4,5]
-[1,2,3,4,5,6,7]	2	4	[4,5]
-[1,2,3,4,5,6,7]	2	4	[4,5]
-[1,2,3,4,5,6,7]	2	4	[4,5]
+[1,2,3,4,5]	2	4	[2,3,4]
+[1,2,3,4,5,6,7]	2	4	[4,5,6]
+[1,2,3,4,5,6,7]	2	4	[4,5,6]
+[1,2,3,4,5,6,7]	2	4	[4,5,6]
+[1,2,3,4,5,6,7]	2	4	[4,5,6]
+[1,2,3,4,5,6,7]	2	4	[4,5,6]
 arrayPushFront(arr1, 1)
 []	[1]
 []	[1]
@@ -504,12 +504,12 @@ arraySlice(arr1, -4, 2)
 arraySlice(arr1, -4, -1)
 []	2	4	[]
 []	2	4	[]
-[1,2,3,4,5]	2	4	[]
-[1,NULL,3,4,NULL,6,7]	2	4	[4,NULL]
-[1,2,3,NULL,5,6,7]	2	4	[NULL,5]
-[1,2,3,4,5,NULL,7]	2	4	[4,5]
-[1,2,3,4,5,6,7]	2	4	[4,5]
-[1,NULL,3,NULL,5,6,7]	2	4	[NULL,5]
+[1,2,3,4,5]	2	4	[2,3,4]
+[1,NULL,3,4,NULL,6,7]	2	4	[4,NULL,6]
+[1,2,3,NULL,5,6,7]	2	4	[NULL,5,6]
+[1,2,3,4,5,NULL,7]	2	4	[4,5,NULL]
+[1,2,3,4,5,6,7]	2	4	[4,5,6]
+[1,NULL,3,NULL,5,6,7]	2	4	[NULL,5,6]
 arrayPushFront(arr1, 1)
 []	[1]
 []	[1]
@@ -817,12 +817,12 @@ arraySlice(arr1, -4, 2)
 arraySlice(arr1, -4, -1)
 []	2	4	[]
 []	2	4	[]
-[1,2,3,4,5]	2	4	[]
-[1,NULL,3,4,NULL,6,7]	2	4	[4,NULL]
-[1,2,3,NULL,5,6,7]	2	4	[NULL,5]
-[1,2,3,4,5,NULL,7]	2	4	[4,5]
-[1,2,3,4,5,6,7]	2	4	[4,5]
-[1,NULL,3,NULL,5,6,7]	2	4	[NULL,5]
+[1,2,3,4,5]	2	4	[2,3,4]
+[1,NULL,3,4,NULL,6,7]	2	4	[4,NULL,6]
+[1,2,3,NULL,5,6,7]	2	4	[NULL,5,6]
+[1,2,3,4,5,NULL,7]	2	4	[4,5,NULL]
+[1,2,3,4,5,6,7]	2	4	[4,5,6]
+[1,NULL,3,NULL,5,6,7]	2	4	[NULL,5,6]
 arrayPushFront(arr1, 1)
 []	[1]
 []	[1]
@@ -1104,12 +1104,12 @@ arraySlice(arr1, -4, 2)
 arraySlice(arr1, -4, -1)
 []	2	4	[]
 []	2	4	[]
-['1','2','3','4','5']	2	4	[]
-['1',NULL,'3','4',NULL,'6','7']	2	4	['4',NULL]
-['1','2','3',NULL,'5','6','7']	2	4	[NULL,'5']
-['1','2','3','4','5',NULL,'7']	2	4	['4','5']
-['1','2','3','4','5','6','7']	2	4	['4','5']
-['1',NULL,'3',NULL,'5','6','7']	2	4	[NULL,'5']
+['1','2','3','4','5']	2	4	['2','3','4']
+['1',NULL,'3','4',NULL,'6','7']	2	4	['4',NULL,'6']
+['1','2','3',NULL,'5','6','7']	2	4	[NULL,'5','6']
+['1','2','3','4','5',NULL,'7']	2	4	['4','5',NULL]
+['1','2','3','4','5','6','7']	2	4	['4','5','6']
+['1',NULL,'3',NULL,'5','6','7']	2	4	[NULL,'5','6']
 arrayPushFront(arr1, 1)
 []	['1']
 []	['1']
diff --git a/tests/queries/0_stateless/00970_substring_arg_validation.reference b/tests/queries/0_stateless/00970_substring_arg_validation.reference
index 13e7564ea0c8..8b137891791f 100644
--- a/tests/queries/0_stateless/00970_substring_arg_validation.reference
+++ b/tests/queries/0_stateless/00970_substring_arg_validation.reference
@@ -1,1 +1,1 @@
-o
+
diff --git a/tests/queries/0_stateless/02185_arraySlice_negative_offset_size.reference b/tests/queries/0_stateless/02185_arraySlice_negative_offset_size.reference
new file mode 100644
index 000000000000..4e3701e4f868
--- /dev/null
+++ b/tests/queries/0_stateless/02185_arraySlice_negative_offset_size.reference
@@ -0,0 +1,15 @@
+[]
+[]
+[]
+[7]
+[7]
+[7]
+-
+[]
+[]
+[]
+[6]
+[]
+[]
+[]
+[]
diff --git a/tests/queries/0_stateless/02185_arraySlice_negative_offset_size.sql b/tests/queries/0_stateless/02185_arraySlice_negative_offset_size.sql
new file mode 100644
index 000000000000..7a6ac0417c72
--- /dev/null
+++ b/tests/queries/0_stateless/02185_arraySlice_negative_offset_size.sql
@@ -0,0 +1,25 @@
+select arraySlice([1, 2, 3, 4, 5, 6, 7, 8], -2, -2);
+select arraySlice(materialize([1, 2, 3, 4, 5, 6, 7, 8]), -2, -2);
+select arraySlice(materialize([1, 2, 3, 4, 5, 6, 7, 8]), materialize(-2), materialize(-2));
+
+select arraySlice([1, 2, 3, 4, 5, 6, 7, 8], -2, -1);
+select arraySlice(materialize([1, 2, 3, 4, 5, 6, 7, 8]), -2, -1);
+select arraySlice(materialize([1, 2, 3, 4, 5, 6, 7, 8]), materialize(-2), materialize(-1));
+
+select '-';
+drop table if exists t;
+create table t
+(
+    s Array(Int),
+    l Int8,
+    r Int8
+) engine = Memory;
+
+insert into t values ([1, 2, 3, 4, 5, 6, 7, 8], -2, -2), ([1, 2, 3, 4, 5, 6, 7, 8], -3, -3);
+
+select arraySlice(s, -2, -2) from t;
+select arraySlice(s, l, -2) from t;
+select arraySlice(s, -2, r) from t;
+select arraySlice(s, l, r) from t;
+
+drop table t;
diff --git a/tests/queries/0_stateless/02185_substring_negative_offset_size.reference b/tests/queries/0_stateless/02185_substring_negative_offset_size.reference
new file mode 100644
index 000000000000..107f083a4d8d
--- /dev/null
+++ b/tests/queries/0_stateless/02185_substring_negative_offset_size.reference
@@ -0,0 +1,31 @@
+
+
+
+g
+g
+g
+-
+
+
+
+g
+g
+g
+-
+
+
+
+6
+
+
+
+
+-
+
+
+
+6
+
+
+
+
diff --git a/tests/queries/0_stateless/02185_substring_negative_offset_size.sql b/tests/queries/0_stateless/02185_substring_negative_offset_size.sql
new file mode 100644
index 000000000000..715dff30369f
--- /dev/null
+++ b/tests/queries/0_stateless/02185_substring_negative_offset_size.sql
@@ -0,0 +1,48 @@
+select substring('abcdefgh', -2, -2);
+select substring(materialize('abcdefgh'), -2, -2);
+select substring(materialize('abcdefgh'), materialize(-2), materialize(-2));
+
+select substring('abcdefgh', -2, -1);
+select substring(materialize('abcdefgh'), -2, -1);
+select substring(materialize('abcdefgh'), materialize(-2), materialize(-1));
+
+select '-';
+select substring(cast('abcdefgh' as FixedString(8)), -2, -2);
+select substring(materialize(cast('abcdefgh' as FixedString(8))), -2, -2);
+select substring(materialize(cast('abcdefgh' as FixedString(8))), materialize(-2), materialize(-2));
+
+select substring(cast('abcdefgh' as FixedString(8)), -2, -1);
+select substring(materialize(cast('abcdefgh' as FixedString(8))), -2, -1);
+select substring(materialize(cast('abcdefgh' as FixedString(8))), materialize(-2), materialize(-1));
+
+select '-';
+drop table if exists t;
+create table t
+(
+    s String,
+    l Int8,
+    r Int8
+) engine = Memory;
+
+insert into t values ('abcdefgh', -2, -2),('12345678', -3, -3);
+
+select substring(s, -2, -2) from t;
+select substring(s, l, -2) from t;
+select substring(s, -2, r) from t;
+select substring(s, l, r) from t;
+
+select '-';
+drop table if exists t;
+create table t(
+                  s FixedString(8),
+                  l Int8,
+                  r Int8
+) engine = Memory;
+insert into t values ('abcdefgh', -2, -2),('12345678', -3, -3);
+
+select substring(s, -2, -2) from t;
+select substring(s, l, -2) from t;
+select substring(s, -2, r) from t;
+select substring(s, l, r) from t;
+
+drop table if exists t;
