diff --git a/dbms/tests/queries/0_stateless/00544_insert_with_select.reference b/dbms/tests/queries/0_stateless/00544_insert_with_select.reference
new file mode 100644
index 000000000000..6cbacc59210b
--- /dev/null
+++ b/dbms/tests/queries/0_stateless/00544_insert_with_select.reference
@@ -0,0 +1,3 @@
+0	0
+1	2
+2	4
diff --git a/dbms/tests/queries/0_stateless/00544_insert_with_select.sql b/dbms/tests/queries/0_stateless/00544_insert_with_select.sql
new file mode 100644
index 000000000000..63ff2b7fd932
--- /dev/null
+++ b/dbms/tests/queries/0_stateless/00544_insert_with_select.sql
@@ -0,0 +1,9 @@
+DROP TABLE IF EXISTS test.test;
+
+CREATE TABLE test.test(number UInt64, num2 UInt64) ENGINE = Log;
+
+INSERT INTO test.test WITH number * 2 AS num2 SELECT number, num2 FROM system.numbers LIMIT 3;
+
+SELECT * FROM test.test;
+
+DROP TABLE test.test;
