diff --git a/tests/queries/0_stateless/02815_first_line.reference b/tests/queries/0_stateless/02815_first_line.reference
new file mode 100644
index 000000000000..cdc86229cc8e
--- /dev/null
+++ b/tests/queries/0_stateless/02815_first_line.reference
@@ -0,0 +1,9 @@
+foo
+foo
+foo
+foobarbaz
+== vector
+1	foo
+2	quux
+3	single line
+4	windows
diff --git a/tests/queries/0_stateless/02815_first_line.sql b/tests/queries/0_stateless/02815_first_line.sql
new file mode 100644
index 000000000000..8c0affaebd3c
--- /dev/null
+++ b/tests/queries/0_stateless/02815_first_line.sql
@@ -0,0 +1,12 @@
+select firstLine('foo
bar
baz');
+select firstLine('foo\rbar\rbaz');
+select firstLine('foo\r
bar\r
baz');
+select firstLine('foobarbaz');
+
+select '== vector';
+
+drop table if exists 02815_first_line_vector;
+create table 02815_first_line_vector (n Int32, text String) engine = MergeTree order by n;
+
+insert into 02815_first_line_vector values (1, 'foo
bar
baz'), (2, 'quux
'), (3, 'single line'), (4, 'windows\r
line breaks');
+select n, firstLine(text) from 02815_first_line_vector order by n;
