diff --git a/tests/queries/0_stateless/01801_s3_cluster_count.reference b/tests/queries/0_stateless/01801_s3_cluster_count.reference
new file mode 100644
index 000000000000..c094c553f81a
--- /dev/null
+++ b/tests/queries/0_stateless/01801_s3_cluster_count.reference
@@ -0,0 +1,2 @@
+12
+12
diff --git a/tests/queries/0_stateless/01801_s3_cluster_count.sql b/tests/queries/0_stateless/01801_s3_cluster_count.sql
new file mode 100644
index 000000000000..8a4fb8049679
--- /dev/null
+++ b/tests/queries/0_stateless/01801_s3_cluster_count.sql
@@ -0,0 +1,5 @@
+-- Tags: no-fasttest
+-- Tag no-fasttest: Depends on AWS
+
+select COUNT() from s3Cluster('test_cluster_two_shards_localhost', 'http://localhost:11111/test/{a,b,c}.tsv');
+select COUNT(*) from s3Cluster('test_cluster_two_shards_localhost', 'http://localhost:11111/test/{a,b,c}.tsv');
