diff --git a/src/IO/ReadBuffer.h b/src/IO/ReadBuffer.h
index 3adb5174ca9d..e4260d1eb1e1 100644
--- a/src/IO/ReadBuffer.h
+++ b/src/IO/ReadBuffer.h
@@ -252,7 +252,8 @@ class ReadBuffer : public BufferBase
     size_t nextimpl_working_buffer_offset = 0;
 
 private:
-    /** Read the next data and fill a buffer with it.
+    /** Read the next data and fill a buffer with it. It should also account for `nextimpl_working_buffer_offset` out parameter if set
+      * so that after this value is applied to `pos` (see next() method) buffer still contains available data.
       * Return `false` in case of the end, `true` otherwise.
       * Throw an exception if something is wrong.
       */
diff --git a/src/IO/ZlibInflatingReadBuffer.cpp b/src/IO/ZlibInflatingReadBuffer.cpp
index 6e416f6e7ffd..9c47521d0d77 100644
--- a/src/IO/ZlibInflatingReadBuffer.cpp
+++ b/src/IO/ZlibInflatingReadBuffer.cpp
@@ -100,6 +100,9 @@ bool ZlibInflatingReadBuffer::nextImpl()
             if (rc != Z_OK)
                 throw Exception(
                     ErrorCodes::ZLIB_INFLATE_FAILED, "inflateReset failed: {}{}", zError(rc), getExceptionEntryWithFileName(*in));
+            if (working_buffer.empty())
+                continue;
+
             return true;
         }
 
