diff --git a/tests/queries/0_stateless/03216_arrayWithConstant_limits.reference b/tests/queries/0_stateless/03216_arrayWithConstant_limits.reference
new file mode 100644
index 000000000000..825319e1c5ba
--- /dev/null
+++ b/tests/queries/0_stateless/03216_arrayWithConstant_limits.reference
@@ -0,0 +1,1 @@
+10000000
diff --git a/tests/queries/0_stateless/03216_arrayWithConstant_limits.sql b/tests/queries/0_stateless/03216_arrayWithConstant_limits.sql
new file mode 100644
index 000000000000..c46524c50e6d
--- /dev/null
+++ b/tests/queries/0_stateless/03216_arrayWithConstant_limits.sql
@@ -0,0 +1,3 @@
+SELECT arrayWithConstant(96142475, ['qMUF']); -- { serverError TOO_LARGE_ARRAY_SIZE }
+SELECT arrayWithConstant(100000000, materialize([[[[[[[[[['Hello, world!']]]]]]]]]])); -- { serverError TOO_LARGE_ARRAY_SIZE }
+SELECT length(arrayWithConstant(10000000, materialize([[[[[[[[[['Hello world']]]]]]]]]])));
