diff --git a/tests/queries/0_stateless/02372_now_in_block.reference b/tests/queries/0_stateless/02372_nowInBlock.reference
similarity index 100%
rename from tests/queries/0_stateless/02372_now_in_block.reference
rename to tests/queries/0_stateless/02372_nowInBlock.reference
diff --git a/tests/queries/0_stateless/02372_now_in_block.sql b/tests/queries/0_stateless/02372_nowInBlock.sql
similarity index 99%
rename from tests/queries/0_stateless/02372_now_in_block.sql
rename to tests/queries/0_stateless/02372_nowInBlock.sql
index d0aec471801b..ee067f5e4df9 100644
--- a/tests/queries/0_stateless/02372_now_in_block.sql
+++ b/tests/queries/0_stateless/02372_nowInBlock.sql
@@ -1,4 +1,5 @@
 SET max_rows_to_read = 0, max_bytes_to_read = 0;
+
 SELECT count() FROM (SELECT DISTINCT nowInBlock(), nowInBlock('Pacific/Pitcairn') FROM system.numbers LIMIT 2);
 SELECT nowInBlock(1); -- { serverError ILLEGAL_TYPE_OF_ARGUMENT }
 SELECT nowInBlock(NULL) IS NULL;
diff --git a/tests/queries/0_stateless/03567_nowInBlock64.reference b/tests/queries/0_stateless/03567_nowInBlock64.reference
new file mode 100644
index 000000000000..f00580c40607
--- /dev/null
+++ b/tests/queries/0_stateless/03567_nowInBlock64.reference
@@ -0,0 +1,2 @@
+3
+1
diff --git a/tests/queries/0_stateless/03567_nowInBlock64.sql b/tests/queries/0_stateless/03567_nowInBlock64.sql
new file mode 100644
index 000000000000..577b4fd2bc4d
--- /dev/null
+++ b/tests/queries/0_stateless/03567_nowInBlock64.sql
@@ -0,0 +1,11 @@
+SET max_rows_to_read = 0, max_bytes_to_read = 0;
+
+SELECT nowInBlock64(3, 'America/Sao_Paulo', 3); --{serverError TOO_MANY_ARGUMENTS_FOR_FUNCTION}
+SELECT nowInBlock64(10); --{serverError ARGUMENT_OUT_OF_BOUND}
+SELECT nowInBlock64('string'); --{serverError ILLEGAL_TYPE_OF_ARGUMENT}
+SELECT nowInBlock64(3, true); --{serverError ILLEGAL_TYPE_OF_ARGUMENT}
+SELECT nowInBlock64(3, 3); --{serverError ILLEGAL_TYPE_OF_ARGUMENT}
+SELECT nowInBlock64(3, 'string'); --{serverError BAD_ARGUMENTS}
+
+SELECT count() FROM (SELECT DISTINCT nowInBlock64(), nowInBlock64(3), nowInBlock64(3, 'Pacific/Pitcairn') FROM system.numbers LIMIT 3);
+SELECT nowInBlock64(NULL) IS NULL;
