diff --git a/tests/clickhouse-test-server b/tests/clickhouse-test-server
index a6b6bd86dfae..4087468b597c 100755
--- a/tests/clickhouse-test-server
+++ b/tests/clickhouse-test-server
@@ -60,6 +60,7 @@ if [ "$DATA_DIR_PATTERN" != "$DATA_DIR" ]; then
     cat ${CONFIG_SERVER_DIR}/ints_dictionary.xml | sed -e s!9000!$CLICKHOUSE_PORT_TCP! > $DATA_DIR/etc/ints_dictionary.xml
     cat ${CONFIG_SERVER_DIR}/strings_dictionary.xml | sed -e s!9000!$CLICKHOUSE_PORT_TCP! > $DATA_DIR/etc/strings_dictionary.xml
     cat ${CONFIG_SERVER_DIR}/decimals_dictionary.xml | sed -e s!9000!$CLICKHOUSE_PORT_TCP! > $DATA_DIR/etc/decimals_dictionary.xml
+    cat ${CONFIG_SERVER_DIR}/executable_pool_dictionary.xml | sed -e s!9000!$CLICKHOUSE_PORT_TCP! > $DATA_DIR/etc/executable_pool_dictionary.xml
 fi
 
 CLICKHOUSE_EXTRACT_CONFIG=${CLICKHOUSE_EXTRACT_CONFIG:="${CLICKHOUSE_EXTRACT} --config=$CLICKHOUSE_CONFIG"}
diff --git a/tests/config/executable_pool_dictionary.xml b/tests/config/executable_pool_dictionary.xml
new file mode 100644
index 000000000000..13f34f0048ee
--- /dev/null
+++ b/tests/config/executable_pool_dictionary.xml
@@ -0,0 +1,154 @@
+<dictionaries>
+
+<dictionary>
+    <name>executable_pool_simple</name>
+
+    <structure>
+        <id>
+            <name>x</name>
+        </id>
+        <attribute>
+            <name>a</name>
+            <type>String</type>
+            <null_value></null_value>
+        </attribute>
+        <attribute>
+            <name>b</name>
+            <type>String</type>
+            <null_value></null_value>
+        </attribute>
+    </structure>
+
+    <source>
+        <executable_pool>
+            <format>TabSeparated</format>
+            <command>while read read_data; do printf "$read_data\t$read_data a\t$read_data b
"; done</command>
+            <size>5</size>
+        </executable_pool>
+    </source>
+
+    <layout>
+        <direct />
+    </layout>
+
+    <lifetime>300</lifetime>
+</dictionary>
+
+<dictionary>
+    <name>executable_pool_complex</name>
+
+    <structure>
+        <key>
+            <attribute>
+                <name>x</name>
+                <type>String</type>
+            </attribute>
+            <attribute>
+                <name>y</name>
+                <type>String</type>
+            </attribute>
+        </key>
+        <attribute>
+            <name>a</name>
+            <type>String</type>
+            <null_value></null_value>
+        </attribute>
+        <attribute>
+            <name>b</name>
+            <type>String</type>
+            <null_value></null_value>
+        </attribute>
+    </structure>
+
+    <source>
+        <executable>
+            <format>TabSeparated</format>
+            <command>while read read_data; do printf "$read_data\tvalue a\tvalue b
"; done</command>
+        </executable>
+    </source>
+
+    <layout>
+        <complex_key_direct />
+    </layout>
+
+    <lifetime>300</lifetime>
+</dictionary>
+
+<dictionary>
+    <name>executable_pool_simple_implicit_key</name>
+
+    <structure>
+        <id>
+            <name>x</name>
+        </id>
+        <attribute>
+            <name>a</name>
+            <type>String</type>
+            <null_value></null_value>
+        </attribute>
+        <attribute>
+            <name>b</name>
+            <type>String</type>
+            <null_value></null_value>
+        </attribute>
+    </structure>
+
+    <source>
+        <executable_pool>
+            <format>TabSeparated</format>
+            <command>while read read_data; do printf "$read_data a\t$read_data b
"; done</command>
+            <size>5</size>
+            <implicit_key>true</implicit_key>
+        </executable_pool>
+    </source>
+
+    <layout>
+        <direct />
+    </layout>
+
+    <lifetime>300</lifetime>
+</dictionary>
+
+<dictionary>
+    <name>executable_pool_complex_implicit_key</name>
+
+    <structure>
+        <key>
+            <attribute>
+                <name>x</name>
+                <type>String</type>
+            </attribute>
+            <attribute>
+                <name>y</name>
+                <type>String</type>
+            </attribute>
+        </key>
+        <attribute>
+            <name>a</name>
+            <type>String</type>
+            <null_value></null_value>
+        </attribute>
+        <attribute>
+            <name>b</name>
+            <type>String</type>
+            <null_value></null_value>
+        </attribute>
+    </structure>
+
+    <source>
+        <executable_pool>
+            <format>TabSeparated</format>
+            <command>while read read_data; do printf "data a\tdata b
"; done</command>
+            <size>5</size>
+            <implicit_key>true</implicit_key>
+        </executable_pool>
+    </source>
+
+    <layout>
+        <complex_key_direct />
+    </layout>
+
+    <lifetime>300</lifetime>
+</dictionary>
+
+</dictionaries>
diff --git a/tests/config/install.sh b/tests/config/install.sh
index de6ba2a7a091..17859d37c5ff 100755
--- a/tests/config/install.sh
+++ b/tests/config/install.sh
@@ -44,6 +44,7 @@ ln -sf $SRC_PATH/ints_dictionary.xml $DEST_SERVER_PATH/
 ln -sf $SRC_PATH/strings_dictionary.xml $DEST_SERVER_PATH/
 ln -sf $SRC_PATH/decimals_dictionary.xml $DEST_SERVER_PATH/
 ln -sf $SRC_PATH/executable_dictionary.xml $DEST_SERVER_PATH/
+ln -sf $SRC_PATH/executable_pool_dictionary.xml $DEST_SERVER_PATH/
 
 ln -sf $SRC_PATH/top_level_domains $DEST_SERVER_PATH/
 
diff --git a/tests/queries/0_stateless/01746_executable_pool_dictionary.reference b/tests/queries/0_stateless/01746_executable_pool_dictionary.reference
new file mode 100644
index 000000000000..cf79a75e3f15
--- /dev/null
+++ b/tests/queries/0_stateless/01746_executable_pool_dictionary.reference
@@ -0,0 +1,10 @@
+executable_pool_simple
+1 a
+1 b
+2 a
+2 b
+executable_pool_complex
+value a
+value b
+value a
+value b
diff --git a/tests/queries/0_stateless/01746_executable_pool_dictionary.sql b/tests/queries/0_stateless/01746_executable_pool_dictionary.sql
new file mode 100644
index 000000000000..aa59d209b510
--- /dev/null
+++ b/tests/queries/0_stateless/01746_executable_pool_dictionary.sql
@@ -0,0 +1,15 @@
+SELECT 'executable_pool_simple';
+
+SELECT dictGet('executable_pool_simple', 'a', toUInt64(1));
+SELECT dictGet('executable_pool_simple', 'b', toUInt64(1));
+
+SELECT dictGet('executable_pool_simple', 'a', toUInt64(2));
+SELECT dictGet('executable_pool_simple', 'b', toUInt64(2));
+
+SELECT 'executable_pool_complex';
+
+SELECT dictGet('executable_pool_complex', 'a', ('First_1', 'Second_1'));
+SELECT dictGet('executable_pool_complex', 'b', ('First_1', 'Second_1'));
+
+SELECT dictGet('executable_pool_complex', 'a', ('First_2', 'Second_2'));
+SELECT dictGet('executable_pool_complex', 'b', ('First_2', 'Second_2'));
diff --git a/tests/queries/0_stateless/01747_executable_pool_dictionary_implicit_key.reference b/tests/queries/0_stateless/01747_executable_pool_dictionary_implicit_key.reference
new file mode 100644
index 000000000000..ab44ba51330c
--- /dev/null
+++ b/tests/queries/0_stateless/01747_executable_pool_dictionary_implicit_key.reference
@@ -0,0 +1,10 @@
+executable_pool_simple_implicit_key
+1 a
+1 b
+2 a
+2 b
+executable_pool_complex_implicit_key
+data a
+data b
+data a
+data b
diff --git a/tests/queries/0_stateless/01747_executable_pool_dictionary_implicit_key.sql b/tests/queries/0_stateless/01747_executable_pool_dictionary_implicit_key.sql
new file mode 100644
index 000000000000..6c6d33c7d43a
--- /dev/null
+++ b/tests/queries/0_stateless/01747_executable_pool_dictionary_implicit_key.sql
@@ -0,0 +1,15 @@
+SELECT 'executable_pool_simple_implicit_key';
+
+SELECT dictGet('executable_pool_simple_implicit_key', 'a', toUInt64(1));
+SELECT dictGet('executable_pool_simple_implicit_key', 'b', toUInt64(1));
+
+SELECT dictGet('executable_pool_simple_implicit_key', 'a', toUInt64(2));
+SELECT dictGet('executable_pool_simple_implicit_key', 'b', toUInt64(2));
+
+SELECT 'executable_pool_complex_implicit_key';
+
+SELECT dictGet('executable_pool_complex_implicit_key', 'a', ('First_1', 'Second_1'));
+SELECT dictGet('executable_pool_complex_implicit_key', 'b', ('First_1', 'Second_1'));
+
+SELECT dictGet('executable_pool_complex_implicit_key', 'a', ('First_2', 'Second_2'));
+SELECT dictGet('executable_pool_complex_implicit_key', 'b', ('First_2', 'Second_2'));
