diff --git a/src/function/scalar/date/date_diff.cpp b/src/function/scalar/date/date_diff.cpp
index d3808e1c25ff..491eda5781c3 100644
--- a/src/function/scalar/date/date_diff.cpp
+++ b/src/function/scalar/date/date_diff.cpp
@@ -50,7 +50,7 @@ struct DateDiff {
 	struct DayOperator {
 		template <class TA, class TB, class TR>
 		static inline TR Operation(TA startdate, TB enddate) {
-			return Date::EpochDays(enddate) - Date::EpochDays(startdate);
+			return TR(Date::EpochDays(enddate)) - TR(Date::EpochDays(startdate));
 		}
 	};
 
