diff --git a/tools/nodejs/package.json b/tools/nodejs/package.json
index 90c757aa6d34..54dab4eba8dc 100644
--- a/tools/nodejs/package.json
+++ b/tools/nodejs/package.json
@@ -25,8 +25,8 @@
     "test": "test"
   },
   "devDependencies": {
-    "mocha": "^8.3.0",
-    "aws-sdk": "^2.790.0"
+    "aws-sdk": "^2.790.0",
+    "mocha": "^8.3.0"
   },
   "repository": {
     "type": "git",
diff --git a/tools/nodejs/src/statement.cpp b/tools/nodejs/src/statement.cpp
index 40820b104e7d..a0fec39846b0 100644
--- a/tools/nodejs/src/statement.cpp
+++ b/tools/nodejs/src/statement.cpp
@@ -171,6 +171,9 @@ static Napi::Value convert_chunk(Napi::Env &env, std::vector<std::string> names,
 			case duckdb::LogicalTypeId::VARCHAR: {
 				value = Napi::String::New(env, dval.str_value);
 			} break;
+			case duckdb::LogicalTypeId::BOOLEAN: {
+				value = Napi::Boolean::New(env, dval.value_.boolean);
+			} break;
 			case duckdb::LogicalTypeId::BLOB: {
 				value = Napi::Buffer<char>::Copy(env, dval.str_value.c_str(), dval.str_value.length());
 			} break;
