diff --git a/docker/test/fasttest/run.sh b/docker/test/fasttest/run.sh
index 7555b5591d00..17cec7ae2868 100755
--- a/docker/test/fasttest/run.sh
+++ b/docker/test/fasttest/run.sh
@@ -251,8 +251,12 @@ function run_tests
         00701_rollup
         00834_cancel_http_readonly_queries_on_client_close
         00911_tautological_compare
+
+        # Hyperscan
         00926_multimatch
         00929_multi_match_edit_distance
+        01681_hyperscan_debug_assertion
+
         01031_mutations_interpreter_and_context
         01053_ssd_dictionary # this test mistakenly requires acces to /var/lib/clickhouse -- can't run this locally, disabled
         01083_expressions_in_engine_arguments
diff --git a/tests/queries/0_stateless/01681_hyperscan_debug_assertion.reference b/tests/queries/0_stateless/01681_hyperscan_debug_assertion.reference
new file mode 100644
index 000000000000..7326d9603970
--- /dev/null
+++ b/tests/queries/0_stateless/01681_hyperscan_debug_assertion.reference
@@ -0,0 +1,1 @@
+Ok
diff --git a/tests/queries/0_stateless/01681_hyperscan_debug_assertion.sh b/tests/queries/0_stateless/01681_hyperscan_debug_assertion.sh
new file mode 100755
index 000000000000..0bbf8942c1af
--- /dev/null
+++ b/tests/queries/0_stateless/01681_hyperscan_debug_assertion.sh
@@ -0,0 +1,23 @@
+#!/usr/bin/env bash
+
+CLICKHOUSE_CLIENT_SERVER_LOGS_LEVEL=fatal
+
+CURDIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+# shellcheck source=../shell_config.sh
+. "$CURDIR"/../shell_config.sh
+
+# We throw our own exception from operator new.
+# In previous versions of Hyperscan it triggered debug assertion as it only expected std::bad_alloc.
+
+M=1000000
+
+while true
+do
+    $CLICKHOUSE_CLIENT --allow_hyperscan 1 --max_memory_usage $M --format Null --query "
+        SELECT [1, 2, 3, 11] = arraySort(multiMatchAllIndices('фабрикант', ['', 'рикан', 'а', 'f[a${RANDOM}e]b[ei]rl', 'ф[иа${RANDOM}эе]б[еэи][рпл]', 'афиукд', 'a[f${RANDOM}t],th', '^ф[аие${RANDOM}э]?б?[еэи]?$', 'бе${RANDOM}рлик', 'fa${RANDOM}b', 'фа[беьв]+е?[рл${RANDOM}ко]']))
+    " 2>&1 | grep -q 'Memory limit' || break;
+
+    M=$((M + 100000))
+done
+
+echo 'Ok'
