diff --git a/tests/config/config.d/enable_access_control_improvements.xml b/tests/config/config.d/enable_access_control_improvements.xml
index 5a1865480984..564b656a0ad0 100644
--- a/tests/config/config.d/enable_access_control_improvements.xml
+++ b/tests/config/config.d/enable_access_control_improvements.xml
@@ -4,5 +4,6 @@
         <on_cluster_queries_require_cluster_grant>true</on_cluster_queries_require_cluster_grant>
         <select_from_system_db_requires_grant>true</select_from_system_db_requires_grant>
         <select_from_information_schema_requires_grant>true</select_from_information_schema_requires_grant>
+        <settings_constraints_replace_previous>true</settings_constraints_replace_previous>
     </access_control_improvements>
 </clickhouse>
diff --git a/tests/integration/helpers/0_common_instance_config.xml b/tests/integration/helpers/0_common_instance_config.xml
index 64f0ce9e361c..27563e47c355 100644
--- a/tests/integration/helpers/0_common_instance_config.xml
+++ b/tests/integration/helpers/0_common_instance_config.xml
@@ -24,5 +24,6 @@
         <on_cluster_queries_require_cluster_grant>true</on_cluster_queries_require_cluster_grant>
         <select_from_system_db_requires_grant>true</select_from_system_db_requires_grant>
         <select_from_information_schema_requires_grant>true</select_from_information_schema_requires_grant>
+        <settings_constraints_replace_previous>true</settings_constraints_replace_previous>
     </access_control_improvements>
 </clickhouse>
diff --git a/tests/integration/test_disabled_access_control_improvements/configs/config.d/disable_access_control_improvements.xml b/tests/integration/test_disabled_access_control_improvements/configs/config.d/disable_access_control_improvements.xml
index 7969c638fd70..a335c7f8a1f1 100644
--- a/tests/integration/test_disabled_access_control_improvements/configs/config.d/disable_access_control_improvements.xml
+++ b/tests/integration/test_disabled_access_control_improvements/configs/config.d/disable_access_control_improvements.xml
@@ -3,5 +3,6 @@
         <users_without_row_policies_can_read_rows remove="remove"/>
         <select_from_system_db_requires_grant remove="remove"/>
         <select_from_information_schema_requires_grant remove="remove"/>
+        <settings_constraints_replace_previous remove="remove"/>
     </access_control_improvements>
 </clickhouse>
diff --git a/tests/integration/test_disk_access_storage/test.py b/tests/integration/test_disk_access_storage/test.py
index 273a00adffef..bcfc9d718a7b 100644
--- a/tests/integration/test_disk_access_storage/test.py
+++ b/tests/integration/test_disk_access_storage/test.py
@@ -93,7 +93,7 @@ def test_alter():
     instance.query("GRANT SELECT ON mydb.mytable TO u1")
     instance.query("GRANT SELECT ON mydb.* TO rx WITH GRANT OPTION")
     instance.query(
-        "ALTER SETTINGS PROFILE s1 SETTINGS max_memory_usage = 987654321 READONLY"
+        "ALTER SETTINGS PROFILE s1 SETTINGS max_memory_usage = 987654321 CONST"
     )
 
     def check():
@@ -124,7 +124,7 @@ def check():
         )
         assert (
             instance.query("SHOW CREATE SETTINGS PROFILE s1")
-            == "CREATE SETTINGS PROFILE s1 SETTINGS max_memory_usage = 987654321 READONLY
"
+            == "CREATE SETTINGS PROFILE s1 SETTINGS max_memory_usage = 987654321 CONST
"
         )
         assert (
             instance.query("SHOW CREATE SETTINGS PROFILE s2")
diff --git a/tests/integration/test_settings_profile/test.py b/tests/integration/test_settings_profile/test.py
index 13af1cf650dd..3358315cca72 100644
--- a/tests/integration/test_settings_profile/test.py
+++ b/tests/integration/test_settings_profile/test.py
@@ -216,7 +216,18 @@ def test_settings_from_granted_role():
             "\\N",
             "\\N",
         ],
-        ["xyz", "\\N", "\\N", 1, "max_ast_depth", 2000, "\\N", "\\N", "\\N", "\\N"],
+        [
+            "xyz",
+            "\\N",
+            "\\N",
+            1,
+            "max_ast_depth",
+            2000,
+            "\\N",
+            "\\N",
+            "\\N",
+            "\\N",
+        ],
     ]
     assert system_settings_profile_elements(role_name="worker") == [
         ["\\N", "\\N", "worker", 0, "\\N", "\\N", "\\N", "\\N", "\\N", "xyz"]
@@ -288,12 +299,12 @@ def test_settings_from_granted_role():
 
 def test_inheritance():
     instance.query(
-        "CREATE SETTINGS PROFILE xyz SETTINGS max_memory_usage = 100000002 READONLY"
+        "CREATE SETTINGS PROFILE xyz SETTINGS max_memory_usage = 100000002 CONST"
     )
     instance.query("CREATE SETTINGS PROFILE alpha SETTINGS PROFILE xyz TO robin")
     assert (
         instance.query("SHOW CREATE SETTINGS PROFILE xyz")
-        == "CREATE SETTINGS PROFILE xyz SETTINGS max_memory_usage = 100000002 READONLY
"
+        == "CREATE SETTINGS PROFILE xyz SETTINGS max_memory_usage = 100000002 CONST
"
     )
     assert (
         instance.query("SHOW CREATE SETTINGS PROFILE alpha")
@@ -315,7 +326,18 @@ def test_inheritance():
         ["xyz", "local directory", 1, 0, "[]", "[]"]
     ]
     assert system_settings_profile_elements(profile_name="xyz") == [
-        ["xyz", "\\N", "\\N", 0, "max_memory_usage", 100000002, "\\N", "\\N", 1, "\\N"]
+        [
+            "xyz",
+            "\\N",
+            "\\N",
+            0,
+            "max_memory_usage",
+            100000002,
+            "\\N",
+            "\\N",
+            "CONST",
+            "\\N",
+        ]
     ]
     assert system_settings_profile("alpha") == [
         ["alpha", "local directory", 1, 0, "['robin']", "[]"]
@@ -366,6 +388,66 @@ def test_alter_and_drop():
     instance.query("SET max_memory_usage = 120000000", user="robin")
 
 
+def test_changeable_in_readonly():
+    instance.query(
+        "CREATE SETTINGS PROFILE xyz SETTINGS max_memory_usage = 100000003 MIN 90000000 MAX 110000000 CHANGEABLE_IN_READONLY SETTINGS readonly = 1 TO robin"
+    )
+    assert (
+        instance.query(
+            "SELECT value FROM system.settings WHERE name = 'max_memory_usage'",
+            user="robin",
+        )
+        == "100000003
"
+    )
+    assert (
+        instance.query(
+            "SELECT value FROM system.settings WHERE name = 'readonly'",
+            user="robin",
+        )
+        == "1
"
+    )
+    assert (
+        "Setting max_memory_usage shouldn't be less than 90000000"
+        in instance.query_and_get_error("SET max_memory_usage = 80000000", user="robin")
+    )
+    assert (
+        "Setting max_memory_usage shouldn't be greater than 110000000"
+        in instance.query_and_get_error(
+            "SET max_memory_usage = 120000000", user="robin"
+        )
+    )
+
+    assert system_settings_profile_elements(profile_name="xyz") == [
+        [
+            "xyz",
+            "\\N",
+            "\\N",
+            0,
+            "max_memory_usage",
+            100000003,
+            90000000,
+            110000000,
+            "CHANGEABLE_IN_READONLY",
+            "\\N",
+        ],
+        [
+            "xyz",
+            "\\N",
+            "\\N",
+            1,
+            "readonly",
+            1,
+            "\\N",
+            "\\N",
+            "\\N",
+            "\\N",
+        ],
+    ]
+
+    instance.query("SET max_memory_usage = 90000000", user="robin")
+    instance.query("SET max_memory_usage = 110000000", user="robin")
+
+
 def test_show_profiles():
     instance.query("CREATE SETTINGS PROFILE xyz")
     assert instance.query("SHOW SETTINGS PROFILES") == "default
readonly
xyz
"
diff --git a/tests/queries/0_stateless/01292_create_user.reference b/tests/queries/0_stateless/01292_create_user.reference
index 997a9504bb51..f723412c636f 100644
--- a/tests/queries/0_stateless/01292_create_user.reference
+++ b/tests/queries/0_stateless/01292_create_user.reference
@@ -58,9 +58,9 @@ CREATE USER u2_01292 SETTINGS PROFILE default
 CREATE USER u3_01292 SETTINGS max_memory_usage = 5000000
 CREATE USER u4_01292 SETTINGS max_memory_usage MIN 5000000
 CREATE USER u5_01292 SETTINGS max_memory_usage MAX 5000000
-CREATE USER u6_01292 SETTINGS max_memory_usage READONLY
+CREATE USER u6_01292 SETTINGS max_memory_usage CONST
 CREATE USER u7_01292 SETTINGS max_memory_usage WRITABLE
-CREATE USER u8_01292 SETTINGS max_memory_usage = 5000000 MIN 4000000 MAX 6000000 READONLY
+CREATE USER u8_01292 SETTINGS max_memory_usage = 5000000 MIN 4000000 MAX 6000000 CONST
 CREATE USER u9_01292 SETTINGS PROFILE default, max_memory_usage = 5000000 WRITABLE
 CREATE USER u1_01292 SETTINGS readonly = 1
 CREATE USER u2_01292 SETTINGS readonly = 1
@@ -111,7 +111,7 @@ u4_01292	local directory	double_sha1_password	{}	['::/0']	[]	[]	[]	1	[]	['r1_012
 -- system.settings_profile_elements
 \N	u1_01292	\N	0	readonly	1	\N	\N	\N	\N
 \N	u2_01292	\N	0	\N	\N	\N	\N	\N	default
-\N	u3_01292	\N	0	max_memory_usage	5000000	4000000	6000000	0	\N
+\N	u3_01292	\N	0	max_memory_usage	5000000	4000000	6000000	WRITABLE	\N
 \N	u4_01292	\N	0	\N	\N	\N	\N	\N	default
 \N	u4_01292	\N	1	max_memory_usage	5000000	\N	\N	\N	\N
 \N	u4_01292	\N	2	readonly	1	\N	\N	\N	\N
diff --git a/tests/queries/0_stateless/01292_create_user.sql b/tests/queries/0_stateless/01292_create_user.sql
index a9582376825d..d0f157d36b09 100644
--- a/tests/queries/0_stateless/01292_create_user.sql
+++ b/tests/queries/0_stateless/01292_create_user.sql
@@ -122,9 +122,9 @@ CREATE USER u2_01292 SETTINGS PROFILE 'default';
 CREATE USER u3_01292 SETTINGS max_memory_usage=5000000;
 CREATE USER u4_01292 SETTINGS max_memory_usage MIN=5000000;
 CREATE USER u5_01292 SETTINGS max_memory_usage MAX=5000000;
-CREATE USER u6_01292 SETTINGS max_memory_usage READONLY;
+CREATE USER u6_01292 SETTINGS max_memory_usage CONST;
 CREATE USER u7_01292 SETTINGS max_memory_usage WRITABLE;
-CREATE USER u8_01292 SETTINGS max_memory_usage=5000000 MIN 4000000 MAX 6000000 READONLY;
+CREATE USER u8_01292 SETTINGS max_memory_usage=5000000 MIN 4000000 MAX 6000000 CONST;
 CREATE USER u9_01292 SETTINGS PROFILE 'default', max_memory_usage=5000000 WRITABLE;
 SHOW CREATE USER u1_01292;
 SHOW CREATE USER u2_01292;
diff --git a/tests/queries/0_stateless/01293_create_role.reference b/tests/queries/0_stateless/01293_create_role.reference
index 8d9a259ecf5c..9b3c4eabd478 100644
--- a/tests/queries/0_stateless/01293_create_role.reference
+++ b/tests/queries/0_stateless/01293_create_role.reference
@@ -15,9 +15,9 @@ CREATE ROLE r2_01293 SETTINGS PROFILE default
 CREATE ROLE r3_01293 SETTINGS max_memory_usage = 5000000
 CREATE ROLE r4_01293 SETTINGS max_memory_usage MIN 5000000
 CREATE ROLE r5_01293 SETTINGS max_memory_usage MAX 5000000
-CREATE ROLE r6_01293 SETTINGS max_memory_usage READONLY
+CREATE ROLE r6_01293 SETTINGS max_memory_usage CONST
 CREATE ROLE r7_01293 SETTINGS max_memory_usage WRITABLE
-CREATE ROLE r8_01293 SETTINGS max_memory_usage = 5000000 MIN 4000000 MAX 6000000 READONLY
+CREATE ROLE r8_01293 SETTINGS max_memory_usage = 5000000 MIN 4000000 MAX 6000000 CONST
 CREATE ROLE r9_01293 SETTINGS PROFILE default, max_memory_usage = 5000000 WRITABLE
 CREATE ROLE r1_01293 SETTINGS readonly = 1
 CREATE ROLE r2_01293 SETTINGS readonly = 1
@@ -32,7 +32,7 @@ r1_01293	local directory
 -- system.settings_profile_elements
 \N	\N	r1_01293	0	readonly	1	\N	\N	\N	\N
 \N	\N	r2_01293	0	\N	\N	\N	\N	\N	default
-\N	\N	r3_01293	0	max_memory_usage	5000000	4000000	6000000	0	\N
+\N	\N	r3_01293	0	max_memory_usage	5000000	4000000	6000000	WRITABLE	\N
 \N	\N	r4_01293	0	\N	\N	\N	\N	\N	default
 \N	\N	r4_01293	1	max_memory_usage	5000000	\N	\N	\N	\N
 \N	\N	r4_01293	2	readonly	1	\N	\N	\N	\N
diff --git a/tests/queries/0_stateless/01293_create_role.sql b/tests/queries/0_stateless/01293_create_role.sql
index 963a1020e3fe..f22edfeec3ef 100644
--- a/tests/queries/0_stateless/01293_create_role.sql
+++ b/tests/queries/0_stateless/01293_create_role.sql
@@ -31,9 +31,9 @@ CREATE ROLE r2_01293 SETTINGS PROFILE 'default';
 CREATE ROLE r3_01293 SETTINGS max_memory_usage=5000000;
 CREATE ROLE r4_01293 SETTINGS max_memory_usage MIN=5000000;
 CREATE ROLE r5_01293 SETTINGS max_memory_usage MAX=5000000;
-CREATE ROLE r6_01293 SETTINGS max_memory_usage READONLY;
+CREATE ROLE r6_01293 SETTINGS max_memory_usage CONST;
 CREATE ROLE r7_01293 SETTINGS max_memory_usage WRITABLE;
-CREATE ROLE r8_01293 SETTINGS max_memory_usage=5000000 MIN 4000000 MAX 6000000 READONLY;
+CREATE ROLE r8_01293 SETTINGS max_memory_usage=5000000 MIN 4000000 MAX 6000000 CONST;
 CREATE ROLE r9_01293 SETTINGS PROFILE 'default', max_memory_usage=5000000 WRITABLE;
 SHOW CREATE ROLE r1_01293;
 SHOW CREATE ROLE r2_01293;
diff --git a/tests/queries/0_stateless/01294_create_settings_profile.reference b/tests/queries/0_stateless/01294_create_settings_profile.reference
index da47b084070e..dbb73bca8511 100644
--- a/tests/queries/0_stateless/01294_create_settings_profile.reference
+++ b/tests/queries/0_stateless/01294_create_settings_profile.reference
@@ -11,9 +11,9 @@ CREATE SETTINGS PROFILE s2_01294 SETTINGS INHERIT default
 CREATE SETTINGS PROFILE s3_01294 SETTINGS max_memory_usage = 5000000
 CREATE SETTINGS PROFILE s4_01294 SETTINGS max_memory_usage MIN 5000000
 CREATE SETTINGS PROFILE s5_01294 SETTINGS max_memory_usage MAX 5000000
-CREATE SETTINGS PROFILE s6_01294 SETTINGS max_memory_usage READONLY
+CREATE SETTINGS PROFILE s6_01294 SETTINGS max_memory_usage CONST
 CREATE SETTINGS PROFILE s7_01294 SETTINGS max_memory_usage WRITABLE
-CREATE SETTINGS PROFILE s8_01294 SETTINGS max_memory_usage = 5000000 MIN 4000000 MAX 6000000 READONLY
+CREATE SETTINGS PROFILE s8_01294 SETTINGS max_memory_usage = 5000000 MIN 4000000 MAX 6000000 CONST
 CREATE SETTINGS PROFILE s9_01294 SETTINGS INHERIT default, max_memory_usage = 5000000 WRITABLE
 CREATE SETTINGS PROFILE s10_01294 SETTINGS INHERIT s1_01294, INHERIT s3_01294, INHERIT default, readonly = 0, max_memory_usage MAX 6000000
 CREATE SETTINGS PROFILE s1_01294 SETTINGS readonly = 0
@@ -47,11 +47,11 @@ CREATE SETTINGS PROFILE s3_01294 TO r1_01294
 CREATE SETTINGS PROFILE s4_01294 TO r1_01294
 -- readonly ambiguity
 CREATE SETTINGS PROFILE s1_01294 SETTINGS readonly = 1
-CREATE SETTINGS PROFILE s2_01294 SETTINGS readonly READONLY
+CREATE SETTINGS PROFILE s2_01294 SETTINGS readonly CONST
 CREATE SETTINGS PROFILE s3_01294 SETTINGS INHERIT readonly
 CREATE SETTINGS PROFILE s4_01294 SETTINGS INHERIT readonly, INHERIT readonly
 CREATE SETTINGS PROFILE s5_01294 SETTINGS INHERIT readonly, readonly = 1
-CREATE SETTINGS PROFILE s6_01294 SETTINGS INHERIT readonly, readonly READONLY
+CREATE SETTINGS PROFILE s6_01294 SETTINGS INHERIT readonly, readonly CONST
 -- system.settings_profiles
 s1_01294	local directory	0	0	[]	[]
 s2_01294	local directory	1	0	['r1_01294']	[]
@@ -61,8 +61,8 @@ s5_01294	local directory	3	0	['u1_01294']	[]
 s6_01294	local directory	0	1	[]	['r1_01294','u1_01294']
 -- system.settings_profile_elements
 s2_01294	\N	\N	0	readonly	0	\N	\N	\N	\N
-s3_01294	\N	\N	0	max_memory_usage	5000000	4000000	6000000	1	\N
+s3_01294	\N	\N	0	max_memory_usage	5000000	4000000	6000000	CONST	\N
 s4_01294	\N	\N	0	max_memory_usage	5000000	\N	\N	\N	\N
 s5_01294	\N	\N	0	\N	\N	\N	\N	\N	default
 s5_01294	\N	\N	1	readonly	0	\N	\N	\N	\N
-s5_01294	\N	\N	2	max_memory_usage	\N	\N	6000000	0	\N
+s5_01294	\N	\N	2	max_memory_usage	\N	\N	6000000	WRITABLE	\N
diff --git a/tests/queries/0_stateless/01294_create_settings_profile.sql b/tests/queries/0_stateless/01294_create_settings_profile.sql
index b7dd91ad6ed2..565b4e70367b 100644
--- a/tests/queries/0_stateless/01294_create_settings_profile.sql
+++ b/tests/queries/0_stateless/01294_create_settings_profile.sql
@@ -25,9 +25,9 @@ CREATE PROFILE s2_01294 SETTINGS INHERIT 'default';
 CREATE PROFILE s3_01294 SETTINGS max_memory_usage=5000000;
 CREATE PROFILE s4_01294 SETTINGS max_memory_usage MIN=5000000;
 CREATE PROFILE s5_01294 SETTINGS max_memory_usage MAX=5000000;
-CREATE PROFILE s6_01294 SETTINGS max_memory_usage READONLY;
+CREATE PROFILE s6_01294 SETTINGS max_memory_usage CONST;
 CREATE PROFILE s7_01294 SETTINGS max_memory_usage WRITABLE;
-CREATE PROFILE s8_01294 SETTINGS max_memory_usage=5000000 MIN 4000000 MAX 6000000 READONLY;
+CREATE PROFILE s8_01294 SETTINGS max_memory_usage=5000000 MIN 4000000 MAX 6000000 CONST;
 CREATE PROFILE s9_01294 SETTINGS INHERIT 'default', max_memory_usage=5000000 WRITABLE;
 CREATE PROFILE s10_01294 SETTINGS INHERIT s1_01294, s3_01294, INHERIT default, readonly=0, max_memory_usage MAX 6000000;
 SHOW CREATE PROFILE s1_01294;
@@ -106,7 +106,7 @@ DROP PROFILE s1_01294, s2_01294, s3_01294, s4_01294, s5_01294, s6_01294;
 SELECT '-- system.settings_profiles';
 CREATE PROFILE s1_01294;
 CREATE PROFILE s2_01294 SETTINGS readonly=0 TO r1_01294;;
-CREATE PROFILE s3_01294 SETTINGS max_memory_usage=5000000 MIN 4000000 MAX 6000000 READONLY TO r1_01294;
+CREATE PROFILE s3_01294 SETTINGS max_memory_usage=5000000 MIN 4000000 MAX 6000000 CONST TO r1_01294;
 CREATE PROFILE s4_01294 SETTINGS max_memory_usage=5000000 TO r1_01294;
 CREATE PROFILE s5_01294 SETTINGS INHERIT default, readonly=0, max_memory_usage MAX 6000000 WRITABLE TO u1_01294;
 CREATE PROFILE s6_01294 TO ALL EXCEPT u1_01294, r1_01294;
diff --git a/tests/queries/0_stateless/02117_show_create_table_system.reference b/tests/queries/0_stateless/02117_show_create_table_system.reference
index d087bb556223..588788647ab5 100644
--- a/tests/queries/0_stateless/02117_show_create_table_system.reference
+++ b/tests/queries/0_stateless/02117_show_create_table_system.reference
@@ -956,7 +956,7 @@ CREATE TABLE system.settings_profile_elements
     `value` Nullable(String),
     `min` Nullable(String),
     `max` Nullable(String),
-    `readonly` Nullable(UInt8),
+    `writability` Nullable(Enum8('WRITABLE' = 0, 'CONST' = 1, 'CHANGEABLE_IN_READONLY' = 2)),
     `inherit_profile` Nullable(String)
 )
 ENGINE = SystemSettingsProfileElements
