diff --git a/tests/queries/0_stateless/03272_partition_pruning_monotonic_func_bug.reference b/tests/queries/0_stateless/03272_partition_pruning_monotonic_func_bug.reference
new file mode 100644
index 000000000000..cbb68477a293
--- /dev/null
+++ b/tests/queries/0_stateless/03272_partition_pruning_monotonic_func_bug.reference
@@ -0,0 +1,11 @@
+1
+2
+Expression
+  ReadFromMergeTree
+  Indexes:
+    PrimaryKey
+      Keys: 
+        dateTrunc(\'hour\', ts)
+      Condition: and((dateTrunc(\'hour\', ts) in (-Inf, 1731592800]), (dateTrunc(\'hour\', ts) in [1731506400, +Inf)))
+      Parts: 1/1
+      Granules: 1/1
diff --git a/tests/queries/0_stateless/03272_partition_pruning_monotonic_func_bug.sql b/tests/queries/0_stateless/03272_partition_pruning_monotonic_func_bug.sql
new file mode 100644
index 000000000000..b69b7b8a7541
--- /dev/null
+++ b/tests/queries/0_stateless/03272_partition_pruning_monotonic_func_bug.sql
@@ -0,0 +1,19 @@
+SET session_timezone = 'Etc/UTC';
+
+DROP TABLE IF EXISTS tt;
+CREATE TABLE tt
+(
+    `id` Int64,
+    `ts` DateTime
+)
+ENGINE = MergeTree()
+ORDER BY dateTrunc('hour', ts)
+SETTINGS index_granularity = 8192;
+
+INSERT INTO tt VALUES (1, '2024-11-14 00:00:00'), (2, '2024-11-14 00:00:00');
+
+SELECT id FROM tt PREWHERE ts BETWEEN toDateTime(1731506400) AND toDateTime(1731594420);
+
+explain indexes=1, description=0 SELECT id FROM tt PREWHERE ts BETWEEN toDateTime(1731506400) AND toDateTime(1731594420);
+
+DROP TABLE tt;
