diff --git a/tests/queries/0_stateless/02475_bad_cast_low_cardinality_to_string_bug.reference b/tests/queries/0_stateless/02475_bad_cast_low_cardinality_to_string_bug.reference
new file mode 100644
index 000000000000..4ffc8576e57d
--- /dev/null
+++ b/tests/queries/0_stateless/02475_bad_cast_low_cardinality_to_string_bug.reference
@@ -0,0 +1,2 @@
+bbbbb
+bbbbb
diff --git a/tests/queries/0_stateless/02475_bad_cast_low_cardinality_to_string_bug.sql b/tests/queries/0_stateless/02475_bad_cast_low_cardinality_to_string_bug.sql
new file mode 100644
index 000000000000..3b2abfb3c428
--- /dev/null
+++ b/tests/queries/0_stateless/02475_bad_cast_low_cardinality_to_string_bug.sql
@@ -0,0 +1,1 @@
+SELECT if(materialize(0), extract(materialize(CAST('aaaaaa', 'LowCardinality(String)')), '\\w'), extract(materialize(CAST('bbbbb', 'LowCardinality(String)')), '\\w*')) AS res FROM numbers(2);
