diff --git a/src/Interpreters/RedundantFunctionsInOrderByVisitor.h b/src/Interpreters/RedundantFunctionsInOrderByVisitor.h
index 6e943c83c652..d737e877f01c 100644
--- a/src/Interpreters/RedundantFunctionsInOrderByVisitor.h
+++ b/src/Interpreters/RedundantFunctionsInOrderByVisitor.h
@@ -5,6 +5,7 @@
 #include <Parsers/ASTIdentifier.h>
 #include <Parsers/ASTOrderByElement.h>
 #include <Parsers/ASTSelectQuery.h>
+#include <Parsers/ASTExpressionList.h>
 
 namespace DB
 {
@@ -75,7 +76,8 @@ class RedundantFunctionsInOrderByMatcher
 
     static bool needChildVisit(const ASTPtr & node, const ASTPtr &)
     {
-        return node->as<ASTFunction>();
+        /// Visit functions and their arguments, that are stored in ASTExpressionList.
+        return node->as<ASTFunction>() || node->as<ASTExpressionList>();
     }
 };
 
