diff --git a/tests/queries/0_stateless/01323_redundant_functions_in_order_by.reference b/tests/queries/0_stateless/01323_redundant_functions_in_order_by.reference
index fd52438e9b0a..c8421b9869ec 100644
--- a/tests/queries/0_stateless/01323_redundant_functions_in_order_by.reference
+++ b/tests/queries/0_stateless/01323_redundant_functions_in_order_by.reference
@@ -63,7 +63,9 @@ SELECT
     key,
     a
 FROM test
-ORDER BY key ASC
+ORDER BY
+    key ASC,
+    exp(key + a) ASC
 [0,1,2]
 [0,1,2]
 [0,1,2]
diff --git a/tests/queries/0_stateless/01593_functions_in_order_by.reference b/tests/queries/0_stateless/01593_functions_in_order_by.reference
new file mode 100644
index 000000000000..534b29af3e68
--- /dev/null
+++ b/tests/queries/0_stateless/01593_functions_in_order_by.reference
@@ -0,0 +1,14 @@
+SELECT
+    msg,
+    toDateTime(intDiv(ms, 1000)) AS time
+FROM 
+(
+    SELECT
+        \'hello\' AS msg,
+        toUInt64(t) * 1000 AS ms
+    FROM generateRandom(\'t DateTime\')
+    LIMIT 10
+)
+ORDER BY
+    msg ASC,
+    time ASC
diff --git a/tests/queries/0_stateless/01593_functions_in_order_by.sql b/tests/queries/0_stateless/01593_functions_in_order_by.sql
new file mode 100644
index 000000000000..2d38e45fff74
--- /dev/null
+++ b/tests/queries/0_stateless/01593_functions_in_order_by.sql
@@ -0,0 +1,11 @@
+EXPLAIN SYNTAX
+SELECT msg, toDateTime(intDiv(ms, 1000)) AS time
+FROM 
+(
+    SELECT
+        'hello' AS msg,
+        toUInt64(t) * 1000 AS ms
+    FROM generateRandom('t DateTime')
+    LIMIT 10
+)
+ORDER BY msg, time;
