diff --git a/tests/queries/0_stateless/01604_explain_ast_of_nonselect_query.reference b/tests/queries/0_stateless/01604_explain_ast_of_nonselect_query.reference
index 3545278ec1a1..3c92ffb0a8cd 100644
--- a/tests/queries/0_stateless/01604_explain_ast_of_nonselect_query.reference
+++ b/tests/queries/0_stateless/01604_explain_ast_of_nonselect_query.reference
@@ -1,6 +1,6 @@
 AlterQuery  t1 (children 1)
  ExpressionList (children 1)
-  AlterCommand 32 (children 1)
+  AlterCommand 33 (children 1)
    Function equals (children 1)
     ExpressionList (children 2)
      Identifier date
diff --git a/tests/queries/0_stateless/02097_remove_sample_by.reference b/tests/queries/0_stateless/02097_remove_sample_by.reference
new file mode 100644
index 000000000000..0747bbd5d1f1
--- /dev/null
+++ b/tests/queries/0_stateless/02097_remove_sample_by.reference
@@ -0,0 +1,3 @@
+CREATE TABLE default.t_remove_sample_by
(
    `id` UInt64
)
ENGINE = MergeTree
ORDER BY id
SETTINGS index_granularity = 8192
+CREATE TABLE default.t_remove_sample_by
(
    `id` UInt64
)
ENGINE = ReplicatedMergeTree(\'/clickhouse/default/t_remove_sample_by\', \'1\')
ORDER BY id
SETTINGS index_granularity = 8192
+CREATE TABLE default.t_remove_sample_by
(
    `id` String
)
ENGINE = MergeTree
ORDER BY id
SETTINGS index_granularity = 8192
diff --git a/tests/queries/0_stateless/02097_remove_sample_by.sql b/tests/queries/0_stateless/02097_remove_sample_by.sql
new file mode 100644
index 000000000000..89fbfe0c4c57
--- /dev/null
+++ b/tests/queries/0_stateless/02097_remove_sample_by.sql
@@ -0,0 +1,44 @@
+-- Tags: zookeeper
+
+DROP TABLE IF EXISTS t_remove_sample_by;
+
+CREATE TABLE t_remove_sample_by(id UInt64) ENGINE = MergeTree ORDER BY id SAMPLE BY id;
+
+ALTER TABLE t_remove_sample_by REMOVE SAMPLE BY;
+SHOW CREATE TABLE t_remove_sample_by;
+
+ALTER TABLE t_remove_sample_by REMOVE SAMPLE BY; -- { serverError 36 }
+SELECT * FROM t_remove_sample_by SAMPLE 1 / 10; -- { serverError 141 }
+
+DROP TABLE t_remove_sample_by;
+
+CREATE TABLE t_remove_sample_by(id UInt64)
+ENGINE = ReplicatedMergeTree('/clickhouse/{database}/t_remove_sample_by', '1')
+ORDER BY id SAMPLE BY id;
+
+ALTER TABLE t_remove_sample_by REMOVE SAMPLE BY;
+SHOW CREATE TABLE t_remove_sample_by;
+
+DROP TABLE t_remove_sample_by;
+
+CREATE TABLE t_remove_sample_by(id UInt64) ENGINE = Memory;
+ALTER TABLE t_remove_sample_by REMOVE SAMPLE BY; -- { serverError 36 }
+
+DROP TABLE t_remove_sample_by;
+
+CREATE TABLE t_remove_sample_by(id String)
+ENGINE = MergeTree ORDER BY id SAMPLE BY id
+SETTINGS check_sample_column_is_correct = 0;
+
+ALTER TABLE t_remove_sample_by RESET SETTING check_sample_column_is_correct;
+
+DETACH TABLE t_remove_sample_by;
+ATTACH TABLE t_remove_sample_by;
+
+INSERT INTO t_remove_sample_by VALUES (1);
+SELECT * FROM t_remove_sample_by SAMPLE 1 / 10; -- { serverError 59 }
+
+ALTER TABLE t_remove_sample_by REMOVE SAMPLE BY;
+SHOW CREATE TABLE t_remove_sample_by;
+
+DROP TABLE t_remove_sample_by;
