diff --git a/tests/queries/0_stateless/25340_logical_optimizer_alias_bug.reference b/tests/queries/0_stateless/25340_logical_optimizer_alias_bug.reference
new file mode 100644
index 000000000000..573541ac9702
--- /dev/null
+++ b/tests/queries/0_stateless/25340_logical_optimizer_alias_bug.reference
@@ -0,0 +1,1 @@
+0
diff --git a/tests/queries/0_stateless/25340_logical_optimizer_alias_bug.sql b/tests/queries/0_stateless/25340_logical_optimizer_alias_bug.sql
new file mode 100644
index 000000000000..5b13eea5e4be
--- /dev/null
+++ b/tests/queries/0_stateless/25340_logical_optimizer_alias_bug.sql
@@ -0,0 +1,2 @@
+create table test_local (id UInt32, path LowCardinality(String)) engine = MergeTree order by id;
+WITH ((position(path, '/a') > 0) AND (NOT (position(path, 'a') > 0))) OR (path = '/b') OR (path = '/b/') as alias1 SELECT max(alias1) FROM remote('127.0.0.{1,2}', currentDatabase(), test_local) WHERE (id = 299386662);
