diff --git a/src/Core/MySQL/MySQLReplication.cpp b/src/Core/MySQL/MySQLReplication.cpp
index 6f3ac1b40e95..b211d746a2fd 100644
--- a/src/Core/MySQL/MySQLReplication.cpp
+++ b/src/Core/MySQL/MySQLReplication.cpp
@@ -118,7 +118,7 @@ namespace MySQLReplication
         }
         else if (query.starts_with("SAVEPOINT"))
         {
-            throw ReplicationError("ParseQueryEvent: Unsupported query event:" + query, ErrorCodes::LOGICAL_ERROR);
+            typ = QUERY_SAVEPOINT;
         }
     }
 
@@ -941,6 +941,7 @@ namespace MySQLReplication
                 {
                     case QUERY_EVENT_MULTI_TXN_FLAG:
                     case QUERY_EVENT_XA:
+                    case QUERY_SAVEPOINT:
                     {
                         event = std::make_shared<DryRunEvent>(std::move(query->header));
                         break;
diff --git a/src/Core/MySQL/MySQLReplication.h b/src/Core/MySQL/MySQLReplication.h
index 8900eee01021..5825924d10b1 100644
--- a/src/Core/MySQL/MySQLReplication.h
+++ b/src/Core/MySQL/MySQLReplication.h
@@ -368,7 +368,8 @@ namespace MySQLReplication
     {
         QUERY_EVENT_DDL = 0,
         QUERY_EVENT_MULTI_TXN_FLAG = 1,
-        QUERY_EVENT_XA = 2
+        QUERY_EVENT_XA = 2,
+        QUERY_SAVEPOINT = 3,
     };
 
     class QueryEvent : public EventBase
