diff --git a/tools/shell/shell-test.py b/tools/shell/shell-test.py
index a9026d229187..72ffeff56ae8 100644
--- a/tools/shell/shell-test.py
+++ b/tools/shell/shell-test.py
@@ -378,6 +378,13 @@ def tf():
 SELECT 42;
 ''', out="42", err="Query Profiling Information")
 
+# escapes in query profiling
+test("""
+PRAGMA enable_profiling=json;
+CREATE TABLE "foo"("hello world" INT);
+SELECT "hello world", '\r\t
\b\f\\' FROM "foo";
+""", err="""SELECT \\"hello world\\", '\\r\\t\
\\b\\f\\\\' FROM \\"foo""")
+
 test('.system echo 42', out="42")
 test('.shell echo 42', out="42")
 
