diff --git a/dbms/tests/queries/0_stateless/00963_startsWith_force_primary_key.reference b/dbms/tests/queries/0_stateless/00963_startsWith_force_primary_key.reference
new file mode 100644
index 000000000000..f1e5eeed2d93
--- /dev/null
+++ b/dbms/tests/queries/0_stateless/00963_startsWith_force_primary_key.reference
@@ -0,0 +1,2 @@
+3
+2
diff --git a/dbms/tests/queries/0_stateless/00963_startsWith_force_primary_key.sql b/dbms/tests/queries/0_stateless/00963_startsWith_force_primary_key.sql
new file mode 100644
index 000000000000..b3895a93b641
--- /dev/null
+++ b/dbms/tests/queries/0_stateless/00963_startsWith_force_primary_key.sql
@@ -0,0 +1,6 @@
+DROP TABLE IF EXISTS test_startsWith;
+CREATE TABLE test_startsWith (a String) Engine = MergeTree PARTITION BY tuple() ORDER BY a;
+INSERT INTO test_startsWith (a) values ('a'), ('abcd'), ('bbb'), (''), ('abc');
+SELECT count() from test_startsWith where startsWith(a, 'a') settings force_primary_key=1;
+SELECT count() from test_startsWith where startsWith(a, 'abc') settings force_primary_key=1;
+DROP TABLE test_startsWith;
