diff --git a/test/issues/general/test_13252.test b/test/issues/general/test_13252.test
new file mode 100644
index 000000000000..95ef3bcb1728
--- /dev/null
+++ b/test/issues/general/test_13252.test
@@ -0,0 +1,13 @@
+# name: test/issues/general/test_13252.test
+# description: DuckDB Internal Error with the UNION_EXTRACT
+# group: [general]
+
+require json
+
+statement ok
+PRAGMA enable_verification
+
+statement error
+WITH tbl1 AS (SELECT NULL AS u) SELECT UNION_EXTRACT(CAST(u AS JSON), 'a') FROM tbl1;
+----
+union_extract can only take a union parameter
diff --git a/test/sql/function/generic/test_least_greatest.test b/test/sql/function/generic/test_least_greatest.test
index 2b7883bd6f23..e743f1888f81 100644
--- a/test/sql/function/generic/test_least_greatest.test
+++ b/test/sql/function/generic/test_least_greatest.test
@@ -11,6 +11,11 @@ SELECT LEAST(1)
 ----
 1
 
+query I
+SELECT LEAST('hello world')
+----
+hello world
+
 query I
 SELECT LEAST(1, 3)
 ----
