diff --git a/helio b/helio
index f76c73fc6ca8..d819bf4f1bc6 160000
--- a/helio
+++ b/helio
@@ -1,1 +1,1 @@
-Subproject commit f76c73fc6ca8cf1ada04edbb7e64a2465aa0f5b1
+Subproject commit d819bf4f1bc61e72d9fbe0c37b2ce40354d9689d
diff --git a/src/server/main_service.cc b/src/server/main_service.cc
index fcc00b48235f..9ecbf4b1e791 100644
--- a/src/server/main_service.cc
+++ b/src/server/main_service.cc
@@ -1022,7 +1022,7 @@ std::optional<ErrorReply> Service::VerifyCommandState(const CommandId* cid, CmdA
       allowed_by_state = false;
       break;
     case GlobalState::TAKEN_OVER:
-      allowed_by_state = cid->name() == "REPLCONF" || cid->name() == "SAVE";
+      allowed_by_state = !cid->IsWriteOnly();
       break;
     default:
       break;
