diff --git a/dbms/src/Interpreters/tests/internal_iotop.cpp b/dbms/src/Interpreters/tests/internal_iotop.cpp
index b3cc720bc8e2..487390c83217 100644
--- a/dbms/src/Interpreters/tests/internal_iotop.cpp
+++ b/dbms/src/Interpreters/tests/internal_iotop.cpp
@@ -5,6 +5,7 @@
 #include <Common/TaskStatsInfoGetter.h>
 #include <Poco/File.h>
 #include <Common/Stopwatch.h>
+#include <common/getThreadId.h>
 #include <IO/WriteBufferFromString.h>
 #include <linux/taskstats.h>
 #include <sys/time.h>
@@ -47,7 +48,7 @@ using namespace DB;
 static void do_io(size_t id)
 {
     ::taskstats stat;
-    int tid = TaskStatsInfoGetter::getCurrentTID();
+    int tid = getThreadId();
     TaskStatsInfoGetter get_info;
 
     get_info.getStat(stat, tid);
@@ -103,7 +104,7 @@ static void test_perf()
 {
 
     ::taskstats stat;
-    int tid = TaskStatsInfoGetter::getCurrentTID();
+    int tid = getThreadId();
     TaskStatsInfoGetter get_info;
 
     rusage rusage;
diff --git a/dbms/tests/queries/0_stateless/00634_performance_introspection_and_logging.sh b/dbms/tests/queries/0_stateless/00634_performance_introspection_and_logging.sh
index c8cd8606cc14..1225cd43c5ec 100755
--- a/dbms/tests/queries/0_stateless/00634_performance_introspection_and_logging.sh
+++ b/dbms/tests/queries/0_stateless/00634_performance_introspection_and_logging.sh
@@ -45,7 +45,7 @@ SELECT
     -- duration, threads_realtime, threads_time_user_system_io,
     threads_realtime >= 0.99 * duration,
     threads_realtime >= threads_time_user_system_io,
-    any(length(thread_numbers)) >= 1
+    any(length(thread_ids)) >= 1
     FROM
         (SELECT * FROM system.query_log PREWHERE query='$heavy_cpu_query' WHERE event_date >= today()-1 AND type=2 ORDER BY event_time DESC LIMIT 1)
     ARRAY JOIN ProfileEvents.Names AS PN, ProfileEvents.Values AS PV"
@@ -57,24 +57,24 @@ $CLICKHOUSE_CLIENT $settings --max_threads=3 -q "$heavy_cpu_query"
 $CLICKHOUSE_CLIENT $settings -q "SYSTEM FLUSH LOGS"
 query_id=`$CLICKHOUSE_CLIENT $settings -q "SELECT query_id FROM system.query_log WHERE event_date >= today()-1 AND type=2 AND query='$heavy_cpu_query' ORDER BY event_time DESC LIMIT 1"`
 query_elapsed=`$CLICKHOUSE_CLIENT $settings -q "SELECT query_duration_ms*1000 FROM system.query_log WHERE event_date >= today()-1 AND type=2 AND query_id='$query_id' ORDER BY event_time DESC LIMIT 1"`
-threads=`$CLICKHOUSE_CLIENT $settings -q "SELECT length(thread_numbers) FROM system.query_log WHERE event_date >= today()-1 AND type=2 AND query_id='$query_id' ORDER BY event_time DESC LIMIT 1"`
+threads=`$CLICKHOUSE_CLIENT $settings -q "SELECT length(thread_ids) FROM system.query_log WHERE event_date >= today()-1 AND type=2 AND query_id='$query_id' ORDER BY event_time DESC LIMIT 1"`
 
 $CLICKHOUSE_CLIENT $settings -q "
 SELECT
     -- max(thread_realtime), $query_elapsed, max(thread_time_user_system_io),
     0.9 * $query_elapsed <= max(thread_realtime) AND max(thread_realtime) <= 1.1 * $query_elapsed,
     0.7 * $query_elapsed <= max(thread_time_user_system_io) AND max(thread_time_user_system_io) <= 1.3 * $query_elapsed,
-    uniqExact(thread_number) = $threads
+    uniqExact(thread_id) = $threads
 FROM
 (
     SELECT
-        thread_number,
+        thread_id,
         sumIf(PV, PN = 'RealTimeMicroseconds') AS thread_realtime,
         sumIf(PV, PN IN ('UserTimeMicroseconds', 'SystemTimeMicroseconds', 'OSIOWaitMicroseconds', 'OSCPUWaitMicroseconds')) AS thread_time_user_system_io
         FROM
             (SELECT * FROM system.query_thread_log PREWHERE query_id='$query_id' WHERE event_date >= today()-1)
         ARRAY JOIN ProfileEvents.Names AS PN, ProfileEvents.Values AS PV
-        GROUP BY thread_number
+        GROUP BY thread_id
 )
 "
 
