diff --git a/tests/queries/0_stateless/02462_match_regexp_pk.reference b/tests/queries/0_stateless/02462_match_regexp_pk.reference
new file mode 100644
index 000000000000..428d6556f4cc
--- /dev/null
+++ b/tests/queries/0_stateless/02462_match_regexp_pk.reference
@@ -0,0 +1,5 @@
+4
+1
+3
+4
+4
diff --git a/tests/queries/0_stateless/02462_match_regexp_pk.sql b/tests/queries/0_stateless/02462_match_regexp_pk.sql
new file mode 100644
index 000000000000..1a944b961962
--- /dev/null
+++ b/tests/queries/0_stateless/02462_match_regexp_pk.sql
@@ -0,0 +1,9 @@
+CREATE TABLE mt_match_pk (v String) ENGINE = MergeTree ORDER BY v SETTINGS index_granularity = 1;
+INSERT INTO mt_match_pk VALUES ('a'), ('aaa'), ('aba'), ('bac'), ('acccca');
+
+SET force_primary_key = 1;
+SELECT count() FROM mt_match_pk WHERE match(v, '^a');
+SELECT count() FROM mt_match_pk WHERE match(v, '^ab');
+SELECT count() FROM mt_match_pk WHERE match(v, '^a.');
+SELECT count() FROM mt_match_pk WHERE match(v, '^ab*');
+SELECT count() FROM mt_match_pk WHERE match(v, '^ac?');
