diff --git a/src/core/qlist_test.cc b/src/core/qlist_test.cc
index 9e5d67c07fff..64d1b52788ea 100644
--- a/src/core/qlist_test.cc
+++ b/src/core/qlist_test.cc
@@ -282,6 +282,15 @@ TEST_F(QListTest, CompressionPlain) {
   EXPECT_EQ(500, i);
 }
 
+TEST_F(QListTest, LargeValues) {
+  string val(100000, 'a');
+  ql_.Push(val, QList::HEAD);
+  ql_.Push(val, QList::HEAD);
+  ql_.Pop(QList::HEAD);
+  auto items = ToItems();
+  EXPECT_THAT(items, ElementsAre(val));
+}
+
 using FillCompress = tuple<int, unsigned>;
 
 class PrintToFillCompress {
diff --git a/tests/dragonfly/seeder/script-genlib.lua b/tests/dragonfly/seeder/script-genlib.lua
index 45d672a0a199..2203df175f32 100644
--- a/tests/dragonfly/seeder/script-genlib.lua
+++ b/tests/dragonfly/seeder/script-genlib.lua
@@ -56,9 +56,7 @@ end
 
 function LG_funcs.add_list(key, keys)
     local is_huge = keys[key]
-    --- TODO -- investigate why second case of replication_test_all fails
-    --- we somehow create a quicklist that is circular and we deadlock
-    redis.apcall('LPUSH', key, unpack(randstr_sequence(false)))
+    redis.apcall('LPUSH', key, unpack(randstr_sequence(is_huge)))
 end
 
 function LG_funcs.mod_list(key, keys)
