diff --git a/src/Interpreters/ProcessList.cpp b/src/Interpreters/ProcessList.cpp
index 37b2992d657a..3c91f00ebf4c 100644
--- a/src/Interpreters/ProcessList.cpp
+++ b/src/Interpreters/ProcessList.cpp
@@ -17,6 +17,11 @@
 #include <chrono>
 
 
+namespace CurrentMetrics
+{
+    extern const Metric Query;
+}
+
 namespace DB
 {
 
@@ -313,6 +318,7 @@ QueryStatus::QueryStatus(
     , client_info(client_info_)
     , priority_handle(std::move(priority_handle_))
     , query_kind(query_kind_)
+    , num_queries_increment(CurrentMetrics::Query)
 {
     auto settings = getContext()->getSettings();
     limits.max_execution_time = settings.max_execution_time;
diff --git a/src/Interpreters/ProcessList.h b/src/Interpreters/ProcessList.h
index 545e5b073453..2ba0b0814ee9 100644
--- a/src/Interpreters/ProcessList.h
+++ b/src/Interpreters/ProcessList.h
@@ -121,6 +121,10 @@ class QueryStatus : public WithContext
 
     IAST::QueryKind query_kind;
 
+    /// This field is unused in this class, but it
+    /// increments/decrements metric in constructor/destructor.
+    CurrentMetrics::Increment num_queries_increment;
+
 public:
 
     QueryStatus(
