diff --git a/tests/queries/0_stateless/02190_current_metrics_query.reference b/tests/queries/0_stateless/02190_current_metrics_query.reference
new file mode 100644
index 000000000000..d00491fd7e5b
--- /dev/null
+++ b/tests/queries/0_stateless/02190_current_metrics_query.reference
@@ -0,0 +1,1 @@
+1
diff --git a/tests/queries/0_stateless/02190_current_metrics_query.sql b/tests/queries/0_stateless/02190_current_metrics_query.sql
new file mode 100644
index 000000000000..e8b22e92a99f
--- /dev/null
+++ b/tests/queries/0_stateless/02190_current_metrics_query.sql
@@ -0,0 +1,2 @@
+-- This query itself is also accounted in metric.
+SELECT value > 0 FROM system.metrics WHERE metric = 'Query';
