diff --git a/tests/queries/0_stateless/03203_fill_missed_subcolumns.reference b/tests/queries/0_stateless/03203_fill_missed_subcolumns.reference
index 9faa3b358489..ce9454ac66d9 100644
--- a/tests/queries/0_stateless/03203_fill_missed_subcolumns.reference
+++ b/tests/queries/0_stateless/03203_fill_missed_subcolumns.reference
@@ -23,6 +23,11 @@
 3	[1]	[1]
 4	[1,0]	[0,1]
 5	[1]	[1]
+2
+1
+1
+2
+1
 1	('foo','bar')	[1,NULL,3]
 2	('aaa','bbb')	[1,NULL,3]
 3	('ccc','ddd')	[4,5,6]
diff --git a/tests/queries/0_stateless/03203_fill_missed_subcolumns.sql b/tests/queries/0_stateless/03203_fill_missed_subcolumns.sql
index 2789c9de35cf..9fd69c1105c9 100644
--- a/tests/queries/0_stateless/03203_fill_missed_subcolumns.sql
+++ b/tests/queries/0_stateless/03203_fill_missed_subcolumns.sql
@@ -25,6 +25,7 @@ INSERT INTO t_missed_subcolumns VALUES (4, [NULL, 'bbb'], ['ddd', NULL]), (5, [N
 SELECT id, n.a, n.b FROM t_missed_subcolumns ORDER BY id;
 SELECT id, n.a.size0, n.b.size0 FROM t_missed_subcolumns ORDER BY id;
 SELECT id, n.a.null, n.b.null FROM t_missed_subcolumns ORDER BY id;
+SELECT n.b.size0 FROM t_missed_subcolumns ORDER BY id;
 
 DROP TABLE IF EXISTS t_missed_subcolumns;
 
diff --git a/tests/queries/0_stateless/03352_concurrent_rename_alter.sh b/tests/queries/0_stateless/03352_concurrent_rename_alter.sh
index 4571d24350f9..dd7fcbf13765 100755
--- a/tests/queries/0_stateless/03352_concurrent_rename_alter.sh
+++ b/tests/queries/0_stateless/03352_concurrent_rename_alter.sh
@@ -4,6 +4,9 @@ CUR_DIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
 # shellcheck source=../shell_config.sh
 . "$CUR_DIR"/../shell_config.sh
 
+# shellcheck source=./mergetree_mutations.lib
+. "$CUR_DIR"/mergetree_mutations.lib
+
 $CLICKHOUSE_CLIENT --query "
     DROP TABLE IF EXISTS t_rename_alter SYNC;
 
