diff --git a/dbms/tests/queries/0_stateless/02001_empty_join_and_totals.reference b/dbms/tests/queries/0_stateless/02001_empty_join_and_totals.reference
new file mode 100644
index 000000000000..e69de29bb2d1
diff --git a/dbms/tests/queries/0_stateless/02001_empty_join_and_totals.sql b/dbms/tests/queries/0_stateless/02001_empty_join_and_totals.sql
new file mode 100644
index 000000000000..2424e7bbae91
--- /dev/null
+++ b/dbms/tests/queries/0_stateless/02001_empty_join_and_totals.sql
@@ -0,0 +1,11 @@
+select * from system.one t1
+join system.one t2
+on t1.dummy = t2.dummy
+limit 0
+FORMAT TabSeparated;
+
+select * from system.one t1
+join system.one t2
+on t1.dummy = t2.dummy
+where t2.dummy > 0
+FORMAT TabSeparated;
