diff --git a/docs/en/sql-reference/table-functions/index.md b/docs/en/sql-reference/table-functions/index.md
index d09adcd13d61..94b23bc695c6 100644
--- a/docs/en/sql-reference/table-functions/index.md
+++ b/docs/en/sql-reference/table-functions/index.md
@@ -39,3 +39,7 @@ You can’t use table functions if the [allow_ddl](../../operations/settings/per
 | [s3](../../sql-reference/table-functions/s3.md)                  | Creates a [S3](../../engines/table-engines/integrations/s3.md)-engine table.                                                           |
 | [sqlite](../../sql-reference/table-functions/sqlite.md)          | Creates a [sqlite](../../engines/table-engines/integrations/sqlite.md)-engine table.                                                       |
 
+:::note
+Only these table functions are enabled in readonly mode :
+null, view, viewIfPermitted, numbers, numbers_mt, generateRandom, values, cluster, clusterAllReplicas 
+:::
\ No newline at end of file
diff --git a/src/Storages/System/StorageSystemTableFunctions.cpp b/src/Storages/System/StorageSystemTableFunctions.cpp
index 308cbc5686db..07a504edc5e7 100644
--- a/src/Storages/System/StorageSystemTableFunctions.cpp
+++ b/src/Storages/System/StorageSystemTableFunctions.cpp
@@ -1,16 +1,23 @@
 #include <Storages/System/StorageSystemTableFunctions.h>
-
 #include <TableFunctions/TableFunctionFactory.h>
+#include <DataTypes/DataTypesNumber.h>
+
 namespace DB
 {
 
+namespace ErrorCodes
+{
+    extern const int UNKNOWN_FUNCTION;
+}
+
 NamesAndTypesList StorageSystemTableFunctions::getNamesAndTypes()
 {
     return
-    {
-        {"name", std::make_shared<DataTypeString>()},
-        {"description", std::make_shared<DataTypeString>()}
-    };
+        {
+            {"name", std::make_shared<DataTypeString>()},
+            {"description", std::make_shared<DataTypeString>()},
+            {"allow_readonly", std::make_shared<DataTypeUInt8>()}
+       };
 }
 
 void StorageSystemTableFunctions::fillData(MutableColumns & res_columns, ContextPtr, const SelectQueryInfo &) const
@@ -20,7 +27,15 @@ void StorageSystemTableFunctions::fillData(MutableColumns & res_columns, Context
     for (const auto & function_name : functions_names)
     {
         res_columns[0]->insert(function_name);
-        res_columns[1]->insert(factory.getDocumentation(function_name).description);
+
+        auto properties = factory.tryGetProperties(function_name);
+        if (properties)
+        {
+            res_columns[1]->insert(properties->documentation.description);
+            res_columns[2]->insert(properties->allow_readonly);
+        }
+        else
+            throw Exception(ErrorCodes::UNKNOWN_FUNCTION, "Unknown table function {}", function_name);
     }
 }
 
diff --git a/src/TableFunctions/ITableFunction.cpp b/src/TableFunctions/ITableFunction.cpp
index 82b6230dc304..da0de7e47f6e 100644
--- a/src/TableFunctions/ITableFunction.cpp
+++ b/src/TableFunctions/ITableFunction.cpp
@@ -4,6 +4,7 @@
 #include <Storages/StorageTableFunction.h>
 #include <Access/Common/AccessFlags.h>
 #include <Common/ProfileEvents.h>
+#include <TableFunctions/TableFunctionFactory.h>
 
 
 namespace ProfileEvents
@@ -25,8 +26,8 @@ StoragePtr ITableFunction::execute(const ASTPtr & ast_function, ContextPtr conte
     ProfileEvents::increment(ProfileEvents::TableFunctionExecute);
 
     AccessFlags required_access = getSourceAccessType();
-    String function_name = getName();
-    if ((function_name != "null") && (function_name != "view") && (function_name != "viewIfPermitted"))
+    auto table_function_properties = TableFunctionFactory::instance().tryGetProperties(getName());
+    if (!(table_function_properties && table_function_properties->allow_readonly))
         required_access |= AccessType::CREATE_TEMPORARY_TABLE;
     context->checkAccess(required_access);
 
diff --git a/src/TableFunctions/ITableFunction.h b/src/TableFunctions/ITableFunction.h
index 4b9a87b93f17..a05edcd32c82 100644
--- a/src/TableFunctions/ITableFunction.h
+++ b/src/TableFunctions/ITableFunction.h
@@ -4,6 +4,7 @@
 #include <Storages/IStorage_fwd.h>
 #include <Storages/ColumnsDescription.h>
 #include <Access/Common/AccessType.h>
+#include <Common/Documentation.h>
 
 #include <memory>
 #include <string>
@@ -79,6 +80,14 @@ class ITableFunction : public std::enable_shared_from_this<ITableFunction>
     virtual const char * getStorageTypeName() const = 0;
 };
 
+/// Properties of table function that are independent of argument types and parameters.
+struct TableFunctionProperties
+{
+    Documentation documentation;
+    bool allow_readonly = false;
+};
+
+
 using TableFunctionPtr = std::shared_ptr<ITableFunction>;
 
 
diff --git a/src/TableFunctions/TableFunctionExplain.cpp b/src/TableFunctions/TableFunctionExplain.cpp
index 02493aa9b0e7..02b9308ed224 100644
--- a/src/TableFunctions/TableFunctionExplain.cpp
+++ b/src/TableFunctions/TableFunctionExplain.cpp
@@ -91,7 +91,7 @@ InterpreterExplainQuery TableFunctionExplain::getInterpreter(ContextPtr context)
 
 void registerTableFunctionExplain(TableFunctionFactory & factory)
 {
-    factory.registerFunction<TableFunctionExplain>({R"(
+    factory.registerFunction<TableFunctionExplain>({.documentation = {R"(
 Returns result of EXPLAIN query.
 
 The function should not be called directly but can be invoked via `SELECT * FROM (EXPLAIN <query>)`.
@@ -103,7 +103,7 @@ You can use this query to process the result of EXPLAIN further using SQL (e.g.,
 
 )",
 {{"1", "SELECT explain FROM (EXPLAIN AST SELECT * FROM system.numbers) WHERE explain LIKE '%Asterisk%'"}}
-});
+}});
 
 }
 
diff --git a/src/TableFunctions/TableFunctionFactory.cpp b/src/TableFunctions/TableFunctionFactory.cpp
index 5ed22e393005..3d2a72ddc9e6 100644
--- a/src/TableFunctions/TableFunctionFactory.cpp
+++ b/src/TableFunctions/TableFunctionFactory.cpp
@@ -16,16 +16,15 @@ namespace ErrorCodes
     extern const int LOGICAL_ERROR;
 }
 
-
 void TableFunctionFactory::registerFunction(
-    const std::string & name, TableFunctionCreator creator, Documentation doc, CaseSensitiveness case_sensitiveness)
+    const std::string & name, Value value, CaseSensitiveness case_sensitiveness)
 {
-    if (!table_functions.emplace(name, TableFunctionFactoryData{creator, doc}).second)
+    if (!table_functions.emplace(name, value).second)
         throw Exception("TableFunctionFactory: the table function name '" + name + "' is not unique",
             ErrorCodes::LOGICAL_ERROR);
 
     if (case_sensitiveness == CaseInsensitive
-        && !case_insensitive_table_functions.emplace(Poco::toLower(name), TableFunctionFactoryData{creator, doc}).second)
+        && !case_insensitive_table_functions.emplace(Poco::toLower(name), value).second)
         throw Exception("TableFunctionFactory: the case insensitive table function name '" + name + "' is not unique",
                         ErrorCodes::LOGICAL_ERROR);
 }
@@ -59,13 +58,13 @@ TableFunctionPtr TableFunctionFactory::tryGet(
     auto it = table_functions.find(name);
     if (table_functions.end() != it)
     {
-        res = it->second.first();
+        res = it->second.creator();
     }
     else
     {
         it = case_insensitive_table_functions.find(Poco::toLower(name));
         if (case_insensitive_table_functions.end() != it)
-            res = it->second.first();
+            res = it->second.creator();
     }
 
     if (!res)
@@ -86,13 +85,29 @@ bool TableFunctionFactory::isTableFunctionName(const std::string & name) const
     return table_functions.contains(name);
 }
 
-Documentation TableFunctionFactory::getDocumentation(const std::string & name) const
+std::optional<TableFunctionProperties> TableFunctionFactory::tryGetProperties(const String & name) const
 {
-    auto it = table_functions.find(name);
-    if (it == table_functions.end())
-        throw Exception(ErrorCodes::UNKNOWN_FUNCTION, "Unknown table function {}", name);
+    return tryGetPropertiesImpl(name);
+}
+
+std::optional<TableFunctionProperties> TableFunctionFactory::tryGetPropertiesImpl(const String & name_param) const
+{
+    String name = getAliasToOrName(name_param);
+    Value found;
+
+    /// Find by exact match.
+    if (auto it = table_functions.find(name); it != table_functions.end())
+    {
+        found = it->second;
+    }
+
+    if (auto jt = case_insensitive_table_functions.find(Poco::toLower(name)); jt != case_insensitive_table_functions.end())
+        found = jt->second;
+
+    if (found.creator)
+        return found.properties;
 
-    return it->second.second;
+    return {};
 }
 
 TableFunctionFactory & TableFunctionFactory::instance()
diff --git a/src/TableFunctions/TableFunctionFactory.h b/src/TableFunctions/TableFunctionFactory.h
index 8ff352ff9acf..2cc648ba1815 100644
--- a/src/TableFunctions/TableFunctionFactory.h
+++ b/src/TableFunctions/TableFunctionFactory.h
@@ -3,7 +3,6 @@
 #include <TableFunctions/ITableFunction.h>
 #include <Common/IFactoryWithAliases.h>
 #include <Common/NamePrompter.h>
-#include <Common/Documentation.h>
 
 #include <functional>
 #include <memory>
@@ -18,7 +17,24 @@ namespace DB
 class Context;
 
 using TableFunctionCreator = std::function<TableFunctionPtr()>;
-using TableFunctionFactoryData = std::pair<TableFunctionCreator, Documentation>;
+
+struct TableFunctionFactoryData
+{
+    TableFunctionCreator creator;
+    TableFunctionProperties properties;
+
+    TableFunctionFactoryData() = default;
+    TableFunctionFactoryData(const TableFunctionFactoryData &) = default;
+    TableFunctionFactoryData & operator = (const TableFunctionFactoryData &) = default;
+
+    template <typename Creator>
+        requires (!std::is_same_v<Creator, TableFunctionFactoryData>)
+    TableFunctionFactoryData(Creator creator_, TableFunctionProperties properties_ = {}) /// NOLINT
+        : creator(std::forward<Creator>(creator_)), properties(std::move(properties_))
+    {
+    }
+};
+
 
 /** Lets you get a table function by its name.
   */
@@ -31,15 +47,16 @@ class TableFunctionFactory final: private boost::noncopyable, public IFactoryWit
     /// No locking, you must register all functions before usage of get.
     void registerFunction(
         const std::string & name,
-        TableFunctionCreator creator,
-        Documentation doc = {},
+        Value value,
         CaseSensitiveness case_sensitiveness = CaseSensitive);
 
     template <typename Function>
-    void registerFunction(Documentation doc = {}, CaseSensitiveness case_sensitiveness = CaseSensitive)
+    void registerFunction(TableFunctionProperties properties = {}, CaseSensitiveness case_sensitiveness = CaseSensitive)
     {
         auto creator = []() -> TableFunctionPtr { return std::make_shared<Function>(); };
-        registerFunction(Function::name, std::move(creator), std::move(doc), case_sensitiveness);
+        registerFunction(Function::name,
+                         TableFunctionFactoryData{std::move(creator), {std::move(properties)}} ,
+                         case_sensitiveness);
     }
 
     /// Throws an exception if not found.
@@ -48,7 +65,7 @@ class TableFunctionFactory final: private boost::noncopyable, public IFactoryWit
     /// Returns nullptr if not found.
     TableFunctionPtr tryGet(const std::string & name, ContextPtr context) const;
 
-    Documentation getDocumentation(const std::string & name) const;
+    std::optional<TableFunctionProperties> tryGetProperties(const String & name) const;
 
     bool isTableFunctionName(const std::string & name) const;
 
@@ -61,6 +78,8 @@ class TableFunctionFactory final: private boost::noncopyable, public IFactoryWit
 
     String getFactoryName() const override { return "TableFunctionFactory"; }
 
+    std::optional<TableFunctionProperties> tryGetPropertiesImpl(const String & name) const;
+
     TableFunctions table_functions;
     TableFunctions case_insensitive_table_functions;
 };
diff --git a/src/TableFunctions/TableFunctionGenerateRandom.cpp b/src/TableFunctions/TableFunctionGenerateRandom.cpp
index 8d1c06c7c4dd..1ddbb48962d4 100644
--- a/src/TableFunctions/TableFunctionGenerateRandom.cpp
+++ b/src/TableFunctions/TableFunctionGenerateRandom.cpp
@@ -91,7 +91,7 @@ StoragePtr TableFunctionGenerateRandom::executeImpl(const ASTPtr & /*ast_functio
 
 void registerTableFunctionGenerate(TableFunctionFactory & factory)
 {
-    factory.registerFunction<TableFunctionGenerateRandom>();
+    factory.registerFunction<TableFunctionGenerateRandom>({.documentation = {}, .allow_readonly = true});
 }
 
 }
diff --git a/src/TableFunctions/TableFunctionNull.cpp b/src/TableFunctions/TableFunctionNull.cpp
index 1a35830b06c4..9ff07cc19460 100644
--- a/src/TableFunctions/TableFunctionNull.cpp
+++ b/src/TableFunctions/TableFunctionNull.cpp
@@ -52,6 +52,6 @@ StoragePtr TableFunctionNull::executeImpl(const ASTPtr & /*ast_function*/, Conte
 
 void registerTableFunctionNull(TableFunctionFactory & factory)
 {
-    factory.registerFunction<TableFunctionNull>();
+    factory.registerFunction<TableFunctionNull>({.documentation = {}, .allow_readonly = true});
 }
 }
diff --git a/src/TableFunctions/TableFunctionNumbers.cpp b/src/TableFunctions/TableFunctionNumbers.cpp
index 0ef1d3815d9d..2056cd838f58 100644
--- a/src/TableFunctions/TableFunctionNumbers.cpp
+++ b/src/TableFunctions/TableFunctionNumbers.cpp
@@ -51,8 +51,8 @@ StoragePtr TableFunctionNumbers<multithreaded>::executeImpl(const ASTPtr & ast_f
 
 void registerTableFunctionNumbers(TableFunctionFactory & factory)
 {
-    factory.registerFunction<TableFunctionNumbers<true>>();
-    factory.registerFunction<TableFunctionNumbers<false>>();
+    factory.registerFunction<TableFunctionNumbers<true>>({.documentation = {}, .allow_readonly = true});
+    factory.registerFunction<TableFunctionNumbers<false>>({.documentation = {}, .allow_readonly = true});
 }
 
 template <bool multithreaded>
diff --git a/src/TableFunctions/TableFunctionRemote.cpp b/src/TableFunctions/TableFunctionRemote.cpp
index 2c62e29810f6..14e0774cf06c 100644
--- a/src/TableFunctions/TableFunctionRemote.cpp
+++ b/src/TableFunctions/TableFunctionRemote.cpp
@@ -346,8 +346,8 @@ void registerTableFunctionRemote(TableFunctionFactory & factory)
 {
     factory.registerFunction("remote", [] () -> TableFunctionPtr { return std::make_shared<TableFunctionRemote>("remote"); });
     factory.registerFunction("remoteSecure", [] () -> TableFunctionPtr { return std::make_shared<TableFunctionRemote>("remote", /* secure = */ true); });
-    factory.registerFunction("cluster", [] () -> TableFunctionPtr { return std::make_shared<TableFunctionRemote>("cluster"); });
-    factory.registerFunction("clusterAllReplicas", [] () -> TableFunctionPtr { return std::make_shared<TableFunctionRemote>("clusterAllReplicas"); });
+    factory.registerFunction("cluster", {[] () -> TableFunctionPtr { return std::make_shared<TableFunctionRemote>("cluster"); }, {.documentation = {}, .allow_readonly = true}});
+    factory.registerFunction("clusterAllReplicas", {[] () -> TableFunctionPtr { return std::make_shared<TableFunctionRemote>("clusterAllReplicas"); }, {.documentation = {}, .allow_readonly = true}});
 }
 
 }
diff --git a/src/TableFunctions/TableFunctionValues.cpp b/src/TableFunctions/TableFunctionValues.cpp
index 7b8de69a4f89..055748252755 100644
--- a/src/TableFunctions/TableFunctionValues.cpp
+++ b/src/TableFunctions/TableFunctionValues.cpp
@@ -147,7 +147,7 @@ StoragePtr TableFunctionValues::executeImpl(const ASTPtr & ast_function, Context
 
 void registerTableFunctionValues(TableFunctionFactory & factory)
 {
-    factory.registerFunction<TableFunctionValues>({}, TableFunctionFactory::CaseInsensitive);
+    factory.registerFunction<TableFunctionValues>({.documentation = {}, .allow_readonly = true}, TableFunctionFactory::CaseInsensitive);
 }
 
 }
diff --git a/src/TableFunctions/TableFunctionView.cpp b/src/TableFunctions/TableFunctionView.cpp
index c53d26a794b8..c999cba08e9e 100644
--- a/src/TableFunctions/TableFunctionView.cpp
+++ b/src/TableFunctions/TableFunctionView.cpp
@@ -55,7 +55,7 @@ StoragePtr TableFunctionView::executeImpl(
 
 void registerTableFunctionView(TableFunctionFactory & factory)
 {
-    factory.registerFunction<TableFunctionView>();
+    factory.registerFunction<TableFunctionView>({.documentation = {}, .allow_readonly = true});
 }
 
 }
diff --git a/src/TableFunctions/TableFunctionViewIfPermitted.cpp b/src/TableFunctions/TableFunctionViewIfPermitted.cpp
index ba3d2cb9d160..6128fe0a36f8 100644
--- a/src/TableFunctions/TableFunctionViewIfPermitted.cpp
+++ b/src/TableFunctions/TableFunctionViewIfPermitted.cpp
@@ -107,7 +107,7 @@ bool TableFunctionViewIfPermitted::isPermitted(const ContextPtr & context, const
 
 void registerTableFunctionViewIfPermitted(TableFunctionFactory & factory)
 {
-    factory.registerFunction<TableFunctionViewIfPermitted>();
+    factory.registerFunction<TableFunctionViewIfPermitted>({.documentation = {}, .allow_readonly = true});
 }
 
 }
diff --git a/src/TableFunctions/TableFunctionZeros.cpp b/src/TableFunctions/TableFunctionZeros.cpp
index 126166b547b5..5874fca67e64 100644
--- a/src/TableFunctions/TableFunctionZeros.cpp
+++ b/src/TableFunctions/TableFunctionZeros.cpp
@@ -48,7 +48,7 @@ StoragePtr TableFunctionZeros<multithreaded>::executeImpl(const ASTPtr & ast_fun
 
 void registerTableFunctionZeros(TableFunctionFactory & factory)
 {
-    factory.registerFunction<TableFunctionZeros<true>>({R"(
+    factory.registerFunction<TableFunctionZeros<true>>({.documentation = {R"(
 Generates a stream of zeros (a table with one column 'zero' of type 'UInt8') of specified size.
 
 This table function is used in performance tests, where you want to spend as little time as possible to data generation while testing some other parts of queries.
@@ -62,9 +62,9 @@ This query will test the speed of `randomPrintableASCII` function using single t
 See also the `system.zeros` table.
 )",
 {{"1", "SELECT count() FROM zeros(100000000) WHERE NOT ignore(randomPrintableASCII(10))"}}
-});
+}});
 
-    factory.registerFunction<TableFunctionZeros<false>>({R"(
+    factory.registerFunction<TableFunctionZeros<false>>({.documentation = {R"(
 Generates a stream of zeros (a table with one column 'zero' of type 'UInt8') of specified size.
 
 This table function is used in performance tests, where you want to spend as little time as possible to data generation while testing some other parts of queries.
@@ -78,7 +78,7 @@ This query will test the speed of `randomPrintableASCII` function using multiple
 See also the `system.zeros` table.
 )",
 {{"1", "SELECT count() FROM zeros_mt(1000000000) WHERE NOT ignore(randomPrintableASCII(10))"}}
-});
+}});
 }
 
 template <bool multithreaded>
