diff --git a/tests/queries/0_stateless/01355_defaultValueOfArgumentType_bug.reference b/tests/queries/0_stateless/01355_defaultValueOfArgumentType_bug.reference
new file mode 100644
index 000000000000..4165503c4b59
--- /dev/null
+++ b/tests/queries/0_stateless/01355_defaultValueOfArgumentType_bug.reference
@@ -0,0 +1,1 @@
+	LowCardinality(String)
diff --git a/tests/queries/0_stateless/01355_defaultValueOfArgumentType_bug.sql b/tests/queries/0_stateless/01355_defaultValueOfArgumentType_bug.sql
new file mode 100644
index 000000000000..e3168eb09a06
--- /dev/null
+++ b/tests/queries/0_stateless/01355_defaultValueOfArgumentType_bug.sql
@@ -0,0 +1,4 @@
+SELECT 
+    materialize(toLowCardinality('')) AS lc,
+    toTypeName(lc)
+WHERE lc = defaultValueOfArgumentType(lc)
