diff --git a/tests/integration/test_s3_style_link/__init__.py b/tests/integration/test_s3_style_link/__init__.py
new file mode 100644
index 000000000000..e69de29bb2d1
diff --git a/tests/integration/test_s3_style_link/configs/config.d/minio.xml b/tests/integration/test_s3_style_link/configs/config.d/minio.xml
new file mode 100644
index 000000000000..072d07b7f739
--- /dev/null
+++ b/tests/integration/test_s3_style_link/configs/config.d/minio.xml
@@ -0,0 +1,10 @@
+<?xml version="1.0"?>
+
+<!-- Using named collections 22.4+ -->
+<clickhouse>
+    <url_scheme_mappers>
+        <minio>
+            <to>http://minio1:9001/root/{bucket}</to>
+        </minio>
+    </url_scheme_mappers>
+</clickhouse>
diff --git a/tests/integration/test_s3_style_link/configs/users.d/users.xml b/tests/integration/test_s3_style_link/configs/users.d/users.xml
new file mode 100644
index 000000000000..4b6ba057ecb1
--- /dev/null
+++ b/tests/integration/test_s3_style_link/configs/users.d/users.xml
@@ -0,0 +1,9 @@
+<clickhouse>
+    <users>
+        <default>
+            <password></password>
+            <profile>default</profile>
+            <named_collection_control>1</named_collection_control>
+        </default>
+    </users>
+</clickhouse>
diff --git a/tests/integration/test_s3_style_link/test.py b/tests/integration/test_s3_style_link/test.py
new file mode 100644
index 000000000000..7ecf4e633e31
--- /dev/null
+++ b/tests/integration/test_s3_style_link/test.py
@@ -0,0 +1,57 @@
+import logging
+import pytest
+from helpers.cluster import ClickHouseCluster
+
+
+cluster = ClickHouseCluster(__file__)
+node = cluster.add_instance(
+    "node",
+    main_configs=[
+        "configs/config.d/minio.xml",
+    ],
+    user_configs=[
+        "configs/users.d/users.xml",
+    ],
+    with_minio=True,
+)
+
+
+@pytest.fixture(scope="module")
+def started_cluster():
+    try:
+        logging.info("Starting cluster...")
+        cluster.start()
+        logging.info("Cluster started")
+
+        yield cluster
+    finally:
+        logging.info("Stopping cluster")
+        cluster.shutdown()
+        logging.info("Cluster stopped")
+
+
+def test_s3_table_functions(started_cluster):
+    """
+    Simple test to check s3 table function functionalities
+    """
+    node.query(
+        """
+            INSERT INTO FUNCTION s3
+                (
+                    'minio://data/test_file.tsv.gz', 'minio', 'minio123'
+                )
+            SELECT * FROM numbers(1000000);
+        """
+    )
+
+    assert (
+        node.query(
+            """
+            SELECT count(*) FROM s3
+            (
+                'minio://data/test_file.tsv.gz', 'minio', 'minio123'
+            );
+        """
+        )
+        == "1000000
"
+    )
