diff --git a/tests/queries/0_stateless/02843_insertion_table_schema_infer.reference b/tests/queries/0_stateless/02843_insertion_table_schema_infer.reference
new file mode 100644
index 000000000000..aec86406a982
--- /dev/null
+++ b/tests/queries/0_stateless/02843_insertion_table_schema_infer.reference
@@ -0,0 +1,3 @@
+user127	1
+user405	1
+user902	1
diff --git a/tests/queries/0_stateless/02843_insertion_table_schema_infer.sh b/tests/queries/0_stateless/02843_insertion_table_schema_infer.sh
new file mode 100755
index 000000000000..d806b6784563
--- /dev/null
+++ b/tests/queries/0_stateless/02843_insertion_table_schema_infer.sh
@@ -0,0 +1,16 @@
+#!/usr/bin/env bash
+# Tags: no-parallel, no-fasttest
+
+set -e
+
+CUR_DIR=$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)
+# shellcheck source=../shell_config.sh
+. "$CUR_DIR"/../shell_config.sh
+
+DATA_DIR=$CUR_DIR/data_tsv
+
+$CLICKHOUSE_LOCAL --multiquery \
+"CREATE VIEW users AS SELECT * FROM file('$DATA_DIR/mock_data.tsv', TSVWithNamesAndTypes); 
+ CREATE TABLE users_output (name String, tag UInt64)ENGINE = Memory;
+ INSERT INTO users_output WITH (SELECT groupUniqArrayArray(mapKeys(Tags)) FROM users) AS unique_tags SELECT UserName AS name, length(unique_tags) AS tag FROM users;
+ SELECT * FROM users_output;"
diff --git a/tests/queries/0_stateless/data_tsv/mock_data.tsv b/tests/queries/0_stateless/data_tsv/mock_data.tsv
new file mode 100644
index 000000000000..fcf2b300b15b
--- /dev/null
+++ b/tests/queries/0_stateless/data_tsv/mock_data.tsv
@@ -0,0 +1,5 @@
+UserName	Age	Tags
+String	Int8	Map(String, UInt64)
+user127	20	{'test': 123}
+user405	43	{'test': 123}
+user902	43	{'test': 123}
