diff --git a/tests/queries/0_stateless/02746_index_analysis_binary_operator_with_null.reference b/tests/queries/0_stateless/02746_index_analysis_binary_operator_with_null.reference
new file mode 100644
index 000000000000..e69de29bb2d1
diff --git a/tests/queries/0_stateless/02746_index_analysis_binary_operator_with_null.sql b/tests/queries/0_stateless/02746_index_analysis_binary_operator_with_null.sql
new file mode 100644
index 000000000000..f9613735bbf8
--- /dev/null
+++ b/tests/queries/0_stateless/02746_index_analysis_binary_operator_with_null.sql
@@ -0,0 +1,12 @@
+drop table if exists tab;
+
+create table tab (x DateTime) engine MergeTree order by x;
+
+SELECT toDateTime(65537, toDateTime(NULL), NULL)
+FROM tab
+WHERE ((x + CAST('1', 'Nullable(UInt8)')) <= 2) AND ((x + CAST('', 'Nullable(UInt8)')) <= 256)
+ORDER BY
+    toDateTime(toDateTime(-2, NULL, NULL) + 100.0001, NULL, -2, NULL) DESC NULLS LAST,
+    x ASC NULLS LAST;
+
+drop table tab;
