diff --git a/src/Interpreters/RewriteSumIfFunctionVisitor.cpp b/src/Interpreters/RewriteSumIfFunctionVisitor.cpp
index 7f725c1d8a5a..50e6bec28f38 100644
--- a/src/Interpreters/RewriteSumIfFunctionVisitor.cpp
+++ b/src/Interpreters/RewriteSumIfFunctionVisitor.cpp
@@ -25,7 +25,8 @@ void RewriteSumIfFunctionMatcher::visit(const ASTFunction & func, ASTPtr & ast,
 
     auto lower_name = Poco::toLower(func.name);
 
-    if (lower_name != "sum" && lower_name != "sumif")
+    /// sumIf, SumIf or sUMIf are valid function names, but sumIF or sumiF are not
+    if (lower_name != "sum" && (lower_name != "sumif" || !endsWith(func.name, "If")))
         return;
 
     const auto & func_arguments = func.arguments->children;
