diff --git a/tests/queries/0_stateless/02177_sum_if_not_found.reference b/tests/queries/0_stateless/02177_sum_if_not_found.reference
new file mode 100644
index 000000000000..bb0b1cf658d1
--- /dev/null
+++ b/tests/queries/0_stateless/02177_sum_if_not_found.reference
@@ -0,0 +1,3 @@
+0
+0
+0
diff --git a/tests/queries/0_stateless/02177_sum_if_not_found.sql b/tests/queries/0_stateless/02177_sum_if_not_found.sql
new file mode 100644
index 000000000000..c888f8b39aa7
--- /dev/null
+++ b/tests/queries/0_stateless/02177_sum_if_not_found.sql
@@ -0,0 +1,35 @@
+SELECT sumIf(1, 0);
+SELECT SumIf(1, 0);
+SELECT sUmIf(1, 0);
+SELECT sumIF(1, 0); -- { serverError 46 }
+
+DROP TABLE IF EXISTS data;
+DROP TABLE IF EXISTS agg;
+
+CREATE TABLE data
+(
+    `n` UInt32,
+    `t` DateTime
+)
+ENGINE = Null;
+
+CREATE TABLE agg
+ENGINE = AggregatingMergeTree
+ORDER BY tuple() AS
+SELECT
+    t,
+    sumIF(n, 0)
+FROM data
+GROUP BY t; -- { serverError 46}
+
+CREATE TABLE agg
+ENGINE = AggregatingMergeTree
+ORDER BY tuple() AS
+SELECT
+    t,
+    sumIf(n, 0)
+FROM data
+GROUP BY t;
+
+DROP TABLE data;
+DROP TABLE agg;
