diff --git a/src/server/json_family_test.cc b/src/server/json_family_test.cc
index 6cf5606950f4..c0560debc382 100644
--- a/src/server/json_family_test.cc
+++ b/src/server/json_family_test.cc
@@ -581,12 +581,21 @@ TEST_F(JsonFamilyTest, ObjKeys) {
   auto resp = Run({"JSON.SET", "json", ".", json});
   ASSERT_THAT(resp, "OK");
 
+  resp = Run({"JSON.OBJKEYS", "json"});
+  EXPECT_THAT(resp.GetVec(), ElementsAre("a", "b", "c", "d", "e"));
+
   resp = Run({"JSON.OBJKEYS", "json", "$.a"});
   EXPECT_THAT(resp, ArgType(RespExpr::NIL_ARRAY));
 
   resp = Run({"JSON.OBJKEYS", "json", "$.b"});
   EXPECT_THAT(resp.GetVec(), ElementsAre("a"));
 
+  resp = Run({"JSON.OBJKEYS", "json", ".b"});
+  EXPECT_THAT(resp.GetVec(), ElementsAre("a"));
+
+  resp = Run({"JSON.OBJKEYS", "json", "b"});
+  EXPECT_THAT(resp.GetVec(), ElementsAre("a"));
+
   resp = Run({"JSON.OBJKEYS", "json", "$.*"});
   ASSERT_THAT(resp, ArrLen(5));
   const auto& arr = resp.GetVec();
